

Signed-off-by: Paolo 'Blaisorblade' Giarrusso <blaisorblade_spam@yahoo.it>
---

 linux-2.6.9-current-paolo/arch/um/Makefile |    2 --
 1 files changed, 2 deletions(-)

diff -puN arch/um/Makefile~uml-no-extraversion-in-mainline-finish arch/um/Makefile
--- linux-2.6.9-current/arch/um/Makefile~uml-no-extraversion-in-mainline-finish	2004-09-19 18:27:36.410503760 +0200
+++ linux-2.6.9-current-paolo/arch/um/Makefile	2004-09-19 18:27:36.412503456 +0200
@@ -66,8 +66,6 @@ ifeq ($(CONFIG_MODE_SKAS), y)
 $(SYS_HEADERS) : $(TOPDIR)/$(ARCH_DIR)/include/skas_ptregs.h
 endif
 
-include/linux/version.h: arch/$(ARCH)/Makefile
-
 $(ARCH_DIR)/vmlinux.lds.S :
 	touch $@
 
_
