Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 drivers/scsi/aacraid/aacraid.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.25.4-rt4/drivers/scsi/aacraid/aacraid.h
===================================================================
--- linux-2.6.25.4-rt4.orig/drivers/scsi/aacraid/aacraid.h	2008-05-29 09:46:29.000000000 -0400
+++ linux-2.6.25.4-rt4/drivers/scsi/aacraid/aacraid.h	2008-05-29 09:47:20.000000000 -0400
@@ -787,7 +787,7 @@ struct fib {
 	 *	This is the event the sendfib routine will wait on if the
 	 *	caller did not pass one and this is synch io.
 	 */
-	struct semaphore	event_wait;
+	struct compat_semaphore	event_wait;
 	spinlock_t		event_lock;
 
 	u32			done;	/* gets set to 1 when fib is complete */
