---
 include/linux/linkage.h |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.22/include/linux/linkage.h
===================================================================
--- linux-2.6.22.orig/include/linux/linkage.h	2007-07-24 08:56:37.000000000 +0200
+++ linux-2.6.22/include/linux/linkage.h	2007-07-24 08:57:04.000000000 +0200
@@ -3,6 +3,8 @@
 
 #include <asm/linkage.h>
 
+#define notrace __attribute ((no_instrument_function))
+
 #ifdef __cplusplus
 #define CPP_ASMLINKAGE extern "C"
 #else
