Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86/kernel/io_apic_64.c |    9 +++++++++
 1 file changed, 9 insertions(+)

Index: linux-2.6.25.4-rt1/arch/x86/kernel/io_apic_64.c
===================================================================
--- linux-2.6.25.4-rt1.orig/arch/x86/kernel/io_apic_64.c	2008-05-17 08:27:42.000000000 -0400
+++ linux-2.6.25.4-rt1/arch/x86/kernel/io_apic_64.c	2008-05-17 08:28:10.000000000 -0400
@@ -1511,6 +1511,15 @@ static void ack_apic_level(unsigned int 
 			move_masked_irq(irq);
 		unmask_IO_APIC_irq(irq);
 	}
+#if (defined(CONFIG_GENERIC_PENDING_IRQ) || defined(CONFIG_IRQBALANCE)) && \
+	defined(CONFIG_PREEMPT_HARDIRQS)
+	/*
+	 * With threaded interrupts, we always have IRQ_INPROGRESS
+	 * when acking.
+	 */
+	else if (unlikely(irq_desc[irq].status & IRQ_MOVE_PENDING))
+		move_masked_irq(irq);
+#endif
 }
 
 static struct irq_chip ioapic_chip __read_mostly = {
