Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 arch/x86/kernel/apic_32.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

--- linux-2.6.27-RT.orig/arch/x86/kernel/apic_32.c
+++ linux-2.6.27-RT/arch/x86/kernel/apic_32.c
@@ -1346,10 +1346,9 @@ void smp_error_interrupt(struct pt_regs
 	   6: Received illegal vector
 	   7: Illegal register address
 	*/
-#if 0
 	printk(KERN_DEBUG "APIC error on CPU%d: %02lx(%02lx)\n",
 		smp_processor_id(), v , v1);
-#endif
+	dump_stack();
 	irq_exit();
 }
 
