Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
 arch/x86/kernel/apic_32.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.25-RT/arch/x86/kernel/apic_32.c
===================================================================
--- linux-2.6.25-RT.orig/arch/x86/kernel/apic_32.c
+++ linux-2.6.25-RT/arch/x86/kernel/apic_32.c
@@ -1285,9 +1285,10 @@ void smp_error_interrupt(struct pt_regs 
 	   6: Received illegal vector
 	   7: Illegal register address
 	*/
-#if 0
+#ifndef CONFIG_PREEMPT_RT
 	printk(KERN_DEBUG "APIC error on CPU%d: %02lx(%02lx)\n",
 		smp_processor_id(), v , v1);
+	dump_stack();
 #endif
 	irq_exit();
 }
