 arch/i386/kernel/apic.c |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.22/arch/i386/kernel/apic.c
===================================================================
--- linux-2.6.22.orig/arch/i386/kernel/apic.c	2007-07-24 08:56:53.000000000 +0200
+++ linux-2.6.22/arch/i386/kernel/apic.c	2007-07-24 08:57:04.000000000 +0200
@@ -1309,6 +1309,7 @@ void smp_error_interrupt(struct pt_regs 
 	*/
 	printk (KERN_DEBUG "APIC error on CPU%d: %02lx(%02lx)\n",
 		smp_processor_id(), v , v1);
+	dump_stack();
 	irq_exit();
 }
 
