Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
 arch/arm/Kconfig |   13 +------------
 1 file changed, 1 insertion(+), 12 deletions(-)

Index: linux-2.6.25.4-rt2/arch/arm/Kconfig
===================================================================
--- linux-2.6.25.4-rt2.orig/arch/arm/Kconfig	2008-05-19 16:55:29.000000000 -0400
+++ linux-2.6.25.4-rt2/arch/arm/Kconfig	2008-05-19 16:55:32.000000000 -0400
@@ -695,18 +695,7 @@ config LOCAL_TIMERS
 	  accounting to be spread across the timer interval, preventing a
 	  "thundering herd" at every timer tick.
 
-config PREEMPT
-	bool "Preemptible Kernel (EXPERIMENTAL)"
-	depends on EXPERIMENTAL
-	help
-	  This option reduces the latency of the kernel when reacting to
-	  real-time or interactive events by allowing a low priority process to
-	  be preempted even if it is in kernel mode executing a system call.
-	  This allows applications to run more reliably even when the system is
-	  under load.
-
-	  Say Y here if you are building a kernel for a desktop, embedded
-	  or real-time system.  Say N if you are unsure.
+source kernel/Kconfig.preempt
 
 config NO_IDLE_HZ
 	bool "Dynamic tick timer"
