---
 include/asm-arm/atomic.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.22/include/asm-arm/atomic.h
===================================================================
--- linux-2.6.22.orig/include/asm-arm/atomic.h	2007-07-24 08:57:13.000000000 +0200
+++ linux-2.6.22/include/asm-arm/atomic.h	2007-07-24 08:57:14.000000000 +0200
@@ -189,10 +189,10 @@ static inline unsigned long __cmpxchg(vo
 	volatile unsigned long *p = ptr;
 
 	if (size == 4) {
-		local_irq_save(flags);
+		raw_local_irq_save(flags);
 		if ((prev = *p) == old)
 			*p = new;
-		local_irq_restore(flags);
+		raw_local_irq_restore(flags);
 		return(prev);
 	} else
 		return wrong_size_cmpxchg(ptr);
