Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/asm-arm/atomic.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.25.4-rt2/include/asm-arm/atomic.h
===================================================================
--- linux-2.6.25.4-rt2.orig/include/asm-arm/atomic.h	2008-05-19 16:55:31.000000000 -0400
+++ linux-2.6.25.4-rt2/include/asm-arm/atomic.h	2008-05-19 16:55:32.000000000 -0400
@@ -189,10 +189,10 @@ static inline unsigned long __cmpxchg(vo
 	volatile unsigned long *p = ptr;
 
 	if (size == 4) {
-		local_irq_save(flags);
+		raw_local_irq_save(flags);
 		if ((prev = *p) == old)
 			*p = new;
-		local_irq_restore(flags);
+		raw_local_irq_restore(flags);
 		return(prev);
 	} else
 		return wrong_size_cmpxchg(ptr);
