Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: clockevents: Remove unused inline function

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

---
 include/linux/clockchips.h |    1 -
 1 file changed, 1 deletion(-)

Index: linux-2.6.23-rc4/include/linux/clockchips.h
===================================================================
--- linux-2.6.23-rc4.orig/include/linux/clockchips.h	2007-08-31 21:51:20.000000000 +0200
+++ linux-2.6.23-rc4/include/linux/clockchips.h	2007-08-31 21:51:21.000000000 +0200
@@ -130,7 +130,6 @@ extern void clockevents_notify(unsigned 
 
 #else
 
-static inline void clockevents_resume_events(void) { }
 #define clockevents_notify(reason, arg) do { } while (0)
 
 #endif
