Subject: clockevents: Remove unused inline function

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

---
 include/linux/clockchips.h |    1 -
 1 file changed, 1 deletion(-)

Index: linux-2.6.22/include/linux/clockchips.h
===================================================================
--- linux-2.6.22.orig/include/linux/clockchips.h	2007-07-24 08:56:52.000000000 +0200
+++ linux-2.6.22/include/linux/clockchips.h	2007-07-24 08:56:54.000000000 +0200
@@ -131,7 +131,6 @@ extern void clockevents_notify(unsigned 
 
 #else
 
-static inline void clockevents_resume_events(void) { }
 #define clockevents_notify(reason, arg) do { } while (0)
 
 #endif
