Subject: edac-2.6.23-to-2.6.22.patch back-port
From: http://superb-east.dl.sourceforge.net/sourceforge/bluesmoke/edac-2007-oct-1.tgz

Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>

Index: linux-2.6.23-rc2/drivers/edac/edac_mc.c
===================================================================
--- linux-2.6.23-rc2.orig/drivers/edac/edac_mc.c
+++ linux-2.6.23-rc2/drivers/edac/edac_mc.c
@@ -684,6 +684,9 @@ void edac_mc_handle_ce(struct mem_ctl_in
 	mci->csrows[row].ce_count++;
 	mci->csrows[row].channels[channel].ce_count++;
 
+	/* perform poll notify operation */
+	edac_mc_sysfs_notify_ce_count(mci);
+
 	if (mci->scrub_mode & SCRUB_SW_SRC) {
 		/*
 		 * Some MC's can remap memory so that it is still available
@@ -712,6 +715,9 @@ void edac_mc_handle_ce_no_info(struct me
 
 	mci->ce_noinfo_count++;
 	mci->ce_count++;
+
+	/* perform poll notify operation */
+	edac_mc_sysfs_notify_ce_noinfo_count(mci);
 }
 EXPORT_SYMBOL_GPL(edac_mc_handle_ce_no_info);
 
@@ -765,6 +771,9 @@ void edac_mc_handle_ue(struct mem_ctl_in
 
 	mci->ue_count++;
 	mci->csrows[row].ue_count++;
+
+	/* perform poll notify operation */
+	edac_mc_sysfs_notify_ue_count(mci);
 }
 EXPORT_SYMBOL_GPL(edac_mc_handle_ue);
 
@@ -778,6 +787,9 @@ void edac_mc_handle_ue_no_info(struct me
 			"UE - no information available: %s\n", msg);
 	mci->ue_noinfo_count++;
 	mci->ue_count++;
+
+	/* perform poll notify operation */
+	edac_mc_sysfs_notify_ue_noinfo_count(mci);
 }
 EXPORT_SYMBOL_GPL(edac_mc_handle_ue_no_info);
 
Index: linux-2.6.23-rc2/drivers/edac/edac_mc_sysfs.c
===================================================================
--- linux-2.6.23-rc2.orig/drivers/edac/edac_mc_sysfs.c
+++ linux-2.6.23-rc2/drivers/edac/edac_mc_sysfs.c
@@ -622,6 +622,33 @@ static struct mcidev_sysfs_attribute *mc
 	NULL
 };
 
+/*
+ * function for the edac_mc code to call, which in turn will
+ * perform a sysfs_notify call for the respective counter
+ */
+void edac_mc_sysfs_notify_ce_count(struct mem_ctl_info *mci)
+{
+	sysfs_notify(&mci->edac_mci_kobj, NULL,
+			(char *) mci_attr_ce_count.attr.name);
+}
+
+void edac_mc_sysfs_notify_ue_count(struct mem_ctl_info *mci)
+{
+	sysfs_notify(&mci->edac_mci_kobj, NULL,
+			(char *) mci_attr_ue_count.attr.name);
+}
+
+void edac_mc_sysfs_notify_ce_noinfo_count(struct mem_ctl_info *mci)
+{
+	sysfs_notify(&mci->edac_mci_kobj, NULL,
+			(char *) mci_attr_ce_noinfo_count.attr.name);
+}
+
+void edac_mc_sysfs_notify_ue_noinfo_count(struct mem_ctl_info *mci)
+{
+	sysfs_notify(&mci->edac_mci_kobj, NULL,
+			(char *) mci_attr_ue_noinfo_count.attr.name);
+}
 
 /*
  * Release of a MC controlling instance
Index: linux-2.6.23-rc2/drivers/edac/edac_module.h
===================================================================
--- linux-2.6.23-rc2.orig/drivers/edac/edac_module.h
+++ linux-2.6.23-rc2/drivers/edac/edac_module.h
@@ -27,6 +27,10 @@ extern int edac_mc_register_sysfs_main_k
 extern void edac_mc_unregister_sysfs_main_kobj(struct mem_ctl_info *mci);
 extern int edac_create_sysfs_mci_device(struct mem_ctl_info *mci);
 extern void edac_remove_sysfs_mci_device(struct mem_ctl_info *mci);
+extern void edac_mc_sysfs_notify_ce_count(struct mem_ctl_info *mci);
+extern void edac_mc_sysfs_notify_ue_count(struct mem_ctl_info *mci);
+extern void edac_mc_sysfs_notify_ce_noinfo_count(struct mem_ctl_info *mci);
+extern void edac_mc_sysfs_notify_ue_noinfo_count(struct mem_ctl_info *mci);
 extern void edac_check_mc_devices(void);
 extern int edac_get_log_ue(void);
 extern int edac_get_log_ce(void);
@@ -43,6 +47,10 @@ extern void edac_device_unregister_sysfs
 				struct edac_device_ctl_info *edac_dev);
 extern int edac_device_create_sysfs(struct edac_device_ctl_info *edac_dev);
 extern void edac_device_remove_sysfs(struct edac_device_ctl_info *edac_dev);
+extern void edac_device_sysfs_notify_ce_count(
+				struct edac_device_ctl_info *edac_dev);
+extern void edac_device_sysfs_notify_ue_count(
+				struct edac_device_ctl_info *edac_dev);
 extern struct sysdev_class *edac_get_edac_class(void);
 
 /* edac core workqueue: single CPU mode */
Index: linux-2.6.23-rc2/drivers/edac/edac_device.c
===================================================================
--- linux-2.6.23-rc2.orig/drivers/edac/edac_device.c
+++ linux-2.6.23-rc2/drivers/edac/edac_device.c
@@ -686,6 +686,9 @@ void edac_device_handle_ce(struct edac_d
 	instance->counters.ce_count++;
 	edac_dev->counters.ce_count++;
 
+	/* perform sysfs notify */
+	edac_device_sysfs_notify_ce_count(edac_dev);
+
 	if (edac_device_get_log_ce(edac_dev))
 		edac_device_printk(edac_dev, KERN_WARNING,
 				"CE: %s instance: %s block: %s '%s'\n",
@@ -732,6 +735,9 @@ void edac_device_handle_ue(struct edac_d
 	instance->counters.ue_count++;
 	edac_dev->counters.ue_count++;
 
+	/* perform sysfs notify */
+	edac_device_sysfs_notify_ue_count(edac_dev);
+
 	if (edac_device_get_log_ue(edac_dev))
 		edac_device_printk(edac_dev, KERN_EMERG,
 				"UE: %s instance: %s block: %s '%s'\n",
Index: linux-2.6.23-rc2/drivers/edac/edac_device_sysfs.c
===================================================================
--- linux-2.6.23-rc2.orig/drivers/edac/edac_device_sysfs.c
+++ linux-2.6.23-rc2/drivers/edac/edac_device_sysfs.c
@@ -407,6 +407,20 @@ static struct instance_attribute *device
 	NULL,
 };
 
+/* sysfs notify call for the ce_count and ue_count */
+void edac_device_sysfs_notify_ce_count(struct edac_device_ctl_info *edac_dev)
+{
+        sysfs_notify(&edac_dev->kobj, NULL,
+                        (char *) attr_instance_ce_count.attr.name);
+}
+
+void edac_device_sysfs_notify_ue_count(struct edac_device_ctl_info *edac_dev)
+{
+        sysfs_notify(&edac_dev->kobj, NULL,
+                        (char *) attr_instance_ue_count.attr.name);
+}
+
+
 /* The 'ktype' for each edac_dev 'instance' */
 static struct kobj_type ktype_instance_ctrl = {
 	.release = edac_device_ctrl_instance_release,
