Subject: Linux-RT 2.6.24-rc8-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/rcupreempt.c |   11 +++++++++++
 1 file changed, 11 insertions(+)

Index: linux-2.6.23-rc8-rt1/kernel/rcupreempt.c
===================================================================
--- linux-2.6.23-rc8-rt1.orig/kernel/rcupreempt.c	2007-09-26 14:27:23.000000000 -0400
+++ linux-2.6.23-rc8-rt1/kernel/rcupreempt.c	2007-09-26 14:27:23.000000000 -0400
@@ -338,6 +338,17 @@ void __synchronize_sched(void)
 	sched_setaffinity(0, oldmask);
 }
 
+/*
+ * Check to see if any future RCU-related work will need to be done
+ * by the current CPU, even if none need be done immediately, returning
+ * 1 if so.  This function is part of the RCU implementation; it is -not-
+ * an exported member of the RCU API.
+ */
+int rcu_needs_cpu(int cpu)
+{
+	return !!rcu_data.waitlist || rcu_pending(cpu);
+}
+
 int rcu_pending(int cpu)
 {
 	return (rcu_data.donelist != NULL ||
