---
 kernel/rcupreempt.c |   11 +++++++++++
 1 file changed, 11 insertions(+)

Index: linux-2.6.22/kernel/rcupreempt.c
===================================================================
--- linux-2.6.22.orig/kernel/rcupreempt.c	2007-07-24 08:57:13.000000000 +0200
+++ linux-2.6.22/kernel/rcupreempt.c	2007-07-24 08:57:13.000000000 +0200
@@ -338,6 +338,17 @@ void __synchronize_sched(void)
 	sched_setaffinity(0, oldmask);
 }
 
+/*
+ * Check to see if any future RCU-related work will need to be done
+ * by the current CPU, even if none need be done immediately, returning
+ * 1 if so.  This function is part of the RCU implementation; it is -not-
+ * an exported member of the RCU API.
+ */
+int rcu_needs_cpu(int cpu)
+{
+	return !!rcu_data.waitlist || rcu_pending(cpu);
+}
+
 int rcu_pending(int cpu)
 {
 	return (rcu_data.donelist != NULL ||
