Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>


---
 kernel/workqueue.c |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.26.3-rt6/kernel/workqueue.c
===================================================================
--- linux-2.6.26.3-rt6.orig/kernel/workqueue.c	2008-09-05 00:23:33.000000000 -0400
+++ linux-2.6.26.3-rt6/kernel/workqueue.c	2008-09-05 00:23:34.000000000 -0400
@@ -666,6 +666,7 @@ out:
 
 	return err;
 }
+EXPORT_SYMBOL(schedule_on_each_cpu);
 
 /**
  * schedule_on_each_cpu_wq - call a function on each online CPU on a per-CPU wq
