Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 mm/filemap.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.26.3-rt6/mm/filemap.c
===================================================================
--- linux-2.6.26.3-rt6.orig/mm/filemap.c	2008-09-05 00:23:22.000000000 -0400
+++ linux-2.6.26.3-rt6/mm/filemap.c	2008-09-05 00:23:39.000000000 -0400
@@ -1803,7 +1803,9 @@ size_t iov_iter_copy_from_user_atomic(st
 	char *kaddr;
 	size_t copied;
 
+#ifndef CONFIG_PREEMPT_RT
 	BUG_ON(!in_atomic());
+#endif
 	kaddr = kmap_atomic(page, KM_USER0);
 	if (likely(i->nr_segs == 1)) {
 		int left;
