Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 mm/filemap.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.24-rt1/mm/filemap.c
===================================================================
--- linux-2.6.24-rt1.orig/mm/filemap.c	2008-01-25 15:08:08.000000000 -0500
+++ linux-2.6.24-rt1/mm/filemap.c	2008-01-25 15:08:29.000000000 -0500
@@ -1763,7 +1763,9 @@ size_t iov_iter_copy_from_user_atomic(st
 	char *kaddr;
 	size_t copied;
 
+#ifndef CONFIG_PREEMPT_RT
 	BUG_ON(!in_atomic());
+#endif
 	kaddr = kmap_atomic(page, KM_USER0);
 	if (likely(i->nr_segs == 1)) {
 		int left;
