Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
 arch/x86/pci/Makefile_32 |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.26/arch/x86/pci/Makefile_32
===================================================================
--- linux-2.6.26.orig/arch/x86/pci/Makefile_32
+++ linux-2.6.26/arch/x86/pci/Makefile_32
@@ -5,12 +5,13 @@ obj-$(CONFIG_PCI_MMCONFIG)	+= mmconfig_3
 obj-$(CONFIG_PCI_DIRECT)	+= direct.o
 obj-$(CONFIG_PCI_OLPC)		+= olpc.o
 
+obj-$(CONFIG_ACPI)		+= acpi.o
+
 pci-y				:= fixup.o
 
 # Do not change the ordering here. There is a nasty init function
 # ordering dependency which breaks when you move acpi.o below
 # legacy/irq.o
-pci-$(CONFIG_ACPI)		+= acpi.o
 pci-y				+= legacy.o irq.o
 
 # Careful: VISWS and NUMAQ overrule the pci-y above. The colons are
