Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From: Steven Rostedt <srostedt@redhat.com>
Subject: ftrace: dont trace markers

It is redundant to trace the marker code.

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
diff --git a/kernel/Makefile b/kernel/Makefile
index 6257af7..57dd8d2 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -22,6 +22,8 @@ CFLAGS_REMOVE_rtmutex-debug.o = -pg
 CFLAGS_REMOVE_cgroup-debug.o = -pg
 CFLAGS_REMOVE_sched_clock.o = -pg
 CFLAGS_REMOVE_sched.o = -mno-spe -pg
+CFLAGS_REMOVE_marker.o = -pg
+CFLAGS_REMOVE_tracepoint.o = -pg
 endif
 
 obj-$(CONFIG_PROFILING) += profile.o
