Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/asm-x86/unistd_32.h |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6-rt-test/include/asm-x86/unistd_32.h
===================================================================
--- linux-2.6-rt-test.orig/include/asm-x86/unistd_32.h	2008-05-07 11:34:13.000000000 -0400
+++ linux-2.6-rt-test/include/asm-x86/unistd_32.h	2008-05-07 12:01:55.000000000 -0400
@@ -333,6 +333,8 @@
 #define __NR_timerfd_settime	325
 #define __NR_timerfd_gettime	326
 
+#define NR_syscalls 327
+
 #ifdef __KERNEL__
 
 #define __ARCH_WANT_IPC_PARSE_VERSION
