Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: ftrace-fix-missing-kprobe-include.pathc
From: Thomas Gleixner <tglx@linutronix.de>
Date: Tue, 29 Jul 2008 23:31:46 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/kprobes.h |    5 +++++
 1 file changed, 5 insertions(+)

Index: linux-2.6.26/include/linux/kprobes.h
===================================================================
--- linux-2.6.26.orig/include/linux/kprobes.h
+++ linux-2.6.26/include/linux/kprobes.h
@@ -259,6 +259,11 @@ void recycle_rp_inst(struct kretprobe_in
 struct jprobe;
 struct kretprobe;
 
+static inline struct kprobe *get_kprobe(void *addr)
+{
+	return NULL;
+}
+
 static inline struct kprobe *kprobe_running(void)
 {
 	return NULL;
