Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/trace/trace.h |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

Index: linux-2.6-rt-test/kernel/trace/trace.h
===================================================================
--- linux-2.6-rt-test.orig/kernel/trace/trace.h	2008-05-07 13:36:45.000000000 -0400
+++ linux-2.6-rt-test/kernel/trace/trace.h	2008-05-07 13:47:26.000000000 -0400
@@ -338,11 +338,17 @@ void tracing_event_sysret(struct trace_a
 			  unsigned long ip,
 			  unsigned long ret);
 
-void tracing_start_function_trace(void);
-void tracing_stop_function_trace(void);
 int register_tracer(struct tracer *type);
 void unregister_tracer(struct tracer *type);
 
+#ifdef CONFIG_FTRACE
+void tracing_start_function_trace(void);
+void tracing_stop_function_trace(void);
+#else
+# define tracing_start_function_trace()	do { } while (0)
+# define tracing_stop_function_trace()	do { } while (0)
+#endif
+
 extern unsigned long nsecs_to_usecs(unsigned long nsecs);
 
 extern unsigned long tracing_max_latency;
