Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/futex.c  |    6 ++++--
 kernel/sysctl.c |    9 +++++++++
 2 files changed, 13 insertions(+), 2 deletions(-)

Index: linux-2.6.25.4-rt2/kernel/futex.c
===================================================================
--- linux-2.6.25.4-rt2.orig/kernel/futex.c	2008-05-19 16:55:56.000000000 -0400
+++ linux-2.6.25.4-rt2/kernel/futex.c	2008-05-19 16:56:04.000000000 -0400
@@ -129,12 +129,14 @@ static struct futex_hash_bucket futex_qu
 /* Futex-fs vfsmount entry: */
 static struct vfsmount *futex_mnt;
 
+int futex_performance_hack;
+
 /*
  * Take mm->mmap_sem, when futex is shared
  */
 static inline void futex_lock_mm(struct rw_semaphore *fshared)
 {
-	if (fshared)
+	if (fshared && !futex_performance_hack)
 		down_read(fshared);
 }
 
@@ -143,7 +145,7 @@ static inline void futex_lock_mm(struct 
  */
 static inline void futex_unlock_mm(struct rw_semaphore *fshared)
 {
-	if (fshared)
+	if (fshared && !futex_performance_hack)
 		up_read(fshared);
 }
 
Index: linux-2.6.25.4-rt2/kernel/sysctl.c
===================================================================
--- linux-2.6.25.4-rt2.orig/kernel/sysctl.c	2008-05-19 16:55:59.000000000 -0400
+++ linux-2.6.25.4-rt2/kernel/sysctl.c	2008-05-19 16:56:04.000000000 -0400
@@ -69,6 +69,7 @@ extern int sysctl_overcommit_ratio;
 extern int sysctl_panic_on_oom;
 extern int sysctl_oom_kill_allocating_task;
 extern int sysctl_oom_dump_tasks;
+extern int futex_performance_hack;
 extern int max_threads;
 extern int core_uses_pid;
 extern int suid_dumpable;
@@ -360,6 +361,14 @@ static struct ctl_table kern_table[] = {
 #endif
 	{
 		.ctl_name	= CTL_UNNUMBERED,
+		.procname	= "futex_performance_hack",
+		.data		= &futex_performance_hack,
+		.maxlen		= sizeof(int),
+		.mode		= 0644,
+		.proc_handler	= &proc_dointvec,
+	},
+	{
+		.ctl_name	= CTL_UNNUMBERED,
 		.procname	= "prof_pid",
 		.data		= &prof_pid,
 		.maxlen		= sizeof(int),
