Subject: Linux-RT 2.6.24-rc5-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/futex.c  |    6 ++++--
 kernel/sysctl.c |    9 +++++++++
 2 files changed, 13 insertions(+), 2 deletions(-)

Index: linux-2.6.24-rc5-rt1/kernel/futex.c
===================================================================
--- linux-2.6.24-rc5-rt1.orig/kernel/futex.c
+++ linux-2.6.24-rc5-rt1/kernel/futex.c
@@ -124,12 +124,14 @@ static struct futex_hash_bucket futex_qu
 /* Futex-fs vfsmount entry: */
 static struct vfsmount *futex_mnt;
 
+int futex_performance_hack;
+
 /*
  * Take mm->mmap_sem, when futex is shared
  */
 static inline void futex_lock_mm(struct rw_semaphore *fshared)
 {
-	if (fshared)
+	if (fshared && !futex_performance_hack)
 		down_read(fshared);
 }
 
@@ -138,7 +140,7 @@ static inline void futex_lock_mm(struct 
  */
 static inline void futex_unlock_mm(struct rw_semaphore *fshared)
 {
-	if (fshared)
+	if (fshared && !futex_performance_hack)
 		up_read(fshared);
 }
 
Index: linux-2.6.24-rc5-rt1/kernel/sysctl.c
===================================================================
--- linux-2.6.24-rc5-rt1.orig/kernel/sysctl.c
+++ linux-2.6.24-rc5-rt1/kernel/sysctl.c
@@ -68,6 +68,7 @@ extern int sysctl_overcommit_memory;
 extern int sysctl_overcommit_ratio;
 extern int sysctl_panic_on_oom;
 extern int sysctl_oom_kill_allocating_task;
+extern int futex_performance_hack;
 extern int max_threads;
 extern int core_uses_pid;
 extern int suid_dumpable;
@@ -350,6 +351,14 @@ static struct ctl_table kern_table[] = {
 	},
 	{
 		.ctl_name	= CTL_UNNUMBERED,
+		.procname	= "futex_performance_hack",
+		.data		= &futex_performance_hack,
+		.maxlen		= sizeof(int),
+		.mode		= 0644,
+		.proc_handler	= &proc_dointvec,
+	},
+	{
+		.ctl_name	= CTL_UNNUMBERED,
 		.procname	= "prof_pid",
 		.data		= &prof_pid,
 		.maxlen		= sizeof(int),
