---
 kernel/futex.c  |    6 ++++--
 kernel/sysctl.c |    9 +++++++++
 2 files changed, 13 insertions(+), 2 deletions(-)

Index: linux-2.6.22/kernel/futex.c
===================================================================
--- linux-2.6.22.orig/kernel/futex.c	2007-07-24 08:57:30.000000000 +0200
+++ linux-2.6.22/kernel/futex.c	2007-07-24 08:57:35.000000000 +0200
@@ -120,12 +120,14 @@ static struct futex_hash_bucket futex_qu
 /* Futex-fs vfsmount entry: */
 static struct vfsmount *futex_mnt;
 
+int futex_performance_hack;
+
 /*
  * Take mm->mmap_sem, when futex is shared
  */
 static inline void futex_lock_mm(struct rw_semaphore *fshared)
 {
-	if (fshared)
+	if (fshared && !futex_performance_hack)
 		down_read(fshared);
 }
 
@@ -134,7 +136,7 @@ static inline void futex_lock_mm(struct 
  */
 static inline void futex_unlock_mm(struct rw_semaphore *fshared)
 {
-	if (fshared)
+	if (fshared && !futex_performance_hack)
 		up_read(fshared);
 }
 
Index: linux-2.6.22/kernel/sysctl.c
===================================================================
--- linux-2.6.22.orig/kernel/sysctl.c	2007-07-24 08:57:31.000000000 +0200
+++ linux-2.6.22/kernel/sysctl.c	2007-07-24 08:57:35.000000000 +0200
@@ -66,6 +66,7 @@ extern int C_A_D;
 extern int sysctl_overcommit_memory;
 extern int sysctl_overcommit_ratio;
 extern int sysctl_panic_on_oom;
+extern int futex_performance_hack;
 extern int max_threads;
 extern int core_uses_pid;
 extern int suid_dumpable;
@@ -294,6 +295,14 @@ static ctl_table kern_table[] = {
 		.proc_handler	= &proc_dointvec,
 	},
 	{
+		.ctl_name	= CTL_UNNUMBERED,
+		.procname	= "futex_performance_hack",
+		.data		= &futex_performance_hack,
+		.maxlen		= sizeof(int),
+		.mode		= 0644,
+		.proc_handler	= &proc_dointvec,
+	},
+	{
 		.ctl_name	= KERN_PANIC,
 		.procname	= "prof_pid",
 		.data		= &prof_pid,
