Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 kernel/timer.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/kernel/timer.c b/kernel/timer.c
index ddc7fb3..6df6453 100644
--- a/kernel/timer.c
+++ b/kernel/timer.c
@@ -1139,6 +1139,13 @@ static inline void update_times(void)
 	static unsigned long last_tick = INITIAL_JIFFIES;
 	unsigned long ticks, flags;
 
+	/*
+	 * Dont take the xtime_lock from every CPU in
+	 * every tick - only when needed:
+	 */
+	if (jiffies == last_tick)
+		return;
+
 	write_seqlock_irqsave(&xtime_lock, flags);
 	ticks = jiffies - last_tick;
 	if (ticks) {
