Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/timer.c |    7 +++++++
 1 file changed, 7 insertions(+)

Index: linux-2.6.26/kernel/timer.c
===================================================================
--- linux-2.6.26.orig/kernel/timer.c
+++ linux-2.6.26/kernel/timer.c
@@ -1147,6 +1147,13 @@ static inline void update_times(void)
 	static unsigned long last_tick = INITIAL_JIFFIES;
 	unsigned long ticks, flags;
 
+	/*
+	 * Dont take the xtime_lock from every CPU in
+	 * every tick - only when needed:
+	 */
+	if (jiffies == last_tick)
+		return;
+
 	write_seqlock_irqsave(&xtime_lock, flags);
 	ticks = jiffies - last_tick;
 	if (ticks) {
