---
 kernel/timer.c |    7 +++++++
 1 file changed, 7 insertions(+)

Index: linux-2.6.22/kernel/timer.c
===================================================================
--- linux-2.6.22.orig/kernel/timer.c	2007-07-24 08:57:33.000000000 +0200
+++ linux-2.6.22/kernel/timer.c	2007-07-24 08:57:35.000000000 +0200
@@ -976,6 +976,13 @@ static inline void update_times(void)
 	static unsigned long last_tick;
 	unsigned long ticks, flags;
 
+	/*
+	 * Dont take the xtime_lock from every CPU in
+	 * every tick - only when needed:
+	 */
+	if (jiffies == last_tick)
+		return;
+
 	write_seqlock_irqsave(&xtime_lock, flags);
 	ticks = jiffies - last_tick;
 	if (ticks) {
