Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86/kernel/io_apic_32.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.24.2-rt2/arch/x86/kernel/io_apic_32.c
===================================================================
--- linux-2.6.24.2-rt2.orig/arch/x86/kernel/io_apic_32.c	2008-02-22 16:12:04.000000000 -0500
+++ linux-2.6.24.2-rt2/arch/x86/kernel/io_apic_32.c	2008-02-23 10:11:56.000000000 -0500
@@ -1346,7 +1346,7 @@ static void __init setup_IO_APIC_irqs(vo
 			vector = assign_irq_vector(irq);
 			entry.vector = vector;
 			ioapic_register_intr(irq, vector, IOAPIC_AUTO,
-					     apic > 0);
+					     1);
 		
 			if (!apic && (irq < 16))
 				disable_8259A_irq(irq);
@@ -2864,7 +2864,7 @@ int io_apic_set_pci_routing (int ioapic,
 		mp_ioapics[ioapic].mpc_apicid, pin, entry.vector, irq,
 		edge_level, active_high_low);
 
-	ioapic_register_intr(irq, entry.vector, edge_level, ioapic > 0);
+	ioapic_register_intr(irq, entry.vector, edge_level, 1);
 
 	if (!ioapic && (irq < 16))
 		disable_8259A_irq(irq);
