Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 mm/highmem.c |   10 ----------
 1 files changed, 0 insertions(+), 10 deletions(-)

diff --git a/mm/highmem.c b/mm/highmem.c
index e16e152..22f4ba6 100644
--- a/mm/highmem.c
+++ b/mm/highmem.c
@@ -105,16 +105,6 @@ static void flush_all_zero_pkmaps(void)
 	flush_tlb_kernel_range(PKMAP_ADDR(0), PKMAP_ADDR(LAST_PKMAP));
 }
 
-/**
- * kmap_flush_unused - flush all unused kmap mappings in order to remove stray mappings
- */
-void kmap_flush_unused(void)
-{
-	spin_lock(&kmap_lock);
-	flush_all_zero_pkmaps();
-	spin_unlock(&kmap_lock);
-}
-
 static inline unsigned long map_new_virtual(struct page *page)
 {
 	unsigned long vaddr;
