Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 mm/highmem.c |   10 ----------
 1 file changed, 10 deletions(-)

Index: linux-2.6.26/mm/highmem.c
===================================================================
--- linux-2.6.26.orig/mm/highmem.c
+++ linux-2.6.26/mm/highmem.c
@@ -104,16 +104,6 @@ static void flush_all_zero_pkmaps(void)
 	flush_tlb_kernel_range(PKMAP_ADDR(0), PKMAP_ADDR(LAST_PKMAP));
 }
 
-/**
- * kmap_flush_unused - flush all unused kmap mappings in order to remove stray mappings
- */
-void kmap_flush_unused(void)
-{
-	spin_lock(&kmap_lock);
-	flush_all_zero_pkmaps();
-	spin_unlock(&kmap_lock);
-}
-
 static inline unsigned long map_new_virtual(struct page *page)
 {
 	unsigned long vaddr;
