Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 mm/highmem.c |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/mm/highmem.c b/mm/highmem.c
index 1a390da..d4e8870 100644
--- a/mm/highmem.c
+++ b/mm/highmem.c
@@ -214,6 +214,14 @@ static unsigned long pkmap_insert(struct page *page)
 	return vaddr;
 }
 
+/*
+ * Flush all unused kmap mappings in order to remove stray mappings.
+ */
+void kmap_flush_unused(void)
+{
+	WARN_ON_ONCE(1);
+}
+
  void *kmap_high(struct page *page)
 {
 	unsigned long vaddr;
