---
 mm/highmem.c |    9 ---------
 1 file changed, 9 deletions(-)

Index: linux-2.6.22/mm/highmem.c
===================================================================
--- linux-2.6.22.orig/mm/highmem.c	2007-07-24 08:56:04.000000000 +0200
+++ linux-2.6.22/mm/highmem.c	2007-07-24 08:57:39.000000000 +0200
@@ -99,15 +99,6 @@ static void flush_all_zero_pkmaps(void)
 	flush_tlb_kernel_range(PKMAP_ADDR(0), PKMAP_ADDR(LAST_PKMAP));
 }
 
-/* Flush all unused kmap mappings in order to remove stray
-   mappings. */
-void kmap_flush_unused(void)
-{
-	spin_lock(&kmap_lock);
-	flush_all_zero_pkmaps();
-	spin_unlock(&kmap_lock);
-}
-
 static inline unsigned long map_new_virtual(struct page *page)
 {
 	unsigned long vaddr;
