Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 drivers/char/hpet.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.23-rc4/drivers/char/hpet.c
===================================================================
--- linux-2.6.23-rc4.orig/drivers/char/hpet.c	2007-08-31 21:51:20.000000000 +0200
+++ linux-2.6.23-rc4/drivers/char/hpet.c	2007-08-31 21:51:20.000000000 +0200
@@ -909,6 +909,8 @@ int hpet_alloc(struct hpet_data *hdp)
 
 	hpetp->hp_delta = hpet_calibrate(hpetp);
 
+#ifdef CONFIG_IA64
+	/* This clocksource driver currently only works on ia64 */
 	if (!hpet_clocksource) {
 		hpet_mctr = (void __iomem *)&hpetp->hp_hpet->hpet_mc;
 		CLKSRC_FSYS_MMIO_SET(clocksource_hpet.fsys_mmio, hpet_mctr);
@@ -918,6 +920,7 @@ int hpet_alloc(struct hpet_data *hdp)
 		hpetp->hp_clocksource = &clocksource_hpet;
 		hpet_clocksource = &clocksource_hpet;
 	}
+#endif
 
 	return 0;
 }
