Subject: compile fix for HRT-disabled configurations
From: Patrick Mullaney <pmullaney@novell.com>

Index: linux-2.6.22/kernel/hrtimer.c
===================================================================
--- linux-2.6.22.orig/kernel/hrtimer.c
+++ linux-2.6.22/kernel/hrtimer.c
@@ -616,7 +616,7 @@ static inline int hrtimer_switch_to_hres
 {
 	return 0;
 }
-static inline void hrtimer_force_reprogram(struct hrtimer_cpu_base *base) { }
+static inline void hrtimer_force_reprogram(struct hrtimer_cpu_base *base, int force) { }
 static inline int hrtimer_enqueue_reprogram(struct hrtimer *timer,
 					    struct hrtimer_clock_base *base)
 {
