---
 kernel/hrtimer.c          |    3 +--
 kernel/time/timekeeping.c |    2 ++
 2 files changed, 3 insertions(+), 2 deletions(-)

Index: linux-2.6.22/kernel/hrtimer.c
===================================================================
--- linux-2.6.22.orig/kernel/hrtimer.c	2007-07-24 08:57:18.000000000 +0200
+++ linux-2.6.22/kernel/hrtimer.c	2007-07-24 08:57:34.000000000 +0200
@@ -589,8 +589,7 @@ static int hrtimer_switch_to_hres(void)
 	/* "Retrigger" the interrupt to get things going */
 	retrigger_next_event(NULL);
 	local_irq_restore(flags);
-	printk(KERN_INFO "Switched to high resolution mode on CPU %d\n",
-	       smp_processor_id());
+
 	return 1;
 }
 
Index: linux-2.6.22/kernel/time/timekeeping.c
===================================================================
--- linux-2.6.22.orig/kernel/time/timekeeping.c	2007-07-24 08:57:33.000000000 +0200
+++ linux-2.6.22/kernel/time/timekeeping.c	2007-07-24 08:57:34.000000000 +0200
@@ -254,8 +254,10 @@ static void change_clocksource(void)
 
 	tick_clock_notify();
 
+#ifndef CONFIG_PREEMPT_RT
 	printk(KERN_INFO "Time: %s clocksource has been installed.\n",
 	       clock->name);
+#endif
 }
 #else
 static inline void change_clocksource(void) { }
