---
 include/linux/hrtimer.h |    3 +++
 kernel/hrtimer.c        |   22 ++++++++++++++++++++++
 2 files changed, 25 insertions(+)

Index: linux-2.6.22/include/linux/hrtimer.h
===================================================================
--- linux-2.6.22.orig/include/linux/hrtimer.h	2007-07-24 08:56:39.000000000 +0200
+++ linux-2.6.22/include/linux/hrtimer.h	2007-07-24 08:57:02.000000000 +0200
@@ -297,6 +297,9 @@ static inline int hrtimer_is_queued(stru
 /* Forward a hrtimer so it expires after now: */
 extern unsigned long
 hrtimer_forward(struct hrtimer *timer, ktime_t now, ktime_t interval);
+/* Overrun count: */
+extern unsigned long
+hrtimer_overrun(struct hrtimer *timer, ktime_t now, ktime_t interval);
 
 /* Precise sleep: */
 extern long hrtimer_nanosleep(struct timespec *rqtp,
Index: linux-2.6.22/kernel/hrtimer.c
===================================================================
--- linux-2.6.22.orig/kernel/hrtimer.c	2007-07-24 08:56:53.000000000 +0200
+++ linux-2.6.22/kernel/hrtimer.c	2007-07-24 08:57:02.000000000 +0200
@@ -674,6 +674,28 @@ hrtimer_forward(struct hrtimer *timer, k
 }
 EXPORT_SYMBOL_GPL(hrtimer_forward);
 
+unsigned long
+hrtimer_overrun(struct hrtimer *timer, ktime_t now, ktime_t interval)
+{
+	unsigned long orun = 1;
+	ktime_t delta;
+
+	delta = ktime_sub(now, timer->expires);
+
+	if (delta.tv64 < 0)
+		return 0;
+
+	if (interval.tv64 < timer->base->resolution.tv64)
+		interval.tv64 = timer->base->resolution.tv64;
+
+	if (unlikely(delta.tv64 >= interval.tv64))
+		orun = ktime_divns(delta, ktime_to_ns(interval)) + 1;
+
+	return orun;
+}
+EXPORT_SYMBOL_GPL(hrtimer_overrun);
+
+
 /*
  * enqueue_hrtimer - internal function to (re)start a timer
  *
