Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/asm-x86/atomic_32.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.24-rt1/include/asm-x86/atomic_32.h
===================================================================
--- linux-2.6.24-rt1.orig/include/asm-x86/atomic_32.h	2008-01-25 15:06:46.000000000 -0500
+++ linux-2.6.24-rt1/include/asm-x86/atomic_32.h	2008-01-25 15:07:09.000000000 -0500
@@ -195,10 +195,10 @@ static __inline__ int atomic_add_return(
 
 #ifdef CONFIG_M386
 no_xadd: /* Legacy 386 processor */
-	local_irq_save(flags);
+	raw_local_irq_save(flags);
 	__i = atomic_read(v);
 	atomic_set(v, i + __i);
-	local_irq_restore(flags);
+	raw_local_irq_restore(flags);
 	return i + __i;
 #endif
 }
