Subject: i386: Remove the useless #ifdef in i8253.h

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Chris Wright <chrisw@sous-sol.org>
Signed-off-by: Ingo Molnar <mingo@elte.hu>

---
 include/asm-i386/i8253.h |    4 ----
 1 file changed, 4 deletions(-)

Index: linux-2.6.22/include/asm-i386/i8253.h
===================================================================
--- linux-2.6.22.orig/include/asm-i386/i8253.h	2007-07-24 08:56:56.000000000 +0200
+++ linux-2.6.22/include/asm-i386/i8253.h	2007-07-24 08:56:58.000000000 +0200
@@ -8,12 +8,8 @@
 
 extern spinlock_t i8253_lock;
 
-#ifdef CONFIG_GENERIC_CLOCKEVENTS
-
 extern struct clock_event_device *global_clock_event;
 
 extern void setup_pit_timer(void);
 
-#endif
-
 #endif	/* __ASM_I8253_H__ */
