---
 include/asm-i386/atomic.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.22/include/asm-i386/atomic.h
===================================================================
--- linux-2.6.22.orig/include/asm-i386/atomic.h	2007-07-24 08:56:39.000000000 +0200
+++ linux-2.6.22/include/asm-i386/atomic.h	2007-07-24 08:57:01.000000000 +0200
@@ -195,10 +195,10 @@ static __inline__ int atomic_add_return(
 
 #ifdef CONFIG_M386
 no_xadd: /* Legacy 386 processor */
-	local_irq_save(flags);
+	raw_local_irq_save(flags);
 	__i = atomic_read(v);
 	atomic_set(v, i + __i);
-	local_irq_restore(flags);
+	raw_local_irq_restore(flags);
 	return i + __i;
 #endif
 }
