Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/asm-x86/atomic_32.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.26.3-rt6/include/asm-x86/atomic_32.h
===================================================================
--- linux-2.6.26.3-rt6.orig/include/asm-x86/atomic_32.h	2008-09-05 00:22:06.000000000 -0400
+++ linux-2.6.26.3-rt6/include/asm-x86/atomic_32.h	2008-09-05 00:22:33.000000000 -0400
@@ -188,10 +188,10 @@ static inline int atomic_add_return(int 
 
 #ifdef CONFIG_M386
 no_xadd: /* Legacy 386 processor */
-	local_irq_save(flags);
+	raw_local_irq_save(flags);
 	__i = atomic_read(v);
 	atomic_set(v, i + __i);
-	local_irq_restore(flags);
+	raw_local_irq_restore(flags);
 	return i + __i;
 #endif
 }
