---
 drivers/acpi/processor_idle.c |    8 ++++++++
 1 file changed, 8 insertions(+)

Index: linux-2.6.22/drivers/acpi/processor_idle.c
===================================================================
--- linux-2.6.22.orig/drivers/acpi/processor_idle.c	2007-07-24 08:56:58.000000000 +0200
+++ linux-2.6.22/drivers/acpi/processor_idle.c	2007-07-24 08:57:08.000000000 +0200
@@ -836,6 +836,12 @@ static inline void acpi_idle_update_bm_r
  */
 static inline void acpi_idle_do_entry(struct acpi_processor_cx *cx)
 {
+	/*
+	 * We have irqs disabled here, so stop latency tracing
+	 * at this point and restart it after we return:
+	 */
+	stop_critical_timing();
+
 	if (cx->space_id == ACPI_CSTATE_FFH) {
 		/* Call into architectural FFH based C-state */
 		acpi_processor_ffh_cstate_enter(cx);
@@ -848,6 +854,8 @@ static inline void acpi_idle_do_entry(st
 		   gets asserted in time to freeze execution properly. */
 		unused = inl(acpi_gbl_FADT.xpm_timer_block.address);
 	}
+
+	touch_critical_timing();
 }
 
 /**
