Subject: Disable KDB IST stacks for debug/stack fault on PREEMPT_RT
From: Sven-Thorsten Dietrich <sdietrich@suse.de>
Signed-off-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Acked-by: Tony Jones <tonyj@suse.de>

--- linux-2.6.26-RT.orig/arch/x86/kdb/kdba_bt.c
+++ linux-2.6.26-RT/arch/x86/kdb/kdba_bt.c
@@ -4875,10 +4875,14 @@ kdba_get_stack_info_alternate(kdb_machre
 		unsigned int nested_size;
 		unsigned int next;
 	} *sdp, stack_data[] = {
+#if STACKFAULT_STACK > 0
 		[STACKFAULT_STACK - 1] =  { "stackfault",    EXCEPTION_STKSZ, EXCEPTION_STKSZ, EXCEPTION_STKSZ - 2*sizeof(void *) },
+#endif
 		[DOUBLEFAULT_STACK - 1] = { "doublefault",   EXCEPTION_STKSZ, EXCEPTION_STKSZ, EXCEPTION_STKSZ - 2*sizeof(void *) },
 		[NMI_STACK - 1] =         { "nmi",           EXCEPTION_STKSZ, EXCEPTION_STKSZ, EXCEPTION_STKSZ - 2*sizeof(void *) },
+#if DEBUG_STACK > 0
 		[DEBUG_STACK - 1] =       { "debug",         DEBUG_STKSZ,     EXCEPTION_STKSZ, EXCEPTION_STKSZ - 2*sizeof(void *) },
+#endif
 		[MCE_STACK - 1] =         { "machine check", EXCEPTION_STKSZ, EXCEPTION_STKSZ, EXCEPTION_STKSZ - 2*sizeof(void *) },
 		[INTERRUPT_STACK - 1] =   { "interrupt",     IRQSTACKSIZE,    IRQSTACKSIZE,    IRQSTACKSIZE    -   sizeof(void *) },
 	};
