Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86/mm/highmem_32.c |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6-rt-test/arch/x86/mm/highmem_32.c
===================================================================
--- linux-2.6-rt-test.orig/arch/x86/mm/highmem_32.c	2008-05-07 11:37:40.000000000 -0400
+++ linux-2.6-rt-test/arch/x86/mm/highmem_32.c	2008-05-07 13:02:28.000000000 -0400
@@ -37,6 +37,7 @@ struct page *kmap_to_page(void *ptr)
 	page = pte_page(pkmap_page_table[PKMAP_NR((unsigned long)ptr)]);
 	return page;
 }
+EXPORT_SYMBOL_GPL(kmap_to_page); /* PREEMPT_RT converts some modules to use this */
 
 static void debug_kmap_atomic_prot(enum km_type type)
 {
