Subject: Linux-RT 2.6.24-rc8-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/latency_trace.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.24-rc8-rt1/kernel/latency_trace.c
===================================================================
--- linux-2.6.24-rc8-rt1.orig/kernel/latency_trace.c	2008-01-16 22:28:13.000000000 -0500
+++ linux-2.6.24-rc8-rt1/kernel/latency_trace.c	2008-01-16 22:28:14.000000000 -0500
@@ -42,6 +42,9 @@ int trace_use_raw_cycles = 0;
 #define __raw_spinlock_t raw_spinlock_t
 #define need_resched_delayed() 0
 
+#define __raw_spinlock_t raw_spinlock_t
+#define need_resched_delayed() 0
+
 #ifdef CONFIG_EVENT_TRACE
 /*
  * Convert raw cycles to usecs.
