---
 kernel/latency_trace.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.22/kernel/latency_trace.c
===================================================================
--- linux-2.6.22.orig/kernel/latency_trace.c	2007-07-24 08:57:07.000000000 +0200
+++ linux-2.6.22/kernel/latency_trace.c	2007-07-24 08:57:07.000000000 +0200
@@ -40,6 +40,9 @@
 
 int trace_use_raw_cycles = 0;
 
+#define __raw_spinlock_t raw_spinlock_t
+#define need_resched_delayed() 0
+
 #ifdef CONFIG_EVENT_TRACE
 /*
  * Convert raw cycles to usecs.
