Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From: Daniel Walker <dwalker@mvista.com>
Subject: [PATCH -rt] Add wakeup timing to expose /proc/preempt_max_latency

Fix my original mistake of using CONFIG_CRITICAL_TIMING , which only
covers preempt and irqs-off timing.. Convert to CONFIG_LATENCY_TIMING,
so wakeup timing also has these.

Signed-Off-By: Daniel Walker <dwalker@mvista.com>

---
 kernel/sysctl.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24-rt1/kernel/sysctl.c
===================================================================
--- linux-2.6.24-rt1.orig/kernel/sysctl.c	2008-01-25 15:07:18.000000000 -0500
+++ linux-2.6.24-rt1/kernel/sysctl.c	2008-01-25 15:07:19.000000000 -0500
@@ -359,7 +359,7 @@ static struct ctl_table kern_table[] = {
 		.proc_handler	= &proc_dointvec,
 	},
 #endif
-#if defined(CONFIG_CRITICAL_TIMING)
+#if defined(CONFIG_LATENCY_TIMING)
 	{
 		.ctl_name	= CTL_UNNUMBERED,
 		.procname	= "preempt_max_latency",
