Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 include/linux/lockdep.h |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/include/linux/lockdep.h b/include/linux/lockdep.h
index 118fd1a..e8cba5c 100644
--- a/include/linux/lockdep.h
+++ b/include/linux/lockdep.h
@@ -371,8 +371,8 @@ do {								\
 	if (!f_try(&(_lock)->lock)) {				\
 		lock_contended(&(_lock)->dep_map, _RET_IP_);	\
 		f_lock(&(_lock)->lock);				\
-		lock_acquired(&(_lock)->dep_map);		\
 	}							\
+	lock_acquired(&(_lock)->dep_map);			\
 } while (0)
 
 
@@ -382,9 +382,9 @@ do {								\
 	if (!f_try(&(_lock)->lock)) {				\
 		lock_contended(&(_lock)->dep_map, _RET_IP_);	\
 		ret = f_lock(&(_lock)->lock);			\
- 		if (!ret)					\
-			lock_acquired(&(_lock)->dep_map);	\
 	}							\
+	if (!ret)						\
+		lock_acquired(&(_lock)->dep_map);		\
  	ret;							\
 })
 
