---
 include/linux/lockdep.h |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

Index: linux-2.6.22/include/linux/lockdep.h
===================================================================
--- linux-2.6.22.orig/include/linux/lockdep.h	2007-07-24 08:57:43.000000000 +0200
+++ linux-2.6.22/include/linux/lockdep.h	2007-07-24 08:57:43.000000000 +0200
@@ -347,8 +347,8 @@ do {								\
 	if (!f_try(&(_lock)->lock)) {				\
 		lock_contended(&(_lock)->dep_map, _RET_IP_);	\
 		f_lock(&(_lock)->lock);				\
-		lock_acquired(&(_lock)->dep_map);		\
 	}							\
+	lock_acquired(&(_lock)->dep_map);			\
 } while (0)
 
 
@@ -358,9 +358,9 @@ do {								\
 	if (!f_try(&(_lock)->lock)) {				\
 		lock_contended(&(_lock)->dep_map, _RET_IP_);	\
 		ret = f_lock(&(_lock)->lock);			\
- 		if (!ret)					\
-			lock_acquired(&(_lock)->dep_map);	\
 	}							\
+	if (!ret)						\
+		lock_acquired(&(_lock)->dep_map);		\
  	ret;							\
 })
 
