Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/linux/lockdep.h |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

Index: linux-2.6.24-rc2-rt1/include/linux/lockdep.h
===================================================================
--- linux-2.6.24-rc2-rt1.orig/include/linux/lockdep.h
+++ linux-2.6.24-rc2-rt1/include/linux/lockdep.h
@@ -366,8 +366,8 @@ do {								\
 	if (!f_try(&(_lock)->lock)) {				\
 		lock_contended(&(_lock)->dep_map, _RET_IP_);	\
 		f_lock(&(_lock)->lock);				\
-		lock_acquired(&(_lock)->dep_map);		\
 	}							\
+	lock_acquired(&(_lock)->dep_map);			\
 } while (0)
 
 
@@ -377,9 +377,9 @@ do {								\
 	if (!f_try(&(_lock)->lock)) {				\
 		lock_contended(&(_lock)->dep_map, _RET_IP_);	\
 		ret = f_lock(&(_lock)->lock);			\
- 		if (!ret)					\
-			lock_acquired(&(_lock)->dep_map);	\
 	}							\
+	if (!ret)						\
+		lock_acquired(&(_lock)->dep_map);		\
  	ret;							\
 })
 
