Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 lib/locking-selftest.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.26.3-rt6/lib/locking-selftest.c
===================================================================
--- linux-2.6.26.3-rt6.orig/lib/locking-selftest.c	2008-09-05 00:23:07.000000000 -0400
+++ linux-2.6.26.3-rt6/lib/locking-selftest.c	2008-09-05 00:24:04.000000000 -0400
@@ -948,7 +948,7 @@ static void dotest(void (*testcase_fn)(v
 	unsigned long saved_preempt_count = preempt_count();
 	int expected_failure = 0;
 #if defined(CONFIG_DEBUG_PREEMPT) && defined(CONFIG_DEBUG_RT_MUTEXES)
-        int saved_lock_count = current->lock_count;
+        atomic_t saved_lock_count = current->lock_count;
 #endif
 
 	WARN_ON(irqs_disabled());
