Subject: Linux-RT 2.6.24-rc5-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>


Signed-off-by: Peter Zijlstra <a.p.zijlstra@chello.nl>
---
 kernel/lockdep.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.24-rc5-rt1/kernel/lockdep.c
===================================================================
--- linux-2.6.24-rc5-rt1.orig/kernel/lockdep.c
+++ linux-2.6.24-rc5-rt1/kernel/lockdep.c
@@ -2755,6 +2755,9 @@ lock_set_subclass(struct lockdep_map *lo
 {
 	unsigned long flags;
 
+	if (unlikely(!lock_stat && !prove_locking))
+		return;
+
 	if (unlikely(current->lockdep_recursion))
 		return;
 
