Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
From fb918e8b87e699c8174d8f7fbff1ab558c7389b0 Mon Sep 17 00:00:00 2001
From: Sebastian Siewior <bigeasy@linutronix.de>
Date: Fri, 18 Apr 2008 17:02:29 +0200
Subject: [PATCH] m68knommu: make cmpxchg RT-safe

Signed-off-by: Sebastian Siewior <bigeasy@linutronix.de>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/asm-m68knommu/system.h |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

Index: linux-2.6.26.3-rt6/arch/m68knommu/include/asm/system.h
===================================================================
--- linux-2.6.26.3-rt6.orig/arch/m68knommu/include/asm/system.h	2008-09-05 00:22:17.000000000 -0400
+++ linux-2.6.26.3-rt6/arch/m68knommu/include/asm/system.h	2008-09-05 00:22:39.000000000 -0400
@@ -2,9 +2,10 @@
 #define _M68KNOMMU_SYSTEM_H
 
 #include <linux/linkage.h>
+#include <linux/kernel.h>
+#include <linux/irqflags.h>
 #include <asm/segment.h>
 #include <asm/entry.h>
-
 /*
  * switch_to(n) should switch tasks to task ptr, first checking that
  * ptr isn't the current task, in which case it does nothing.  This
@@ -130,7 +131,7 @@ static inline unsigned long __xchg(unsig
 {
   unsigned long tmp, flags;
 
-  local_irq_save(flags);
+  raw_local_irq_save(flags);
 
   switch (size) {
   case 1:
@@ -152,7 +153,7 @@ static inline unsigned long __xchg(unsig
     : "=&d" (tmp) : "d" (x), "m" (*__xg(ptr)) : "memory");
     break;
   }
-  local_irq_restore(flags);
+  raw_local_irq_restore(flags);
   return tmp;
 }
 #else
