Subject: Linux-RT 2.6.24-rc8-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 drivers/pci/msi.c |    4 ++++
 1 file changed, 4 insertions(+)

Index: linux-2.6.24-rc8-rt1/drivers/pci/msi.c
===================================================================
--- linux-2.6.24-rc8-rt1.orig/drivers/pci/msi.c	2008-01-16 22:27:46.000000000 -0500
+++ linux-2.6.24-rc8-rt1/drivers/pci/msi.c	2008-01-16 22:28:09.000000000 -0500
@@ -241,6 +241,10 @@ static void __pci_restore_msi_state(stru
 		return;
 
 	entry = get_irq_msi(dev->irq);
+	if (!entry) {
+		WARN_ON(1);
+		return;
+	}
 	pos = entry->msi_attrib.pos;
 
 	pci_intx_for_msi(dev, 0);
