Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 net/core/dev.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24-rt1/net/core/dev.c
===================================================================
--- linux-2.6.24-rt1.orig/net/core/dev.c	2008-01-25 15:06:37.000000000 -0500
+++ linux-2.6.24-rt1/net/core/dev.c	2008-01-25 15:07:37.000000000 -0500
@@ -1801,8 +1801,8 @@ int netif_rx_ni(struct sk_buff *skb)
 {
 	int err;
 
-	preempt_disable();
 	err = netif_rx(skb);
+	preempt_disable();
 	if (local_softirq_pending())
 		do_softirq();
 	preempt_enable();
