Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 net/core/dev.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.25.2-rt1/net/core/dev.c
===================================================================
--- linux-2.6.25.2-rt1.orig/net/core/dev.c	2008-05-08 16:18:11.000000000 -0400
+++ linux-2.6.25.2-rt1/net/core/dev.c	2008-05-08 16:18:53.000000000 -0400
@@ -1805,8 +1805,8 @@ int netif_rx_ni(struct sk_buff *skb)
 {
 	int err;
 
-	preempt_disable();
 	err = netif_rx(skb);
+	preempt_disable();
 	if (local_softirq_pending())
 		do_softirq();
 	preempt_enable();
