---
 net/core/dev.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/net/core/dev.c
===================================================================
--- linux-2.6.22.orig/net/core/dev.c	2007-07-24 08:56:20.000000000 +0200
+++ linux-2.6.22/net/core/dev.c	2007-07-24 08:57:21.000000000 +0200
@@ -1679,8 +1679,8 @@ int netif_rx_ni(struct sk_buff *skb)
 {
 	int err;
 
-	preempt_disable();
 	err = netif_rx(skb);
+	preempt_disable();
 	if (local_softirq_pending())
 		do_softirq();
 	preempt_enable();
