Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/net/netfilter/nf_conntrack.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.25.4-rt2/include/net/netfilter/nf_conntrack.h
===================================================================
--- linux-2.6.25.4-rt2.orig/include/net/netfilter/nf_conntrack.h	2008-05-19 16:55:42.000000000 -0400
+++ linux-2.6.25.4-rt2/include/net/netfilter/nf_conntrack.h	2008-05-19 16:56:15.000000000 -0400
@@ -262,7 +262,7 @@ DECLARE_PER_CPU(struct ip_conntrack_stat
 #define NF_CT_STAT_INC_ATOMIC(count)			\
 do {							\
 	local_bh_disable();				\
-	__get_cpu_var(nf_conntrack_stat).count++;	\
+	__raw_get_cpu_var(nf_conntrack_stat).count++;	\
 	local_bh_enable();				\
 } while (0)
 
