Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 kernel/panic.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.27-RT/kernel/panic.c
===================================================================
--- linux-2.6.27-RT.orig/kernel/panic.c
+++ linux-2.6.27-RT/kernel/panic.c
@@ -105,7 +105,7 @@ NORET_TYPE void panic(const char * fmt,
 	 * unfortunately means it may not be hardened to work in a panic
 	 * situation.
 	 */
-	smp_send_stop();
+//	smp_send_stop();
 #endif
 
 	atomic_notifier_call_chain(&panic_notifier_list, 0, buf);
