---
 kernel/panic.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/kernel/panic.c
===================================================================
--- linux-2.6.22.orig/kernel/panic.c	2007-07-24 08:57:31.000000000 +0200
+++ linux-2.6.22/kernel/panic.c	2007-07-24 08:57:35.000000000 +0200
@@ -96,7 +96,7 @@ NORET_TYPE void panic(const char * fmt, 
 	 * unfortunately means it may not be hardened to work in a panic
 	 * situation.
 	 */
-	smp_send_stop();
+//	smp_send_stop();
 #endif
 
 	atomic_notifier_call_chain(&panic_notifier_list, 0, buf);
