---
 arch/i386/kernel/paravirt.c |   12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

Index: linux-2.6.22/arch/i386/kernel/paravirt.c
===================================================================
--- linux-2.6.22.orig/arch/i386/kernel/paravirt.c	2007-07-24 08:55:59.000000000 +0200
+++ linux-2.6.22/arch/i386/kernel/paravirt.c	2007-07-24 08:57:44.000000000 +0200
@@ -228,6 +228,16 @@ static int __init print_banner(void)
 }
 core_initcall(print_banner);
 
+#ifdef CONFIG_HIGHPTE
+/*
+ * kmap_atomic() might be an inline or a macro:
+ */
+static void *kmap_atomic_func(struct page *page, enum km_type idx)
+{
+	return kmap_atomic(page, idx);
+}
+#endif
+
 struct paravirt_ops paravirt_ops = {
 	.name = "bare hardware",
 	.paravirt_enabled = 0,
@@ -316,7 +326,7 @@ struct paravirt_ops paravirt_ops = {
 	.pte_update_defer = paravirt_nop,
 
 #ifdef CONFIG_HIGHPTE
-	.kmap_atomic_pte = kmap_atomic,
+	.kmap_atomic_pte = kmap_atomic_func,
 #endif
 
 #ifdef CONFIG_X86_PAE
