Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: pmtmr: allow command line override of ioport
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 21 May 2008 21:14:58 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 drivers/clocksource/acpi_pm.c |   19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

Index: linux-2.6.26/drivers/clocksource/acpi_pm.c
===================================================================
--- linux-2.6.26.orig/drivers/clocksource/acpi_pm.c
+++ linux-2.6.26/drivers/clocksource/acpi_pm.c
@@ -215,3 +215,22 @@ pm_good:
  * but we still need to load before device_initcall
  */
 fs_initcall(init_acpi_pm_clocksource);
+
+/*
+ * Allow an override of the IOPort. Stupid BIOSes do not tell us about
+ * the PMTimer, but we might know where it is.
+ */
+static int __init parse_pmtmr(char *arg)
+{
+	unsigned long base;
+	char *e;
+
+	base = simple_strtoul(arg, &e, 16);
+
+	printk(KERN_INFO "PMTMR IOPort override: 0x%04lx -> 0x%04lx\n",
+	       (unsigned long) pmtmr_ioport, base);
+	pmtmr_ioport = base;
+
+	return 0;
+}
+__setup("pmtmr=", parse_pmtmr);
