Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From tony@bakeyournoodle.com Wed Sep 26 10:29:23 2007
Date: Tue, 04 Sep 2007 17:09:02 +1000
From: Tony Breeds <tony@bakeyournoodle.com>
To: linux-rt-users@vger.kernel.org
Subject: [PATCH 3/5] [POWERPC] Match __rw_yeild function declaration to
    prototype.

Signed-off-by: Tony Breeds <tony@bakeyournoodle.com>

---

 arch/powerpc/lib/locks.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.26/arch/powerpc/lib/locks.c
===================================================================
--- linux-2.6.26.orig/arch/powerpc/lib/locks.c
+++ linux-2.6.26/arch/powerpc/lib/locks.c
@@ -55,7 +55,7 @@ void __spin_yield(__raw_spinlock_t *lock
  * This turns out to be the same for read and write locks, since
  * we only know the holder if it is write-locked.
  */
-void __rw_yield(raw_rwlock_t *rw)
+void __rw_yield(__raw_rwlock_t *rw)
 {
 	int lock_value;
 	unsigned int holder_cpu, yield_count;
