Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/powerpc/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.24-rt1/arch/powerpc/Kconfig
===================================================================
--- linux-2.6.24-rt1.orig/arch/powerpc/Kconfig	2008-01-25 15:06:50.000000000 -0500
+++ linux-2.6.24-rt1/arch/powerpc/Kconfig	2008-01-25 15:06:55.000000000 -0500
@@ -79,6 +79,7 @@ config ARCH_NO_VIRT_TO_BUS
 config PPC
 	bool
 	default y
+	select HAVE_MCOUNT
 
 config EARLY_PRINTK
 	bool
