---
 arch/powerpc/kernel/irq.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/arch/powerpc/kernel/irq.c
===================================================================
--- linux-2.6.22.orig/arch/powerpc/kernel/irq.c	2007-07-24 08:56:35.000000000 +0200
+++ linux-2.6.22/arch/powerpc/kernel/irq.c	2007-07-24 08:57:06.000000000 +0200
@@ -114,7 +114,7 @@ static inline void set_soft_enabled(unsi
 	: : "r" (enable), "i" (offsetof(struct paca_struct, soft_enabled)));
 }
 
-void local_irq_restore(unsigned long en)
+void notrace local_irq_restore(unsigned long en)
 {
 	/*
 	 * get_paca()->soft_enabled = en;
