---
 arch/powerpc/kernel/time.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/arch/powerpc/kernel/time.c
===================================================================
--- linux-2.6.22.orig/arch/powerpc/kernel/time.c	2007-07-24 08:57:11.000000000 +0200
+++ linux-2.6.22/arch/powerpc/kernel/time.c	2007-07-24 08:57:11.000000000 +0200
@@ -898,7 +898,7 @@ void div128_by_32(u64 dividend_high, u64
 
 #include <linux/clocksource.h>
 
-static cycle_t timebase_read(void)
+static cycle_t notrace timebase_read(void)
 {
 	return (cycle_t)get_tb();
 }
