Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/powerpc/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.25.4-rt2/arch/powerpc/Kconfig
===================================================================
--- linux-2.6.25.4-rt2.orig/arch/powerpc/Kconfig	2008-05-19 16:55:10.000000000 -0400
+++ linux-2.6.25.4-rt2/arch/powerpc/Kconfig	2008-05-19 16:55:23.000000000 -0400
@@ -90,6 +90,7 @@ config PPC
 	select HAVE_IDE
 	select HAVE_OPROFILE
 	select HAVE_KPROBES
+	select HAVE_MCOUNT
 	select HAVE_KRETPROBES
 
 config EARLY_PRINTK
