Subject: Linux-RT 2.6.24-rc8-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/powerpc/kernel/time.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24-rc8-rt1/arch/powerpc/kernel/time.c
===================================================================
--- linux-2.6.24-rc8-rt1.orig/arch/powerpc/kernel/time.c	2008-01-16 22:27:55.000000000 -0500
+++ linux-2.6.24-rc8-rt1/arch/powerpc/kernel/time.c	2008-01-16 22:28:20.000000000 -0500
@@ -751,7 +751,7 @@ static cycle_t rtc_read(void)
 	return (cycle_t)get_rtc();
 }
 
-static cycle_t timebase_read(void)
+static cycle_t notrace timebase_read(void)
 {
 	return (cycle_t)get_tb();
 }
