Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/powerpc/kernel/time.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.25.4-rt2/arch/powerpc/kernel/time.c
===================================================================
--- linux-2.6.25.4-rt2.orig/arch/powerpc/kernel/time.c	2008-05-19 16:55:08.000000000 -0400
+++ linux-2.6.25.4-rt2/arch/powerpc/kernel/time.c	2008-05-19 16:55:31.000000000 -0400
@@ -796,7 +796,7 @@ static cycle_t rtc_read(void)
 	return (cycle_t)get_rtc();
 }
 
-static cycle_t timebase_read(void)
+static cycle_t notrace timebase_read(void)
 {
 	return (cycle_t)get_tb();
 }
