Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86/kernel/process.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

Index: linux-2.6.26/arch/x86/kernel/process.c
===================================================================
--- linux-2.6.26.orig/arch/x86/kernel/process.c
+++ linux-2.6.26/arch/x86/kernel/process.c
@@ -107,7 +107,9 @@ static void mwait_idle(void)
 static void poll_idle(void)
 {
 	local_irq_enable();
-	cpu_relax();
+	do {
+		cpu_relax();
+	} while (!need_resched());
 }
 
 /*
