---
 drivers/block/paride/pseudo.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/drivers/block/paride/pseudo.h
===================================================================
--- linux-2.6.22.orig/drivers/block/paride/pseudo.h	2007-07-24 08:56:11.000000000 +0200
+++ linux-2.6.22/drivers/block/paride/pseudo.h	2007-07-24 08:57:31.000000000 +0200
@@ -43,7 +43,7 @@ static unsigned long ps_timeout;
 static int ps_tq_active = 0;
 static int ps_nice = 0;
 
-static DEFINE_SPINLOCK(ps_spinlock __attribute__((unused)));
+static __attribute__((unused)) DEFINE_SPINLOCK(ps_spinlock);
 
 static DECLARE_DELAYED_WORK(ps_tq, ps_tq_int);
 
