Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/time/tick-sched.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.26/kernel/time/tick-sched.c
===================================================================
--- linux-2.6.26.orig/kernel/time/tick-sched.c
+++ linux-2.6.26/kernel/time/tick-sched.c
@@ -225,6 +225,8 @@ void tick_nohz_stop_sched_tick(int inidl
 	if (need_resched())
 		goto end;
 
+
+#ifndef CONFIG_PREEMPT_RT
 	if (unlikely(local_softirq_pending())) {
 		static int ratelimit;
 
@@ -234,6 +236,7 @@ void tick_nohz_stop_sched_tick(int inidl
 			ratelimit++;
 		}
 	}
+#endif
 
 	ts->idle_calls++;
 	/* Read jiffies and the time when jiffies were updated last */
