Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/sched_cpupri.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.25.4-rt2/kernel/sched_cpupri.h
===================================================================
--- linux-2.6.25.4-rt2.orig/kernel/sched_cpupri.h	2008-05-19 16:55:17.000000000 -0400
+++ linux-2.6.25.4-rt2/kernel/sched_cpupri.h	2008-05-19 16:55:56.000000000 -0400
@@ -12,7 +12,7 @@
 /* values 2-101 are RT priorities 0-99 */
 
 struct cpupri_vec {
-	spinlock_t lock;
+	raw_spinlock_t lock;
 	int        count;
 	cpumask_t  mask;
 };
