---
 kernel/rcupreempt.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

Index: linux-2.6.22/kernel/rcupreempt.c
===================================================================
--- linux-2.6.22.orig/kernel/rcupreempt.c	2007-07-24 08:57:13.000000000 +0200
+++ linux-2.6.22/kernel/rcupreempt.c	2007-07-24 08:57:32.000000000 +0200
@@ -55,7 +55,7 @@
  */
 
 struct rcu_data {
-	spinlock_t	lock;
+	raw_spinlock_t	lock;
 	long		completed;	/* Number of last completed batch. */
 	struct rcu_head *nextlist;
 	struct rcu_head **nexttail;
@@ -68,12 +68,12 @@ struct rcu_data {
 #endif /* #ifdef CONFIG_RCU_TRACE */
 };
 struct rcu_ctrlblk {
-	spinlock_t	fliplock;
+	raw_spinlock_t	fliplock;
 	long		completed;	/* Number of last completed batch. */
 };
 static struct rcu_data rcu_data;
 static struct rcu_ctrlblk rcu_ctrlblk = {
-	.fliplock = SPIN_LOCK_UNLOCKED,
+	.fliplock = RAW_SPIN_LOCK_UNLOCKED(rcu_ctrlblk.fliplock),
 	.completed = 0,
 };
 static DEFINE_PER_CPU(atomic_t [2], rcu_flipctr) =
@@ -348,7 +348,7 @@ int rcu_needs_cpu(int cpu)
 	return !!rcu_data.waitlist || rcu_pending(cpu);
 }
 
-int rcu_pending(int cpu)
+int notrace rcu_pending(int cpu)
 {
 	return (rcu_data.donelist != NULL ||
 		rcu_data.waitlist != NULL ||
