Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>

  To fix the following boot time error by removing ack member added by
the rt patch.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processor 1 found.
Brought up 2 CPUs
------------[ cut here ]------------
kernel BUG at arch/powerpc/platforms/cell/interrupt.c:86!
pu 0x1: Vector: 700 (Program Check) at [c00000000fff3c80]
    pc: c000000000033f9c: .iic_eoi+0x58/0x64
    lr: c00000000009add8: .handle_percpu_irq+0xd4/0xf4
    sp: c00000000fff3f00
   msr: 9000000000021032
  current = 0xc000000000fee040
  paca    = 0xc000000000509e80
    pid   = 0, comm = swapper
kernel BUG at arch/powerpc/platforms/cell/interrupt.c:86!
enter ? for help
[link register   ] c00000000009add8 .handle_percpu_irq+0xd4/0xf4
[c00000000fff3f00] c00000000009ada8 .handle_percpu_irq+0xa4/0xf4 (unreliable)
[c00000000fff3f90] c000000000023bb8 .call_handle_irq+0x1c/0x2c
[c000000000ff7950] c00000000000c910 .do_IRQ+0xf8/0x1b8
[c000000000ff79f0] c000000000034f34 .cbe_system_reset_exception+0x74/0xb4
[c000000000ff7a70] c000000000022610 .system_reset_exception+0x40/0xe0
[c000000000ff7af0] c000000000003378 system_reset_common+0xf8/0x100
---
 arch/powerpc/platforms/cell/interrupt.c |    1 -
 1 file changed, 1 deletion(-)

Index: linux-2.6.25.4-rt4/arch/powerpc/platforms/cell/interrupt.c
===================================================================
--- linux-2.6.25.4-rt4.orig/arch/powerpc/platforms/cell/interrupt.c	2008-05-29 09:46:23.000000000 -0400
+++ linux-2.6.25.4-rt4/arch/powerpc/platforms/cell/interrupt.c	2008-05-29 09:46:23.000000000 -0400
@@ -90,7 +90,6 @@ static struct irq_chip iic_chip = {
 	.typename = " CELL-IIC ",
 	.mask = iic_mask,
 	.unmask = iic_unmask,
-	.ack = iic_eoi,
 	.eoi = iic_eoi,
 };
 
