Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86/kernel/i8259_64.c |    1 +
 arch/x86/kernel/time_64.c  |    3 ++-
 2 files changed, 3 insertions(+), 1 deletion(-)

Index: linux-2.6.24-rt1/arch/x86/kernel/i8259_64.c
===================================================================
--- linux-2.6.24-rt1.orig/arch/x86/kernel/i8259_64.c	2008-01-25 15:06:40.000000000 -0500
+++ linux-2.6.24-rt1/arch/x86/kernel/i8259_64.c	2008-01-25 15:07:27.000000000 -0500
@@ -397,6 +397,7 @@ device_initcall(i8259A_init_sysfs);
 
 static struct irqaction irq2 = {
 	.handler = no_action,
+	.flags = IRQF_NODELAY,
 	.mask = CPU_MASK_NONE,
 	.name = "cascade",
 };
Index: linux-2.6.24-rt1/arch/x86/kernel/time_64.c
===================================================================
--- linux-2.6.24-rt1.orig/arch/x86/kernel/time_64.c	2008-01-25 15:06:40.000000000 -0500
+++ linux-2.6.24-rt1/arch/x86/kernel/time_64.c	2008-01-25 15:07:27.000000000 -0500
@@ -259,7 +259,8 @@ static unsigned int __init tsc_calibrate
 
 static struct irqaction irq0 = {
 	.handler	= timer_event_interrupt,
-	.flags		= IRQF_DISABLED | IRQF_IRQPOLL | IRQF_NOBALANCING,
+	.flags		= IRQF_DISABLED | IRQF_IRQPOLL | IRQF_NOBALANCING |
+			  IRQF_NODELAY,
 	.mask		= CPU_MASK_NONE,
 	.name		= "timer"
 };
