Subject: Linux-RT 2.6.24-rc5-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 init/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.24-rc5-rt1/init/Kconfig
===================================================================
--- linux-2.6.24-rc5-rt1.orig/init/Kconfig
+++ linux-2.6.24-rc5-rt1/init/Kconfig
@@ -647,6 +647,7 @@ config SLAB
 
 config SLUB
 	bool "SLUB (Unqueued Allocator)"
+	depends on !PREEMPT_RT
 	help
 	   SLUB is a slab allocator that minimizes cache line usage
 	   instead of managing queues of cached objects (SLAB approach).
