Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 init/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.26.3-rt6/init/Kconfig
===================================================================
--- linux-2.6.26.3-rt6.orig/init/Kconfig	2008-09-05 00:23:21.000000000 -0400
+++ linux-2.6.26.3-rt6/init/Kconfig	2008-09-05 00:23:31.000000000 -0400
@@ -769,6 +769,7 @@ config SLAB
 
 config SLUB
 	bool "SLUB (Unqueued Allocator)"
+	depends on !PREEMPT_RT
 	help
 	   SLUB is a slab allocator that minimizes cache line usage
 	   instead of managing queues of cached objects (SLAB approach).
