---
 kernel/cpu.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.22/kernel/cpu.c
===================================================================
--- linux-2.6.22.orig/kernel/cpu.c	2007-07-24 08:56:09.000000000 +0200
+++ linux-2.6.22/kernel/cpu.c	2007-07-24 08:57:33.000000000 +0200
@@ -37,12 +37,14 @@ void lock_cpu_hotplug(void)
 	struct task_struct *tsk = current;
 
 	if (tsk == recursive) {
+#ifdef CONFIG_PREEMPT_RT
 		static int warnings = 10;
 		if (warnings) {
 			printk(KERN_ERR "Lukewarm IQ detected in hotplug locking\n");
 			WARN_ON(1);
 			warnings--;
 		}
+#endif
 		recursive_depth++;
 		return;
 	}
