---
 drivers/char/rtc.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.22/drivers/char/rtc.c
===================================================================
--- linux-2.6.22.orig/drivers/char/rtc.c	2007-07-24 08:57:12.000000000 +0200
+++ linux-2.6.22/drivers/char/rtc.c	2007-07-24 08:57:34.000000000 +0200
@@ -1333,7 +1333,9 @@ static void rtc_dropped_irq(unsigned lon
 
 	spin_unlock_irq(&rtc_lock);
 
+#ifndef CONFIG_PREEMPT_RT
 	printk(KERN_WARNING "rtc: lost some interrupts at %ldHz.\n", freq);
+#endif
 
 	/* Now we have new data */
 	wake_up_interruptible(&rtc_wait);
