Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 drivers/net/tulip/tulip_core.c |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.27-RT/drivers/net/tulip/tulip_core.c
===================================================================
--- linux-2.6.27-RT.orig/drivers/net/tulip/tulip_core.c
+++ linux-2.6.27-RT/drivers/net/tulip/tulip_core.c
@@ -1801,6 +1801,7 @@ static void __devexit tulip_remove_one (
 	pci_iounmap(pdev, tp->base_addr);
 	free_netdev (dev);
 	pci_release_regions (pdev);
+	pci_disable_device (pdev);
 	pci_set_drvdata (pdev, NULL);
 
 	/* pci_power_off (pdev, -1); */
