---
 drivers/net/loopback.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/drivers/net/loopback.c
===================================================================
--- linux-2.6.22.orig/drivers/net/loopback.c	2007-07-24 08:57:09.000000000 +0200
+++ linux-2.6.22/drivers/net/loopback.c	2007-07-24 08:57:34.000000000 +0200
@@ -159,7 +159,7 @@ static int loopback_xmit(struct sk_buff 
 	lb_stats->packets++;
 	put_cpu();
 
-	netif_rx(skb);
+	netif_rx_ni(skb);
 
 	return 0;
 }
