Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 drivers/char/rtc.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.24-rt1/drivers/char/rtc.c
===================================================================
--- linux-2.6.24-rt1.orig/drivers/char/rtc.c	2008-01-25 15:07:18.000000000 -0500
+++ linux-2.6.24-rt1/drivers/char/rtc.c	2008-01-25 15:08:00.000000000 -0500
@@ -1341,8 +1341,10 @@ static void rtc_dropped_irq(unsigned lon
 
 	spin_unlock_irq(&rtc_lock);
 
+#ifndef CONFIG_PREEMPT_RT
 	if (printk_ratelimit())
 		printk(KERN_WARNING "rtc: lost some interrupts at %ldHz.\n", freq);
+#endif
 
 	/* Now we have new data */
 	wake_up_interruptible(&rtc_wait);
