---
 init/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.22/init/Kconfig
===================================================================
--- linux-2.6.22.orig/init/Kconfig	2007-07-24 08:55:59.000000000 +0200
+++ linux-2.6.22/init/Kconfig	2007-07-24 08:57:44.000000000 +0200
@@ -568,6 +568,7 @@ config SLAB
 
 config SLUB
 	bool "SLUB (Unqueued Allocator)"
+	depends on !PREEMPT_RT
 	help
 	   SLUB is a slab allocator that minimizes cache line usage
 	   instead of managing queues of cached objects (SLAB approach).
