 drivers/ide/pci/cs5530.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

Index: linux-2.6.22/drivers/ide/pci/cs5530.c
===================================================================
--- linux-2.6.22.orig/drivers/ide/pci/cs5530.c	2007-07-24 08:56:19.000000000 +0200
+++ linux-2.6.22/drivers/ide/pci/cs5530.c	2007-07-24 08:57:23.000000000 +0200
@@ -227,8 +227,8 @@ static unsigned int __devinit init_chips
 		goto out;
 	}
 
-	spin_lock_irqsave(&ide_lock, flags);
-		/* all CPUs (there should only be one CPU with this chipset) */
+	/* Local CPU. ide_lock is acquired in do_ide_setup_pci_device. */
+	local_irq_save(flags);
 
 	/*
 	 * Enable BusMaster and MemoryWriteAndInvalidate for the cs5530:
@@ -280,7 +280,7 @@ static unsigned int __devinit init_chips
 	pci_write_config_byte(master_0, 0x42, 0x00);
 	pci_write_config_byte(master_0, 0x43, 0xc1);
 
-	spin_unlock_irqrestore(&ide_lock, flags);
+	local_irq_restore(flags);
 
 out:
 	pci_dev_put(master_0);
