Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 drivers/net/tulip/tulip_core.c |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.26.3-rt6/drivers/net/tulip/tulip_core.c
===================================================================
--- linux-2.6.26.3-rt6.orig/drivers/net/tulip/tulip_core.c	2008-09-05 00:21:55.000000000 -0400
+++ linux-2.6.26.3-rt6/drivers/net/tulip/tulip_core.c	2008-09-05 00:23:10.000000000 -0400
@@ -1799,6 +1799,7 @@ static void __devexit tulip_remove_one (
 	pci_iounmap(pdev, tp->base_addr);
 	free_netdev (dev);
 	pci_release_regions (pdev);
+	pci_disable_device (pdev);
 	pci_set_drvdata (pdev, NULL);
 
 	/* pci_power_off (pdev, -1); */
