---
 arch/arm/mach-shark/leds.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/arch/arm/mach-shark/leds.c
===================================================================
--- linux-2.6.22.orig/arch/arm/mach-shark/leds.c	2007-07-24 08:56:17.000000000 +0200
+++ linux-2.6.22/arch/arm/mach-shark/leds.c	2007-07-24 08:57:25.000000000 +0200
@@ -32,7 +32,7 @@ static char led_state;
 static short hw_led_state;
 static short saved_state;
 
-static DEFINE_SPINLOCK(leds_lock);
+static DEFINE_RAW_SPINLOCK(leds_lock);
 
 short sequoia_read(int addr) {
   outw(addr,0x24);
