Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86/kernel/process_32.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

Index: linux-2.6.24-rc2-rt1/arch/x86/kernel/process_32.c
===================================================================
--- linux-2.6.24-rc2-rt1.orig/arch/x86/kernel/process_32.c
+++ linux-2.6.24-rc2-rt1/arch/x86/kernel/process_32.c
@@ -134,7 +134,9 @@ EXPORT_SYMBOL(default_idle);
  */
 static void poll_idle (void)
 {
-	cpu_relax();
+	do {
+		cpu_relax();
+	} while (!need_resched() && !need_resched_delayed());
 }
 
 #ifdef CONFIG_HOTPLUG_CPU
