Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 init/main.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.27-RT/init/main.c
===================================================================
--- linux-2.6.27-RT.orig/init/main.c
+++ linux-2.6.27-RT/init/main.c
@@ -364,6 +364,9 @@ static int __init rdinit_setup(char *str
 {
 	unsigned int i;
 
+#ifdef CONFIG_PREEMPT_RT
+	ftrace_disable_daemon();
+#endif
 	ramdisk_execute_command = str;
 	/* See "auto" comment in init_setup */
 	for (i = 1; i < MAX_INIT_ARGS; i++)
