Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 init/Kconfig |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.25.4-rt2/init/Kconfig
===================================================================
--- linux-2.6.25.4-rt2.orig/init/Kconfig	2008-05-19 16:56:09.000000000 -0400
+++ linux-2.6.25.4-rt2/init/Kconfig	2008-05-19 16:56:16.000000000 -0400
@@ -727,6 +727,7 @@ config SLAB
 
 config SLUB
 	bool "SLUB (Unqueued Allocator)"
+	depends on !PREEMPT_RT
 	help
 	   SLUB is a slab allocator that minimizes cache line usage
 	   instead of managing queues of cached objects (SLAB approach).
