---
 kernel/time/timekeeping.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

Index: linux-2.6.22/kernel/time/timekeeping.c
===================================================================
--- linux-2.6.22.orig/kernel/time/timekeeping.c	2007-07-24 08:57:33.000000000 +0200
+++ linux-2.6.22/kernel/time/timekeeping.c	2007-07-24 08:57:33.000000000 +0200
@@ -366,9 +366,13 @@ static int timekeeping_suspend(struct sy
 {
 	unsigned long flags;
 
+	/*
+	 * Read the CMOS outside the raw xtime_lock:
+	 */
+	timekeeping_suspend_time = read_persistent_clock();
+
 	write_seqlock_irqsave(&xtime_lock, flags);
 	timekeeping_suspended = 1;
-	timekeeping_suspend_time = read_persistent_clock();
 	write_sequnlock_irqrestore(&xtime_lock, flags);
 
 	clockevents_notify(CLOCK_EVT_NOTIFY_SUSPEND, NULL);
