---
 include/asm-arm/arch-pxa/timex.h |    6 ++++++
 1 file changed, 6 insertions(+)

Index: linux-2.6.22/include/asm-arm/arch-pxa/timex.h
===================================================================
--- linux-2.6.22.orig/include/asm-arm/arch-pxa/timex.h	2007-07-24 08:56:17.000000000 +0200
+++ linux-2.6.22/include/asm-arm/arch-pxa/timex.h	2007-07-24 08:57:25.000000000 +0200
@@ -16,9 +16,15 @@
 #define CLOCK_TICK_RATE 3686400
 #elif defined(CONFIG_PXA27x)
 /* PXA27x timer base */
+#include <asm-arm/arch-pxa/hardware.h>
+#include <asm-arm/arch-pxa/pxa-regs.h>
 #ifdef CONFIG_MACH_MAINSTONE
 #define CLOCK_TICK_RATE 3249600
 #else
 #define CLOCK_TICK_RATE 3250000
 #endif
 #endif
+
+#define mach_read_cycles() OSCR
+#define mach_cycles_to_usecs(d) (((d) * ((1000000LL << 32) / CLOCK_TICK_RATE)) >> 32)
+#define mach_usecs_to_cycles(d) (((d) * (((long long)CLOCK_TICK_RATE << 32) / 1000000)) >> 32)
