Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 lib/radix-tree.c |   24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

Index: linux-2.6.26/lib/radix-tree.c
===================================================================
--- linux-2.6.26.orig/lib/radix-tree.c
+++ linux-2.6.26/lib/radix-tree.c
@@ -79,6 +79,26 @@ static unsigned long height_to_maxindex[
 #ifdef CONFIG_RADIX_TREE_CONCURRENT
 static struct lock_class_key radix_node_class[RADIX_TREE_MAX_PATH];
 #endif
+#ifdef CONFIG_DEBUG_LOCK_ALLOC
+static const char *radix_node_key_string[RADIX_TREE_MAX_PATH] = {
+	"radix-node-00",
+	"radix-node-01",
+	"radix-node-02",
+	"radix-node-03",
+	"radix-node-04",
+	"radix-node-05",
+	"radix-node-06",
+	"radix-node-07",
+	"radix-node-08",
+	"radix-node-09",
+	"radix-node-10",
+	"radix-node-11",
+	"radix-node-12",
+	"radix-node-13",
+	"radix-node-14",
+	"radix-node-15",
+};
+#endif
 
 #ifdef CONFIG_RADIX_TREE_OPTIMISTIC
 static DEFINE_PER_CPU(unsigned long[RADIX_TREE_MAX_PATH+1], optimistic_histogram);
@@ -282,7 +302,9 @@ radix_tree_node_alloc(struct radix_tree_
 	BUG_ON(radix_tree_is_indirect_ptr(ret));
 #ifdef CONFIG_RADIX_TREE_CONCURRENT
 	spin_lock_init(&ret->lock);
-	lockdep_set_class(&ret->lock, &radix_node_class[height]);
+	lockdep_set_class_and_name(&ret->lock,
+			&radix_node_class[height],
+			radix_node_key_string[height]);
 #endif
 	ret->height = height;
 	return ret;
