Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>

Hack fix to prevent a bunch of warnings about using smp_processor_id
in preemptable code. The real fix would be to pass down a cpu variable
and have an RT sleeping spinlock to protect it.

Signed-off-by: Steven Rostedt <srostedt@redhat.com>

---
 lib/radix-tree.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

Index: linux-2.6.25.4-rt4/lib/radix-tree.c
===================================================================
--- linux-2.6.25.4-rt4.orig/lib/radix-tree.c	2008-05-29 09:46:57.000000000 -0400
+++ linux-2.6.25.4-rt4/lib/radix-tree.c	2008-05-29 09:47:20.000000000 -0400
@@ -105,10 +105,20 @@ static DEFINE_PER_CPU(unsigned long[RADI
 
 static void optimistic_hit(unsigned long height)
 {
+	unsigned long flags;
+
 	if (height > RADIX_TREE_MAX_PATH)
 		height = RADIX_TREE_MAX_PATH;
 
+	/*
+	 * HACK:
+	 *  On PREEMPT_RT this protects the percpu var.
+	 *  We really need to fix this to pass in cpu var protected
+	 *  with a RT sleeping spinlock.
+	 */
+	local_irq_save(flags);
 	__get_cpu_var(optimistic_histogram)[height]++;
+	local_irq_restore(flags);
 }
 
 #ifdef CONFIG_PROC_FS
