Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
From: Steven Rostedt <srostedt@redhat.com>
Subject: lockdep: add +1 to radix tree array

The radix tree array was off by 1 used for lockdep annotation.

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 lib/radix-tree.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.26.3-rt6/lib/radix-tree.c
===================================================================
--- linux-2.6.26.3-rt6.orig/lib/radix-tree.c	2008-09-05 00:23:47.000000000 -0400
+++ linux-2.6.26.3-rt6/lib/radix-tree.c	2008-09-05 00:23:57.000000000 -0400
@@ -77,10 +77,10 @@ struct radix_tree_path {
 static unsigned long height_to_maxindex[RADIX_TREE_MAX_PATH + 1] __read_mostly;
 
 #ifdef CONFIG_RADIX_TREE_CONCURRENT
-static struct lock_class_key radix_node_class[RADIX_TREE_MAX_PATH];
+static struct lock_class_key radix_node_class[RADIX_TREE_MAX_PATH + 1];
 #endif
 #ifdef CONFIG_DEBUG_LOCK_ALLOC
-static const char *radix_node_key_string[RADIX_TREE_MAX_PATH] = {
+static const char *radix_node_key_string[RADIX_TREE_MAX_PATH + 1] = {
 	"radix-node-00",
 	"radix-node-01",
 	"radix-node-02",
