Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 kernel/sched.c |    1 -
 1 file changed, 1 deletion(-)

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -7700,7 +7700,6 @@ static void detach_destroy_domains(const
 
 	for_each_cpu_mask_nr(i, *cpu_map)
 		cpu_attach_domain(NULL, &def_root_domain, i);
-	synchronize_sched();
 	arch_destroy_sched_domains(cpu_map, &tmpmask);
 }
 
