 kernel/rcutorture.c |    4 ++++
 1 file changed, 4 insertions(+)

Index: linux-2.6.22/kernel/rcutorture.c
===================================================================
--- linux-2.6.22.orig/kernel/rcutorture.c	2007-07-24 08:56:29.000000000 +0200
+++ linux-2.6.22/kernel/rcutorture.c	2007-07-24 08:57:13.000000000 +0200
@@ -327,7 +327,11 @@ static void rcu_bh_torture_read_unlock(i
 
 static int rcu_bh_torture_completed(void)
 {
+#ifdef CONFIG_PREEMPT_RCU
+	return rcu_batches_completed();
+#else
 	return rcu_batches_completed_bh();
+#endif
 }
 
 static void rcu_bh_torture_deferred_free(struct rcu_torture *p)
