Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/Kconfig.preempt |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.24-rt1/kernel/Kconfig.preempt
===================================================================
--- linux-2.6.24-rt1.orig/kernel/Kconfig.preempt	2008-01-25 15:08:24.000000000 -0500
+++ linux-2.6.24-rt1/kernel/Kconfig.preempt	2008-01-25 15:08:24.000000000 -0500
@@ -160,6 +160,7 @@ endchoice
 config PREEMPT_RCU_BOOST
 	bool "Enable priority boosting of RCU read-side critical sections"
 	depends on PREEMPT_RCU
+	default y if PREEMPT_RT
 	help
 	  This option permits priority boosting of RCU read-side critical
 	  sections tat have been preempted and a RT process is waiting
