Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/Kconfig.preempt |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.25.2-rt1/kernel/Kconfig.preempt
===================================================================
--- linux-2.6.25.2-rt1.orig/kernel/Kconfig.preempt	2008-05-08 16:18:21.000000000 -0400
+++ linux-2.6.25.2-rt1/kernel/Kconfig.preempt	2008-05-08 16:18:22.000000000 -0400
@@ -69,6 +69,7 @@ config PREEMPT_RCU
 config PREEMPT_RCU_BOOST
 	bool "Enable priority boosting of RCU read-side critical sections"
 	depends on PREEMPT_RCU
+	default y if PREEMPT_RT
 	help
 	  This option permits priority boosting of RCU read-side critical
 	  sections tat have been preempted and a RT process is waiting
