Subject: Linux-RT 2.6.24-rc8-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/rcupreempt.c |    5 +++++
 1 file changed, 5 insertions(+)

Index: linux-2.6.23-rc8-rt1/kernel/rcupreempt.c
===================================================================
--- linux-2.6.23-rc8-rt1.orig/kernel/rcupreempt.c	2007-09-26 14:27:23.000000000 -0400
+++ linux-2.6.23-rc8-rt1/kernel/rcupreempt.c	2007-09-26 14:27:24.000000000 -0400
@@ -129,6 +129,9 @@ void __rcu_read_lock(void)
 			atomic_inc(current->rcu_flipctr2);
 			smp_mb__after_atomic_inc();  /* might optimize out... */
 		}
+	} else {
+		/* nesting limit: */
+		WARN_ON_ONCE(current->rcu_read_lock_nesting > 30);
 	}
 	local_irq_restore(oldirq);
 }
@@ -154,6 +157,8 @@ void __rcu_read_unlock(void)
 			atomic_dec(current->rcu_flipctr2);
 			current->rcu_flipctr2 = NULL;
 		}
+	} else {
+		WARN_ON_ONCE(current->rcu_read_lock_nesting < 0);
 	}
 
 	local_irq_restore(oldirq);
