Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
 kernel/rcutorture.c |    4 ++++
 1 file changed, 4 insertions(+)

Index: linux-2.6.23.1-rt11/kernel/rcutorture.c
===================================================================
--- linux-2.6.23.1-rt11.orig/kernel/rcutorture.c
+++ linux-2.6.23.1-rt11/kernel/rcutorture.c
@@ -390,7 +390,11 @@ static void rcu_bh_torture_read_unlock(i
 
 static int rcu_bh_torture_completed(void)
 {
+#ifdef CONFIG_PREEMPT_RCU
+	return rcu_batches_completed();
+#else
 	return rcu_batches_completed_bh();
+#endif
 }
 
 static void rcu_bh_torture_deferred_free(struct rcu_torture *p)
