Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/linux/Kbuild |    1 -
 1 file changed, 1 deletion(-)

Index: linux-2.6.24.3-rt3/include/linux/Kbuild
===================================================================
--- linux-2.6.24.3-rt3.orig/include/linux/Kbuild	2008-02-26 23:29:13.000000000 -0500
+++ linux-2.6.24.3-rt3/include/linux/Kbuild	2008-02-26 23:31:28.000000000 -0500
@@ -217,7 +217,6 @@ unifdef-y += i2o-dev.h
 unifdef-y += icmp.h
 unifdef-y += icmpv6.h
 unifdef-y += if_addr.h
-unifdef-y += if_addrlabel.h
 unifdef-y += if_arp.h
 unifdef-y += if_bridge.h
 unifdef-y += if_ec.h
