Subject: Linux-RT 2.6.24-rc5-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From: Remy Bohmer <linux@bohmer.net>
Subject: fix warning on dec_rt_tasks on UP

Signed-off-by: Remy Bohmer <linux@bohmer.net>

---
 kernel/sched_rt.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.24-rc5-rt1/kernel/sched_rt.c
===================================================================
--- linux-2.6.24-rc5-rt1.orig/kernel/sched_rt.c
+++ linux-2.6.24-rc5-rt1/kernel/sched_rt.c
@@ -86,8 +86,9 @@ static inline void inc_rt_tasks(struct t
 
 static inline void dec_rt_tasks(struct task_struct *p, struct rq *rq)
 {
+#ifdef CONFIG_SMP
 	int highest_prio = rq->rt.highest_prio;
-
+#endif
 	WARN_ON(!rt_task(p));
 	WARN_ON(!rq->rt.rt_nr_running);
 	rq->rt.rt_nr_running--;
