---
 net/ipv4/route.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/net/ipv4/route.c
===================================================================
--- linux-2.6.22.orig/net/ipv4/route.c	2007-07-24 08:56:24.000000000 +0200
+++ linux-2.6.22/net/ipv4/route.c	2007-07-24 08:57:18.000000000 +0200
@@ -238,7 +238,7 @@ static spinlock_t	*rt_hash_locks;
 			spin_lock_init(&rt_hash_locks[i]); \
 		}
 #else
-# define rt_hash_lock_addr(slot) NULL
+# define rt_hash_lock_addr(slot) ((spinlock_t *)NULL)
 # define rt_hash_lock_init()
 #endif
 
