Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: rt-mutex-cleanup.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 20 Jun 2008 12:20:09 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/rtmutex.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

Index: linux-2.6.26/kernel/rtmutex.c
===================================================================
--- linux-2.6.26.orig/kernel/rtmutex.c
+++ linux-2.6.26/kernel/rtmutex.c
@@ -125,9 +125,12 @@ static inline void mark_rt_rwlock_check(
 #endif
 
 #ifdef CONFIG_PREEMPT_RT
-#define task_is_reader(task) ((task) == RT_RW_READER)
+static inline int task_is_reader(struct task_struct *task)
+{
+	return task == RT_RW_READER;
+}
 #else
-#define task_is_reader(task) (0)
+static inline int task_is_reader(struct task_struct *task) { return 0; }
 #endif
 
 int pi_initialized;
