Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/arm/kernel/semaphore.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.25.2-rt1/arch/arm/kernel/semaphore.c
===================================================================
--- linux-2.6.25.2-rt1.orig/arch/arm/kernel/semaphore.c	2008-05-08 16:18:51.000000000 -0400
+++ linux-2.6.25.2-rt1/arch/arm/kernel/semaphore.c	2008-05-08 16:18:51.000000000 -0400
@@ -154,7 +154,7 @@ EXPORT_SYMBOL(__compat_down_interruptibl
  * single "cmpxchg" without failure cases,
  * but then it wouldn't work on a 386.
  */
- int __attribute_used__ __compat_down_trylock(struct compat_semaphore * sem)
+ int __attribute_used__ __sched __compat_down_trylock(struct compat_semaphore * sem)
 {
 	int sleepers;
 	unsigned long flags;
@@ -176,7 +176,7 @@ EXPORT_SYMBOL(__compat_down_interruptibl
 
 EXPORT_SYMBOL(__compat_down_trylock);
 
- int compat_sem_is_locked(struct compat_semaphore *sem)
+ int __sched compat_sem_is_locked(struct compat_semaphore *sem)
 {
 	return (int) atomic_read(&sem->count) < 0;
 }
