---
 net/ipv4/tcp.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.22/net/ipv4/tcp.c
===================================================================
--- linux-2.6.22.orig/net/ipv4/tcp.c	2007-07-24 08:57:11.000000000 +0200
+++ linux-2.6.22/net/ipv4/tcp.c	2007-07-24 08:57:18.000000000 +0200
@@ -1144,10 +1144,10 @@ int tcp_recvmsg(struct kiocb *iocb, stru
 	preempt_disable();
 	if ((len > sysctl_tcp_dma_copybreak) && !(flags & MSG_PEEK) &&
 	    !sysctl_tcp_low_latency && __get_cpu_var(softnet_data).net_dma) {
-		preempt_enable_no_resched();
+		preempt_enable();
 		tp->ucopy.pinned_list = dma_pin_iovec_pages(msg->msg_iov, len);
 	} else
-		preempt_enable_no_resched();
+		preempt_enable();
 #endif
 
 	do {
