Subject: Linux-RT 2.6.24-rc8-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 init/Makefile       |    2 +-
 scripts/mkcompile_h |    4 +++-
 2 files changed, 4 insertions(+), 2 deletions(-)

Index: linux-2.6.24-rc8-rt1/init/Makefile
===================================================================
--- linux-2.6.24-rc8-rt1.orig/init/Makefile	2008-01-16 22:27:20.000000000 -0500
+++ linux-2.6.24-rc8-rt1/init/Makefile	2008-01-16 22:29:10.000000000 -0500
@@ -30,4 +30,4 @@ $(obj)/version.o: include/linux/compile.
 include/linux/compile.h: FORCE
 	@echo '  CHK     $@'
 	$(Q)$(CONFIG_SHELL) $(srctree)/scripts/mkcompile_h $@ \
-	"$(UTS_MACHINE)" "$(CONFIG_SMP)" "$(CONFIG_PREEMPT)" "$(CC) $(KBUILD_CFLAGS)"
+	"$(UTS_MACHINE)" "$(CONFIG_SMP)" "$(CONFIG_PREEMPT)" "$(CONFIG_PREEMPT_RT)" "$(CC) $(KBUILD_CFLAGS)"
Index: linux-2.6.24-rc8-rt1/scripts/mkcompile_h
===================================================================
--- linux-2.6.24-rc8-rt1.orig/scripts/mkcompile_h	2008-01-16 22:27:20.000000000 -0500
+++ linux-2.6.24-rc8-rt1/scripts/mkcompile_h	2008-01-16 22:29:10.000000000 -0500
@@ -2,7 +2,8 @@ TARGET=$1
 ARCH=$2
 SMP=$3
 PREEMPT=$4
-CC=$5
+PREEMPT_RT=$5
+CC=$6
 
 # If compile.h exists already and we don't own autoconf.h
 # (i.e. we're not the same user who did make *config), don't
@@ -43,6 +44,7 @@ UTS_VERSION="#$VERSION"
 CONFIG_FLAGS=""
 if [ -n "$SMP" ] ; then CONFIG_FLAGS="SMP"; fi
 if [ -n "$PREEMPT" ] ; then CONFIG_FLAGS="$CONFIG_FLAGS PREEMPT"; fi
+if [ -n "$PREEMPT_RT" ] ; then CONFIG_FLAGS="$CONFIG_FLAGS RT"; fi
 UTS_VERSION="$UTS_VERSION $CONFIG_FLAGS $TIMESTAMP"
 
 # Truncate to maximum length
