---
 init/Makefile       |    2 +-
 scripts/mkcompile_h |    4 +++-
 2 files changed, 4 insertions(+), 2 deletions(-)

Index: linux-2.6.22/init/Makefile
===================================================================
--- linux-2.6.22.orig/init/Makefile	2007-07-24 08:55:59.000000000 +0200
+++ linux-2.6.22/init/Makefile	2007-07-24 08:57:43.000000000 +0200
@@ -30,4 +30,4 @@ $(obj)/version.o: include/linux/compile.
 include/linux/compile.h: FORCE
 	@echo '  CHK     $@'
 	$(Q)$(CONFIG_SHELL) $(srctree)/scripts/mkcompile_h $@ \
-	"$(UTS_MACHINE)" "$(CONFIG_SMP)" "$(CONFIG_PREEMPT)" "$(CC) $(CFLAGS)"
+	"$(UTS_MACHINE)" "$(CONFIG_SMP)" "$(CONFIG_PREEMPT)" "$(CONFIG_PREEMPT_RT)" "$(CC) $(CFLAGS)"
Index: linux-2.6.22/scripts/mkcompile_h
===================================================================
--- linux-2.6.22.orig/scripts/mkcompile_h	2007-07-24 08:55:59.000000000 +0200
+++ linux-2.6.22/scripts/mkcompile_h	2007-07-24 08:57:43.000000000 +0200
@@ -2,7 +2,8 @@ TARGET=$1
 ARCH=$2
 SMP=$3
 PREEMPT=$4
-CC=$5
+PREEMPT_RT=$5
+CC=$6
 
 # If compile.h exists already and we don't own autoconf.h
 # (i.e. we're not the same user who did make *config), don't
@@ -43,6 +44,7 @@ UTS_VERSION="#$VERSION"
 CONFIG_FLAGS=""
 if [ -n "$SMP" ] ; then CONFIG_FLAGS="SMP"; fi
 if [ -n "$PREEMPT" ] ; then CONFIG_FLAGS="$CONFIG_FLAGS PREEMPT"; fi
+if [ -n "$PREEMPT_RT" ] ; then CONFIG_FLAGS="$CONFIG_FLAGS RT"; fi
 UTS_VERSION="$UTS_VERSION $CONFIG_FLAGS $TIMESTAMP"
 
 # Truncate to maximum length
