Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: rtmutex-debug-fix.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 20 Jun 2008 12:27:50 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/rtmutex-debug.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-2.6.26/kernel/rtmutex-debug.c
===================================================================
--- linux-2.6.26.orig/kernel/rtmutex-debug.c
+++ linux-2.6.26/kernel/rtmutex-debug.c
@@ -59,6 +59,9 @@ void rt_mutex_debug_task_free(struct tas
 {
 	DEBUG_LOCKS_WARN_ON(!plist_head_empty(&task->pi_waiters));
 	DEBUG_LOCKS_WARN_ON(task->pi_blocked_on);
+#ifdef CONFIG_PREEMPT_RT
+	WARN_ON(task->reader_lock_count);
+#endif
 }
 
 /*
