Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From: Steven Rostedt <srostedt@redhat.com>
Subject: rwlock: fix torture test to handle non-rt


Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 kernel/rwlock_torture.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.26.3-rt6/kernel/rwlock_torture.c
===================================================================
--- linux-2.6.26.3-rt6.orig/kernel/rwlock_torture.c	2008-09-05 00:23:57.000000000 -0400
+++ linux-2.6.26.3-rt6/kernel/rwlock_torture.c	2008-09-05 00:24:00.000000000 -0400
@@ -570,7 +570,9 @@ static int run_test(unsigned long time)
 		run_two_locks(time, read);
 	}
 
+#ifdef CONFIG_PREEMPT_RT
 	WARN_ON_ONCE(current->reader_lock_count);
+#endif
 
 	return ret;
 }
