Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
---
 kernel/rtmutex.c |   19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

Index: linux-2.6.26.3-rt6/kernel/rtmutex.c
===================================================================
--- linux-2.6.26.3-rt6.orig/kernel/rtmutex.c	2008-09-05 00:23:56.000000000 -0400
+++ linux-2.6.26.3-rt6/kernel/rtmutex.c	2008-09-05 00:23:58.000000000 -0400
@@ -1916,7 +1916,6 @@ rt_write_slowunlock(struct rw_mutex *rwm
 	if (!rt_mutex_has_waiters(mutex))
 		goto out;
 
-	spin_lock(&pendowner->pi_lock);
 	/*
 	 * Wake up all readers.
 	 * This gets a bit more complex. More than one reader can't
@@ -1931,13 +1930,17 @@ rt_write_slowunlock(struct rw_mutex *rwm
 	while (waiter && !waiter->write_lock) {
 		struct task_struct *reader = waiter->task;
 
+		spin_lock(&pendowner->pi_lock);
 		plist_del(&waiter->list_entry, &mutex->wait_list);
 
 		/* nop if not on a list */
 		plist_del(&waiter->pi_list_entry, &pendowner->pi_waiters);
+		spin_unlock(&pendowner->pi_lock);
 
+		spin_lock(&reader->pi_lock);
 		waiter->task = NULL;
 		reader->pi_blocked_on = NULL;
+		spin_unlock(&reader->pi_lock);
 
 		if (savestate)
 			wake_up_process_mutex(reader);
@@ -1955,6 +1958,8 @@ rt_write_slowunlock(struct rw_mutex *rwm
 		struct rt_mutex_waiter *next;
 
 		next = rt_mutex_top_waiter(mutex);
+
+		spin_lock(&pendowner->pi_lock);
 		/* delete incase we didn't go through the loop */
 		plist_del(&next->pi_list_entry, &pendowner->pi_waiters);
 
@@ -1962,13 +1967,12 @@ rt_write_slowunlock(struct rw_mutex *rwm
 		if (next->write_lock)
 			/* add back in as top waiter */
 			plist_add(&next->pi_list_entry, &pendowner->pi_waiters);
+		spin_unlock(&pendowner->pi_lock);
 
 		rwm->prio = next->task->prio;
 	} else
 		rwm->prio = MAX_PRIO;
 
-	spin_unlock(&pendowner->pi_lock);
-
  out:
 
 	spin_unlock_irqrestore(&mutex->wait_lock, flags);
@@ -2048,18 +2052,21 @@ rt_mutex_downgrade_write(struct rw_mutex
 	 * waiting, until we hit the reader limit, or a writer.
 	 */
 
-	spin_lock(&current->pi_lock);
 	waiter = rt_mutex_top_waiter(mutex);
 	while (waiter && !waiter->write_lock) {
 		struct task_struct *reader = waiter->task;
 
+		spin_lock(&current->pi_lock);
 		plist_del(&waiter->list_entry, &mutex->wait_list);
 
 		/* nop if not on a list */
 		plist_del(&waiter->pi_list_entry, &current->pi_waiters);
+		spin_unlock(&current->pi_lock);
 
+		spin_lock(&reader->pi_lock);
 		waiter->task = NULL;
 		reader->pi_blocked_on = NULL;
+		spin_unlock(&reader->pi_lock);
 
 		/* downgrade is only for mutexes */
 		wake_up_process(reader);
@@ -2079,14 +2086,14 @@ rt_mutex_downgrade_write(struct rw_mutex
 		/* setup this mutex prio for read */
 		rwm->prio = next->task->prio;
 
+		spin_lock(&current->pi_lock);
 		/* delete incase we didn't go through the loop */
 		plist_del(&next->pi_list_entry, &current->pi_waiters);
+		spin_unlock(&current->pi_lock);
 		/* No need to add back since readers don't have PI waiters */
 	} else
 		rwm->prio = MAX_PRIO;
 
-	spin_unlock(&current->pi_lock);
-
 	rt_mutex_set_owner(mutex, RT_RW_READER, 0);
 
 	spin_unlock_irqrestore(&mutex->wait_lock, flags);
