---
 fs/proc/generic.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.22/fs/proc/generic.c
===================================================================
--- linux-2.6.22.orig/fs/proc/generic.c	2007-07-24 08:57:36.000000000 +0200
+++ linux-2.6.22/fs/proc/generic.c	2007-07-24 08:57:36.000000000 +0200
@@ -564,7 +564,6 @@ static void proc_kill_inodes(struct proc
 	/*
 	 * Actually it's a partial revoke().
 	 */
-	filevec_add_drain_all();
 	lock_list_for_each_entry(filp, &sb->s_files, f_u.fu_llist) {
 		struct dentry * dentry = filp->f_path.dentry;
 		struct inode * inode;
@@ -725,6 +724,8 @@ void remove_proc_entry(const char *name,
 		goto out;
 	len = strlen(fn);
 
+	filevec_add_drain_all();
+
 	spin_lock(&proc_subdir_lock);
 	for (p = &parent->subdir; *p; p=&(*p)->next ) {
 		if (!proc_match(len, fn, *p))
