---
 fs/file_table.c |   21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

Index: linux-2.6.22/fs/file_table.c
===================================================================
--- linux-2.6.22.orig/fs/file_table.c	2007-07-24 08:57:36.000000000 +0200
+++ linux-2.6.22/fs/file_table.c	2007-07-24 08:57:37.000000000 +0200
@@ -332,6 +332,22 @@ static void __filevec_add(struct filevec
 	filevec_reinit(fvec);
 }
 
+/*
+ * Flush files per-CPU workqueue:
+ */
+static struct workqueue_struct *flush_files_workqueue;
+
+int __init flush_files_init(void)
+{
+        flush_files_workqueue = create_workqueue("flush_filesd");
+        if (!flush_files_workqueue)
+                panic("Failed to create flush_filesd\n");
+
+	return 0;
+}
+
+__initcall(flush_files_init);
+
 static void filevec_add_drain(void)
 {
 	int cpu;
@@ -341,14 +357,15 @@ static void filevec_add_drain(void)
 	put_cpu_var_locked(sb_fvec, cpu);
 }
 
-static void filevec_add_drain_per_cpu(void *dummy)
+static void filevec_add_drain_per_cpu(struct work_struct *none)
 {
 	filevec_add_drain();
 }
 
 int filevec_add_drain_all(void)
 {
-	return schedule_on_each_cpu(filevec_add_drain_per_cpu, NULL);
+	return schedule_on_each_cpu_wq(flush_files_workqueue,
+				       filevec_add_drain_per_cpu);
 }
 EXPORT_SYMBOL_GPL(filevec_add_drain_all);
 
