Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/sched.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.26.3-rt6/kernel/sched.c
===================================================================
--- linux-2.6.26.3-rt6.orig/kernel/sched.c	2008-09-05 00:23:52.000000000 -0400
+++ linux-2.6.26.3-rt6/kernel/sched.c	2008-09-05 00:23:53.000000000 -0400
@@ -1392,7 +1392,7 @@ void wake_up_idle_cpu(int cpu)
 {
 	struct rq *rq = cpu_rq(cpu);
 
-	if (cpu == smp_processor_id())
+	if (cpu == raw_smp_processor_id())
 		return;
 
 	/*
