Subject: Linux-RT 2.6.24-rc5-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/workqueue.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.24-rc5-rt1/kernel/workqueue.c
===================================================================
--- linux-2.6.24-rc5-rt1.orig/kernel/workqueue.c
+++ linux-2.6.24-rc5-rt1/kernel/workqueue.c
@@ -643,7 +643,7 @@ int schedule_on_each_cpu(void (*func)(vo
 		}
 	}
 
-	preempt_disable();		/* CPU hotplug */
+	lock_cpu_hotplug();
 	for_each_online_cpu(cpu) {
 		struct schedule_on_each_cpu_work *work;
 
@@ -656,7 +656,7 @@ int schedule_on_each_cpu(void (*func)(vo
 		set_bit(WORK_STRUCT_PENDING, work_data_bits(&work->work));
 		__queue_work(per_cpu_ptr(keventd_wq->cpu_wq, cpu), &work->work);
 	}
-	preempt_enable();
+	unlock_cpu_hotplug();
 
 out:
 	for_each_possible_cpu(cpu) {
