Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
Subject: sched-fix-dequeued-race.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Thu, 24 Jul 2008 16:14:31 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/sched.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.27-RT/kernel/sched.c
===================================================================
--- linux-2.6.27-RT.orig/kernel/sched.c
+++ linux-2.6.27-RT/kernel/sched.c
@@ -4464,7 +4464,8 @@ void scheduler_tick(void)
 	spin_lock(&rq->lock);
 	update_rq_clock(rq);
 	update_cpu_load(rq);
-	curr->sched_class->task_tick(rq, curr, 0);
+	if (curr != rq->idle && curr->se.on_rq)
+		curr->sched_class->task_tick(rq, curr, 0);
 	spin_unlock(&rq->lock);
 
 #ifdef CONFIG_SMP
