Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/asm-x86/semaphore_32.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6-rt-test/include/asm-x86/semaphore_32.h
===================================================================
--- linux-2.6-rt-test.orig/include/asm-x86/semaphore_32.h	2008-05-07 11:37:38.000000000 -0400
+++ linux-2.6-rt-test/include/asm-x86/semaphore_32.h	2008-05-07 11:44:44.000000000 -0400
@@ -179,7 +179,7 @@ static inline void compat_up(struct comp
 		:"memory","ax");
 }
 
-extern int FASTCALL(compat_sem_is_locked(struct compat_semaphore *sem));
+extern int compat_sem_is_locked(struct compat_semaphore *sem);
 
 #define compat_sema_count(sem) atomic_read(&(sem)->count)
 
