Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86/kernel/nmi.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.26-RT/arch/x86/kernel/nmi.c
===================================================================
--- linux-2.6.26-RT.orig/arch/x86/kernel/nmi.c
+++ linux-2.6.26-RT/arch/x86/kernel/nmi.c
@@ -594,8 +594,10 @@ void smp_send_nmi_allbutself(void)
 {
 #ifdef CONFIG_SMP
 	cpumask_t mask = cpu_online_map;
+	preempt_disable();
 	cpu_clear(safe_smp_processor_id(), mask);
 	if (!cpus_empty(mask))
 		send_IPI_mask(mask, NMI_VECTOR);
+	preempt_enable();
 #endif
 }
