Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
 drivers/serial/8250.c |   15 +++++----------
 1 file changed, 5 insertions(+), 10 deletions(-)

Index: linux-2.6.24-RT/drivers/serial/8250.c
===================================================================
--- linux-2.6.24-RT.orig/drivers/serial/8250.c	2008-01-26 03:06:08.317212268 -0800
+++ linux-2.6.24-RT/drivers/serial/8250.c	2008-01-26 03:10:44.099954075 -0800
@@ -2501,15 +2501,11 @@ serial8250_console_write(struct console 
 	int locked = 1;
 
 	touch_nmi_watchdog();
-
-	local_irq_save(flags);
-	if (up->port.sysrq) {
-		/* serial8250_handle_port() already took the lock */
-		locked = 0;
-	} else if (oops_in_progress || KDB_8250()) {
-		locked = spin_trylock(&up->port.lock);
-	} else
-		spin_lock(&up->port.lock);
+	
+	if (up->port.sysrq || oops_in_progress || KDB_8250())
+		locked = spin_trylock_irqsave(&up->port.lock, flags);
+	else
+		spin_lock_irqsave(&up->port.lock, flags);
 
 	/*
 	 *	First save the IER then disable the interrupts
@@ -2541,8 +2537,7 @@ serial8250_console_write(struct console 
 		check_modem_status(up);
 
 	if (locked)
-		spin_unlock(&up->port.lock);
-	local_irq_restore(flags);
+		spin_unlock_irqrestore(&up->port.lock, flags);
 }
 
 static int __init serial8250_console_setup(struct console *co, char *options)
