Subject: Linux-RT 2.6.24-rc8-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/latency_trace.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24-rc8-rt1/kernel/latency_trace.c
===================================================================
--- linux-2.6.24-rc8-rt1.orig/kernel/latency_trace.c	2008-01-16 22:28:14.000000000 -0500
+++ linux-2.6.24-rc8-rt1/kernel/latency_trace.c	2008-01-16 22:28:14.000000000 -0500
@@ -192,7 +192,7 @@ static int report_latency(cycle_t delta)
 /*
  * Number of per-CPU trace entries:
  */
-#define MAX_TRACE (65536UL*16UL)
+#define MAX_TRACE (65536UL)
 
 #define CMDLINE_BYTES 16
 
