---
 kernel/latency_trace.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/kernel/latency_trace.c
===================================================================
--- linux-2.6.22.orig/kernel/latency_trace.c	2007-07-24 08:57:07.000000000 +0200
+++ linux-2.6.22/kernel/latency_trace.c	2007-07-24 08:57:07.000000000 +0200
@@ -190,7 +190,7 @@ static int report_latency(cycle_t delta)
 /*
  * Number of per-CPU trace entries:
  */
-#define MAX_TRACE (65536UL*16UL)
+#define MAX_TRACE (65536UL)
 
 #define CMDLINE_BYTES 16
 
