---
 drivers/pci/pcie/aer/aerdrv.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.22/drivers/pci/pcie/aer/aerdrv.c
===================================================================
--- linux-2.6.22.orig/drivers/pci/pcie/aer/aerdrv.c	2007-07-24 08:56:40.000000000 +0200
+++ linux-2.6.22/drivers/pci/pcie/aer/aerdrv.c	2007-07-24 08:57:00.000000000 +0200
@@ -157,7 +157,7 @@ static struct aer_rpc* aer_alloc_rpc(str
 	 * Initialize Root lock access, e_lock, to Root Error Status Reg,
 	 * Root Error ID Reg, and Root error producer/consumer index.
 	 */
-	rpc->e_lock = SPIN_LOCK_UNLOCKED;
+	spin_lock_init(&rpc->e_lock);
 
 	rpc->rpd = dev;
 	INIT_WORK(&rpc->dpc_handler, aer_isr);
