Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From: Steven Rostedt <srostedt@redhat.com>
Subject: rt: dont disable irqs in usb

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 drivers/usb/core/hcd.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.26.3-rt6/drivers/usb/core/hcd.c
===================================================================
--- linux-2.6.26.3-rt6.orig/drivers/usb/core/hcd.c	2008-09-05 00:21:42.000000000 -0400
+++ linux-2.6.26.3-rt6/drivers/usb/core/hcd.c	2008-09-05 00:24:00.000000000 -0400
@@ -1700,7 +1700,7 @@ irqreturn_t usb_hcd_irq (int irq, void *
 	 * when the first handler doesn't use it.  So let's just
 	 * assume it's never used.
 	 */
-	local_irq_save(flags);
+	local_irq_save_nort(flags);
 
 	if (unlikely(hcd->state == HC_STATE_HALT ||
 		     !test_bit(HCD_FLAG_HW_ACCESSIBLE, &hcd->flags))) {
@@ -1715,7 +1715,7 @@ irqreturn_t usb_hcd_irq (int irq, void *
 		rc = IRQ_HANDLED;
 	}
 
-	local_irq_restore(flags);
+	local_irq_restore_nort(flags);
 	return rc;
 }
 
