Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/latency_trace.c |   36 ++++++++++++++++++++----------------
 1 file changed, 20 insertions(+), 16 deletions(-)

Index: linux-2.6.24-rt1/kernel/latency_trace.c
===================================================================
--- linux-2.6.24-rt1.orig/kernel/latency_trace.c	2008-01-25 15:07:15.000000000 -0500
+++ linux-2.6.24-rt1/kernel/latency_trace.c	2008-01-25 15:07:15.000000000 -0500
@@ -880,29 +880,33 @@ static void notrace print_name(struct se
 	 * Special trace values:
 	 */
 	if (((long)eip < 100000L) && ((long)eip > -100000L)) {
-		seq_printf(m, "(%5ld)", eip);
+		seq_printf(m, "<%ld>", eip);
 		return;
 	}
 	sym_name = kallsyms_lookup(eip, &size, &offset, &modname, namebuf);
 	if (sym_name)
-		seq_puts(m, sym_name);
+		seq_printf(m, "%s+%#lx/%#lx",
+			   sym_name, offset, size);
 	else
 		seq_printf(m, "<%08lx>", eip);
 }
 
-static void notrace print_name_offset(struct seq_file *m, unsigned long eip)
+static void notrace print_name_eip(struct seq_file *m, unsigned long eip)
 {
 	char namebuf[KSYM_NAME_LEN+1];
 	unsigned long size, offset;
 	const char *sym_name;
 	char *modname;
 
-	sym_name = kallsyms_lookup(eip, &size, &offset, &modname, namebuf);
-	if (sym_name)
-		seq_printf(m, "%s+%#lx/%#lx <%08lx>",
-					sym_name, offset, size, eip);
-	else
-		seq_printf(m, "<%08lx>", eip);
+	if (eip) {
+		sym_name = kallsyms_lookup(eip, &size, &offset, &modname, namebuf);
+		if (sym_name)
+			seq_printf(m, "%s+%#lx/%#lx <%08lx>",
+				   sym_name, offset, size, eip);
+		else
+			seq_printf(m, "<%08lx>", eip);
+	} else
+		seq_printf(m, "0");
 }
 
 static unsigned long out_sequence = -1;
@@ -1267,9 +1271,9 @@ static void * notrace l_start(struct seq
 		seq_puts(m, "    -----------------\n");
 		if (trace_user_triggered) {
 			seq_puts(m, " => started at: ");
-			print_name_offset(m, tr->critical_start);
+			print_name_eip(m, tr->critical_start);
 			seq_puts(m, "\n => ended at:   ");
-			print_name_offset(m, tr->critical_end);
+			print_name_eip(m, tr->critical_end);
 			seq_puts(m, "\n");
 		}
 		seq_puts(m, "\n");
@@ -1379,9 +1383,9 @@ static int notrace l_show_fn(struct seq_
 			entry->preempt_count, trace_idx,
 			entry->timestamp, abs_usecs/1000,
 			abs_usecs % 1000, rel_usecs/1000, rel_usecs % 1000);
-		print_name_offset(m, entry->u.fn.eip);
+		print_name_eip(m, entry->u.fn.eip);
 		seq_puts(m, " (");
-		print_name_offset(m, entry->u.fn.parent_eip);
+		print_name_eip(m, entry->u.fn.parent_eip);
 		seq_puts(m, ")\n");
 	} else {
 		print_generic(m, entry);
@@ -1406,7 +1410,7 @@ static int notrace l_show_special(struct
 	print_generic(m, entry);
 	print_timestamp(m, abs_usecs, rel_usecs);
 	if (trace_verbose)
-		print_name_offset(m, entry->u.special.eip);
+		print_name_eip(m, entry->u.special.eip);
 	else
 		print_name(m, entry->u.special.eip);
 
@@ -1450,7 +1454,7 @@ l_show_special_pid(struct seq_file *m, u
 	print_generic(m, entry);
 	print_timestamp(m, abs_usecs, rel_usecs);
 	if (trace_verbose)
-		print_name_offset(m, entry->u.special.eip);
+		print_name_eip(m, entry->u.special.eip);
 	else
 		print_name(m, entry->u.special.eip);
 	seq_printf(m, " <%.8s-%d> (%ld %ld)\n",
@@ -1473,7 +1477,7 @@ l_show_special_sym(struct seq_file *m, u
 	print_generic(m, entry);
 	print_timestamp(m, abs_usecs, rel_usecs);
 	if (trace_verbose)
-		print_name_offset(m, entry->u.special.eip);
+		print_name_eip(m, entry->u.special.eip);
 	else
 		print_name(m, entry->u.special.eip);
 
