Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/arm/kernel/irq.c |    4 ++++
 1 file changed, 4 insertions(+)

Index: linux-2.6.26/arch/arm/kernel/irq.c
===================================================================
--- linux-2.6.26.orig/arch/arm/kernel/irq.c
+++ linux-2.6.26/arch/arm/kernel/irq.c
@@ -37,6 +37,8 @@
 #include <linux/kallsyms.h>
 #include <linux/proc_fs.h>
 
+#include <linux/ftrace.h>
+
 #include <asm/system.h>
 #include <asm/mach/time.h>
 
@@ -113,6 +115,8 @@ asmlinkage void __exception asm_do_IRQ(u
 	struct pt_regs *old_regs = set_irq_regs(regs);
 	struct irq_desc *desc = irq_desc + irq;
 
+	ftrace_event_irq(irq, user_mode(regs), instruction_pointer(regs));
+
 	/*
 	 * Some hardware gives randomly wrong interrupts.  Rather
 	 * than crashing, do something sensible.
