Subject: Linux-RT 2.6.24-rc5-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 kernel/latency_trace.c |    6 ------
 1 file changed, 6 deletions(-)

Index: linux-2.6.24-rc5-rt1/kernel/latency_trace.c
===================================================================
--- linux-2.6.24-rc5-rt1.orig/kernel/latency_trace.c
+++ linux-2.6.24-rc5-rt1/kernel/latency_trace.c
@@ -38,12 +38,6 @@
 
 int trace_use_raw_cycles = 0;
 
-#define __raw_spinlock_t raw_spinlock_t
-#define need_resched_delayed() 0
-
-#define __raw_spinlock_t raw_spinlock_t
-#define need_resched_delayed() 0
-
 #ifdef CONFIG_EVENT_TRACE
 /*
  * Convert raw cycles to usecs.
