Subject: 2.6.22.1-rt[78] wait.h extern declarations update

Synchronize extern sleep_on declarations with changes 
to sched.c introduced between 2.6.22.1-rt6 and 2.6.22.1-rt7

signed-off-by: Sven-Thorsten Dietrich <sdietrich@novell.com>

Index: linux-2.6.22/include/linux/wait.h
===================================================================
--- linux-2.6.22.orig/include/linux/wait.h
+++ linux-2.6.22/include/linux/wait.h
@@ -369,12 +369,10 @@ static inline void remove_wait_queue_loc
  * They are racy.  DO NOT use them, use the wait_event* interfaces above.  
  * We plan to remove these interfaces during 2.7.
  */
-extern void FASTCALL(sleep_on(wait_queue_head_t *q));
-extern long FASTCALL(sleep_on_timeout(wait_queue_head_t *q,
-				      signed long timeout));
-extern void FASTCALL(interruptible_sleep_on(wait_queue_head_t *q));
-extern long FASTCALL(interruptible_sleep_on_timeout(wait_queue_head_t *q,
-						    signed long timeout));
+extern void sleep_on(wait_queue_head_t *q);
+extern long sleep_on_timeout(wait_queue_head_t *q, long timeout);
+extern void interruptible_sleep_on(wait_queue_head_t *q);
+extern long interruptible_sleep_on_timeout(wait_queue_head_t *q, long timeout);
 
 /*
  * Waitqueues which are removed from the waitqueue_head at wakeup time
