Subject: Linux-RT 2.6.24-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: add -rt extra-version
From: Ingo Molnar <mingo@elte.hu>

add -rt extra-version.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Steven Rostedt <rostedt@goodmis.org>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24-rt1/Makefile
===================================================================
--- linux-2.6.24-rt1.orig/Makefile	2008-01-25 15:06:52.000000000 -0500
+++ linux-2.6.24-rt1/Makefile	2008-01-25 15:08:36.000000000 -0500
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 24
-EXTRAVERSION =
+EXTRAVERSION = -rt1
 NAME = Arr Matey! A Hairy Bilge Rat!
 
 # *DOCUMENTATION*
