Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: add -rt extra-version
From: Ingo Molnar <mingo@elte.hu>

add -rt extra-version.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Steven Rostedt <rostedt@goodmis.org>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.25.4-rt2/Makefile
===================================================================
--- linux-2.6.25.4-rt2.orig/Makefile	2008-05-19 16:55:17.000000000 -0400
+++ linux-2.6.25.4-rt2/Makefile	2008-05-19 16:56:38.000000000 -0400
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 25
-EXTRAVERSION = .4
+EXTRAVERSION = .4-rt2
 NAME = Funky Weasel is Jiggy wit it
 
 # *DOCUMENTATION*
