Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
Subject: Linux-RT 2.6.26-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From: Steven Rostedt <srostedt@redhat.com>
Subject: remove warn on for scatterlist in preempt rt

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 lib/scatterlist.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.26/lib/scatterlist.c
===================================================================
--- linux-2.6.26.orig/lib/scatterlist.c
+++ linux-2.6.26/lib/scatterlist.c
@@ -313,7 +313,7 @@ static size_t sg_copy_buffer(struct scat
 	size_t buf_off = 0;
 	int i;
 
-	WARN_ON(!irqs_disabled());
+	WARN_ON_NONRT(!irqs_disabled());
 
 	for_each_sg(sgl, sg, nents, i) {
 		struct page *page;
