Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 include/linux/spinlock.h |    7 +++++++
 1 file changed, 7 insertions(+)

Index: linux-2.6.25.4-rt2/include/linux/spinlock.h
===================================================================
--- linux-2.6.25.4-rt2.orig/include/linux/spinlock.h	2008-05-19 16:55:09.000000000 -0400
+++ linux-2.6.25.4-rt2/include/linux/spinlock.h	2008-05-19 16:55:27.000000000 -0400
@@ -295,6 +295,13 @@ do {						\
 	1 : ({ local_irq_restore(flags); 0; }); \
 })
 
+#define write_trylock_irqsave(lock, flags) \
+({ \
+	local_irq_save(flags);				\
+	write_trylock(lock) ?				\
+		1 : ({ local_irq_restore(flags); 0; });	\
+})
+
 /*
  * Locks two spinlocks l1 and l2.
  * l1_first indicates if spinlock l1 should be taken first.
