---
 arch/x86_64/kernel/smpboot.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.22/arch/x86_64/kernel/smpboot.c
===================================================================
--- linux-2.6.22.orig/arch/x86_64/kernel/smpboot.c	2007-07-24 08:56:57.000000000 +0200
+++ linux-2.6.22/arch/x86_64/kernel/smpboot.c	2007-07-24 08:57:03.000000000 +0200
@@ -331,8 +331,8 @@ void __cpuinit start_secondary(void)
 	barrier();
 
 	/*
-  	 * Check TSC sync first:
- 	 */
+	 * Check TSC sync first:
+	 */
 	check_tsc_sync_target();
 
 	Dprintk("cpu %d: setting up apic clock\n", smp_processor_id()); 	
