Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 arch/x86_64/kernel/smpboot.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.23-rc4/arch/x86_64/kernel/smpboot.c
===================================================================
--- linux-2.6.23-rc4.orig/arch/x86_64/kernel/smpboot.c	2007-08-31 21:51:22.000000000 +0200
+++ linux-2.6.23-rc4/arch/x86_64/kernel/smpboot.c	2007-08-31 21:51:26.000000000 +0200
@@ -331,8 +331,8 @@ void __cpuinit start_secondary(void)
 	barrier();
 
 	/*
-  	 * Check TSC sync first:
- 	 */
+	 * Check TSC sync first:
+	 */
 	check_tsc_sync_target();
 
 	Dprintk("cpu %d: setting up apic clock\n", smp_processor_id()); 	
