---
 arch/x86_64/kernel/tsc_sync.c |    4 ++++
 1 file changed, 4 insertions(+)

Index: linux-2.6.22/arch/x86_64/kernel/tsc_sync.c
===================================================================
--- linux-2.6.22.orig/arch/x86_64/kernel/tsc_sync.c	2007-07-24 08:56:40.000000000 +0200
+++ linux-2.6.22/arch/x86_64/kernel/tsc_sync.c	2007-07-24 08:57:01.000000000 +0200
@@ -97,6 +97,7 @@ static __cpuinit void check_tsc_warp(voi
  */
 void __cpuinit check_tsc_sync_source(int cpu)
 {
+	unsigned long flags;
 	int cpus = 2;
 
 	/*
@@ -117,8 +118,11 @@ void __cpuinit check_tsc_sync_source(int
 	/*
 	 * Wait for the target to arrive:
 	 */
+	local_save_flags(flags);
+	local_irq_enable();
 	while (atomic_read(&start_count) != cpus-1)
 		cpu_relax();
+	local_irq_restore(flags);
 	/*
 	 * Trigger the target to continue into the measurement too:
 	 */
