From ffe2045114ce7148e0bed9dec77788983329bd09 Mon Sep 17 00:00:00 2001
From: Al Viro <viro@ZenIV.linux.org.uk>
Date: Thu, 30 Oct 2014 05:42:09 +0000
Subject: [PATCH 5/5] fix a leak in bch_cached_dev_run()
Patch-mainline: Submitted
References: bnc#910440

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/md/bcache/super.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/md/bcache/super.c
+++ b/drivers/md/bcache/super.c
@@ -880,8 +880,11 @@ void bch_cached_dev_run(struct cached_de
 	buf[SB_LABEL_SIZE] = '\0';
 	env[2] = kasprintf(GFP_KERNEL, "CACHED_LABEL=%s", buf);
 
-	if (atomic_xchg(&dc->running, 1))
+	if (atomic_xchg(&dc->running, 1)) {
+		kfree(env[1]);
+		kfree(env[2]);
 		return;
+	}
 
 	if (!d->c &&
 	    BDEV_STATE(&dc->sb) != BDEV_STATE_NONE) {
