From: Takashi Iwai <tiwai@suse.de>
Subject: Remove invalid debug prints
Patch-mainline: not yet
References: 

Removed invalid debug prints.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

--- linux-2.6.11/sound/pci/emu10k1/voice.c	2005-03-14 14:31:48.000000000 +0100
+++ linux/sound/pci/emu10k1/voice.c	2005-03-14 14:32:30.000000000 +0100
@@ -81,12 +81,8 @@ static int voice_alloc(emu10k1_t *emu, e
 		}
 	}
 	
-	if (first_voice == last_voice) {
-		printk("BUG (or not enough voices), number %d, next free %d!\n",
-				number,
-				emu->next_free_voice);
+	if (first_voice == last_voice)
 		return -ENOMEM;
-	}	
 	
 	for (i=0; i < number; i++) {
 		voice = &emu->voices[(first_voice + i) % NUM_G];
