From: Takashi Iwai <tiwai@suse.de>
Subject: Add AC97 quirk for Dell
Patch-mainline: not yet
References: 

Added AC97 quirk entry for Dell Precision 650.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

--- linux/sound/pci/intel8x0.c	14 Feb 2005 15:17:23 -0000	1.191
+++ linux/sound/pci/intel8x0.c	24 Feb 2005 11:52:50 -0000	1.192
@@ -1763,6 +1763,12 @@
 	},
 	{
 		.vendor = 0x1028,
+		.device = 0x012c,
+		.name = "Dell Precision 650",	/* AD1981A */
+		.type = AC97_TUNE_HP_ONLY
+	},
+	{
+		.vendor = 0x1028,
 		.device = 0x012d,
 		.name = "Dell Precision 450",	/* AD1981B*/
 		.type = AC97_TUNE_HP_ONLY
