From: Takashi Iwai <tiwai@suse.de>
Subject: [ALSA] intel8x0 - Add quirk for IBM NetVisa A30p
Patch-mainline: 
References: 114683

Added ac97_quirk entry for IBM NetVisa A30p.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

---


diff -u linux/sound/pci/intel8x0.c:1.208 linux/sound/pci/intel8x0.c:1.209
--- linux/sound/pci/intel8x0.c:1.208	Thu Aug 25 12:30:27 2005
+++ linux/sound/pci/intel8x0.c	Thu Sep  1 05:56:23 2005
@@ -1759,6 +1759,12 @@
 		.type = AC97_TUNE_ALC_JACK
 	},
 	{
+		.subvendor = 0x1014,
+		.subdevice = 0x0267,
+		.name = "IBM NetVista A30p",	/* AD1981B */
+		.type = AC97_TUNE_HP_ONLY
+	},
+	{
 		.subvendor = 0x1028,
 		.subdevice = 0x00d8,
 		.name = "Dell Precision 530",	/* AD1885 */
