From: Takashi Iwai <tiwai@suse.de>
Subject: Fix the wrong allocation size for user controls
Patch-mainline: not yet
References: 

Fixed the wrong allocation size for user controls.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/core/control.c linux/sound/core/control.c
--- linux-2.6.11/sound/core/control.c	2005-03-08 17:49:41.000000000 +0100
+++ linux/sound/core/control.c	2005-03-07 15:22:19.000000000 +0100
@@ -928,7 +928,7 @@ static int snd_ctl_elem_add(snd_ctl_file
 	if (ue == NULL)
 		return -ENOMEM;
 	ue->info = *info;
-	ue->elem_data = (char *)ue + sizeof(ue);
+	ue->elem_data = (char *)ue + sizeof(*ue);
 	ue->elem_data_size = private_size;
 	kctl.private_free = snd_ctl_elem_user_free;
 	_kctl = snd_ctl_new(&kctl, access);
