From: Takashi Iwai <tiwai@suse.de>
Subject: Fix the calculation of appl_ptr
Patch-mainline: not yet
References: 

Fixed the calculation of PCM appl_ptr.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/core/pcm_lib.c linux/sound/core/pcm_lib.c
--- linux-2.6.11/sound/core/pcm_lib.c	2005-03-02 08:38:07.000000000 +0100
+++ linux/sound/core/pcm_lib.c	2005-03-07 15:22:19.000000000 +0100
@@ -1863,9 +1863,8 @@ static void snd_pcm_system_tick_set(snd_
 	if (ticks == 0)
 		del_timer(&runtime->tick_timer);
 	else {
+		ticks += (1000000 / HZ) - 1;
 		ticks /= (1000000 / HZ);
-		if (ticks % (1000000 / HZ))
-			ticks++;
 		mod_timer(&runtime->tick_timer, jiffies + ticks);
 	}
 }
@@ -2142,11 +2141,9 @@ static snd_pcm_sframes_t snd_pcm_lib_wri
 			break;
 		}
 		appl_ptr += frames;
-		if (appl_ptr >= runtime->boundary) {
-			runtime->control->appl_ptr = 0;
-		} else {
-			runtime->control->appl_ptr = appl_ptr;
-		}
+		if (appl_ptr >= runtime->boundary)
+			appl_ptr -= runtime->boundary;
+		runtime->control->appl_ptr = appl_ptr;
 		if (substream->ops->ack)
 			substream->ops->ack(substream);
 
@@ -2441,11 +2438,9 @@ static snd_pcm_sframes_t snd_pcm_lib_rea
 			break;
 		}
 		appl_ptr += frames;
-		if (appl_ptr >= runtime->boundary) {
-			runtime->control->appl_ptr = 0;
-		} else {
-			runtime->control->appl_ptr = appl_ptr;
-		}
+		if (appl_ptr >= runtime->boundary)
+			appl_ptr -= runtime->boundary;
+		runtime->control->appl_ptr = appl_ptr;
 		if (substream->ops->ack)
 			substream->ops->ack(substream);
 
