From: Takashi Iwai <tiwai@suse.de>
Subject: Add DXS support for MSI K8T Neo2-FI
Patch-mainline: not yet
References: 

Added the DXS entry for  MSI K8T Neo2-FI.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/pci/via82xx.c linux/sound/pci/via82xx.c
--- linux-2.6.11/sound/pci/via82xx.c	2005-03-11 20:44:46.000000000 +0100
+++ linux/sound/pci/via82xx.c	2005-03-11 12:28:39.000000000 +0100
@@ -2127,6 +2127,7 @@ static int __devinit check_dxs_list(stru
 		{ .vendor = 0x1458, .device = 0xa002, .action = VIA_DXS_ENABLE }, /* Gigabyte GA-7VAXP */
 		{ .vendor = 0x1462, .device = 0x3800, .action = VIA_DXS_ENABLE }, /* MSI KT266 */
 		{ .vendor = 0x1462, .device = 0x5901, .action = VIA_DXS_NO_VRA }, /* MSI KT6 Delta-SR */
+		{ .vendor = 0x1462, .device = 0x7023, .action = VIA_DXS_NO_VRA }, /* MSI K8T Neo2-FI */
 		{ .vendor = 0x1462, .device = 0x7120, .action = VIA_DXS_ENABLE }, /* MSI KT4V */
 		{ .vendor = 0x147b, .device = 0x1401, .action = VIA_DXS_ENABLE }, /* ABIT KD7(-RAID) */
 		{ .vendor = 0x147b, .device = 0x1411, .action = VIA_DXS_ENABLE }, /* ABIT VA-20 */
