From: Takashi Iwai <tiwai@suse.de>
Subject: Fix SPDIF route control on ALC655/658/850
Patch-mainline: not yet
References: 

Fixed the IEC958 (SPDIF) route control on ALC655/658/850.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/pci/ac97/ac97_patch.c linux/sound/pci/ac97/ac97_patch.c
--- linux-2.6.11/sound/pci/ac97/ac97_patch.c	2005-03-11 20:44:41.000000000 +0100
+++ linux/sound/pci/ac97/ac97_patch.c	2005-03-11 16:14:36.000000000 +0100
@@ -1598,7 +1598,7 @@ static int alc655_iec958_route_put(snd_k
 	ac97_t *ac97 = snd_kcontrol_chip(kcontrol);
 
 	return ac97_update_bits_page(ac97, AC97_ALC650_MULTICH, 3 << 12,
-				     (unsigned short)ucontrol->value.enumerated.item[0],
+				     (unsigned short)ucontrol->value.enumerated.item[0] << 12,
 				     0);
 }
 
