From: Takashi Iwai <tiwai@suse.de>
Subject: [ALSA] Disable auto-toggling of DRC on TiBooks
Patch-mainline: 
References: 113630

Disabled auto-toggling of DRC on TiBooks.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

--

--- linux/sound/ppc/tumbler.c	26 Jul 2005 10:52:45 -0000	1.49
+++ linux/sound/ppc/tumbler.c	1 Sep 2005 16:00:13 -0000
@@ -48,6 +48,9 @@
 #define DBG(fmt...)
 #endif
 
+/* Define this if to toggle off DRC automatically when headphone is plugged in */
+/* #define AUTO_TOGGLE_DRC */
+
 /* i2c address for tumbler */
 #define TAS_I2C_ADDR	0x34
 
@@ -948,7 +951,9 @@
 			msleep(10);
 		check_mute(chip, &mix->amp_mute, 1, mix->auto_mute_notify,
 			   chip->speaker_sw_ctl);
+#ifdef AUTO_TOGGLE_DRC
 		mix->drc_enable = 0;
+#endif
 	} else {
 		/* unmute speaker, mute others */
 		check_mute(chip, &mix->amp_mute, 0, mix->auto_mute_notify,
@@ -960,20 +965,26 @@
 		if (mix->line_mute.addr != 0)
 			check_mute(chip, &mix->line_mute, 1, mix->auto_mute_notify,
 				   chip->lineout_sw_ctl);
+#ifdef AUTO_TOGGLE_DRC
 		mix->drc_enable = 1;
+#endif
 	}
 	if (mix->auto_mute_notify) {
 		snd_ctl_notify(chip->card, SNDRV_CTL_EVENT_MASK_VALUE,
 				       &chip->hp_detect_ctl->id);
+#ifdef AUTO_TOGGLE_DRC
 		snd_ctl_notify(chip->card, SNDRV_CTL_EVENT_MASK_VALUE,
 			       &chip->drc_sw_ctl->id);
+#endif
 	}
 
+#ifdef AUTO_TOGGLE_DRC
 	/* first set the DRC so the speaker do not explode -ReneR */
 	if (chip->model == PMAC_TUMBLER)
 		tumbler_set_drc(mix);
 	else
 		snapper_set_drc(mix);
+#endif
 
 	/* reset the master volume so the correct amplification is applied */
 	tumbler_set_master_volume(mix);
@@ -1370,6 +1381,17 @@
 	if ((err = snd_ctl_add(chip->card, chip->drc_sw_ctl)) < 0)
 		return err;
 
+	/* set initial DRC range to 60% */
+	if (chip->model == PMAC_TUMBLER)
+		mix->drc_range = (TAS3001_DRC_MAX * 6) / 10;
+	else
+		mix->drc_range = (TAS3004_DRC_MAX * 6) / 10;
+	mix->drc_enable = 1; /* will be changed later if AUTO_TOGGLE_DRC is set */
+	if (chip->model == PMAC_TUMBLER)
+		tumbler_set_drc(mix);
+	else
+		snapper_set_drc(mix);
+
 #ifdef CONFIG_PM
 	chip->suspend = tumbler_suspend;
 	chip->resume = tumbler_resume;
