From: Takashi Iwai <tiwai@suse.de>
Subject: Don't enable DRA for ALI5455
Patch-mainline: not yet
References: 

Don't enable DRA support for ALI5455.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

--- linux-2.6.11/sound/pci/intel8x0.c	2005-03-08 17:49:42.000000000 +0100
+++ linux/sound/pci/intel8x0.c	2005-03-08 18:10:40.000000000 +0100
@@ -2032,7 +2032,8 @@
 	/* FIXME: my test board doesn't work well with VRA... */
 	if (chip->device_type == DEVICE_ALI)
 		pbus->no_vra = 1;
-	pbus->dra = 1;
+	else
+		pbus->dra = 1;
 	chip->ac97_bus = pbus;
 
 	ac97.pci = chip->pci;
