From: Takashi Iwai <tiwai@suse.de>
Subject: Fix Audigy SPDIF register bits
Patch-mainline: not yet
References: 

Fixed the definition of Audigy SPDIF register bits.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

diff -ruN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/include/sound/emu10k1.h linux/include/sound/emu10k1.h
--- linux-2.6.11/include/sound/emu10k1.h	2005-03-08 17:49:42.000000000 +0100
+++ linux/include/sound/emu10k1.h	2005-03-07 15:22:22.000000000 +0100
@@ -765,8 +765,8 @@
 #define A_SPDIF_SAMPLERATE	0x76		/* Set the sample rate of SPDIF output		*/
 #define A_SPDIF_RATE_MASK	0x000000c0
 #define A_SPDIF_48000		0x00000000
-#define A_SPDIF_44100		0x00000040
-#define A_SPDIF_96000		0x00000080
+#define A_SPDIF_44100		0x00000080
+#define A_SPDIF_96000		0x00000040
 
 /* 0x77,0x78,0x79 "something i2s-related" - default to 0x01080000 on my audigy 2 ZS --rlrevell	*/
 /* 0x7a, 0x7b - lookup tables */
