From: Takashi Iwai <tiwai@suse.de>
Subject: Alias fixes for ALSA modules
Patch-mainline: not yet
References: 

Fix module aliases for some ALSA drivers.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/core/rtctimer.c linux/sound/core/rtctimer.c
--- linux-2.6.11/sound/core/rtctimer.c	2005-03-02 08:38:20.000000000 +0100
+++ linux/sound/core/rtctimer.c	2005-03-07 15:22:19.000000000 +0100
@@ -183,4 +183,6 @@ MODULE_PARM_DESC(rtctimer_freq, "timer f
 
 MODULE_LICENSE("GPL");
 
+MODULE_ALIAS("snd-timer-" __stringify(SNDRV_TIMER_GLOBAL_RTC));
+
 #endif /* CONFIG_RTC || CONFIG_RTC_MODULE */
diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/core/seq/seq_dummy.c linux/sound/core/seq/seq_dummy.c
--- linux-2.6.11/sound/core/seq/seq_dummy.c	2005-03-02 08:38:09.000000000 +0100
+++ linux/sound/core/seq/seq_dummy.c	2005-03-07 15:22:20.000000000 +0100
@@ -63,6 +63,7 @@
 MODULE_AUTHOR("Takashi Iwai <tiwai@suse.de>");
 MODULE_DESCRIPTION("ALSA sequencer MIDI-through client");
 MODULE_LICENSE("GPL");
+MODULE_ALIAS("snd-seq-client-" __stringify(SNDRV_SEQ_CLIENT_DUMMY));
 
 static int ports = 1;
 static int duplex = 0;
diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/core/sound.c linux/sound/core/sound.c
--- linux-2.6.11/sound/core/sound.c	2005-03-08 17:49:41.000000000 +0100
+++ linux/sound/core/sound.c	2005-03-07 15:22:19.000000000 +0100
@@ -48,7 +48,6 @@ module_param(major, int, 0444);
 MODULE_PARM_DESC(major, "Major # for sound driver.");
 module_param(cards_limit, int, 0444);
 MODULE_PARM_DESC(cards_limit, "Count of auto-loadable soundcards.");
-MODULE_ALIAS_CHARDEV_MAJOR(CONFIG_SND_MAJOR);
 #ifdef CONFIG_DEVFS_FS
 module_param(device_mode, int, 0444);
 MODULE_PARM_DESC(device_mode, "Device file permission mask for devfs.");
