From: Takashi Iwai <tiwai@suse.de>
Subject: [ALSA] Add identification for Live 5.1 [SB0220]
Patch-mainline: 2.6.14

Add identification for Live 5.1 [SB0220]

Signed-off-by: Lee Revell <rlrevell@joe-job.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

---


diff -u linux/sound/pci/emu10k1/emu10k1_main.c:1.66 linux/sound/pci/emu10k1/emu10k1_main.c:1.67
--- linux/sound/pci/emu10k1/emu10k1_main.c:1.66	Fri Aug 26 09:26:40 2005
+++ linux/sound/pci/emu10k1/emu10k1_main.c	Mon Aug 29 09:42:00 2005
@@ -741,6 +741,13 @@
 	 .emu10k1_chip = 1,
 	 .ac97_chip = 1,
 	 .sblive51 = 1} ,
+	/* Tested by Thomas Zehetbauer 27th Aug 2005 */
+	{.vendor = 0x1102, .device = 0x0002, .subsystem = 0x80651102,
+	 .driver = "EMU10K1", .name = "SB Live 5.1 [SB0220]", 
+	 .id = "Live",
+	 .emu10k1_chip = 1,
+	 .ac97_chip = 1,
+	 .sblive51 = 1} ,
 	{.vendor = 0x1102, .device = 0x0002, .subsystem = 0x80641102,
 	 .driver = "EMU10K1", .name = "SB Live 5.1", 
 	 .id = "Live",
