From: Takashi Iwai <tiwai@suse.de>
Subject: Fix the chip query of cm8338
Patch-mainline: not yet
References: 

Fixed the version query on CMI8338.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

--- linus-2.6/sound/pci/cmipci.c	2005-02-22 10:44:27.000000000 +0100
+++ linux/sound/pci/cmipci.c	2005-02-22 21:16:33.000000000 +0100
@@ -2554,13 +2554,6 @@ static void __devinit query_chip(cmipci_
 			printk(KERN_ERR "chip %x version not supported\n", detect);
 		}
 	}
-
-	/* added -MCx suffix for chip supporting multi-channels */
-	if (cm->can_multi_ch)
-		sprintf(cm->card->driver + strlen(cm->card->driver),
-			"-MC%d", cm->max_channels);
-	else if (cm->can_ac3_sw)
-		strcpy(cm->card->driver + strlen(cm->card->driver), "-SWIEC");
 }
 
 
@@ -2664,7 +2657,15 @@ static int __devinit snd_cmipci_create(s
 	cm->max_channels = 2;
 	cm->do_soft_ac3 = soft_ac3[dev];
 
-	query_chip(cm);
+	if (pci->device != PCI_DEVICE_ID_CMEDIA_CM8338A &&
+	    pci->device != PCI_DEVICE_ID_CMEDIA_CM8338B)
+		query_chip(cm);
+	/* added -MCx suffix for chip supporting multi-channels */
+	if (cm->can_multi_ch)
+		sprintf(cm->card->driver + strlen(cm->card->driver),
+			"-MC%d", cm->max_channels);
+	else if (cm->can_ac3_sw)
+		strcpy(cm->card->driver + strlen(cm->card->driver), "-SWIEC");
 
 	cm->dig_status = SNDRV_PCM_DEFAULT_CON_SPDIF;
 	cm->dig_pcm_status = SNDRV_PCM_DEFAULT_CON_SPDIF;
