From: Takashi Iwai <tiwai@suse.de>
Subject: [ALSA] emu10k1 - Add missing ac97 support on SBLive! Player 5.1
Patch-mainline: 
References: 113149

Added the missing ac97 support on SBLive! Player 5.1.

Signed-off-by: Takashi Iwai <tiwai@suse.de>
---


diff -u linux/sound/pci/emu10k1/emu10k1_main.c:1.65 linux/sound/pci/emu10k1/emu10k1_main.c:1.66
--- linux/sound/pci/emu10k1/emu10k1_main.c:1.65	Fri Aug 12 15:54:58 2005
+++ linux/sound/pci/emu10k1/emu10k1_main.c	Fri Aug 26 09:26:40 2005
@@ -752,6 +752,7 @@
 	 .driver = "EMU10K1", .name = "SBLive! Player 5.1 [SB0060]", 
 	 .id = "Live",
 	 .emu10k1_chip = 1,
+	 .ac97_chip = 1,
 	 .sblive51 = 1} ,
 	{.vendor = 0x1102, .device = 0x0002, .subsystem = 0x80511102,
 	 .driver = "EMU10K1", .name = "SBLive! Value [CT4850]", 
