From: Takashi Iwai <tiwai@suse.de>
Subject: Fix voice allocation corruption on emu10k1
Patch-mainline: not yet
References: 

Fixed the corruption of voice allocation on emu10k1.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

--- linux-2.6.11/sound/pci/emu10k1/emupcm.c	2005-03-11 20:44:42.000000000 +0100
+++ linux/sound/pci/emu10k1/emupcm.c	2005-03-09 18:17:23.000000000 +0100
@@ -109,14 +109,17 @@
 		snd_emu10k1_voice_free(epcm->emu, epcm->voices[1]);
 		epcm->voices[1] = NULL;
 	}
-	if (voices == 1 && epcm->voices[0] != NULL)
-		return 0;		/* already allocated */
-	if (voices == 2 && epcm->voices[0] != NULL && epcm->voices[1] != NULL)
-		return 0;
-	if (voices > 1) {
-		if (epcm->voices[0] != NULL && epcm->voices[1] == NULL) {
-			snd_emu10k1_voice_free(epcm->emu, epcm->voices[0]);
-			epcm->voices[0] = NULL;
+	for (i = 0; i < voices; i++) {
+		if (epcm->voices[i] == NULL)
+			break;
+	}
+	if (i == voices)
+		return 0; /* already allocated */
+
+	for (i = 0; i < ARRAY_SIZE(epcm->voices); i++) {
+		if (epcm->voices[i]) {
+			snd_emu10k1_voice_free(epcm->emu, epcm->voices[i]);
+			epcm->voices[i] = NULL;
 		}
 	}
 	err = snd_emu10k1_voice_alloc(epcm->emu,
