From: Takashi Iwai <tiwai@suse.de>
Subject: [ALSA] atiixp - Fix PM resume
Patch-mainline: 
References: 

Fixed PM resume on atiixp driver.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

---


diff -u linux/sound/pci/atiixp.c:1.37 linux/sound/pci/atiixp.c:1.38
--- linux/sound/pci/atiixp.c:1.37	Thu Aug 18 05:43:12 2005
+++ linux/sound/pci/atiixp.c	Thu Sep  1 03:56:34 2005
@@ -1453,6 +1453,7 @@
 			atiixp_dma_t *dma = &chip->dmas[i];
 			if (dma->substream && dma->suspended) {
 				dma->ops->enable_dma(chip, 1);
+				dma->substream->ops->prepare(dma->substream);
 				writel((u32)dma->desc_buf.addr | ATI_REG_LINKPTR_EN,
 				       chip->remap_addr + dma->ops->llp_offset);
 				writel(dma->saved_curptr, chip->remap_addr + dma->ops->dt_cur);
