Subject: patch to fix cell_edac polling mode
From: Benjamin Herrenschmidt <benh@kernel.crashing.org>                                                                                                                
References: 437757 - LTC49143

Signed-off-by: Olaf Hering <olh@suse.de>

---
 drivers/edac/cell_edac.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/edac/cell_edac.c
+++ b/drivers/edac/cell_edac.c
@@ -9,6 +9,7 @@
  */
 #undef DEBUG
 
+#include <linux/edac.h>
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/platform_device.h>
@@ -164,6 +165,8 @@ static int __devinit cell_edac_probe(str
 	if (regs == NULL)
 		return -ENODEV;
 
+	edac_op_state = EDAC_OPSTATE_POLL;
+
 	/* Get channel population */
 	reg = in_be64(&regs->mic_mnt_cfg);
 	dev_dbg(&pdev->dev, "MIC_MNT_CFG = 0x%016lx\n", reg);
