diff -purN linux-2.6.2-rc2/drivers/input/evdev.c linux-2.6.2-rc2.wacom/drivers/input/evdev.c
--- linux-2.6.2-rc2/drivers/input/evdev.c	2004-01-15 11:05:53.000000000 +0000
+++ linux-2.6.2-rc2.wacom/drivers/input/evdev.c	2004-01-30 12:33:19.000000000 +0000
@@ -402,7 +402,21 @@ static struct input_handle *evdev_connec
 	struct evdev *evdev;
 	int minor;
 
-	for (minor = 0; minor < EVDEV_MINORS && evdev_table[minor]; minor++);
+	for (minor = 0; minor < EVDEV_MINORS && evdev_table[minor]; minor++)
+	{
+		if (!evdev_table[minor]->exist && evdev_table[minor]->open) /* reuse */
+		{
+			evdev = evdev_table[minor];
+			evdev->handle.dev = dev;
+			evdev->handle.handler = handler;
+			evdev->handle.private = evdev; /* should already be */
+			evdev->exist = 1;
+			printk(KERN_INFO "evdev: reusing event%d for vendor/product 0x%x/0x%x\n", minor, dev->id.vendor, dev->id.product);
+			input_open_device(&evdev->handle);
+			wake_up_interruptible(&evdev->wait);
+			return &evdev->handle;
+		}
+	}
 	if (minor == EVDEV_MINORS) {
 		printk(KERN_ERR "evdev: no more free evdev devices\n");
 		return NULL;
diff -purN linux-2.6.2-rc2/drivers/input/mousedev.c linux-2.6.2-rc2.wacom/drivers/input/mousedev.c
--- linux-2.6.2-rc2/drivers/input/mousedev.c	2004-01-15 11:06:00.000000000 +0000
+++ linux-2.6.2-rc2.wacom/drivers/input/mousedev.c	2004-01-30 12:33:19.000000000 +0000
@@ -459,6 +459,9 @@ static struct input_handle *mousedev_con
 	struct mousedev *mousedev;
 	int minor = 0;
 
+	/* Ignore all wacom tablets */
+	if (dev->id.vendor == 0x56a) return NULL;
+
 	for (minor = 0; minor < MOUSEDEV_MINORS && mousedev_table[minor]; minor++);
 	if (minor == MOUSEDEV_MINORS) {
 		printk(KERN_ERR "mousedev: no more free mousedev devices\n");
