51_dvb-fixups
diff -up linux-2.6.9-rc2/drivers/media/video/dpc7146.c linux/drivers/media/video/dpc7146.c
--- linux-2.6.9-rc2/drivers/media/video/dpc7146.c	2004-09-14 10:35:16.000000000 +0200
+++ linux/drivers/media/video/dpc7146.c	2004-09-17 14:57:30.436925952 +0200
@@ -79,8 +79,8 @@ static struct saa7146_extension_ioctls i
 
 struct dpc
 {
-	struct video_device	video_dev;
-	struct video_device	vbi_dev;
+	struct video_device	*video_dev;
+	struct video_device	*vbi_dev;
 
 	struct i2c_adapter	i2c_adapter;	
 	struct i2c_client	*saa7111a;
@@ -106,7 +106,7 @@ static int dpc_probe(struct saa7146_dev*
 	   video port pins should be enabled here ?! */
 	saa7146_write(dev, MC1, (MASK_08 | MASK_24 | MASK_10 | MASK_26));
 
-	saa7146_i2c_adapter_prepare(dev, &dpc->i2c_adapter, I2C_CLASS_TV_ANALOG, SAA7146_I2C_BUS_BIT_RATE_480);
+	saa7146_i2c_adapter_prepare(dev, &dpc->i2c_adapter, SAA7146_I2C_BUS_BIT_RATE_480);
 	if(i2c_add_adapter(&dpc->i2c_adapter) < 0) {
 		DEB_S(("cannot register i2c-device. skipping.\n"));
 		kfree(dpc);
diff -up linux-2.6.9-rc2/drivers/media/video/hexium_gemini.c linux/drivers/media/video/hexium_gemini.c
--- linux-2.6.9-rc2/drivers/media/video/hexium_gemini.c	2004-09-14 10:34:57.000000000 +0200
+++ linux/drivers/media/video/hexium_gemini.c	2004-09-17 14:57:30.434926327 +0200
@@ -78,7 +78,7 @@ static struct v4l2_queryctrl hexium_cont
 struct hexium
 {
 	int type;
-	struct video_device	video_dev;
+	struct video_device	*video_dev;
 	struct i2c_adapter	i2c_adapter;
 		
 	int 		cur_input;	/* current input */
@@ -250,7 +250,7 @@ static int hexium_attach(struct saa7146_
 	/* enable i2c-port pins */
 	saa7146_write(dev, MC1, (MASK_08 | MASK_24 | MASK_10 | MASK_26));
 
-	saa7146_i2c_adapter_prepare(dev, &hexium->i2c_adapter, I2C_CLASS_TV_ANALOG, SAA7146_I2C_BUS_BIT_RATE_480);
+	saa7146_i2c_adapter_prepare(dev, &hexium->i2c_adapter, SAA7146_I2C_BUS_BIT_RATE_480);
 	if (i2c_add_adapter(&hexium->i2c_adapter) < 0) {
 		DEB_S(("cannot register i2c-device. skipping.\n"));
 		kfree(hexium);
diff -up linux-2.6.9-rc2/drivers/media/video/hexium_orion.c linux/drivers/media/video/hexium_orion.c
--- linux-2.6.9-rc2/drivers/media/video/hexium_orion.c	2004-09-14 10:38:17.000000000 +0200
+++ linux/drivers/media/video/hexium_orion.c	2004-09-17 14:57:30.431926889 +0200
@@ -68,7 +68,7 @@ static struct saa7146_extension_ioctls i
 struct hexium
 {
 	int type;
-	struct video_device	video_dev;
+	struct video_device	*video_dev;
 	struct i2c_adapter	i2c_adapter;	
 	int cur_input;	/* current input */
 };
@@ -237,7 +237,7 @@ static int hexium_probe(struct saa7146_d
 	saa7146_write(dev, DD1_STREAM_B, 0x00000000);
 	saa7146_write(dev, MC2, (MASK_09 | MASK_25 | MASK_10 | MASK_26));
 
-	saa7146_i2c_adapter_prepare(dev, &hexium->i2c_adapter, I2C_CLASS_TV_ANALOG, SAA7146_I2C_BUS_BIT_RATE_480);
+	saa7146_i2c_adapter_prepare(dev, &hexium->i2c_adapter, SAA7146_I2C_BUS_BIT_RATE_480);
 	if (i2c_add_adapter(&hexium->i2c_adapter) < 0) {
 		DEB_S(("cannot register i2c-device. skipping.\n"));
 		kfree(hexium);
diff -up linux-2.6.9-rc2/drivers/media/video/mxb.c linux/drivers/media/video/mxb.c
--- linux-2.6.9-rc2/drivers/media/video/mxb.c	2004-09-14 10:34:40.000000000 +0200
+++ linux/drivers/media/video/mxb.c	2004-09-17 14:57:30.429927264 +0200
@@ -128,8 +128,8 @@ static struct saa7146_extension_ioctls i
 
 struct mxb
 {
-	struct video_device	video_dev;
-	struct video_device	vbi_dev;
+	struct video_device	*video_dev;
+	struct video_device	*vbi_dev;
 
 	struct i2c_adapter	i2c_adapter;	
 
@@ -183,7 +183,7 @@ static int mxb_probe(struct saa7146_dev*
 	}
 	memset(mxb, 0x0, sizeof(struct mxb));	
 
-	saa7146_i2c_adapter_prepare(dev, &mxb->i2c_adapter, I2C_CLASS_TV_ANALOG, SAA7146_I2C_BUS_BIT_RATE_480);
+	saa7146_i2c_adapter_prepare(dev, &mxb->i2c_adapter, SAA7146_I2C_BUS_BIT_RATE_480);
 	if(i2c_add_adapter(&mxb->i2c_adapter) < 0) {
 		DEB_S(("cannot register i2c-device. skipping.\n"));
 		kfree(mxb);
