Index: linux-2.6.8/drivers/media/dvb/frontends/alps_tdmb7.c
===================================================================
--- linux-2.6.8.orig/drivers/media/dvb/frontends/alps_tdmb7.c	2004-10-01 10:26:02.000000000 +0200
+++ linux-2.6.8/drivers/media/dvb/frontends/alps_tdmb7.c	2004-10-01 10:56:43.972401666 +0200
@@ -423,6 +423,9 @@ static int attach_adapter (struct i2c_ad
 
 	dprintk ("%s\n", __FUNCTION__);
 
+	if (0 == strncmp(adapter->name,"cx88",4))
+		return -ENODEV;
+
 	if (i2c_transfer(adapter, msg, 2) != 2)
 		return -ENODEV;
 
