From: Karsten Keil <kkeil@novell.com>
Subject: skb->tail is not longer always a pointer

Fix the e1000 7.6.5 version to always use skb_tail_pointer().

--- linux-2.6.22.orig/drivers/net/e1000/e1000_main.c 2007-09-17 11:40:35.154651000 +0200
+++ linux-2.6.22/drivers/net/e1000/e1000_main.c 2007-09-17 11:40:35.173636000 +0200
@@ -5285,7 +5285,7 @@ static boolean_t e1000_clean_rx_irq_ps(s
 				PCI_DMA_FROMDEVICE);
 			vaddr = kmap_atomic(ps_page->ps_page[0],
 			                    KM_SKB_DATA_SOFTIRQ);
-			memcpy(skb->tail, vaddr, l1);
+			memcpy(skb_tail_pointer(skb), vaddr, l1);
 			kunmap_atomic(vaddr, KM_SKB_DATA_SOFTIRQ);
 			pci_dma_sync_single_for_device(pdev,
 				ps_page_dma->ps_page_dma[0],
