Subject: add alias entry for portN properties
From: olh@suse.de
References: 435215 - LTC48564

---
 drivers/net/ehea/ehea_main.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/net/ehea/ehea_main.c
+++ b/drivers/net/ehea/ehea_main.c
@@ -116,6 +116,10 @@ static struct of_device_id ehea_device_t
 		.name = "lhea",
 		.compatible = "IBM,lhea",
 	},
+	{
+		.type = "network",
+		.compatible = "IBM,lhea-ethernet",
+	},
 	{},
 };
 MODULE_DEVICE_TABLE(of, ehea_device_table);
