Date: Fri, 20 Aug 2004 13:43:57 +0200
From: garloff@suse.de
Subject: Add support for i915 AGP
References: 43414

diff -uNrp linux-2.6.8/drivers/char/agp/intel-agp.c linux-2.6.8.agpi915/drivers/char/agp/intel-agp.c
--- linux-2.6.8/drivers/char/agp/intel-agp.c	2004-08-20 12:58:17.906648259 +0200
+++ linux-2.6.8.agpi915/drivers/char/agp/intel-agp.c	2004-08-20 13:34:58.147973441 +0200
@@ -1768,6 +1768,7 @@ static struct pci_device_id agp_intel_pc
 	ID(PCI_DEVICE_ID_INTEL_82875_HB),
 	ID(PCI_DEVICE_ID_INTEL_7505_0),
 	ID(PCI_DEVICE_ID_INTEL_7205_0),
+	ID(PCI_DEVICE_ID_INTEL_82915G_HB),
 	{ }
 };
 
