--- ./arch/ia64/sn/io/drivers/ioc4/ioc4_serial.c~	2004-05-02 19:38:45.000000000 +0200
+++ ./arch/ia64/sn/io/drivers/ioc4/ioc4_serial.c	2004-05-02 19:41:26.000000000 +0200
@@ -3954,7 +3954,7 @@ static int ioc4_serial_open(struct tty_s
 	if (info->state->count == 1)
 		ioc4_ss_change_speed(info, (struct termios *)0);
 
-	tty->session = current->session;
+	tty->session = current->signal->session;
 	tty->pgrp = process_group(current);
 
 	/* start timer if we are polling */
