From: Fujita Tomonori
Subject: iscsitarget svn update

Update iscsitarget to svn r1262 / 0.4.13

Signed-off-by: Hannes Reinecke <hare@suse.de>

diff -pur linux-2.6.14/drivers/scsi/iscsitarget.orig/iscsi.c linux-2.6.14/drivers/scsi/iscsitarget/iscsi.c
--- linux-2.6.14/drivers/scsi/iscsitarget.orig/iscsi.c	2005-11-03 09:41:18.444882239 +0100
+++ linux-2.6.14/drivers/scsi/iscsitarget/iscsi.c	2005-09-21 12:49:39.000000000 +0200
@@ -248,7 +248,7 @@ static void do_send_data_rsp(struct iscs
 	LIST_HEAD(send);
 
 	dprintk(D_GENERIC, "%p\n", cmnd);
-	pdusize = conn->session->param.max_data_pdu_length;
+	pdusize = conn->session->param.max_xmit_data_length;
 	expsize = cmnd_read_size(cmnd);
 	size = min(expsize, tio->size);
 	offset = 0;
@@ -664,7 +664,7 @@ static int cmnd_recv_pdu(struct iscsi_co
 	offset &= ~PAGE_CACHE_MASK;
 
 	conn->read_msg.msg_iov = conn->read_iov;
-	conn->read_size = (size + 3) & -4;
+	conn->read_size = size = (size + 3) & -4;
 	conn->read_overflow = 0;
 
 	i = 0;
@@ -1547,9 +1547,9 @@ static int check_segment_length(struct i
 	struct iscsi_conn *conn = cmnd->conn;
 	struct iscsi_sess_param *param = &conn->session->param;
 
-	if (cmnd->pdu.datasize > param->max_data_pdu_length) {
+	if (cmnd->pdu.datasize > param->max_recv_data_length) {
 		eprintk("too lond data %x %u %u\n", cmnd_itt(cmnd),
-			cmnd->pdu.datasize, param->max_data_pdu_length);
+			cmnd->pdu.datasize, param->max_recv_data_length);
 
 		if (get_pgcnt(cmnd->pdu.datasize, 0) > ISCSI_CONN_IOV_MAX) {
 			conn_close(conn);
diff -pur linux-2.6.14/drivers/scsi/iscsitarget.orig/iscsi_dbg.h linux-2.6.14/drivers/scsi/iscsitarget/iscsi_dbg.h
--- linux-2.6.14/drivers/scsi/iscsitarget.orig/iscsi_dbg.h	2005-11-03 09:41:18.448882482 +0100
+++ linux-2.6.14/drivers/scsi/iscsitarget/iscsi_dbg.h	2005-09-21 12:49:39.000000000 +0200
@@ -104,11 +104,12 @@ static inline void iscsi_dump_pdu(struct
 
 #define show_param(param)\
 {\
-	eprintk("%d %d %d %d %d %d %d %d %d %d %d %d %d %d\n",\
+	eprintk("%d %d %d %d %d %d %d %d %d %d %d %d %d %d %d\n",\
 		(param)->initial_r2t,\
 		(param)->immediate_data,\
 		(param)->max_connections,\
-		(param)->max_data_pdu_length,\
+		(param)->max_recv_data_length,\
+		(param)->max_xmit_data_length,\
 		(param)->max_burst_length,\
 		(param)->first_burst_length,\
 		(param)->default_wait_time,\
diff -pur linux-2.6.14/drivers/scsi/iscsitarget.orig/iscsi.h linux-2.6.14/drivers/scsi/iscsitarget/iscsi.h
--- linux-2.6.14/drivers/scsi/iscsitarget.orig/iscsi.h	2005-11-03 09:41:18.448882482 +0100
+++ linux-2.6.14/drivers/scsi/iscsitarget/iscsi.h	2005-09-21 12:49:39.000000000 +0200
@@ -20,7 +20,8 @@ struct iscsi_sess_param {
 	int initial_r2t;
 	int immediate_data;
 	int max_connections;
-	int max_data_pdu_length;
+	int max_recv_data_length;
+	int max_xmit_data_length;
 	int max_burst_length;
 	int first_burst_length;
 	int default_wait_time;
diff -pur linux-2.6.14/drivers/scsi/iscsitarget.orig/param.c linux-2.6.14/drivers/scsi/iscsitarget/param.c
--- linux-2.6.14/drivers/scsi/iscsitarget.orig/param.c	2005-11-03 09:41:18.448882482 +0100
+++ linux-2.6.14/drivers/scsi/iscsitarget/param.c	2005-09-21 12:49:39.000000000 +0200
@@ -44,7 +44,9 @@ static void sess_param_check(struct iscs
 	u32 *iparam = info->session_param;
 
 	CHECK_PARAM(info, iparam, max_connections, 1, 65535);
-	CHECK_PARAM(info, iparam, max_data_pdu_length, 512,
+	CHECK_PARAM(info, iparam, max_recv_data_length, 512,
+		    (u32) ((ISCSI_CONN_IOV_MAX - 1) * PAGE_CACHE_SIZE));
+	CHECK_PARAM(info, iparam, max_xmit_data_length, 512,
 		    (u32) ((ISCSI_CONN_IOV_MAX - 1) * PAGE_CACHE_SIZE));
 	CHECK_PARAM(info, iparam, error_recovery_level, 0, 0);
 	CHECK_PARAM(info, iparam, data_pdu_inorder, 1, 1);
@@ -64,7 +66,8 @@ static void sess_param_set(struct iscsi_
 	SET_PARAM(param, info, iparam, initial_r2t);
 	SET_PARAM(param, info, iparam, immediate_data);
 	SET_PARAM(param, info, iparam, max_connections);
-	SET_PARAM(param, info, iparam, max_data_pdu_length);
+	SET_PARAM(param, info, iparam, max_recv_data_length);
+	SET_PARAM(param, info, iparam, max_xmit_data_length);
 	SET_PARAM(param, info, iparam, max_burst_length);
 	SET_PARAM(param, info, iparam, first_burst_length);
 	SET_PARAM(param, info, iparam, default_wait_time);
@@ -88,7 +91,8 @@ static void sess_param_get(struct iscsi_
 	GET_PARAM(param, info, iparam, initial_r2t);
 	GET_PARAM(param, info, iparam, immediate_data);
 	GET_PARAM(param, info, iparam, max_connections);
-	GET_PARAM(param, info, iparam, max_data_pdu_length);
+	GET_PARAM(param, info, iparam, max_recv_data_length);
+	GET_PARAM(param, info, iparam, max_xmit_data_length);
 	GET_PARAM(param, info, iparam, max_burst_length);
 	GET_PARAM(param, info, iparam, first_burst_length);
 	GET_PARAM(param, info, iparam, default_wait_time);
diff -pur linux-2.6.14/drivers/scsi/iscsitarget.orig/target.c linux-2.6.14/drivers/scsi/iscsitarget/kernel/target.c
--- linux-2.6.14/drivers/scsi/iscsitarget.orig/target.c	2005-11-03 09:41:18.448882482 +0100
+++ linux-2.6.14/drivers/scsi/iscsitarget/target.c	2005-09-21 12:49:39.000000000 +0200
@@ -19,7 +19,8 @@ static struct iscsi_sess_param default_s
 	.initial_r2t = 1,
 	.immediate_data = 1,
 	.max_connections = 1,
-	.max_data_pdu_length = 8192,
+	.max_recv_data_length = 8192,
+	.max_xmit_data_length = 8192,
 	.max_burst_length = 262144,
 	.first_burst_length = 65536,
 	.default_wait_time = 2,
diff -pur linux-2.6.14/drivers/scsi/iscsitarget.orig/iet_u.h linux-2.6.14/drivers/scsi/iscsitarget/iet_u.h
--- linux-2.6.14/drivers/scsi/iscsitarget.orig/iet_u.h	2005-11-03 09:41:18.444882239 +0100
+++ linux-2.6.14/drivers/scsi/iscsitarget/iet_u.h	2005-09-21 12:49:39.000000000 +0200
@@ -1,7 +1,7 @@
 #ifndef _IET_U_H
 #define _IET_U_H
 
-#define IET_VERSION_STRING	"0.4.11"
+#define IET_VERSION_STRING	"0.4.13"
 
 /* The maximum length of 223 bytes in the RFC. */
 #define ISCSI_NAME_LEN	256
@@ -51,7 +51,8 @@ enum {
 	key_initial_r2t,
 	key_immediate_data,
 	key_max_connections,
-	key_max_data_pdu_length,
+	key_max_recv_data_length,
+	key_max_xmit_data_length,
 	key_max_burst_length,
 	key_first_burst_length,
 	key_default_wait_time,
@@ -108,10 +109,10 @@ struct iet_event {
 #define	MAX_NR_WTHREADS		128
 
 #define	DEFAULT_NR_QUEUED_CMNDS	32
-#define	MIN_NR_QUEUED_CMNDS	4
+#define	MIN_NR_QUEUED_CMNDS	1
 #define	MAX_NR_QUEUED_CMNDS	256
 
-#define NETLINK_IET	12
+#define NETLINK_IET	21
 
 #define ADD_TARGET _IOW('i', 0, struct target_info)
 #define DEL_TARGET _IOW('i', 1, struct target_info)
