--- linux-2.6.4/drivers/char/lirc/Kconfig.lirc	2004-03-29 14:33:26.026827439 +0200
+++ linux-2.6.4/drivers/char/lirc/Kconfig	2004-03-29 15:18:27.582254678 +0200
@@ -20,23 +20,25 @@
 		default "2"
 		depends on LIRC_SUPPORT
 
-	config LIRC_I2C
-		tristate "I2C Driver"
-		depends on LIRC_SUPPORT && VIDEO_BT848 && I2C && I2C_ALGOBIT
-		help
-		  Say Y here if you need support for the following cards:
-
-		  Pixelview IR
-		  Hauppauage IR
-		  PV951 IR
-		  TV-Box IR
-		  KNC ONE IR
-
-		  If these dont make sense to you, then dont use the module.
-
-	config LIRC_GPIO
-		tristate "GPIO Driver"
-		depends on LIRC_SUPPORT && VIDEO_BT848
+#	config LIRC_I2C
+#		tristate "I2C Driver"
+#		depends on LIRC_SUPPORT && VIDEO_BT848 && I2C && I2C_ALGOBIT
+#		help
+#		  Say Y here if you need support for the following cards:
+#
+#		  Pixelview IR
+#		  Hauppauage IR
+#		  PV951 IR
+#		  TV-Box IR
+#		  KNC ONE IR
+#
+#		  If these dont make sense to you, then dont use the module.
+#			-- obsoleted by ir-kbd-i2c
+
+#	config LIRC_GPIO
+#		tristate "GPIO Driver"
+#		depends on LIRC_SUPPORT && VIDEO_BT848
+#			-- obsoleted by ir-kbd-gpio
 
 	config LIRC_BT829
 		tristate "BT829 Driver"
@@ -50,42 +52,29 @@
 		tristate "ATI USB Driver"
 		depends on LIRC_SUPPORT && USB
 
-	config LIRC_MCEUSB
-		tristate "MCE USB Driver"
-		depends on LIRC_SUPPORT && USB
+#	config LIRC_MCEUSB
+#		tristate "MCE USB Driver"
+#		depends on LIRC_SUPPORT && USB
+#			- doesn't build
 
 	config LIRC_PARALLEL
 		tristate "Parallel Driver"
 		depends on LIRC_SUPPORT && !SMP && PARPORT
 
-		choice
-			prompt "Parallel Port"
-			depends on LIRC_PARALLEL
-			config LIRC_PARALLEL_LPT1
-				bool "LPT1 (0x378, 7)"
-			config LIRC_PARALLEL_LPT2
-				bool "LPT2 (0x278, 5)"
-			config LIRC_PARALLEL_LPT3
-				bool "COM3 (0x3bc, none)"
-			config LIRC_PARALLEL_OTHER
-				bool "Other (custom values)"
-		endchoice
-
+		# simply default to lpt1, can be overritten with insmod
+		# options anyway ...
 		config LIRC_PORT_PARALLEL
-			hex "I/O Port"
-				default "0x378" if LIRC_PARALLEL_LPT1
-				default "0x278" if LIRC_PARALLEL_LPT2
-				default "0x3bc" if LIRC_PARALLEL_LPT3
+			hex
+			default "0x378"
 			depends on LIRC_PARALLEL
 
 		config LIRC_IRQ_PARALLEL
-			hex "IRQ"
-				default "7" if LIRC_PARALLEL_LPT1
-				default "5" if LIRC_PARALLEL_LPT2
+			hex
+			default "7"
 			depends on LIRC_PARALLEL
 
 		config LIRC_TIMER
-			int "Timer"
+			int
 			default "65535"
 			depends on LIRC_PARALLEL
 
@@ -93,116 +82,33 @@
 		tristate "Serial Driver"
 		depends on LIRC_SUPPORT && SERIAL_8250
 
-		choice
-			prompt "Serial Receiver Type"
-			depends on LIRC_SERIAL
-
-			config LIRC_HOMEBREW
-				bool "Homebrew"
-
-			config LIRC_SERIAL_ANIMAX
-				bool "Animax"
-
-			config LIRC_SERIAL_IRDEO
-				bool "IRdeo"
-
-			config LIRC_SERIAL_IRDEO_REMOTE
-				bool "IRdeo Remote"
-
-		endchoice
-
-		config LIRC_SERIAL_TRANSMITTER
-			bool "With transmitter diode"
-			depends on LIRC_SERIAL && !LIRC_SERIAL_ANIMAX
-
-		config LIRC_SERIAL_SOFTCARRIER
-			bool "With software carrier"
-			depends on LIRC_SERIAL_TRANSMITTER
-
-		config LIRC_SERIAL_IGOR
-			bool "Igor Ceska's variation"
-			depends on LIRC_SERIAL
-
-		choice
-			prompt "Serial Port"
-			depends on LIRC_SERIAL
-			config LIRC_SERIAL_COM1
-				bool "COM1 (0x3f8, 4)"
-			config LIRC_SERIAL_COM2
-				bool "COM2 (0x2f8, 3)"
-			config LIRC_SERIAL_COM3
-				bool "COM3 (0x3e8, 4)"
-			config LIRC_SERIAL_COM4
-				bool "COM4 (0x2e8, 3)"
-			config LIRC_SERIAL_OTHER
-				bool "Other (custom values)"
-		endchoice
+		config LIRC_HOMEBREW
+			bool
+			default yes
 
 		config LIRC_PORT_SERIAL
-			hex "I/O Port"
-				default "0x3f8" if LIRC_SERIAL_COM1
-				default "0x2f8" if LIRC_SERIAL_COM2
-				default "0x3e8" if LIRC_SERIAL_COM3
-				default "0x2e8" if LIRC_SERIAL_COM4
+			hex
+			default "0x3f8"
 			depends on LIRC_SERIAL
 
 		config LIRC_IRQ_SERIAL
-			hex "IRQ"
-				default "4" if LIRC_SERIAL_COM1 || LIRC_SERIAL_COM3
-				default "3" if LIRC_SERIAL_COM2 || LIRC_SERIAL_COM4
+			hex
+			default "4"
 			depends on LIRC_SERIAL
 
 	config LIRC_SIR
 		tristate "SIR Driver"
 		depends on LIRC_SUPPORT
-
-		config LIRC_ON_SA1100
-			bool "LIRC driver for StrongARM SA1100 embedded microprocessor"
-			depends on LIRC_SIR
-
-		choice
-			prompt "SIR Type"
-			depends on LIRC_SIR && !LIRC_ON_SA1100
-
-			config LIRC_SIR_IRDA
-				bool "SIR IrDA (built-in IR ports)"
-
-			config LIRC_SIR_TEKRAM
-				bool "Tekram Irmate 210 (16x50 UART compatible serial port)"
-
-			config LIRC_SIR_ACTISYS_ACT200L
-				bool "Actisys Act200L SIR driver support"
-
-		endchoice
-
-		choice
-			prompt "Serial Port"
-			depends on LIRC_SIR
-			config LIRC_SIR_COM1
-				bool "COM1 (0x3f8, 4)"
-			config LIRC_SIR_COM2
-				bool "COM2 (0x2f8, 3)"
-			config LIRC_SIR_COM3
-				bool "COM3 (0x3e8, 4)"
-			config LIRC_SIR_COM4
-				bool "COM4 (0x2e8, 3)"
-			config LIRC_SIR_OTHER
-				bool "Other (custom values)"
-		endchoice
+		select LIRC_SIR_IRDA
 
 		config LIRC_PORT_SIR
-			hex "I/O Port"
-				default "0x3f8" if LIRC_SIR_COM1
-				default "0x2f8" if LIRC_SIR_COM2
-				default "0x3e8" if LIRC_SIR_COM3
-				default "0x2e8" if LIRC_SIR_COM4
+			hex
+			default "0x3f8"
 			depends on LIRC_SIR
 
 		config LIRC_IRQ_SIR
-			hex "IRQ"
-				default "4" if LIRC_SIR_COM1 || LIRC_SIR_COM3
-				default "3" if LIRC_SIR_COM2 || LIRC_SIR_COM4
+			hex
+			default "4"
 			depends on LIRC_SIR
 
 endmenu
-
--- linux-2.6.4/drivers/char/lirc/Makefile.lirc	2004-03-29 15:14:14.444004828 +0200
+++ linux-2.6.4/drivers/char/lirc/Makefile	2004-03-29 15:14:31.590770424 +0200
@@ -12,3 +12,5 @@
 obj-$(CONFIG_LIRC_ATIUSB) += lirc_atiusb.o
 obj-$(CONFIG_LIRC_MCEUSB) += lirc_mceusb.o
 obj-$(CONFIG_LIRC_I2C) += lirc_i2c.o
+
+EXTRA_CFLAGS := -DLIRC26=1
--- linux-2.6.4/drivers/char/lirc/lirc_mceusb.c.lirc	2004-03-29 15:16:15.118241864 +0200
+++ linux-2.6.4/drivers/char/lirc/lirc_mceusb.c	2004-03-29 15:16:32.961875970 +0200
@@ -108,7 +108,7 @@
 #endif
 
 #include <linux/lirc.h>
-#include "drivers/lirc_dev/lirc_dev.h"
+#include "lirc_dev.h"
 
 /* Version Information */
 #define DRIVER_VERSION "v0.1"
--- linux-2.6.4/drivers/char/lirc/lirc_atiusb.c.lirc	2004-03-29 15:23:22.467628253 +0200
+++ linux-2.6.4/drivers/char/lirc/lirc_atiusb.c	2004-03-29 15:23:48.548708311 +0200
@@ -371,7 +371,7 @@
 
 #if KERNEL26
 	dev = interface_to_usbdev(intf);
-	idesc = &intf->altsetting[intf->act_altsetting];
+	idesc = intf->cur_altsetting;
 	if (idesc->desc.bNumEndpoints != 2)
 		return -ENODEV;
 	ep_in = &idesc->endpoint[0].desc;
--- linux-2.6.4/drivers/char/lirc/lirc_it87.c.lirc	2004-03-29 16:09:58.416206567 +0200
+++ linux-2.6.4/drivers/char/lirc/lirc_it87.c	2004-03-29 16:10:13.334396313 +0200
@@ -35,6 +35,7 @@
 
 
 #include <linux/sched.h>
+#include <linux/init.h>
 #include <linux/errno.h>
 #include <linux/signal.h>
 #include <linux/fs.h>
--- linux-2.6.4/drivers/char/lirc/lirc_serial.c.lirc	2004-03-29 16:10:18.611402239 +0200
+++ linux-2.6.4/drivers/char/lirc/lirc_serial.c	2004-03-29 16:10:28.714499027 +0200
@@ -67,6 +67,7 @@
 #include <linux/interrupt.h>
 #include <linux/ioport.h>
 #include <linux/kernel.h>
+#include <linux/init.h>
 #include <linux/major.h>
 #include <linux/serial_reg.h>
 #include <linux/time.h>
--- linux-2.6.4/drivers/char/lirc/lirc_sir.c.lirc	2004-03-29 16:10:36.720990766 +0200
+++ linux-2.6.4/drivers/char/lirc/lirc_sir.c	2004-03-29 16:10:45.767286621 +0200
@@ -59,6 +59,7 @@
 #include <linux/time.h>
 #include <linux/string.h>
 #include <linux/types.h>
+#include <linux/init.h>
 #include <linux/wait.h>
 #include <linux/mm.h>
 #include <linux/delay.h>
