Subject: lirc kconfig stuff
From: kraxel@suse.de
Patch-mainline: never, depends on lirc-drivers patch

kconfig bits for the lirc drivers.

Index: linux-2.6.12/drivers/char/Kconfig
===================================================================
--- linux-2.6.12.orig/drivers/char/Kconfig	2005-07-28 11:51:49.000000000 +0200
+++ linux-2.6.12/drivers/char/Kconfig	2005-07-28 11:57:50.000000000 +0200
@@ -1004,3 +1004,5 @@ source "drivers/char/tpm/Kconfig"
 
 endmenu
 
+source "drivers/char/lirc/Kconfig"
+
Index: linux-2.6.12/drivers/char/Makefile
===================================================================
--- linux-2.6.12.orig/drivers/char/Makefile	2005-07-28 11:51:49.000000000 +0200
+++ linux-2.6.12/drivers/char/Makefile	2005-07-28 11:57:50.000000000 +0200
@@ -7,7 +7,7 @@
 #
 FONTMAPFILE = cp437.uni
 
-obj-y	 += mem.o random.o tty_io.o n_tty.o tty_ioctl.o
+obj-y	 += mem.o random.o tty_io.o n_tty.o tty_ioctl.o lirc/
 
 obj-$(CONFIG_LEGACY_PTYS)	+= pty.o
 obj-$(CONFIG_UNIX98_PTYS)	+= pty.o
Index: linux-2.6.12/drivers/char/lirc/Kconfig
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.12/drivers/char/lirc/Kconfig	2005-07-28 11:57:50.000000000 +0200
@@ -0,0 +1,36 @@
+# LIRC http://lirc.sf.net/
+# Kernel patch by Flameeyes <dgp85@users.sf.net>
+# Check for new patch at http://flameeyes.web.ctonet.it
+#
+# Thanks to Koos Vriezen <koos.vriezen@xs4all.nl> for the Homebrew support.
+# Thanks to Jeff Clark <jeff@tmtrading.com> for support when I wasn't able
+# to update it and for his patch (found at http://www.clarkmania.com/~jclark/
+# Thanks to Bernhard Rosenkraenzer <bero@arklinux.org> for SMP patch.
+# Thanks to Vince <fuzzy77@free.fr> for the temporary lirc_atiusb driver.
+# Thanks to Paul Miller <pmiller9@users.sourceforge.net> for the new working
+# lirc_atiusb driver.
+
+config LIRC_SUPPORT
+	tristate "Linux InfraRed Controller"
+	depends on X86
+
+config LIRC_MAX_DEV
+	int
+	depends on LIRC_SUPPORT
+	default "2"
+
+config LIRC_I2C
+	tristate "I2C Driver"
+	depends on LIRC_SUPPORT && VIDEO_BT848 && I2C && I2C_ALGOBIT
+
+config LIRC_GPIO
+	tristate "GPIO Driver"
+	depends on LIRC_SUPPORT && VIDEO_BT848
+
+config LIRC_PARALLEL
+	tristate "Parallel Driver"
+	depends on LIRC_SUPPORT && !SMP && PARPORT
+
+config LIRC_SERIAL
+	tristate "Serial Driver"
+	depends on LIRC_SUPPORT && SERIAL_8250
Index: linux-2.6.12/drivers/char/lirc/Makefile
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.12/drivers/char/lirc/Makefile	2005-07-28 13:13:28.000000000 +0200
@@ -0,0 +1,19 @@
+#
+# Makefile for the lirc drivers
+#
+
+obj-$(CONFIG_LIRC_SUPPORT) += lirc_dev.o lirc_atiusb.o lirc_bt829.o \
+	lirc_igorplugusb.o lirc_it87.o lirc_mceusb.o lirc_sir.o \
+	lirc_streamzap.o lirc_imon.o lirc_sasem.o
+
+obj-$(CONFIG_LIRC_I2C)		+= lirc_i2c.o
+obj-$(CONFIG_LIRC_GPIO)		+= lirc_gpio.o
+obj-$(CONFIG_LIRC_PARALLEL)	+= lirc_parallel.o
+obj-$(CONFIG_LIRC_SERIAL)	+= lirc_serial.o
+
+EXTRA_CFLAGS += -DKERNEL26=1
+EXTRA_CFLAGS += -DIRCTL_DEV_MAJOR=60
+EXTRA_CFLAGS += -DDEV_LIRC='"lirc"'
+EXTRA_CFLAGS += -DLIRC_PORT=0x3f8
+EXTRA_CFLAGS += -DLIRC_IRQ=4
+EXTRA_CFLAGS += -DLIRC_TIMER=1
