From foo@baz Tue Apr  9 12:12:43 2002
To: Greg KH <greg@kroah.com>
Date: 25 Feb 2005 15:48:28 -08:00
From: khali@linux-fr.org
Subject: [PATCH] PCI: One more Asus SMBus quirk

One more Asus laptop requiring the SMBus quirk (W1N model).

Signed-off-by: Jean Delvare <khali@linux-fr.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


diff -Nru a/drivers/pci/quirks.c b/drivers/pci/quirks.c
--- a/drivers/pci/quirks.c	2005-03-02 23:04:06 -08:00
+++ b/drivers/pci/quirks.c	2005-03-02 23:04:06 -08:00
@@ -786,6 +786,7 @@
 			}
 		if (dev->device == PCI_DEVICE_ID_INTEL_82855PM_HB)
 			switch (dev->subsystem_device) {
+			case 0x184b: /* W1N notebook */
 			case 0x186a: /* M6Ne notebook */
 				asus_hides_smbus = 1;
 			}
