From foo@baz Tue Apr  9 12:12:43 2002
To: Greg KH <greg@kroah.com>
Date: 25 Feb 2005 08:26:11 -08:00
From: gregkh@suse.de
Subject: PCI: remove pci_find_device usage from pci sysfs code.

Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


diff -Nru a/drivers/pci/pci-sysfs.c b/drivers/pci/pci-sysfs.c
--- a/drivers/pci/pci-sysfs.c	2005-03-02 23:04:22 -08:00
+++ b/drivers/pci/pci-sysfs.c	2005-03-02 23:04:22 -08:00
@@ -481,7 +481,7 @@
 	struct pci_dev *pdev = NULL;
 	
 	sysfs_initialized = 1;
-	while ((pdev = pci_find_device(PCI_ANY_ID, PCI_ANY_ID, pdev)) != NULL)
+	while ((pdev = pci_get_device(PCI_ANY_ID, PCI_ANY_ID, pdev)) != NULL)
 		pci_create_sysfs_dev_files(pdev);
 
 	return 0;
