diff -purN /tmp/linux-2.6.4/include/asm-ppc/pci.h ./include/asm-ppc/pci.h
--- /tmp/linux-2.6.4/include/asm-ppc/pci.h	2004-03-26 16:28:20.000000000 +0000
+++ ./include/asm-ppc/pci.h	2004-03-26 17:40:01.000000000 +0000
@@ -313,6 +313,11 @@ pcibios_resource_to_bus(struct pci_dev *
 
 extern void pcibios_add_platform_entries(struct pci_dev *dev);
 
+#define PCI_DMA_ERROR_CODE      (~(dma_addr_t)0x0)
+static inline int pci_dma_mapping_error(dma_addr_t dma_addr)
+{
+	return (dma_addr == PCI_DMA_ERROR_CODE);
+}
 #endif	/* __KERNEL__ */
 
 #endif /* __PPC_PCI_H */
