--- ./drivers/scsi/qla2xxx/qla_iocb.c	2004-05-06 18:35:37.000000000 -0500
+++ ./drivers/scsi/qla2xxx/qla_iocb.c	2004-05-07 18:04:30.763021960 -0500
@@ -379,6 +379,8 @@ qla2x00_start_scsi(srb_t *sp)
 			sg = (struct scatterlist *) cmd->request_buffer;
 			sp->tot_dsds = pci_map_sg(ha->pdev, sg, cmd->use_sg,
 			    cmd->sc_data_direction);
+			if (sp->tot_dsds == 0)
+				return (QLA_FUNCTION_FAILED);
 		} else if (cmd->request_bufflen) {
 		    sp->tot_dsds++;
 		}
