diff -Naur linux-2.6.1/drivers/scsi/scsi.c linux-2.6.1-sacl/drivers/scsi/scsi.c
--- linux-2.6.1/drivers/scsi/scsi.c	Fri Jan  9 00:59:34 2004
+++ linux-2.6.1-sacl/drivers/scsi/scsi.c	Thu Jan 22 14:17:50 2004
@@ -104,7 +104,7 @@
 	"Communications   ",
 	"Unknown          ",
 	"Unknown          ",
-	"Unknown          ",
+	"RAID             ",
 	"Enclosure        ",
 };
 
diff -Naur linux-2.6.1/include/scsi/scsi.h linux-2.6.1-sacl/include/scsi/scsi.h
--- linux-2.6.1/include/scsi/scsi.h	Fri Jan  9 01:00:12 2004
+++ linux-2.6.1-sacl/include/scsi/scsi.h	Thu Jan 22 14:19:03 2004
@@ -193,6 +193,7 @@
 #define TYPE_MEDIUM_CHANGER 0x08
 #define TYPE_COMM           0x09    /* Communications device */
 #define TYPE_ENCLOSURE      0x0d    /* Enclosure Services Device */
+#define TYPE_RAID           0x0c
 #define TYPE_NO_LUN         0x7f
 
 /*
