Subject: [Qce-ga-devel] 2.6.11-rc1 patch.                                                                                                                            
From: Jonathan McDowell <noodles@earth.li>                                                                                                                           

quickcam: Unknown symbol remap_page_range

Signed-off-by: Olaf Hering <olh@suse.de>

--- ./drivers/usb/media/qc/qc-memory.c	2003-12-15 09:31:46.000000000 +0000
+++ ./drivers/usb/media/qc/qc-memory.c	2005-02-08 22:44:49.000000000 +0000
@@ -230,7 +230,11 @@
 	pos = (unsigned long)src;
 	while ((long)size > 0) {
 		physaddr = kvirt_to_pa(pos);
-		if (remap_page_range(vma, start, physaddr, PAGE_SIZE, PAGE_SHARED))
+		if (remap_pfn_range(vma,
+					start,
+					physaddr >> PAGE_SHIFT,
+					PAGE_SIZE,
+					PAGE_SHARED))
 			return -EAGAIN;
 		start += PAGE_SIZE;
 		pos += PAGE_SIZE;
