From: gregkh@suse.de
Subject: USB: minor fixups from the sysfs-info patches on the error paths.

Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


diff -Naur -X /home/gregkh/linux/dontdiff linux-2.6.10/drivers/usb/core/config.c linux-2.6.10-gregkh/drivers/usb/core/config.c
--- linux-2.6.10/drivers/usb/core/config.c	2005-02-27 22:13:51.623487696 -0800
+++ linux-2.6.10-gregkh/drivers/usb/core/config.c	2005-02-27 22:16:37.565260712 -0800
@@ -420,8 +420,8 @@
 	for (c = 0; c < dev->descriptor.bNumConfigurations; c++) {
 		struct usb_host_config *cf = &dev->config[c];
 
-		if (cf->string)
-			kfree(cf->string);
+		kfree(cf->string);
+		cf->string = NULL;
 
 		for (i = 0; i < cf->desc.bNumInterfaces; i++) {
 			if (cf->intf_cache[i])
diff -Naur -X /home/gregkh/linux/dontdiff linux-2.6.10/drivers/usb/core/hub.c linux-2.6.10-gregkh/drivers/usb/core/hub.c
--- linux-2.6.10/drivers/usb/core/hub.c	2005-02-27 22:13:51.583493776 -0800
+++ linux-2.6.10-gregkh/drivers/usb/core/hub.c	2005-02-27 22:16:37.568260256 -0800
@@ -1126,6 +1126,8 @@
 		return;
 	if (usb_string(udev, index, buf, 256) > 0)
 		*string = buf;
+	else
+		kfree(buf);
 }
 
 
