From: olh@suse.de
Subject: let the luser specify the udelay value

adds modprobe usb-storage usb_genesys_transport_delay=200 


diff -purN linux-2.6.11-rc2.orig/drivers/usb/storage/transport.c linux-2.6.11-rc2.genesys/drivers/usb/storage/transport.c
--- linux-2.6.11-rc2.orig/drivers/usb/storage/transport.c	2005-01-22 02:47:48.000000000 +0100
+++ linux-2.6.11-rc2.genesys/drivers/usb/storage/transport.c	2005-02-01 08:51:08.497368910 +0100
@@ -995,7 +995,7 @@ int usb_stor_Bulk_transport(struct scsi_
 	 * command phase and the data phase.  Some devices need a little
 	 * more than that, probably because of clock rate inaccuracies. */
 	if (le16_to_cpu(us->pusb_dev->descriptor.idVendor) == USB_VENDOR_ID_GENESYS)
-		udelay(110);
+		udelay(usb_genesys_transport_delay);
 
 	if (transfer_length) {
 		unsigned int pipe = srb->sc_data_direction == DMA_FROM_DEVICE ? 
diff -purN linux-2.6.11-rc2.orig/drivers/usb/storage/usb.c linux-2.6.11-rc2.genesys/drivers/usb/storage/usb.c
--- linux-2.6.11-rc2.orig/drivers/usb/storage/usb.c	2005-01-22 02:47:31.000000000 +0100
+++ linux-2.6.11-rc2.genesys/drivers/usb/storage/usb.c	2005-02-01 08:52:41.062050480 +0100
@@ -101,6 +101,9 @@ MODULE_LICENSE("GPL");
 static unsigned int delay_use = 5;
 module_param(delay_use, uint, S_IRUGO | S_IWUSR);
 MODULE_PARM_DESC(delay_use, "seconds to delay before using a new device");
+unsigned int usb_genesys_transport_delay = 110;
+module_param(usb_genesys_transport_delay, uint, S_IRUGO | S_IWUSR);
+MODULE_PARM_DESC(usb_genesys_transport_delay, "udelay value in transport.c, only for vendor genesys, default == 110");
 
 
 static int storage_probe(struct usb_interface *iface,
diff -purN linux-2.6.11-rc2.orig/drivers/usb/storage/usb.h linux-2.6.11-rc2.genesys/drivers/usb/storage/usb.h
--- linux-2.6.11-rc2.orig/drivers/usb/storage/usb.h	2005-01-22 02:48:28.000000000 +0100
+++ linux-2.6.11-rc2.genesys/drivers/usb/storage/usb.h	2005-02-01 08:50:55.652707547 +0100
@@ -180,5 +180,6 @@ extern void fill_inquiry_response(struct
 
 /* Vendor ID list for devices that require special handling */
 #define USB_VENDOR_ID_GENESYS		0x05e3	/* Genesys Logic */
+extern unsigned int usb_genesys_transport_delay;
 
 #endif
