diff -up linux-2.6.5-rc3/drivers/media/video/cx88/Makefile linux/drivers/media/video/cx88/Makefile
--- linux-2.6.5-rc3/drivers/media/video/cx88/Makefile	2004-03-31 13:47:59.000000000 +0200
+++ linux/drivers/media/video/cx88/Makefile	2004-03-31 14:59:53.136223272 +0200
@@ -1,5 +1,5 @@
 cx88xx-objs	:= cx88-cards.o cx88-core.o
-cx8800-objs	:= cx88-video.o cx88-tvaudio.o cx88-i2c.o
+cx8800-objs	:= cx88-video.o cx88-tvaudio.o cx88-i2c.o cx88-vbi.o
 
 obj-$(CONFIG_VIDEO_CX88) += cx88xx.o cx8800.o
 
