From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 18 Oct 2017 09:38:33 +0200
Subject: [PATCH] Linux 4.13.8
References: bnc#1012628
Patch-mainline: 4.13.8
Git-commit: d913bc0b6dac2b93233a0aff4ff418be8c615613

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0d4f1b19869d..66ec023da822 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.14.2

