From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 12 Oct 2017 11:56:20 +0200
Subject: [PATCH] Linux 4.13.6
References: bnc#1012628
Patch-mainline: 4.13.6
Git-commit: b058014bd9ee72304191b2d8482f7ea63f34b534

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 189f1a748e4c..9e1af1af327b 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.14.2

