From 94cd0e9dcdc13d2712d3a9ed749eac301dad160d Mon Sep 17 00:00:00 2001
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 10 Sep 2017 07:45:52 +0200
Subject: [PATCH] Linux 4.13.1
References: bnc#1012628
Patch-mainline: 4.13.1
Git-commit: 94cd0e9dcdc13d2712d3a9ed749eac301dad160d

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ed65d7278bb3..41a976854cad 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.14.2

