From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 2 Nov 2017 09:54:50 +0100
Subject: [PATCH] Linux 4.13.11
References: bnc#1012628
Patch-mainline: 4.13.11
Git-commit: 3996e9c638b8fe280971dc7f7c1f5baf3a6b4578

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0e30a0d282e8..8280953c8a45 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 10
+SUBLEVEL = 11
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.15.0

