From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 27 Sep 2017 14:43:35 +0200
Subject: [PATCH] Linux 4.13.4
References: bnc#1012628
Patch-mainline: 4.13.4
Git-commit: 6eb9c0fc1bca163dd084da77d77bb11c4b1639bc

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0f31ef4aea7b..159901979dec 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 3
+SUBLEVEL = 4
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.14.2

