From 07dd6cc1fff160143e82cf5df78c1db0b6e03355 Mon Sep 17 00:00:00 2001
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 13 Sep 2017 14:21:49 -0700
Subject: [PATCH] Linux 4.13.2
References: bnc#1012628
Patch-mainline: 4.13.2
Git-commit: 07dd6cc1fff160143e82cf5df78c1db0b6e03355

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 41a976854cad..8aad6bc50d52 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.14.2

