From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 27 Oct 2017 10:39:17 +0200
Subject: [PATCH] Linux 4.13.10
References: bnc#1012628
Patch-mainline: 4.13.10
Git-commit: 5f1d25a9a8c999ddb1a18a6b167f1284a46e4dd1

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index aa02679504..0e30a0d282 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.14.2

