From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 24 Oct 2017 14:56:48 +0200
Subject: [PATCH] Linux 4.13.10-rc1
References: bnc#1012628
Patch-mainline: 4.13.10
Git-commit: 3604f7dd42f6c47302607a02f4aa8e70212c0312

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index aa02679504..675a3f726f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 4
 PATCHLEVEL = 13
-SUBLEVEL = 9
-EXTRAVERSION =
+SUBLEVEL = 10
+EXTRAVERSION = -rc1
 NAME = Fearless Coyote
 
 # *DOCUMENTATION*
-- 
2.14.2

