From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 10 Dec 2017 13:40:45 +0100
Subject: [PATCH] Linux 4.14.5
References: bnc#1012628
Patch-mainline: 4.14.5
Git-commit: 64138f0adb25ca8f34baa57af33260b05efe2874

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ba1648c093fe..43ac7bdb10ad 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 14
-SUBLEVEL = 4
+SUBLEVEL = 5
 EXTRAVERSION =
 NAME = Petit Gorille
 
-- 
2.15.1

