From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 10 Jan 2018 09:31:23 +0100
Subject: [PATCH] Linux 4.14.13
References: bnc#1012628
Patch-mainline: 4.14.13
Git-commit: b8447222eb207d5a5ec20a0f357065963dabdcd0

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 20f7d4de0f1c..a67c5179052a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 14
-SUBLEVEL = 12
+SUBLEVEL = 13
 EXTRAVERSION =
 NAME = Petit Gorille
 
-- 
2.15.1

