From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 21 Nov 2017 09:49:25 +0100
Subject: [PATCH] Linux 4.14.1
References: bnc#1012628
Patch-mainline: 4.14.1
Git-commit: 780a781dd6f1af9dfac15b8eeba1cb678c9fc380

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index ccd981892ef2..01f9df1af256 100644
--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,9 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 14
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
-NAME = Fearless Coyote
+NAME = Petit Gorille
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
-- 
2.15.0

