From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 7 Feb 2018 11:14:15 -0800
Subject: [PATCH] Linux 4.15.2
References: bnc#1012628
Patch-mainline: 4.15.2
Git-commit: db22ec452bb4ab9f3ecff00fe935da17d048f14d

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index af101b556ba0..54f1bc10b531 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 15
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.1

