From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 31 Mar 2018 18:05:46 +0200
Subject: [PATCH] Linux 4.15.15
References: bnc#1012628
Patch-mainline: 4.15.15
Git-commit: ae0a11b2bd3317491d88d809dea044b4a7fa12a8

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index a5e561900daf..20c9b7bfeed4 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 15
-SUBLEVEL = 14
+SUBLEVEL = 15
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.3

