From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 3 Feb 2018 17:58:44 +0100
Subject: [PATCH] Linux 4.15.1
References: bnc#1012628
Patch-mainline: 4.15.1
Git-commit: d55dce9083d0308333c83d6abb8ea3560a8293b8

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index c8b8e902d5a4..af101b556ba0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 15
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.1

