From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 8 Mar 2018 22:47:50 -0800
Subject: [PATCH] Linux 4.15.8
References: bnc#1012628
Patch-mainline: 4.15.8
Git-commit: 968369e0e44dbd9a362541038049ddc0b13175ea

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 49f524444050..eb18d200a603 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 15
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.2

