From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 1 May 2018 12:47:31 -0700
Subject: [PATCH] Linux 4.16.7
References: bnc#1012628
Patch-mainline: 4.16.7
Git-commit: 9dc30ff9a115559cc55673d0b1d3c576402d073e

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 41f07b2b7905..1c5d5d8c45e2 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 16
-SUBLEVEL = 6
+SUBLEVEL = 7
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.3

