From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 19 May 2018 10:19:37 +0200
Subject: [PATCH] Linux 4.16.10
References: bnc#1012628
Patch-mainline: 4.16.10
Git-commit: b3fdf8284efbc5020dfbd0a28150637189076115

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ea3cb221d7c5..33f3c94f02ca 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 16
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.3

