From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 12 Apr 2018 12:30:01 +0200
Subject: [PATCH] Linux 4.16.2
References: bnc#1012628
Patch-mainline: 4.16.2
Git-commit: 216f33936eaa006a8b4f5bb992592e34f6432fc2

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 1773c718074e..f0040b05df30 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 16
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.3

