From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 22 May 2018 18:56:31 +0200
Subject: [PATCH] Linux 4.16.11
References: bnc#1012628
Patch-mainline: 4.16.11
Git-commit: 57d3dcea476e924eb6eeeb7a585692edf5c8bd23

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 33f3c94f02ca..79c191442771 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 16
-SUBLEVEL = 10
+SUBLEVEL = 11
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.17.0

