From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 9 May 2018 09:53:14 +0200
Subject: [PATCH] Linux 4.16.8
References: bnc#1012628
Patch-mainline: 4.16.8
Git-commit: dec316ea18281d2892324a4bfeb4d5a8a6605e69

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 1c5d5d8c45e2..5da6ffd69209 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 16
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Fearless Coyote
 
-- 
2.16.3

