From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 18 Jul 2018 07:56:38 +0200
Subject: [PATCH] Linux 4.17.8
References: bnc#1012628
Patch-mainline: 4.17.8
Git-commit: 5606f577a707aa4ccc391714dca815933aeba508

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 5c9f331f29c0..7cc36fe18dbb 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

