From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 8 Jul 2018 15:32:21 +0200
Subject: [PATCH] Linux 4.17.5
References: bnc#1012628
Patch-mainline: 4.17.5
Git-commit: c97bfb7e6e7955dbd6c9b6bff6db2b24c0a3e15e

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 1d740dbe676d..e4ddbad49636 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 4
+SUBLEVEL = 5
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

