From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 26 Jun 2018 07:51:32 +0800
Subject: [PATCH] Linux 4.17.3
References: bnc#1012628
Patch-mainline: 4.17.3
Git-commit: 83e05c58300605f3ab87819c19d1822f54eff81e

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f43cd522b175..31dc3a08295a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 2
+SUBLEVEL = 3
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

