From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 16 Jun 2018 09:18:24 +0200
Subject: [PATCH] Linux 4.17.2
References: bnc#1012628
Patch-mainline: 4.17.2
Git-commit: 3816828a8cd06c5e0a0a9e0456013d61d979d975

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e551c9af6a06..f43cd522b175 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.17.1

