From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 25 Jul 2018 11:26:13 +0200
Subject: [PATCH] Linux 4.17.10
References: bnc#1012628
Patch-mainline: 4.17.10
Git-commit: 50f9e029a6f9f440b8da4259fb7f9b879361368a

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 693fde3aa317..0ab689c38e82 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

