From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 22 Jul 2018 15:16:09 +0200
Subject: [PATCH] Linux 4.17.9
References: bnc#1012628
Patch-mainline: 4.17.9
Git-commit: 339d95b9a7a4e1c911ed3627307c7ba1e8675228

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7cc36fe18dbb..693fde3aa317 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 8
+SUBLEVEL = 9
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

