From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 11 Jun 2018 22:43:19 +0200
Subject: [PATCH] Linux 4.17.1
References: bnc#1012628
Patch-mainline: 4.17.1
Git-commit: d0c077266ecbe4ebbaac24c0fe5bd81c5304c5a2

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 554dcaddbce4..e551c9af6a06 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.17.1

