From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 3 Aug 2018 07:48:07 +0200
Subject: [PATCH] Linux 4.17.12
References: bnc#1012628
Patch-mainline: 4.17.12
Git-commit: 506f6fba269679b77f5e7b71df07ef0bb3013a36

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e2664c641109..790e8faf0ddc 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 17
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

