From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 18 Oct 2018 09:18:20 +0200
Subject: [PATCH] Linux 4.18.15
References: bnc#1012628
Patch-mainline: 4.18.15
Git-commit: 2724bf10255a5bd0fe578b4a4c7a0f7ab15f73cd

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f3909ebe4e64..968eb96a0553 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 14
+SUBLEVEL = 15
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.19.1

