From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 15 Aug 2018 17:37:34 +0200
Subject: [PATCH] Linux 4.18.1
References: bnc#1012628
Patch-mainline: 4.18.1
Git-commit: 529bea37411759c2b5b41a187b3020723c67c16d

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 863f58503bee..5edf963148e8 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

