From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 10 Oct 2018 08:56:08 +0200
Subject: [PATCH] Linux 4.18.13
References: bnc#1012628
Patch-mainline: 4.18.13
Git-commit: 04a3fbba60ae3422f50a63b1672970a6eb0f7018

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 466e07af8473..4442e9ea4b6d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 12
+SUBLEVEL = 13
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.19.0

