From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 20 Oct 2018 09:47:10 +0200
Subject: [PATCH] Linux 4.18.16
References: bnc#1012628
Patch-mainline: 4.18.16
Git-commit: 6b3252287aa248cb49befa5de2a68fed6413c814

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 968eb96a0553..034dd990b0ae 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 15
+SUBLEVEL = 16
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.19.1

