From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 3 Oct 2018 16:59:28 -0700
Subject: [PATCH] Linux 4.18.12
References: bnc#1012628
Patch-mainline: 4.18.12
Git-commit: 7da07a3216a00aa3c523db4539fc6914497d0576

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index de0ecace693a..466e07af8473 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.19.0

