From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 19 Sep 2018 22:41:37 +0200
Subject: [PATCH] Linux 4.18.9
References: bnc#1012628
Patch-mainline: 4.18.9
Git-commit: 86e014f514f979312cae23a15284cb81d1ee7336

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0d73431f66cd..1178348fb9ca 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 8
+SUBLEVEL = 9
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.19.0

