From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 17 Aug 2018 21:03:49 +0200
Subject: [PATCH] Linux 4.18.2
References: bnc#1012628
Patch-mainline: 4.18.2
Git-commit: e8c3aced75bd8803f62ead25941fd68302f9efcb

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 5edf963148e8..fd409a0fd4e1 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.18.0

