From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 26 Sep 2018 08:39:43 +0200
Subject: [PATCH] Linux 4.18.10
References: bnc#1012628
Patch-mainline: 4.18.10
Git-commit: 183348489d36bbe1ace1364830425009e98bcc25

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index af5507fb1b82..ffab15235ff0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.19.0

