From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 29 Sep 2018 02:56:03 -0700
Subject: [PATCH] Linux 4.18.11
References: bnc#1012628
Patch-mainline: 4.18.11
Git-commit: 2f411a0873a9daa20ae16cf5879d11bbea267582

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ffab15235ff0..de0ecace693a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 18
-SUBLEVEL = 10
+SUBLEVEL = 11
 EXTRAVERSION =
 NAME = Merciless Moray
 
-- 
2.19.0

