From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 5 Dec 2018 19:32:14 +0100
Subject: [PATCH] Linux 4.19.7
References: bnc#1012628
Patch-mainline: 4.19.7
Git-commit: 61c68f2a2af0f3dc531053524f980a7875e24e7d

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 20cbb8e84650..d2b4efcfb388 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 6
+SUBLEVEL = 7
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.19.2

