From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 4 Nov 2018 14:50:54 +0100
Subject: [PATCH] Linux 4.19.1
References: bnc#1012628
Patch-mainline: 4.19.1
Git-commit: 07a03b97b9ce2a6430344386eeab9b16283b893f

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 69fa5c0310d8..abcd8ca4966f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.19.1

