From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 8 Dec 2018 12:59:10 +0100
Subject: [PATCH] Linux 4.19.8
References: bnc#1012628
Patch-mainline: 4.19.8
Git-commit: 178574b66509c9ff7df4ad26c84a8884567e93b4

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d2b4efcfb388..34bc4c752c49 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.19.2

