From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 21 Dec 2018 14:15:25 +0100
Subject: [PATCH] Linux 4.19.12
References: bnc#1012628
Patch-mainline: 4.19.12
Git-commit: 2a7cb228d29c3882c1414c10a44c5f3f59bfa44d

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 676155d4dc3e..9770f29a690a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.20.1

