From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 19 Dec 2018 19:19:54 +0100
Subject: [PATCH] Linux 4.19.11
References: bnc#1012628
Patch-mainline: 4.19.11
Git-commit: 2a5f80c5bd72487ddf35638c078c6a753894db87

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 36d9de42def3..676155d4dc3e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 10
+SUBLEVEL = 11
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.20.1

