From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 13 Nov 2018 11:09:00 -0800
Subject: [PATCH] Linux 4.19.2
References: bnc#1012628
Patch-mainline: 4.19.2
Git-commit: 7950eb316adf792283cac5743dfe5a11e74833dc

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index abcd8ca4966f..c8fe567f18ab 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.19.1

