From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 1 Dec 2018 09:37:35 +0100
Subject: [PATCH] Linux 4.19.6
References: bnc#1012628
Patch-mainline: 4.19.6
Git-commit: 96db90800c06d3fe3fa08eb6222fe201286bb778

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index a07830185bdf..20cbb8e84650 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.19.2

