From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 23 Nov 2018 08:17:07 +0100
Subject: [PATCH] Linux 4.19.4
References: bnc#1012628
Patch-mainline: 4.19.4
Git-commit: 98a661ea85a8b46aa3e4bff1d900e289011e2d2a

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e4064fa16f11..1f3c7adeea63 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 19
-SUBLEVEL = 3
+SUBLEVEL = 4
 EXTRAVERSION =
 NAME = "People's Front"
 
-- 
2.19.1

