From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 23 Feb 2019 09:08:07 +0100
Subject: [PATCH] Linux 4.20.12
References: bnc#1012628
Patch-mainline: 4.20.12
Git-commit: c91951f15978f1a0c6b65f063d30f7ea7bc6fb42

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 193cfe3a3d70..0a92b4e11621 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 20
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.20.1

