From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 31 Jan 2019 08:15:47 +0100
Subject: [PATCH] Linux 4.20.6
References: bnc#1012628
Patch-mainline: 4.20.6
Git-commit: 65f42a73e55331961153006e902c4fe0bb4a69bb

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 690f6a9d9f1b..523922ea9c97 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 20
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.20.1

