From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 22 Jan 2019 21:10:01 +0100
Subject: [PATCH] Linux 4.20.4
References: bnc#1012628
Patch-mainline: 4.20.4
Git-commit: 3d4332cd825848da1a40c7116eb777b3269df0ea

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 3e70871059be..a056dba5ede0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 20
-SUBLEVEL = 3
+SUBLEVEL = 4
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.20.1

