From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 13 Jan 2019 09:24:10 +0100
Subject: [PATCH] Linux 4.20.2
References: bnc#1012628
Patch-mainline: 4.20.2
Git-commit: 8aab2b4410a257349539c4b09ac9038f369094f5

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 84d2f8deea30..4ba3dd0bf35d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 20
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.20.1

