From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 9 Jan 2019 17:46:05 +0100
Subject: [PATCH] Linux 4.20.1
References: bnc#1012628
Patch-mainline: 4.20.1
Git-commit: 8c3f48e8c28823378274d2342a2ff1442a4af55f

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7a2a9a175756..84d2f8deea30 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 4
 PATCHLEVEL = 20
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.20.1

