From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 23 Mar 2019 20:11:41 +0100
Subject: [PATCH] Linux 5.0.4
References: bnc#1012628
Patch-mainline: 5.0.4
Git-commit: 239cc2c5a3c8555efd2330cc5b09affd5b503669

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fb888787e7d1..06fda21614bc 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 0
-SUBLEVEL = 3
+SUBLEVEL = 4
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

