From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 13 Mar 2019 14:01:32 -0700
Subject: [PATCH] Linux 5.0.2
References: bnc#1012628
Patch-mainline: 5.0.2
Git-commit: 9e6fdea016e45c5feab60589241e0d3607d39945

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 3cd7163fe164..bb2f7664594a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 0
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

