From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 5 May 2019 14:43:41 +0200
Subject: [PATCH] Linux 5.0.13
References: bnc#1012628
Patch-mainline: 5.0.13
Git-commit: e5b9547b1aa39164a8df1d01f2996391c0356d71

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fd044f594bbf..51a819544505 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 0
-SUBLEVEL = 12
+SUBLEVEL = 13
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

