From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 10 Mar 2019 07:09:02 +0100
Subject: [PATCH] Linux 5.0.1
References: bnc#1012628
Patch-mainline: 5.0.1
Git-commit: 283506fcd65de1bc10fcc6e2ca633a7b63171ffa

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d5713e7b1e50..3cd7163fe164 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 0
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

