From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 4 May 2019 09:21:23 +0200
Subject: [PATCH] Linux 5.0.12
References: bnc#1012628
Patch-mainline: 5.0.12
Git-commit: 6006d5b025224f2db8751d317e2e6bcdad69d4b6

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 12870303a029..fd044f594bbf 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 0
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

