From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 20 Apr 2019 09:17:05 +0200
Subject: [PATCH] Linux 5.0.9
References: bnc#1012628
Patch-mainline: 5.0.9
Git-commit: e4abcebedac3415cf347e95749209a4a7b6f3074

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f7666051de66..ef192ca04330 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 0
-SUBLEVEL = 8
+SUBLEVEL = 9
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

