From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 11 May 2019 07:49:56 +0200
Subject: [PATCH] Linux 5.1.1
References: bnc#1012628
Patch-mainline: 5.1.1
Git-commit: b724e93564045d97ec1179d3acacdac60e658569

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 26c92f892d24..bf604f77e5e5 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 1
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

