From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 14 May 2019 19:15:32 +0200
Subject: [PATCH] Linux 5.1.2
References: bnc#1012628
Patch-mainline: 5.1.2
Git-commit: eb5d65a82f5c5dbc85caa0bafeec8c8a78cc6267

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index bf604f77e5e5..58ec07990e76 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 1
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

