From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 15 Jun 2019 11:53:06 +0200
Subject: [PATCH] Linux 5.1.10
References: bnc#1012628
Patch-mainline: 5.1.10
Git-commit: 7e1bdd68ffeecb9ef476f87b791b61910e8c853c

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 2884a8d3b6d6..e7d1973d9c26 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 1
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.22.0

