From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 31 May 2019 06:43:59 -0700
Subject: [PATCH] Linux 5.1.6
References: bnc#1012628
Patch-mainline: 5.1.6
Git-commit: 98e4b991db5ac04be8c23cb2dec277ee31fd22f3

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 24a16a544ffd..d8bdd2bb55dc 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 1
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

