From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 22 May 2019 07:39:58 +0200
Subject: [PATCH] Linux 5.1.4
References: bnc#1012628
Patch-mainline: 5.1.4
Git-commit: e0e8106a6cf13b6abfccf6dac15bac32c9513f3e

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 718e519f789e..acab93537f63 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 1
-SUBLEVEL = 3
+SUBLEVEL = 4
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

