From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 22 Jun 2019 08:09:16 +0200
Subject: [PATCH] Linux 5.1.13
References: bnc#1012628
Patch-mainline: 5.1.13
Git-commit: 7da3d9e6cd75cca48bfd99ba55c874900c0135d9

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6d7bfe9fcd7d..dfcd51a35824 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 1
-SUBLEVEL = 12
+SUBLEVEL = 13
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.22.0

