From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 25 May 2019 18:16:51 +0200
Subject: [PATCH] Linux 5.1.5
References: bnc#1012628
Patch-mainline: 5.1.5
Git-commit: 835365932f0dc25468840753e071c05ad6abc76f

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index acab93537f63..24a16a544ffd 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 1
-SUBLEVEL = 4
+SUBLEVEL = 5
 EXTRAVERSION =
 NAME = Shy Crocodile
 
-- 
2.21.0

