From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 25 Mar 2021 10:13:34 +0100
Subject: [PATCH] Linux 5.11.10
References: bsc#1012628
Patch-mainline: 5.11.10
Git-commit: 6cc049b8e0d05e1519d71afcf2d40d3aa5a48366

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 23403c8e0838..824d15c14be0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 11
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = 💕 Valentine's Day Edition 💕
 
-- 
2.31.0

