From: Sasha Levin <sashal@kernel.org>
Date: Thu, 26 Aug 2021 08:48:43 -0400
Subject: [PATCH] Linux 5.13.13
References: bsc#1012628
Patch-mainline: 5.13.13
Git-commit: 7aeadb5bb82ad21ffbcd54c81d77727b7a05e6c1

Tested-by: Linux Kernel Functional Testing <lkft@linaro.org>
Tested-by: Guenter Roeck <linux@roeck-us.net>
Tested-by: Shuah Khan <skhan@linuxfoundation.org>
Signed-off-by: Sasha Levin <sashal@kernel.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 2458a4abcbcc..ed4031db3894 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 13
-SUBLEVEL = 12
+SUBLEVEL = 13
 EXTRAVERSION =
 NAME = Opossums on Parade
 
-- 
2.33.0

