From: Sasha Levin <sashal@kernel.org>
Date: Wed, 7 Jul 2021 08:24:58 -0400
Subject: [PATCH] Linux 5.13.1
References: bsc#1012628
Patch-mainline: 5.13.1
Git-commit: aaa1f5834d71fe7b687a0c41834bd8d4cc733d90

Tested-by: Fox Chen <foxhlchen@gmail.com>
Tested-by: Guenter Roeck <linux@roeck-us.net>
Tested-by: Linux Kernel Functional Testing <lkft@linaro.org>
Tested-by: Shuah Khan <skhan@linuxfoundation.org>
Signed-off-by: Sasha Levin <sashal@kernel.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0565caea0362..069607cfe283 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 13
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Opossums on Parade
 
-- 
2.32.0

