From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 16 Dec 2021 15:17:49 +0100
Subject: [PATCH] Linux 5.15.9
References: bsc#1012628
Patch-mainline: 5.15.9
Git-commit: 14655b6d42046a537950bfb0dee371742a3dd1fc

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 72344b214bba..e6d2ea920a1d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 15
-SUBLEVEL = 8
+SUBLEVEL = 9
 EXTRAVERSION =
 NAME = Trick or Treat
 
-- 
2.34.1

