From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 15 Jul 2022 10:12:06 +0200
Subject: [PATCH] Linux 5.18.12
References: bsc#1012628
Patch-mainline: 5.18.12
Git-commit: c2e9702659dfc309dfda6116da48f200fe425aab

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 323032d60ac3..f8e2445b6044 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 18
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Superb Owl
 
-- 
2.37.1

