From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 26 Jul 2019 09:11:12 +0200
Subject: [PATCH] Linux 5.2.3
References: bnc#1012628
Patch-mainline: 5.2.3
Git-commit: 577669778464e991c12291f5833bd627d2102bcb

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d6c65b678d21..bcb6a2465e21 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 2
-SUBLEVEL = 2
+SUBLEVEL = 3
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.22.0

