From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 31 Jul 2019 07:25:04 +0200
Subject: [PATCH] Linux 5.2.5
References: bnc#1012628
Patch-mainline: 5.2.5
Git-commit: 2519374d2a6b8aa5d395393f21e74232409c2e82

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fa0f48c43ab2..78bd926c8439 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 2
-SUBLEVEL = 4
+SUBLEVEL = 5
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.22.0

