From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 10 Sep 2019 10:35:27 +0100
Subject: [PATCH] Linux 5.2.14
References: bnc#1012628
Patch-mainline: 5.2.14
Git-commit: 997fee5473ce59c9b1461f54dd2975c57b258a6e

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 288284de8858..d019994462ba 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 2
-SUBLEVEL = 13
+SUBLEVEL = 14
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.23.0

