From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 28 Jul 2019 08:27:24 +0200
Subject: [PATCH] Linux 5.2.4
References: bnc#1012628
Patch-mainline: 5.2.4
Git-commit: fc89179bfa10dd243fce24af71cf622267f31f2f

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index bcb6a2465e21..68ee97784c4d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 2
-SUBLEVEL = 3
+SUBLEVEL = 4
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.22.0

