From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 21 Jul 2019 09:00:45 +0200
Subject: [PATCH] Linux 5.2.2
References: bnc#1012628
Patch-mainline: 5.2.2
Git-commit: e9b75c60f91a359cb0e1b2d0a9ed1c81485215e2

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d8f5dbfd6b76..d6c65b678d21 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 2
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.22.0

