From: Sasha Levin <sashal@kernel.org>
Date: Sun, 25 Aug 2019 10:13:54 -0400
Subject: [PATCH] Linux 5.2.10
References: bnc#1012628
Patch-mainline: 5.2.10
Git-commit: f7d5b3dc4792a5fe0a4d6b8106a8f3eb20c3c24c

Signed-off-by: Sasha Levin <sashal@kernel.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index cfc667fe9959..35fee16d5006 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 2
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.22.0

