From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 9 Aug 2019 17:51:49 +0200
Subject: [PATCH] Linux 5.2.8
References: bnc#1012628
Patch-mainline: 5.2.8
Git-commit: d36a8d2fb62c7c9415213bea9cf576d8b1f9071f

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 359a6b49e576..bad87c4c8117 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 2
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.22.0

