From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 29 Nov 2019 10:10:32 +0100
Subject: [PATCH] Linux 5.4.1
References: bnc#1012628
Patch-mainline: 5.4.1
Git-commit: 79438f37a69a80822333c86acb06a71abce1bce3

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d4d36c61940b..641a62423fd6 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 4
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.24.0

