From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 4 Dec 2019 22:31:09 +0100
Subject: [PATCH] Linux 5.4.2
References: bnc#1012628
Patch-mainline: 5.4.2
Git-commit: 5f8bc2bb0e0f456e2217bbd1caac2acf211431c9

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 641a62423fd6..e67f2e95b71d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 4
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.24.0

