From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 9 Jan 2020 10:25:53 +0100
Subject: [PATCH] Linux 5.4.10
References: bnc#1012628
Patch-mainline: 5.4.10
Git-commit: 7a02c193298ec15f2ba1344b6bcd5d578a41b2e0

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 3ba15c3528c8..726bb3dacd5b 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 4
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.24.1

