From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 5 Mar 2020 16:45:26 +0100
Subject: [PATCH] Linux 5.5.8
References: bnc#1012628
Patch-mainline: 5.5.8
Git-commit: 97453d9b9b2b2356f4389824a9e77fd0c1899d4b

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0f64b92fa39a..a1e5190e4721 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 5
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.25.1

