From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 21 Mar 2020 08:15:56 +0100
Subject: [PATCH] Linux 5.5.11
References: bnc#1012628
Patch-mainline: 5.5.11
Git-commit: 18f57e2782ecc0c4de7e27b17a3ee4885988f80c

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index dcc0053ed18a..b3a42cb0042d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 5
-SUBLEVEL = 10
+SUBLEVEL = 11
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.25.1

