From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 25 Mar 2020 08:27:11 +0100
Subject: [PATCH] Linux 5.5.12
References: bnc#1012628
Patch-mainline: 5.5.12
Git-commit: 7a5ab69495d1a1cd7489abccc7eb6855dba792a4

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index b3a42cb0042d..d962fe0f26ce 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 5
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.26.0

