From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 4 Feb 2020 18:18:03 +0000
Subject: [PATCH] Linux 5.5.2
References: bnc#1012628
Patch-mainline: 5.5.2
Git-commit: 567f0ad8b0b773cc886906514b6c1ccbd1335a41

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ec2caaef8b2d..cfbdf5f920fd 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 5
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.25.0

