From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 24 Feb 2020 08:38:51 +0100
Subject: [PATCH] Linux 5.5.6
References: bnc#1012628
Patch-mainline: 5.5.6
Git-commit: d542c06f7665ee93cf6983dac472a1dd12050401

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 142042ac62e2..7fb236f30926 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 5
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.25.1

