From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 1 Feb 2020 09:33:01 +0000
Subject: [PATCH] Linux 5.5.1
References: bnc#1012628
Patch-mainline: 5.5.1
Git-commit: 70c707aae3e586f6ebc677ad733a9c3cea32f7ac

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6a01b073915e..ec2caaef8b2d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 5
-SUBLEVEL = 0
+SUBLEVEL = 1
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.25.0

