From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 8 Apr 2020 09:11:10 +0200
Subject: [PATCH] Linux 5.6.3
References: bnc#1012628
Patch-mainline: 5.6.3
Git-commit: f9fb85751506e75ecffaa498896efbb0c886adda

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 680b2d52405f..41aafb394d25 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 6
-SUBLEVEL = 2
+SUBLEVEL = 3
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.26.0

