From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 10 Jun 2020 20:21:45 +0200
Subject: [PATCH] Linux 5.7.2
References: bnc#1012628
Patch-mainline: 5.7.2
Git-commit: c58091a3ea4c1fa8a71f38842c877fbd4c54b45d

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 2dd4f37c9f10..a6dda75d18cd 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 7
-SUBLEVEL = 1
+SUBLEVEL = 2
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.27.0

