From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 29 Jul 2020 10:20:01 +0200
Subject: [PATCH] Linux 5.7.11
References: bnc#1012628
Patch-mainline: 5.7.11
Git-commit: 09fd4bcb610a34127afdb97a5df73364c3d7b01c

Tested-by: Shuah Khan <skhan@linuxfoundation.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f569cbc2ef87..12777a95833f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 7
-SUBLEVEL = 10
+SUBLEVEL = 11
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.28.0

