From: Sasha Levin <sashal@kernel.org>
Date: Tue, 30 Jun 2020 16:21:22 -0400
Subject: [PATCH] Linux 5.7.7
References: bnc#1012628
Patch-mainline: 5.7.7
Git-commit: fb6c7905238010236cf0750b209ce9162222eb97

Signed-off-by: Sasha Levin <sashal@kernel.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f928cd1dfdc1..5a5e329d9241 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 7
-SUBLEVEL = 6
+SUBLEVEL = 7
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.27.0

