From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 5 Nov 2020 17:50:14 +0100
Subject: [PATCH] Linux 5.9.6
References: bsc#1012628
Patch-mainline: 5.9.6
Git-commit: 2d182f2dc3b7d38fb1f2a74aabab50a3e0d27c3d

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 27d4fe12da24..2fed32cac74e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 9
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.29.2

