From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 10 Nov 2020 21:16:17 +0100
Subject: [PATCH] Linux 5.9.8
References: bsc#1012628
Patch-mainline: 5.9.8
Git-commit: 951cbbc386ff01b50da4f46387e994e81d9ab431

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 035d86a0d291..ac292d6dd262 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 9
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-- 
2.29.2

