From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 26 Oct 2022 12:53:32 +0200
Subject: [PATCH] Linux 6.0.5
References: bsc#1012628
Patch-mainline: 6.0.5
Git-commit: 3829606fc5dffeccdf80aebeed3aa75255257f35

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f2e41dcc..62a7398c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 0
-SUBLEVEL = 4
+SUBLEVEL = 5
 EXTRAVERSION =
 NAME = Hurr durr I'ma ninja sloth
 
-- 
2.35.3

