From: Helge Deller <deller@gmx.de>
Date: Tue, 23 May 2023 09:06:40 +0200
Subject: [PATCH] parisc: Enable LOCKDEP support
References: bsc#1012628
Patch-mainline: 6.1.31
Git-commit: adf8e96a7ea670d45b5de7594acc67e8f4787ae6

commit adf8e96a7ea670d45b5de7594acc67e8f4787ae6 upstream.

Cc: <stable@vger.kernel.org> # v6.0+
Signed-off-by: Helge Deller <deller@gmx.de>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 arch/parisc/Kconfig | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/parisc/Kconfig b/arch/parisc/Kconfig
index a98940e64243..67c26e81e215 100644
--- a/arch/parisc/Kconfig
+++ b/arch/parisc/Kconfig
@@ -129,6 +129,10 @@ config PM
 config STACKTRACE_SUPPORT
 	def_bool y
 
+config LOCKDEP_SUPPORT
+	bool
+	default y
+
 config ISA_DMA_API
 	bool
 
-- 
2.42.0

