From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 24 Jul 2023 18:55:35 +0200
Subject: [PATCH] Linux 6.1.41
References: bsc#1012628
Patch-mainline: 6.1.41
Git-commit: 5302e81aa2091d7dd59803f0727d1ba35ed2f929

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 2bb0c01391a9..6c940cce3656 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 1
-SUBLEVEL = 40
+SUBLEVEL = 41
 EXTRAVERSION =
 NAME = Curry Ramen
 
-- 
2.42.0

