From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 13 Mar 2023 10:21:32 +0100
Subject: [PATCH] Linux 6.1.19
References: bsc#1012628
Patch-mainline: 6.1.19
Git-commit: 6449a0ba6843fe70523eeb7855984054f36f6d24

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index a825361f7162..ea18c4c20738 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 1
-SUBLEVEL = 18
+SUBLEVEL = 19
 EXTRAVERSION =
 NAME = Hurr durr I'ma ninja sloth
 
-- 
2.42.0

