From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 11 Mar 2023 11:03:45 +0100
Subject: [PATCH] Linux 6.1.17
References: bsc#1012628
Patch-mainline: 6.1.17
Git-commit: 11585e2f8b9d5b4f0a4c51f12adcaab1573811f1

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 5ac6895229e9..db482a420dca 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 1
-SUBLEVEL = 16
+SUBLEVEL = 17
 EXTRAVERSION =
 NAME = Hurr durr I'ma ninja sloth
 
-- 
2.42.0

