From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 14 Jul 2025 16:02:59 +0200
Subject: [PATCH] Linux 6.12.38
References: bsc#1234429
Patch-mainline: 6.12.38
Git-commit: 259f4977409c87a980fa2227b7c76a2fe3fb8c2f

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ca3225cbf130..28c9acdd9b35 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 12
-SUBLEVEL = 37
+SUBLEVEL = 38
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.49.0

