From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 5 May 2025 11:00:29 +0200
Subject: [PATCH] Linux 6.12.27
References: bsc#1234429
Patch-mainline: 6.12.27
Git-commit: b801eaa96a5a2f2a8fe0b5982ffe4a1fba3c1e93

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 467d820fa23f..77f5d180902c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 12
-SUBLEVEL = 26
+SUBLEVEL = 27
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.49.0

