From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 11 Sep 2025 17:21:47 +0200
Subject: [PATCH] Linux 6.12.47
References: bsc#1234429
Patch-mainline: 6.12.47
Git-commit: f6cf124428f51e3ef07a8e54c743873face9d2b2

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 4a4c5e04417d..2500f343c6c8 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 12
-SUBLEVEL = 46
+SUBLEVEL = 47
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.51.0

