From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 11 Sep 2025 17:23:23 +0200
Subject: [PATCH] Linux 6.16.7
References: bsc#1012628
Patch-mainline: 6.16.7
Git-commit: 131e2001572ba68b6728bcba91c58647168d237f

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0200497da26c..86359283ccc9 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 16
-SUBLEVEL = 6
+SUBLEVEL = 7
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.51.0

