From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 28 Aug 2026 08:22:54 +0200
Subject: [PATCH] Linux 6.18.48
References: bsc#1258210
Patch-mainline: 6.18.48
Git-commit: 5bbb9c9f8f808710e2123f2b30f0d61d7d698f52

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 81053b878665..f40baa6f6c23 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 18
-SUBLEVEL = 47
+SUBLEVEL = 48
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.53.0

