From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 11 May 2026 08:20:52 +0200
Subject: [PATCH] Linux 6.18.29
References: bsc#1258210
Patch-mainline: 6.18.29
Git-commit: d31a849ff5011dad5c271b53819a0b279e367d68

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 8d068587e6bd..ca41f54bea2e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 18
-SUBLEVEL = 28
+SUBLEVEL = 29
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.53.0

