From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 8 May 2026 08:40:23 +0200
Subject: [PATCH] Linux 6.18.28
References: bsc#1258210
Patch-mainline: 6.18.28
Git-commit: c31b94a4231c974dca12e6b83f2af1d1d95dfc5e

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index eb796fab3380..8d068587e6bd 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 18
-SUBLEVEL = 27
+SUBLEVEL = 28
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.53.0

