From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 27 Feb 2026 16:05:10 -0500
Subject: [PATCH] Linux 6.18.15
References: bsc#1258210
Patch-mainline: 6.18.15
Git-commit: df0dc1b06fb6b6461f9838694bf84079eca7562a

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d166d0695099..e4aa2e76ea56 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 18
-SUBLEVEL = 14
+SUBLEVEL = 15
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.51.0

