From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 16 Feb 2026 17:10:28 +0100
Subject: [PATCH] Linux 6.18.12
References: bsc#1258210
Patch-mainline: 6.18.12
Git-commit: 2784b1b43af3711a6268a616eb8f41f9788c6e3a

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 1d8d4b2c1da7..09153bd3bc5d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 18
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.51.0

