From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 6 Aug 2026 19:27:09 +0200
Subject: [PATCH] Linux 6.18.43
References: bsc#1258210
Patch-mainline: 6.18.43
Git-commit: 7b923c78b50d2ec52690c4353e5aad8302e80599

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 05ccceb3a87c..5976ab6f0128 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 18
-SUBLEVEL = 42
+SUBLEVEL = 43
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.53.0

