From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 11 Mar 2023 11:04:54 +0100
Subject: [PATCH] Linux 6.2.4
References: bsc#1012628
Patch-mainline: 6.2.4
Git-commit: 3c9c464267468acc8cccf61db4beadee35595cb5

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index eef164b4..83cbbc3a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 2
-SUBLEVEL = 3
+SUBLEVEL = 4
 EXTRAVERSION =
 NAME = Hurr durr I'ma ninja sloth
 
-- 
2.35.3

