From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 24 Jul 2023 18:56:41 +0200
Subject: [PATCH] Linux 6.4.6
References: bsc#1012628
Patch-mainline: 6.4.6
Git-commit: 79562f63d621517795c125092620cf4e5778ad44

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index c3245291..23ddaa3f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 4
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Hurr durr I'ma ninja sloth
 
-- 
2.35.3

