From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 2 Mar 2024 18:23:09 +0100
Subject: [PATCH] Linux 6.6.20
References: bsc#1218260
Patch-mainline: 6.6.20
Git-commit: 9b4a8eac17f0d840729384618b4b1e876233026c

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6130b6bd8d6c..a3bdd583afcc 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 6
-SUBLEVEL = 19
+SUBLEVEL = 20
 EXTRAVERSION =
 NAME = Hurr durr I'ma ninja sloth
 
-- 
2.43.0

