From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 10 Oct 2024 12:50:06 +0200
Subject: [PATCH] Linux 6.6.56
References: bsc#1218260
Patch-mainline: 6.6.56
Git-commit: d4576c5670090297b903771726a526f81fdd8c7c

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Robert Frohl <rfrohl@suse.com>
---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 6e297758842d..301c5694995c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,9 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 6
-SUBLEVEL = 55
+SUBLEVEL = 56
 EXTRAVERSION =
-NAME = Hurr durr I'ma ninja sloth
+NAME = Pinguïn Aangedreven
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
-- 
2.46.0

