From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 2 Mar 2024 18:23:50 +0100
Subject: [PATCH] Linux 6.7.8
References: bsc#1012628
Patch-mainline: 6.7.8
Git-commit: d6d6c49dbf4512f1421f5e42896e2d70dc121f9a

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 72ee3609..6569f225 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 6
 PATCHLEVEL = 7
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Hurr durr I'ma ninja sloth
 
-- 
2.44.0

