From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 8 May 2026 08:42:26 +0200
Subject: [PATCH] Linux 7.0.5
References: bsc#1012628
Patch-mainline: 7.0.5
Git-commit: 3fd2ca34a3e7d1f52e7deaa27016379935ae0845

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 316c0c4ebe5c..6694d125285e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 7
 PATCHLEVEL = 0
-SUBLEVEL = 4
+SUBLEVEL = 5
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.54.0

