From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Mon, 11 May 2026 08:21:59 +0200
Subject: [PATCH] Linux 7.0.6
References: bsc#1012628
Patch-mainline: 7.0.6
Git-commit: 5d83f95062a860326fd9c69a9d7a1f01063270c1

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6694d125285e..dbc380a9339e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 7
 PATCHLEVEL = 0
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.54.0

