From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 6 Aug 2026 19:28:04 +0200
Subject: [PATCH] Linux 7.1.7
References: bsc#1012628
Patch-mainline: 7.1.7
Git-commit: c7ba9d6de43e9d9bd755b1f3c19501a38898c6b6

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 75e5c9573863..4643c9e03a20 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 7
 PATCHLEVEL = 1
-SUBLEVEL = 6
+SUBLEVEL = 7
 EXTRAVERSION =
 NAME = Baby Opossum Posse
 
-- 
2.55.0

