--- linux-2.6.7.proc.orig/arch/ppc/kernel/ppc_htab.c
+++ linux-2.6.7.proc/arch/ppc/kernel/ppc_htab.c
@@ -110,7 +110,7 @@ static ssize_t ppc_htab_read(struct file
 
 	/* FIXME - needs seek/pos locking */
 	
-	if (count < 0 || pos > 512)
+	if (pos < 0 || pos > 512)
 		return -EINVAL;
 
 	if (cur_cpu_spec[0]->cpu_features & CPU_FTR_604_PERF_MON) {
@@ -197,7 +197,7 @@ return_string:
 		n = strlen(buffer) - pos;
 	if (n > count)
 		n = count;
-	if (copy_to_user(buf, buffer + *ppos, n))
+	if (copy_to_user(buf, buffer + pos, n))
 		return -EFAULT;
 	*ppos = pos + n;
 	return n;
