--- linux-2.6.7.proc.orig/arch/ppc64/kernel/rtas_flash.c
+++ linux-2.6.7.proc/arch/ppc64/kernel/rtas_flash.c
@@ -238,7 +238,7 @@ static ssize_t rtas_flash_read(struct fi
 	if (msglen > count)
 		msglen = count;
 
-	if (ppos && *ppos != 0)
+	if (*ppos != 0)
 		return 0;	/* be cheap */
 
 	error = verify_area(VERIFY_WRITE, buf, msglen);
@@ -248,8 +248,7 @@ static ssize_t rtas_flash_read(struct fi
 	if (copy_to_user(buf, msg, msglen))
 		return -EFAULT;
 
-	if (ppos)
-		*ppos = msglen;
+	*ppos = msglen;
 	return msglen;
 }
 
@@ -375,7 +374,7 @@ static ssize_t manage_flash_read(struct 
 	if (msglen > count)
 		msglen = count;
 
-	if (ppos && *ppos != 0)
+	if (*ppos != 0)
 		return 0;	/* be cheap */
 
 	error = verify_area(VERIFY_WRITE, buf, msglen);
@@ -385,8 +384,7 @@ static ssize_t manage_flash_read(struct 
 	if (copy_to_user(buf, msg, msglen))
 		return -EFAULT;
 
-	if (ppos)
-		*ppos = msglen;
+	*ppos = msglen;
 	return msglen;
 }
 
@@ -481,7 +479,7 @@ static ssize_t validate_flash_read(struc
 
 	args_buf = (struct rtas_validate_flash_t *) dp->data;
 
-	if (ppos && *ppos != 0)
+	if (*ppos != 0)
 		return 0;	/* be cheap */
 	
 	msglen = get_validate_flash_msg(args_buf, msg);
@@ -495,8 +493,7 @@ static ssize_t validate_flash_read(struc
 	if (copy_to_user(buf, msg, msglen))
 		return -EFAULT;
 
-	if (ppos)
-		*ppos = msglen;
+	*ppos = msglen;
 	return msglen;
 }
 
