--- linux-2.6.7/arch/ia64/kernel/salinfo.c	2004-06-16 21:10:24.000000000 +0100
+++ linux-2.6.7-viroized/arch/ia64/kernel/salinfo.c	2004-06-27 17:51:10.060116008 +0100
@@ -430,6 +430,9 @@
 	size_t size;
 	u8 *buf;
 	u64 bufsize;
+	loff_t pos = *ppos;
+	
+	/* FIXME: needs seek/parallel-lock */
 
 	if (data->state == STATE_LOG_RECORD) {
 		buf = data->log_buffer;
@@ -441,7 +444,7 @@
 		buf = NULL;
 		bufsize = 0;
 	}
-	if (*ppos >= bufsize)
+	if (pos >= bufsize)
 		return 0;
 
 	saldata = buf + file->f_pos;
@@ -451,7 +454,7 @@
 	if (copy_to_user(buffer, saldata, size))
 		return -EFAULT;
 
-	*ppos += size;
+	*ppos = pos + size;
 	return size;
 }
 
