--- linux-2.6.7.proc.orig/arch/ppc64/kernel/nvram.c
+++ linux-2.6.7.proc/arch/ppc64/kernel/nvram.c
@@ -92,9 +92,9 @@ static ssize_t dev_nvram_read(struct fil
 
 	if (verify_area(VERIFY_WRITE, buf, count))
 		return -EFAULT;
-	if (pos >= size)
+	if (pos >= size || pos < 0)
 		return 0;
-	if (count > size) 
+	if (count > size|| size < 0 ) 
 		count = size;
 
 	tmp_buffer = (char *) kmalloc(count, GFP_KERNEL);
@@ -135,9 +135,9 @@ static ssize_t dev_nvram_write(struct fi
 
 	if (verify_area(VERIFY_READ, buf, count))
 		return -EFAULT;
-	if (pos >= size)
+	if (pos >= size || pos < 0)
 		return 0;
-	if (count > size)
+	if (count > size || size < 0)
 		count = size;
 
 	tmp_buffer = (char *) kmalloc(count, GFP_KERNEL);
