--- linux-2.6.7/arch/ppc/kernel/ppc_htab.c	2004-06-16 21:11:35.000000000 +0100
+++ linux-2.6.7-viroized/arch/ppc/kernel/ppc_htab.c	2004-06-27 17:54:21.158064704 +0100
@@ -106,8 +106,11 @@
 	PTE *ptr;
 #endif /* CONFIG_PPC_STD_MMU */
 	char buffer[512];
+	loff_t pos = *ppos;
 
-	if (count < 0)
+	/* FIXME - needs seek/pos locking */
+	
+	if (count < 0 || pos > 512)
 		return -EINVAL;
 
 	if (cur_cpu_spec[0]->cpu_features & CPU_FTR_604_PERF_MON) {
@@ -188,15 +191,15 @@
 		      "Non-error misses: %lu\n"
 		      "Error misses\t: %lu\n",
 		      pte_misses, pte_errors);
-	if (*ppos >= strlen(buffer))
+	if (pos >= strlen(buffer))
 		return 0;
-	if (n > strlen(buffer) - *ppos)
-		n = strlen(buffer) - *ppos;
+	if (n > strlen(buffer) - pos)
+		n = strlen(buffer) - pos;
 	if (n > count)
 		n = count;
 	if (copy_to_user(buf, buffer + *ppos, n))
 		return -EFAULT;
-	*ppos += n;
+	*ppos = pos + n;
 	return n;
 }
 
