--- linux-2.6.7/arch/s390/kernel/debug.c	2004-06-16 21:11:35.000000000 +0100
+++ linux-2.6.7-viroized/arch/s390/kernel/debug.c	2004-06-27 18:20:44.844307800 +0100
@@ -427,7 +427,7 @@
 
 	p_info = ((file_private_info_t *) file->private_data);
 	if (*offset != p_info->offset) 
-		return -EPIPE;
+		return -ESPIPE;
 	if(p_info->act_area >= p_info->debug_info_snap->nr_areas)
 		return 0;
 
@@ -449,7 +449,7 @@
 				goto out;
 	}
 out:
-	p_info->offset           = *offset + count;
+	p_info->offset           += count;
 	p_info->act_entry_offset = size;	
 	*offset = p_info->offset;
 	return count;
@@ -964,7 +964,7 @@
 		       input_buf[0]);
 	}
       out:
-	*offset += in_buf_size;
+	*offset = in_buf_size;
 	return rc;		/* number of input characters */
 }
 
@@ -1031,7 +1031,7 @@
         printk(KERN_INFO "debug: area `%c` is not valid\n", input_buf[0]);
 
       out:
-        *offset += in_buf_size;
+        *offset = in_buf_size;
         return rc;              /* number of input characters */
 }
 
