--- linux-2.6.7/arch/ppc64/kernel/rtas_flash.c	2004-06-16 21:11:35.000000000 +0100
+++ linux-2.6.7-viroized/arch/ppc64/kernel/rtas_flash.c	2004-07-02 21:56:16.000000000 +0100
@@ -238,7 +238,7 @@
 	if (msglen > count)
 		msglen = count;
 
-	if (ppos && *ppos != 0)
+	if (*ppos != 0)
 		return 0;	/* be cheap */
 
 	error = verify_area(VERIFY_WRITE, buf, msglen);
@@ -248,8 +248,7 @@
 	if (copy_to_user(buf, msg, msglen))
 		return -EFAULT;
 
-	if (ppos)
-		*ppos = msglen;
+	*ppos = msglen;
 	return msglen;
 }
 
@@ -375,7 +374,7 @@
 	if (msglen > count)
 		msglen = count;
 
-	if (ppos && *ppos != 0)
+	if (*ppos != 0)
 		return 0;	/* be cheap */
 
 	error = verify_area(VERIFY_WRITE, buf, msglen);
@@ -385,8 +384,7 @@
 	if (copy_to_user(buf, msg, msglen))
 		return -EFAULT;
 
-	if (ppos)
-		*ppos = msglen;
+	*ppos = msglen;
 	return msglen;
 }
 
@@ -481,7 +479,7 @@
 
 	args_buf = (struct rtas_validate_flash_t *) dp->data;
 
-	if (ppos && *ppos != 0)
+	if (*ppos != 0)
 		return 0;	/* be cheap */
 	
 	msglen = get_validate_flash_msg(args_buf, msg);
@@ -495,8 +493,7 @@
 	if (copy_to_user(buf, msg, msglen))
 		return -EFAULT;
 
-	if (ppos)
-		*ppos = msglen;
+	*ppos = msglen;
 	return msglen;
 }
 
