Index: linux-2.6.5/arch/ia64/kernel/salinfo.c
===================================================================
--- linux-2.6.5.orig/arch/ia64/kernel/salinfo.c
+++ linux-2.6.5/arch/ia64/kernel/salinfo.c
@@ -430,6 +430,9 @@ salinfo_log_read(struct file *file, char
 	size_t size;
 	u8 *buf;
 	u64 bufsize;
+	loff_t pos = *ppos;
+	
+	/* FIXME: needs seek/parallel-lock */
 
 	if (data->state == STATE_LOG_RECORD) {
 		buf = data->log_buffer;
@@ -441,17 +444,17 @@ salinfo_log_read(struct file *file, char
 		buf = NULL;
 		bufsize = 0;
 	}
-	if (*ppos >= bufsize)
+	if (pos >= bufsize)
 		return 0;
 
-	saldata = buf + file->f_pos;
-	size = bufsize - file->f_pos;
+	saldata = buf + pos;
+	size = bufsize - pos;
 	if (size > count)
 		size = count;
 	if (copy_to_user(buffer, saldata, size))
 		return -EFAULT;
 
-	*ppos += size;
+	*ppos = pos + size;
 	return size;
 }
 
