--- linux-2.6.7/arch/ppc64/kernel/nvram.c	2004-06-16 21:10:20.000000000 +0100
+++ linux-2.6.7-viroized/arch/ppc64/kernel/nvram.c	2004-06-27 18:08:30.358966544 +0100
@@ -57,6 +57,7 @@
 static loff_t dev_nvram_llseek(struct file *file, loff_t offset, int origin)
 {
 	int size;
+	/* XXX needs locking */
 
 	if (ppc_md.nvram_size == NULL)
 		return -ENODEV;
@@ -83,6 +84,7 @@
 	ssize_t len;
 	char *tmp_buffer;
 	int size;
+	loff_t pos = *ppos;
 
 	if (ppc_md.nvram_size == NULL)
 		return -ENODEV;
@@ -90,7 +92,7 @@
 
 	if (verify_area(VERIFY_WRITE, buf, count))
 		return -EFAULT;
-	if (*ppos >= size)
+	if (pos >= size)
 		return 0;
 	if (count > size) 
 		count = size;
@@ -101,7 +103,9 @@
 		return -ENOMEM;
 	}
 
-	len = ppc_md.nvram_read(tmp_buffer, count, ppos);
+	len = ppc_md.nvram_read(tmp_buffer, count, &pos);
+	*ppos = pos;
+	
 	if ((long)len <= 0) {
 		kfree(tmp_buffer);
 		return len;
@@ -123,6 +127,7 @@
 	ssize_t len;
 	char * tmp_buffer;
 	int size;
+	loff_t pos = *ppos;
 
 	if (ppc_md.nvram_size == NULL)
 		return -ENODEV;
@@ -130,7 +135,7 @@
 
 	if (verify_area(VERIFY_READ, buf, count))
 		return -EFAULT;
-	if (*ppos >= size)
+	if (pos >= size)
 		return 0;
 	if (count > size)
 		count = size;
@@ -146,7 +151,9 @@
 		return -EFAULT;
 	}
 
-	len = ppc_md.nvram_write(tmp_buffer, count, ppos);
+	len = ppc_md.nvram_write(tmp_buffer, count, &pos);
+	*ppos = pos;
+	
 	if ((long)len <= 0) {
 		kfree(tmp_buffer);
 		return len;
