--- linux-2.6.7/arch/ppc64/kernel/proc_ppc64.c	2004-06-16 21:10:19.000000000 +0100
+++ linux-2.6.7-viroized/arch/ppc64/kernel/proc_ppc64.c	2004-06-27 18:11:06.896169272 +0100
@@ -140,6 +140,7 @@
 
 static loff_t page_map_seek( struct file *file, loff_t off, int whence)
 {
+	/* XXX - locking needed */
 	loff_t new;
 	struct proc_dir_entry *dp = PDE(file->f_dentry->d_inode);
 
@@ -163,7 +164,7 @@
 
 static ssize_t page_map_read( struct file *file, char *buf, size_t nbytes, loff_t *ppos)
 {
-	unsigned pos = *ppos;
+	loff_t pos = *ppos;
 	struct proc_dir_entry *dp = PDE(file->f_dentry->d_inode);
 
 	if ( pos >= dp->size )
