Index: linux-2.6.5/drivers/ieee1394/pcilynx.c
===================================================================
--- linux-2.6.5.orig/drivers/ieee1394/pcilynx.c
+++ linux-2.6.5/drivers/ieee1394/pcilynx.c
@@ -1052,12 +1052,13 @@ static ssize_t mem_read(struct file *fil
         ssize_t retval;
         void *membase;
 
-        if ((off + count) > PCILYNX_MAX_MEMORY+1) {
-                count = PCILYNX_MAX_MEMORY+1 - off;
-        }
-        if (count == 0 || off > PCILYNX_MAX_MEMORY) {
+	/* XXX - should do proper locking instead */
+	if (off < 0)
+		return -EINVAL;
+	if (off > PCILYNX_MAX_MEMORY)
                 return -ENOSPC;
-        }
+        if (count > PCILYNX_MAX_MEMORY + 1 - off)
+                count = PCILYNX_MAX_MEMORY + 1 - off;
 
         switch (md->type) {
         case rom:
@@ -1078,6 +1079,7 @@ static ssize_t mem_read(struct file *fil
 
         if (count < mem_mindma) {
                 memcpy_fromio(md->lynx->mem_dma_buffer, membase+off, count);
+                off += count;
                 goto out;
         }
 
@@ -1108,14 +1110,16 @@ static ssize_t mem_read(struct file *fil
         if (bcount) {
                 memcpy_fromio(md->lynx->mem_dma_buffer + count - bcount,
                               membase+off, bcount);
+                off += bcount;
         }
 
  out:
         retval = copy_to_user(buffer, md->lynx->mem_dma_buffer, count);
         up(&md->lynx->mem_dma_mutex);
 
-	if (retval) return -EFAULT;
-        *offset += count;
+	if (retval < 0)  
+		return -EFAULT;
+        *offset = off;
         return count;
 }
 
@@ -1124,32 +1128,34 @@ static ssize_t mem_write(struct file *fi
                          loff_t *offset)
 {
         struct memdata *md = (struct memdata *)file->private_data;
+        loff_t off = * offset;
+        
+	if (off < 0)
+		return -EINVAL;
+	if (off > PCILYNX_MAX_MEMORY)
+		return -ENOSPC;
 
-        if (((*offset) + count) > PCILYNX_MAX_MEMORY+1) {
-                count = PCILYNX_MAX_MEMORY+1 - *offset;
-        }
-        if (count == 0 || *offset > PCILYNX_MAX_MEMORY) {
-                return -ENOSPC;
-        }
+        if (count > PCILYNX_MAX_MEMORY + 1 - off)
+                count = PCILYNX_MAX_MEMORY + 1 - off;
 
         /* FIXME: dereferencing pointers to PCI mem doesn't work everywhere */
         switch (md->type) {
         case aux:
-		if (copy_from_user(md->lynx->aux_port+(*offset), buffer, count))
+		if (copy_from_user(md->lynx->aux_port+off, buffer, count))
 			return -EFAULT;
                 break;
         case ram:
-		if (copy_from_user(md->lynx->local_ram+(*offset), buffer, count))
+		if (copy_from_user(md->lynx->local_ram+off, buffer, count))
 			return -EFAULT;
                 break;
         case rom:
                 /* the ROM may be writeable */
-		if (copy_from_user(md->lynx->local_rom+(*offset), buffer, count))
+		if (copy_from_user(md->lynx->local_rom+off, buffer, count))
 			return -EFAULT;
                 break;
         }
 
-        file->f_pos += count;
+        *offset = off + count;
         return count;
 }
 #endif /* CONFIG_IEEE1394_PCILYNX_PORTS */
