--- linux-2.6.7.proc.orig/drivers/char/vc_screen.c
+++ linux-2.6.7.proc/drivers/char/vc_screen.c
@@ -107,7 +107,7 @@ vcs_read(struct file *file, char __user 
 {
 	struct inode *inode = file->f_dentry->d_inode;
 	unsigned int currcons = iminor(inode);
-	long pos;
+	loff_t pos;
 	long viewed, attr, read;
 	int col, maxcol;
 	unsigned short *org = NULL;
@@ -278,7 +278,7 @@ vcs_write(struct file *file, const char 
 {
 	struct inode *inode = file->f_dentry->d_inode;
 	unsigned int currcons = iminor(inode);
-	long pos;
+	loff_t pos;
 	long viewed, attr, size, written;
 	char *con_buf0;
 	int col, maxcol;
