--- linux-2.6.7.proc.orig/drivers/char/nwflash.c
+++ linux-2.6.7.proc/drivers/char/nwflash.c
@@ -134,7 +134,7 @@ static int flash_ioctl(struct inode *ino
 
 static ssize_t flash_read(struct file *file, char *buf, size_t size, loff_t * ppos)
 {
-	unsigned long p;
+	loff_t p;
 	unsigned int count = size;
 	int ret = 0;
 
@@ -162,10 +162,7 @@ static ssize_t flash_read(struct file *f
 			ret = count;
 			*ppos += count;
 		} else
-		{
-			up(s);
 			ret = -EFAULT;
-		}
 	}
 	up(&nwflash_sem);
 	return ret;
@@ -173,7 +170,7 @@ static ssize_t flash_read(struct file *f
 
 static ssize_t flash_write(struct file *file, const char *buf, size_t size, loff_t * ppos)
 {
-	unsigned long p;
+	loff_t p;
 	unsigned int count = size;
 	int written;
 	int nBlock, temp, rc;
