--- linux-2.6.7/drivers/usb/core/devio.c	2004-06-16 21:11:36.000000000 +0100
+++ linux-2.6.7-viroized/drivers/usb/core/devio.c	2004-06-27 19:15:04.193811360 +0100
@@ -80,9 +80,10 @@
 static loff_t usbdev_lseek(struct file *file, loff_t offset, int orig)
 {
 	loff_t ret;
+	struct dev_state *ps = (struct dev_state *)file->private_data;
+	struct usb_device *dev = ps->dev;
 
-	lock_kernel();
-
+	down(&dev->serialize);
 	switch (orig) {
 	case 0:
 		file->f_pos = offset;
@@ -96,8 +97,7 @@
 	default:
 		ret = -EINVAL;
 	}
-
-	unlock_kernel();
+	up(&dev->serialize);
 	return ret;
 }
 
@@ -110,8 +110,8 @@
 	loff_t pos;
 	int i;
 
-	pos = *ppos;
 	down(&dev->serialize);
+	pos = *ppos;
 	if (!connected(dev)) {
 		ret = -ENODEV;
 		goto err;
