--- linux-2.6.7/drivers/scsi/st.c	2004-06-16 21:11:36.000000000 +0100
+++ linux-2.6.7-viroized/drivers/scsi/st.c	2004-06-27 20:06:02.315906136 +0100
@@ -1220,7 +1220,7 @@
 
 	if (ppos != &filp->f_pos) {
 		/* "A request was outside the capabilities of the device." */
-		retval = (-ENXIO);
+		retval = (-ESPIPE);
 		goto out;
 	}
 
@@ -3715,6 +3715,7 @@
 	.read =		st_read,
 	.write =	st_write,
 	.ioctl =	st_ioctl,
+	.llseek =	no_llseek,
 	.open =		st_open,
 	.flush =	st_flush,
 	.release =	st_release,
