--- linux-2.6.7.proc.orig/drivers/video/fbmem.c
+++ linux-2.6.7.proc/drivers/video/fbmem.c
@@ -886,12 +886,12 @@ err:
 	return err;
 }
 
-loff_t fbmem_llseek(struct file *file, loff_t offset, int origin)
+loff_t fb_llseek(struct file *file, loff_t offset, int origin)
 {
 	struct inode *inode = file->f_dentry->d_inode;
 	int fbidx = iminor(inode);
 	struct fb_info *info = registered_fb[fbidx];
-	long long retval;
+	loff_t retval;
 	
 	down(&info->mutex);
 	switch (origin) {
