--- linux-2.6.7.proc.orig/drivers/ieee1394/pcilynx.c
+++ linux-2.6.7.proc/drivers/ieee1394/pcilynx.c
@@ -1101,7 +1101,12 @@ static ssize_t mem_read(struct file *fil
         while (bcount >= 4) {
                 retval = mem_dmaread(md, md->lynx->mem_dma_buffer_dma
                                      + count - bcount, bcount, off);
-                if (retval < 0) return retval;
+                if (retval < 0)
+                {
+                	/* FIXME - ought to report short read on error */
+		        up(&md->lynx->mem_dma_mutex);
+			return retval;
+                }
 
                 bcount -= retval;
                 off += retval;
@@ -1133,7 +1138,7 @@ static ssize_t mem_write(struct file *fi
 
 	if (!count)
 		return 0;
-	if (off < PCILYNX_MAX_MEMORY)
+	if (off < 0)
 		return -EINVAL;
 	if (off > PCILYNX_MAX_MEMORY)
 		return -ENOSPC;
