Index: linux-2.6.5/drivers/char/mem.c
===================================================================
--- linux-2.6.5.orig/drivers/char/mem.c
+++ linux-2.6.5/drivers/char/mem.c
@@ -143,7 +143,7 @@ static ssize_t do_write_mem(void *p, uns
 		return -EFAULT;
 	}
 	written += count;
-	*ppos += written;
+	*ppos = realp + written;
 	return written;
 }
 
@@ -180,7 +180,7 @@ static ssize_t read_mem(struct file * fi
 	if (copy_to_user(buf, __va(p), count))
 		return -EFAULT;
 	read += count;
-	*ppos += read;
+	*ppos = p + read;
 	return read;
 }
 
