--- linux-2.6.7/drivers/usb/media/vicam.c	2004-06-16 21:11:36.000000000 +0100
+++ linux-2.6.7-viroized/drivers/usb/media/vicam.c	2004-06-27 19:15:18.832585928 +0100
@@ -864,6 +864,8 @@
 {
 	/* vicam_decode_color - Convert from Vicam Y-Cr-Cb to RGB
 	 * Copyright (C) 2002 Monroe Williams (monroe@pobox.com)
+	 *
+	 * FIXME: camera should use V4L2 YCrCb return type now
 	 */
 
 	int i, prevY, nextY;
@@ -1009,8 +1011,10 @@
 
 	DBG("read %d bytes.\n", (int) count);
 
+	down(&cam->cam_lock);
 	if (*ppos >= VICAM_MAX_FRAME_SIZE) {
 		*ppos = 0;
+		up(&cam->cam_lock);
 		return 0;
 	}
 
@@ -1032,7 +1036,7 @@
 	if (count == VICAM_MAX_FRAME_SIZE) {
 		*ppos = 0;
 	}
-
+	up(&cam->cam_lock);
 	return count;
 }
 
