--- linux-2.6.7/drivers/block/acsi_slm.c	2004-06-16 21:10:03.000000000 +0100
+++ linux-2.6.7-viroized/drivers/block/acsi_slm.c	2004-06-27 18:25:29.019106664 +0100
@@ -366,10 +366,12 @@
 {
 	struct inode *node = file->f_dentry->d_inode;
 	unsigned long page;
+	loff_t pos = *ppos;
 	int length;
 	int end;
 
-	if (count < 0)
+	/* XXX - seek locking required */
+	if (count < 0 || pos < 0)
 		return( -EINVAL );
 	if (!(page = __get_free_page( GFP_KERNEL )))
 		return( -ENOMEM );
@@ -379,18 +381,18 @@
 		count = length;
 		goto out;
 	}
-	if (file->f_pos >= length) {
+	if (pos >= length) {
 		count = 0;
 		goto out;
 	}
-	if (count + file->f_pos > length)
-		count = length - file->f_pos;
-	end = count + file->f_pos;
-	if (copy_to_user(buf, (char *)page + file->f_pos, count)) {
+	if (count + pos > length)
+		count = length - pos;
+	end = count + pos;
+	if (copy_to_user(buf, (char *)page + pos, count)) {
 		count = -EFAULT;
 		goto out;
 	}
-	file->f_pos = end;
+	*ppos = end;
 out:	free_page( page );
 	return( count );
 }
