Index: linux-2.6.5/drivers/usb/core/devices.c
===================================================================
--- linux-2.6.5.orig/drivers/usb/core/devices.c
+++ linux-2.6.5/drivers/usb/core/devices.c
@@ -560,31 +560,39 @@ static ssize_t usb_device_read(struct fi
 {
 	struct list_head *buslist;
 	struct usb_bus *bus;
-	ssize_t ret, total_written = 0;
-	loff_t skip_bytes = *ppos;
+	ssize_t ret = 0, total_written = 0;
+	loff_t skip_bytes;
 
-	if (*ppos < 0)
-		return -EINVAL;
+	down (&usb_bus_list_lock);
+	
+	skip_bytes = *ppos;
+	if (skip_bytes < 0)
+	{
+		ret =  -EINVAL;
+		goto error;
+	}
 	if (nbytes <= 0)
-		return 0;
+		goto error;
 	if (!access_ok(VERIFY_WRITE, buf, nbytes))
-		return -EFAULT;
-
+	{
+		ret =  -EFAULT;
+		goto error;
+	}
 	/* enumerate busses */
-	down (&usb_bus_list_lock);
 	for (buslist = usb_bus_list.next; buslist != &usb_bus_list; buslist = buslist->next) {
 		/* print devices for this bus */
 		bus = list_entry(buslist, struct usb_bus, bus_list);
 		/* recurse through all children of the root hub */
 		ret = usb_device_dump(&buf, &nbytes, &skip_bytes, ppos, bus->root_hub, bus, 0, 0, 0);
 		if (ret < 0) {
-			up(&usb_bus_list_lock);
-			return ret;
+			goto error;
 		}
 		total_written += ret;
 	}
+	ret = total_written;
+error:
 	up (&usb_bus_list_lock);
-	return total_written;
+	return ret;
 }
 
 /* Kernel lock for "lastev" protection */
@@ -646,7 +654,8 @@ static loff_t usb_device_lseek(struct fi
 {
 	loff_t ret;
 
-	lock_kernel();
+	/* This already guards the other users */
+	down(&usb_bus_list_lock);
 
 	switch (orig) {
 	case 0:
@@ -662,7 +671,7 @@ static loff_t usb_device_lseek(struct fi
 		ret = -EINVAL;
 	}
 
-	unlock_kernel();
+	up(&usb_bus_list_lock);
 	return ret;
 }
 
