--- linux-2.6.7/drivers/char/mem.c	2004-06-16 21:11:35.000000000 +0100
+++ linux-2.6.7-viroized/drivers/char/mem.c	2004-06-28 14:48:13.580997984 +0100
@@ -143,7 +143,7 @@
 		return -EFAULT;
 	}
 	written += count;
-	*ppos += written;
+	*ppos = realp + written;
 	return written;
 }
 
@@ -180,7 +180,7 @@
 	if (copy_to_user(buf, __va(p), count))
 		return -EFAULT;
 	read += count;
-	*ppos += read;
+	*ppos = p + read;
 	return read;
 }
 
