Index: linux-2.6.5/drivers/usb/core/devio.c
===================================================================
--- linux-2.6.5.orig/drivers/usb/core/devio.c
+++ linux-2.6.5/drivers/usb/core/devio.c
@@ -63,9 +63,10 @@ struct async {
 static loff_t usbdev_lseek(struct file *file, loff_t offset, int orig)
 {
 	loff_t ret;
+	struct dev_state *ps = (struct dev_state *)file->private_data;
+	struct usb_device *dev = ps->dev;
 
-	lock_kernel();
-
+	down_read(&ps->devsem);
 	switch (orig) {
 	case 0:
 		file->f_pos = offset;
@@ -79,8 +80,7 @@ static loff_t usbdev_lseek(struct file *
 	default:
 		ret = -EINVAL;
 	}
-
-	unlock_kernel();
+	up_read(&ps->devsem);
 	return ret;
 }
 
@@ -92,8 +92,8 @@ static ssize_t usbdev_read(struct file *
 	loff_t pos;
 	int i;
 
-	pos = *ppos;
 	down_read(&ps->devsem);
+	pos = *ppos;
 	if (!ps->dev) {
 		ret = -ENODEV;
 		goto err;
