--- linux-2.6.7.proc.orig/drivers/zorro/proc.c
+++ linux-2.6.7.proc/drivers/zorro/proc.c
@@ -44,8 +44,9 @@ proc_bus_zorro_lseek(struct file *file, 
 		up(&zorro_proc_lock);
 		return -EINVAL;
 	}
+	file->f_pos = new;
 	up(&zorro_proc_lock);
-	return (file->f_pos = new);
+	return file->f_pos;
 }
 
 static ssize_t
@@ -56,9 +57,9 @@ proc_bus_zorro_read(struct file *file, c
 	struct zorro_dev *z = dp->data;
 	struct ConfigDev cd;
 	loff_t pos;
-	
+
 	down(&zorro_proc_lock);
-	
+
 	pos = *ppos;
 	if (pos >= sizeof(struct ConfigDev))
 	{
