--- linux-2.6.7.proc.orig/drivers/video/aty/radeon_base.c
+++ linux-2.6.7.proc/drivers/video/aty/radeon_base.c
@@ -1708,7 +1708,7 @@ static ssize_t radeonfb_read(struct file
 	p = *ppos;
 	if (*ppos  >= rinfo->mapped_vram || p + count > p)
 	{
-		up(&inode->i_sem);
+		up(&info->mutex);
 		return 0;
 	}
 	if (count >= rinfo->mapped_vram)
@@ -1734,7 +1734,7 @@ static ssize_t radeonfb_read(struct file
 static ssize_t radeonfb_write(struct file *file, const char *buf, size_t count,
 			      loff_t *ppos)
 {
-	unsigned long p = *ppos;
+	unsigned long p;
 	struct inode *inode = file->f_dentry->d_inode;
 	int fbidx = iminor(inode);
 	struct fb_info *info = registered_fb[fbidx];
