--- linux-2.6.7.proc.orig/drivers/pnp/isapnp/proc.c
+++ linux-2.6.7.proc/drivers/pnp/isapnp/proc.c
@@ -54,8 +54,9 @@ static loff_t isapnp_proc_bus_lseek(stru
 		up(&isapnp_mutex);
 		return -EINVAL;
 	}
+	file->f_pos = new;
 	up(&isapnp_mutex);
-	return (file->f_pos = new);
+	return new;
 }
 
 static ssize_t isapnp_proc_bus_read(struct file *file, char __user *buf, size_t nbytes, loff_t *ppos)
@@ -82,7 +83,7 @@ static ssize_t isapnp_proc_bus_read(stru
 	if (!access_ok(VERIFY_WRITE, buf, cnt))
 	{
 		up(&isapnp_mutex);
-		return -EINVAL;
+		return -EFAULT;
 	}
 
 	isapnp_cfg_begin(dev->card->number, dev->number);
