--- linux-2.6.7/drivers/usb/misc/auerswald.c	2004-06-16 21:10:08.000000000 +0100
+++ linux-2.6.7-viroized/drivers/usb/misc/auerswald.c	2004-06-27 19:15:18.833585776 +0100
@@ -1591,7 +1591,7 @@
 	/* Error checking */
 	if (!ccp)
 		return -EIO;
-	if (*ppos)
+	if (ppos != &file->f_pos)
  		return -ESPIPE;
         if (count == 0)
 		return 0;
@@ -1723,8 +1723,8 @@
 	/* Error checking */
 	if (!ccp)
 		return -EIO;
-        if (*ppos)
-		return -ESPIPE;
+	if (ppos != &file->f_pos)
+ 		return -ESPIPE;
         if (len == 0)
                 return 0;
 
