Index: linux-2.6.5/drivers/scsi/osst.c
===================================================================
--- linux-2.6.5.orig/drivers/scsi/osst.c
+++ linux-2.6.5/drivers/scsi/osst.c
@@ -3187,7 +3187,7 @@ static ssize_t osst_write(struct file * 
 	
 	if (ppos != &filp->f_pos) {
 	 	/* "A request was outside the capabilities of the device." */
-		retval = (-ENXIO);
+		retval = (-ESPIPE);
 		goto out;
 	}
 
@@ -3514,7 +3514,7 @@ static ssize_t osst_read(struct file * f
 	
 	if (ppos != &filp->f_pos) {
 		/* "A request was outside the capabilities of the device." */
-		retval = (-ENXIO);
+		retval = (-ESPIPE);
 		goto out;
 	}
 
@@ -5386,6 +5386,7 @@ static struct file_operations osst_fops 
 	.owner =        THIS_MODULE,
 	.read =         osst_read,
 	.write =        osst_write,
+	.llseek = 	no_llseek,
 	.ioctl =        osst_ioctl,
 	.open =         os_scsi_tape_open,
 	.flush =        os_scsi_tape_flush,
