--- linux-2.6.7/drivers/isdn/divert/divert_procfs.c	2004-06-16 21:10:00.000000000 +0100
+++ linux-2.6.7-viroized/drivers/isdn/divert/divert_procfs.c	2004-06-27 18:42:56.176914480 +0100
@@ -79,6 +79,7 @@
 isdn_divert_read(struct file *file, char *buf, size_t count, loff_t * off)
 {
 	struct divert_info *inf;
+	loff_t pos = *off;
 	int len;
 
 	if (!*((struct divert_info **) file->private_data)) {
@@ -94,8 +95,8 @@
 	if ((len = strlen(inf->info_start)) <= count) {
 		if (copy_to_user(buf, inf->info_start, len))
 			return -EFAULT;
-		file->f_pos += len;
-		return (len);
+		*off = pos + len;
+		return len;
 	}
 	return (0);
 }				/* isdn_divert_read */
