--- linux-2.6.7/drivers/telephony/ixj.c	2004-06-16 21:09:53.000000000 +0100
+++ linux-2.6.7-viroized/drivers/telephony/ixj.c	2004-06-27 19:15:04.192811512 +0100
@@ -2857,7 +2857,7 @@
 
 static ssize_t ixj_read(struct file * file_p, char *buf, size_t length, loff_t * ppos)
 {
-	unsigned long i = *ppos;
+	unsigned long i;
 	IXJ * j = get_ixj(NUM(file_p->f_dentry->d_inode));
 
 	DECLARE_WAITQUEUE(wait, current);
@@ -2937,7 +2937,7 @@
 
 static ssize_t ixj_write(struct file *file_p, const char *buf, size_t count, loff_t * ppos)
 {
-	unsigned long i = *ppos;
+	unsigned long i;
 	IXJ *j = file_p->private_data;
 
 	DECLARE_WAITQUEUE(wait, current);
@@ -2983,7 +2983,7 @@
 		j->flags.inwrite = 0;
 		return -EFAULT;
 	}
-       if(j->play_codec == ALAW)
+        if(j->play_codec == ALAW)
                alaw2ulaw(j->write_buffer_wp, min(count, j->write_buffer_size));
 	j->flags.inwrite = 0;
 	return min(count, j->write_buffer_size);
