--- linux-2.6.7.proc.orig/drivers/char/nvram.c
+++ linux-2.6.7.proc/drivers/char/nvram.c
@@ -288,7 +288,7 @@ nvram_write(struct file *file, const cha
 	unsigned char contents[NVRAM_BYTES];
 	loff_t i;
 	unsigned char *tmp;
-	int len;
+	int len = count;
 
 	if (len > NVRAM_BYTES)
 		len = NVRAM_BYTES;
@@ -306,7 +306,7 @@ nvram_write(struct file *file, const cha
 	if (!__nvram_check_checksum())
 		goto checksum_err;
 
-	for (tmp = contents; count-- > 0 && i < NVRAM_BYTES; ++i, ++tmp)
+	for (tmp = contents; len-- > 0 && i < NVRAM_BYTES; ++i, ++tmp)
 		__nvram_write_byte(*tmp, i);
 
 	__nvram_set_checksum();
@@ -318,7 +318,6 @@ nvram_write(struct file *file, const cha
 
 checksum_err:
 	spin_unlock_irq(&rtc_lock);
- 	struct semaphore *s = &nvram_lock;	/* Single lock for now */
 	return -EIO;
 }
 
