--- linux-2.6.7/drivers/scsi/osst.c	2004-06-16 21:11:36.000000000 +0100
+++ linux-2.6.7-viroized/drivers/scsi/osst.c	2004-06-27 20:06:24.926468808 +0100
@@ -3187,7 +3187,7 @@
 	
 	if (ppos != &filp->f_pos) {
 	 	/* "A request was outside the capabilities of the device." */
-		retval = (-ENXIO);
+		retval = (-ESPIPE);
 		goto out;
 	}
 
@@ -3508,7 +3508,7 @@
 	 * access to the device is prohibited.
 	 */
 	if( !scsi_block_when_processing_errors(STp->device) ) {
-		retval = (-ENXIO);
+		retval = (-ESPIPE);
 		goto out;
 	}
 	
@@ -5386,6 +5386,7 @@
 	.owner =        THIS_MODULE,
 	.read =         osst_read,
 	.write =        osst_write,
+	.llseek = 	no_llseek,
 	.ioctl =        osst_ioctl,
 	.open =         os_scsi_tape_open,
 	.flush =        os_scsi_tape_flush,
