--- linux-2.6.7/drivers/isdn/hysdn/hysdn_proclog.c	2004-06-16 21:09:59.000000000 +0100
+++ linux-2.6.7-viroized/drivers/isdn/hysdn/hysdn_proclog.c	2004-06-27 18:45:01.807815680 +0100
@@ -210,6 +210,7 @@
 	struct proc_dir_entry *pde = PDE(file->f_dentry->d_inode);
 	struct procdata *pd = NULL;
 	hysdn_card *card;
+	loff_t pos = *off;
 
 	if (!*((struct log_data **) file->private_data)) {
 		if (file->f_flags & O_NONBLOCK)
@@ -237,7 +238,7 @@
 	if ((len = strlen(inf->log_start)) <= count) {
 		if (copy_to_user(buf, inf->log_start, len))
 			return -EFAULT;
-		file->f_pos += len;
+		*off = pos + len;
 		return (len);
 	}
 	return (0);
