Index: linux-2.6.5/drivers/isdn/i4l/isdn_common.c
===================================================================
--- linux-2.6.5.orig/drivers/isdn/i4l/isdn_common.c
+++ linux-2.6.5/drivers/isdn/i4l/isdn_common.c
@@ -947,7 +947,7 @@ isdn_info_update(void)
 }
 
 static ssize_t
-isdn_read(struct file *file, char *buf, size_t count, loff_t * off)
+isdn_read(struct file *file, char *buf, size_t count, loff_t * poff)
 {
 	uint minor = MINOR(file->f_dentry->d_inode->i_rdev);
 	int len = 0;
@@ -955,11 +955,14 @@ isdn_read(struct file *file, char *buf, 
 	int chidx;
 	int retval;
 	char *p;
+	loff_t off;
 
-	if (off != &file->f_pos)
+	if (poff != &file->f_pos)
 		return -ESPIPE;
 
 	lock_kernel();
+	off = *poff;
+	
 	if (minor == ISDN_MINOR_STATUS) {
 		if (!file->private_data) {
 			if (file->f_flags & O_NONBLOCK) {
@@ -975,7 +978,7 @@ isdn_read(struct file *file, char *buf, 
 				retval = -EFAULT;
 				goto out;
 			}
-			*off += len;
+			off += len;
 			retval = len;
 			goto out;
 		}
@@ -1004,7 +1007,7 @@ isdn_read(struct file *file, char *buf, 
 		}
 		len = isdn_readbchan(drvidx, chidx, p, 0, count,
 				     &dev->drv[drvidx]->rcv_waitq[chidx]);
-		*off += len;
+		off += len;
 		if (copy_to_user(buf,p,len)) 
 			len = -EFAULT;
 		kfree(p);
@@ -1037,7 +1040,7 @@ isdn_read(struct file *file, char *buf, 
 			dev->drv[drvidx]->stavail -= len;
 		else
 			dev->drv[drvidx]->stavail = 0;
-		*off += len;
+		off += len;
 		retval = len;
 		goto out;
 	}
@@ -1049,6 +1052,7 @@ isdn_read(struct file *file, char *buf, 
 #endif
 	retval = -ENODEV;
  out:
+ 	*poff = off;
 	unlock_kernel();
 	return retval;
 }
