--- linux-2.6.7/drivers/s390/char/tape_char.c	2004-06-16 21:11:36.000000000 +0100
+++ linux-2.6.7-viroized/drivers/s390/char/tape_char.c	2004-06-27 19:02:47.354827912 +0100
@@ -142,7 +142,10 @@
 	struct tape_request *request;
 	size_t block_size;
 	int rc;
+	loff_t pos = *ppos;
 
+	/* XXX - driver needs proper read/write/seek locks it seems */
+	
 	DBF_EVENT(6, "TCHAR:read\n");
 	device = (struct tape_device *) filp->private_data;
 	/* Check position. */
@@ -153,7 +156,7 @@
 		 * read work...
 		 */
 		DBF_EVENT(6, "TCHAR:ppos wrong\n");
-		return -EOVERFLOW;
+		return -ESPIPE;
 	}
 
 	/*
@@ -196,11 +199,13 @@
 	if (rc == 0) {
 		rc = block_size - request->rescnt;
 		DBF_EVENT(6, "TCHAR:rbytes:  %x\n", rc);
-		filp->f_pos += rc;
+		pos += rc;
 		/* Copy data from idal buffer to user space. */
 		if (idal_buffer_to_user(device->char_data.idal_buf,
 					data, rc) != 0)
 			rc = -EFAULT;
+		else
+			*ppos = pos;
 	}
 	tape_free_request(request);
 	return rc;
@@ -218,6 +223,9 @@
 	size_t written;
 	int nblocks;
 	int i, rc;
+	loff_t pos = *ppos;
+
+	/* XXX - driver needs proper read/write/seek locks it seems */
 
 	DBF_EVENT(6, "TCHAR:write\n");
 	device = (struct tape_device *) filp->private_data;
@@ -225,7 +233,7 @@
 	if (ppos != &filp->f_pos) {
 		/* "A request was outside the capabilities of the device." */
 		DBF_EVENT(6, "TCHAR:ppos wrong\n");
-		return -EOVERFLOW;
+		return -ESPIPE;
 	}
 	/* Find out block size and number of blocks */
 	if (device->char_data.block_size != 0) {
@@ -270,7 +278,7 @@
 			break;
 		DBF_EVENT(6, "TCHAR:wbytes: %lx\n",
 			  block_size - request->rescnt);
-		filp->f_pos += block_size - request->rescnt;
+		pos += block_size - request->rescnt;
 		written += block_size - request->rescnt;
 		if (request->rescnt != 0)
 			break;
@@ -297,7 +305,10 @@
 	 * tapemark it doesn't hurt to write two marks again.
 	 */
 	if (!rc)
+	{
 		device->required_tapemarks = 2;
+		*ppos = pos;
+	}
 
 	return rc ? rc : written;
 }
