Index: linux-2.6.5/fs/cifs/file.c
===================================================================
--- linux-2.6.5.orig/fs/cifs/file.c
+++ linux-2.6.5/fs/cifs/file.c
@@ -580,6 +580,7 @@ cifs_write(struct file * file, const cha
 	struct cifsTconInfo *pTcon;
 	int xid, long_op;
 	struct cifsFileInfo * open_file;
+	loff_t offset = *poffset;
 
 	if(file->f_dentry == NULL)
 		return -EBADF;
@@ -604,7 +605,7 @@ cifs_write(struct file * file, const cha
 		return -EBADF;
 	}
 
-	if (*poffset > file->f_dentry->d_inode->i_size)
+	if (offset > file->f_dentry->d_inode->i_size)
 		long_op = 2;  /* writes past end of file can take a long time */
 	else
 		long_op = 1;
@@ -645,8 +646,10 @@ cifs_write(struct file * file, const cha
 				FreeXid(xid);
 				return rc;
 			}
-		} else
-			*poffset += bytes_written;
+		} else {
+			offset += bytes_written;
+			*poffset = offset;
+		}
 		long_op = FALSE; /* subsequent writes fast - 15 seconds is plenty */
 	}
 
@@ -952,6 +955,7 @@ cifs_read(struct file * file, char *read
 	int xid;
 	char * current_offset;
 	struct cifsFileInfo * open_file;
+	loff_t loffset = * poffset;
 
 	xid = GetXid();
 	cifs_sb = CIFS_SB(file->f_dentry->d_sb);
@@ -981,7 +985,7 @@ cifs_read(struct file * file, char *read
 
 			rc = CIFSSMBRead(xid, pTcon,
 				 open_file->netfid,
-				 current_read_size, *poffset,
+				 current_read_size, loffset,
 				 &bytes_read, &current_offset);
 		}
 		if (rc || (bytes_read == 0)) {
@@ -992,7 +996,8 @@ cifs_read(struct file * file, char *read
 				return rc;
 			}
 		} else {
-			*poffset += bytes_read;
+			loffset += bytes_read;
+			*poffset = loffset;
 		}
 	}
 
