--- linux-2.6.7/fs/proc/base.c	2004-06-16 21:11:36.000000000 +0100
+++ linux-2.6.7-viroized/fs/proc/base.c	2004-06-27 19:49:08.375048576 +0100
@@ -519,6 +519,7 @@
 	ssize_t length;
 	ssize_t end;
 	struct task_struct *task = proc_task(inode);
+	loff_t pos = *ppos;
 
 	if (count > PROC_BLOCK_SIZE)
 		count = PROC_BLOCK_SIZE;
@@ -532,14 +533,14 @@
 		return length;
 	}
 	/* Static 4kB (or whatever) block capacity */
-	if (*ppos >= length) {
+	if (pos < 0 || pos >= length) {
 		free_page(page);
 		return 0;
 	}
-	if (count + *ppos > length)
-		count = length - *ppos;
-	end = count + *ppos;
-	if (copy_to_user(buf, (char *) page + *ppos, count))
+	if (count + pos > length)
+		count = length - pos;
+	end = count + pos;
+	if (copy_to_user(buf, (char *) page + pos, count))
 		count = -EFAULT;
 	else
 		*ppos = end;
