--- linux-2.6.7/fs/proc/proc_misc.c	2004-06-16 21:11:36.000000000 +0100
+++ linux-2.6.7-viroized/fs/proc/proc_misc.c	2004-06-27 19:33:48.983817288 +0100
@@ -564,12 +564,12 @@
 static ssize_t
 read_profile(struct file *file, char __user *buf, size_t count, loff_t *ppos)
 {
-	unsigned long p = *ppos;
+	loff_t p = *ppos;
 	ssize_t read;
 	char * pnt;
 	unsigned int sample_step = 1 << prof_shift;
 
-	if (p >= (prof_len+1)*sizeof(unsigned int))
+	if (p < 0 || p >= (prof_len+1)*sizeof(unsigned int))
 		return 0;
 	if (count > (prof_len+1)*sizeof(unsigned int) - p)
 		count = (prof_len+1)*sizeof(unsigned int) - p;
@@ -583,7 +583,7 @@
 	if (copy_to_user(buf,(void *)pnt,count))
 		return -EFAULT;
 	read += count;
-	*ppos += read;
+	*ppos = p;
 	return read;
 }
 
