--- linux-2.6.7/net/sunrpc/cache.c	2004-06-16 21:11:37.000000000 +0100
+++ linux-2.6.7-viroized/net/sunrpc/cache.c	2004-06-28 14:20:27.643259016 +0100
@@ -1176,7 +1176,7 @@
 
 	sprintf(tbuf, "%lu\n", cd->flush_time);
 	len = strlen(tbuf);
-	if (p >= len)
+	if (p != (unsigned int)p || p >= len)
 		return 0;
 	len -= p;
 	if (len > count) len = count;
@@ -1207,7 +1207,7 @@
 	cd->nextcheck = get_seconds();
 	cache_flush();
 
-	*ppos += count;
+	*ppos = count;
 	return count;
 }
 
