--- linux-2.6.7.proc.orig/sound/core/info.c
+++ linux-2.6.7.proc/sound/core/info.c
@@ -181,6 +181,7 @@ static ssize_t snd_info_entry_read(struc
 	struct snd_info_entry *entry;
 	snd_info_buffer_t *buf;
 	size_t size = 0;
+	loff_t pos = *offset;
 
 	data = snd_magic_cast(snd_info_private_data_t, file->private_data, return -ENXIO);
 	snd_assert(data != NULL, return -ENXIO);
@@ -190,13 +191,13 @@ static ssize_t snd_info_entry_read(struc
 		buf = data->rbuffer;
 		if (buf == NULL)
 			return -EIO;
-		if (file->f_pos >= (long)buf->size)
+		if (pos < 0 || pos >= buf->size)
 			return 0;
-		size = buf->size - file->f_pos;
+		size = buf->size - pos;
 		size = min(count, size);
-		if (copy_to_user(buffer, buf->buffer + file->f_pos, size))
+		if (copy_to_user(buffer, buf->buffer + pos, size))
 			return -EFAULT;
-		file->f_pos += size;
+		*ppos = pos + size;
 		break;
 	case SNDRV_INFO_CONTENT_DATA:
 		if (entry->c.ops->read)
@@ -215,6 +216,7 @@ static ssize_t snd_info_entry_write(stru
 	struct snd_info_entry *entry;
 	snd_info_buffer_t *buf;
 	size_t size = 0;
+	loff_t pos = *ppos;
 
 	data = snd_magic_cast(snd_info_private_data_t, file->private_data, return -ENXIO);
 	snd_assert(data != NULL, return -ENXIO);
@@ -224,17 +226,17 @@ static ssize_t snd_info_entry_write(stru
 		buf = data->wbuffer;
 		if (buf == NULL)
 			return -EIO;
-		if (file->f_pos < 0)
+		if (pos < 0)
 			return -EINVAL;
-		if (file->f_pos >= (long)buf->len)
+		if (pos >= buf->len)
 			return -ENOMEM;
-		size = buf->len - file->f_pos;
+		size = buf->len - pos;
 		size = min(count, size);
-		if (copy_from_user(buf->buffer + file->f_pos, buffer, size))
+		if (copy_from_user(buf->buffer + pos, buffer, size))
 			return -EFAULT;
-		if ((long)buf->size < file->f_pos + size)
-			buf->size = file->f_pos + size;
-		file->f_pos += size;
+		if ((long)buf->size < pos + size)
+			buf->size = pos + size;
+		*ppos = pos + size;
 		break;
 	case SNDRV_INFO_CONTENT_DATA:
 		if (entry->c.ops->write)
