Index: linux-2.6.3-rc3.orig/fs/xattr_acl.c
===================================================================
--- linux-2.6.3-rc3.orig.orig/fs/xattr_acl.c
+++ linux-2.6.3-rc3.orig/fs/xattr_acl.c
@@ -29,7 +29,7 @@ posix_acl_from_xattr(const void *value, 
 	if (size < sizeof(posix_acl_xattr_header))
 		 return ERR_PTR(-EINVAL);
 	if (header->a_version != cpu_to_le32(POSIX_ACL_XATTR_VERSION))
-		return ERR_PTR(-EINVAL);
+		return ERR_PTR(-EOPNOTSUPP);
 
 	count = posix_acl_xattr_count(size);
 	if (count < 0)
