From: Thomas Renninger <trenn@suse.de>
Subject: Change exception msg of non present hotplug CPUs to a debug msg
Patch-mainline: yes or soon
References: 179685

Signed-off-by: Thomas Renninger <trenn@suse.de>

Index: linux-2.6.16/drivers/acpi/processor_core.c
===================================================================
--- linux-2.6.16.orig/drivers/acpi/processor_core.c
+++ linux-2.6.16/drivers/acpi/processor_core.c
@@ -690,7 +690,7 @@ static int is_processor_present(acpi_han
 
 	status = acpi_evaluate_integer(handle, "_STA", NULL, &sta);
 	if (ACPI_FAILURE(status) || !(sta & ACPI_STA_PRESENT)) {
-		ACPI_EXCEPTION((AE_INFO, status, "Processor Device is not present"));
+		printk(KERN_DEBUG "Hotpluggable processor device is not present\n");
 		return_VALUE(0);
 	}
 	return_VALUE(1);
