From: Jean Delvare <khali@linux-fr.org>
Subject: ACPI: Fix a duplicate log level
Patch-mainline: not yet (submitted 2008-02-26)

Signed-off-by: Jean Delvare <jdelvare@suse.de>
---
 drivers/acpi/osl.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- linux-2.6.25-rc3.orig/drivers/acpi/osl.c	2008-02-16 10:37:18.000000000 +0100
+++ linux-2.6.25-rc3/drivers/acpi/osl.c	2008-02-27 19:08:01.000000000 +0100
@@ -1237,7 +1237,7 @@ int acpi_check_resource_conflict(struct 
 
 	if (clash) {
 		if (acpi_enforce_resources != ENFORCE_RESOURCES_NO) {
-			printk(KERN_INFO "%sACPI: %s resource %s [0x%llx-0x%llx]"
+			printk("%sACPI: %s resource %s [0x%llx-0x%llx]"
 			       " conflicts with ACPI region %s"
 			       " [0x%llx-0x%llx]\n",
 			       acpi_enforce_resources == ENFORCE_RESOURCES_LAX
