From: Thomas Renninger <trenn@suse.de>
Subject: Blacklist another Thinkpad R40e Bios to not use deeper C-states
Patch-mainline: yes or soon
References: kernel #3549

Signed-off-by: Thomas Renninger <trenn@suse.de>

--- linux/drivers/acpi/processor_idle.c.orig	2006-05-22 20:04:35.000000000 +0200
+++ linux/drivers/acpi/processor_idle.c	2006-05-29 01:10:31.000000000 +0200
@@ -99,6 +99,9 @@ static int set_max_cstate(struct dmi_sys
 static struct dmi_system_id __cpuinitdata processor_power_dmi_table[] = {
 	{ set_max_cstate, "IBM ThinkPad R40e", {
 	  DMI_MATCH(DMI_BIOS_VENDOR,"IBM"),
+	  DMI_MATCH(DMI_BIOS_VERSION,"1SET70WW")}, (void *)1},
+	{ set_max_cstate, "IBM ThinkPad R40e", {
+	  DMI_MATCH(DMI_BIOS_VENDOR,"IBM"),
 	  DMI_MATCH(DMI_BIOS_VERSION,"1SET60WW")}, (void *)1},
 	{ set_max_cstate, "IBM ThinkPad R40e", {
 	  DMI_MATCH(DMI_BIOS_VENDOR,"IBM"),
