Subject: Set _device not powermananageable_ to debug message
From: Thomas Renninger <trenn@suse.de>

 bus.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletion(-)

Signed-off-by: Thomas Renninger <trenn@suse.de>

Index: linux-2.6.15/drivers/acpi/bus.c
===================================================================
*** linux-2.6.15.orig/drivers/acpi/bus.c
--- linux-2.6.15/drivers/acpi/bus.c
***************
*** 196,202 ****
  	/* Make sure this is a valid target state */
  
  	if (!device->flags.power_manageable) {
! 		ACPI_INFO((AE_INFO, "Device is not power manageable"));
  		return_VALUE(-ENODEV);
  	}
  	/*
--- 196,204 ----
  	/* Make sure this is a valid target state */
  
  	if (!device->flags.power_manageable) {
! 		ACPI_DEBUG_PRINT((ACPI_DB_INFO,
! 			      "Device %s is not power manageable",
! 			      device->kobj.name));
  		return_VALUE(-ENODEV);
  	}
  	/*
