From: Frank Seidel <fseidel@suse.de>
Subject: Fix check whether the module should get loaded
Patch-mainline: not yet
References: 212936

 drivers/acpi/asus_acpi.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

Signed-off-by: Thomas Renninger <trenn@suse.de>
Signed-off-by: Frank Seidel <fseidel@suse.de>

--- linux-2.6.19.orig/drivers/acpi/asus_acpi.c
+++ linux-2.6.19/drivers/acpi/asus_acpi.c
@@ -1412,7 +1412,7 @@ static int __init asus_acpi_init(void)
 	if (!asus_hotk_found) {
 		acpi_bus_unregister_driver(&asus_hotk_driver);
 		remove_proc_entry(PROC_ASUS, acpi_root_dir);
-		return result;
+		return -ENODEV;
 	}
 
 	asus_backlight_device = backlight_device_register("asus",NULL,NULL,
