From: Thomas Renninger <trenn@suse.de>
Subject: Always activate fan if it should be active, don't trust former values
Patch-mainline: no
References: 239101

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 drivers/acpi/thermal.c |    6 ++++++
 1 files changed, 6 insertions(+)

--- linux-2.6.24.orig/drivers/acpi/thermal.c
+++ linux-2.6.24/drivers/acpi/thermal.c
@@ -652,8 +652,14 @@ static void acpi_thermal_active(struct a
 			if (active->temperature > maxtemp)
 				tz->state.active_index = i;
 			maxtemp = active->temperature;
+/*
+  Allways try to enable fan, to workaround unsynchronized fans on resume
+  and possibly breaking away fans on high ACPI processing (probably due
+  locking/thread problems)
+
 			if (active->flags.enabled)
 				continue;
+*/
 			for (j = 0; j < active->devices.count; j++) {
 				result =
 				    acpi_bus_set_power(active->devices.
