
From: Andi Kleen <ak@suse.de>

Add POSIX mqueue support to x86-64.


Index: linux-2.6.5/arch/x86_64/ia32/ia32entry.S
===================================================================
--- linux-2.6.5.orig/arch/x86_64/ia32/ia32entry.S
+++ linux-2.6.5/arch/x86_64/ia32/ia32entry.S
@@ -578,6 +578,16 @@ ia32_sys_call_table:
 	.quad sys_tgkill
 	.quad compat_sys_utimes
 	.quad sys32_fadvise64_64
+	.quad sys_ni_syscall	/* sys_vserver */
+	.quad sys_ni_syscall	/* sys_mbind */
+	.quad sys_ni_syscall	/* 275 sys_get_mempolicy */
+	.quad sys_ni_syscall	/* sys_set_mempolicy */
+	.quad compat_sys_mq_open
+	.quad sys_mq_unlink
+	.quad compat_sys_mq_timedsend
+	.quad compat_sys_mq_timedreceive	/* 280 */
+	.quad compat_sys_mq_notify
+	.quad compat_sys_mq_getsetattr
 	/* don't forget to change IA32_NR_syscalls */
 ia32_syscall_end:		
 	.rept IA32_NR_syscalls-(ia32_syscall_end-ia32_sys_call_table)/8
Index: linux-2.6.5/include/asm-x86_64/unistd.h
===================================================================
--- linux-2.6.5.orig/include/asm-x86_64/unistd.h
+++ linux-2.6.5/include/asm-x86_64/unistd.h
@@ -538,8 +538,20 @@ __SYSCALL(__NR_mbind, sys_mbind)
 __SYSCALL(__NR_set_mempolicy, sys_set_mempolicy)
 #define __NR_get_mempolicy 239
 __SYSCALL(__NR_get_mempolicy, sys_get_mempolicy)
+#define __NR_mq_open		240
+__SYSCALL(__NR_mq_open, sys_mq_open)
+#define __NR_mq_unlink		241
+__SYSCALL(__NR_mq_unlink, sys_mq_unlink)
+#define __NR_mq_timedsend	242
+__SYSCALL(__NR_mq_timedsend, sys_mq_timedsend)
+#define __NR_mq_timedreceive	243
+__SYSCALL(__NR_mq_timedreceive, sys_mq_timedreceive)
+#define __NR_mq_notify		244
+__SYSCALL(__NR_mq_notify, sys_mq_notify)
+#define __NR_mq_getsetattr	245
+__SYSCALL(__NR_mq_getsetattr, sys_mq_getsetattr)
 
-#define __NR_syscall_max __NR_get_mempolicy
+#define __NR_syscall_max __NR_mq_getsetattr
 #ifndef __NO_STUBS
 
 /* user-visible error numbers are in the range -1 - -4095 */
