diff -urNp linux-2.6.4/sound/drivers/serial-u16550.c linux-2.6.4.SUSE/sound/drivers/serial-u16550.c
--- linux-2.6.4/sound/drivers/serial-u16550.c	2004-03-11 03:55:22.000000000 +0100
+++ linux-2.6.4.SUSE/sound/drivers/serial-u16550.c	2004-03-16 10:19:24.000000000 +0100
@@ -342,7 +342,7 @@ static int __init snd_uart16550_detect(s
 	unsigned char c;
 
 	/* Do some vague tests for the presence of the uart */
-	if (io_base == 0) {
+	if (io_base == 0 || io_base == SNDRV_AUTO_PORT) {
 		return -ENODEV;	/* Not configured */
 	}
 
@@ -804,10 +804,9 @@ static int __init snd_uart16550_create(s
 		return err;
 	}
 
-	if (irq >= 0) {
+	if (irq >= 0 && irq != SNDRV_AUTO_IRQ) {
 		if (request_irq(irq, snd_uart16550_interrupt,
 				SA_INTERRUPT, "Serial MIDI", (void *) uart)) {
-			uart->irq = -1;
 			snd_printk("irq %d busy. Using Polling.\n", irq);
 		} else {
 			uart->irq = irq;
diff -urNp linux-2.6.4/sound/i2c/cs8427.c linux-2.6.4.SUSE/sound/i2c/cs8427.c
--- linux-2.6.4/sound/i2c/cs8427.c	2004-03-11 03:55:21.000000000 +0100
+++ linux-2.6.4.SUSE/sound/i2c/cs8427.c	2004-03-16 10:19:24.000000000 +0100
@@ -553,6 +553,7 @@ int snd_cs8427_iec958_pcm(snd_i2c_device
 			       SNDRV_CTL_EVENT_MASK_VALUE,
 			       &chip->playback.pcm_ctl->id);
 	reset = chip->rate != rate;
+	chip->rate = rate;
 	snd_i2c_unlock(cs8427->bus);
 	if (reset)
 		snd_cs8427_reset(cs8427);
diff -urNp linux-2.6.4/sound/pci/emu10k1/emu10k1_callback.c linux-2.6.4.SUSE/sound/pci/emu10k1/emu10k1_callback.c
--- linux-2.6.4/sound/pci/emu10k1/emu10k1_callback.c	2004-03-16 10:18:16.000000000 +0100
+++ linux-2.6.4.SUSE/sound/pci/emu10k1/emu10k1_callback.c	2004-03-16 10:19:24.000000000 +0100
@@ -426,7 +426,7 @@ start_voice(snd_emux_voice_t *vp)
 		snd_emu10k1_ptr_write(hw, CDF, ch, sample);
 
 		/* invalidate maps */
-		temp = ((unsigned int)hw->silent_page_dmaaddr << 1) | MAP_PTI_MASK;
+		temp = ((unsigned int)hw->silent_page.addr << 1) | MAP_PTI_MASK;
 		snd_emu10k1_ptr_write(hw, MAPA, ch, temp);
 		snd_emu10k1_ptr_write(hw, MAPB, ch, temp);
 		
diff -urNp linux-2.6.4/sound/pci/emu10k1/memory.c linux-2.6.4.SUSE/sound/pci/emu10k1/memory.c
--- linux-2.6.4/sound/pci/emu10k1/memory.c	2004-03-16 10:18:16.000000000 +0100
+++ linux-2.6.4.SUSE/sound/pci/emu10k1/memory.c	2004-03-16 10:19:24.000000000 +0100
@@ -62,7 +62,7 @@ static inline void set_silent_ptb(emu10k
 	page *= UNIT_PAGES;
 	for (i = 0; i < UNIT_PAGES; i++, page++)
 		/* do not increment ptr */
-		__set_ptb_entry(emu, page, emu->silent_page_dmaaddr);
+		__set_ptb_entry(emu, page, emu->silent_page.addr);
 }
 #endif /* PAGE_SIZE */
 
diff -urNp linux-2.6.4/sound/pci/es1968.c linux-2.6.4.SUSE/sound/pci/es1968.c
--- linux-2.6.4/sound/pci/es1968.c	2004-03-16 10:18:16.000000000 +0100
+++ linux-2.6.4.SUSE/sound/pci/es1968.c	2004-03-16 10:19:24.000000000 +0100
@@ -162,10 +162,10 @@ MODULE_PARM_DESC(clock, "Clock on " CARD
 MODULE_PARM_SYNTAX(clock, SNDRV_ENABLED);
 MODULE_PARM(use_pm, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(use_pm, "Toggle power-management.  (0 = off, 1 = on, 2 = auto)");
-MODULE_PARM_SYNTAX(use_pm, SNDRV_ENABLED ",allows:{{0,1,2}},skill:advanced");
+MODULE_PARM_SYNTAX(use_pm, SNDRV_ENABLED ",allows:{{0,1,2}},default:2,skill:advanced");
 MODULE_PARM(enable_mpu, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(enable_mpu, "Enable MPU401.  (0 = off, 1 = on, 2 = auto)");
-MODULE_PARM_SYNTAX(enable_mpu, SNDRV_ENABLED "," SNDRV_BOOLEAN_TRUE_DESC);
+MODULE_PARM_SYNTAX(enable_mpu, SNDRV_ENABLED ",allows:{{0,2}},default:2");
 #ifdef SUPPORT_JOYSTICK
 MODULE_PARM(joystick, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(joystick, "Enable joystick.");
diff -urNp linux-2.6.4/sound/pci/trident/trident_memory.c linux-2.6.4.SUSE/sound/pci/trident/trident_memory.c
--- linux-2.6.4/sound/pci/trident/trident_memory.c	2004-03-16 10:18:16.000000000 +0100
+++ linux-2.6.4.SUSE/sound/pci/trident/trident_memory.c	2004-03-16 10:19:24.000000000 +0100
@@ -76,8 +76,8 @@ static inline void set_tlb_bus(trident_t
 static inline void set_silent_tlb(trident_t *trident, int page)
 {
 	page <<= 1;
-	__set_tlb_bus(trident, page, (unsigned long)trident->tlb.silent_page, trident->tlb.silent_page_dmaaddr);
-	__set_tlb_bus(trident, page+1, (unsigned long)trident->tlb.silent_page, trident->tlb.silent_page_dmaaddr);
+	__set_tlb_bus(trident, page, (unsigned long)trident->tlb.silent_page.area, trident->tlb.silent_page.addr);
+	__set_tlb_bus(trident, page+1, (unsigned long)trident->tlb.silent_page.area, trident->tlb.silent_page.addr);
 }
 
 #else
@@ -111,7 +111,7 @@ static inline void set_silent_tlb(triden
 	int i;
 	page *= UNIT_PAGES;
 	for (i = 0; i < UNIT_PAGES; i++, page++)
-		__set_tlb_bus(trident, page, (unsigned long)trident->tlb.silent_page, trident->tlb.silent_page_dmaaddr);
+		__set_tlb_bus(trident, page, (unsigned long)trident->tlb.silent_page.area, trident->tlb.silent_page.addr);
 }
 
 #endif /* PAGE_SIZE */
