diff -urNp linux-2.6.4/sound/pci/es1968.c linux-2.6.4.SUSE/sound/pci/es1968.c
--- linux-2.6.4/sound/pci/es1968.c	2004-03-16 16:14:47.000000000 +0100
+++ linux-2.6.4.SUSE/sound/pci/es1968.c	2004-03-16 16:14:58.000000000 +0100
@@ -162,10 +162,10 @@ MODULE_PARM_DESC(clock, "Clock on " CARD
 MODULE_PARM_SYNTAX(clock, SNDRV_ENABLED);
 MODULE_PARM(use_pm, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(use_pm, "Toggle power-management.  (0 = off, 1 = on, 2 = auto)");
-MODULE_PARM_SYNTAX(use_pm, SNDRV_ENABLED ",allows:{{0,1,2}},skill:advanced");
+MODULE_PARM_SYNTAX(use_pm, SNDRV_ENABLED ",allows:{{0,1,2}},default:2,skill:advanced");
 MODULE_PARM(enable_mpu, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(enable_mpu, "Enable MPU401.  (0 = off, 1 = on, 2 = auto)");
-MODULE_PARM_SYNTAX(enable_mpu, SNDRV_ENABLED "," SNDRV_BOOLEAN_TRUE_DESC);
+MODULE_PARM_SYNTAX(enable_mpu, SNDRV_ENABLED ",allows:{{0,2}},default:2");
 #ifdef SUPPORT_JOYSTICK
 MODULE_PARM(joystick, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(joystick, "Enable joystick.");
diff -urNp linux-2.6.4/sound/pci/intel8x0.c linux-2.6.4.SUSE/sound/pci/intel8x0.c
--- linux-2.6.4/sound/pci/intel8x0.c	2004-03-16 16:14:47.000000000 +0100
+++ linux-2.6.4.SUSE/sound/pci/intel8x0.c	2004-03-16 16:14:58.000000000 +0100
@@ -96,7 +96,7 @@ MODULE_PARM_DESC(ac97_clock, "AC'97 code
 MODULE_PARM_SYNTAX(ac97_clock, SNDRV_ENABLED ",default:0");
 MODULE_PARM(ac97_quirk, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(ac97_quirk, "AC'97 workaround for strange hardware.");
-MODULE_PARM_SYNTAX(ac97_quirk, SNDRV_ENABLED ",allows:{{-1,3}},dialog:list,default:-1");
+MODULE_PARM_SYNTAX(ac97_quirk, SNDRV_ENABLED ",allows:{{-1,4}},dialog:list,default:-1");
 #ifdef SUPPORT_JOYSTICK
 MODULE_PARM(joystick, "1-" __MODULE_STRING(SNDRV_CARDS) "i");
 MODULE_PARM_DESC(joystick, "Enable joystick for Intel i8x0 soundcard.");
