Index: linux-2.6.3/sound/pci/mixart/mixart_core.c
===================================================================
--- linux-2.6.3.orig/sound/pci/mixart/mixart_core.c	2004-03-10 21:35:26.000000000 +0000
+++ linux-2.6.3/sound/pci/mixart/mixart_core.c	2004-03-10 22:31:22.000000000 +0000
@@ -91,7 +91,7 @@
 	}
 	size -= MSG_DESCRIPTOR_SIZE;
 
-	memcpy_fromio(resp->data, MIXART_MEM(mgr, msg_frame_address + MSG_HEADER_SIZE ), size);
+	memcpy_fromio(resp->data, (void *)MIXART_MEM(mgr, msg_frame_address + MSG_HEADER_SIZE ), size);
 	resp->size = size;
 
 	/* swap if necessary */
