Index: linux-2.6.4/Documentation/DocBook/parportbook.tmpl
===================================================================
--- linux-2.6.4.orig/Documentation/DocBook/parportbook.tmpl
+++ linux-2.6.4/Documentation/DocBook/parportbook.tmpl
@@ -2730,7 +2730,7 @@ to permit their use in free software.
 
 </book>
 <!-- Additional function to be documented:
---! Ddrivers/parport/init.c (this file doesn't exist any more)
+    Ddrivers/parport/init.c (this file doesn't exist any more)
 -->
 <!-- Local Variables: -->
 <!-- sgml-indent-step: 1 -->
