--- ./drivers/net/e1000/e1000_main.c.orig	2004-05-05 13:47:06.000000000 -0500
+++ ./drivers/net/e1000/e1000_main.c	2004-05-07 15:53:37.682823200 -0500
@@ -536,9 +536,6 @@ e1000_probe(struct pci_dev *pdev,
 	INIT_WORK(&adapter->tx_timeout_task,
 		(void (*)(void *))e1000_tx_timeout_task, netdev);
 
-	if((err = register_netdev(netdev)))
-		goto err_register;
-
 	/* we're going to reset, so assume we have no link for now */
 
 	netif_carrier_off(netdev);
@@ -579,6 +576,9 @@ e1000_probe(struct pci_dev *pdev,
 
 	e1000_reset(adapter);
 
+	if((err = register_netdev(netdev)))
+		goto err_register;
+
 	cards_found++;
 	return 0;
 
