diff -urNp linux-2.6.8/drivers/net/e1000/e1000_main.c linux-2.6.8.SUSE/drivers/net/e1000/e1000_main.c
--- linux-2.6.8/drivers/net/e1000/e1000_main.c	2004-08-30 16:11:06.569980528 +0200
+++ linux-2.6.8.SUSE/drivers/net/e1000/e1000_main.c	2004-08-30 16:12:12.869367809 +0200
@@ -1708,7 +1708,7 @@ e1000_tx_queue(struct e1000_adapter *ada
 	 * know there are new descriptors to fetch.  (Only
 	 * applicable for weak-ordered memory model archs,
 	 * such as IA-64). */
-	wmb();
+	mb();
 
 	tx_ring->next_to_use = i;
 	E1000_WRITE_REG(&adapter->hw, TDT, i);
@@ -2209,6 +2209,9 @@ e1000_clean_tx_irq(struct e1000_adapter 
 					       buffer_info->dma,
 					       buffer_info->length,
 					       PCI_DMA_TODEVICE);
+
+				wmb();
+
 				buffer_info->dma = 0;
 			}
 
@@ -2405,7 +2408,7 @@ e1000_alloc_rx_buffers(struct e1000_adap
 			 * know there are new descriptors to fetch.  (Only
 			 * applicable for weak-ordered memory model archs,
 			 * such as IA-64). */
-			wmb();
+			mb();
 
 			E1000_WRITE_REG(&adapter->hw, RDT, i);
 		}
