Index: linux-2.6.7/drivers/net/e100.c
===================================================================
--- linux-2.6.7.orig/drivers/net/e100.c	2004-08-05 13:46:40.187702572 +0200
+++ linux-2.6.7/drivers/net/e100.c	2004-08-05 13:52:01.672526745 +0200
@@ -828,6 +828,8 @@
 	mb();
 	cb->prev->command &= cpu_to_le16(~cb_s);
 
+	mb();
+
 	while(nic->cb_to_send != nic->cb_to_use) {
 		if(unlikely(e100_exec_cmd(nic, nic->cuc_cmd,
 			nic->cb_to_send->dma_addr))) {
