--- ./drivers/net/e1000/e1000_main.c.orig	2004-05-05 13:47:06.000000000 -0500
+++ ./drivers/net/e1000/e1000_main.c	2004-05-07 15:53:37.682823200 -0500
@@ -559,15 +559,11 @@ e1000_probe(struct pci_dev *pdev,
 	INIT_WORK(&adapter->tx_timeout_task,
 		(void (*)(void *))e1000_tx_timeout_task, netdev);
 
-	if((err = register_netdev(netdev)))
-		goto err_register;
-
 	/* we're going to reset, so assume we have no link for now */
 
 	netif_carrier_off(netdev);
 	netif_stop_queue(netdev);
 
-	DPRINTK(PROBE, INFO, "Intel(R) PRO/1000 Network Connection\n");
 	e1000_check_options(adapter);
 
 	/* Initial Wake on LAN setting
@@ -601,6 +597,11 @@ e1000_probe(struct pci_dev *pdev,
 
 	e1000_reset(adapter);
 
+	if((err = register_netdev(netdev)))
+		goto err_register;
+	
+	DPRINTK(PROBE, INFO, "Intel(R) PRO/1000 Network Connection\n");
+	
 	cards_found++;
 	return 0;
 
