--- linux-2.6.7-mm6/init/main.c.orig	2004-07-09 11:01:04.000000000 +0200
+++ linux-2.6.7-mm6/init/main.c	2004-07-09 11:14:19.000000000 +0200
@@ -93,6 +93,7 @@
 extern void populate_rootfs(void);
 extern void driver_init(void);
 extern void prepare_namespace(void);
+extern void usermodehelper_init(void);
 
 #ifdef CONFIG_TC
 extern void tc_init(void);
@@ -599,6 +600,10 @@
  */
 static void __init do_basic_setup(void)
 {
+	/* drivers will send hotplug events */
+	init_workqueues();
+	usermodehelper_init();
+
 	driver_init();
 
 #ifdef CONFIG_SYSCTL
@@ -608,7 +613,6 @@
 	/* Networking initialization needs a process context */ 
 	sock_init();
 
-	init_workqueues();
 	do_initcalls();
 }
 
--- linux-2.6.5/kernel/kmod.c.orig	2004-07-28 14:12:52.867485229 +0200
+++ linux-2.6.5/kernel/kmod.c	2004-07-28 14:23:13.758635645 +0200
@@ -263,13 +263,11 @@
 }
 EXPORT_SYMBOL(call_usermodehelper);
 
-static __init int usermodehelper_init(void)
+__init void usermodehelper_init(void)
 {
 	khelper_wq = create_singlethread_workqueue("khelper");
 	BUG_ON(!khelper_wq);
-	return 0;
 }
-__initcall(usermodehelper_init);
 #else
 int call_usermodehelper(char *path, char **argv, char **envp, int wait)
 {
