Index: linux-2.6.5/drivers/block/ll_rw_blk.c
===================================================================
--- linux-2.6.5.orig/drivers/block/ll_rw_blk.c
+++ linux-2.6.5/drivers/block/ll_rw_blk.c
@@ -145,6 +145,8 @@ struct backing_dev_info *blk_get_backing
 	return ret;
 }
 
+EXPORT_SYMBOL(blk_get_backing_dev_info);
+
 void blk_queue_activity_fn(request_queue_t *q, activity_fn *fn, void *data)
 {
 	q->activity_fn = fn;
