From: Andreas Gruenbacher <agruen@suse.de>
Subject: Fix another gcc4 compile-time error

Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

Index: linux-2.6.11/drivers/net/hamradio/baycom_epp.c
===================================================================
--- linux-2.6.11.orig/drivers/net/hamradio/baycom_epp.c
+++ linux-2.6.11/drivers/net/hamradio/baycom_epp.c
@@ -400,3 +400,3 @@ static inline void do_kiss_params(struct
 #define ENCODEITERA(j)                         \
-({                                             \
+do {                                             \
         if (!(notbitstream & (0x1f0 << j)))    \
@@ -404,6 +404,6 @@ static inline void do_kiss_params(struct
   encodeend##j:    	;                      \
-})
+} while(0)
 
 #define ENCODEITERB(j)                                          \
-({                                                              \
+do {                                                              \
   stuff##j:                                                     \
@@ -415,3 +415,3 @@ static inline void do_kiss_params(struct
         goto encodeend##j;                                      \
-})
+} while(0)
 
@@ -633,3 +633,3 @@ static void do_rxpacket(struct net_devic
 #define DECODEITERA(j)                                                        \
-({                                                                            \
+do {                                                                            \
         if (!(notbitstream & (0x0fc << j)))              /* flag or abort */  \
@@ -639,6 +639,6 @@ static void do_rxpacket(struct net_devic
   enditer##j:      ;                                                           \
-})
+} while(0)
 
 #define DECODEITERB(j)                                                                 \
-({                                                                                     \
+do {                                                                                     \
   flgabrt##j:                                                                          \
@@ -661,3 +661,3 @@ static void do_rxpacket(struct net_devic
         goto enditer##j;                                                               \
-})
+} while(0)
         
