diff -u linux-2.5.69/arch/x86_64/kernel/x8664_ksyms.c linux/arch/x86_64/kernel/x8664_ksyms.c
--- linux-2.5.69/arch/x86_64/kernel/x8664_ksyms.c	2003-05-08 13:30:01.000000000 +0200
+++ linux/arch/x86_64/kernel/x8664_ksyms.c	2003-05-08 13:55:11.000000000 +0200
@@ -110,6 +110,8 @@
 EXPORT_SYMBOL(mmx_clear_page);
 EXPORT_SYMBOL(mmx_copy_page);
 #endif
+EXPORT_SYMBOL(clear_page);
+EXPORT_SYMBOL(copy_page);
 
 #ifdef CONFIG_SMP
 EXPORT_SYMBOL(cpu_data);
