--- linux-2.6.8/drivers/usb/input/hid-input.c.orig	2004-09-23 09:27:14.000000000 +0200
+++ linux-2.6.8/drivers/usb/input/hid-input.c	2004-09-23 13:19:22.677275836 +0200
@@ -598,7 +598,7 @@
 				hidinput->input.id.vendor = dev->descriptor.idVendor;
 				hidinput->input.id.product = dev->descriptor.idProduct;
 				hidinput->input.id.version = dev->descriptor.bcdDevice;
-				hidinput->input.dev = &hid->intf->dev;
+				hidinput->input.dev = get_device(&hid->intf->dev);
 				sprintf(hidinput->input.cdev.class_id,"usbhid%d",
 					hidinput_num++);
 			}
@@ -641,6 +641,7 @@
 
 	list_for_each_safe (lh, next, &hid->inputs) {
 		hidinput = list_entry(lh, struct hid_input, list);
+		put_device(hidinput->input.dev);
 		input_unregister_device(&hidinput->input);
 		list_del(&hidinput->list);
 		kfree(hidinput);
--- linux-2.6.8/drivers/input/input.c.orig	2004-09-24 16:40:09.049148803 +0200
+++ linux-2.6.8/drivers/input/input.c	2004-09-24 16:41:14.860405217 +0200
@@ -487,7 +487,9 @@
 
 static void input_device_class_release( struct class_device *class_dev )
 {
+#if 0
 	put_device(class_dev->dev);
+#endif
 }
 
 static struct class input_device_class = {
