--- linux-2.6.5-7.21/include/linux/cpu.h	2004-05-05 12:40:27.000000000 -0500
+++ linux-2.6.5-7.21.new/include/linux/cpu.h	2004-05-06 17:45:55.000000000 -0500
@@ -60,7 +60,8 @@ extern struct semaphore cpucontrol;
 #define unlock_cpu_hotplug()	up(&cpucontrol)
 #define lock_cpu_hotplug_interruptible() down_interruptible(&cpucontrol)
 #define hotcpu_notifier(fn, pri) {				\
-	static struct notifier_block fn##_nb = { fn, pri };	\
+	static struct notifier_block fn##_nb =			\
+		{ .notifier_call = fn, .priority = pri };	\
 	register_cpu_notifier(&fn##_nb);			\
 }
 int cpu_down(unsigned int cpu);
