diff -x infiniband -purNX /tmp/kernel_exclude.txt linux-2.6.4.orig/arch/ia64/hp/common/sba_iommu.c linux-2.6.4/arch/ia64/hp/common/sba_iommu.c
--- linux-2.6.4.orig/arch/ia64/hp/common/sba_iommu.c	2004-03-15 16:43:19.174619000 +0000
+++ linux-2.6.4/arch/ia64/hp/common/sba_iommu.c	2004-03-15 17:43:20.318750450 +0000
@@ -1055,13 +1055,13 @@ sba_alloc_coherent (struct device *dev, 
 	*dma_handle = virt_to_phys(addr);
 
 #ifdef ALLOW_IOV_BYPASS
-	ASSERT(to_pci_dev(dev)->consistent_dma_mask);
+	ASSERT(to_pci_dev(dev)->dev.coherent_dma_mask);
 	/*
  	** Check if the PCI device can DMA to ptr... if so, just return ptr
  	*/
-	if (likely((*dma_handle & ~to_pci_dev(dev)->consistent_dma_mask) == 0)) {
+	if (likely((*dma_handle & ~to_pci_dev(dev)->dev.coherent_dma_mask) == 0)) {
 		DBG_BYPASS("sba_alloc_coherent() bypass mask/addr: 0x%lx/0x%lx\n",
-		           to_pci_dev(dev)->consistent_dma_mask, *dma_handle);
+		           to_pci_dev(dev)->dev.coherent_dma_mask, *dma_handle);
 
 		return addr;
 	}
diff -x infiniband -purNX /tmp/kernel_exclude.txt linux-2.6.4.orig/include/asm-ia64/machvec.h linux-2.6.4/include/asm-ia64/machvec.h
--- linux-2.6.4.orig/include/asm-ia64/machvec.h	2004-03-15 16:43:19.855261000 +0000
+++ linux-2.6.4/include/asm-ia64/machvec.h	2004-03-15 16:58:56.978058141 +0000
@@ -154,10 +154,10 @@ struct ia64_machine_vector {
 	ia64_mv_dma_unmap_single *dma_unmap_single;
 	ia64_mv_dma_map_sg *dma_map_sg;
 	ia64_mv_dma_unmap_sg *dma_unmap_sg;
-	ia64_mv_dma_sync_single *dma_sync_single_for_cpu;
-	ia64_mv_dma_sync_sg *dma_sync_sg_for_cpu;
-	ia64_mv_dma_sync_single *dma_sync_single_for_device;
-	ia64_mv_dma_sync_sg *dma_sync_sg_for_device;
+	ia64_mv_dma_sync_single_for_cpu *dma_sync_single_for_cpu;
+	ia64_mv_dma_sync_sg_for_cpu *dma_sync_sg_for_cpu;
+	ia64_mv_dma_sync_single_for_device *dma_sync_single_for_device;
+	ia64_mv_dma_sync_sg_for_device *dma_sync_sg_for_device;
 	ia64_mv_dma_supported *dma_supported;
 	ia64_mv_irq_desc *irq_desc;
 	ia64_mv_irq_to_vector *irq_to_vector;
