Work around an internal compiler error.

--- linux-2.6.5.orig/include/linux/fshooks.h	2004-04-21 02:26:40.243268904 +0200
+++ linux-2.6.5/include/linux/fshooks.h	2004-04-20 23:13:30.000000000 +0200
@@ -21,7 +21,7 @@
 typedef enum {
 	false,
 	true
-} __attribute__((__mode__(__QI__))) boolean_t;
+} boolean_t __attribute__((__mode__(__QI__)));
 
 typedef union fshook_info {
 	struct fshook_generic_info {
