From: Vojtech Pavlik <vojtech@suse.cz>
Date: Fri Sep  9 13:49:50 CEST 2005
Subject: Add Sony Vaio FSC-115b to i8042 MUX blacklist
References: #80358
Patch-mainline: 2.6.17

Signed-off-by: Vojtech Pavlik <vojtech@suse.cz>


Index: linux-2.6.13/drivers/input/serio/i8042-x86ia64io.h
===================================================================
--- linux-2.6.13.orig/drivers/input/serio/i8042-x86ia64io.h
+++ linux-2.6.13/drivers/input/serio/i8042-x86ia64io.h
@@ -151,6 +151,13 @@ static struct dmi_system_id __initdata i
 			DMI_MATCH(DMI_PRODUCT_NAME, "Sentia"),
 		},
 	},
+	{
+		.ident = "Sony Vaio FS-115b",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Sony Corporation"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "VGN-FS115B"),
+		},
+	},
 	{ }
 };
 
