From: Vojtech Pavlik <vojtech@suse.cz>
Date: Fri Sep  9 13:49:50 CEST 2005
Subject: Add Sony Vaio FSC-115b to i8042 MUX blacklist
References: #80358

Signed-off-by: Vojtech Pavlik <vojtech@suse.cz>


Index: dmitry/drivers/input/serio/i8042-x86ia64io.h
===================================================================
--- dmitry.orig/drivers/input/serio/i8042-x86ia64io.h	2005-09-05 11:36:43.000000000 +0200
+++ dmitry/drivers/input/serio/i8042-x86ia64io.h	2005-09-09 13:49:28.000000000 +0200
@@ -158,6 +158,15 @@
 			DMI_MATCH(DMI_PRODUCT_NAME, "Sentia"),
 		},
 	},
+	{
+		.ident = "Sony Vaio FS-115b",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Sony Corporation"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "VGN-FS115B"),
+		},
+	},
+	{
+		.ident = "",
 	{ }
 };
 
