diff -rup linux-2.6.8/fs/isofs/inode.c linux-2.6.8.2/fs/isofs/inode.c
--- linux-2.6.8/fs/isofs/inode.c	2004-08-15 18:44:40.082882384 -0400
+++ linux-2.6.8.2/fs/isofs/inode.c	2004-08-15 18:49:41.381078096 -0400
@@ -715,6 +715,7 @@ root_found:
 	  }
 
 	s->s_magic = ISOFS_SUPER_MAGIC;
+	s->s_maxbytes = 0xffffffff; /* We can handle files up to 4 GB */
 
 	/* The CDROM is read-only, has no nodes (devices) on it, and since
 	   all of the files appear to be owned by root, we really do not want
