diff -urp linux-sles9-beta4/fs/jfs/jfs_mount.c linux/fs/jfs/jfs_mount.c
--- linux-sles9-beta4/fs/jfs/jfs_mount.c	2004-05-13 13:45:36.000000000 -0500
+++ linux/fs/jfs/jfs_mount.c	2004-05-17 09:28:43.366769840 -0500
@@ -288,6 +288,11 @@ int jfs_mount_rw(struct super_block *sb,
 	 */
 	logMOUNT(sb);
 
+	/*
+	 * Set page cache allocation policy
+	 */
+	mapping_set_gfp_mask(sb->s_bdev->bd_inode->i_mapping, GFP_NOFS);
+
 	return rc;
 }
 
diff -urp linux-sles9-beta4/fs/jfs/jfs_umount.c linux/fs/jfs/jfs_umount.c
--- linux-sles9-beta4/fs/jfs/jfs_umount.c	2004-05-13 13:45:36.000000000 -0500
+++ linux/fs/jfs/jfs_umount.c	2004-05-17 09:30:25.343267056 -0500
@@ -121,7 +121,10 @@ int jfs_umount(struct super_block *sb)
 	 * list (to signify skip logredo()).
 	 */
 	if (log) {		/* log = NULL if read-only mount */
-		rc = updateSuper(sb, FM_CLEAN);
+		updateSuper(sb, FM_CLEAN);
+
+		/* Restore default gfp_mask for bdev */
+		mapping_set_gfp_mask(bdev_mapping, GFP_USER);
 
 		/*
 		 * close log: 
@@ -168,5 +171,8 @@ int jfs_umount_rw(struct super_block *sb
 
 	updateSuper(sb, FM_CLEAN);
 
+	/* Restore default gfp_mask for bdev */
+	mapping_set_gfp_mask(bdev_mapping, GFP_USER);
+
 	return lmLogClose(sb);
 }
