From: Andreas Gruenbacher <agruen@suse.de>
Subject: Add missing EXPORT_SYMBOL(kdb_current_task)

Used in kdbm_x86.ko.

Index: linux-2.6.16/kdb/kdbmain.c
===================================================================
--- linux-2.6.16.orig/kdb/kdbmain.c
+++ linux-2.6.16/kdb/kdbmain.c
@@ -64,6 +64,7 @@ static volatile int kdb_new_cpu;		/* Whi
 volatile int kdb_state[NR_CPUS];		/* Per cpu state */
 
 const struct task_struct *kdb_current_task;
+EXPORT_SYMBOL(kdb_current_task);
 struct pt_regs *kdb_current_regs;
 
 #ifdef	CONFIG_KDB_OFF
