diff -Nru a/lib/kobject.c b/lib/kobject.c
--- a/lib/kobject.c	Tue Dec 16 17:10:16 2003
+++ b/lib/kobject.c	Tue Dec 16 17:10:16 2003
@@ -344,16 +344,16 @@
 		/* 
 		 * Need more space? Allocate it and try again 
 		 */
-		name = kmalloc(need,GFP_KERNEL);
+		limit = need + 1;
+		name = kmalloc(limit,GFP_KERNEL);
 		if (!name) {
 			error = -ENOMEM;
 			goto Done;
 		}
-		limit = need;
 		need = vsnprintf(name,limit,fmt,args);
 
 		/* Still? Give up. */
-		if (need > limit) {
+		if (need >= limit) {
 			kfree(name);
 			error = -EFAULT;
 			goto Done;
