From: agraf@suse.de
Subject: Ignore apic polarity
Patch-mainline: unknown
References: bnc#556564

---
 virt/kvm/ioapic.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/virt/kvm/ioapic.c
+++ b/virt/kvm/ioapic.c
@@ -202,7 +202,8 @@ int kvm_ioapic_set_irq(struct kvm_ioapic
 	old_irr = ioapic->irr;
 	if (irq >= 0 && irq < IOAPIC_NUM_PINS) {
 		entry = ioapic->redirtbl[irq];
-		level ^= entry.fields.polarity;
+// polarity is always active high in qemu
+//		level ^= entry.fields.polarity;
 		if (!level)
 			ioapic->irr &= ~mask;
 		else {
