--- linux/drivers/scsi/libata-core.c~	2004-02-18 17:06:21.718963718 +0100
+++ linux/drivers/scsi/libata-core.c	2004-02-18 17:06:52.780615572 +0100
@@ -2953,8 +2953,10 @@
 
 err_out:
 	for (i = 0; i < count; i++) {
-		ata_host_remove(host_set->ports[i], 1);
-		scsi_host_put(host_set->ports[i]->host);
+		if (host_set->ports[i]) {
+			ata_host_remove(host_set->ports[i], 1);
+			scsi_host_put(host_set->ports[i]->host);
+		}
 	}
 	kfree(host_set);
 	VPRINTK("EXIT, returning 0\n");

-- 
Jens Axboe

