diff -purN linux-2.6.5/include/linux/pci_ids.h linux-2.6.6/include/linux/pci_ids.h
--- linux-2.6.5/include/linux/pci_ids.h	2004-04-04 03:36:57.000000000 +0000
+++ linux-2.6.6/include/linux/pci_ids.h	2004-05-10 02:32:28.000000000 +0000
@@ -1871,6 +1881,13 @@
 #define PCI_DEVICE_ID_ALTIMA_AC9100	0x03ea
 #define PCI_DEVICE_ID_ALTIMA_AC1003	0x03eb
 
+#define PCI_VENDOR_ID_S2IO		0x17d5
+#define	PCI_DEVICE_ID_S2IO_WIN		0x5731
+#define	PCI_DEVICE_ID_S2IO_UNI		0x5831
+
+#define PCI_VENDOR_ID_ARC               0x192E
+#define PCI_DEVICE_ID_ARC_EHCI          0x0101
+
 #define PCI_VENDOR_ID_SYMPHONY		0x1c1c
 #define PCI_DEVICE_ID_SYMPHONY_101	0x0001
 
