diff -purN linux-2.6.5/drivers/media/dvb/ttusb-budget/dvb-ttusb-budget.c linux-2.6.6/drivers/media/dvb/ttusb-budget/dvb-ttusb-budget.c
--- linux-2.6.5/drivers/media/dvb/ttusb-budget/dvb-ttusb-budget.c	2004-04-04 03:38:22.000000000 +0000
+++ linux-2.6.6/drivers/media/dvb/ttusb-budget/dvb-ttusb-budget.c	2004-05-10 02:33:21.000000000 +0000
@@ -1008,7 +1010,6 @@ static int ttusb_stop_feed(struct dvb_de
 
 static int ttusb_setup_interfaces(struct ttusb *ttusb)
 {
-	usb_set_configuration(ttusb->dev, 1);
 	usb_set_interface(ttusb->dev, 1, 1);
 
 	ttusb->bulk_out_pipe = usb_sndbulkpipe(ttusb->dev, 1);
@@ -1077,6 +1078,19 @@ static int ttusb_probe(struct usb_interf
 
 	udev = interface_to_usbdev(intf);
 
+	/* Device has already been reset; its configuration was chosen.
+	 * If this fault happens, use a hotplug script to choose the
+	 * right configuration (write bConfigurationValue in sysfs).
+	 */
+	if (udev->actconfig->desc.bConfigurationValue != 1) {
+		dev_err(&intf->dev, "device config is #%d, need #1\n",
+			udev->actconfig->desc.bConfigurationValue);
+		return -ENODEV;
+	}
+
+
+        if (intf->altsetting->desc.bInterfaceNumber != 1) return -ENODEV;
+
 	if (!(ttusb = kmalloc(sizeof(struct ttusb), GFP_KERNEL)))
 		return -ENOMEM;
 
