From: Andreas Gruenbacher <agruen@suse.de>
Subject: Add missing symbol exports

Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

Index: linux-2.6.12/arch/ia64/kernel/traps.c
===================================================================
--- linux-2.6.12.orig/arch/ia64/kernel/traps.c
+++ linux-2.6.12/arch/ia64/kernel/traps.c
@@ -43,6 +43,7 @@ int register_die_notifier(struct notifie
 	spin_unlock_irqrestore(&die_notifier_lock, flags);
 	return err;
 }
+EXPORT_SYMBOL(register_die_notifier);
 
 void __init
 trap_init (void)
Index: linux-2.6.12/arch/ppc64/kernel/traps.c
===================================================================
--- linux-2.6.12.orig/arch/ppc64/kernel/traps.c
+++ linux-2.6.12/arch/ppc64/kernel/traps.c
@@ -74,6 +74,7 @@ int register_die_notifier(struct notifie
 	spin_unlock_irqrestore(&die_notifier_lock, flags);
 	return err;
 }
+EXPORT_SYMBOL(register_die_notifier);
 
 /*
  * Trap & Exception support
Index: linux-2.6.12/arch/sparc64/kernel/traps.c
===================================================================
--- linux-2.6.12.orig/arch/sparc64/kernel/traps.c
+++ linux-2.6.12/arch/sparc64/kernel/traps.c
@@ -53,6 +53,7 @@ int register_die_notifier(struct notifie
 	spin_unlock_irqrestore(&die_notifier_lock, flags);
 	return err;
 }
+EXPORT_SYMBOL(register_die_notifier);
 
 /* When an irrecoverable trap occurs at tl > 0, the trap entry
  * code logs the trap state registers at every level in the trap
