From: Andreas Gruenbacher <agruen@suse.de>
Subject: Add missing symbol exports
Patch-mainline: yes

Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

Index: linux-2.6.14/arch/sparc64/kernel/traps.c
===================================================================
--- linux-2.6.14.orig/arch/sparc64/kernel/traps.c	2005-10-28 02:02:08.000000000 +0200
+++ linux-2.6.14/arch/sparc64/kernel/traps.c	2005-12-01 16:00:38.199861238 +0100
@@ -54,6 +54,7 @@ int register_die_notifier(struct notifie
 	spin_unlock_irqrestore(&die_notifier_lock, flags);
 	return err;
 }
+EXPORT_SYMBOL(register_die_notifier);
 
 /* When an irrecoverable trap occurs at tl > 0, the trap entry
  * code logs the trap state registers at every level in the trap
Index: linux-2.6.14/arch/powerpc/kernel/traps.c
===================================================================
--- linux-2.6.14.orig/arch/powerpc/kernel/traps.c	2005-12-01 16:00:29.551859216 +0100
+++ linux-2.6.14/arch/powerpc/kernel/traps.c	2005-12-01 16:00:38.199861238 +0100
@@ -86,6 +86,7 @@ int register_die_notifier(struct notifie
 	spin_unlock_irqrestore(&die_notifier_lock, flags);
 	return err;
 }
+EXPORT_SYMBOL(register_die_notifier);
 
 /*
  * Trap & Exception support
