Index: linux-2.6.3/arch/i386/kernel/i386_ksyms.c
===================================================================
--- linux-2.6.3.orig/arch/i386/kernel/i386_ksyms.c	2004-02-18 03:59:52.000000000 +0000
+++ linux-2.6.3/arch/i386/kernel/i386_ksyms.c	2004-02-25 21:29:38.000000000 +0000
@@ -207,3 +207,7 @@
 #endif
 
 EXPORT_SYMBOL(csum_partial);
+
+#ifdef CONFIG_X86_GENERICARCH
+EXPORT_SYMBOL(genapic);
+#endif
