Index: linux-2.6.3/arch/i386/mach-generic/probe.c
===================================================================
--- linux-2.6.3.orig/arch/i386/mach-generic/probe.c
+++ linux-2.6.3/arch/i386/mach-generic/probe.c
@@ -20,6 +20,7 @@ extern struct genapic apic_bigsmp;
 extern struct genapic apic_default;
 
 struct genapic *genapic = &apic_default;
+EXPORT_SYMBOL(genapic);
 
 struct genapic *apic_probe[] __initdata = { 
 	&apic_summit,
