diff -pur linux-2.6.4-sysfs/drivers/net/amd8111e.c linux-2.6.4/drivers/net/amd8111e.c
--- linux-2.6.4-sysfs/drivers/net/amd8111e.c	2004-03-29 17:31:53.000000000 +0200
+++ linux-2.6.4/drivers/net/amd8111e.c	2004-03-29 19:46:32.000000000 +0200
@@ -1845,6 +1845,7 @@ static int __devinit amd8111e_probe_one(
 	}
 
 	SET_MODULE_OWNER(dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 #if AMD8111E_VLAN_TAG_USED
 	dev->features |= NETIF_F_HW_VLAN_TX | NETIF_F_HW_VLAN_RX ;
diff -pur linux-2.6.4-sysfs/drivers/net/irda/donauboe.c linux-2.6.4/drivers/net/irda/donauboe.c
--- linux-2.6.4-sysfs/drivers/net/irda/donauboe.c	2004-03-11 03:55:26.000000000 +0100
+++ linux-2.6.4/drivers/net/irda/donauboe.c	2004-03-29 19:53:15.000000000 +0200
@@ -1713,6 +1713,7 @@ toshoboe_open (struct pci_dev *pci_dev, 
 #endif
 
   SET_MODULE_OWNER(dev);
+  SET_NETDEV_DEV(dev, &pci_dev->dev);
   dev->hard_start_xmit = toshoboe_hard_xmit;
   dev->open = toshoboe_net_open;
   dev->stop = toshoboe_net_close;
diff -pur linux-2.6.4-sysfs/drivers/net/irda/vlsi_ir.c linux-2.6.4/drivers/net/irda/vlsi_ir.c
--- linux-2.6.4-sysfs/drivers/net/irda/vlsi_ir.c	2004-03-29 17:31:54.000000000 +0200
+++ linux-2.6.4/drivers/net/irda/vlsi_ir.c	2004-03-29 19:55:28.000000000 +0200
@@ -1735,6 +1735,8 @@ static int vlsi_irda_init(struct net_dev
 	ndev->tx_timeout      = vlsi_tx_timeout;
 	ndev->watchdog_timeo  = 500*HZ/1000;	/* max. allowed turn time for IrLAP */
 
+	SET_NETDEV_DEV(ndev, &pdev->dev);
+
 	return 0;
 }	
 
diff -pur linux-2.6.4-sysfs/drivers/net/saa9730.c linux-2.6.4/drivers/net/saa9730.c
--- linux-2.6.4-sysfs/drivers/net/saa9730.c	2004-03-11 03:55:28.000000000 +0100
+++ linux-2.6.4/drivers/net/saa9730.c	2004-03-29 19:56:41.000000000 +0200
@@ -1146,6 +1146,7 @@ static int __devinit saa9730_init_one(st
 	}
 
 	pci_set_drvdata(pdev, dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 	return 0;
 	
 out2:
diff -pur linux-2.6.4-sysfs/drivers/net/sk98lin/skge.c linux-2.6.4/drivers/net/sk98lin/skge.c
--- linux-2.6.4-sysfs/drivers/net/sk98lin/skge.c	2004-03-29 17:31:54.000000000 +0200
+++ linux-2.6.4/drivers/net/sk98lin/skge.c	2004-03-29 19:57:20.000000000 +0200
@@ -382,6 +382,7 @@ static int __init skge_probe (void)
 		dev->do_ioctl =		&SkGeIoctl;
 		dev->change_mtu =	&SkGeChangeMtu;
 		dev->flags &= 		~IFF_RUNNING;
+		SET_NETDEV_DEV(dev, &pdev->dev);
 
 #ifdef SK_ZEROCOPY
 #ifdef USE_SK_TX_CHECKSUM
diff -pur linux-2.6.4-sysfs/drivers/net/tc35815.c linux-2.6.4/drivers/net/tc35815.c
--- linux-2.6.4-sysfs/drivers/net/tc35815.c	2004-03-11 03:55:20.000000000 +0100
+++ linux-2.6.4/drivers/net/tc35815.c	2004-03-29 19:58:26.000000000 +0200
@@ -597,6 +597,7 @@ static int __devinit tc35815_probe1(stru
 	dev->get_stats		= tc35815_get_stats;
 	dev->set_multicast_list = tc35815_set_multicast_list;
 	SET_MODULE_OWNER(dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	ret = register_netdev(dev);
 	if (ret)
diff -pur linux-2.6.4-sysfs/drivers/net/tokenring/3c359.c linux-2.6.4/drivers/net/tokenring/3c359.c
--- linux-2.6.4-sysfs/drivers/net/tokenring/3c359.c	2004-03-29 17:31:54.000000000 +0200
+++ linux-2.6.4/drivers/net/tokenring/3c359.c	2004-03-29 19:58:45.000000000 +0200
@@ -345,6 +345,7 @@ int __devinit xl_probe(struct pci_dev *p
 	dev->get_stats=&xl_get_stats ;
 	dev->set_mac_address=&xl_set_mac_address ; 
 	SET_MODULE_OWNER(dev); 
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	pci_set_drvdata(pdev,dev) ; 
 	if ((i = register_netdev(dev))) { 
diff -pur linux-2.6.4-sysfs/drivers/net/tokenring/abyss.c linux-2.6.4/drivers/net/tokenring/abyss.c
--- linux-2.6.4-sysfs/drivers/net/tokenring/abyss.c	2004-03-29 17:31:54.000000000 +0200
+++ linux-2.6.4/drivers/net/tokenring/abyss.c	2004-03-29 19:59:12.000000000 +0200
@@ -167,6 +167,7 @@ static int __devinit abyss_attach(struct
 	dev->stop = abyss_close;
 
 	pci_set_drvdata(pdev, dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	ret = register_netdev(dev);
 	if (ret)
diff -pur linux-2.6.4-sysfs/drivers/net/tokenring/lanstreamer.c linux-2.6.4/drivers/net/tokenring/lanstreamer.c
--- linux-2.6.4-sysfs/drivers/net/tokenring/lanstreamer.c	2004-03-11 03:55:51.000000000 +0100
+++ linux-2.6.4/drivers/net/tokenring/lanstreamer.c	2004-03-29 19:59:44.000000000 +0200
@@ -332,6 +332,7 @@ static int __devinit streamer_init_one(s
 	dev->set_mac_address = &streamer_set_mac_address;
 	dev->irq = pdev->irq;
 	dev->base_addr=pio_start;
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	streamer_priv->streamer_card_name = (char *)pdev->resource[0].name;
 	streamer_priv->pci_dev = pdev;
diff -pur linux-2.6.4-sysfs/drivers/net/tokenring/olympic.c linux-2.6.4/drivers/net/tokenring/olympic.c
--- linux-2.6.4-sysfs/drivers/net/tokenring/olympic.c	2004-03-29 17:31:54.000000000 +0200
+++ linux-2.6.4/drivers/net/tokenring/olympic.c	2004-03-29 20:00:10.000000000 +0200
@@ -259,6 +259,7 @@ static int __devinit olympic_probe(struc
 	dev->get_stats=&olympic_get_stats ;
 	dev->set_mac_address=&olympic_set_mac_address ;  
 	SET_MODULE_OWNER(dev) ; 
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	pci_set_drvdata(pdev,dev) ; 
 	register_netdev(dev) ; 
diff -pur linux-2.6.4-sysfs/drivers/net/tokenring/tmspci.c linux-2.6.4/drivers/net/tokenring/tmspci.c
--- linux-2.6.4-sysfs/drivers/net/tokenring/tmspci.c	2004-03-11 03:55:44.000000000 +0100
+++ linux-2.6.4/drivers/net/tokenring/tmspci.c	2004-03-29 20:00:31.000000000 +0200
@@ -164,6 +164,7 @@ static int __devinit tms_pci_attach(stru
 	dev->open = tms380tr_open;
 	dev->stop = tms380tr_close;
 	pci_set_drvdata(pdev, dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	ret = register_netdev(dev);
 	if (ret)
diff -pur linux-2.6.4-sysfs/drivers/net/wan/sbni.c linux-2.6.4/drivers/net/wan/sbni.c
--- linux-2.6.4-sysfs/drivers/net/wan/sbni.c	2004-03-11 03:55:25.000000000 +0100
+++ linux-2.6.4/drivers/net/wan/sbni.c	2004-03-29 20:09:26.000000000 +0200
@@ -333,8 +333,10 @@ sbni_pci_probe( struct net_device  *dev 
 			release_region( pci_ioaddr, SBNI_IO_EXTENT );
 			return  -EIO;
 		}
-		if( sbni_probe1( dev, pci_ioaddr, pci_irq_line ) )
+		if( sbni_probe1( dev, pci_ioaddr, pci_irq_line ) ) {
+			SET_NETDEV_DEV(dev, &pdev->dev);
 			return  0;
+		}
 	}
 	return  -ENODEV;
 }
diff -pur linux-2.6.4-sysfs/drivers/net/wireless/airo.c linux-2.6.4/drivers/net/wireless/airo.c
--- linux-2.6.4-sysfs/drivers/net/wireless/airo.c	2004-03-29 17:31:55.000000000 +0200
+++ linux-2.6.4/drivers/net/wireless/airo.c	2004-03-30 00:10:56.000000000 +0200
@@ -2727,6 +2727,11 @@
 	dev->irq = irq;
 	dev->base_addr = port;
 
+	/* what is with PCMCIA ??? */
+	if (pci) {
+		SET_NETDEV_DEV(dev, &pci->dev);
+	}
+
 	if (test_bit(FLAG_MPI,&ai->flags))
 		reset_mpi_card (dev);
 
diff -pur linux-2.6.4-sysfs/drivers/net/wireless/orinoco_pci.c linux-2.6.4/drivers/net/wireless/orinoco_pci.c
--- linux-2.6.4-sysfs/drivers/net/wireless/orinoco_pci.c	2004-03-11 03:55:36.000000000 +0100
+++ linux-2.6.4/drivers/net/wireless/orinoco_pci.c	2004-03-29 20:11:25.000000000 +0200
@@ -220,6 +220,7 @@ static int orinoco_pci_init_one(struct p
 	dev->mem_end = pci_iorange + pci_iolen - 1;
 
 	SET_MODULE_OWNER(dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	printk(KERN_DEBUG
 	       "Detected Orinoco/Prism2 PCI device at %s, mem:0x%lX to 0x%lX -> 0x%p, irq:%d\n",
diff -pur linux-2.6.4-sysfs/drivers/net/wireless/orinoco_plx.c linux-2.6.4/drivers/net/wireless/orinoco_plx.c
--- linux-2.6.4-sysfs/drivers/net/wireless/orinoco_plx.c	2004-03-11 03:55:49.000000000 +0100
+++ linux-2.6.4/drivers/net/wireless/orinoco_plx.c	2004-03-29 20:11:58.000000000 +0200
@@ -233,6 +233,7 @@ static int orinoco_plx_init_one(struct p
 	priv = dev->priv;
 	dev->base_addr = pccard_ioaddr;
 	SET_MODULE_OWNER(dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	printk(KERN_DEBUG
 	       "Detected Orinoco/Prism2 PLX device at %s irq:%d, io addr:0x%lx\n",
diff -pur linux-2.6.4-sysfs/drivers/net/wireless/orinoco_tmd.c linux-2.6.4/drivers/net/wireless/orinoco_tmd.c
--- linux-2.6.4-sysfs/drivers/net/wireless/orinoco_tmd.c	2004-03-11 03:55:22.000000000 +0100
+++ linux-2.6.4/drivers/net/wireless/orinoco_tmd.c	2004-03-29 20:12:08.000000000 +0200
@@ -125,6 +125,7 @@ static int orinoco_tmd_init_one(struct p
 	priv = dev->priv;
 	dev->base_addr = pccard_ioaddr;
 	SET_MODULE_OWNER(dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 
 	printk(KERN_DEBUG
 	       "Detected Orinoco/Prism2 TMD device at %s irq:%d, io addr:0x%lx\n",
diff -ur linux-2.6.4.org/drivers/net/3c509.c linux/drivers/net/3c509.c
--- linux-2.6.4.org/drivers/net/3c509.c	2004-03-30 09:44:31.000000000 +0200
+++ linux/drivers/net/3c509.c	2004-03-30 12:18:07.728637288 +0200
@@ -426,6 +426,7 @@
 			}
 
 			SET_MODULE_OWNER(dev);
+			SET_NETDEV_DEV(dev, &idev->dev);
 			pnp_cards++;
 
 			netdev_boot_setup_check(dev);
diff -ur linux-2.6.4.org/drivers/net/3c515.c linux/drivers/net/3c515.c
--- linux-2.6.4.org/drivers/net/3c515.c	2004-03-11 03:55:22.000000000 +0100
+++ linux/drivers/net/3c515.c	2004-03-30 12:17:58.924975648 +0200
@@ -538,6 +538,7 @@
 		     		inl(ioaddr + 0x2002), inw(ioaddr + 0x2000));
 			/* irq = inw(ioaddr + 0x2002) & 15; */ /* Use the irq from isapnp */
 			corkscrew_setup(dev, ioaddr, idev, cards_found++);
+			SET_NETDEV_DEV(dev, &idev->dev);
 			pnp_cards++;
 			err = register_netdev(dev);
 			if (!err)
