--- 2.6.6-mm5/arch/ppc64/kernel/smp.c	2004-05-23 09:48:40.000000000 -0500
+++ 2.6.6-mm5-ltc8794/arch/ppc64/kernel/smp.c	2004-05-23 23:25:48.000000000 -0500
@@ -261,6 +261,9 @@ int __cpu_disable(void)
 	/* FIXME: go put this in a header somewhere */
 	extern void xics_migrate_irqs_away(void);
 
+	if (systemcfg->platform != PLATFORM_PSERIES_LPAR)
+		return -ENOSYS;
+
 	systemcfg->processorCount--;
 
 	/*fix boot_cpuid here*/
@@ -881,6 +884,10 @@ int __devinit __cpu_up(unsigned int cpu)
 	if (system_state == SYSTEM_BOOTING && !cpu_present_at_boot(cpu))
 		return -ENOENT;
 
+	if (system_state != SYSTEM_BOOTING &&
+	    systemcfg->platform != PLATFORM_PSERIES_LPAR)
+		return -ENOSYS;
+
 	paca[cpu].prof_counter = 1;
 	paca[cpu].prof_multiplier = 1;
 	paca[cpu].default_decr = tb_ticks_per_jiffy / decr_overclock;
