From: Pankaj Biswas <bpankaj@novell.com>
Subject: Fix checking of login id
References: bnc#626119
Patch-mainline: No

Signed-off-by: Pankaj Biswas <bpankaj@novell.com>
Acked-by: Jan Kara <jack@suse.cz>

--- a/fs/novfs/vfs.h	2011-02-23 17:15:27.000000000 +0530
+++ b/fs/novfs/vfs.h	2011-03-04 11:57:46.000000000 +0530
@@ -54,6 +54,8 @@ struct novfs_schandle {
 
 #define PATH_LENGTH_BUFFER	PATH_MAX
 #define NW_MAX_PATH_LENGTH	255
+#define NW_MAX_DN_CHARS        256
+#define NW_MAX_DN_BYTES        (2*(NW_MAX_DN_CHARS+1))
 
 #define XA_BUFFER		(8 * 1024)
 
--- a/fs/novfs/nwcapi.c	2011-03-02 12:00:46.000000000 +0530
+++ b/fs/novfs/nwcapi.c	2011-03-04 11:57:42.000000000 +0530
@@ -480,7 +480,7 @@ int novfs_login_id(struct novfs_xplat *p
 	DbgPrint("DomainName\n");
 	novfs_dump(sizeof(nwcStr), &nwcStr);
 
-	if (nwcStr.DataLen > MAX_NAME_LEN)
+	if (nwcStr.DataLen > MAX_DOMAIN_LEN)
 		return -EINVAL;
 
 	if ((server.buffer = kmalloc(nwcStr.DataLen, GFP_KERNEL))) {
@@ -492,7 +492,7 @@ int novfs_login_id(struct novfs_xplat *p
 
 			cpylen = copy_from_user(&nwcStr, lgn.pObjectName, sizeof(nwcStr));
 			DbgPrint("ObjectName");
-			if (nwcStr.DataLen > MAX_OBJECT_NAME_LENGTH) {
+			if (nwcStr.DataLen > NW_MAX_DN_BYTES) {
 				retCode = -EINVAL;
 				goto out;
 			}
