diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/arch/i386/mm/pageattr.c sles/arch/i386/mm/pageattr.c
--- sles-ref/arch/i386/mm/pageattr.c	2004-05-21 06:34:52.058411440 +0200
+++ sles/arch/i386/mm/pageattr.c	2004-05-21 06:37:05.921061224 +0200
@@ -118,20 +118,20 @@ __change_page_attr(struct page *page, pg
 			pte_t standard = mk_pte(page, PAGE_KERNEL); 
 			set_pte_atomic(kpte, mk_pte(page, prot)); 
 			if (pte_same(old,standard))
-				atomic_inc(&kpte_page->count);
+				get_page(kpte_page);
 		} else {
 			struct page *split = split_large_page(address, prot); 
 			if (!split)
 				return -ENOMEM;
-			atomic_inc(&kpte_page->count);
+			get_page(kpte_page);
 			set_pmd_pte(kpte,address,mk_pte(split, PAGE_KERNEL));
 		}	
 	} else if ((pte_val(*kpte) & _PAGE_PSE) == 0) { 
 		set_pte_atomic(kpte, mk_pte(page, PAGE_KERNEL));
-		atomic_dec(&kpte_page->count); 
+		__put_page(kpte_page);
 	}
 
-	if (cpu_has_pse && (atomic_read(&kpte_page->count) == 1)) { 
+	if (cpu_has_pse && (page_count(kpte_page) == 1)) {
 		list_add(&kpte_page->lru, &df_list);
 		revert_page(kpte_page, address);
 	} 
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/arch/mips/mm/init.c sles/arch/mips/mm/init.c
--- sles-ref/arch/mips/mm/init.c	2004-03-11 08:27:26.000000000 +0100
+++ sles/arch/mips/mm/init.c	2004-05-21 06:37:17.799255464 +0200
@@ -235,7 +235,7 @@ void __init mem_init(void)
 		}
 		ClearPageReserved(page);
 		set_bit(PG_highmem, &page->flags);
-		atomic_set(&page->count, 1);
+		set_page_count(page, 1);
 		__free_page(page);
 		totalhigh_pages++;
 	}
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/arch/ppc/mm/init.c sles/arch/ppc/mm/init.c
--- sles-ref/arch/ppc/mm/init.c	2004-02-20 17:26:33.000000000 +0100
+++ sles/arch/ppc/mm/init.c	2004-05-21 06:37:05.922061072 +0200
@@ -458,7 +458,7 @@ void __init mem_init(void)
 
 			ClearPageReserved(page);
 			set_bit(PG_highmem, &page->flags);
-			atomic_set(&page->count, 1);
+			set_page_count(page, 1);
 			__free_page(page);
 			totalhigh_pages++;
 		}
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/arch/sparc/mm/init.c sles/arch/sparc/mm/init.c
--- sles-ref/arch/sparc/mm/init.c	2004-04-04 08:08:57.000000000 +0200
+++ sles/arch/sparc/mm/init.c	2004-05-21 06:37:05.924060768 +0200
@@ -389,7 +389,7 @@ void map_high_region(unsigned long start
 
 		ClearPageReserved(page);
 		set_bit(PG_highmem, &page->flags);
-		atomic_set(&page->count, 1);
+		set_page_count(page, 1);
 		__free_page(page);
 		totalhigh_pages++;
 	}
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/arch/um/kernel/mem.c sles/arch/um/kernel/mem.c
--- sles-ref/arch/um/kernel/mem.c	2004-05-21 06:33:59.454408464 +0200
+++ sles/arch/um/kernel/mem.c	2004-05-21 06:37:33.775826656 +0200
@@ -56,7 +56,7 @@ static void setup_highmem(unsigned long 
 		page = &mem_map[highmem_pfn + i];
 		ClearPageReserved(page);
 		set_bit(PG_highmem, &page->flags);
-		atomic_set(&page->count, 1);
+		set_page_count(page, 1);
 		__free_page(page);
 	}
 }
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/arch/x86_64/mm/pageattr.c sles/arch/x86_64/mm/pageattr.c
--- sles-ref/arch/x86_64/mm/pageattr.c	2004-05-21 06:33:42.056053416 +0200
+++ sles/arch/x86_64/mm/pageattr.c	2004-05-21 06:38:25.159015224 +0200
@@ -130,21 +130,21 @@ __change_page_attr(unsigned long address
 
 			set_pte(kpte, mk_pte(page, prot)); 
 			if (pte_same(old,standard))
-				atomic_inc(&kpte_page->count);
+				get_page(kpte_page);
 		} else {
 			struct page *split;
 			split = split_large_page(address, prot, ref_prot); 
 			if (!split)
 				return -ENOMEM;
-			atomic_inc(&kpte_page->count);
+			get_page(kpte_page);
 			set_pte(kpte,mk_pte(split, ref_prot));
 		}	
 	} else if ((kpte_flags & _PAGE_PSE) == 0) { 
 		set_pte(kpte, mk_pte(page, ref_prot));
-		atomic_dec(&kpte_page->count); 
+		__put_page(kpte_page);
 	}
 
-	if (atomic_read(&kpte_page->count) == 1) { 
+	if (page_count(kpte_page) == 1) {
 		save_page(address, kpte_page); 		     
 		revert_page(kpte_page, address, ref_prot);
 	} 
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/drivers/char/drm/ati_pcigart.h sles/drivers/char/drm/ati_pcigart.h
--- sles-ref/drivers/char/drm/ati_pcigart.h	2003-07-17 01:54:05.000000000 +0200
+++ sles/drivers/char/drm/ati_pcigart.h	2004-05-21 06:37:05.931059704 +0200
@@ -67,7 +67,7 @@ static unsigned long DRM(ati_alloc_pciga
 	page = virt_to_page( address );
 
 	for ( i = 0 ; i < ATI_PCIGART_TABLE_PAGES ; i++, page++ ) {
-		atomic_inc( &page->count );
+		get_page(page);
 		SetPageReserved( page );
 	}
 
@@ -84,7 +84,7 @@ static void DRM(ati_free_pcigart_table)(
 	page = virt_to_page( address );
 
 	for ( i = 0 ; i < ATI_PCIGART_TABLE_PAGES ; i++, page++ ) {
-		atomic_dec( &page->count );
+		__put_page(page);
 		ClearPageReserved( page );
 	}
 
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/drivers/char/drm/drm_vm.h sles/drivers/char/drm/drm_vm.h
--- sles-ref/drivers/char/drm/drm_vm.h	2004-01-15 18:36:09.000000000 +0100
+++ sles/drivers/char/drm/drm_vm.h	2004-05-21 06:39:21.651427072 +0200
@@ -131,7 +131,7 @@ struct page *DRM(vm_nopage)(struct vm_ar
 
 		DRM_DEBUG("baddr = 0x%lx page = 0x%p, offset = 0x%lx, count=%d\n",
 			  baddr, __va(agpmem->memory->memory[offset]), offset,
-			  atomic_read(&page->count));
+			  page_count(page));
 
 		if (type)
 			*type = VM_FAULT_MINOR;
diff -urNp --exclude CVS --exclude BitKeeper --exclude {arch} --exclude .arch-ids sles-ref/drivers/scsi/sg.c sles/drivers/scsi/sg.c
--- sles-ref/drivers/scsi/sg.c	2004-05-21 06:34:31.937470288 +0200
+++ sles/drivers/scsi/sg.c	2004-05-21 06:37:05.930059856 +0200
@@ -1113,10 +1113,10 @@ sg_rb_correct4mmap(Sg_scatter_hold * rsv
 				page_ptr = sg_scatg2virt(sclp) + m;
 				page = virt_to_page(page_ptr);
 				if (startFinish)
-					atomic_inc(&page->count);
+					get_page(page);
 				else {
 					if (page_count(page) > 0)
-						atomic_dec(&page->count);
+						__put_page(page);
 				}
 			}
 		}
@@ -1125,10 +1125,10 @@ sg_rb_correct4mmap(Sg_scatter_hold * rsv
 			page_ptr = (unsigned char *) rsv_schp->buffer + m;
 			page = virt_to_page(page_ptr);
 			if (startFinish)
-				atomic_inc(&page->count);
+				get_page(page);
 			else {
 				if (page_count(page) > 0)
-					atomic_dec(&page->count);
+					__put_page(page);
 			}
 		}
 	}
