From: Joachim Gleissner <jg@suse.de>
Subject: fix build of pcmcia modules
Patch-mainline: no
References: none

fix build of pcmcia modules

Acked-by: mantel@suse.de

diff -urNp linux-2.6.12/include/pcmcia/ds.h linux-2.6.12.SUSE/include/pcmcia/ds.h
--- linux-2.6.12/include/pcmcia/ds.h	2005-07-11 14:41:51.990669525 +0200
+++ linux-2.6.12.SUSE/include/pcmcia/ds.h	2005-07-11 14:42:32.052278446 +0200
@@ -16,10 +16,13 @@
 #ifndef _LINUX_DS_H
 #define _LINUX_DS_H
 
+#ifdef __KERNEL
+#include <linux/mod_devicetable.h>
+#endif
+
 #include <pcmcia/bulkmem.h>
 #include <pcmcia/cs_types.h>
 #include <pcmcia/device_id.h>
-#include <linux/mod_devicetable.h>
 
 typedef struct tuple_parse_t {
     tuple_t		tuple;
