--- linux-sles9-rc5/drivers/pci/hotplug/rpaphp_pci.c	2004-07-12 14:46:55.000000000 -0500
+++ sles9-rc5/drivers/pci/hotplug/rpaphp_pci.c	2004-07-13 18:35:09.000000000 -0500
@@ -361,7 +361,7 @@ static void rpaphp_eeh_remove_bus_device
 int rpaphp_unconfig_pci_adapter(struct slot *slot)
 {
 	int retval = 0;
-	struct list_head *ln;
+	struct list_head *ln, *tmp;
 
 	dbg("Entry %s: slot[%s]\n", __FUNCTION__, slot->name);
 	if (list_empty(&slot->dev.pci_funcs)) {
@@ -372,7 +372,7 @@ int rpaphp_unconfig_pci_adapter(struct s
 		goto exit;
 	}
 	/* remove the devices from the pci core */
-	list_for_each (ln, &slot->dev.pci_funcs) {
+	list_for_each_safe (ln, tmp, &slot->dev.pci_funcs) {
 		struct rpaphp_pci_func *func;
 	
 		func = list_entry(ln, struct rpaphp_pci_func, sibling);
