Index: linux-2.6.5/kernel/sys.c
===================================================================
--- linux-2.6.5.orig/kernel/sys.c
+++ linux-2.6.5/kernel/sys.c
@@ -1499,7 +1499,7 @@ asmlinkage long sys_setdomainname(char _
 
 asmlinkage long sys_getrlimit(unsigned int resource, struct rlimit __user *rlim)
 {
-	if (resource >= RLIM_NLIMITS)
+	if (resource >= RLIM_NLIMITS - 1)
 		return -EINVAL;
 	else
 		return copy_to_user(rlim, current->rlim + resource, sizeof(*rlim))
@@ -1515,7 +1515,7 @@ asmlinkage long sys_getrlimit(unsigned i
 asmlinkage long sys_old_getrlimit(unsigned int resource, struct rlimit __user *rlim)
 {
 	struct rlimit x;
-	if (resource >= RLIM_NLIMITS)
+	if (resource >= RLIM_NLIMITS - 1)
 		return -EINVAL;
 
 	memcpy(&x, current->rlim + resource, sizeof(*rlim));
@@ -1533,7 +1533,7 @@ asmlinkage long sys_setrlimit(unsigned i
 	struct rlimit new_rlim, *old_rlim;
 	int retval;
 
-	if (resource >= RLIM_NLIMITS)
+	if (resource >= RLIM_NLIMITS - 1)
 		return audit_intercept(AUDIT_setrlimit, resource, NULL), audit_result(-EINVAL);
 	if(copy_from_user(&new_rlim, rlim, sizeof(*rlim)))
 		return -EFAULT;
