From: schwidefsky@de.ibm.com
Subject: Bug 67212 - LTC14261-raw device ioctl failed

Replace ioctl_by_bdev with blkdev_ioctl in raw.c.

Signed-off-by: Olaf Hering <olh@suse.de>

--- ./drivers/char/raw.c	17 Aug 2004 11:44:48 -0000	1.27
+++ ./drivers/char/raw.c	3 Mar 2005 13:52:05 -0000
@@ -122,7 +122,7 @@
 {
 	struct block_device *bdev = filp->private_data;
 
-	return ioctl_by_bdev(bdev, command, arg);
+	return blkdev_ioctl(bdev->bd_inode, filp, command, arg);
 }
 
 static void bind_device(struct raw_config_request *rq)
--- ../linux-2.6.5/drivers/block/ioctl.c	2005-03-07 18:19:15.123215576 +0100
+++ ./drivers/block/ioctl.c	2005-03-07 18:21:08.082033846 +0100
@@ -218,3 +218,5 @@ int blkdev_ioctl(struct inode *inode, st
 	}
 	return -ENOTTY;
 }
+
+EXPORT_SYMBOL(blkdev_ioctl);
