Index: linux-2.6.3/include/asm/spinlock.h
===================================================================
--- linux-2.6.3.orig/include/asm/spinlock.h	2004-02-18 03:57:20.000000000 +0000
+++ linux-2.6.3/include/asm/spinlock.h	2004-02-28 04:43:47.000000000 +0000
@@ -7,7 +7,7 @@
 #include <linux/config.h>
 #include <linux/compiler.h>
 
-extern int printk(const char * fmt, ...)
+extern asmlinkage int printk(const char * fmt, ...)
 	__attribute__ ((format (printf, 1, 2)));
 
 /*
