--- linux-2.6.5.1/fs/reiserfs/namei.c	2004-05-03 13:52:20.661164712 -0400
+++ linux-2.6.5.2/fs/reiserfs/namei.c	2004-05-03 13:52:29.748783184 -0400
@@ -643,7 +643,7 @@
 
 out_failed:
     if (locked)
-        reiserfs_write_unlock (dir->i_sb);
+        reiserfs_write_unlock_xattrs (dir->i_sb);
     reiserfs_write_unlock(dir->i_sb);
     return retval;
 }
@@ -715,7 +715,7 @@
 
 out_failed:
     if (locked)
-        reiserfs_write_unlock (dir->i_sb);
+        reiserfs_write_unlock_xattrs (dir->i_sb);
     reiserfs_write_unlock(dir->i_sb);
     return retval;
 }
