--- linux-2.6.5.orig/fs/reiserfs/namei.c
+++ linux-2.6.5/fs/reiserfs/namei.c
@@ -643,7 +643,7 @@
 
 out_failed:
     if (locked)
-        reiserfs_write_unlock (dir->i_sb);
+        reiserfs_write_unlock_xattrs (dir->i_sb);
     reiserfs_write_unlock(dir->i_sb);
     return retval;
 }
@@ -715,7 +715,7 @@
 
 out_failed:
     if (locked)
-        reiserfs_write_unlock (dir->i_sb);
+        reiserfs_write_unlock_xattrs (dir->i_sb);
     reiserfs_write_unlock(dir->i_sb);
     return retval;
 }
@@ -797,7 +797,7 @@
     d_instantiate(dentry, inode);
     journal_end(&th, dir->i_sb, jbegin_count) ;
 out_failed:
-    if (retval)
+    if (locked)
         reiserfs_write_unlock_xattrs (dir->i_sb);
     reiserfs_write_unlock(dir->i_sb);
     return retval;
