

Remove the now-unneeded page.list field.


Index: linux.t/include/linux/mm.h
===================================================================
--- linux.t.orig/include/linux/mm.h	2004-03-15 15:20:47.689914071 -0500
+++ linux.t/include/linux/mm.h	2004-03-15 15:20:54.686079127 -0500
@@ -178,7 +178,6 @@ struct page {
 	page_flags_t flags;		/* atomic flags, some possibly
 					   updated asynchronously */
 	atomic_t count;			/* Usage count, see below. */
-	struct list_head list;		/* ->mapping has some page lists. */
 	struct address_space *mapping;	/* The inode (or ...) we belong to. */
 	pgoff_t index;			/* Our offset within mapping. */
 	struct list_head lru;		/* Pageout list, eg. active_list;
Index: linux.t/mm/page_alloc.c
===================================================================
--- linux.t.orig/mm/page_alloc.c	2004-03-15 15:20:47.687914311 -0500
+++ linux.t/mm/page_alloc.c	2004-03-15 15:20:54.688078888 -0500
@@ -699,8 +699,6 @@ nopage:
 	return NULL;
 got_pg:
 	kernel_map_pages(page, 1 << order, 1);
-	INIT_LIST_HEAD(&page->list);
-	INIT_LIST_HEAD(&page->lru);
 	return page;
 }
 
