From: KOSAKI Motohiro <kosaki.motohiro@jp.fujitsu.com>
Subject: mm: remove UP version of lru_add_drain_all()
References: bnc#458380
Patch-mainline: yes
Git: 6841c8e26357904ef462650273f5d5015f7bb370

Acked-by: Nick Piggin <npiggin@suse.de>
---
 mm/swap.c |   13 -------------
 1 file changed, 13 deletions(-)

Index: linux-sles11/mm/swap.c
===================================================================
--- linux-sles11.orig/mm/swap.c
+++ linux-sles11/mm/swap.c
@@ -244,7 +244,6 @@ void lru_add_drain(void)
 	put_cpu();
 }
 
-#ifdef CONFIG_NUMA
 static void lru_add_drain_per_cpu(struct work_struct *dummy)
 {
 	lru_add_drain();
@@ -258,18 +257,6 @@ int lru_add_drain_all(void)
 	return schedule_on_each_cpu(lru_add_drain_per_cpu);
 }
 
-#else
-
-/*
- * Returns 0 for success
- */
-int lru_add_drain_all(void)
-{
-	lru_add_drain();
-	return 0;
-}
-#endif
-
 /*
  * Batched page_cache_release().  Decrement the reference count on all the
  * passed pages.  If it fell to zero then remove the page from the LRU and
