--- linux-2.6.0-test11-4/include/asm-s390/bitops.h.orig	2003-12-16 08:44:44.413574567 +0100
+++ linux-2.6.0-test11-4/include/asm-s390/bitops.h	2003-12-16 08:45:26.213580722 +0100
@@ -1015,7 +1015,7 @@
                 "    jnz  4f\n"
                 "    ahi  %0,1\n"
                 "4:"
-                : "=&d" (r), "+d" (x) : : "cc" );
+                : "+d" (r), "+d" (x) : : "cc" );
         return r;
 }
 
