--- ./include/scsi/scsi_device.h~	2004-05-13 18:22:20.000000000 +0200
+++ ./include/scsi/scsi_device.h	2004-05-13 18:27:37.000000000 +0200
@@ -57,7 +57,7 @@ struct scsi_device {
 
 	void *hostdata;		/* available to low-level driver */
 	char devfs_name[256];	/* devfs junk */
-	char type;
+	signed char type;
 	char scsi_level;
 	unsigned char inquiry_len;	/* valid bytes in 'inquiry' */
 	unsigned char * inquiry;	/* INQUIRY response data */
