--- linux-2.6.5.orig/Makefile	2004-05-08 13:04:16.000000000 +0200
+++ linux-2.6.5/Makefile	2004-05-08 21:14:06.808888847 +0200
@@ -1001,6 +1001,8 @@ define all-sources
 	       -name '*.[chS]' -print; \
 	  find arch/$(ARCH) $(RCS_FIND_IGNORE) \
 	       -name '*.[chS]' -print; \
+	  find security/selinux/include $(RCS_FIND_IGNORE) \
+	       -name '*.[chS]' -print; \
 	  find include $(RCS_FIND_IGNORE) \
 	       \( -name config -o -name 'asm-*' \) -prune \
 	       -o -name '*.[chS]' -print; \
