Index: linux-2.6.4/fs/namei.c
===================================================================
--- linux-2.6.4.orig/fs/namei.c
+++ linux-2.6.4/fs/namei.c
@@ -1027,8 +1027,9 @@ static inline int may_delete(struct inod
 		return error;
 	if (IS_APPEND(dir))
 		return -EPERM;
-	if (check_sticky(dir, victim->d_inode)||IS_APPEND(victim->d_inode)||
-	    IS_IMMUTABLE(victim->d_inode))
+	if (check_sticky(dir, victim->d_inode))
+		return -EACCES;
+	if (IS_APPEND(victim->d_inode) || IS_IMMUTABLE(victim->d_inode))
 		return -EPERM;
 	if (isdir) {
 		if (!S_ISDIR(victim->d_inode->i_mode))
