diff -Naurp linux-2.6.5-20040518/fs/proc/proc_misc.c linux-2.6.5-20040518-patch/fs/proc/proc_misc.c
--- linux-2.6.5-20040518/fs/proc/proc_misc.c	2004-05-19 16:05:46.166626433 -0700
+++ linux-2.6.5-20040518-patch/fs/proc/proc_misc.c	2004-05-19 15:09:56.000000000 -0700
@@ -390,7 +390,7 @@ int show_stat(struct seq_file *p, void *
 		(unsigned long long)jiffies_64_to_clock_t(iowait),
 		(unsigned long long)jiffies_64_to_clock_t(irq),
 		(unsigned long long)jiffies_64_to_clock_t(softirq));
-	for_each_cpu(i) {
+	for_each_online_cpu(i) {
 
 		/* Copy values here to work around gcc-2.95.3, gcc-2.96 */
 		user = kstat_cpu(i).cpustat.user;
