


---

 25-akpm/mm/vmscan.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN mm/vmscan.c~shrink_slab-handle-GFP_NOFS-fix mm/vmscan.c
--- 25/mm/vmscan.c~shrink_slab-handle-GFP_NOFS-fix	2004-05-01 14:34:25.446391008 -0700
+++ 25-akpm/mm/vmscan.c	2004-05-01 14:34:37.424570048 -0700
@@ -156,7 +156,7 @@ static int shrink_slab(unsigned long sca
 			shrinker->nr = LONG_MAX;	/* It wrapped! */
 
 		if (shrinker->nr <= SHRINK_BATCH)
-			break;
+			continue;
 		while (shrinker->nr) {
 			long this_scan = shrinker->nr;
 			int shrink_ret;

_
