diff -pur linux-2.6.4-sysfs/drivers/net/sk98lin/skge.c linux-2.6.4/drivers/net/sk98lin/skge.c
--- linux-2.6.4-sysfs/drivers/net/sk98lin/skge.c	2004-03-29 17:31:54.000000000 +0200
+++ linux-2.6.4/drivers/net/sk98lin/skge.c	2004-03-29 19:57:20.000000000 +0200
@@ -437,6 +437,8 @@
 	dev->change_mtu =	&SkGeChangeMtu;
 	dev->flags &= 		~IFF_RUNNING;
 
+	SET_NETDEV_DEV(dev, &pdev->dev);
+
 	pAC->Index = sk98lin_boards_found;
 
 	if (SkGeBoardInit(dev, pAC)) {
