

Remove remaining references to address_space.clean_pages.


Index: linux.t/fs/inode.c
===================================================================
--- linux.t.orig/fs/inode.c	2004-03-15 15:08:24.350828647 -0500
+++ linux.t/fs/inode.c	2004-03-15 15:09:00.553506291 -0500
@@ -176,7 +176,6 @@ void inode_init_once(struct inode *inode
 {
 	memset(inode, 0, sizeof(*inode));
 	INIT_HLIST_NODE(&inode->i_hash);
-	INIT_LIST_HEAD(&inode->i_data.clean_pages);
 	INIT_LIST_HEAD(&inode->i_dentry);
 	INIT_LIST_HEAD(&inode->i_devices);
 	sema_init(&inode->i_sem, 1);
Index: linux.t/include/linux/fs.h
===================================================================
--- linux.t.orig/include/linux/fs.h	2004-03-15 15:08:24.358827692 -0500
+++ linux.t/include/linux/fs.h	2004-03-15 15:09:00.558505694 -0500
@@ -325,7 +325,6 @@ struct address_space {
 	struct inode		*host;		/* owner: inode, block_device */
 	struct radix_tree_root	page_tree;	/* radix tree of all pages */
 	spinlock_t		tree_lock;	/* and spinlock protecting it */
-	struct list_head	clean_pages;	/* list of clean pages */
 	unsigned long		nrpages;	/* number of total pages */
 	struct address_space_operations *a_ops;	/* methods */
 	struct list_head	i_mmap;		/* list of private mappings */
Index: linux.t/include/linux/pagemap.h
===================================================================
--- linux.t.orig/include/linux/pagemap.h	2004-03-15 15:05:38.689600641 -0500
+++ linux.t/include/linux/pagemap.h	2004-03-15 15:09:00.559505575 -0500
@@ -142,7 +142,6 @@ static inline unsigned long get_page_cac
 static inline void ___add_to_page_cache(struct page *page,
 		struct address_space *mapping, unsigned long index)
 {
-	list_add(&page->list, &mapping->clean_pages);
 	page->mapping = mapping;
 	page->index = index;
 
Index: linux.t/mm/swap_state.c
===================================================================
--- linux.t.orig/mm/swap_state.c	2004-03-15 15:08:24.354828169 -0500
+++ linux.t/mm/swap_state.c	2004-03-15 15:09:00.554506172 -0500
@@ -27,7 +27,6 @@ extern struct address_space_operations s
 struct address_space swapper_space = {
 	.page_tree	= RADIX_TREE_INIT(GFP_ATOMIC),
 	.tree_lock	= SPIN_LOCK_UNLOCKED,
-	.clean_pages	= LIST_HEAD_INIT(swapper_space.clean_pages),
 	.a_ops		= &swap_aops,
 	.backing_dev_info = &swap_backing_dev_info,
 	.i_mmap		= LIST_HEAD_INIT(swapper_space.i_mmap),
