Index: linux-2.6.4/kernel/power/swsusp.c
===================================================================
--- linux-2.6.4.orig/kernel/power/swsusp.c	2004-04-01 10:12:23.000000000 +0000
+++ linux-2.6.4/kernel/power/swsusp.c	2004-04-01 13:12:12.000000000 +0000
@@ -364,7 +364,7 @@
 }
 
 struct highmem_page {
-	char *data;
+	long data;
 	struct page *page;
 	struct highmem_page *next;
 };
@@ -406,7 +406,7 @@
 		if (!save->data)
 			panic("Not enough memory");
 		kaddr = kmap_atomic(page, KM_USER0);
-		memcpy(save->data, kaddr, PAGE_SIZE);
+		memcpy((void *)save->data, kaddr, PAGE_SIZE);
 		kunmap_atomic(kaddr, KM_USER0);
 		highmem_copy = save;
 	}
@@ -422,7 +422,7 @@
 		highmem_copy = save->next;
 		
 		kaddr = kmap_atomic(save->page, KM_USER0);
-		memcpy(kaddr, save->data, PAGE_SIZE);
+		memcpy(kaddr, (void *)save->data, PAGE_SIZE);
 		kunmap_atomic(kaddr, KM_USER0);
 		free_page(save->data);
 		kfree(save);
