Subject: Bug 158568 - LTC22155-ethtool -t (self test) will cause bcm5780S to be inoperable
From: whchang@tw.ibm.com


Signed-off-by: Olaf Hering <olh@suse.de>

---
 drivers/net/tg3.c |    7 +++++++
 1 file changed, 7 insertions(+)

Index: linux-2.6.16/drivers/net/tg3.c
===================================================================
--- linux-2.6.16.orig/drivers/net/tg3.c
+++ linux-2.6.16/drivers/net/tg3.c
@@ -140,6 +140,8 @@
 
 #define TG3_NUM_TEST		6
 
+static void tg3_link_report(struct tg3 *);
+
 static char version[] __devinitdata =
 	DRV_MODULE_NAME ".c:v" DRV_MODULE_VERSION " (" DRV_MODULE_RELDATE ")\n";
 
@@ -965,6 +967,11 @@ static int tg3_phy_reset(struct tg3 *tp)
 	if (err != 0)
 		return -EBUSY;
 
+	if (netif_running(tp->dev) && netif_carrier_ok(tp->dev)) {
+		netif_carrier_off(tp->dev);
+		tg3_link_report(tp);
+	}
+
 	if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5703 ||
 	    GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5704 ||
 	    GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5705) {
