--- linux/drivers/net/tulip/tulip_core.c	19 Apr 2004 17:04:06 -0000
+++ linux.orig/drivers/net/tulip/tulip_core.c	26 Apr 2004 01:08:31 -0000
@@ -1761,11 +1761,11 @@
 		return;
 
 	tp = netdev_priv(dev);
+	unregister_netdev(dev);
 	pci_free_consistent (pdev,
 			     sizeof (struct tulip_rx_desc) * RX_RING_SIZE +
 			     sizeof (struct tulip_tx_desc) * TX_RING_SIZE,
 			     tp->rx_ring, tp->rx_ring_dma);
-	unregister_netdev (dev);
 	if (tp->mtable)
 		kfree (tp->mtable);
 #ifndef USE_IO_OPS
