From: Jiri Benc <jbenc@suse.cz> 
Subject: Make ALi (ULi) integrated 100BaseTX Ethernet work

=====================================================================================

diff -ur linux-2.6.10/drivers/net/tulip/media.c linux-2.6.10work/drivers/net/tulip/media.c
--- linux-2.6.10/drivers/net/tulip/media.c	2005-03-01 14:10:49.000000000 +0100
+++ linux-2.6.10work/drivers/net/tulip/media.c	2005-03-01 15:53:56.000000000 +0100
@@ -517,10 +517,11 @@
 		/* Enable autonegotiation: some boards default to off. */
 		if (tp->default_port == 0) {
 			new_bmcr = mii_reg0 | BMCR_ANENABLE;
-			if (new_bmcr != mii_reg0) {
-				new_bmcr |= BMCR_ANRESTART;
-				ane_switch = 1;
-			}
+			/* DM9161E PHY seems to need to restart
+			 * autonegotiation even if it defaults to enabled.
+			 */
+			new_bmcr |= BMCR_ANRESTART;
+			ane_switch = 1;
 		}
 		/* ...or disable nway, if forcing media */
 		else {
diff -ur linux-2.6.10/drivers/net/tulip/tulip_core.c linux-2.6.10work/drivers/net/tulip/tulip_core.c
--- linux-2.6.10/drivers/net/tulip/tulip_core.c	2005-03-01 14:11:21.000000000 +0100
+++ linux-2.6.10work/drivers/net/tulip/tulip_core.c	2005-03-01 15:54:31.000000000 +0100
@@ -383,6 +383,11 @@
 				goto media_picked;
 			}
 	}
+	if (tp->chip_id == ULI526X) {
+		for (i = tp->mtable->leafcount - 1; i >= 0; i--)
+			if (tulip_media_cap[tp->mtable->mleaf[i].media] & MediaIsMII)
+				goto media_picked;
+	}
 	/* Start sensing first non-full-duplex media. */
 	for (i = tp->mtable->leafcount - 1;
 		 (tulip_media_cap[tp->mtable->mleaf[i].media] & MediaAlwaysFD) && i > 0; i--)
