From: olh@suse.de
Subject: /proc/bus/usb/devices causes trouble

dont read it

diff -purNx tags ../linux-2.6.11-rc3.orig/drivers/usb/core/devices.c ./drivers/usb/core/devices.c
--- ../linux-2.6.11-rc3.orig/drivers/usb/core/devices.c	2005-02-03 02:56:53.000000000 +0100
+++ ./drivers/usb/core/devices.c	2005-02-07 10:37:39.917591731 +0100
@@ -583,6 +583,9 @@ static ssize_t usb_device_read(struct fi
 		return 0;
 	if (!access_ok(VERIFY_WRITE, buf, nbytes))
 		return -EFAULT;
+	if (!*ppos)
+		printk(KERN_ALERT "program %s is still not converted to use sysfs. slackers!\n",
+				current->comm);
 
 	down (&usb_bus_list_lock);
 	/* print devices for all busses */
