
	pigs can fly

diff -purN linux-2.6.9-rc2.orig/drivers/usb/storage/transport.c linux-2.6.9-rc2/drivers/usb/storage/transport.c
--- linux-2.6.9-rc2.orig/drivers/usb/storage/transport.c	2004-09-13 07:32:17.000000000 +0200
+++ linux-2.6.9-rc2/drivers/usb/storage/transport.c	2004-09-22 11:40:55.947309036 +0200
@@ -982,10 +982,10 @@ int usb_stor_Bulk_transport(struct scsi_
 	/* DATA STAGE */
 	/* send/receive data payload, if there is any */
 
-	/* Genesys Logic interface chips need a 100us delay between the
+	/* Genesys Logic interface chips need a 200us delay between the
 	 * command phase and the data phase */
 	if (us->pusb_dev->descriptor.idVendor == USB_VENDOR_ID_GENESYS)
-		udelay(100);
+		udelay(200);
 
 	if (transfer_length) {
 		unsigned int pipe = srb->sc_data_direction == DMA_FROM_DEVICE ? 
