From: Oliver Neukum <oneukum@suse.de>
Subject:  add MODULE_DEVICE_TABLE to cdc-wdm driver to compute a proper alias
Patch-mainline: submitted
References: bnc#433817

This makes autoloading work for this module

diff --git a/drivers/usb/class/cdc-wdm.c b/drivers/usb/class/cdc-wdm.c
index 7e8e123..023a4e9 100644
--- a/drivers/usb/class/cdc-wdm.c
+++ b/drivers/usb/class/cdc-wdm.c
@@ -42,6 +42,8 @@ static struct usb_device_id wdm_ids[] = {
 	{ }
 };
 
+MODULE_DEVICE_TABLE (usb, wdm_ids);
+
 #define WDM_MINOR_BASE	176
 
 
