Index: linux-2.6.3/arch/x86_64/kernel/x8664_ksyms.c
===================================================================
--- linux-2.6.3.orig/arch/x86_64/kernel/x8664_ksyms.c
+++ linux-2.6.3/arch/x86_64/kernel/x8664_ksyms.c
@@ -225,3 +225,6 @@ EXPORT_SYMBOL_GPL(flush_tlb_all);
 #endif
 
 EXPORT_SYMBOL(sys_ioctl);
+
+EXPORT_SYMBOL(memcpy_toio);
+EXPORT_SYMBOL(memcpy_fromio);
Index: linux-2.6.3/arch/x86_64/lib/io.c
===================================================================
--- linux-2.6.3.orig/arch/x86_64/lib/io.c
+++ linux-2.6.3/arch/x86_64/lib/io.c
@@ -11,7 +11,3 @@ void *memcpy_fromio(void *dst,const void
 {
 	return __inline_memcpy(dst,__io_virt(src),len);
 }
-
-EXPORT_SYMBOL(memcpy_toio);
-EXPORT_SYMBOL(memcpy_fromio);
-
Index: linux-2.6.3/drivers/char/sx.c
===================================================================
--- linux-2.6.3.orig/drivers/char/sx.c
+++ linux-2.6.3/drivers/char/sx.c
@@ -1025,8 +1025,9 @@ static void sx_transmit_chars (struct sx
 		if (c == 0) break;
 
 
-		memcpy_toio (port->board->base + CHAN_OFFSET(port,hi_txbuf) + tx_ip, 
-		             port->gs.xmit_buf + port->gs.xmit_tail, c);
+		memcpy_toio ((char *)port->board->base +
+			       CHAN_OFFSET(port,hi_txbuf) + tx_ip, 
+		             (char *)port->gs.xmit_buf + port->gs.xmit_tail, c);
 
 		/* Update the pointer in the card */
 		sx_write_channel_byte (port, hi_txipos, (tx_ip+c) & 0xff);
@@ -1095,7 +1096,8 @@ static inline void sx_receive_chars (str
 		            read_sx_byte (port->board, CHAN_OFFSET(port,hi_rxbuf) + rx_op),
 		            CHAN_OFFSET(port, hi_rxbuf)); 
 		memcpy_fromio (tty->flip.char_buf_ptr, 
-		               port->board->base + CHAN_OFFSET(port,hi_rxbuf) + rx_op, c);
+		               (char *)port->board->base +
+			         CHAN_OFFSET(port,hi_rxbuf) + rx_op, c);
 		memset(tty->flip.flag_buf_ptr, TTY_NORMAL, c);
 
 		/* Update the kernel buffer end */
Index: linux-2.6.3/drivers/mtd/maps/elan-104nc.c
===================================================================
--- linux-2.6.3.orig/drivers/mtd/maps/elan-104nc.c
+++ linux-2.6.3/drivers/mtd/maps/elan-104nc.c
@@ -146,7 +146,8 @@ static void elan_104nc_copy_from(struct 
 		
 		spin_lock(&elan_104nc_spin);
 		elan_104nc_page(map, from);
-		memcpy_fromio(to, iomapadr + (from & WINDOW_MASK), thislen);
+		memcpy_fromio(to, (char *)iomapadr + (from & WINDOW_MASK),
+			      thislen);
 		spin_unlock(&elan_104nc_spin);
 		to += thislen;
 		from += thislen;
@@ -187,7 +188,8 @@ static void elan_104nc_copy_to(struct ma
 		
 		spin_lock(&elan_104nc_spin);
 		elan_104nc_page(map, to);
-		memcpy_toio(iomapadr + (to & WINDOW_MASK), from, thislen);
+		memcpy_toio((char *)iomapadr + (to & WINDOW_MASK), from,
+			    thislen);
 		spin_unlock(&elan_104nc_spin);
 		to += thislen;
 		from += thislen;
Index: linux-2.6.3/drivers/mtd/maps/ich2rom.c
===================================================================
--- linux-2.6.3.orig/drivers/mtd/maps/ich2rom.c
+++ linux-2.6.3/drivers/mtd/maps/ich2rom.c
@@ -64,7 +64,7 @@ static __u32 ich2rom_read32(struct map_i
 
 static void ich2rom_copy_from(struct map_info *map, void *to, unsigned long from, ssize_t len)
 {
-	memcpy_fromio(to, addr(map, from), len);
+	memcpy_fromio(to, (void *)addr(map, from), len);
 }
 
 static void ich2rom_write8(struct map_info *map, __u8 d, unsigned long ofs)
@@ -87,7 +87,7 @@ static void ich2rom_write32(struct map_i
 
 static void ich2rom_copy_to(struct map_info *map, unsigned long to, const void *from, ssize_t len)
 {
-	memcpy_toio(addr(map, to), from, len);
+	memcpy_toio((void *)addr(map, to), from, len);
 }
 
 static struct ich2rom_map_info ich2rom_map = {
Index: linux-2.6.3/drivers/mtd/maps/map_funcs.c
===================================================================
--- linux-2.6.3.orig/drivers/mtd/maps/map_funcs.c
+++ linux-2.6.3/drivers/mtd/maps/map_funcs.c
@@ -70,12 +70,12 @@ static void simple_map_write64(struct ma
 
 static void simple_map_copy_from(struct map_info *map, void *to, unsigned long from, ssize_t len)
 {
-	memcpy_fromio(to, map->virt + from, len);
+	memcpy_fromio(to, (char *)map->virt + from, len);
 }
 
 static void simple_map_copy_to(struct map_info *map, unsigned long to, const void *from, ssize_t len)
 {
-	memcpy_toio(map->virt + to, from, len);
+	memcpy_toio((char *)map->virt + to, from, len);
 }
 
 void simple_map_init(struct map_info *map)
Index: linux-2.6.3/drivers/mtd/maps/octagon-5066.c
===================================================================
--- linux-2.6.3.orig/drivers/mtd/maps/octagon-5066.c
+++ linux-2.6.3/drivers/mtd/maps/octagon-5066.c
@@ -101,7 +101,7 @@ static void oct5066_copy_from(struct map
 		
 		spin_lock(&oct5066_spin);
 		oct5066_page(map, from);
-		memcpy_fromio(to, iomapadr + from, thislen);
+		memcpy_fromio(to, (char *)iomapadr + from, thislen);
 		spin_unlock(&oct5066_spin);
 		to += thislen;
 		from += thislen;
@@ -142,7 +142,7 @@ static void oct5066_copy_to(struct map_i
 		
 		spin_lock(&oct5066_spin);
 		oct5066_page(map, to);
-		memcpy_toio(iomapadr + to, from, thislen);
+		memcpy_toio((char *)iomapadr + to, from, thislen);
 		spin_unlock(&oct5066_spin);
 		to += thislen;
 		from += thislen;
Index: linux-2.6.3/drivers/mtd/maps/sbc_gxx.c
===================================================================
--- linux-2.6.3.orig/drivers/mtd/maps/sbc_gxx.c
+++ linux-2.6.3/drivers/mtd/maps/sbc_gxx.c
@@ -153,7 +153,8 @@ static void sbc_gxx_copy_from(struct map
 		
 		spin_lock(&sbc_gxx_spin);
 		sbc_gxx_page(map, from);
-		memcpy_fromio(to, iomapadr + (from & WINDOW_MASK), thislen);
+		memcpy_fromio(to, (char *)iomapadr + (from & WINDOW_MASK),
+			      thislen);
 		spin_unlock(&sbc_gxx_spin);
 		to += thislen;
 		from += thislen;
@@ -194,7 +195,8 @@ static void sbc_gxx_copy_to(struct map_i
 		
 		spin_lock(&sbc_gxx_spin);
 		sbc_gxx_page(map, to);
-		memcpy_toio(iomapadr + (to & WINDOW_MASK), from, thislen);
+		memcpy_toio((char *)iomapadr + (to & WINDOW_MASK), from,
+			    thislen);
 		spin_unlock(&sbc_gxx_spin);
 		to += thislen;
 		from += thislen;
Index: linux-2.6.3/drivers/mtd/maps/vmax301.c
===================================================================
--- linux-2.6.3.orig/drivers/mtd/maps/vmax301.c
+++ linux-2.6.3/drivers/mtd/maps/vmax301.c
@@ -92,7 +92,7 @@ static void vmax301_copy_from(struct map
 			thislen = WINDOW_LENGTH-(from & WINDOW_MASK);
 		spin_lock(&vmax301_spin);
 		vmax301_page(map, from);
-		memcpy_fromio(to, map->map_priv_2 + from, thislen);
+		memcpy_fromio(to, (char *)map->map_priv_2 + from, thislen);
 		spin_unlock(&vmax301_spin);
 		to += thislen;
 		from += thislen;
@@ -133,7 +133,7 @@ static void vmax301_copy_to(struct map_i
 
 		spin_lock(&vmax301_spin);
 		vmax301_page(map, to);
-		memcpy_toio(map->map_priv_2 + to, from, thislen);
+		memcpy_toio((char *)map->map_priv_2 + to, from, thislen);
 		spin_unlock(&vmax301_spin);		
 		to += thislen;
 		from += thislen;
Index: linux-2.6.3/drivers/scsi/fdomain.c
===================================================================
--- linux-2.6.3.orig/drivers/scsi/fdomain.c	2004-02-18 03:58:47.000000000 +0000
+++ linux-2.6.3/drivers/scsi/fdomain.c	2004-03-10 22:31:40.000000000 +0000
@@ -1650,7 +1650,7 @@
    } else if (bios_major == 3
 	      && bios_minor >= 0
 	      && bios_minor < 4) { /* 3.0 and 3.2 BIOS */
-      memcpy_fromio( &i, bios_base + 0x1f71 + drive * 10,
+      memcpy_fromio( &i, (void *)(bios_base + 0x1f71 + drive * 10),
 		     sizeof( struct drive_info ) );
       info_array[0] = i.heads + 1;
       info_array[1] = i.sectors;
