From: Joachim Deguara <joachim.deguara@amd.com>
Subject: Workaround a hpet BIOS bug which is common on latest AMD driven boards
Patch-Mainline: not yet
References: bnc#387053

Signed-off-by: Thomas Renninger <trenn@suse.de>

--- x/arch/x86/kernel/early-quirks.c.bak	2008-07-11 14:48:51.000000000 +0800
+++ y/arch/x86/kernel/early-quirks.c	2008-07-11 14:37:41.000000000 +0800
@@ -21,6 +21,8 @@
 #include <asm/gart.h>
 #endif
 
+extern int hpet_rework;
+
 static void __init fix_hypertransport_config(int num, int slot, int func)
 {
 	u32 htcfg;
@@ -109,6 +111,16 @@
 #endif
 }
 
+static void __init amd_sb700_hpet(int num, int slot, int func)
+{
+	int rev;
+	rev = read_pci_config_byte(num, slot, func, 0x08);
+	if (rev <= 0x3a && rev >=0x30) {
+		hpet_rework = 1;
+		printk(KERN_INFO "SB700 rev 0x3a under detected!\n");
+	}
+}
+
 #define QFLAG_APPLY_ONCE 	0x1
 #define QFLAG_APPLIED		0x2
 #define QFLAG_DONE		(QFLAG_APPLY_ONCE|QFLAG_APPLIED)
@@ -130,6 +142,8 @@
 	  PCI_CLASS_BRIDGE_PCI, PCI_ANY_ID, QFLAG_APPLY_ONCE, ati_bugs },
 	{ PCI_VENDOR_ID_AMD, PCI_DEVICE_ID_AMD_K8_NB,
 	  PCI_CLASS_BRIDGE_HOST, PCI_ANY_ID, 0, fix_hypertransport_config },
+	{ PCI_VENDOR_ID_ATI, PCI_DEVICE_ID_ATI_SBX00_SMBUS,
+	  PCI_CLASS_SERIAL_SMBUS, PCI_ANY_ID, 0, amd_sb700_hpet },
 	{}
 };
 

--- x/arch/x86/kernel/hpet.c.bak	2008-06-19 10:42:23.000000000 +0800
+++ y/arch/x86/kernel/hpet.c	2008-07-11 14:09:01.000000000 +0800
@@ -23,6 +23,9 @@
  * HPET address is set in acpi/boot.c, when an ACPI entry exists
  */
 unsigned long hpet_address;
+
+int hpet_rework __initdata = 0;
+
 static void __iomem *hpet_virt_address;
 
 unsigned long hpet_readl(unsigned long a)
@@ -387,6 +390,11 @@
 	 * Read the period and check for a sane value:
 	 */
 	hpet_period = hpet_readl(HPET_PERIOD);
+
+	if (hpet_rework) {
+		int timeout = 1000;
+		while (0xffffffff == hpet_readl(HPET_CFG) && timeout-- != 0);
+	}
 	if (hpet_period < HPET_MIN_PERIOD || hpet_period > HPET_MAX_PERIOD)
 		goto out_nohpet;
 
