Index: linux-2.6.4/fs/xfs/xfs_attr_leaf.c
===================================================================
--- linux-2.6.4.orig/fs/xfs/xfs_attr_leaf.c
+++ linux-2.6.4/fs/xfs/xfs_attr_leaf.c
@@ -575,8 +575,8 @@ xfs_attr_shortform_list(xfs_attr_list_co
 	for ( ; i < nsbuf; i++, sbp++) {
 		attrnames_t	*namesp;
 
-		namesp = (sfe->flags & XFS_ATTR_SECURE) ? &attr_secure :
-			((sfe->flags & XFS_ATTR_ROOT) ? &attr_trusted :
+		namesp = (sbp->flags & XFS_ATTR_SECURE) ? &attr_secure :
+			((sbp->flags & XFS_ATTR_ROOT) ? &attr_trusted :
 			  &attr_user);
 
 		if (cursor->hashval != INT_GET(sbp->hash, ARCH_CONVERT)) {
