From: Thomas Renninger <trenn@suse.de>
Subject: Also autoload the bay driver, was forgotten...


---
 drivers/acpi/bay.c |    6 ++++++
 1 file changed, 6 insertions(+)

Index: linux-2.6.22-SL103_BRANCH/drivers/acpi/bay.c
===================================================================
--- linux-2.6.22-SL103_BRANCH.orig/drivers/acpi/bay.c
+++ linux-2.6.22-SL103_BRANCH/drivers/acpi/bay.c
@@ -46,6 +46,12 @@ MODULE_LICENSE("GPL");
 	printk(KERN_DEBUG PREFIX "%s: %s\n", prefix, s); }
 static void bay_notify(acpi_handle handle, u32 event, void *data);
 
+static const struct acpi_device_id bay_device_ids[] = {
+	{"LNXIOBAY", 0},
+	{"", 0},
+};
+MODULE_DEVICE_TABLE(acpi, bay_device_ids);
+
 struct bay {
 	acpi_handle handle;
 	char *name;
