From: Alexander Graf <agraf@suse.de>
Date: Weg, 20 Mar 2013 15:17:25 +0100
Subject: [PATCH] VT8500: fix miscompile in clk driver
Patch-Mainline: No

The prate variable doesn't exist. Use parent_rate instead.

Signed-off-by: Alexander Graf <agraf@suse.de>

Index: linux-3.9-rc3-master/drivers/clk/clk-vt8500.c
===================================================================
--- linux-3.9-rc3-master.orig/drivers/clk/clk-vt8500.c
+++ linux-3.9-rc3-master/drivers/clk/clk-vt8500.c
@@ -157,7 +157,7 @@ static int vt8500_dclk_set_rate(struct c
 	divisor =  parent_rate / rate;
 
 	/* If prate / rate would be decimal, incr the divisor */
-	if (rate * divisor < *prate)
+	if (rate * divisor < parent_rate)
 		divisor++;
 
 	if (divisor == cdev->div_mask + 1)
