From: Alexander Graf <agraf@suse.de>
Subject: [PATCH] Disable sparsemem for exynos
Patch-Mainline: No, this is a test patch

Sparsemem breaks when loading an initrd. Temporarily disable it.

Signed-off-by: Alexander Graf <agraf@suse.de>

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 48c19d4..414cc05 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -889,7 +889,6 @@ config ARCH_S5PV210
 config ARCH_EXYNOS
	bool "SAMSUNG EXYNOS"
	select CPU_V7
-	select ARCH_SPARSEMEM_ENABLE
	select ARCH_HAS_HOLES_MEMORYMODEL
	select GENERIC_GPIO
	select HAVE_CLK
