From: dmueller@suse.com
Subject: Fix syntax error in ux500 driver
Patch-mainline: no

Fix syntax error in ux500 driver

Signed-off-by: Dirk Mueller <dmueller@suse.com>
--- linux-3.6-rc7-master/drivers/crypto/ux500/cryp/cryp_core.c
+++ linux-3.6-rc7-master/drivers/crypto/ux500/cryp/cryp_core.c
@@ -1750,7 +1750,7 @@
 	.shutdown = ux500_cryp_shutdown,
 	.driver = {
 		.owner = THIS_MODULE,
-		.name  = "cryp1"
+		.name  = "cryp1",
 		.pm    = &ux500_cryp_pm,
 	}
 };
