From: olh@suse.de
Subject: compile fix for pc_spkr
pc_spkr needs this header, from linux-2.6.9

diff -Nru a/include/asm-ppc/8253pit.h b/include/asm-ppc/8253pit.h
--- /dev/null	Wed Dec 31 16:00:00 196900
+++ b/include/asm-ppc/8253pit.h	2004-10-18 14:57:24 -07:00
@@ -0,0 +1,10 @@
+/*
+ * 8253/8254 Programmable Interval Timer
+ */
+
+#ifndef _8253PIT_H
+#define _8253PIT_H
+
+#define PIT_TICK_RATE 	1193182UL
+
+#endif
diff -Nru a/include/asm-ppc64/8253pit.h b/include/asm-ppc64/8253pit.h
--- /dev/null	Wed Dec 31 16:00:00 196900
+++ b/include/asm-ppc64/8253pit.h	2004-10-18 14:57:24 -07:00
@@ -0,0 +1,10 @@
+/*
+ * 8253/8254 Programmable Interval Timer
+ */
+
+#ifndef _8253PIT_H
+#define _8253PIT_H
+
+#define PIT_TICK_RATE 	1193182UL
+
+#endif
