diff -urNpa linux-2.6.5-7/arch/ppc64/kernel/sys_ppc32.c linux-2.6.5-7-new/arch/ppc64/kernel/sys_ppc32.c
--- linux-2.6.5-7/arch/ppc64/kernel/sys_ppc32.c	2004-04-08 20:26:02.000000000 +0530
+++ linux-2.6.5-7-new/arch/ppc64/kernel/sys_ppc32.c	2004-04-27 20:42:48.452908064 +0530
@@ -1478,6 +1478,8 @@ do_sys32_shmctl(int first, int second, v
 	case IPC_RMID:
 	case SHM_LOCK:
 	case SHM_UNLOCK:
+		if (second == (IPC_INFO|IPC_64))
+			second = IPC_INFO;
 		err = sys_shmctl(first, second, (struct shmid_ds *)uptr);
 		break;
 	case IPC_SET:
