--- linux-2.6.5-20040525/arch/ppc64/mm/numa.c	2004-05-26 16:39:23.000000000 -0500
+++ linux-2.6.5-20040525-ntl/arch/ppc64/mm/numa.c	2004-05-27 12:08:47.000000000 -0500
@@ -42,6 +42,8 @@
 bootmem_data_t plat_node_bdata[MAX_NUMNODES];
 static unsigned long node0_io_hole_size;
 
+static int numa_disabled;
+
 EXPORT_SYMBOL(node_data);
 EXPORT_SYMBOL(numa_cpu_lookup_table);
 EXPORT_SYMBOL(numa_memory_lookup_table);
@@ -264,7 +266,7 @@
 	if (action == CPU_UP_PREPARE) {
 		int depth = find_min_common_depth();
 
-		if (depth == -1)
+		if (depth == -1 || numa_disabled)
 			map_cpu_to_node(lcpu, 0);
 		else
 			numa_setup_cpu(lcpu, depth);
@@ -288,6 +290,7 @@
 
 	if (strstr(saved_command_line, "numa=off")) {
 		printk(KERN_WARNING "NUMA disabled by user\n");
+		numa_disabled = 1;
 		return -1;
 	}
 
