--- ./arch/ppc64/kernel/eeh.c.global-enable	2004-05-06 17:18:01.000000000 -0500
+++ ./arch/ppc64/kernel/eeh.c	2004-05-10 11:45:58.000000000 -0500
@@ -492,10 +492,6 @@ unsigned long eeh_check_failure(void *to
 	    dn->eeh_mode & EEH_MODE_NOCHECK) 
 		goto ok_return;
 
-        /* Make sure we aren't ISA */
-        if (!strcmp(dn->type, "isa")) 
-                goto ok_return;
-
 	if (!dn->eeh_config_addr) 
 		goto ok_return;
 
@@ -588,6 +584,10 @@ static void *early_enable_eeh(struct dev
 	if ((*class_code >> 16) == PCI_BASE_CLASS_DISPLAY)
 		enable = 0;
 
+	/* Make sure we aren't ISA */
+	if (dn->type && !strcmp(dn->type, "isa"))
+		enable = 0;
+
 	if (!eeh_check_opts_config(dn, *class_code, *vendor_id, *device_id,
 				   enable)) {
 		if (enable) {
