--- ppc64-linux-2.5/arch/ppc64/kernel/head.S	2004-04-19 17:54:59.000000000 +1000
+++ test25/arch/ppc64/kernel/head.S	2004-04-30 16:43:53.026440901 +1000
@@ -265,6 +265,9 @@
 	mfspr	r23,SPRG2;		/* Save r20 in exc. frame      */ \
 	std	r23,EX_R20(r21);	                                  \
                                                                          \
+	mfmsr	r22;			    /* set MSR.RI		   */ \
+	ori	r22,r22,MSR_RI;						      \
+	mtmsrd	r22,1;							      \
 	mfcr    r23;                        /* save CR in r23              */
 
 /*
