--- arch/ppc64/kernel/eeh.c.bug9283	2004-06-14 16:17:55.000000000 -0500
+++ arch/ppc64/kernel/eeh.c	2004-06-15 14:59:37.000000000 -0500
@@ -231,7 +231,6 @@ static void __pci_addr_cache_insert_devi
 	if (!dn) {
 		printk(KERN_WARNING "PCI: no pci dn found for dev=%s %s\n",
 			pci_name(dev), pci_pretty_name(dev));
-		pci_dev_put(dev);
 		return;
 	}
 
@@ -242,9 +241,9 @@ static void __pci_addr_cache_insert_devi
 		printk(KERN_INFO "PCI: skip building address cache for=%s %s\n",
 		       pci_name(dev), pci_pretty_name(dev));
 #endif
-		pci_dev_put(dev);
 		return;
 	}
+	pci_dev_get(dev);
 
 	/* Walk resources on this device, poke them into the tree */
 	for (i = 0; i < DEVICE_COUNT_RESOURCE; i++) {
@@ -334,7 +333,6 @@ void __init pci_addr_cache_build(void)
 	while ((dev = pci_get_device(PCI_ANY_ID, PCI_ANY_ID, dev)) != NULL) {
 		/* Ignore PCI bridges ( XXX why ??) */
 		if ((dev->class >> 16) == PCI_BASE_CLASS_BRIDGE) {
-			pci_dev_put(dev);
 			continue;
 		}
 		pci_addr_cache_insert_device(dev);
@@ -782,7 +780,6 @@ void eeh_add_device_late(struct pci_dev 
 	       pci_pretty_name(dev));
 #endif
 
-	pci_dev_get (dev);
 	pci_addr_cache_insert_device (dev);
 }
 EXPORT_SYMBOL(eeh_add_device_late);
