--- linux-2.6.5-7.97/arch/ppc64/kernel/ptrace32.c.orig	2004-07-29 13:52:42.000000000 -0500
+++ linux-2.6.5-7.97/arch/ppc64/kernel/ptrace32.c	2004-07-29 13:43:02.000000000 -0500
@@ -413,7 +413,9 @@
 		break;
 	}
 
-
+       case PTRACE_GETEVENTMSG:   
+                ret = put_user(child->ptrace_message, (unsigned int __user *) data);
+                break;
 
 	default:
 		ret = ptrace_request(child, request, addr, data);
