diff -Naur 2.6.6-rc3-mm1/arch/ppc64/kernel/smp.c 2.6.6-rc3-mm1.ltc8091/arch/ppc64/kernel/smp.c
--- 2.6.6-rc3-mm1/arch/ppc64/kernel/smp.c	2004-05-03 13:03:22.000000000 -0500
+++ 2.6.6-rc3-mm1.ltc8091/arch/ppc64/kernel/smp.c	2004-05-03 13:00:49.000000000 -0500
@@ -909,13 +909,20 @@
 	 * use this value that I found through experimentation.
 	 * -- Cort
 	 */
-	for (c = 5000; c && !cpu_callin_map[cpu]; c--)
-		udelay(100);
+	for (c = 50000; c && !cpu_callin_map[cpu]; c--) {
+		if (system_state == SYSTEM_BOOTING)
+			udelay(100);
+		else {
+			set_current_state(TASK_UNINTERRUPTIBLE);
+			schedule_timeout(HZ/5);
+		}
+	}
 
 	if (!cpu_callin_map[cpu]) {
 		printk("Processor %u is stuck.\n", cpu);
 		return -ENOENT;
-	}
+	} else
+		printk("Found processor %u with %d tries left\n", cpu, c);
 
 	printk("Processor %u found.\n", cpu);
 
