diff -u --recursive --new-file --exclude '*.cmd' --exclude '*.mod' linux-2.6.0-test5.orig/arch/s390/kernel/compat_ioctl.c linux-2.6.0-test5/arch/s390/kernel/compat_ioctl.c
--- linux-2.6.0-test5.orig/arch/s390/kernel/compat_ioctl.c	2003-09-08 21:50:12.000000000 +0200
+++ linux-2.6.0-test5/arch/s390/kernel/compat_ioctl.c	2003-09-24 13:50:40.000000000 +0200
@@ -26,7 +26,6 @@
 #include <linux/blkpg.h>
 #include <linux/cdrom.h>
 #include <linux/dm-ioctl.h>
-#include <linux/elevator.h>
 #include <linux/file.h>
 #include <linux/fs.h>
 #include <linux/hdreg.h>
diff -u --recursive --new-file --exclude '*.cmd' --exclude '*.mod' linux-2.6.0-test5.orig/drivers/s390/net/fsm.c linux-2.6.0-test5/drivers/s390/net/fsm.c
--- linux-2.6.0-test5.orig/drivers/s390/net/fsm.c	2003-09-08 21:50:04.000000000 +0200
+++ linux-2.6.0-test5/drivers/s390/net/fsm.c	2003-09-29 12:16:36.000000000 +0200
@@ -10,6 +10,10 @@
 #include <linux/module.h>
 #include <linux/timer.h>
 
+MODULE_AUTHOR("IBM Corp.");
+MODULE_DESCRIPTION("generic FSM");
+MODULE_LICENSE("GPL");
+
 fsm_instance *
 init_fsm(char *name, const char **state_names, const char **event_names, int nr_states,
 		int nr_events, const fsm_node *tmpl, int tmpl_len, int order)
diff -u --recursive --new-file --exclude '*.cmd' --exclude '*.mod' linux-2.6.0-test5.orig/include/asm-s390/pgtable.h linux-2.6.0-test5/include/asm-s390/pgtable.h
--- linux-2.6.0-test5.orig/include/asm-s390/pgtable.h	2003-09-08 21:50:21.000000000 +0200
+++ linux-2.6.0-test5/include/asm-s390/pgtable.h	2003-09-24 14:09:13.000000000 +0200
@@ -285,5 +285,6 @@
 #define PAGE_RO_SHARED	  __pgprot(_PAGE_RO)
 #define PAGE_RO_PRIVATE	  __pgprot(_PAGE_RO|_PAGE_ISCLEAN)
 #define PAGE_COPY	  __pgprot(_PAGE_RO|_PAGE_ISCLEAN)
+#define PAGE_READONLY	  __pgprot(_PAGE_RO)
 #define PAGE_SHARED	  __pgprot(0)
 #define PAGE_KERNEL	  __pgprot(_PAGE_ISCLEAN)
--- linux-2.6.0-test6/arch/s390/Makefile~	2003-09-28 02:50:10.000000000 +0200
+++ linux-2.6.0-test6/arch/s390/Makefile	2003-10-02 14:47:43.000000000 +0200
@@ -22,7 +22,6 @@
 
 ifdef CONFIG_ARCH_S390X
 LDFLAGS		:= -m elf64_s390
-MODFLAGS	+= -fpic -D__PIC__
 CFLAGS		+= -m64
 AFLAGS		+= -m64
 UTS_MACHINE	:= s390x
