Subject: imac g5 and current g5 support
From: olh@suse.de
Patch-mainline: almost

http://perso.magic.fr/l_indien/linux-ImacG5/index.htm

diff -urNbBp /usr/src/linux-2.6.10-rc3/arch/ppc64/kernel/pmac_feature.c linux-2.6.10-rc3/arch/ppc64/kernel/pmac_feature.c
--- /usr/src/linux-2.6.10-rc3/arch/ppc64/kernel/pmac_feature.c	2004-12-12 11:26:42.000000000 +0100
+++ linux-2.6.10-rc3/arch/ppc64/kernel/pmac_feature.c	2004-12-12 12:45:31.000000000 +0100
@@ -347,6 +347,10 @@ static struct pmac_mb_def pmac_mb_defs[]
 		PMAC_TYPE_POWERMAC_G5,		g5_features,
 		0,
 	},
+	{	"PowerMac8,1",			"IMac G5",
+		PMAC_TYPE_POWERMAC_G5,		g5_features,
+		0,
+	},
 	{       "RackMac3,1",                   "XServe G5",
 		PMAC_TYPE_POWERMAC_G5,          g5_features,
 		0,

diff -urNbBp /usr/src/linux-2.6.10-rc3/drivers/net/sungem.c linux-2.6.10-rc3/drivers/net/sungem.c
--- /usr/src/linux-2.6.10-rc3/drivers/net/sungem.c	2004-12-12 11:26:42.000000000 +0100
+++ linux-2.6.10-rc3/drivers/net/sungem.c	2004-12-12 12:45:31.000000000 +0100
@@ -109,6 +109,8 @@ static struct pci_device_id gem_pci_tbl[
 	  PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0UL },
 	{ PCI_VENDOR_ID_APPLE, PCI_DEVICE_ID_APPLE_K2_GMAC,
 	  PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0UL },
+       { PCI_VENDOR_ID_APPLE, PCI_DEVICE_ID_APPLE_SH_SUNGEM,
+          PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0UL },
 	{0, }
 };
 

