From: olh@suse.de
Subject: enable CONFIG_ISA to get console output on legacy 8250 ports
Patch-mainline: to be discussed

Signed-off-by: Olaf Hering <olh@suse.de>

 arch/powerpc/Kconfig |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.15/arch/powerpc/Kconfig
===================================================================
--- linux-2.6.15.orig/arch/powerpc/Kconfig
+++ linux-2.6.15/arch/powerpc/Kconfig
@@ -712,7 +712,7 @@ menu "Bus options"
 
 config ISA
 	bool "Support for ISA-bus hardware"
-	depends on PPC_PREP || PPC_CHRP
+	depends on PPC_PREP || PPC_CHRP || PPC_PSERIES
 	select PPC_I8259
 	help
 	  Find out whether you have ISA slots on your motherboard.  ISA is the
