---
 arch/powerpc/platforms/ps3/repository.c |    2 ++
 1 file changed, 2 insertions(+)

Index: b/arch/powerpc/platforms/ps3/repository.c
===================================================================
--- a/arch/powerpc/platforms/ps3/repository.c
+++ b/arch/powerpc/platforms/ps3/repository.c
@@ -542,6 +542,7 @@ int ps3_repository_find_bus(enum ps3_bus
 	*bus_index = UINT_MAX;
 	return -ENODEV;
 }
+EXPORT_SYMBOL_GPL(ps3_repository_find_bus);
 
 static int find_device(unsigned int bus_index, unsigned int num_dev,
 	unsigned int start_dev_index, enum ps3_dev_type dev_type,
@@ -688,6 +689,7 @@ int ps3_repository_find_interrupt(const 
 
 	return result;
 }
+EXPORT_SYMBOL_GPL(ps3_repository_find_interrupt);
 
 int ps3_repository_find_reg(const struct ps3_repository_device *dev,
 	enum ps3_reg_type reg_type, u64 *bus_addr, u64 *len)
