From: olh@suse.de
Subject: CONFIG_NUMA does currently not work

---
 arch/powerpc/platforms/cell/setup.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.16-rc6-olh/arch/powerpc/platforms/cell/setup.c
===================================================================
--- linux-2.6.16-rc6-olh.orig/arch/powerpc/platforms/cell/setup.c
+++ linux-2.6.16-rc6-olh/arch/powerpc/platforms/cell/setup.c
@@ -69,7 +69,8 @@ static void cell_show_cpuinfo(struct seq
 	of_node_put(root);
 }
 
-#ifdef CONFIG_SPARSEMEM
+#if defined(CONFIG_SPARSEMEM) && defined(CONFIG_BROKEN)
+#error CONFIG_NUMA
 static int __init find_spu_node_id(struct device_node *spe)
 {
 	unsigned int *id;
