From: olh@suse.de
Subject: create /sys/module/gelic/drivers

---
 drivers/net/gelic_net.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/gelic_net.c
+++ b/drivers/net/gelic_net.c
@@ -1844,6 +1844,7 @@ static struct ps3_system_bus_driver ps3_
 	.probe = ps3_gelic_driver_probe,
 	.remove = ps3_gelic_driver_remove,
 	.core = {
+		.owner = THIS_MODULE,
 		.name = "ps3_gelic_driver",
 	},
 };
