Subject: fix compile error
From: olh@suse.de

---
 arch/powerpc/platforms/iseries/setup.c |    2 ++
 1 files changed, 2 insertions(+)

--- linux-2.6.19.orig/arch/powerpc/platforms/iseries/setup.c
+++ linux-2.6.19/arch/powerpc/platforms/iseries/setup.c
@@ -304,9 +304,11 @@ static void __init iSeries_init_early(vo
 		initrd_start = (unsigned long)__va(naca.xRamDisk);
 		initrd_end = initrd_start + naca.xRamDiskSize * HW_PAGE_SIZE;
 		initrd_below_start_ok = 1;	// ramdisk in kernel space
+#if defined(CONFIG_BLK_DEV_RAM)
 		ROOT_DEV = Root_RAM0;
 		if (((rd_size * 1024) / HW_PAGE_SIZE) < naca.xRamDiskSize)
 			rd_size = (naca.xRamDiskSize * HW_PAGE_SIZE) / 1024;
+#endif
 	} else
 #endif /* CONFIG_BLK_DEV_INITRD */
 	{
