---
 drivers/net/gelic_net.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/net/gelic_net.c
+++ b/drivers/net/gelic_net.c
@@ -630,9 +630,9 @@ static void gelic_net_set_txdescr_cmdsta
 		/* is packet ip?
 		 * if yes: tcp? udp? */
 		if (skb->protocol == htons(ETH_P_IP)) {
-			if (ip_hdr(skb)->protocol == IPPROTO_TCP)
+			if (skb->nh.iph->protocol == IPPROTO_TCP)
 				descr->dmac_cmd_status = GELIC_NET_DMAC_CMDSTAT_TCPCS | eofr;
-			else if (ip_hdr(skb)->protocol == IPPROTO_UDP)
+			else if (skb->nh.iph->protocol == IPPROTO_UDP)
 				descr->dmac_cmd_status = GELIC_NET_DMAC_CMDSTAT_UDPCS | eofr;
 			else	/*
 				 * the stack should checksum non-tcp and non-udp
