Subject: export for alsa
From: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Patch-mainline: yes

Signed-off-by: Olaf Hering <olh@suse.de>

 arch/ppc/platforms/pmac_feature.c |    1 +
 1 files changed, 1 insertion(+)

Index: linux-2.6.11/arch/ppc/platforms/pmac_feature.c
===================================================================
--- linux-2.6.11.orig/arch/ppc/platforms/pmac_feature.c
+++ linux-2.6.11/arch/ppc/platforms/pmac_feature.c
@@ -88,6 +88,7 @@ macio_find(struct device_node* child, in
 	}
 	return NULL;
 }
+EXPORT_SYMBOL_GPL(macio_find);
 
 static const char* macio_names[] __pmacdata =
 {
