---
 drivers/ps3/ps3av.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/ps3/ps3av.c	2007-08-27 14:01:21.000000000 -0400
+++ b/drivers/ps3/ps3av.c	2007-08-27 14:01:21.000000000 -0400
@@ -713,6 +713,9 @@ static const struct ps3av_monitor_quirk 
 		.monitor_name	= "L226WTQ",
 		.clear_60	= PS3AV_RESBIT_1920x1080I |
 				  PS3AV_RESBIT_1920x1080P
+	}, {
+		.monitor_name	= "SyncMaster",
+		.clear_60	= PS3AV_RESBIT_1920x1080I
 	}
 };
 
