Subject: run device-init only on PS3
From: olh@suse.de

---
 arch/powerpc/platforms/ps3/device-init.c |    3 +++
 1 file changed, 3 insertions(+)

Index: b/arch/powerpc/platforms/ps3/device-init.c
===================================================================
--- a/arch/powerpc/platforms/ps3/device-init.c
+++ b/arch/powerpc/platforms/ps3/device-init.c
@@ -441,6 +441,9 @@ ps3_register_known_devices (void)
 
 	pr_debug(" -> %s:%d\n", __func__, __LINE__);
 
+	if (!firmware_has_feature(FW_FEATURE_PS3_LV1))
+		return -ENODEV;
+
 	//ps3_repository_dump_bus_info();
 
 	result = ps3_register_ohci_0();
