Subject: hide CONFIG_DEBUG_SLAB in CONFIG_NETFILTER is enabled
From: olh@suse.de
References: 155004


---
 lib/Kconfig.debug |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

Index: linux-2.6.16/lib/Kconfig.debug
===================================================================
--- linux-2.6.16.orig/lib/Kconfig.debug
+++ linux-2.6.16/lib/Kconfig.debug
@@ -77,9 +77,15 @@ config SCHEDSTATS
 	  application, you can say N to avoid the very slight overhead
 	  this adds.
 
+# There is an alignment bug with slab debugging, some iptable modules
+# will not work. As a result, SuSEfirewall will lock you out. Great.
+
+comment "Debug memory allocations can not work because netfilter/iptables support is enabled"
+	depends on PPC32 && NETFILTER
+
 config DEBUG_SLAB
 	bool "Debug memory allocations"
-	depends on DEBUG_KERNEL && SLAB
+	depends on DEBUG_KERNEL && SLAB && (!PPC32 || (PPC32 && NETFILTER=n))
 	help
 	  Say Y here to have the kernel do limited verification on memory
 	  allocation as well as poisoning memory on free to catch use of freed
