From: olh@suse.de
Subject: create /sys/module/ps3_storage/drivers

---
 drivers/block/ps3_storage.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/block/ps3_storage.c
+++ b/drivers/block/ps3_storage.c
@@ -2620,6 +2620,7 @@ static void ps3_stor_driver_shutdown(str
 
 static struct platform_driver ps3_stor_platform_driver = {
 	.driver = {
+		.owner = THIS_MODULE,
 		.name = "ps3_storage"
 	},
 	.probe          = ps3_stor_driver_probe,
