--- linux-2.6.4.orig/arch/ppc/boot/simple/Makefile	2004-03-17 20:08:07.000000000 +0000
+++ linux-2.6.4/arch/ppc/boot/simple/Makefile	2004-03-17 20:33:30.000000000 +0000
@@ -114,7 +114,7 @@ zimageinitrd-$(CONFIG_SPRUCE)		:= zImage
         misc-$(CONFIG_SPRUCE)		+= misc-spruce.o
 
 # SMP images should have a '.smp' suffix.
-         end-$(CONFIG_SMP)		+= .smp
+         end-$(CONFIG_SMP)		:= $(end-y).smp
 
 # This is a treeboot that needs init functions until the
 # boot rom is sorted out (i.e. this is short lived)
