--- linux-2.6.5/arch/ppc64/kernel/mf_proc.c	2004-04-04 05:36:57.000000000 +0200
+++ linux-2.6.5/arch/ppc64/kernel/mf_proc.c	2004-04-27 18:02:18.000000000 +0200
@@ -177,10 +177,14 @@ static int proc_mf_change_cmdline(struct
 static int proc_mf_change_vmlinux(struct file *file, const char *buffer,
 		unsigned long count, void *data)
 {
+	int rc;
 	if (!capable(CAP_SYS_ADMIN))
 		return -EACCES;
 
-	mf_setVmlinuxChunk(buffer, count, file->f_pos, (u64)data);
+	rc = mf_setVmlinuxChunk(buffer, count, file->f_pos, (u64)data);
+	if (rc < 0)
+		return rc;
+
 	file->f_pos += count;
 
 	return count;			
