diff -Naurp linux-2.6.5-7.orig/arch/ppc64/kernel/xics.c linux-2.6.5-7/arch/ppc64/kernel/xics.c
--- linux-2.6.5-7.orig/arch/ppc64/kernel/xics.c	2004-04-22 11:21:56.483972352 -0700
+++ linux-2.6.5-7/arch/ppc64/kernel/xics.c	2004-04-22 11:20:59.426993936 -0700
@@ -615,7 +615,7 @@ static void xics_set_affinity(unsigned i
 	cpumask_t tmp = CPU_MASK_NONE;
 
 	irq = virt_irq_to_real(irq_offset_down(virq));
-	if (irq == XICS_IPI)
+	if (irq == XICS_IPI || irq == NO_IRQ)
 		return;
 
 	status = rtas_call(ibm_get_xive, 1, 3, (void *)&xics_status, irq);
