From: olh@suse.de
Subject: Bug 56865 - LTC9238-ibmveth module not loaded on machines w/ virtual ethernet adapters

Signed-off-by: Olaf Hering <olh@suse.de>

 arch/ppc64/kernel/vio.c |   27 +++++++++++++++++++++++++++
 1 files changed, 27 insertions(+)

Index: linux-2.6.14-olh/arch/ppc64/kernel/vio.c
===================================================================
--- linux-2.6.14-olh.orig/arch/ppc64/kernel/vio.c
+++ linux-2.6.14-olh/arch/ppc64/kernel/vio.c
@@ -21,6 +21,7 @@
 #include <asm/iommu.h>
 #include <asm/dma.h>
 #include <asm/vio.h>
+#include <asm/prom.h>
 
 static const struct vio_device_id *vio_match_device(
 		const struct vio_device_id *, const struct vio_dev *);
@@ -255,7 +256,37 @@ static int vio_bus_match(struct device *
 	return (ids != NULL) && (vio_match_device(ids, vio_dev) != NULL);
 }
 
+static int vio_hotplug (struct device *dev, char **envp, int num_envp,
+                          char *buffer, int buffer_size)
+{
+#ifdef CONFIG_PPC_PSERIES
+	const struct vio_dev *vio_dev = to_vio_dev(dev);
+	char *cp;
+	int length;
+
+	if (!num_envp)
+		return -ENOMEM;
+
+	if (!vio_dev->dev.platform_data)
+		return -ENODEV;
+	cp = (char *)get_property(vio_dev->dev.platform_data, "compatible", &length);
+	if (!cp)
+		return -ENODEV;
+
+	envp[0] = buffer;
+	length = scnprintf (buffer, buffer_size, "MODALIAS=vio:T%sS%s",
+	                     vio_dev->type, cp);
+	if (buffer_size - length <= 0)
+		return -ENOMEM;
+	envp[1] = NULL;
+	return 0;
+#else
+	return -ENODEV;
+#endif
+}
+
 struct bus_type vio_bus_type = {
 	.name = "vio",
+	.hotplug = vio_hotplug,
 	.match = vio_bus_match,
 };
