--- ./drivers/scsi/ibmvscsi/rpa_vscsi.c~	2004-02-13 18:56:02.000000000 +0100
+++ ./drivers/scsi/ibmvscsi/rpa_vscsi.c	2004-02-13 19:41:13.000000000 +0100
@@ -28,6 +28,7 @@
  */
 
 #include <linux/module.h>
+#include <linux/pci.h>
 #include <asm/vio.h>
 #include <asm/pci_dma.h>
 #include <asm/hvcall.h>
--- ./drivers/scsi/ibmvscsi/ibmvscsi.c~	2004-02-13 18:56:02.000000000 +0100
+++ ./drivers/scsi/ibmvscsi/ibmvscsi.c	2004-02-13 19:38:14.000000000 +0100
@@ -63,6 +63,7 @@
  */
 
 #include <linux/module.h>
+#include <linux/pci.h>
 #include <asm/vio.h>
 #include "ibmvscsi.h"
 
