From: olh@suse.de
Subject: [Bug middle-end/24644] [4.1 Regression] gcc-4.1 compiled ppc64 kernels do not boot                                                                          

 include/asm-ppc64/paca.h |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.14-olh/include/asm-ppc64/paca.h
===================================================================
--- linux-2.6.14-olh.orig/include/asm-ppc64/paca.h
+++ linux-2.6.14-olh/include/asm-ppc64/paca.h
@@ -22,7 +22,7 @@
 #include	<asm/iSeries/ItLpRegSave.h>
 #include	<asm/mmu.h>
 
-register struct paca_struct *local_paca asm("r13");
+register struct paca_struct * volatile local_paca asm("r13");
 #define get_paca()	local_paca
 
 struct task_struct;
