--- linux-2.6.5/include/asm-s390/processor.h.orig	2004-04-23 10:04:22.000000000 +0200
+++ linux-2.6.5/include/asm-s390/processor.h	2004-04-23 10:04:50.000000000 +0200
@@ -72,6 +72,7 @@
 # define TASK31_SIZE		(0x80000000UL)
 # define TASK_UNMAPPED_BASE	(test_thread_flag(TIF_31BIT) ? \
 					(TASK31_SIZE / 2) : (TASK_SIZE / 2))
+# define __TASK_UNMAPPED_BASE	(TASK31_SIZE / 2)
 
 #endif /* __s390x__ */
 
