From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: Kernel panic while testing softdog
References: 161997 - LTC22671

Kernel panic while working with softdog 

Acked-by: Andi Kleen <ak@suse.de>
Acked-by: Hannes Reinecke <hare@suse.de>

diff -urpN linux-2.6/arch/s390/kernel/setup.c linux-2.6-softdog/arch/s390/kernel/setup.c
--- linux-2.6/arch/s390/kernel/setup.c	2006-03-20 16:56:34.000000000 +0100
+++ linux-2.6-softdog/arch/s390/kernel/setup.c	2006-05-04 15:27:12.000000000 +0200
@@ -297,19 +297,34 @@ void (*_machine_power_off)(void) = do_ma
 
 void machine_restart(char *command)
 {
-	console_unblank();
+	if (!in_interrupt() || oops_in_progress)
+		/*
+		 * Only unblank the console if we are called in enabled
+		 * context or a bust_spinlocks cleared the way for us.
+		 */
+		console_unblank();
 	_machine_restart(command);
 }
 
 void machine_halt(void)
 {
-	console_unblank();
+	if (!in_interrupt() || oops_in_progress)
+		/*
+		 * Only unblank the console if we are called in enabled
+		 * context or a bust_spinlocks cleared the way for us.
+		 */
+		console_unblank();
 	_machine_halt();
 }
 
 void machine_power_off(void)
 {
-	console_unblank();
+	if (!in_interrupt() || oops_in_progress)
+		/*
+		 * Only unblank the console if we are called in enabled
+		 * context or a bust_spinlocks cleared the way for us.
+		 */
+		console_unblank();
 	_machine_power_off();
 }
 
