 less output when booted with 'quiet'

diff -purN linux-2.6.8-rc2/drivers/ide/ppc/pmac.c linux-2.6.8-rc2.quiet/drivers/ide/ppc/pmac.c
--- linux-2.6.8-rc2/drivers/ide/ppc/pmac.c	2004-06-16 07:18:58.000000000 +0200
+++ linux-2.6.8-rc2.quiet/drivers/ide/ppc/pmac.c	2004-07-19 12:02:18.852743725 +0200
@@ -53,7 +53,7 @@
 
 extern void ide_do_request(ide_hwgroup_t *hwgroup, int masked_irq);
 
-#define IDE_PMAC_DEBUG
+#undef IDE_PMAC_DEBUG
 
 #define DMA_WAIT_TIMEOUT	50
 
