diff -purN linux-2.6.6/arch/ppc64/kernel/setup.c linux-2.6.6.autocons/arch/ppc64/kernel/setup.c
--- linux-2.6.6/arch/ppc64/kernel/setup.c	2004-05-10 04:32:29.000000000 +0200
+++ linux-2.6.6.autocons/arch/ppc64/kernel/setup.c	2004-05-20 11:58:14.000000000 +0200
@@ -476,6 +476,7 @@ static int __init set_preferred_console(
 {
 	struct device_node *prom_stdout;
 	char *name;
+	int offset;
 
 	/* The user has requested a console so this is already set up. */
 	if (strstr(saved_command_line, "console="))
@@ -493,7 +494,6 @@ static int __init set_preferred_console(
 		int i;
 		u32 *reg = (u32 *)get_property(prom_stdout, "reg", &i);
 		if (i > 8) {
-			int offset;
 			switch (reg[1]) {
 				case 0x3f8:
 					offset = 0;
@@ -511,15 +511,19 @@ static int __init set_preferred_console(
 					/* We dont recognise the serial port */
 					return -ENODEV;
 			}
-
-			return add_preferred_console("ttyS", offset, NULL);
 		}
-	} else if (strcmp(name, "vty") == 0) {
+	} else if (strcmp(name, "vty") == 0)
 		/* pSeries LPAR virtual console */
 		return add_preferred_console("hvc", 0, NULL);
-	}
+	else if (strcmp(name, "ch-a") == 0)
+			offset = 0;
+	else if (strcmp(name, "ch-b") == 0)
+			offset = 1;
+	else
+		return -ENODEV;
+
+	return add_preferred_console("ttyS", offset, NULL);
 
-	return -ENODEV;
 }
 console_initcall(set_preferred_console);
 
