diff -purN ../linux-2.6.5/arch/x86_64/boot/Makefile ./arch/x86_64/boot/Makefile
--- ../linux-2.6.5/arch/x86_64/boot/Makefile	2004-10-29 19:15:31.415986330 +0200
+++ ./arch/x86_64/boot/Makefile	2004-10-29 19:35:10.045981605 +0200
@@ -31,6 +31,7 @@ targets		:= vmlinux.bin bootsect bootsec
 EXTRA_CFLAGS := -m32
 
 hostprogs-y	:= tools/build
+HOST_EXTRACFLAGS += $(LINUXINCLUDE)
 subdir-		:= compressed/	#Let make clean descend in compressed/
 # ---------------------------------------------------------------------------
 
