From 857b7375c6ea34e585d770e6553c64a5f43dc17f Mon Sep 17 00:00:00 2001
From: Michal Marek <mmarek@suse.cz>
Date: Mon, 21 Feb 2011 22:58:37 +0100
Subject: [PATCH 1/2] spi: Fix Altera SPI build
Patch-mainline: See https://lkml.org/lkml/2011/2/19/143

Signed-off-by: Michal Marek <mmarek@suse.cz>
---
 drivers/spi/spi_altera.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/spi/spi_altera.c b/drivers/spi/spi_altera.c
index 5e001f3..4813a63 100644
--- a/drivers/spi/spi_altera.c
+++ b/drivers/spi/spi_altera.c
@@ -304,7 +304,7 @@ static int __devexit altera_spi_remove(struct platform_device *dev)
 static const struct of_device_id altera_spi_match[] = {
 	{ .compatible = "ALTR,spi-1.0", },
 	{},
-}
+};
 MODULE_DEVICE_TABLE(of, altera_spi_match);
 #else /* CONFIG_OF */
 #define altera_spi_match NULL
-- 
1.7.3.4

