From c8b2af84976a5e4131a9cc6877cbd96fa07eb417 Mon Sep 17 00:00:00 2001
From: Michal Marek <mmarek@suse.cz>
Date: Mon, 21 Feb 2011 23:30:21 +0100
Subject: [PATCH 2/2] spi: fix build of OpenCores tiny driver
Patch-mainline: See https://lkml.org/lkml/2011/2/19/143

Signed-off-by: Michal Marek <mmarek@suse.cz>
---
 drivers/spi/spi_oc_tiny.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/spi/spi_oc_tiny.c b/drivers/spi/spi_oc_tiny.c
index e99e545..f1bde66 100644
--- a/drivers/spi/spi_oc_tiny.c
+++ b/drivers/spi/spi_oc_tiny.c
@@ -390,7 +390,7 @@ static int __devexit tiny_spi_remove(struct platform_device *pdev)
 static const struct of_device_id tiny_spi_match[] = {
 	{ .compatible = "opencores,tiny-spi-rtlsvn2", },
 	{},
-}
+};
 MODULE_DEVICE_TABLE(of, tiny_spi_match);
 #else /* CONFIG_OF */
 #define tiny_spi_match NULL
-- 
1.7.3.4

