From: Masahiro Yamada <masahiroy@kernel.org>
Subject: Fix scripts/dummy-tools/gcc: unknown assembler invoked

References: scripts/dummy-tools/gcc: unknown assembler invoked
Patch-mainline: not yet, merge pending

Acked-by: Michal Suchanek <msuchanek@suse.de>

--- a/scripts/dummy-tools/gcc
+++ b/scripts/dummy-tools/gcc
@@ -67,6 +67,12 @@ if arg_contain -E "$@"; then
 	fi
 fi

+# To set CONFIG_AS_IS_GNU
+if arg_contain -Wa,--version "$@"; then
+	echo "GNU assembler (scripts/dummy-tools) 2.50"
+	exit 0
+fi
+
 if arg_contain -S "$@"; then
 	# For scripts/gcc-x86-*-has-stack-protector.sh
 	if arg_contain -fstack-protector "$@"; then
