From: agruen@suse.de
Subject: Hard-code the RPM release number; don't get it from .config

--- a/Makefile.1	2006-01-12 14:49:16.098116699 +0100
+++ b/Makefile	2006-01-12 14:50:02.639552826 +0100
@@ -335,8 +335,12 @@
 		   -ffreestanding
 AFLAGS		:= -D__ASSEMBLY__
 
-# Read KERNELRELEASE from .kernelrelease (if it exists)
-KERNELRELEASE = $(shell cat .kernelrelease 2> /dev/null)
+ifneq ($(wildcard $(srctree)/rpm-release),)
+RPM_RELEASE := -$(shell cat $(srctree)/rpm-release)
+endif
+
+KERNELRELEASE=$(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)$(RPM_RELEASE)$(LOCALVERSION)
+
 
 export	VERSION PATCHLEVEL SUBLEVEL KERNELRELEASE \
 	ARCH CONFIG_SHELL HOSTCC HOSTCFLAGS CROSS_COMPILE AS LD CC \
