From a1b1c316125e701e739c18e0bc9cb7b63d693267 Mon Sep 17 00:00:00 2001
From: Hans-Peter Jansen <hp.jansen@suse.com>
Date: Sat, 13 Aug 2022 12:25:49 +0200
Patch-mainline: Never, carry this change until upstream provides better method
References: none
Subject: [PATCH] scripts/check-local-export: use /bin/bash in #!

---
 scripts/check-local-export | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/check-local-export b/scripts/check-local-export
index 6ccc2f467416..3930a71c4ce3 100755
--- a/scripts/check-local-export
+++ b/scripts/check-local-export
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/bash
 # SPDX-License-Identifier: GPL-2.0-only
 #
 # Copyright (C) 2022 Masahiro Yamada <masahiroy@kernel.org>
-- 
2.37.1

