From a0eb859d0ac42beb4120b306119d1f6955b528aa Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Thu, 15 Mar 2018 20:49:17 +0100
Subject: [PATCH] Thunderbolt: kABI paddings added
Patch-mainline: Never (SUSE kABI padding)
References: bsc#1070681

just in case

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/thunderbolt.h |    5 +++++
 1 file changed, 5 insertions(+)

--- a/include/linux/thunderbolt.h
+++ b/include/linux/thunderbolt.h
@@ -82,6 +82,7 @@ struct tb {
 	int index;
 	enum tb_security_level security_level;
 	size_t nboot_acl;
+	void *suse_kabi_padding;
 	unsigned long privdata[0];
 };
 
@@ -106,6 +107,7 @@ static inline unsigned int tb_phy_port_f
 struct tb_property_dir {
 	const uuid_t *uuid;
 	struct list_head properties;
+	void *suse_kabi_padding;
 };
 
 enum tb_property_type {
@@ -139,6 +141,7 @@ struct tb_property {
 		char *text;
 		u32 immediate;
 	} value;
+	void *suse_kabi_padding;
 };
 
 struct tb_property_dir *tb_property_parse_dir(const u32 *block,
@@ -443,6 +446,7 @@ struct tb_nhi {
 	bool going_away;
 	struct work_struct interrupt_work;
 	u32 hop_count;
+	void *suse_kabi_padding;
 };
 
 /**
@@ -490,6 +494,7 @@ struct tb_ring {
 	u16 eof_mask;
 	void (*start_poll)(void *data);
 	void *poll_data;
+	void *suse_kabi_padding;
 };
 
 /* Leave ring interrupt enabled on suspend */
