From c5578bea3e0ffae60f02b540f79efc7af4527bea Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.de>
Date: Thu, 3 Jul 2014 14:29:25 +0200
Subject: [PATCH 1/2] Add a void* suse_kabi_padding placeholder to some USB
 structures
Patch-mainline: Never, kabi
References: fate#312013

Signed-off-by: Oliver Neukum <oneukum@suse.de>
---
 include/linux/usb/hcd.h |    2 ++
 1 file changed, 2 insertions(+)

--- a/include/linux/usb/hcd.h
+++ b/include/linux/usb/hcd.h
@@ -544,6 +544,8 @@ struct usb_tt {
 	spinlock_t		lock;
 	struct list_head	clear_list;	/* of usb_tt_clear */
 	struct work_struct	clear_work;
+
+	void *suse_kabi_padding;
 };
 
 struct usb_tt_clear {
