From 98b3b423a9c7f45936f1ecad8a352933ca39a29e Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.de>
Date: Tue, 30 Jul 2013 16:08:04 +0200
Subject: [PATCH] Don't switch camera on HP EB 820 G1
Patch-Mainline: Never
References: bnc#822164

---
 drivers/usb/host/pci-quirks.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/usb/host/pci-quirks.c
+++ b/drivers/usb/host/pci-quirks.c
@@ -565,6 +565,13 @@ static const struct dmi_system_id bt_exc
 		.driver_data = (void *)(64 | 8),
 	},
 	{
+		/* HP EliteBook 820 */
+		.matches = {
+			DMI_MATCH(DMI_PRODUCT_NAME, "HP EliteBook 820 G1"),
+		},
+		.driver_data = (void *)64,
+	},
+	{
 		/* HP EliteBook 840 */
 		.matches = {
 			DMI_MATCH(DMI_PRODUCT_NAME, "HP EliteBook 840 G1"),
