From 3658c799acce7b1af2f4d565c7c0122b9aa13a05 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.de>
Date: Thu, 3 Jul 2014 14:39:17 +0200
Subject: [PATCH 2/2] Add a void* suse_kabi_padding placeholder to some gspca
 structures
Patch-mainline: Never, kabi
References: fate#312013

Signed-off-by: Oliver Neukum <oneukum@suse.de>
---
 drivers/media/usb/gspca/gspca.h |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/media/usb/gspca/gspca.h
+++ b/drivers/media/usb/gspca/gspca.h
@@ -126,6 +126,8 @@ struct sd_desc {
 	   int_pkt_scan is NULL, for cams with non interrupt driven buttons */
 	u8 other_input;
 #endif
+
+	void *suse_kabi_padding;
 };
 
 /* packet types when moving from iso buf to frame buf */
@@ -211,6 +213,8 @@ struct gspca_dev {
 	/* (*) These variables are proteced by both usb_lock and queue_lock,
 	   that is any code setting them is holding *both*, which means that
 	   any code getting them needs to hold at least one of them */
+
+	void *suse_kabi_padding;
 };
 
 int gspca_dev_probe(struct usb_interface *intf,
