From: Tony Jones <tonyj@suse.de>
Subject: Add external support and no support module to bad taint 

Patch-mainline: Never, SLES feature
References: bnc#870450
Target: SLE-12

Acked-by: Lee, Chun-Yi <jlee@suse.com>

---
 kernel/tracepoint.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/kernel/tracepoint.c
+++ b/kernel/tracepoint.c
@@ -331,7 +331,11 @@ EXPORT_SYMBOL_GPL(tracepoint_probe_unreg
 bool trace_module_has_bad_taint(struct module *mod)
 {
 	return mod->taints & ~((1 << TAINT_OOT_MODULE) | (1 << TAINT_CRAP) |
-			       (1 << TAINT_UNSIGNED_MODULE));
+			       (1 << TAINT_UNSIGNED_MODULE)
+#ifdef CONFIG_SUSE_KERNEL_SUPPORTED
+			       | (1 << TAINT_EXTERNAL_SUPPORT) | (1 << TAINT_NO_SUPPORT)
+#endif
+			       );
 }
 
 static BLOCKING_NOTIFIER_HEAD(tracepoint_notify_list);
