From: Tony Jones <tonyj@suse.de>
Subject: Add external support and no support module to bad taint 

Patch-mainline: Never, SLES feature
References: bnc#870450
Target: SLE-12

Acked-by: Lee, Chun-Yi <jlee@suse.com>

---
 kernel/tracepoint.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/kernel/tracepoint.c
+++ b/kernel/tracepoint.c
@@ -572,6 +572,9 @@ bool trace_module_has_bad_taint(struct m
 {
 	return mod->taints & ~((1 << TAINT_OOT_MODULE) | (1 << TAINT_CRAP) |
 				(1 << TAINT_UNSIGNED_MODULE) | (1 << TAINT_TEST) |
+#ifdef CONFIG_SUSE_KERNEL_SUPPORTED
+			        (1 << TAINT_EXTERNAL_SUPPORT) | (1 << TAINT_NO_SUPPORT) |
+#endif
 				(1 << TAINT_LIVEPATCH));
 }
 
