From: Yuval Mintz <yuvalmin@broadcom.com>
Date: Sun, 1 Jul 2012 03:18:57 +0000
Subject: [PATCH 3/9] tg3: set maximal number of default RSS queues
Git-commit: 118008784965003307ea164370094c7d0810546e
Patch-mainline: v3.6-rc1
References: bnc#785554

Signed-off-by: Yuval Mintz <yuvalmin@broadcom.com>
Signed-off-by: Eilon Greenstein <eilong@broadcom.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>

Cc: Matt Carlson <mcarlson@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/tg3.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/tg3.c
+++ b/drivers/net/tg3.c
@@ -9474,7 +9474,7 @@ static bool tg3_enable_msix(struct tg3 *
 	int i, rc;
 	struct msix_entry msix_ent[tp->irq_max];
 
-	tp->irq_cnt = num_online_cpus();
+	tp->irq_cnt = netif_get_num_default_rss_queues();
 	if (tp->irq_cnt > 1) {
 		/* We want as many rx rings enabled as there are cpus.
 		 * In multiqueue MSI-X mode, the first MSI-X vector
