From 1abd1a93771d6e5437bd0089b7fa1003dae9c47b Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Fri, 15 Dec 2023 20:21:52 -0500
Subject: drivers/gpu/drm/i915/i915_memcpy.c: fix missing includes
Git-commit: 86b9357c1bbe993e74a304b3f7783d7d0c79c40c
Patch-mainline: v6.8-rc1
References: drm-backport-placeholder

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
Acked-by: Patrik Jakobsson <pjakobsson@suse.de>
---
 drivers/gpu/drm/i915/i915_memcpy.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_memcpy.c b/drivers/gpu/drm/i915/i915_memcpy.c
index 1b021a4902de..ba82277254b7 100644
--- a/drivers/gpu/drm/i915/i915_memcpy.c
+++ b/drivers/gpu/drm/i915/i915_memcpy.c
@@ -23,6 +23,8 @@
  */
 
 #include <linux/kernel.h>
+#include <linux/string.h>
+#include <linux/cpufeature.h>
 #include <asm/fpu/api.h>
 
 #include "i915_memcpy.h"
-- 
2.46.1

