From: Jeff Layton <jlayton@redhat.com>
Date: Thu, 20 Dec 2012 16:15:38 -0500
Subject: [PATCH 08/28] vfs: fix linkat to retry once on ESTALE errors
Git-commit: 442e31ca5a49e398351b2954b51f578353fdf210
Patch-mainline: v3.8
References: bnc#876463

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: NeilBrown <neilb@suse.de>

---
 fs/namei.c |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

--- linux-3.0-SLE11-SP3.orig/fs/namei.c
+++ linux-3.0-SLE11-SP3/fs/namei.c
@@ -3046,12 +3046,13 @@ SYSCALL_DEFINE5(linkat, int, olddfd, con
 
 	if (flags & AT_SYMLINK_FOLLOW)
 		how |= LOOKUP_FOLLOW;
-
+retry:
 	error = user_path_at(olddfd, oldname, how, &old_path);
 	if (error)
 		return error;
 
-	error = user_path_parent(newdfd, newname, &nd, &to, 0);
+	error = user_path_parent(newdfd, newname, &nd, &to,
+					(how & LOOKUP_REVAL));
 	if (error)
 		goto out;
 	error = -EXDEV;
@@ -3077,6 +3078,11 @@ out_unlock:
 out_release:
 	path_put(&nd.path);
 	putname(to);
+	if (retry_estale(error, how)) {
+		path_put(&old_path);
+		how |= LOOKUP_REVAL;
+		goto retry;
+	}
 out:
 	path_put(&old_path);
 
