From: Jeff Layton <jlayton@redhat.com>
Date: Thu, 20 Dec 2012 16:28:33 -0500
Subject: [PATCH 10/28] vfs: make do_rmdir retry once on ESTALE errors
Git-commit: c6ee920698301febdf10df0b57039173a1edbd43
Patch-mainline: v3.8
References: bnc#876463

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: NeilBrown <neilb@suse.de>

---
 fs/namei.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- linux-3.0-SLE11-SP3.orig/fs/namei.c
+++ linux-3.0-SLE11-SP3/fs/namei.c
@@ -2738,8 +2738,9 @@ static long do_rmdir(int dfd, const char
 	char * name;
 	struct dentry *dentry;
 	struct nameidata nd;
-
-	error = user_path_parent(dfd, pathname, &nd, &name, 0);
+	unsigned int lookup_flags = 0;
+retry:
+	error = user_path_parent(dfd, pathname, &nd, &name, lookup_flags);
 	if (error)
 		return error;
 
@@ -2782,6 +2783,10 @@ exit2:
 exit1:
 	path_put(&nd.path);
 	putname(name);
+	if (retry_estale(error, lookup_flags)) {
+		lookup_flags |= LOOKUP_REVAL;
+		goto retry;
+	}
 	return error;
 }
 
