From: Jeff Layton <jlayton@redhat.com>
Date: Thu, 20 Dec 2012 16:38:04 -0500
Subject: [PATCH 11/28] vfs: make do_unlinkat retry once on ESTALE errors
Git-commit: 5d18f8133cad85ccbb7fa6fd351d75025da32504
Patch-mainline: v3.8
References: bnc#876463

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: NeilBrown <neilb@suse.de>

---
 fs/namei.c |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

--- linux-3.0-SLE11-SP3.orig/fs/namei.c
+++ linux-3.0-SLE11-SP3/fs/namei.c
@@ -2840,8 +2840,9 @@ static long do_unlinkat(int dfd, const c
 	struct dentry *dentry;
 	struct nameidata nd;
 	struct inode *inode = NULL;
-
-	error = user_path_parent(dfd, pathname, &nd, &name, 0);
+	unsigned int lookup_flags = 0;
+retry:
+	error = user_path_parent(dfd, pathname, &nd, &name, lookup_flags);
 	if (error)
 		return error;
 
@@ -2880,6 +2881,11 @@ exit3:
 exit1:
 	path_put(&nd.path);
 	putname(name);
+	if (retry_estale(error, lookup_flags)) {
+		lookup_flags |= LOOKUP_REVAL;
+		inode = NULL;
+		goto retry;
+	}
 	return error;
 
 slashes:
