
From 3808da15e33fb1020c838e2cfd0cf24321e964de Mon Sep 17 00:00:00 2001
From: Yaniv Rosner <yanivr@broadcom.com>
Date: Thu, 27 Oct 2011 05:13:52 +0000
Subject: [PATCH 13/15] bnx2x: Enable changing speed when port type is PORT_DA

Git-commit: bfdb58232f9227149cac89c17454edaf25f43136
Patch-mainline: v3.2-rc1
References: bnc#709064

Signed-off-by: Yaniv Rosner <yanivr@broadcom.com>
Signed-off-by: Eilon Greenstein <eilong@broadcom.com>
Acked-by: Lee, Chun-Yi <jlee@suse.com>
---
 drivers/net/bnx2x/bnx2x_ethtool.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/bnx2x/bnx2x_ethtool.c
+++ b/drivers/net/bnx2x/bnx2x_ethtool.c
@@ -329,6 +329,7 @@ static int bnx2x_set_settings(struct net
 			PORT_HW_CFG_PHY_SELECTION_FIRST_PHY;
 		break;
 	case PORT_FIBRE:
+	case PORT_DA:
 		if (bp->port.supported[cfg_idx] & SUPPORTED_FIBRE)
 			break; /* no port change */
 
