From: Jeff Layton <jlayton@redhat.com>
Date: Thu, 20 Dec 2012 17:08:32 -0500
Subject: [PATCH 16/28] vfs: have chroot retry once on ESTALE error
Git-commit: 2771261ec5b677a38f0cd5fcfc6cefd5393787ef
Patch-mainline: v3.8
References: bnc#876463

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: NeilBrown <neilb@suse.de>

---
 fs/open.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- linux-3.0-SLE11-SP3.orig/fs/open.c
+++ linux-3.0-SLE11-SP3/fs/open.c
@@ -441,8 +441,9 @@ SYSCALL_DEFINE1(chroot, const char __use
 {
 	struct path path;
 	int error;
-
-	error = user_path_dir(filename, &path);
+	unsigned int lookup_flags = LOOKUP_FOLLOW | LOOKUP_DIRECTORY;
+retry:
+	error = user_path_at(AT_FDCWD, filename, lookup_flags, &path);
 	if (error)
 		goto out;
 
@@ -461,6 +462,10 @@ SYSCALL_DEFINE1(chroot, const char __use
 	error = 0;
 dput_and_out:
 	path_put(&path);
+	if (retry_estale(error, lookup_flags)) {
+		lookup_flags |= LOOKUP_REVAL;
+		goto retry;
+	}
 out:
 	return error;
 }
