From: Jeff Layton <jlayton@redhat.com>
Date: Tue, 11 Dec 2012 12:10:14 -0500
Subject: [PATCH 19/28] vfs: fix user_statfs to retry once on ESTALE errors
Git-commit: 96948fc6069b68380abac2944b8b02b43a2e2057
Patch-mainline: v3.8
References: bnc#876463

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: NeilBrown <neilb@suse.de>

---
 fs/statfs.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

--- linux-3.0-SLE11-SP3.orig/fs/statfs.c
+++ linux-3.0-SLE11-SP3/fs/statfs.c
@@ -76,10 +76,17 @@ EXPORT_SYMBOL(vfs_statfs);
 int user_statfs(const char __user *pathname, struct kstatfs *st)
 {
 	struct path path;
-	int error = user_path_at(AT_FDCWD, pathname, LOOKUP_FOLLOW|LOOKUP_AUTOMOUNT, &path);
+	int error;
+	unsigned int lookup_flags = LOOKUP_FOLLOW|LOOKUP_AUTOMOUNT;
+retry:
+	error = user_path_at(AT_FDCWD, pathname, lookup_flags, &path);
 	if (!error) {
 		error = vfs_statfs(&path, st);
 		path_put(&path);
+		if (retry_estale(error, lookup_flags)) {
+			lookup_flags |= LOOKUP_REVAL;
+			goto retry;
+		}
 	}
 	return error;
 }
