From: Jeff Layton <jlayton@redhat.com>
Date: Tue, 11 Dec 2012 12:10:15 -0500
Subject: [PATCH 21/28] vfs: allow setxattr to retry once on ESTALE errors
Git-commit: 68f1bb8bb89e0bb813c893a42373a26ebdab7f9c
Patch-mainline: v3.8
References: bnc#876463

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: NeilBrown <neilb@suse.de>

---
 fs/xattr.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- linux-3.0-SLE11-SP3.orig/fs/xattr.c
+++ linux-3.0-SLE11-SP3/fs/xattr.c
@@ -303,8 +303,9 @@ SYSCALL_DEFINE5(setxattr, const char __u
 {
 	struct path path;
 	int error;
-
-	error = user_path(pathname, &path);
+	unsigned int lookup_flags = LOOKUP_FOLLOW;
+retry:
+	error = user_path_at(AT_FDCWD, pathname, lookup_flags, &path);
 	if (error)
 		return error;
 	error = mnt_want_write(path.mnt);
@@ -313,6 +314,10 @@ SYSCALL_DEFINE5(setxattr, const char __u
 		mnt_drop_write(path.mnt);
 	}
 	path_put(&path);
+	if (retry_estale(error, lookup_flags)) {
+		lookup_flags |= LOOKUP_REVAL;
+		goto retry;
+	}
 	return error;
 }
 
