From: Sathya Perla <sathya.perla@emulex.com>
Date: Mon, 22 Aug 2011 19:41:55 +0000
Subject: [PATCH] be2net: remove unused variable
References: bnc#722092
Patch-mainline: v3.1-rc1
Git-commit: 15133fbbb91ae695f153fb48daa6a1a8af4a5032

Signed-off-by: Sathya Perla <sathya.perla@emulex.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/benet/be.h |    1 -
 1 file changed, 1 deletion(-)

--- a/drivers/net/benet/be.h
+++ b/drivers/net/benet/be.h
@@ -348,7 +348,6 @@ struct be_adapter {
 	u32 beacon_state;	/* for set_phys_id */
 
 	bool eeh_err;
-	bool link_up;
 	u32 port_num;
 	bool promiscuous;
 	bool wol;
