From: Jun Nakajima <jun.nakajima@intel.com>
Date: Wed, 20 Dec 2017 08:04:49 -0800
Subject: Use the "ibrs_inuse" variable.
Patch-mainline: submitted on 2018/1/9
References: bnc#1068032 CVE-2017-5715

Signed-off-by: Jun Nakajima <jun.nakajima@intel.com>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/kvm/svm.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index a04ec1d2b5e8..c4c34d532377 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -4960,7 +4960,7 @@ static void svm_vcpu_run(struct kvm_vcpu *vcpu)
 
 	local_irq_enable();
 
-	if (static_cpu_has(X86_FEATURE_SPEC_CTRL) &&
+	if (ibrs_inuse &&
 	    svm->spec_ctrl != SPEC_CTRL_IBRS)
 		wrmsrl(MSR_IA32_SPEC_CTRL, svm->spec_ctrl);
 
@@ -5056,7 +5056,7 @@ static void svm_vcpu_run(struct kvm_vcpu *vcpu)
 #endif
 		);
 
-	if (static_cpu_has(X86_FEATURE_SPEC_CTRL)) {
+	if (ibrs_inuse) {
 		rdmsrl(MSR_IA32_SPEC_CTRL, svm->spec_ctrl);
 		if (svm->spec_ctrl != SPEC_CTRL_IBRS)
 			wrmsrl(MSR_IA32_SPEC_CTRL, SPEC_CTRL_IBRS);
-- 
2.15.1

