From: Somnath Kotur <somnath.kotur@emulex.com>
Date: Fri, 30 Sep 2011 07:24:28 +0000
Subject: [PATCH] be2net: Modified PCI MaxReadReq size to 4096 bytes
References: bnc#722092
Patch-mainline: v3.1-rc7
Git-commit: 5a56eb10babbcd7b3796dc3c28c271260aa3608d

Signed-off-by: Somnath Kotur <somnath.kotur@emulex.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/benet/be_main.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/drivers/net/benet/be_main.c b/drivers/net/benet/be_main.c
index 1a7b24c..816ce56 100644
--- a/drivers/net/benet/be_main.c
+++ b/drivers/net/benet/be_main.c
@@ -2529,6 +2529,8 @@ static int be_setup(struct be_adapter *adapter)
 	adapter->link_speed = -1;
 
 	be_cmd_get_fw_ver(adapter, adapter->fw_ver, NULL);
+
+	pcie_set_readrq(adapter->pdev, 4096);
 	return 0;
 
 rx_qs_destroy:
-- 
1.5.6.1

