From: Somnath Kotur <somnath.kotur@emulex.com>
Date: Wed, 5 Oct 2011 22:33:39 +0530
Subject: [PATCH] be2net: Changed version to 4.1.227
References: bnc#722092
Patch-mainline: Never, suse specific


Signed-off-by: Somnath Kotur <somnath.kotur@emulex.com>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/benet/be.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/benet/be.h b/drivers/net/benet/be.h
index 4fab71b..d40f551 100644
--- a/drivers/net/benet/be.h
+++ b/drivers/net/benet/be.h
@@ -34,7 +34,7 @@
 
 #include "be_hw.h"
 
-#define DRV_VER			"4.0.100u"
+#define DRV_VER			"4.1.227u"
 #define DRV_NAME		"be2net"
 #define BE_NAME			"ServerEngines BladeEngine2 10Gbps NIC"
 #define BE3_NAME		"ServerEngines BladeEngine3 10Gbps NIC"
-- 
1.7.4

