From 0119ab14c31587198daed671a5bf949df2c14552 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Thu, 17 Feb 2022 23:40:18 -0500
Subject: [PATCH] dm: remove unused mapped_device argument from free_tio
Git-commit: 0119ab14c31587198daed671a5bf949df2c14552
Patch-mainline: v5.18-rc1
References: jsc#PED-2765

Reviewed-by: Christoph Hellwig <hch@lst.de>
Reviewed-by: Mikulas Patocka <mpatocka@redhat.com>
Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Signed-off-by: Coly Li <colyli@suse.de>

---
 drivers/md/dm.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index c7246b00d363..b8cdd8c78c0e 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -556,7 +556,7 @@ static struct dm_io *alloc_io(struct mapped_device *md, struct bio *bio)
 	return io;
 }
 
-static void free_io(struct mapped_device *md, struct dm_io *io)
+static void free_io(struct dm_io *io)
 {
 	bio_put(&io->tio.clone);
 }
@@ -839,7 +839,7 @@ void dm_io_dec_pending(struct dm_io *io, blk_status_t error)
 
 		io_error = io->status;
 		dm_end_io_acct(io);
-		free_io(md, io);
+		free_io(io);
 		smp_wmb();
 		this_cpu_dec(*md->pending_io);
 
-- 
2.35.3

