From 168678d765d3659fddffe80a70668488066a9569 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Wed, 2 Mar 2022 12:19:47 -0500
Subject: [PATCH] dm mpath: use DMINFO instead of printk with KERN_INFO
Git-commit: 168678d765d3659fddffe80a70668488066a9569
Patch-mainline: v5.18-rc1
References: jsc#PED-2765

Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Signed-off-by: Coly Li <colyli@suse.de>

---
 drivers/md/dm-mpath.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/drivers/md/dm-mpath.c b/drivers/md/dm-mpath.c
index 411403b3ec43..70ffe4efc03e 100644
--- a/drivers/md/dm-mpath.c
+++ b/drivers/md/dm-mpath.c
@@ -899,8 +899,7 @@ static int setup_scsi_dh(struct block_device *bdev, struct multipath *m,
 	if (m->hw_handler_name) {
 		r = scsi_dh_attach(q, m->hw_handler_name);
 		if (r == -EBUSY) {
-			printk(KERN_INFO "dm-mpath: retaining handler on device %pg\n",
-			       bdev);
+			DMINFO("retaining handler on device %pg", bdev);
 			goto retain;
 		}
 		if (r < 0) {
-- 
2.35.3

