From: Tomas Winkler <tomas.winkler@intel.com>
Subject: mei: mei_irq_thread_write_handler check for overflow
References: fate#311770
Patch-Mainline: v3.6-rc1
Git-commit: 7d5e0e59542ac452585e3a33abcad6e26b4dec4a


Signed-off-by: Thomas Renninger <trenn@suse.de>

check for overflow when retrieving empty write slots

Signed-off-by: Tomas Winkler <tomas.winkler@intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/drivers/misc/mei/interrupt.c b/drivers/misc/mei/interrupt.c
index 83e80bc..8e4dd74 100644
--- a/drivers/misc/mei/interrupt.c
+++ b/drivers/misc/mei/interrupt.c
@@ -1204,6 +1204,9 @@ static int mei_irq_thread_write_handler(struct mei_io_list *cmpl_list,
 		return 0;
 	}
 	*slots = mei_count_empty_write_slots(dev);
+	if (*slots <= 0)
+		return -EMSGSIZE;
+
 	/* complete all waiting for write CB */
 	dev_dbg(&dev->pdev->dev, "complete all waiting for write cb.\n");
 
