From: Tomas Winkler <tomas.winkler@intel.com>
Subject: mei: mei_wd_host_init: update the comment
References: fate#311770
Patch-Mainline: v3.6-rc1
Git-commit: 7cb1ba9b679afe5fc335205fd6fb25fd3e51e33a


Signed-off-by: Thomas Renninger <trenn@suse.de>

Signed-off-by: Tomas Winkler <tomas.winkler@intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/drivers/misc/mei/wd.c b/drivers/misc/mei/wd.c
index 325d26f..0a89338 100644
--- a/drivers/misc/mei/wd.c
+++ b/drivers/misc/mei/wd.c
@@ -53,11 +53,12 @@ static void mei_wd_set_start_timeout(struct mei_device *dev, u16 timeout)
 }
 
 /**
- * host_init_wd - mei initialization wd.
+ * mei_wd_host_init - connect to the watchdog client
  *
  * @dev: the device structure
  * returns -ENENT if wd client cannot be found
  *         -EIO if write has failed
+ *         0 on success
  */
 int mei_wd_host_init(struct mei_device *dev)
 {
