From: Steve French <stfrench@microsoft.com>
Date: Fri, 9 Jul 2021 13:02:26 -0500
Subject: [PATCH] cifs: update internal version number
Git-commit: 4d069f6022e938bc51667da637f2483a37a77e19
Patch-mainline: v5.14-rc1
References: bsc#1192606

To 2.33

Signed-off-by: Steve French <stfrench@microsoft.com>
Acked-by: Enzo Matsumiya <ematsumiya@suse.de>
---
 fs/cifs/cifsfs.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/cifs/cifsfs.h b/fs/cifs/cifsfs.h
index 177f3e7ab86d..d25a4099b32e 100644
--- a/fs/cifs/cifsfs.h
+++ b/fs/cifs/cifsfs.h
@@ -153,5 +153,5 @@ extern struct dentry *cifs_smb3_do_mount(struct file_system_type *fs_type,
 extern const struct export_operations cifs_export_ops;
 #endif /* CONFIG_CIFS_NFSD_EXPORT */
 
-#define CIFS_VERSION   "2.32"
+#define CIFS_VERSION   "2.33"
 #endif				/* _CIFSFS_H */
-- 
2.33.1

