From 9cd6c339e34ab13479fb139b0b9f5ffc9743bfc1 Mon Sep 17 00:00:00 2001
From: Joonas Lahtinen <joonas.lahtinen@linux.intel.com>
Date: Fri, 27 Sep 2019 15:25:54 +0300
Subject: drm/i915: Update DRIVER_DATE to 20190927
Git-commit: 9cd6c339e34ab13479fb139b0b9f5ffc9743bfc1
Patch-mainline: v5.5-rc1
References: bsc#1152489

Signed-off-by: Joonas Lahtinen <joonas.lahtinen@linux.intel.com>
Acked-by: Thomas Zimmermann <tzimmermann@suse.de>
---
 drivers/gpu/drm/i915/i915_drv.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 0e4e77373ee5..c0f6c1f72b62 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -106,8 +106,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20190822"
-#define DRIVER_TIMESTAMP	1566477988
+#define DRIVER_DATE		"20190927"
+#define DRIVER_TIMESTAMP	1569587154
 
 struct drm_i915_gem_object;
 
-- 
2.28.0

