From 49530f22f02030b9cd7aff52a98f7fe8ec2e81ac Mon Sep 17 00:00:00 2001
From: Rodrigo Vivi <rodrigo.vivi@intel.com>
Date: Mon, 24 Feb 2020 17:45:39 -0800
Subject: drm/i915: Update DRIVER_DATE to 20200224
Git-commit: 12b3788093cb73722b3471335fddb11d8e652ad2
Patch-mainline: v5.7-rc1
References: jsc#SLE-12680, jsc#SLE-12880, jsc#SLE-12882, jsc#SLE-12883, jsc#SLE-13496, jsc#SLE-15322

Signed-off-by: Rodrigo Vivi <rodrigo.vivi@intel.com>
Signed-off-by: Patrik Jakobsson <pjakobsson@suse.de>
---
 drivers/gpu/drm/i915/i915_drv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 9bc94a511403..88e4fb8ac739 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -113,7 +113,7 @@
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
 #define DRIVER_DATE		"20200224"
-#define DRIVER_TIMESTAMP	1582594340
+#define DRIVER_TIMESTAMP	1582595138
 
 struct drm_i915_gem_object;
 
-- 
2.28.0

