From 907f74f6f6798e90a2342395894ca8b00fca4412 Mon Sep 17 00:00:00 2001
From: Rodrigo Vivi <rodrigo.vivi@intel.com>
Date: Tue, 25 Feb 2020 10:41:22 -0800
Subject: drm/i915: Update DRIVER_DATE to 20200225
Git-commit: 53e3ca6749186b5c147964bddc4eb47ba8b5f69e
Patch-mainline: v5.7-rc1
References: jsc#SLE-12680, jsc#SLE-12880, jsc#SLE-12882, jsc#SLE-12883, jsc#SLE-13496, jsc#SLE-15322

Signed-off-by: Rodrigo Vivi <rodrigo.vivi@intel.com>
Signed-off-by: Patrik Jakobsson <pjakobsson@suse.de>
---
 drivers/gpu/drm/i915/i915_drv.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 88e4fb8ac739..ea13fc0b409b 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -112,8 +112,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20200224"
-#define DRIVER_TIMESTAMP	1582595138
+#define DRIVER_DATE		"20200225"
+#define DRIVER_TIMESTAMP	1582656081
 
 struct drm_i915_gem_object;
 
-- 
2.28.0

