From 4d9295ca938da9a305fbd8cbc434259e68542c08 Mon Sep 17 00:00:00 2001
From: Danny Al-Gaaf <danny.al-gaaf@suse.de>
Date: Mon, 3 Jun 2013 18:10:51 +0200
Subject: [PATCH 934/938] Add define for SIZE_MAX as in include/linux/kernel.h
 on recent kernels
References: fate#312983
Patch-mainline: No, when/where was it submitted? -jeffm

Signed-off-by: Danny Al-Gaaf <danny.al-gaaf@suse.de>

---
 drivers/block/rbd.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/block/rbd.c b/drivers/block/rbd.c
index e9fcadf..4e68d16 100644
--- a/drivers/block/rbd.c
+++ b/drivers/block/rbd.c
@@ -44,6 +44,7 @@
 
 #include "rbd_types.h"
 
+#define SIZE_MAX       (~(size_t)0)
 #define RBD_DEBUG	/* Activate rbd_assert() calls */
 
 /*
-- 
1.8.3

