From 761802113bac49f4e25e9cce8b1bfb0d41601015 Mon Sep 17 00:00:00 2001
From: Danny Al-Gaaf <danny.al-gaaf@suse.de>
Date: Mon, 3 Jun 2013 18:15:47 +0200
Subject: [PATCH 935/938] Add define for SIZE_MAX as in include/linux/kernel.h
 on recent kernels
References: fate#312983
Patch-mainline: No, when/where was it submitted? -jeffm

Signed-off-by: Danny Al-Gaaf <danny.al-gaaf@bisect.de>
Acked-by: Danny Al-Gaaf <dalgaaf@suse.de>

---
 net/ceph/messenger.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/net/ceph/messenger.c b/net/ceph/messenger.c
index 3107d3e..8a2d1e6 100644
--- a/net/ceph/messenger.c
+++ b/net/ceph/messenger.c
@@ -100,6 +100,8 @@
 #define CON_FLAG_SOCK_CLOSED	   3  /* socket state changed to closed */
 #define CON_FLAG_BACKOFF           4  /* need to retry queuing delayed work */
 
+#define SIZE_MAX       (~(size_t)0)
+
 static bool con_flag_valid(unsigned long con_flag)
 {
 	switch (con_flag) {
-- 
1.8.3

