From e5b9da1eea78dda4c0bf191b1706f872c0f27644 Mon Sep 17 00:00:00 2001
From: Danny Al-Gaaf <danny.al-gaaf@suse.de>
Date: Mon, 3 Jun 2013 20:13:06 +0200
Subject: [PATCH 939/939] remove include of linux/export.h which isn't needed
References: fate#312983
Patch-mainline: No, when/where was it submitted? -jeffm

Signed-off-by: Danny Al-Gaaf <danny.al-gaaf@suse.de>

---
 net/ceph/snapshot.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/net/ceph/snapshot.c b/net/ceph/snapshot.c
index 154683f..5d97e95 100644
--- a/net/ceph/snapshot.c
+++ b/net/ceph/snapshot.c
@@ -21,7 +21,6 @@
 #include <stddef.h>
 
 #include <linux/types.h>
-#include <linux/export.h>
 #include <linux/ceph/libceph.h>
 
 /*
-- 
1.8.3

