From ddce0caaae6d6ce6978be4f0291a6001f45c16ba Mon Sep 17 00:00:00 2001
From: Francois Dugast <francois.dugast@intel.com>
Date: Fri, 12 May 2023 14:10:04 +0200
Subject: drm/xe: Remove unused define
Git-commit: d0e96f3d5255f62bc9721392b198acc4d302de32
Patch-mainline: v6.8-rc1
References: drm-backport-placeholder

Signed-off-by: Francois Dugast <francois.dugast@intel.com>
Reviewed-by: Matt Atwood <matthew.s.atwood@intel.com>
Signed-off-by: Rodrigo Vivi <rodrigo.vivi@intel.com>
Acked-by: Patrik Jakobsson <pjakobsson@suse.de>
---
 drivers/gpu/drm/xe/xe_drv.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/gpu/drm/xe/xe_drv.h b/drivers/gpu/drm/xe/xe_drv.h
index 0377e5e4e35f..d45b71426cc8 100644
--- a/drivers/gpu/drm/xe/xe_drv.h
+++ b/drivers/gpu/drm/xe/xe_drv.h
@@ -11,7 +11,6 @@
 #define DRIVER_NAME		"xe"
 #define DRIVER_DESC		"Intel Xe Graphics"
 #define DRIVER_DATE		"20201103"
-#define DRIVER_TIMESTAMP	1604406085
 
 /* Interface history:
  *
-- 
2.46.1

