From be6133b8917f8131b76728f11db51a65c91b3fec Mon Sep 17 00:00:00 2001
From: Rodrigo Vivi <rodrigo.vivi@intel.com>
Date: Tue, 13 Aug 2019 23:59:53 -0700
Subject: drm/i915: Update DRIVER_DATE to 20190813
Git-commit: be6133b8917f8131b76728f11db51a65c91b3fec
Patch-mainline: v5.4-rc1
References: bsc#1152489

Signed-off-by: Rodrigo Vivi <rodrigo.vivi@intel.com>
Acked-by: Thomas Zimmermann <tzimmermann@suse.de>
---
 drivers/gpu/drm/i915/i915_drv.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index d6a35f8c7dd9..50effdccc76c 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -104,8 +104,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20190730"
-#define DRIVER_TIMESTAMP	1564512624
+#define DRIVER_DATE		"20190813"
+#define DRIVER_TIMESTAMP	1565765993
 
 struct drm_i915_gem_object;
 
-- 
2.28.0

