From d967e794680efe26ac071380df690fb280a8f9df Mon Sep 17 00:00:00 2001
From: Jani Nikula <jani.nikula@intel.com>
Date: Wed, 15 Jul 2020 14:18:02 +0300
Subject: drm/i915: Update DRIVER_DATE to 20200715
Git-commit: e57bd05ec0d2d82d63725dedf9f5a063f879de25
Patch-mainline: v5.9-rc1
References: jsc#SLE-12680, jsc#SLE-12880, jsc#SLE-12882, jsc#SLE-12883, jsc#SLE-13496, jsc#SLE-15322

Signed-off-by: Jani Nikula <jani.nikula@intel.com>
Signed-off-by: Patrik Jakobsson <pjakobsson@suse.de>
---
 drivers/gpu/drm/i915/i915_drv.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 9ba6cfff9e3f..e4f7f6518945 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -108,8 +108,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20200702"
-#define DRIVER_TIMESTAMP	1593714328
+#define DRIVER_DATE		"20200715"
+#define DRIVER_TIMESTAMP	1594811881
 
 struct drm_i915_gem_object;
 
-- 
2.29.2

