From 1883e2999f045e1fd6f76a7f30288a5312085289 Mon Sep 17 00:00:00 2001
From: Joonas Lahtinen <joonas.lahtinen@linux.intel.com>
Date: Fri, 1 Nov 2019 12:41:13 +0200
Subject: drm/i915: Update DRIVER_DATE to 20191101
Git-commit: 1883e2999f045e1fd6f76a7f30288a5312085289
Patch-mainline: v5.5-rc1
References: bsc#1152489

Signed-off-by: Joonas Lahtinen <joonas.lahtinen@linux.intel.com>
Acked-by: Thomas Zimmermann <tzimmermann@suse.de>
---
 drivers/gpu/drm/i915/i915_drv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 4bfd842e2914..1e6118f62b29 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -111,7 +111,7 @@
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
 #define DRIVER_DATE		"20191101"
-#define DRIVER_TIMESTAMP	1572591869
+#define DRIVER_TIMESTAMP	1572604873
 
 struct drm_i915_gem_object;
 
-- 
2.28.0

