From: jbeulich@suse.com
Subject: x86/retpoline/entry: convert unwind assembly indirect jumps
Patch-mainline: never, SUSE specific
References: bsc#1068032 CVE-2017-5715

--- a/arch/x86/kernel/entry_32.S
+++ b/arch/x86/kernel/entry_32.S
@@ -1083,7 +1083,7 @@ ENTRY(arch_unwind_init_running)
 	movl	%ebx, PT_EFLAGS(%edx)
 	movl	PT_EBX(%edx), %ebx
 	movl	$__KERNEL_DS, PT_OLDSS(%edx)
-	jmpl	*%eax
+	JMP_NOSPEC %eax
 	CFI_ENDPROC
 ENDPROC(arch_unwind_init_running)
 #endif
--- a/arch/x86/kernel/entry_64.S
+++ b/arch/x86/kernel/entry_64.S
@@ -1367,7 +1367,7 @@ ENTRY(arch_unwind_init_running)
 	movq	%rax, EFLAGS(%rdi)
 	movq	%r9, RSP(%rdi)
 	movq	$__KERNEL_DS, SS(%rdi)
-	jmpq	*%rcx
+	JMP_NOSPEC %rcx
 	CFI_ENDPROC
 END(arch_unwind_init_running)
 #endif
