From: Michal Srb <msrb@suse.com>
Patch-mainline: never
Subject: kbuild: fix drm include

Acked-by: Michal Srb <msrb@suse.com>
---
Index: linux-3.0-SLE11-SP3/include/Kbuild
===================================================================
--- linux-3.0-SLE11-SP3.orig/include/Kbuild
+++ linux-3.0-SLE11-SP3/include/Kbuild
@@ -7,6 +7,5 @@ header-y += sound/
 header-y += mtd/
 header-y += rdma/
 header-y += video/
-header-y += drm/
 header-y += xen/
 header-y += scsi/
