From f85f35d3257b79ad85bb3f140976cd24b26ac632 Mon Sep 17 00:00:00 2001
From: David Disseldorp <ddiss@suse.de>
Date: Thu, 14 Aug 2025 10:29:50 +0200
Subject: [PATCH] bcachefs: print message at mount time regarding imminent
 removal
References: boo#1248109
Patch-mainline: Never, Tumbleweed specific

Signed-off-by: David Disseldorp <ddiss@suse.de>
---
 fs/bcachefs/super.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/fs/bcachefs/super.c b/fs/bcachefs/super.c
index c46b1053a02c9..343252d489c6c 100644
--- a/fs/bcachefs/super.c
+++ b/fs/bcachefs/super.c
@@ -1241,6 +1241,10 @@ int bch2_fs_start(struct bch_fs *c)
 		bch_err_msg(c, ret, "starting filesystem");
 	else
 		bch_verbose(c, "done starting filesystem");
+
+	pr_crit("Bcachefs may be removed from the kernel very soon. See:\n"
+	  "https://lore.kernel.org/all/CAHk-=wi+k8E4kWR8c-nREP0+EA4D+=rz5j0Hdk3N6cWgfE03-Q@mail.gmail.com/\n");
+
 	return ret;
 }
 
-- 
2.50.1

