From: Roland Dreier <roland@purestorage.com>
Date: Wed, 29 Feb 2012 20:40:46 +0000
Subject: [PATCH 050/143] be2net: Remove unused OFFSET_IN_PAGE() macro
Patch-mainline: v3.4-rc1
Git-commit: 97767a87f3be8834192dc3fc9412aaccf708d87f
References: bnc#777565 FATE#313819

Signed-off-by: Roland Dreier <roland@purestorage.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/emulex/benet/be.h |    4 ----
 1 file changed, 4 deletions(-)

--- a/drivers/net/ethernet/emulex/benet/be.h
+++ b/drivers/net/ethernet/emulex/benet/be.h
@@ -442,10 +442,6 @@ extern const struct ethtool_ops be_ethto
 		((u32)((((size_t)(_address) & (PAGE_SIZE_4K - 1)) + 	\
 			(size) + (PAGE_SIZE_4K - 1)) >> PAGE_SHIFT_4K))
 
-/* Byte offset into the page corresponding to given address */
-#define OFFSET_IN_PAGE(addr)						\
-		 ((size_t)(addr) & (PAGE_SIZE_4K-1))
-
 /* Returns bit offset within a DWORD of a bitfield */
 #define AMAP_BIT_OFFSET(_struct, field)  				\
 		(((size_t)&(((_struct *)0)->field))%32)
