From: Hannes Reinecke <hare@suse.de>
Subject: blk: Re-add symbols
Date: Tue 25 Mar 2025 01:12:56 PM CET
References: jsc#PED-9651
Patch-Mainline: never, kABI specific

Commit fe6628608627424fb4a6d4c8d2235822457c5d9c removed some symbols,
which we need to re-add to preserve kABI.

Signed-off-by: Hannes Reinecke <hare@suse.de>

diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index 76f0a4e7c2e5..248416ecd01c 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -561,6 +561,7 @@ struct request_queue {
 	struct list_head	flush_list;
 
 	struct mutex		sysfs_lock;
+	struct mutex		sysfs_dir_lock;
 	struct mutex		limits_lock;
 
 	/*
@@ -605,6 +604,8 @@ struct request_queue {
 	 * Serializes all debugfs metadata operations using the above dentries.
 	 */
 	struct mutex		debugfs_mutex;
+
+	bool			mq_sysfs_init_done;
 };
 
 /* Keep blk_queue_flag_name[] in sync with the definitions below */
