From: Weiping Zhang <zhangweiping@didiglobal.com>
Date: Thu, 27 Feb 2020 09:38:46 +0800
Subject: [PATCH] blk-iocost: remove duplicated lines in comments
References: bsc#1175995,jsc#SLE-15608
Git-commit: fa800d73c8d0d36b1f5929198371f421b69e610e
Patch-mainline: v5.7-rc1

Acked-by: Tejun Heo <tj@kernel.org>
Signed-off-by: Weiping Zhang <zhangweiping@didiglobal.com>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/blk-iocost.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/block/blk-iocost.c b/block/blk-iocost.c
index 7b42d1944748..9f95d8cc58dd 100644
--- a/block/blk-iocost.c
+++ b/block/blk-iocost.c
@@ -46,9 +46,6 @@
  * If needed, tools/cgroup/iocost_coef_gen.py can be used to generate
  * device-specific coefficients.
  *
- * If needed, tools/cgroup/iocost_coef_gen.py can be used to generate
- * device-specific coefficients.
- *
  * 2. Control Strategy
  *
  * The device virtual time (vtime) is used as the primary control metric.
-- 
2.16.4

