From: Christoph Hellwig <hch@lst.de>
Date: Wed, 13 May 2020 12:49:32 +0200
Subject: [PATCH] block: mark blk_account_io_completion static
References: bsc#1175995,jsc#SLE-15608
Git-commit: f1394b79881458c3cfb12c1cac813085549f2c7d
Patch-mainline: v5.8-rc1

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/blk-core.c | 2 +-
 block/blk.h      | 1 -
 2 files changed, 1 insertion(+), 2 deletions(-)

diff --git a/block/blk-core.c b/block/blk-core.c
index 5e69c99f7d57..27551e421fba 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -1349,7 +1349,7 @@ unsigned int blk_rq_err_bytes(const struct request *rq)
 }
 EXPORT_SYMBOL_GPL(blk_rq_err_bytes);
 
-void blk_account_io_completion(struct request *req, unsigned int bytes)
+static void blk_account_io_completion(struct request *req, unsigned int bytes)
 {
 	if (req->part && blk_do_io_stat(req)) {
 		const int sgrp = op_stat_group(req_op(req));
diff --git a/block/blk.h b/block/blk.h
index e4386f682183..c91d0abdbcac 100644
--- a/block/blk.h
+++ b/block/blk.h
@@ -186,7 +186,6 @@ bool blk_attempt_plug_merge(struct request_queue *q, struct bio *bio,
 		unsigned int nr_segs, struct request **same_queue_rq);
 
 void blk_account_io_start(struct request *req, bool new_io);
-void blk_account_io_completion(struct request *req, unsigned int bytes);
 void blk_account_io_done(struct request *req, u64 now);
 
 /*
-- 
2.16.4

