From: Christoph Hellwig <hch@lst.de>
Date: Tue, 3 Dec 2019 10:39:03 +0100
Subject: [PATCH] block: remove the empty line at the end of blk-zoned.c
References: bsc#1175995,jsc#SLE-15608
Git-commit: bb55628288fcd96d919a9ecc59dd26704a65493b
Patch-mainline: v5.5-rc1

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/blk-zoned.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/block/blk-zoned.c b/block/blk-zoned.c
index 3e08983c9a87..23b558340a18 100644
--- a/block/blk-zoned.c
+++ b/block/blk-zoned.c
@@ -470,4 +470,3 @@ int blk_revalidate_disk_zones(struct gendisk *disk)
 	return ret;
 }
 EXPORT_SYMBOL_GPL(blk_revalidate_disk_zones);
-
-- 
2.16.4

