From: Christoph Hellwig <hch@lst.de>
Date: Fri, 26 Jun 2020 10:01:57 +0200
Subject: [PATCH] block: remove the unused bd_private field from struct
Git-commit: 47b5e00322a3033851ab304f3c3873aebdfb4979
Patch-mainline: v5.9-rc1
References: bsc#1175995,jsc#SLE-15608
 block_device

Reviewed-by: Johannes Thumshirn <johannes.thumshirn@wdc.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Acked-by: Hannes Reinecke <hare@suse.com>
---
 include/linux/blk_types.h | 7 -------
 1 file changed, 7 deletions(-)

diff --git a/include/linux/blk_types.h b/include/linux/blk_types.h
index 667cd365fd04..b5f7105806e4 100644
--- a/include/linux/blk_types.h
+++ b/include/linux/blk_types.h
@@ -41,13 +41,6 @@ struct block_device {
 	struct gendisk *	bd_disk;
 	struct backing_dev_info *bd_bdi;
 	struct list_head	bd_list;
-	/*
-	 * Private data.  You must have bd_claim'ed the block_device
-	 * to use this.  NOTE:  bd_claim allows an owner to claim
-	 * the same device multiple times, the owner must take special
-	 * care to not mess up bd_private for that case.
-	 */
-	unsigned long		bd_private;
 
 	/* The counter of freeze processes */
 	int			bd_fsfreeze_count;
-- 
2.16.4

