From: Thomas Abraham <tabraham@suse.com>
Subject: [PATCH] block: add flag QUEUE_FLAG_REGISTERED
References: bsc#1047027
Patch-mainline: no, upstream differs

Signed-off-by: Thomas Abraham <tabraham@suse.com>
diff --git a/block/blk-sysfs.c b/block/blk-sysfs.c
index c7d14ee23125..88631c91ec63 100644
--- a/block/blk-sysfs.c
+++ b/block/blk-sysfs.c
@@ -533,6 +533,11 @@ int blk_register_queue(struct gendisk *disk)
 	if (WARN_ON(!q))
 		return -ENXIO;
 
+	WARN_ONCE(test_bit(QUEUE_FLAG_REGISTERED, &q->queue_flags),
+		"%s is registering an already registered queue\n",
+		kobject_name(&dev->kobj));
+	queue_flag_set_unlocked(QUEUE_FLAG_REGISTERED, q);
+
 	ret = blk_trace_init_sysfs(dev);
 	if (ret)
 		return ret;
diff --git a/block/elevator.c b/block/elevator.c
index d06905f34905..db1f03933828 100644
--- a/block/elevator.c
+++ b/block/elevator.c
@@ -973,6 +973,10 @@ int elevator_change(struct request_queue *q, const char *name)
 	if (!q->elevator)
 		return -ENXIO;
 
+	/* Make sure queue is not in the middle of being removed */
+	if (!test_bit(QUEUE_FLAG_REGISTERED, &q->queue_flags))
+		return -ENOENT;
+
 	strlcpy(elevator_name, name, sizeof(elevator_name));
 	e = elevator_get(strstrip(elevator_name));
 	if (!e) {
diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index 916b92f3964b..404c41664030 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -416,7 +416,7 @@ struct request_queue
 #define QUEUE_FLAG_SECDISCARD  17	/* supports SECDISCARD */
 #define QUEUE_FLAG_NO_ROUND    18	/* Don't round timeout up to next second */
 #define QUEUE_FLAG_DEAD        19	/* queue tear-down finished */
-
+#define QUEUE_FLAG_REGISTERED  26       /* queue has been registered to a disk */
 #define QUEUE_FLAG_DEFAULT	((1 << QUEUE_FLAG_IO_STAT) |		\
 				 (1 << QUEUE_FLAG_STACKABLE)	|	\
 				 (1 << QUEUE_FLAG_SAME_COMP)	|	\
