From: Yuval Mintz <yuvalmin@broadcom.com>
Date: Mon, 2 Jul 2012 08:56:58 +0300
Subject: [PATCH 08/11] bnx2x: update driver version
Patch-mainline: Never, not applicable, version numbers have diverged
References: bnc#769035

Signed-off-by: Yuval Mintz <yuvalmin@broadcom.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/bnx2x/bnx2x.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/bnx2x/bnx2x.h
+++ b/drivers/net/bnx2x/bnx2x.h
@@ -22,7 +22,7 @@
  * (you will need to reboot afterwards) */
 /* #define BNX2X_STOP_ON_ERROR */
 
-#define DRV_MODULE_VERSION      "1.70.35-0"
+#define DRV_MODULE_VERSION      "1.70.35-1"
 #define DRV_MODULE_RELDATE      "2011/11/10"
 #define BNX2X_BC_VER            0x040200
 
