From cfe64549d6a7bd4187c6cebc4c37b100c796bb98 Mon Sep 17 00:00:00 2001
From: Dmitry Kravkov <dmitry@broadcom.com>
Date: Tue, 23 Aug 2011 14:04:58 +0300
Subject: [PATCH 2/2] bnx2x: advance version
Patch-mainline: never
References: bnc#698050, fate#311461


Signed-off-by: Dmitry Kravkov <dmitry@broadcom.com>
Acked-by: Michal Marek <mmarek@suse.cz>
---
 drivers/net/bnx2x/bnx2x.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/bnx2x/bnx2x.h b/drivers/net/bnx2x/bnx2x.h
index ea7a394..11d4d46 100644
--- a/drivers/net/bnx2x/bnx2x.h
+++ b/drivers/net/bnx2x/bnx2x.h
@@ -22,7 +22,7 @@
  * (you will need to reboot afterwards) */
 /* #define BNX2X_STOP_ON_ERROR */
 
-#define DRV_MODULE_VERSION      "1.70.00-0"
+#define DRV_MODULE_VERSION      "1.70.00-1"
 #define DRV_MODULE_RELDATE      "2011/06/13"
 #define BNX2X_BC_VER            0x040200
 
-- 
1.7.2.2

