From: Yuval Mintz <yuvalmin@broadcom.com>
Date: Sun, 1 Jul 2012 03:18:58 +0000
Subject: [PATCH 23/27] bnx2: set maximal number of default RSS queues
Patch-mainline: v3.6-rc1
Git-commit: 0a742128db29031e55d27e37b1b098433c56033a
References: bnc#790590 FATE#313913

Signed-off-by: Yuval Mintz <yuvalmin@broadcom.com>
Signed-off-by: Eilon Greenstein <eilong@broadcom.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>

Cc: Michael Chan <mchan@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/ethernet/broadcom/bnx2.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/ethernet/broadcom/bnx2.c
+++ b/drivers/net/ethernet/broadcom/bnx2.c
@@ -6249,7 +6249,7 @@ bnx2_enable_msix(struct bnx2 *bp, int ms
 static int
 bnx2_setup_int_mode(struct bnx2 *bp, int dis_msi)
 {
-	int cpus = num_online_cpus();
+	int cpus = netif_get_num_default_rss_queues();
 	int msix_vecs;
 
 	if (!bp->num_req_rx_rings)
