From: stephen hemminger <stephen@networkplumber.org>
Date: Wed, 10 Apr 2013 10:53:40 +0000
Subject: [PATCH v2 04/31] bnx2: make cnic_probe static
Patch-mainline: v3.10-rc1

Git-commit: 61c2fc4b50b9db025f7bb78cd1d4ab9de7d740cf
References: bsc#908686 FATE#317538


Signed-off-by: Stephen Hemminger <stephen@networkplumber.org>
Acked-by: Michael Chan <mchan@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Cho, Yu-Chen <acho@suse.com>
---
 drivers/net/ethernet/broadcom/bnx2.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/ethernet/broadcom/bnx2.c
+++ b/drivers/net/ethernet/broadcom/bnx2.c
@@ -419,7 +419,7 @@ static int bnx2_unregister_cnic(struct n
 	return 0;
 }
 
-struct cnic_eth_dev *bnx2_cnic_probe(struct net_device *dev)
+static struct cnic_eth_dev *bnx2_cnic_probe(struct net_device *dev)
 {
 	struct bnx2 *bp = netdev_priv(dev);
 	struct cnic_eth_dev *cp = &bp->cnic_eth_dev;
