From: David Sterba <dsterba@suse.cz>
Date: Mon, 02 Jan 2012 13:40:28 +0100
Subject: [PATCH] btrfs: use correct device for maps
Reference: bnc#672923 bnc#769545
Patch-mainline: no

Signed-off-by: David Sterba <dsterba@suse.cz>
---

Index: linux-3.1-openSUSE-12.1/fs/proc/task_mmu.c
===================================================================
--- linux-3.1-openSUSE-12.1.orig/fs/proc/task_mmu.c
+++ linux-3.1-openSUSE-12.1/fs/proc/task_mmu.c
@@ -220,7 +220,15 @@ static void show_map_vma(struct seq_file
 
 	if (file) {
 		struct inode *inode = vma->vm_file->f_path.dentry->d_inode;
-		dev = inode->i_sb->s_dev;
+
+		if (inode->i_sb->s_magic == BTRFS_SUPER_MAGIC) {
+			struct kstat stat;
+
+			vfs_getattr(file->f_path.mnt, file->f_path.dentry, &stat);
+			dev = stat.dev;
+		} else {
+			dev = inode->i_sb->s_dev;
+		}
 		ino = inode->i_ino;
 		pgoff = ((loff_t)vma->vm_pgoff) << PAGE_SHIFT;
 	}
Index: linux-3.1-openSUSE-12.1/fs/proc/task_nommu.c
===================================================================
--- linux-3.1-openSUSE-12.1.orig/fs/proc/task_nommu.c
+++ linux-3.1-openSUSE-12.1/fs/proc/task_nommu.c
@@ -148,7 +148,15 @@ static int nommu_vma_show(struct seq_fil
 
 	if (file) {
 		struct inode *inode = vma->vm_file->f_path.dentry->d_inode;
-		dev = inode->i_sb->s_dev;
+
+		if (inode->i_sb->s_magic == BTRFS_SUPER_MAGIC) {
+			struct kstat stat;
+
+			vfs_getattr(file->f_path.mnt, file->f_path.dentry, &stat);
+			dev = stat.dev;
+		} else {
+			dev = inode->i_sb->s_dev;
+		}
 		ino = inode->i_ino;
 		pgoff = (loff_t)vma->vm_pgoff << PAGE_SHIFT;
 	}
