From: Ilya Dryomov <idryomov@gmail.com>
Date: Fri, 4 Nov 2011 09:41:02 -0400
Patch-mainline: v3.2-rc1
References: FATE#306586
Git-commit: 9510dc4c62252e96dca44a94948327c76774f25b
Subject: [PATCH] Btrfs: stop leaking btrfs_bios on readahead

Signed-off-by: Chris Mason <chris.mason@oracle.com>
Signed-off-by: David Sterba <dsterba@suse.cz>
---
 fs/btrfs/reada.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/fs/btrfs/reada.c
+++ b/fs/btrfs/reada.c
@@ -423,6 +423,7 @@ again:
 	}
 	spin_unlock(&fs_info->reada_lock);
 
+	kfree(bbio);
 	return re;
 
 error:
@@ -447,6 +448,7 @@ error:
 		kref_put(&zone->refcnt, reada_zone_release);
 		spin_unlock(&fs_info->reada_lock);
 	}
+	kfree(bbio);
 	kfree(re);
 	if (looped)
 		goto again;

