From: Thomas Renninger <trenn@suse.de>
Subject: Fix compiler warning, no functional change
References: bnc#740281
Patch-Mainline: never -> these parts seem to be SLE11 SP2 only

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 drivers/block/cciss.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-3.0-SLE11-SP2/drivers/block/cciss.c
===================================================================
--- linux-3.0-SLE11-SP2.orig/drivers/block/cciss.c
+++ linux-3.0-SLE11-SP2/drivers/block/cciss.c
@@ -4268,7 +4268,7 @@ struct cciss_match_arg {
 static int cciss_match_dev(struct device *dev, void *data)
 {
 	struct cciss_match_arg *arg = data;
-	struct pci_device_id *dev_id = cciss_pci_device_id;
+	const struct pci_device_id *dev_id = cciss_pci_device_id;
 	struct pci_dev *pdev = to_pci_dev(dev);
 
 	/* Skip non-HP devices */
