From: Mark Gross <mgross@linux.intel.com>
Subject: [PATCH] Disable CET when calling tboot->shutdown_entry passed in from
 tboot boot parrameters.
References: bsc#1247950
Patch-mainline: Never, test patch


Acked-by: Nikolay Borisov <nik.borisov@suse.com>
---
 arch/x86/kernel/tboot.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/arch/x86/kernel/tboot.c
+++ b/arch/x86/kernel/tboot.c
@@ -28,6 +28,7 @@
 #include <asm/setup.h>
 #include <asm/e820/api.h>
 #include <asm/io.h>
+#include <asm/cpu.h>

 #include "../realmode/rm/wakeup.h"

@@ -247,6 +248,7 @@ void tboot_shutdown(u32 shutdown_type)
 	tboot->shutdown_type = shutdown_type;

 	switch_to_tboot_pt();
+	cet_disable();

 	shutdown = (void(*)(void))(unsigned long)tboot->shutdown_entry;
 	shutdown();
