From: Steve French <stfrench@microsoft.com>
Date: Thu, 19 Jan 2023 15:20:21 -0600
Subject: [PATCH] update internal module version number for cifs.ko
Git-commit: fdbf807215250217c83f1cb715b883cd910102fa
References: bsc#1193629
Patch-mainline: v6.3-rc1

From 2.41 to 2.42

Signed-off-by: Steve French <stfrench@microsoft.com>
Acked-by: Paulo Alcantara <palcantara@suse.de>
---
 fs/cifs/cifsfs.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/cifs/cifsfs.h b/fs/cifs/cifsfs.h
index ea628da503c6..845a0dda5864 100644
--- a/fs/cifs/cifsfs.h
+++ b/fs/cifs/cifsfs.h
@@ -160,5 +160,5 @@ extern const struct export_operations cifs_export_ops;
 
 /* when changing internal version - update following two lines at same time */
 #define SMB3_PRODUCT_BUILD 41
-#define CIFS_VERSION   "2.41"
+#define CIFS_VERSION   "2.42"
 #endif				/* _CIFSFS_H */
-- 
2.39.2


