From 0903a0c8491c1e987dfc6eb294199a36760398bc Mon Sep 17 00:00:00 2001
From: Al Viro <viro@zeniv.linux.org.uk>
Date: Thu, 29 Nov 2012 22:11:06 -0500
Subject: [PATCH] cifs: get rid of blind d_drop() in readdir
References: bnc#831143
Patch-mainline: yes
Git-commit: 0903a0c8491c1e987dfc6eb294199a36760398bc

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: David Disseldorp <ddiss@suse.de>
---
 fs/cifs/readdir.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

Index: linux-3.0-bnc831143_11sp2_master_reparse_backport/fs/cifs/readdir.c
===================================================================
--- linux-3.0-bnc831143_11sp2_master_reparse_backport.orig/fs/cifs/readdir.c
+++ linux-3.0-bnc831143_11sp2_master_reparse_backport/fs/cifs/readdir.c
@@ -85,14 +85,17 @@ cifs_readdir_lookup(struct dentry *paren
 
 	dentry = d_lookup(parent, name);
 	if (dentry) {
+		int err;
 		inode = dentry->d_inode;
 		/* update inode in place if i_ino didn't change */
 		if (inode && CIFS_I(inode)->uniqueid == fattr->cf_uniqueid) {
 			cifs_fattr_to_inode(inode, fattr);
 			return dentry;
 		}
-		d_drop(dentry);
+		err = d_invalidate(dentry);
 		dput(dentry);
+		if (err)
+			return NULL;
 	}
 
 	dentry = d_alloc(parent, name);
