bug 41243, fix ckrm handling of bad values from userland

Index: linux.t/fs/rcfs/magic.c
===================================================================
--- linux.t.orig/fs/rcfs/magic.c	2004-05-26 14:22:00.291272184 -0400
+++ linux.t/fs/rcfs/magic.c	2004-05-26 14:22:04.680604904 -0400
@@ -426,7 +426,7 @@ shares_write(struct file *file, const ch
 	int rc = 0;
 	struct ckrm_core_class *core;
 	int done;
-	char *resname;
+	char *resname = NULL;
 
 	struct ckrm_shares newshares = {
 		CKRM_SHARE_UNCHANGED,
@@ -454,6 +454,11 @@ shares_write(struct file *file, const ch
 	
 	core = ri->core; 
 	optbuf = kmalloc(SHARES_MAX_INPUT_SIZE, GFP_KERNEL);
+	if (!optbuf) {
+		up(&inode->i_sem);
+		return -ENOMEM ;
+	}
+
 	__copy_from_user(optbuf, buf, count);
 	if (optbuf[count-1] == '\n')
 		optbuf[count-1]='\0';
