bug 40937

diff -BbadpruN linux-2.6.5-ckrm-e12/fs/rcfs/dir.c linux-2.6.5-ckrm-e12-enotempty/fs/rcfs/dir.c
--- linux-2.6.5-ckrm-e12/fs/rcfs/dir.c	2004-05-14 16:40:57.000000000 -0700
+++ linux-2.6.5-ckrm-e12-enotempty/fs/rcfs/dir.c	2004-05-14 16:41:21.000000000 -0700
@@ -211,7 +211,7 @@ rcfs_rmdir(struct inode * dir, struct de
 	
 	if (!rcfs_empty(dentry)) {
 		printk(KERN_ERR "rcfs_rmdir: directory not empty\n");
-		goto out;
+		return -ENOTEMPTY;
 	}
 
 	// Core class removal 
