bug 40396 

diff -Nru a/kernel/ckrm/ckrm.c b/kernel/ckrm/ckrm.c
--- a/kernel/ckrm/ckrm.c	Fri May 14 18:15:10 2004
+++ b/kernel/ckrm/ckrm.c	Fri May 14 18:15:10 2004
@@ -901,6 +901,9 @@
 	return rc;
 }
 
+#define legalshare(a)   \
+         ( ((a) >=0) || ((a) == CKRM_SHARE_UNCHANGED) || ((a) == CKRM_SHARE_DONTCARE) ) 
+
 int
 ckrm_class_set_shares(struct ckrm_core_class *core, const char *resname,
 		      struct ckrm_shares *shares)
@@ -909,7 +912,11 @@
 	struct ckrm_res_ctlr *rcbs;
 	int rc;
 
-	printk("ckrm_class_set_shares(%s,%s)\n",core->name,resname);
+	// Check for legal values
+	if ( !legalshare(shares->my_guarantee) || !legalshare(shares->my_limit) || 
+	     !legalshare(shares->total_guarantee) || !legalshare(shares->max_limit)) 
+		return -EINVAL;
+
 	rcbs = ckrm_resctlr_lookup(clstype,resname);
 	if (rcbs == NULL || rcbs->set_share_values == NULL)
 		return -EINVAL; 
