diff -Nru -X ./dontdiff a/net/ipv4/tcp_ipv4.c b/net/ipv4/tcp_ipv4.c
--- a/net/ipv4/tcp_ipv4.c	Mon May  3 10:06:46 2004
+++ b/net/ipv4/tcp_ipv4.c	Mon May  3 15:45:06 2004
@@ -1445,8 +1445,7 @@
 
 #ifdef CONFIG_ACCEPT_QUEUES
 	class = (skb->nfmark <= 0) ? 0 :
-		((skb->nfmark > NUM_ACCEPT_QUEUES) ? NUM_ACCEPT_QUEUES:
-		 skb->nfmark);
+		((skb->nfmark >= NUM_ACCEPT_QUEUES) ? 0: skb->nfmark);
 	/*
 	 * Accept only if the class has shares set or if the default class
 	 * i.e. class 0 has shares
