From b414e8ecd49867d9c83614800d405419e4f58e94 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Sat, 4 Mar 2023 22:45:27 -0500
Subject: [PATCH] closures: Add a missing include
Git-commit: b414e8ecd49867d9c83614800d405419e4f58e94
Patch-mainline: v6.7-rc1
References: jsc#PED-7542

Fixes building in userspace.

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
Signed-off-by: Coly Li <colyli@suse.de>

---
 lib/closure.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/closure.c b/lib/closure.c
index b38ded00b9b0..0855e698ced1 100644
--- a/lib/closure.c
+++ b/lib/closure.c
@@ -9,6 +9,7 @@
 #include <linux/closure.h>
 #include <linux/debugfs.h>
 #include <linux/export.h>
+#include <linux/rcupdate.h>
 #include <linux/seq_file.h>
 #include <linux/sched/debug.h>
 
-- 
2.35.3

