From: Steve Wise <swise@opengridcomputing.com>
Date: Wed, 9 Apr 2014 09:38:27 -0500
Subject: [PATCH 133/223] RDMA/cxgb4: Initialize reserved fields in a FW work
 request
Patch-mainline: v3.15-rc2
Git-commit: c3f98fa29176753a759ade424f18b11f440b19f4
References: bsc#909577 FATE#317550

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
Acked-by: David Chang <dchang@suse.com>
Signed-off-by: John Jolly <jjolly@suse.de>
---
 drivers/infiniband/hw/cxgb4/qp.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/infiniband/hw/cxgb4/qp.c
+++ b/drivers/infiniband/hw/cxgb4/qp.c
@@ -442,6 +442,8 @@ static int build_rdma_send(struct t4_sq
 	default:
 		return -EINVAL;
 	}
+	wqe->send.r3 = 0;
+	wqe->send.r4 = 0;
 
 	plen = 0;
 	if (wr->num_sge) {
