From: Steve Wise <swise@opengridcomputing.com>
Date: Fri, 21 Mar 2014 20:40:30 +0530
Subject: [PATCH 114/223] RDMA/cxgb4: Don't leak skb in c4iw_uld_rx_handler()
Patch-mainline: v3.15-rc1
Git-commit: dbb084cc5f52152f53b5fd22fa76b9bf69904594
References: bsc#909577 FATE#317550

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
Acked-by: David Chang <dchang@suse.com>
Signed-off-by: John Jolly <jjolly@suse.de>
---
 drivers/infiniband/hw/cxgb4/device.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

--- a/drivers/infiniband/hw/cxgb4/device.c
+++ b/drivers/infiniband/hw/cxgb4/device.c
@@ -836,11 +836,13 @@ static int c4iw_uld_rx_handler(void *han
 	}
 
 	opcode = *(u8 *)rsp;
-	if (c4iw_handlers[opcode])
+	if (c4iw_handlers[opcode]) {
 		c4iw_handlers[opcode](dev, skb);
-	else
+	} else {
 		pr_info("%s no handler opcode 0x%x...\n", __func__,
 		       opcode);
+		kfree_skb(skb);
+	}
 
 	return 0;
 nomem:
