From: Steve Wise <swise@opengridcomputing.com>
Date: Fri, 21 Mar 2014 20:40:34 +0530
Subject: [PATCH 118/223] RDMA/cxgb4: Update snd_seq when sending MPA messages
Patch-mainline: v3.15-rc1
Git-commit: 9c88aa003d26e9f1e9ea6e08511768c2ef666654
References: bsc#909577 FATE#317550

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
Acked-by: David Chang <dchang@suse.com>
Signed-off-by: John Jolly <jjolly@suse.de>
---
 drivers/infiniband/hw/cxgb4/cm.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/infiniband/hw/cxgb4/cm.c
+++ b/drivers/infiniband/hw/cxgb4/cm.c
@@ -680,6 +680,7 @@ static void send_mpa_req(struct c4iw_ep
 	start_ep_timer(ep);
 	state_set(&ep->com, MPA_REQ_SENT);
 	ep->mpa_attr.initiator = 1;
+	ep->snd_seq += mpalen;
 	return;
 }
 
@@ -759,6 +760,7 @@ static int send_mpa_reject(struct c4iw_e
 	t4_set_arp_err_handler(skb, NULL, arp_failure_discard);
 	BUG_ON(ep->mpa_skb);
 	ep->mpa_skb = skb;
+	ep->snd_seq += mpalen;
 	return c4iw_l2t_send(&ep->com.dev->rdev, skb, ep->l2t);
 }
 
@@ -843,6 +845,7 @@ static int send_mpa_reply(struct c4iw_ep
 	t4_set_arp_err_handler(skb, NULL, arp_failure_discard);
 	ep->mpa_skb = skb;
 	state_set(&ep->com, MPA_REP_SENT);
+	ep->snd_seq += mpalen;
 	return c4iw_l2t_send(&ep->com.dev->rdev, skb, ep->l2t);
 }
 
