diff -Naur linux.org/include/linux/device.h linux.dev_evlog.patched/include/linux/device.h
--- linux.org/include/linux/device.h	Tue Mar 16 11:30:12 2004
+++ linux.dev_evlog.patched/include/linux/device.h	Tue Mar 16 11:30:12 2004
@@ -394,8 +394,18 @@
 extern void firmware_unregister(struct subsystem *);
 
 /* debugging and troubleshooting/diagnostic helpers. */
+#ifdef CONFIG_EVLOG
+#include <linux/evlog.h>
+#define dev_printk(level, dev, format, arg...)	\
+    do { \
+	printk(level "%s %s: " format , (dev)->driver->name , (dev)->bus_id , ## arg); \
+	evl_printk((dev)->driver->name , 0, (level[1]-'0') , \
+	    "%s %s: " format , (dev)->driver->name , (dev)->bus_id , ## arg); \
+    } while (0)
+#else
 #define dev_printk(level, dev, format, arg...)	\
 	printk(level "%s %s: " format , (dev)->driver->name , (dev)->bus_id , ## arg)
+#endif
 
 #ifdef DEBUG
 #define dev_dbg(dev, format, arg...)		\
