From 45974c3a74bf5ba93066363f9325a0546620962c Mon Sep 17 00:00:00 2001
From: Petr Uzel <petr.uzel@suse.cz>
Date: Thu, 28 Nov 2013 11:13:18 +0100
Subject: [PATCH] Fix return value from lockspace_busy()
Git-Commit: a97f4a66d8ee4faf0f31cc4ad3f4aa0baef23fc7
Patch-Mainline: v3.13
References: https://bugzilla.novell.com/show_bug.cgi?id=827670#c43

Signed-off-by: Petr Uzel <petr.uzel@suse.cz>
Signed-off-by: Mark Fasheh <mfasheh@suse.de>
---
 fs/dlm/lockspace.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/fs/dlm/lockspace.c b/fs/dlm/lockspace.c
index 14cbf40..95efae5 100644
--- a/fs/dlm/lockspace.c
+++ b/fs/dlm/lockspace.c
@@ -660,7 +660,8 @@ static int lockspace_busy(struct dlm_ls *ls)
 			lkb_found = 1;
 			list_for_each_entry(lkb, &ls->ls_lkbtbl[i].list,
 					    lkb_idtbl_list) {
-				if (!lkb->lkb_nodeid) {
+				if (!lkb->lkb_nodeid &&
+					  lkb->lkb_grmode != DLM_LOCK_IV) {
 					read_unlock(&ls->ls_lkbtbl[i].lock);
 					return 2;
 				}
-- 
1.8.1.4

