From: Susan LeGendre-McGhee <slegendr@redhat.com>
Date: Tue, 24 Sep 2024 12:11:43 -0400
Subject: [PATCH] dm-vdo: reset bi_ioprio to the default value when the bio is
 reset
Git-commit: bd7e677c6bc4b577192f96ffeb6f965f2dbc8ecb
Patch-mainline: v6.13-rc1
References: jsc#PED-9651

Signed-off-by: Susan LeGendre-McGhee <slegendr@redhat.com>
Signed-off-by: Matthew Sakai <msakai@redhat.com>
Signed-off-by: Mikulas Patocka <mpatocka@redhat.com>
Acked-by: Hannes Reinecke <hare@suse.com>
---
 drivers/md/dm-vdo/vio.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/md/dm-vdo/vio.c b/drivers/md/dm-vdo/vio.c
index b291578f726f..e710f3c5a972 100644
--- a/drivers/md/dm-vdo/vio.c
+++ b/drivers/md/dm-vdo/vio.c
@@ -202,6 +202,7 @@ int vio_reset_bio(struct vio *vio, char *data, bio_end_io_t callback,
 	if (data == NULL)
 		return VDO_SUCCESS;
 
+	bio->bi_ioprio = 0;
 	bio->bi_io_vec = bio->bi_inline_vecs;
 	bio->bi_max_vecs = vio->block_count + 1;
 	len = VDO_BLOCK_SIZE * vio->block_count;
-- 
2.35.3

