From d70898e4ea7680e4b9d49cf41ace141a464455eb Mon Sep 17 00:00:00 2001
From: Rodrigo Vivi <rodrigo.vivi@intel.com>
Date: Tue, 20 Aug 2019 09:55:48 -0700
Subject: drm/i915: Update DRIVER_DATE to 20190820
Git-commit: d70898e4ea7680e4b9d49cf41ace141a464455eb
Patch-mainline: v5.4-rc1
References: bsc#1152489

Signed-off-by: Rodrigo Vivi <rodrigo.vivi@intel.com>
Acked-by: Thomas Zimmermann <tzimmermann@suse.de>
---
 drivers/gpu/drm/i915/i915_drv.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 67b7722f9d4d..f506cda92f1e 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -105,8 +105,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20190813"
-#define DRIVER_TIMESTAMP	1565765993
+#define DRIVER_DATE		"20190820"
+#define DRIVER_TIMESTAMP	1566320148
 
 struct drm_i915_gem_object;
 
-- 
2.28.0

