From: Egbert Eich <eich@suse.com>
Date: Tue Jul 16 10:43:20 2013 +0200
Subject: drm/mgag200: Add sysfs support for connectors
Patch-Mainline: v3.11-rc4
Git-commit: 3d5a1c5e300483df005c81c55792268d4a7bff9f
Signed-off-by: Egbert Eich <eich@suse.com>

Signed-off-by: Egbert Eich <eich@suse.com>
---
 drivers/gpu/drm/mgag200/mgag200_mode.c |    2 ++
 1 file changed, 2 insertions(+)

--- linux-3.0-SLE11-SP3.orig/drivers/gpu/drm/mgag200/mgag200_mode.c
+++ linux-3.0-SLE11-SP3/drivers/gpu/drm/mgag200/mgag200_mode.c
@@ -1625,6 +1625,8 @@ static struct drm_connector *mga_vga_ini
 
 	drm_connector_helper_add(connector, &mga_vga_connector_helper_funcs);
 
+	drm_sysfs_connector_add(connector);
+
 	mga_connector->i2c = mgag200_i2c_create(dev);
 	if (!mga_connector->i2c)
 		DRM_ERROR("failed to add ddc bus\n");
