From: Fabian Frederick <fabf@skynet.be>
Date: Mon, 14 Jul 2014 18:30:57 +0200
Subject: drop_monitor: remove unnecessary break after return
Patch-mainline: v3.17-rc1
Git-commit: 4d3520cb529e520554906573e4be0efaa38050fe
References: FATE#318825

Signed-off-by: Fabian Frederick <fabf@skynet.be>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 net/core/drop_monitor.c |    2 --
 1 file changed, 2 deletions(-)

--- a/net/core/drop_monitor.c
+++ b/net/core/drop_monitor.c
@@ -284,10 +284,8 @@ static int net_dm_cmd_trace(struct sk_bu
 	switch (info->genlhdr->cmd) {
 	case NET_DM_CMD_START:
 		return set_all_monitor_traces(TRACE_ON);
-		break;
 	case NET_DM_CMD_STOP:
 		return set_all_monitor_traces(TRACE_OFF);
-		break;
 	}
 
 	return -ENOTSUPP;
