From: Richard Cochran <richardcochran@gmail.com>
Date: Wed, 4 Apr 2012 17:43:31 +0000
Subject: e100: Support the get_ts_info ethtool method.
Patch-mainline: v3.5-rc1
Git-commit: abe0c5d165a422c438db51c0817b54eba1c4e85c
References: bnc#798929 FATE#313650

Signed-off-by: Richard Cochran <richardcochran@gmail.com>
Tested-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/intel/e100.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/ethernet/intel/e100.c
+++ b/drivers/net/ethernet/intel/e100.c
@@ -2708,6 +2708,7 @@ static const struct ethtool_ops e100_eth
 	.set_phys_id		= e100_set_phys_id,
 	.get_ethtool_stats	= e100_get_ethtool_stats,
 	.get_sset_count		= e100_get_sset_count,
+	.get_ts_info		= ethtool_op_get_ts_info,
 };
 
 static int e100_do_ioctl(struct net_device *netdev, struct ifreq *ifr, int cmd)
