From: Bruce Allan <bruce.w.allan@intel.com>
Date: Sat, 29 Dec 2012 09:08:50 +0000
Subject: e1000e: remove prototype of non-existent function
Patch-mainline: v3.9-rc1
Git-commit: fc915e93226cf00f4b8d1263ce2711959f9b1b76
References: bug#909495 FATE#317394

Signed-off-by: Bruce Allan <bruce.w.allan@intel.com>
Tested-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/intel/e1000e/e1000.h |    2 --
 1 file changed, 2 deletions(-)

--- a/drivers/net/ethernet/intel/e1000e/e1000.h
+++ b/drivers/net/ethernet/intel/e1000e/e1000.h
@@ -563,8 +563,6 @@ extern void e1000e_write_itr(struct e100
 extern unsigned int copybreak;
 extern int entropy;
 
-extern char *e1000e_get_hw_dev_name(struct e1000_hw *hw);
-
 extern const struct e1000_info e1000_82571_info;
 extern const struct e1000_info e1000_82572_info;
 extern const struct e1000_info e1000_82573_info;
