From: Stephen Hemminger <shemminger@vyatta.com>
Date: Fri, 7 Sep 2012 09:33:15 -0700
Subject: netdev: make pci_error_handlers const
Patch-mainline: v3.7-rc1
Git-commit: 3646f0e5c97703cecfd96cdabb085e78d9b7f680 (partial)
References: bnc#798929 FATE#313650

Signed-off-by: Stephen Hemminger <shemminger@vyatta.com>
Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/intel/e100.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/ethernet/intel/e100.c
+++ b/drivers/net/ethernet/intel/e100.c
@@ -3115,7 +3115,7 @@ static void e100_io_resume(struct pci_de
 	}
 }
 
-static struct pci_error_handlers e100_err_handler = {
+static const struct pci_error_handlers e100_err_handler = {
 	.error_detected = e100_io_error_detected,
 	.slot_reset = e100_io_slot_reset,
 	.resume = e100_io_resume,
