From: Tushar Dave <tushar.n.dave@intel.com>
Date: Tue, 12 Jun 2012 13:03:29 +0000
Subject: [PATCH 36/51] e1000: Combining Bitwise OR in one expression.
Patch-mainline: v3.6-rc1
Git-commit: 7500673be30f9467cd1a0e065c93e75c5213b44d
References: bnc#795304 FATE#313649

Signed-off-by: Tushar Dave <tushar.n.dave@intel.com>
Tested-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Acked-by: John Ronciak <john.ronciak@intel.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/intel/e1000/e1000_main.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/drivers/net/ethernet/intel/e1000/e1000_main.c
+++ b/drivers/net/ethernet/intel/e1000/e1000_main.c
@@ -1086,9 +1086,9 @@ static int __devinit e1000_probe(struct
 		netdev->vlan_features |= NETIF_F_HIGHDMA;
 	}
 
-	netdev->vlan_features |= NETIF_F_TSO;
-	netdev->vlan_features |= NETIF_F_HW_CSUM;
-	netdev->vlan_features |= NETIF_F_SG;
+	netdev->vlan_features |= (NETIF_F_TSO |
+				  NETIF_F_HW_CSUM |
+				  NETIF_F_SG);
 
 	adapter->en_mng_pt = e1000_enable_mng_pass_thru(hw);
 
