From: Tobias Klauser <tklauser@distanz.ch>
Date: Sun, 3 Jul 2011 23:47:04 +0000
Subject: net: e1000e: Use is_multicast_ether_addr helper
Patch-mainline: v3.1-rc1
Git-commit: 3e714ad3c2a07ee120044b72222cc20c14959efb
References: bnc#795174 FATE#313648

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/e1000e/lib.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
--- a/drivers/net/e1000e/lib.c
+++ b/drivers/net/e1000e/lib.c
@@ -221,7 +221,7 @@ s32 e1000_check_alt_mac_addr_generic(str
 	}
 
 	/* if multicast bit is set, the alternate address will not be used */
-	if (alt_mac_addr[0] & 0x01) {
+	if (is_multicast_ether_addr(alt_mac_addr)) {
 		e_dbg("Ignoring Alternate Mac Address with MC bit set\n");
 		goto out;
 	}
