From: Thomas Renninger <trenn@suse.de>
Subject: Higher CPU usage with the latest kernel                                                                  

--- linux-2.6.10_i386_ACPI_CACHE_ENABLED/drivers/acpi/Kconfig.orig	2005-02-09 13:58:58.067585392 +0100
+++ linux-2.6.10_i386_ACPI_CACHE_ENABLED/drivers/acpi/Kconfig	2005-02-09 19:21:32.695234752 +0100
@@ -355,5 +355,11 @@
 	  one found in the initrd, the DSDT from the BIOS is used. Even you do not 
 	  need a new one at the moment, you may want to use a better implemented 
 	  DSDT later. It is save to say yes here
-
+config ACPI_ENABLE_OBJECT_CACHE
+	bool "Enable ACPI object cache"
+	depends on ACPI
+	default y
+	help
+	  Improves performance for ACPI drivers especially when ACPI_DEBUG is
+          compiled in.
 endmenu
