From: Vasanthy Kolluri <vkolluri@cisco.com>
Subject: enic: suse specific version
Patch-mainline: Never, suse specific
References: bnc#638258


Signed-off-by: Vasanthy Kolluri <vkolluri@cisco.com>
Signed-off-by: Brandon Philips <bphilips@suse.de>

---
 drivers/net/enic/enic.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/enic/enic.h
+++ b/drivers/net/enic/enic.h
@@ -32,7 +32,7 @@
 
 #define DRV_NAME		"enic"
 #define DRV_DESCRIPTION		"Cisco VIC Ethernet NIC Driver"
-#define DRV_VERSION		"1.4.1.2"
+#define DRV_VERSION		"1.4.1.2-1"
 #define DRV_COPYRIGHT		"Copyright 2008-2010 Cisco Systems, Inc"
 #define PFX			DRV_NAME ": "
 
