From cb883c622e93f9db176783161b610f3966fc20ed Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Wed, 24 Jan 2018 17:36:27 +0100
Subject: [PATCH] Fix build error in drmem.c

References: bsc#1077428
Patch-mainline: no, not needed

Signed-off-by: Michal Suchanek <msuchanek@suse.de>
---
 arch/powerpc/mm/drmem.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/powerpc/mm/drmem.c b/arch/powerpc/mm/drmem.c
index 1604110c4238..cf27d14248b5 100644
--- a/arch/powerpc/mm/drmem.c
+++ b/arch/powerpc/mm/drmem.c
@@ -15,6 +15,7 @@
 #include <linux/of.h>
 #include <linux/of_fdt.h>
 #include <linux/memblock.h>
+#include <linux/slab.h>
 #include <asm/prom.h>
 #include <asm/drmem.h>
 
-- 
2.13.6

