From: Christoph Hellwig <hch@lst.de>
Date: Sat, 20 Jun 2020 09:16:37 +0200
Subject: [PATCH] fs: remove an unused block_device_operations forward
Git-commit: 7dbac5baa887facf80373825b8f66c626703621f
Patch-mainline: v5.9-rc1
References: bsc#1175995,jsc#SLE-15608
 declaration

Signed-off-by: Christoph Hellwig <hch@lst.de>
Reviewed-by: Bart Van Assche <bvanassche@acm.org>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Acked-by: Hannes Reinecke <hare@suse.com>
---
 include/linux/fs.h | 2 --
 1 file changed, 2 deletions(-)

diff --git a/include/linux/fs.h b/include/linux/fs.h
index b1c960e9b84e..0d282c853691 100644
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -1774,8 +1774,6 @@ struct dir_context {
 	loff_t pos;
 };
 
-struct block_device_operations;
-
 /* These macros are for out of kernel modules to test that
  * the kernel supports the unlocked_ioctl and compat_ioctl
  * fields in struct file_operations. */
-- 
2.16.4

