Subject: [PATCH] export writeback_bdev and writeback_inode                                                                    
From: Olaf Hering <olh@suse.de>                                                                                               

*** Warning: "writeback_bdev" [fs/msdos/msdos.ko] undefined!
*** Warning: "writeback_inode" [fs/msdos/msdos.ko] undefined!

Signed-off-by: Olaf Hering <olh@suse.de>

 fs/fs-writeback.c |    2 ++
 1 files changed, 2 insertions(+)

Index: linux-2.6.15/fs/fs-writeback.c
===================================================================
--- linux-2.6.15.orig/fs/fs-writeback.c
+++ linux-2.6.15/fs/fs-writeback.c
@@ -397,6 +397,7 @@ writeback_bdev(struct super_block *sb)
 	filemap_flush(mapping);
 	blk_run_address_space(mapping);
 }
+EXPORT_SYMBOL_GPL(writeback_bdev);
 
 void
 writeback_inode(struct inode *inode)
@@ -410,6 +411,7 @@ writeback_inode(struct inode *inode)
 	sync_inode(inode, &wbc);
 	filemap_fdatawrite(mapping);
 }
+EXPORT_SYMBOL_GPL(writeback_inode);
 
 /*
  * Start writeback of dirty pagecache data against all unlocked inodes.
