From a63e9b2e76632b3ccb0f33f43484c0c64d601849 Mon Sep 17 00:00:00 2001
From: Jan Kara <jack@suse.cz>
Date: Tue, 12 Jun 2012 16:20:26 +0200
Subject: [PATCH 5/8] gfs2: Push file_update_time() into gfs2_page_mkwrite()
Git-commit: a63e9b2e76632b3ccb0f33f43484c0c64d601849
Patch-mainline: v3.6-rc1
References: bnc#804734

Cc: Steven Whitehouse <swhiteho@redhat.com>
Cc: cluster-devel@redhat.com
Acked-by: Steven Whitehouse <swhiteho@redhat.com>
Signed-off-by: Jan Kara <jack@suse.cz>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>

---
 fs/gfs2/file.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

Index: linux-3.0-SLE11-SP3/fs/gfs2/file.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/fs/gfs2/file.c
+++ linux-3.0-SLE11-SP3/fs/gfs2/file.c
@@ -357,6 +357,9 @@ static int gfs2_page_mkwrite(struct vm_a
 	struct gfs2_alloc *al;
 	int ret;
 
+	/* Update file times before taking page lock */
+	file_update_time(vma->vm_file);
+
 	gfs2_holder_init(ip->i_gl, LM_ST_EXCLUSIVE, 0, &gh);
 	ret = gfs2_glock_nq(&gh);
 	if (ret)
