From 120c2bcad80c0f7c6691ad70f7fb1e709854d725 Mon Sep 17 00:00:00 2001
From: Jan Kara <jack@suse.cz>
Date: Tue, 12 Jun 2012 16:20:25 +0200
Subject: [PATCH 4/8] 9p: Push file_update_time() into v9fs_vm_page_mkwrite()
Git-commit: 120c2bcad80c0f7c6691ad70f7fb1e709854d725
Patch-mainline: v3.6-rc1
References: bnc#804734

Cc: Eric Van Hensbergen <ericvh@gmail.com>
Cc: Ron Minnich <rminnich@sandia.gov>
Cc: Latchesar Ionkov <lucho@ionkov.net>
Cc: v9fs-developer@lists.sourceforge.net
Signed-off-by: Jan Kara <jack@suse.cz>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>

---
 fs/9p/vfs_file.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

Index: linux-3.0-SLE11-SP3/fs/9p/vfs_file.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/fs/9p/vfs_file.c
+++ linux-3.0-SLE11-SP3/fs/9p/vfs_file.c
@@ -610,6 +610,9 @@ v9fs_vm_page_mkwrite(struct vm_area_stru
 	P9_DPRINTK(P9_DEBUG_VFS, "page %p fid %lx\n",
 		   page, (unsigned long)filp->private_data);
 
+	/* Update file times before taking page lock */
+	file_update_time(filp);
+
 	v9inode = V9FS_I(inode);
 	/* make sure the cache has finished storing the page */
 	v9fs_fscache_wait_on_page_write(inode, page);
