From: "Stephen M. Cameron" <scameron@beardog.cce.hp.com>
Date: Fri, 21 Feb 2014 16:24:54 -0600
Subject: hpsa: remove unused kthread.h header
References: bnc#913241,FATE#317338
Patch-Mainline: v3.16
Git-commit: 9f02e5bc60d699017bc1f5e5cc95c4fedc7b29ba

Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/scsi/hpsa.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index 1b174ff..5525d7f2 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -46,8 +46,7 @@
 #include <linux/cciss_ioctl.h>
 #include <linux/string.h>
 #include <linux/bitmap.h>
-#include <asm/atomic.h>
-#include <linux/kthread.h>
+#include <linux/atomic.h>
 #include <linux/jiffies.h>
 #include <asm/div64.h>
 #include "hpsa_cmd.h"
-- 
1.8.5.2

