From: Al Viro <viro@zeniv.linux.org.uk>
Date: Mon, 17 Mar 2025 22:06:04 -0400
Subject: hypfs_create_cpu_files(): add missing check for hypfs_mkdir() failure
Git-commit: 00cdfdcfa0806202aea56b02cedbf87ef1e75df8
Patch-mainline: v6.15-rc1
References: git-fixes bsc#1243839

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: Miroslav Franc <mfranc@suse.cz>
---
 arch/s390/hypfs/hypfs_diag_fs.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/s390/hypfs/hypfs_diag_fs.c b/arch/s390/hypfs/hypfs_diag_fs.c
index 00a6d370a280..280266a74f37 100644
--- a/arch/s390/hypfs/hypfs_diag_fs.c
+++ b/arch/s390/hypfs/hypfs_diag_fs.c
@@ -208,6 +208,8 @@ static int hypfs_create_cpu_files(struct dentry *cpus_dir, void *cpu_info)
 	snprintf(buffer, TMP_SIZE, "%d", cpu_info__cpu_addr(diag204_get_info_type(),
 							    cpu_info));
 	cpu_dir = hypfs_mkdir(cpus_dir, buffer);
+	if (IS_ERR(cpu_dir))
+		return PTR_ERR(cpu_dir);
 	rc = hypfs_create_u64(cpu_dir, "mgmtime",
 			      cpu_info__acc_time(diag204_get_info_type(), cpu_info) -
 			      cpu_info__lp_time(diag204_get_info_type(), cpu_info));

