From 53d0d12127093bc4ab425a5d9cd651fc5d616c3e Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Mon, 20 Jan 2025 13:42:26 +0100
Subject: [PATCH 2/4] i3c: adding kABI padding
References: bsc#1220369
Patch-mainline: Never , kABI padding

to facilitate backporting future fixes

Signed-off-by: oneukum@suse.com
---
 include/linux/i3c/master.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/i3c/master.h b/include/linux/i3c/master.h
index 93fad480a..61275f9e9 100644
--- a/include/linux/i3c/master.h
+++ b/include/linux/i3c/master.h
@@ -483,6 +483,8 @@ struct i3c_master_controller_ops {
 	int (*enable_hotjoin)(struct i3c_master_controller *master);
 	int (*disable_hotjoin)(struct i3c_master_controller *master);
 	int (*set_speed)(struct i3c_master_controller *master, enum i3c_open_drain_speed speed);
+
+	void *suse_kabi_padding;
 };
 
 /**
-- 
2.47.1

