From 9683d0e21e3d6127314bec4744ed4cadec897003 Mon Sep 17 00:00:00 2001
From: Nicolas Morey <nmorey@suse.com>
Date: Tue, 19 May 2026 16:07:32 +0200
Subject: [PATCH 1/1] IB: add suse_kabi_padding to ib_device_ops
Patch-mainline: Never, in-house patch
References: jsc#PED-15314

Signed-off-by: Nicolas Morey <nmorey@suse.com>
Acked-by: Nicolas Morey <nmorey@suse.com>
---
 include/rdma/ib_verbs.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/rdma/ib_verbs.h b/include/rdma/ib_verbs.h
index 41f298848893..327e2ec76ddf 100644
--- a/include/rdma/ib_verbs.h
+++ b/include/rdma/ib_verbs.h
@@ -2786,6 +2786,8 @@ struct ib_device_ops {
 	DECLARE_RDMA_OBJ_SIZE(ib_ucontext);
 	DECLARE_RDMA_OBJ_SIZE(ib_xrcd);
 	DECLARE_RDMA_OBJ_SIZE(rdma_counter);
+
+	void *suse_kabi_padding[4];
 };
 
 struct ib_core_device {
-- 
2.54.0

