Subject: [PATCH] ieee80211_crypt_tkip depends on NET_RADIO                                                                                                           
From: Olaf Hering <olh@suse.de>                                                                                                                                      

*** Warning: ".wireless_send_event" [net/ieee80211/ieee80211_crypt_tkip.ko] undefined!

Signed-off-by: Olaf Hering <olh@suse.de>

 net/ieee80211/Kconfig |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.14/net/ieee80211/Kconfig
===================================================================
--- linux-2.6.14.orig/net/ieee80211/Kconfig
+++ linux-2.6.14/net/ieee80211/Kconfig
@@ -55,7 +55,7 @@ config IEEE80211_CRYPT_CCMP
 
 config IEEE80211_CRYPT_TKIP
 	tristate "IEEE 802.11i TKIP encryption"
-	depends on IEEE80211
+	depends on IEEE80211 && NET_RADIO
 	select CRYPTO
 	select CRYPTO_MICHAEL_MIC
 	---help---
