From: Tobias Klauser <tklauser@distanz.ch>
Date: Sun, 3 Jul 2011 23:50:15 +0000
Subject: net: igb: Use is_multicast_ether_addr helper
Patch-mainline: v3.1-rc1
Git-commit: 58d14d4fbaeb0cede14c59e8a2648b4791c0afe2
References: bnc#795297 FATE#313656

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: David Chang <dchang@suse.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/igb/e1000_mac.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/net/igb/e1000_mac.c
+++ b/drivers/net/igb/e1000_mac.c
@@ -29,6 +29,7 @@
 #include <linux/delay.h>
 #include <linux/pci.h>
 #include <linux/netdevice.h>
+#include <linux/etherdevice.h>
 
 #include "e1000_mac.h"
 
@@ -224,7 +225,7 @@ s32 igb_check_alt_mac_addr(struct e1000_
 	}
 
 	/* if multicast bit is set, the alternate address will not be used */
-	if (alt_mac_addr[0] & 0x01) {
+	if (is_multicast_ether_addr(alt_mac_addr)) {
 		hw_dbg("Ignoring Alternate Mac Address with MC bit set\n");
 		goto out;
 	}
