From 1e9e3281c2c22c00607709c9bffb14372e28bc26 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Mon, 20 Jan 2025 14:10:39 +0100
Subject: [PATCH 3/4] input: add kABI padding
References: bsc#1220369
Patch-mainline: Never , kABI padding

to facilitte future bug fixes

Signed-off-by : oneukum@suse.com
---
 include/linux/input.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/linux/input.h b/include/linux/input.h
index 49790c1bd..a66404eaf 100644
--- a/include/linux/input.h
+++ b/include/linux/input.h
@@ -208,6 +208,8 @@ struct input_dev {
 
 	ktime_t timestamp[INPUT_CLK_MAX];
 
+	void *suse_kabi_padding;
+
 	bool inhibited;
 };
 #define to_input_dev(d) container_of(d, struct input_dev, dev)
@@ -320,6 +322,8 @@ struct input_handler {
 	void (*disconnect)(struct input_handle *handle);
 	void (*start)(struct input_handle *handle);
 
+	void *suse_kabi_padding;
+
 	bool legacy_minors;
 	int minor;
 	const char *name;
@@ -562,6 +566,9 @@ struct ff_device {
 
 	int max_effects;
 	struct ff_effect *effects;
+
+	void *suse_kabi_padding;
+
 	struct file *effect_owners[];
 };
 
-- 
2.47.1

