From: Thomas Renninger <trenn@suse.de>
Subject: intel iommu: Only check for iommu ioat BIOS bug if iommu is detected
References: bnc#733702
Patch-Mainline: submitted, should also show up in next stable kernel

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 drivers/pci/intel-iommu.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-3.0-SLE11-SP2/drivers/pci/intel-iommu.c
===================================================================
--- linux-3.0-SLE11-SP2.orig/drivers/pci/intel-iommu.c
+++ linux-3.0-SLE11-SP2/drivers/pci/intel-iommu.c
@@ -3189,6 +3189,9 @@ static void quirk_ioat_snb_local_iommu(s
 	u32 vtbar;
 	int rc;
 
+	if (!iommu_detected)
+		return;
+
 	/* We know that this device on this chipset has its own IOMMU.
 	 * If we find it under a different IOMMU, then the BIOS is lying
 	 * to us. Hope that the IOMMU for this device is actually
