From: Andreas Gruenbacher <agruen@suse.de>
Subject: Add missing license information

Index: linux-2.6.16/drivers/dma/testclient.c
===================================================================
--- linux-2.6.16.orig/drivers/dma/testclient.c
+++ linux-2.6.16/drivers/dma/testclient.c
@@ -132,3 +132,4 @@ testclient_exit_module(void)
 
 module_exit(testclient_exit_module);
 
+MODULE_LICENSE("GPL");
