From: Christoph Hellwig <hch@lst.de>
Date: Tue, 10 Aug 2021 18:33:04 -0700
Subject: [PATCH] iomap: mark the iomap argument to iomap_sector const
Git-commit: 66b8165ed4b5a2e7ddb7b9bbf3586b7ccdd86a1c
Patch-mainline: v5.15-rc1
References: jsc#PED-1183

Signed-off-by: Christoph Hellwig <hch@lst.de>
Reviewed-by: Darrick J. Wong <djwong@kernel.org>
Signed-off-by: Darrick J. Wong <djwong@kernel.org>
Acked-by: Hannes Reinecke <hare@suse.com>
---
 include/linux/iomap.h | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/include/linux/iomap.h b/include/linux/iomap.h
index 72696a55c137..8030483331d1 100644
--- a/include/linux/iomap.h
+++ b/include/linux/iomap.h
@@ -91,8 +91,7 @@ struct iomap {
 	const struct iomap_page_ops *page_ops;
 };
 
-static inline sector_t
-iomap_sector(struct iomap *iomap, loff_t pos)
+static inline sector_t iomap_sector(const struct iomap *iomap, loff_t pos)
 {
 	return (iomap->addr + pos - iomap->offset) >> SECTOR_SHIFT;
 }
-- 
2.35.3

