From: Vipul Pandya <vipul@chelsio.com>
Date: Mon, 21 May 2012 17:31:13 +0530
Subject: [PATCH 19/25] RDMA/cxgb4: Include vmalloc.h for vmalloc and vfree
Git-commit: e572568fbcc680e5bfce1819bbb4805ca218b5d0
Patch-mainline: v3.5-rc1
References: bnc#784800 FATE#314296

Signed-off-by: Vipul Pandya <vipul@chelsio.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/infiniband/hw/cxgb4/device.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/infiniband/hw/cxgb4/device.c
+++ b/drivers/infiniband/hw/cxgb4/device.c
@@ -32,6 +32,7 @@
 #include <linux/module.h>
 #include <linux/moduleparam.h>
 #include <linux/debugfs.h>
+#include <linux/vmalloc.h>
 
 #include <rdma/ib_verbs.h>
 
