From: Joe Perches <joe@perches.com>
Date: Wed, 17 Aug 2011 06:58:04 -0700
Subject: treewide: Correct spelling of successfully in comments
Patch-mainline: v3.2-rc1
Git-commit: bfb9035c98906aafcd3cf22694fba2550997bf53 (partial)
References: bnc#795301 FATE#313663

Signed-off-by: Joe Perches <joe@perches.com>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ixgbevf/mbx.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
--- a/drivers/net/ixgbevf/mbx.c
+++ b/drivers/net/ixgbevf/mbx.c
@@ -276,7 +276,7 @@ out_no_write:
  *  @msg: The message buffer
  *  @size: Length of buffer
  *
- *  returns 0 if it successfuly read message from buffer
+ *  returns 0 if it successfully read message from buffer
  **/
 static s32 ixgbevf_read_mbx_vf(struct ixgbe_hw *hw, u32 *msg, u16 size)
 {
