From: Emil Tantilov <emil.s.tantilov@intel.com>
Date: Tue, 12 Jul 2011 08:13:41 +0000
Subject: ixgbe: remove ifdef check for non-existent define
Patch-mainline: v3.1-rc1
Git-commit: 34c6ee8181d6ec1ed761ed2fdac4f1b0358e5447
References: bnc#795303 FATE#313662

Signed-off-by: Emil Tantilov <emil.s.tantilov@intel.com>
Tested-by: Phil Schmitt <phillip.j.schmitt@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ixgbe/ixgbe_main.c |    2 --
 1 file changed, 2 deletions(-)
--- a/drivers/net/ixgbe/ixgbe_main.c
+++ b/drivers/net/ixgbe/ixgbe_main.c
@@ -6037,9 +6037,7 @@ static void ixgbe_watchdog_link_is_up(st
 	       (flow_tx ? "TX" : "None"))));
 
 	netif_carrier_on(netdev);
-#ifdef HAVE_IPLINK_VF_CONFIG
 	ixgbe_check_vf_rate_limit(adapter);
-#endif /* HAVE_IPLINK_VF_CONFIG */
 }
 
 /**
