From: Frederic Weisbecker <fweisbecker@suse.de>
Date: Mon, 26 Feb 2024 08:43:38 +0100
Subject: [PATCH] kABI padding for context tracking
Patch-mainline: Never, kABI padding
References: bsc#1220369

Add kABI padding to important BPF types

Signed-off-by: Frederic Weisbecker <fweisbecker@suse.de>
---
--- a/include/linux/context_tracking_state.h
+++ b/include/linux/context_tracking_state.h
@@ -19,10 +19,11 @@ enum ctx_state {
 };
 
 /* Even value for idle, else odd. */
-#define RCU_DYNTICKS_IDX CONTEXT_MAX
+/* Suse/kABI placeholder: leave 5 bits for eg: IPI deferral */
+#define RCU_DYNTICKS_IDX (CONTEXT_MAX << 5)
 
 #define CT_STATE_MASK (CONTEXT_MAX - 1)
-#define CT_DYNTICKS_MASK (~CT_STATE_MASK)
+#define CT_DYNTICKS_MASK (~(RCU_DYNTICKS_IDX - 1))
 
 struct context_tracking {
 #ifdef CONFIG_CONTEXT_TRACKING_USER
