From: Joerg Roedel <jroedel@suse.de>
Date: Mon, 21 Feb 2022 12:42:16 +0100
Subject: [PATCH] kABI: Fix kABI for AMD IOMMU driver
References: git-fixes
Patch-mainline: Never, kABI fix

Signed-off-by: Joerg Roedel <jroedel@suse.de>
---
 drivers/iommu/amd/amd_iommu_types.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/iommu/amd/amd_iommu_types.h b/drivers/iommu/amd/amd_iommu_types.h
index 473b769403a2..b5d1a14a8c4f 100644
--- a/drivers/iommu/amd/amd_iommu_types.h
+++ b/drivers/iommu/amd/amd_iommu_types.h
@@ -605,6 +605,8 @@ struct amd_iommu {
 	/* DebugFS Info */
 	struct dentry *debugfs;
 #endif
+	/* IRQ notifier for IntCapXT interrupt */
+	struct irq_affinity_notify intcapxt_notify;
 };
 
 static inline struct amd_iommu *dev_to_amd_iommu(struct device *dev)
-- 
2.25.0

