diff -purNX kernel_exclude.txt linux-2.6.5/arch/ppc64/kdb/kdba_bp.c linux-2.6.5/arch/ppc64/kdb/kdba_bp.c
--- linux-2.6.5/arch/ppc64/kdb/kdba_bp.c	2004-04-26 09:33:09.000000000 +0000
+++ linux-2.6.5/arch/ppc64/kdb/kdba_bp.c	2004-04-26 11:12:51.000000000 +0000
@@ -169,9 +169,9 @@ kdba_db_trap(kdb_eframe_t ef, int error_
 				KDB_STATE_CLEAR(DOING_SSB);
 				KDB_STATE_CLEAR(DOING_SS);
 			} else {
-				char *argv[] = {"ssb", NULL};
+				const char *argv[] = {"ssb", NULL};
 				rv = KDB_DB_SSB; /* Indicate ssb - dismiss immediately */
-				kdb_ss(0, (char **)argv, NULL, ef);
+				kdb_ss(0, argv, NULL, ef);
 			}
 		} else {
 			/*
