From: Thomas Renninger <trenn@suse.de>
Subject: Fix ia64 - Export kdb_usb_kbds
Patch-Mainline: no
References: none

---
 arch/ia64/kdb/kdba_io.c |    1 +
 1 file changed, 1 insertion(+)

Index: linux-2.6.32-master/arch/ia64/kdb/kdba_io.c
===================================================================
--- linux-2.6.32-master.orig/arch/ia64/kdb/kdba_io.c
+++ linux-2.6.32-master/arch/ia64/kdb/kdba_io.c
@@ -43,6 +43,7 @@
 /* support up to 8 USB keyboards (probably excessive, but...) */
 #define KDB_USB_NUM_KEYBOARDS	8
 struct kdb_usb_kbd_info kdb_usb_kbds[KDB_USB_NUM_KEYBOARDS];
+EXPORT_SYMBOL(kdb_usb_kbds);
 
 extern int kdb_no_usb;
 
