From ff7124b91f52412696e66683b37a5724c303cc11 Mon Sep 17 00:00:00 2001
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 18 Nov 2021 14:01:34 +0100
Subject: [PATCH] Linux 5.14.20
Git-commit: ff7124b91f52412696e66683b37a5724c303cc11
Patch-mainline: v5.14.20
References: stable-5.14.20

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index f4773aee95c4..0e14a3a30d07 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 14
-SUBLEVEL = 19
+SUBLEVEL = 20
 EXTRAVERSION =
 NAME = Opossums on Parade
 
-- 
2.26.2

