From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 17 Dec 2019 20:08:41 +0100
Subject: Linux 5.3.17
Git-commit: 5e2cfee1b25c94aa357fca06f30d3be073c36d16
Patch-mainline: 5.3.17
References: bnc#1151927 5.3.17

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 16
+SUBLEVEL = 17
 EXTRAVERSION =
 NAME = Bobtail Squid
 
