From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 6 Nov 2019 13:09:35 +0100
Subject: Linux 5.3.9
Git-commit: fd272dcd73353e737928d50497ec113fa1d347f2
Patch-mainline: 5.3.9
References: bnc#1151927 5.3.9

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 445f9488d8ba..ad5f5230bbbe 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 8
+SUBLEVEL = 9
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.24.0

