From bccf6d7352fa518d9c2f4d4241ccfeb3a78f7c43 Mon Sep 17 00:00:00 2001
From: Stefan Wahren <stefan.wahren@i2se.com>
Date: Thu, 4 Apr 2019 19:59:49 +0200
Subject: [PATCH] Linux v5.0-rc7: bcm2835 MMC issues

References: bsc#1070872
Patch-mainline: no, under discussion http://lists.infradead.org/pipermail/linux-rpi-kernel/2019-March/008627.html
Acked-by: Michal Suchanek <msuchanek@suse.de>
---
 drivers/dma/bcm2835-dma.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/dma/bcm2835-dma.c b/drivers/dma/bcm2835-dma.c
index 8675ee0648c8..3b334dbb1e0d 100644
--- a/drivers/dma/bcm2835-dma.c
+++ b/drivers/dma/bcm2835-dma.c
@@ -649,7 +649,7 @@ static struct dma_async_tx_descriptor *bcm2835_dma_prep_slave_sg(
 	d = bcm2835_dma_create_cb_chain(chan, direction, false,
 					info, extra,
 					frames, src, dst, 0, 0,
-					GFP_KERNEL);
+					GFP_NOWAIT);
 	if (!d)
 		return NULL;
 
-- 
2.20.1

