From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 29 Nov 2019 10:08:31 +0100
Subject: Linux 5.3.14
Git-commit: b8e167066e85c9e1e9c5d27b82a858d96e6ba22c
Patch-mainline: 5.3.14
References: bnc#1151927 5.3.14

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 13
+SUBLEVEL = 14
 EXTRAVERSION =
 NAME = Bobtail Squid
 
