From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 11 Oct 2019 18:36:59 +0200
Subject: Linux 5.3.6
Git-commit: a2fc8ee6676067f27d2f5c6e4d512adff3d9938c
Patch-mainline: 5.3.6
References: bnc#1151927 5.3.6

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 5
+SUBLEVEL = 6
 EXTRAVERSION =
 NAME = Bobtail Squid
 
