From: someone at intel <someone@intel.com>
Subject: speed up kms even more


Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/gpu/drm/i915/intel_lvds.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/gpu/drm/i915/intel_lvds.c
+++ b/drivers/gpu/drm/i915/intel_lvds.c
@@ -551,7 +551,7 @@ static void intel_lvds_prepare(struct dr
 	dev_priv->backlight_duty_cycle = (dev_priv->saveBLC_PWM_CTL &
 				       BACKLIGHT_DUTY_CYCLE_MASK);
 
-	intel_lvds_set_power(dev, false);
+//	intel_lvds_set_power(dev, false);
 }
 
 static void intel_lvds_commit( struct drm_encoder *encoder)
@@ -563,7 +563,7 @@ static void intel_lvds_commit( struct dr
 		dev_priv->backlight_duty_cycle =
 			intel_lvds_get_max_backlight(dev);
 
-	intel_lvds_set_power(dev, true);
+//	intel_lvds_set_power(dev, true);
 }
 
 static void intel_lvds_mode_set(struct drm_encoder *encoder,
