From: Arjan van de Ven <arjan@infradead.org>
Subject: enable async_enabled by default
Patch-mainline: not yet


Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 kernel/async.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/kernel/async.c
+++ b/kernel/async.c
@@ -68,7 +68,7 @@ static LIST_HEAD(async_pending);
 static LIST_HEAD(async_running);
 static DEFINE_SPINLOCK(async_lock);
 
-static int async_enabled = 0;
+static int async_enabled = 1;
 
 struct async_entry {
 	struct list_head list;
