From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 10 Nov 2019 11:35:00 +0100
Subject: Linux 5.3.10
Git-commit: b260a0862e3a9fccdac23ec3b783911b098c1c74
Patch-mainline: 5.3.10
References: bnc#1151927 5.3.10

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ad5f5230bbbe..e2a8b4534da5 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 9
+SUBLEVEL = 10
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.24.0

