From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 29 Oct 2019 09:22:48 +0100
Subject: Linux 5.3.8
Git-commit: db0655e705be645ad673b0a70160921e088517c0
Patch-mainline: 5.3.8
References: bnc#1151927 5.3.8

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 7
+SUBLEVEL = 8
 EXTRAVERSION =
 NAME = Bobtail Squid
 
