From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sat, 5 Oct 2019 14:48:42 +0200
Subject: Linux 5.3.3
Git-commit: 5a4dd7cf7b980093d507d55571e342ba702cb336
Patch-mainline: 5.3.3
References: bnc#1151927 5.3.3

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 13fa3a409ddd..a5f4e184b552 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 2
+SUBLEVEL = 3
 EXTRAVERSION =
 NAME = Bobtail Squid
 
-- 
2.23.0

