From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 13 Dec 2019 08:49:56 +0100
Subject: Linux 5.3.16
Git-commit: 128f430ae9acb403059d02d9bbcbd9dcf52968a0
Patch-mainline: 5.3.16
References: bnc#1151927 5.3.16

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 15
+SUBLEVEL = 16
 EXTRAVERSION =
 NAME = Bobtail Squid
 
