From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 4 Dec 2019 22:34:32 +0100
Subject: Linux 5.3.15
Git-commit: 8539dfa4fcbcf58c3c2f92ac57b964add884d12b
Patch-mainline: 5.3.15
References: bnc#1151927 5.3.15

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 14
+SUBLEVEL = 15
 EXTRAVERSION =
 NAME = Bobtail Squid
 
