From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Wed, 20 Nov 2019 16:49:17 +0100
Subject: Linux 5.3.12
Git-commit: 807d174bcb26ffc9eeb944d39591a15059aa7cbc
Patch-mainline: 5.3.12
References: bnc#1151927 5.3.12

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 11
+SUBLEVEL = 12
 EXTRAVERSION =
 NAME = Bobtail Squid
 
