From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Sun, 24 Nov 2019 08:17:01 +0100
Subject: Linux 5.3.13
Git-commit: 42adce4180734a883f2d9b6cec24446d49c5c8eb
Patch-mainline: 5.3.13
References: bnc#1151927 5.3.13

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 3
-SUBLEVEL = 12
+SUBLEVEL = 13
 EXTRAVERSION =
 NAME = Bobtail Squid
 
