bug 41509

diff -Naurp rc1-before/drivers/dump/dump_fmt.c rc1-after/drivers/dump/dump_fmt.c
--- rc1-before/drivers/dump/dump_fmt.c	2004-05-26 01:31:39.000000000 +0530
+++ rc1-after/drivers/dump/dump_fmt.c	2004-06-02 19:04:47.000000000 +0530
@@ -93,7 +93,7 @@ struct __dump_header_asm dump_header_asm
 static int lcrash_init_dump_header(const char *panic_str)
 {
 	struct timeval dh_time;
-	unsigned long temp_dha_stack[DUMP_MAX_NUM_CPUS];
+	unsigned long temp_dha_stack[NR_CPUS];
 	u64 temp_memsz = dump_header.dh_memory_size;
 
 	/* make sure the dump header isn't TOO big */
@@ -107,12 +107,12 @@ static int lcrash_init_dump_header(const
 	/* initialize the dump headers to zero */
 	/* save dha_stack pointer because it may contains pointer for stack! */
 	memcpy(&(temp_dha_stack[0]), &(dump_header_asm.dha_stack[0]),
-		DUMP_MAX_NUM_CPUS * sizeof(unsigned long));
+		NR_CPUS * sizeof(unsigned long));
 	memset(&dump_header, 0, sizeof(dump_header));
 	memset(&dump_header_asm, 0, sizeof(dump_header_asm));
 	dump_header.dh_memory_size = temp_memsz;
 	memcpy(&(dump_header_asm.dha_stack[0]), &(temp_dha_stack[0]),
-		DUMP_MAX_NUM_CPUS * sizeof(unsigned long));
+		NR_CPUS * sizeof(unsigned long));
 
 	/* configure dump header values */
 	dump_header.dh_magic_number = DUMP_MAGIC_NUMBER;
diff -Naurp rc1-before/include/asm-i386/dump.h rc1-after/include/asm-i386/dump.h
--- rc1-before/include/asm-i386/dump.h	2004-05-26 01:31:39.000000000 +0530
+++ rc1-after/include/asm-i386/dump.h	2004-06-02 19:00:22.000000000 +0530
@@ -22,9 +22,6 @@
 #define DUMP_ASM_MAGIC_NUMBER	0xdeaddeadULL	/* magic number            */
 #define DUMP_ASM_VERSION_NUMBER	0x3	/* version number          */
 
-/* max number of cpus */
-#define DUMP_MAX_NUM_CPUS 32
-
 /*
  * Structure: __dump_header_asm
  *  Function: This is the header for architecture-specific stuff.  It
@@ -52,10 +49,10 @@ struct __dump_header_asm {
 	/* smp specific */
 	u32		dha_smp_num_cpus;
 	u32		dha_dumping_cpu;
-	struct pt_regs	dha_smp_regs[DUMP_MAX_NUM_CPUS];
-	u32		dha_smp_current_task[DUMP_MAX_NUM_CPUS];
-	u32		dha_stack[DUMP_MAX_NUM_CPUS];
-	u32		dha_stack_ptr[DUMP_MAX_NUM_CPUS];
+	struct pt_regs	dha_smp_regs[NR_CPUS];
+	u32		dha_smp_current_task[NR_CPUS];
+	u32		dha_stack[NR_CPUS];
+	u32		dha_stack_ptr[NR_CPUS];
 } __attribute__((packed));
 
 #ifdef __KERNEL__
diff -Naurp rc1-before/include/asm-ia64/dump.h rc1-after/include/asm-ia64/dump.h
--- rc1-before/include/asm-ia64/dump.h	2004-05-26 01:31:39.000000000 +0530
+++ rc1-after/include/asm-ia64/dump.h	2004-06-02 19:02:25.000000000 +0530
@@ -16,9 +16,6 @@
 #define DUMP_ASM_MAGIC_NUMBER     0xdeaddeadULL  /* magic number            */
 #define DUMP_ASM_VERSION_NUMBER   0x4            /* version number          */
 
-/* max number of cpus */
-#define DUMP_MAX_NUM_CPUS 32
-
 #ifdef __KERNEL__
 #include <linux/efi.h>
 #include <asm/pal.h>
@@ -110,10 +107,10 @@ typedef struct __dump_header_asm {
 	/* smp specific */
 	uint32_t	     dha_smp_num_cpus;
 	uint32_t	     dha_dumping_cpu;	
-	struct pt_regs	     dha_smp_regs[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_smp_current_task[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_stack[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_stack_ptr[DUMP_MAX_NUM_CPUS];
+	struct pt_regs	     dha_smp_regs[NR_CPUS];
+	uint64_t	     dha_smp_current_task[NR_CPUS];
+	uint64_t	     dha_stack[NR_CPUS];
+	uint64_t	     dha_stack_ptr[NR_CPUS];
 
 } __attribute__((packed)) dump_header_asm_t;
 
diff -Naurp rc1-before/include/asm-ppc64/dump.h rc1-after/include/asm-ppc64/dump.h
--- rc1-before/include/asm-ppc64/dump.h	2004-05-26 01:31:38.000000000 +0530
+++ rc1-after/include/asm-ppc64/dump.h	2004-06-02 19:03:39.000000000 +0530
@@ -21,9 +21,6 @@
 #define DUMP_ASM_MAGIC_NUMBER     0xdeaddeadULL  /* magic number            */
 #define DUMP_ASM_VERSION_NUMBER   0x5            /* version number          */
 
-/* max number of cpus */
-#define DUMP_MAX_NUM_CPUS 32
-
 /*
  * Structure: __dump_header_asm
  *  Function: This is the header for architecture-specific stuff.  It
@@ -46,10 +43,10 @@ struct __dump_header_asm {
 	/* smp specific */
 	uint32_t	     dha_smp_num_cpus;
 	int		     dha_dumping_cpu;	
-	struct pt_regs	     dha_smp_regs[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_smp_current_task[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_stack[DUMP_MAX_NUM_CPUS];
-	uint64_t     	     dha_stack_ptr[DUMP_MAX_NUM_CPUS];
+	struct pt_regs	     dha_smp_regs[NR_CPUS];
+	uint64_t	     dha_smp_current_task[NR_CPUS];
+	uint64_t	     dha_stack[NR_CPUS];
+	uint64_t     	     dha_stack_ptr[NR_CPUS];
 } __attribute__((packed));
 
 #ifdef __KERNEL__
diff -Naurp rc1-before/include/asm-x86_64/dump.h rc1-after/include/asm-x86_64/dump.h
--- rc1-before/include/asm-x86_64/dump.h	2004-05-26 01:31:38.000000000 +0530
+++ rc1-after/include/asm-x86_64/dump.h	2004-06-02 19:04:15.000000000 +0530
@@ -16,9 +16,6 @@
 #include <asm/ptrace.h>                          /* for pt_regs             */
 #include <linux/threads.h>
 
-/* Max number of cpus */
-#define DUMP_MAX_NUM_CPUS	32
-
 /* definitions */
 #define DUMP_ASM_MAGIC_NUMBER     0xdeaddeadULL  /* magic number            */
 #define DUMP_ASM_VERSION_NUMBER   0x2            /* version number          */
@@ -46,10 +43,10 @@ struct __dump_header_asm {
 	/* smp specific */
 	uint32_t	     dha_smp_num_cpus;
 	int		     dha_dumping_cpu;	
-	struct pt_regs	     dha_smp_regs[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_smp_current_task[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_stack[DUMP_MAX_NUM_CPUS];
-	uint64_t	     dha_stack_ptr[DUMP_MAX_NUM_CPUS];
+	struct pt_regs	     dha_smp_regs[NR_CPUS];
+	uint64_t	     dha_smp_current_task[NR_CPUS];
+	uint64_t	     dha_stack[NR_CPUS];
+	uint64_t	     dha_stack_ptr[NR_CPUS];
 } __attribute__((packed));
 
 #ifdef __KERNEL__
