--- linux-2.6.3/init/Makefile.orig	2004-02-27 14:31:36.000000000 +0100
+++ linux-2.6.3/init/Makefile	2004-02-27 14:32:24.000000000 +0100
@@ -9,7 +9,7 @@
 mounts-$(CONFIG_BLK_DEV_INITRD)	+= do_mounts_initrd.o
 mounts-$(CONFIG_BLK_DEV_MD)	+= do_mounts_md.o
 
-extra-$(subst m,y,$(CONFIG_CRASH_DUMP))	+= kerntypes.o
+extra-$(CONFIG_KERNTYPES)	+= kerntypes.o
 CFLAGS_kerntypes.o		:= -gstabs
 
 # files to be removed upon make clean
--- linux-2.6.3/arch/i386/Kconfig.orig	2004-02-27 14:32:44.000000000 +0100
+++ linux-2.6.3/arch/i386/Kconfig	2004-02-27 14:36:58.000000000 +0100
@@ -1209,6 +1209,11 @@
 	  or other error occurs. Dumps can also be forced with the SysRq+d
 	  key if MAGIC_SYSRQ is enabled.
 
+config KERNTYPES
+	bool
+	depends on CRASH_DUMP
+	default y
+
 config CRASH_DUMP_BLOCKDEV
 	tristate "Crash dump block device driver"
 	depends on CRASH_DUMP
--- linux-2.6.3/arch/s390/Kconfig.orig	2004-02-27 14:39:18.000000000 +0100
+++ linux-2.6.3/arch/s390/Kconfig	2004-02-27 14:43:21.000000000 +0100
@@ -361,14 +361,13 @@
 
 menu "Kernel hacking"
 
-config CRASH_DUMP
-	tristate "Crash dump support (EXPERIMENTAL)"
-	depends on EXPERIMENTAL
-	default n
+config KERNTYPES
+	bool "Kerntypes debugging information"
+	default y
 	---help---
-	  Say Y here to enable saving an image of system memory when a panic
-	  or other error occurs. Dumps can also be forced with the SysRq+d
-	  key if MAGIC_SYSRQ is enabled.
+	  Say Y here to save additional kernel debugging information in the
+	  file init/kerntypes.o. This information is used by crash analysis
+	  tools such as lcrash to assign structures to kernel addresses.
 
 config DEBUG_KERNEL
 	bool "Kernel debugging"
