From: hari@in.ibm.com

Fixing netdump for lkcd (#39396)

diff -Naurp before/drivers/dump/dump_netdev.c after/drivers/dump/dump_netdev.c
--- before/drivers/dump/dump_netdev.c	2004-04-15 15:51:32.000000000 +0530
+++ after/drivers/dump/dump_netdev.c	2004-04-21 12:30:58.000000000 +0530
@@ -410,6 +410,7 @@ dump_net_release(struct dump_dev *net_de
 static int
 dump_net_silence(struct dump_dev *net_dev)
 {
+	netpoll_set_trap(1);
 	local_irq_save(flags_global);
         startup_handshake = 1;
 	net_dev->curr_offset = 0;
@@ -440,6 +441,7 @@ dump_net_resume(struct dump_dev *net_dev
 	}
 	printk("NETDUMP END!\n");
 	local_irq_restore(flags_global);
+	netpoll_set_trap(0);
 	startup_handshake = 0;
 	return 0;
 }
