diff -Naurp linux-2.6.5-7.orig/drivers/dump/dump_blockdev.c linux-2.6.5-7/drivers/dump/dump_blockdev.c
--- linux-2.6.5-7.orig/drivers/dump/dump_blockdev.c	2004-04-22 11:30:33.427925992 -0700
+++ linux-2.6.5-7/drivers/dump/dump_blockdev.c	2004-04-22 11:30:01.524983600 -0700
@@ -99,6 +99,7 @@ dump_block_map(struct dump_blockdev *dev
 	if (dump_block_map_valid(dev, page, len)) {
 		/* already mapped and usable rightaway */
 		bio->bi_size = len; /* reset size to the whole bio */
+		bio->bi_vcnt = (len + PAGE_SIZE - 1) / PAGE_SIZE; /* Set the proper vector cnt */
 	} else {
 		/* need to map the bio */
 		bio->bi_size = 0;
