diff -urNpa linux-2.6.5/arch/x86_64/kernel/traps.c linux-2.6.5-lkcd/arch/x86_64/kernel/traps.c
--- linux-2.6.5/arch/x86_64/kernel/traps.c	2004-05-11 11:46:19.000000000 -0700
+++ linux-2.6.5-lkcd/arch/x86_64/kernel/traps.c	2004-05-18 04:13:11.836583392 -0700
@@ -552,21 +552,6 @@ static void mem_parity_error(unsigned ch
 	outb(reason, 0x61);
 }
 
-
-#ifdef CONFIG_SMP
-int (*dump_ipi_function_ptr)(struct pt_regs *) = NULL;
-static int dump_ipi(struct pt_regs *regs)
-{
-	if (!(dump_ipi_function_ptr && dump_ipi_function_ptr(regs))) {
-		return 0;
-	}
-	ack_APIC_irq();
-	return 1;
-}
-#else
-#define dump_ipi(regs) 0
-#endif
-
 static void io_check_error(unsigned char reason, struct pt_regs * regs)
 {
 	printk("NMI: IOCK error (debug interrupt?)\n");
@@ -590,8 +575,10 @@ asmlinkage void default_do_nmi(struct pt
 {
 	unsigned char reason = inb(0x61);
 
-	if (dump_ipi(regs))
+	if (dump_oncpu) {
+		ack_APIC_irq();	
 		return;
+	}
 
 	if (!(reason & 0xc0)) {
 		if (notify_die(DIE_NMI_IPI, "nmi_ipi", regs, reason, 0, SIGINT) == NOTIFY_BAD)
diff -urNpa linux-2.6.5/arch/x86_64/kernel/x8664_ksyms.c linux-2.6.5-lkcd/arch/x86_64/kernel/x8664_ksyms.c
--- linux-2.6.5/arch/x86_64/kernel/x8664_ksyms.c	2004-05-11 11:46:19.000000000 -0700
+++ linux-2.6.5-lkcd/arch/x86_64/kernel/x8664_ksyms.c	2004-05-18 04:10:16.774196920 -0700
@@ -214,7 +214,7 @@ EXPORT_SYMBOL(clear_page);
 #ifdef CONFIG_CRASH_DUMP_MODULE
 #ifdef CONFIG_SMP
 extern irq_desc_t irq_desc[NR_IRQS];
-extern unsigned long irq_affinity[NR_IRQS];
+extern cpumask_t irq_affinity[NR_IRQS];
 extern void stop_this_cpu(void *);
 EXPORT_SYMBOL(irq_desc);
 EXPORT_SYMBOL(irq_affinity);
diff -urNpa linux-2.6.5/drivers/dump/dump_x8664.c linux-2.6.5-lkcd/drivers/dump/dump_x8664.c
--- linux-2.6.5/drivers/dump/dump_x8664.c	2004-05-11 11:46:18.000000000 -0700
+++ linux-2.6.5-lkcd/drivers/dump/dump_x8664.c	2004-05-18 04:20:23.263996456 -0700
@@ -75,7 +75,7 @@ __dump_save_regs(struct pt_regs* dest_re
 }
 
 #ifdef CONFIG_SMP
-extern unsigned long irq_affinity[];
+extern cpumask_t irq_affinity[];
 extern irq_desc_t irq_desc[];
 extern void dump_send_ipi(void);
 static int dump_expect_ipi[NR_CPUS];
@@ -134,7 +134,7 @@ __dump_save_other_cpus(void) 
 
 		for (i = 0; i < NR_CPUS; i++)
 			dump_expect_ipi[i] = (i != cpu && cpu_online(i));
-		
+
 		set_nmi_callback(dump_nmi_callback);
 		wmb();
 
@@ -160,13 +160,15 @@ static void
 set_irq_affinity(void)
 {
 	int i;
-	int cpu = smp_processor_id();
+	/*int cpu = smp_processor_id();*/
+	cpumask_t cpu = CPU_MASK_NONE;
 
+	cpu_set(smp_processor_id(), cpu);
 	memcpy(saved_affinity, irq_affinity, NR_IRQS * sizeof(unsigned long));
 	for (i = 0; i < NR_IRQS; i++) {
 		if (irq_desc[i].handler == NULL)
 			continue;
-		irq_affinity[i] = 1UL << cpu;
+		irq_affinity[i] = cpu;
 		if (irq_desc[i].handler->set_affinity != NULL)
 			irq_desc[i].handler->set_affinity(i, irq_affinity[i]);
 	}
diff -urNpa linux-2.6.5/include/asm-x86_64/dump.h linux-2.6.5-lkcd/include/asm-x86_64/dump.h
--- linux-2.6.5/include/asm-x86_64/dump.h	2004-05-11 11:46:18.000000000 -0700
+++ linux-2.6.5-lkcd/include/asm-x86_64/dump.h	2004-05-18 04:19:09.418222720 -0700
@@ -78,7 +78,6 @@ extern struct __dump_header_asm dump_hea
 
 #ifdef CONFIG_SMP
 extern unsigned long irq_affinity[];
-extern int (*dump_ipi_function_ptr)(struct pt_regs *);
 extern void dump_send_ipi(void);
 #else
 #define dump_send_ipi() do { } while(0)
