From: SGI
Subject: enable lkcd polling for mpt fusion
Patch-mainline: never
References: 150096

Acked-by: mason@suse.com
Signed-off-by: 

Index: linux/drivers/message/fusion/mptsas.c
===================================================================
--- linux.orig/drivers/message/fusion/mptsas.c	2006-02-10 09:47:19.519377843 -0800
+++ linux/drivers/message/fusion/mptsas.c	2006-02-10 09:56:13.633060834 -0800
@@ -397,6 +397,7 @@
 	.max_sectors			= 8192,
 	.cmd_per_lun			= 7,
 	.use_clustering			= ENABLE_CLUSTERING,
+	.dump_poll			= mptscsih_poll,
 };
 
 static inline MPT_ADAPTER *phy_to_ioc(struct sas_phy *phy)
