diff -purN linux-2.6.5/arch/ppc64/Kconfig linux-2.6.5.kerntypes/arch/ppc64/Kconfig
--- linux-2.6.5/arch/ppc64/Kconfig	2004-05-06 11:05:39.000000000 +0200
+++ linux-2.6.5.kerntypes/arch/ppc64/Kconfig	2004-05-06 11:07:29.000000000 +0200
@@ -407,6 +407,11 @@ config CRASH_DUMP_COMPRESS_GZIP
 	help
 	  Say Y to allow saving dumps with Gnu Zip compression.
 
+config KERNTYPES
+	bool
+	depends on CRASH_DUMP
+	default y
+
 config DEBUG_KERNEL
 	bool "Kernel debugging"
 	help
