From: Andrew Morton <akpm@linux-foundation.org>
Date: Mon, 31 Oct 2011 17:13:08 -0700
Subject: llist-return-whether-list-is-empty-before-adding-in-llist_add-fix
References: bnc#913030,FATE#317455
Patch-Mainline: v3.3
Git-commit: fc23af34b00ef444eec088f744983b9ca6c7f5d1

clarify comment

Cc: Huang Ying <ying.huang@intel.com>
Cc: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Cc: Peter Zijlstra <peterz@infradead.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 include/linux/llist.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/llist.h b/include/linux/llist.h
index dd0651a..8295946 100644
--- a/include/linux/llist.h
+++ b/include/linux/llist.h
@@ -143,7 +143,7 @@ static inline bool llist_empty(const struct llist_head *head)
  * @new:	new entry to be added
  * @head:	the head for your lock-less list
  *
- * Return whether list is empty before adding.
+ * Returns true if the list was empty prior to adding this entry.
  */
 static inline bool llist_add(struct llist_node *new, struct llist_head *head)
 {
-- 
1.8.5.2

