From: James Smart <james.smart@emulex.com>
Date: Tue, 14 Aug 2012 14:26:20 -0400
Subject: [SCSI] lpfc 8.3.34: Adjust IO Channels to 1 when INTx
Git-commit: 1c13bf664894b34ace55fce49cc49598267787b7
References: FATE#313818
Patch-Mainline: v3.7

Signed-off-by: James Smart <james.smart@emulex.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/scsi/lpfc/lpfc_init.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/scsi/lpfc/lpfc_init.c b/drivers/scsi/lpfc/lpfc_init.c
index a78934b..75217af 100644
--- a/drivers/scsi/lpfc/lpfc_init.c
+++ b/drivers/scsi/lpfc/lpfc_init.c
@@ -9513,7 +9513,7 @@ lpfc_pci_probe_one_s4(struct pci_dev *pdev, const struct pci_device_id *pid)
 		}
 		/* Default to single EQ for non-MSI-X */
 		if (phba->intr_type != MSIX)
-			adjusted_fcp_io_channel = 0;
+			adjusted_fcp_io_channel = 1;
 		else if (phba->sli4_hba.msix_vec_nr <
 					phba->cfg_fcp_io_channel)
 			adjusted_fcp_io_channel = phba->sli4_hba.msix_vec_nr;
-- 
1.7.4.2

