From: James Smart <james.smart@emulex.com>
Date: Tue, 14 Aug 2012 14:25:59 -0400
Subject: [SCSI] lpfc 8.3.34: Correct typecasts for snprintf messages
Git-commit: a51a17831b28001645f4442c21ba5c85329d36d2
References: FATE#313818
Patch-Mainline: v3.7

Signed-off-by: James Smart <james.smart@emulex.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/scsi/lpfc/lpfc_debugfs.c |   18 +++++++++++-------
 1 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/drivers/scsi/lpfc/lpfc_debugfs.c b/drivers/scsi/lpfc/lpfc_debugfs.c
index d5574f3..8cba348 100644
--- a/drivers/scsi/lpfc/lpfc_debugfs.c
+++ b/drivers/scsi/lpfc/lpfc_debugfs.c
@@ -2051,7 +2051,7 @@ lpfc_idiag_queinfo_read(struct file *file, char __user *buf, size_t nbytes,
 				"EQ-STAT[max:x%x noE:x%x "
 				"bs:x%x proc:x%llx]\n",
 				qp->q_cnt_1, qp->q_cnt_2,
-				qp->q_cnt_3, qp->q_cnt_4);
+				qp->q_cnt_3, (unsigned long long)qp->q_cnt_4);
 
 			len += snprintf(pbuffer+len,
 				LPFC_QUE_INFO_GET_BUF_SIZE-len,
@@ -2085,7 +2085,7 @@ proc_cq:
 				"xabt:x%x wq:x%llx]\n",
 				qp->assoc_qid,
 				qp->q_cnt_1, qp->q_cnt_2,
-				qp->q_cnt_3, qp->q_cnt_4);
+				qp->q_cnt_3, (unsigned long long)qp->q_cnt_4);
 			len += snprintf(pbuffer+len,
 				LPFC_QUE_INFO_GET_BUF_SIZE-len,
 				"\tCQID[%02d], "
@@ -2115,7 +2115,7 @@ proc_cq:
 				"AssocCQID[%02d]: "
 				"WQ-STAT[oflow:x%x posted:x%llx]\n",
 				qp->assoc_qid,
-				qp->q_cnt_1, qp->q_cnt_4);
+				qp->q_cnt_1, (unsigned long long)qp->q_cnt_4);
 			len += snprintf(pbuffer+len,
 				LPFC_QUE_INFO_GET_BUF_SIZE-len,
 				"\t\tWQID[%02d], "
@@ -2150,7 +2150,8 @@ proc_cq:
 					"xabt:x%x wq:x%llx]\n",
 					qp->assoc_qid,
 					qp->q_cnt_1, qp->q_cnt_2,
-					qp->q_cnt_3, qp->q_cnt_4);
+					qp->q_cnt_3,
+					(unsigned long long)qp->q_cnt_4);
 				len += snprintf(pbuffer+len,
 					LPFC_QUE_INFO_GET_BUF_SIZE-len,
 					"\tCQID[%02d], "
@@ -2204,7 +2205,8 @@ proc_cq:
 					"xabt:x%x wq:x%llx]\n",
 					qp->assoc_qid,
 					qp->q_cnt_1, qp->q_cnt_2,
-					qp->q_cnt_3, qp->q_cnt_4);
+					qp->q_cnt_3,
+					(unsigned long long)qp->q_cnt_4);
 				len += snprintf(pbuffer+len,
 					LPFC_QUE_INFO_GET_BUF_SIZE-len,
 					"\tCQID [%02d], "
@@ -2235,7 +2237,8 @@ proc_cq:
 					" WQ-STAT[oflow:x%x "
 					"posted:x%llx]\n",
 					qp->assoc_qid,
-					qp->q_cnt_1, qp->q_cnt_4);
+					qp->q_cnt_1,
+					(unsigned long long)qp->q_cnt_4);
 				len += snprintf(pbuffer+len,
 					LPFC_QUE_INFO_GET_BUF_SIZE-len,
 					"\t\tWQID[%02d], "
@@ -2265,7 +2268,8 @@ proc_cq:
 					"trunc:x%x rcv:x%llx]\n",
 					qp->assoc_qid,
 					qp->q_cnt_1, qp->q_cnt_2,
-					qp->q_cnt_3, qp->q_cnt_4);
+					qp->q_cnt_3,
+					(unsigned long long)qp->q_cnt_4);
 				len += snprintf(pbuffer+len,
 					LPFC_QUE_INFO_GET_BUF_SIZE-len,
 					"\t\tHQID[%02d], "
-- 
1.7.4.2

