diff -pruN linux-2.6.16-8/drivers/message/fusion/mptbase.c linux-2.6.16-8-aic94xx/drivers/message/fusion/mptbase.c
--- linux-2.6.16-8/drivers/message/fusion/mptbase.c	2006-04-06 18:49:47.000000000 -0700
+++ linux-2.6.16-8-aic94xx/drivers/message/fusion/mptbase.c	2006-04-06 17:51:30.000000000 -0700
@@ -180,7 +180,6 @@ static void	mpt_sp_ioc_info(MPT_ADAPTER 
 static void	mpt_fc_log_info(MPT_ADAPTER *ioc, u32 log_info);
 static void	mpt_spi_log_info(MPT_ADAPTER *ioc, u32 log_info);
 static void	mpt_sas_log_info(MPT_ADAPTER *ioc, u32 log_info);
-static int	mpt_read_ioc_pg_3(MPT_ADAPTER *ioc);
 
 /* module entry point */
 static int  __init    fusion_init  (void);
@@ -4939,7 +4938,7 @@ done_and_free:
 	return rc;
 }
 
-static int
+int
 mpt_read_ioc_pg_3(MPT_ADAPTER *ioc)
 {
 	IOCPage3_t		*pIoc3;
@@ -6538,6 +6537,7 @@ EXPORT_SYMBOL(mpt_stm_index);
 EXPORT_SYMBOL(mpt_HardResetHandler);
 EXPORT_SYMBOL(mpt_config);
 EXPORT_SYMBOL(mpt_findImVolumes);
+EXPORT_SYMBOL(mpt_read_ioc_pg_3);
 EXPORT_SYMBOL(mpt_alloc_fw_memory);
 EXPORT_SYMBOL(mpt_free_fw_memory);
 EXPORT_SYMBOL(mptbase_sas_persist_operation);
diff -pruN linux-2.6.16-8/drivers/message/fusion/mptbase.h linux-2.6.16-8-aic94xx/drivers/message/fusion/mptbase.h
--- linux-2.6.16-8/drivers/message/fusion/mptbase.h	2006-04-06 18:49:47.000000000 -0700
+++ linux-2.6.16-8-aic94xx/drivers/message/fusion/mptbase.h	2006-04-06 19:57:05.000000000 -0700
@@ -331,6 +331,7 @@ typedef struct _SYSIF_REGS
  *	VirtDevice - FC LUN device or SCSI target device
  */
 typedef struct _VirtTarget {
+	struct scsi_target	 *starget;
 	u8			 tflags;
 	u8			 ioc_id;
 	u8			 target_id;
@@ -361,6 +362,7 @@ typedef struct _VirtDevice {
 #define MPT_TARGET_FLAGS_Q_YES		0x08
 #define MPT_TARGET_FLAGS_VALID_56	0x10
 #define MPT_TARGET_FLAGS_SAF_TE_ISSUED	0x20
+#define MPT_TARGET_FLAGS_RAID_COMPONENT	0x40
 
 /*
  *	/proc/mpt interface
@@ -1037,6 +1039,7 @@ extern int	 mpt_config(MPT_ADAPTER *ioc,
 extern void	 mpt_alloc_fw_memory(MPT_ADAPTER *ioc, int size);
 extern void	 mpt_free_fw_memory(MPT_ADAPTER *ioc);
 extern int	 mpt_findImVolumes(MPT_ADAPTER *ioc);
+extern int	 mpt_read_ioc_pg_3(MPT_ADAPTER *ioc);
 extern int	 mptbase_sas_persist_operation(MPT_ADAPTER *ioc, u8 persist_opcode);
 extern int	 mptbase_GetFcPortPage0(MPT_ADAPTER *ioc, int portnum);
 extern int	 mpt_alt_ioc_wait(MPT_ADAPTER *ioc);
diff -pruN linux-2.6.16-8/drivers/message/fusion/mptscsih.c linux-2.6.16-8-aic94xx/drivers/message/fusion/mptscsih.c
--- linux-2.6.16-8/drivers/message/fusion/mptscsih.c	2006-04-06 18:49:47.000000000 -0700
+++ linux-2.6.16-8-aic94xx/drivers/message/fusion/mptscsih.c	2006-04-06 15:56:20.000000000 -0700
@@ -159,7 +159,7 @@ int		mptscsih_scandv_complete(MPT_ADAPTE
 static int	mptscsih_do_cmd(MPT_SCSI_HOST *hd, INTERNAL_CMD *iocmd);
 static void	mptscsih_synchronize_cache(MPT_SCSI_HOST *hd, VirtDevice *vdevice);
 static void	mptscsih_negotiate_to_asyn_narrow(MPT_SCSI_HOST *hd, VirtDevice *vdevice);
-static int	mptscsih_is_phys_disk(MPT_ADAPTER *ioc, int id);
+int		mptscsih_is_phys_disk(MPT_ADAPTER *ioc, int id);
 
 #ifdef MPTSCSIH_ENABLE_DOMAIN_VALIDATION
 static int	mptscsih_do_raid(MPT_SCSI_HOST *hd, u8 action, INTERNAL_CMD *io);
@@ -1424,7 +1424,7 @@ mptscsih_qcmd(struct scsi_cmnd *SCpnt, v
 
 #ifdef MPTSCSIH_ENABLE_DOMAIN_VALIDATION
 	if (hd->ioc->bus_type == SPI) {
-		int dvStatus = hd->ioc->spi_data.dvStatus[vdev->target_id];
+		int dvStatus = hd->ioc->spi_data.dvStatus[vdev->vtarget->target_id];
 		int issueCmd = 1;
 
 		if (dvStatus || hd->ioc->spi_data.forceDv) {
@@ -3089,7 +3089,7 @@ mptscsih_no_negotiate(MPT_SCSI_HOST *hd,
 	VirtDevice	*vdev;
 
 	if ((vdev = sc->device->hostdata) != NULL)
-		hd->ioc->spi_data.dvStatus[vdev->target_id] |= MPT_SCSICFG_BLK_NEGO;
+		hd->ioc->spi_data.dvStatus[vdev->vtarget->target_id] |= MPT_SCSICFG_BLK_NEGO;
 	return;
 }
 
@@ -4089,24 +4089,6 @@ mptscsih_synchronize_cache(MPT_SCSI_HOST
 		mptscsih_do_cmd(hd, &iocmd);
 }
 
-/* Search IOC page 3 to determine if this is hidden physical disk
- */
-static int
-mptscsih_is_phys_disk(MPT_ADAPTER *ioc, int id)
-{
-	int i;
-
-	if (!ioc->raid_data.isRaid || !ioc->raid_data.pIocPg3)
-		return 0;
-
-	for (i = 0; i < ioc->raid_data.pIocPg3->NumPhysDisks; i++) {
-		if (id == ioc->raid_data.pIocPg3->PhysDisk[i].PhysDiskID)
-			return 1;
-	}
-
-	return 0;
-}
-
 #ifdef MPTSCSIH_ENABLE_DOMAIN_VALIDATION
 /*=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=*/
 /**
Index: linux-2.6.16-20060329183759/include/scsi/scsi_device.h
===================================================================
--- linux-2.6.16-20060329183759.orig/include/scsi/scsi_device.h	2006-03-30 15:32:58.000000000 -0800
+++ linux-2.6.16-20060329183759/include/scsi/scsi_device.h	2006-03-30 15:33:23.000000000 -0800
@@ -281,6 +281,11 @@ extern int scsi_execute_async(struct scs
 			      void (*done)(void *, char *, int, int),
 			      gfp_t gfp);
 
+static inline void scsi_device_reprobe(struct scsi_device *sdev)
+{
+	device_reprobe(&sdev->sdev_gendev);
+}
+
 static inline unsigned int sdev_channel(struct scsi_device *sdev)
 {
 	return sdev->channel;
Index: linux-2.6.16-20060329183759/drivers/base/bus.c
===================================================================
--- linux-2.6.16-20060329183759.orig/drivers/base/bus.c	2006-03-30 15:32:58.000000000 -0800
+++ linux-2.6.16-20060329183759/drivers/base/bus.c	2006-03-30 15:33:23.000000000 -0800
@@ -543,6 +543,28 @@ void bus_rescan_devices(struct bus_type 
 	bus_for_each_dev(bus, NULL, NULL, bus_rescan_devices_helper);
 }
 
+/**
+ * device_reprobe - remove driver for a device and probe for a new driver
+ * @dev: the device to reprobe
+ *
+ * This function detaches the attached driver (if any) for the given
+ * device and restarts the driver probing process.  It is intended
+ * to use if probing criteria changed during a devices lifetime and
+ * driver attachment should change accordingly.
+ */
+void device_reprobe(struct device *dev)
+{
+	if (dev->driver) {
+		if (dev->parent)        /* Needed for USB */
+			down(&dev->parent->sem);
+		device_release_driver(dev);
+		if (dev->parent)
+			up(&dev->parent->sem);
+	}
+
+	bus_rescan_devices_helper(dev, NULL);
+}
+EXPORT_SYMBOL_GPL(device_reprobe);
 
 struct bus_type * get_bus(struct bus_type * bus)
 {
Index: linux-2.6.16-20060329183759/include/linux/device.h
===================================================================
--- linux-2.6.16-20060329183759.orig/include/linux/device.h	2006-03-30 15:32:58.000000000 -0800
+++ linux-2.6.16-20060329183759/include/linux/device.h	2006-03-30 15:33:23.000000000 -0800
@@ -378,6 +378,7 @@ extern void device_bind_driver(struct de
 extern void device_release_driver(struct device * dev);
 extern int  device_attach(struct device * dev);
 extern void driver_attach(struct device_driver * drv);
+extern void device_reprobe(struct device *dev);
 
 
 /*
