bug 40390, ltt causes an uninitialized timer warning with gettimeofday
timestamping

Index: linux-2.6.5/kernel/ltt/ltt-core.c
===================================================================
--- linux-2.6.5.orig/kernel/ltt/ltt-core.c
+++ linux-2.6.5/kernel/ltt/ltt-core.c
@@ -1197,9 +1197,9 @@ int ltt_free_daemon_handle(struct trace_
 
 	if (!active_traces())
 		del_percpu_timers();
 
-	if(!need_heartbeat()) 
+	if(trace->using_tsc && !need_heartbeat()) 
 		del_timer(&heartbeat_timer);
 
 	for (i = 0; i < num_cpus; i++) {
 		if (trace_channel_handle(trace->trace_handle, i) != -1) {
