Index: linux-2.6.5/net/ipv6/Makefile
===================================================================
--- linux-2.6.5.orig/net/ipv6/Makefile
+++ linux-2.6.5/net/ipv6/Makefile
@@ -23,7 +23,6 @@ obj-$(CONFIG_INET6_IPCOMP) += ipcomp6.o
 obj-$(CONFIG_NETFILTER)	+= netfilter/
 
 obj-$(CONFIG_IPV6_TUNNEL)   += ipv6_tunnel.o
-obj-$(CONFIG_IPV6_MOBILITY) += mipglue.o
 obj-$(CONFIG_IPV6_MOBILITY) += mobile_ip6/
 
 obj-y += exthdrs_core.o
