From: Sreekanth Reddy <Sreekanth.Reddy@lsi.com>
Date: Fri, 30 Nov 2012 07:58:10 +0530
Subject: [SCSI] miscdevice: Adding support for MPT3SAS_MINOR(222)
Git-commit: 78ad724ade23c66650dcaba04f4d2307c2884c69
References: FATE#314302
Patch-Mainline: v3.8-rc1

Signed-off-by: Sreekanth Reddy <Sreekanth.Reddy@lsi.com>
Reviewed-by: Nagalakshmi Nandigama <Nagalakshmi.Nandigama@lsi.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 include/linux/miscdevice.h |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/include/linux/miscdevice.h b/include/linux/miscdevice.h
index 18fd130..3a53122 100644
--- a/include/linux/miscdevice.h
+++ b/include/linux/miscdevice.h
@@ -33,6 +33,7 @@
 #define MWAVE_MINOR		219	/* ACP/Mwave Modem */
 #define MPT_MINOR		220
 #define MPT2SAS_MINOR		221
+#define MPT3SAS_MINOR		222
 #define UINPUT_MINOR		223
 #define HPET_MINOR		228
 #define FUSE_MINOR		229
-- 
1.7.4.2

