From: Amir Vadai <amirv@mellanox.com>
Date: Wed, 13 Feb 2013 10:09:48 +0200
Subject: IB/mlx4: remove __devinit from mlx4_init_xrcd_table()
Patch-mainline: No
References: bnc#786036 FATE#314304

Signed-off-by: Amir Vadai <amirv@mellanox.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/mlx4/xrcd.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
--- a/drivers/net/mlx4/xrcd.c
+++ b/drivers/net/mlx4/xrcd.c
@@ -54,7 +54,7 @@ void mlx4_xrcd_free(struct mlx4_dev *dev
 }
 EXPORT_SYMBOL_GPL(mlx4_xrcd_free);
 
-int __devinit mlx4_init_xrcd_table(struct mlx4_dev *dev)
+int mlx4_init_xrcd_table(struct mlx4_dev *dev)
 {
 	struct mlx4_priv *priv = mlx4_priv(dev);
 
