From 158f81bb091f3f93d06e3cc45afba5d9e0df4755 Mon Sep 17 00:00:00 2001
From: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
Date: Thu, 25 Feb 2010 18:30:21 +0200
Subject: [PATCH] mlx4_core: missing device ID
References: bnc#585269
Patch-mainline: Submitted upstream

Signed-off-by: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
Acked-by: John jolly <jjolly@suse.de>
---
 drivers/net/mlx4/main.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/mlx4/main.c
+++ b/drivers/net/mlx4/main.c
@@ -1405,6 +1405,7 @@ static DEFINE_PCI_DEVICE_TABLE(mlx4_pci_
 	{ PCI_VDEVICE(MELLANOX, 0x6764) }, /* MT26468 ConnectX EN 10GigE PCIe gen2*/
 	{ PCI_VDEVICE(MELLANOX, 0x6746) }, /* MT26438 ConnectX EN 40GigE PCIe gen2 5GT/s */
 	{ PCI_VDEVICE(MELLANOX, 0x676e) }, /* MT26478 ConnectX2 40GigE PCIe gen2 */
+	{ PCI_VDEVICE(MELLANOX, 0x6778) }, /* MT26488 ConnectX VPI PCIe 2.0 5GT/s - IB DDR / 10GigE Virt+ */
 	{ PCI_VDEVICE(MELLANOX, 0x1002) }, /* MT25400 Family [ConnectX-2 Virtual Function] */
 	{ PCI_VDEVICE(MELLANOX, 0x1003) }, /* MT27500 Family [ConnectX-3] */
 	{ PCI_VDEVICE(MELLANOX, 0x1004) }, /* MT27500 Family [ConnectX-3 Virtual Function] */
