From: Roland Dreier <roland@purestorage.com>
Date: Mon, 8 Jul 2013 00:13:35 -0700
Subject: [PATCH 005/129] IB/mlx5: Make profile[] static in main.c
Git-commit: ad32b95f8281ac5a292da83ff27616aceee4b770
Patch-mainline: v3.11-rc1
References: bnc#905015 fate#317531

Signed-off-by: Roland Dreier <roland@purestorage.com>
Signed-off-by: Hadar Hen Zion <hadarh@mellanox.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/infiniband/hw/mlx5/main.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/infiniband/hw/mlx5/main.c b/drivers/infiniband/hw/mlx5/main.c
index 809f1b6..249c451 100644
--- a/drivers/infiniband/hw/mlx5/main.c
+++ b/drivers/infiniband/hw/mlx5/main.c
@@ -62,7 +62,7 @@ static char mlx5_version[] =
 	DRIVER_NAME ": Mellanox Connect-IB Infiniband driver v"
 	DRIVER_VERSION " (" DRIVER_RELDATE ")\n";
 
-struct mlx5_profile profile[] = {
+static struct mlx5_profile profile[] = {
 	[0] = {
 		.mask		= 0,
 	},
-- 
1.7.8.2


