From: Sreekanth Reddy <sreekanth.reddy@lsi.com>
Date: Mon, 7 Jan 2013 10:16:41 +0100
Subject: mptfusion: Bump driver vesion to v4.28.00.01suse
Patch-Mainline: n/a
References: bnc#784805, FATE#313895

Signed-off-by: Sreekanth Reddy <sreekanth.reddy@lsi.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/message/fusion/mptbase.h |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/message/fusion/mptbase.h b/drivers/message/fusion/mptbase.h
index b929a25..f5530e8 100644
--- a/drivers/message/fusion/mptbase.h
+++ b/drivers/message/fusion/mptbase.h
@@ -72,13 +72,13 @@
 #define COPYRIGHT	"Copyright (c) 1999-2010 " MODULEAUTHOR
 #endif
 
-#define MPT_LINUX_VERSION_COMMON	"4.28.00.00suse"
-#define MPT_LINUX_PACKAGE_NAME		"@(#)mptlinux-4.28.00.00"
+#define MPT_LINUX_VERSION_COMMON	"4.28.00.01suse"
+#define MPT_LINUX_PACKAGE_NAME		"@(#)mptlinux-4.28.00.01"
 #define WHAT_MAGIC_STRING		"@" "(" "#" ")"
 #define MPT_LINUX_MAJOR_VERSION		4
 #define MPT_LINUX_MINOR_VERSION		28
 #define MPT_LINUX_BUILD_VERSION		00
-#define MPT_LINUX_RELEASE_VERSION	00
+#define MPT_LINUX_RELEASE_VERSION	01
 
 #define show_mptmod_ver(s,ver)  \
 	printk(KERN_INFO "%s %s\n", s, ver);
-- 
1.7.4.2

