From: olh@suse.de
Subject: current git tree broken as of 2.6.13-rc2-git5

 drivers/mtd/Kconfig |    1 +
 1 files changed, 1 insertion(+)

Index: linux-2.6.12/drivers/mtd/Kconfig
===================================================================
--- linux-2.6.12.orig/drivers/mtd/Kconfig
+++ linux-2.6.12/drivers/mtd/Kconfig
@@ -4,6 +4,7 @@ menu "Memory Technology Devices (MTD)"
 
 config MTD
 	tristate "Memory Technology Device (MTD) support"
+	depends on BROKEN
 	help
 	  Memory Technology Devices are flash, RAM and similar chips, often
 	  used for solid state file systems on embedded devices. This option
