From 170d40fe8a20eafd725b0ce3a5685e0c4b758cbf Mon Sep 17 00:00:00 2001
From: Denis Kirjanov <denis.kirjanov@suse.com>
Date: Fri, 28 Jan 2022 19:43:28 +0300
Subject: [PATCH] net: tipc: validate domain record count on input
Patch-mainline: Not yet, will be submitted upstream`
References: bsc#1195254

Signed-off-by: Denis Kirjanov <denis.kirjanov@suse.com>
---
 net/tipc/monitor.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/net/tipc/monitor.c b/net/tipc/monitor.c
index 254ddc2c3914..99fcb42d0af2 100644
--- a/net/tipc/monitor.c
+++ b/net/tipc/monitor.c
@@ -463,6 +463,8 @@ void tipc_mon_rcv(struct net *net, void
 		return;
 	if ((dlen < new_dlen) || ntohs(arrv_dom->len) != new_dlen)
 		return;
+	if (new_member_cnt >= MAX_MON_DOMAIN)
+		return;
 
 	/* Synch generation numbers with peer if link just came up */
 	if (!state->synched) {
-- 
2.16.4

