From: Yue Haibing <yuehaibing@huawei.com>
Date: Sat, 5 Aug 2023 18:48:11 +0800
Subject: net/tls: Remove unused function declarations
Patch-mainline: v6.6-rc1
Git-commit: f6ecb68b38a5c23f20160dd49718d1a9d395a86d
References: jsc#PED-6831

Commit 3c4d7559159b ("tls: kernel TLS support") declared but never implemented
these functions.

Signed-off-by: Yue Haibing <yuehaibing@huawei.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 net/tls/tls.h |    4 ----
 1 file changed, 4 deletions(-)

--- a/net/tls/tls.h
+++ b/net/tls/tls.h
@@ -87,10 +87,6 @@ void tls_ctx_free(struct sock *sk, struc
 void update_sk_prot(struct sock *sk, struct tls_context *ctx);
 
 int wait_on_pending_writer(struct sock *sk, long *timeo);
-int tls_sk_query(struct sock *sk, int optname, char __user *optval,
-		 int __user *optlen);
-int tls_sk_attach(struct sock *sk, int optname, char __user *optval,
-		  unsigned int optlen);
 void tls_err_abort(struct sock *sk, int err);
 
 int tls_set_sw_offload(struct sock *sk, struct tls_context *ctx, int tx);
