From: Sankar P <psankar@novell.com>
Subject: novfs: fix debug message
Patch-mainline: no

Signed-off-by: Sankar P <psankar@novell.com>
Acked-by: Jiri Benc <jbenc@suse.cz>
 
---
 fs/novfs/daemon.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- linux-2.6.32-master.orig/fs/novfs/daemon.c
+++ linux-2.6.32-master/fs/novfs/daemon.c
@@ -1590,7 +1590,7 @@ int novfs_daemon_lib_ioctl(struct inode 
 					break;
 
 				case NWC_SET_CONN_INFO:
-					DbgIocCall("NwGetConnInfo");
+					DbgIocCall("NwSetConnInfo");
 					retCode =
 					    novfs_set_conn_info(&data, dh->session);
 					break;
