From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Mon, 21 Mar 2011 10:28:43 -0400
Subject: NVMe: Version 0.5
Git-commit: 8ef700678f65e2eef1c3a94cdedb79d757608392
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index 90a96ec..d3eeca5 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1630,6 +1630,6 @@ static void __exit nvme_exit(void)
 
 MODULE_AUTHOR("Matthew Wilcox <willy@linux.intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_VERSION("0.4");
+MODULE_VERSION("0.5");
 module_init(nvme_init);
 module_exit(nvme_exit);
-- 
1.7.4.2

