From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Fri, 7 Oct 2011 13:20:37 -0400
Subject: NVMe: Version 0.7
Git-commit: ce38c149576fd0a3360fec3bef4012212d42e736
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index 660aa5d..cfe5932 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1700,6 +1700,6 @@ static void __exit nvme_exit(void)
 
 MODULE_AUTHOR("Matthew Wilcox <willy@linux.intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_VERSION("0.6");
+MODULE_VERSION("0.7");
 module_init(nvme_init);
 module_exit(nvme_exit);
-- 
1.7.4.2

