From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Tue, 22 Feb 2011 14:15:34 -0500
Subject: NVMe: Mark CMD_CTX_CANCELLED as being unlikely
Git-commit: c42705592be2a539f3027b6f3907de8e8f9591a8
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index af45e28..ce919b4 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -518,7 +518,7 @@ static void sync_completion(struct nvme_queue *nvmeq, void *ctx,
 						struct nvme_completion *cqe)
 {
 	struct sync_cmd_info *cmdinfo = ctx;
-	if ((unsigned long)cmdinfo == CMD_CTX_CANCELLED)
+	if (unlikely((unsigned long)cmdinfo == CMD_CTX_CANCELLED))
 		return;
 	if (unlikely((unsigned long)cmdinfo == CMD_CTX_COMPLETED)) {
 		dev_warn(nvmeq->q_dmadev,
-- 
1.7.4.2

