From: Keith Busch <keith.busch@intel.com>
Date: Tue, 25 Jun 2013 15:15:23 -0600
Subject: NVMe: Update nvme_id_power_state with latest spec
References: bnc#913030,FATE#317455
Patch-Mainline: v3.13
Git-commit: 685585c25e8269cc355b59d1cd6fc65b8c5c4878

Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 include/linux/nvme.h |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

--- a/include/linux/nvme.h
+++ b/include/linux/nvme.h
@@ -61,7 +61,8 @@ enum {
 
 struct nvme_id_power_state {
 	__le16			max_power;	/* centiwatts */
-	__u16			rsvd2;
+	__u8			rsvd2;
+	__u8			flags;
 	__le32			entry_lat;	/* microseconds */
 	__le32			exit_lat;	/* microseconds */
 	__u8			read_tput;
@@ -73,6 +74,11 @@ struct nvme_id_power_state {
 
 #define NVME_VS(major, minor)	(major << 16 | minor)
 
+enum {
+	NVME_PS_FLAGS_MAX_POWER_SCALE	= 1 << 0,
+	NVME_PS_FLAGS_NON_OP_STATE	= 1 << 1,
+};
+
 struct nvme_id_ctrl {
 	__le16			vid;
 	__le16			ssvid;
