From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Wed, 11 May 2011 21:38:57 -0400
Subject: NVMe: Version 0.6
Git-commit: be5e09484078e95af20acb13e215cd8aec705893
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index 12062c1..d1cd91b 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1737,6 +1737,6 @@ static void __exit nvme_exit(void)
 
 MODULE_AUTHOR("Matthew Wilcox <willy@linux.intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_VERSION("0.5");
+MODULE_VERSION("0.6");
 module_init(nvme_init);
 module_exit(nvme_exit);
-- 
1.7.4.2

