From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Mon, 14 Feb 2011 17:35:00 -0500
Subject: NVMe: Release 0.3
Git-commit: ad8a5df97cb060aa4d817af25587c99e2d2fda97
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index 606371e..7554625 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1549,6 +1549,6 @@ static void __exit nvme_exit(void)
 
 MODULE_AUTHOR("Matthew Wilcox <willy@linux.intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_VERSION("0.2");
+MODULE_VERSION("0.3");
 module_init(nvme_init);
 module_exit(nvme_exit);
-- 
1.7.4.2

