From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Thu, 24 Feb 2011 16:20:14 -0500
Subject: NVMe: Version 0.4
Git-commit: b57ab0fada358357571f0eb448cdf2f144785321
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index 45bfae1..a8549df 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1617,6 +1617,6 @@ static void __exit nvme_exit(void)
 
 MODULE_AUTHOR("Matthew Wilcox <willy@linux.intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_VERSION("0.3");
+MODULE_VERSION("0.4");
 module_init(nvme_init);
 module_exit(nvme_exit);
-- 
1.7.4.2

