From: Keith Busch <keith.busch@intel.com>
Date: Wed, 1 May 2013 13:07:48 -0600
Subject: NVMe: Check for NULL memory in nvme_dev_add
References: bnc#913030,FATE#317455
Patch-Mainline: v3.11
Git-commit: a9ef4343afbe67a6abf83c0f0294e80db48e513a

Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme-core.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/block/nvme-core.c b/drivers/block/nvme-core.c
index 53dd6f7..02bd4ad 100644
--- a/drivers/block/nvme-core.c
+++ b/drivers/block/nvme-core.c
@@ -1533,6 +1533,8 @@ static int nvme_dev_add(struct nvme_dev *dev)
 
 	mem = dma_alloc_coherent(&dev->pci_dev->dev, 8192, &dma_addr,
 								GFP_KERNEL);
+	if (!mem)
+		return -ENOMEM;
 
 	res = nvme_identify(dev, 0, 1, dma_addr);
 	if (res) {
-- 
1.8.5.2

