From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Thu, 3 Feb 2011 14:36:07 -0500
Subject: NVMe: Release 0.2
Git-commit: db5d0c198d673b6a932b449d4db95a2ad50c755e
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index 7cdf7f6..06a6aea 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1272,6 +1272,6 @@ static void __exit nvme_exit(void)
 
 MODULE_AUTHOR("Matthew Wilcox <willy@linux.intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_VERSION("0.1");
+MODULE_VERSION("0.2");
 module_init(nvme_init);
 module_exit(nvme_exit);
-- 
1.7.4.2

