From: Keith Busch <keith.busch@intel.com>
Date: Fri, 27 Jul 2012 11:53:28 -0600
Subject: NVMe: Fix uninitialized iod compiler warning
Git-commit: c7d36ab8fa04c213328119a9c0d66985fe204ee5
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index b631de3..899f489 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1158,7 +1158,7 @@ static int nvme_user_admin_cmd(struct nvme_dev *dev,
 	struct nvme_admin_cmd cmd;
 	struct nvme_command c;
 	int status, length;
-	struct nvme_iod *iod;
+	struct nvme_iod *uninitialized_var(iod);
 
 	if (!capable(CAP_SYS_ADMIN))
 		return -EACCES;
-- 
1.7.4.2

