From: Matthew Wilcox <matthew.r.wilcox@intel.com>
Date: Tue, 10 Jan 2012 16:30:15 -0500
Subject: NVMe: Version 0.8
Git-commit: 366e8217e5ec6ce9f73aec19c46d983110fb4a98
References: FATE#313627
Patch-Mainline: v3.7-rc1

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/block/nvme.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/block/nvme.c b/drivers/block/nvme.c
index b583603..28c84b1 100644
--- a/drivers/block/nvme.c
+++ b/drivers/block/nvme.c
@@ -1726,6 +1726,6 @@ static void __exit nvme_exit(void)
 
 MODULE_AUTHOR("Matthew Wilcox <willy@linux.intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_VERSION("0.7");
+MODULE_VERSION("0.8");
 module_init(nvme_init);
 module_exit(nvme_exit);
-- 
1.7.4.2

