From 3955727904623a2f1cccacd067a458077e7bd71f Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 24 Jan 2023 10:39:43 +0100
Subject: [PATCH 3/3] paddings for ath 802.11
Patch-mainline: Never (kABI padding)
References: bsc#1207439

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 drivers/net/wireless/ath/ath.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/net/wireless/ath/ath.h b/drivers/net/wireless/ath/ath.h
index f083fb903..4ffe4d671 100644
--- a/drivers/net/wireless/ath/ath.h
+++ b/drivers/net/wireless/ath/ath.h
@@ -178,6 +178,8 @@ struct ath_common {
 	const struct ath_bus_ops *bus_ops;
 	const struct ath_ps_ops *ps_ops;
 
+	void *suse_kabi_padding;
+
 	bool btcoex_enabled;
 	bool disable_ani;
 	bool bt_ant_diversity;
-- 
2.39.0

