From 379954b607cd3c094fced46fecac1d9096859666 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 24 Jan 2023 09:38:05 +0100
Subject: [PATCH 3/3] paddings for gadgets
Patch-mainline: Never (kABI padding)
References: bsc#1207439

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/usb/gadget.h |   12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/include/linux/usb/gadget.h
+++ b/include/linux/usb/gadget.h
@@ -121,6 +121,8 @@ struct usb_request {
 
 	int			status;
 	unsigned		actual;
+
+	void *suse_kabi_padding;
 };
 
 /*-------------------------------------------------------------------------*/
@@ -151,6 +153,8 @@ struct usb_ep_ops {
 
 	int (*fifo_status) (struct usb_ep *ep);
 	void (*fifo_flush) (struct usb_ep *ep);
+
+	void *suse_kabi_padding;
 };
 
 /**
@@ -239,6 +243,8 @@ struct usb_ep {
 	u8			address;
 	const struct usb_endpoint_descriptor	*desc;
 	const struct usb_ss_ep_comp_descriptor	*comp_desc;
+
+	void *suse_kabi_padding;
 };
 
 /*-------------------------------------------------------------------------*/
@@ -331,6 +337,8 @@ struct usb_gadget_ops {
 			struct usb_endpoint_descriptor *,
 			struct usb_ss_ep_comp_descriptor *);
 	int	(*check_config)(struct usb_gadget *gadget);
+
+	void *suse_kabi_padding;
 };
 
 /**
@@ -453,6 +461,8 @@ struct usb_gadget {
 	unsigned			wakeup_armed:1;
 	int				irq;
 	int				id_number;
+
+	void *suse_kabi_padding;
 };
 #define work_to_gadget(w)	(container_of((w), struct usb_gadget, work))
 
@@ -741,6 +751,8 @@ struct usb_gadget_driver {
 	char			*udc_name;
 	unsigned                match_existing_only:1;
 	bool			is_bound:1;
+
+	void *suse_kabi_padding;
 };
 
 
