From 9fc60dc996e3a6095107ac534bc5ff45f0c74844 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 24 Jan 2023 10:07:01 +0100
Subject: [PATCH] for TB and USB4 XDomain structures
Patch-mainline: Never (kABI padding)
References: bsc#1207439

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/pci.h         |    4 ++++
 include/linux/thunderbolt.h |   10 ++++++++++
 2 files changed, 14 insertions(+)

--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -775,6 +775,8 @@ struct pci_ops {
 	void __iomem *(*map_bus)(struct pci_bus *bus, unsigned int devfn, int where);
 	int (*read)(struct pci_bus *bus, unsigned int devfn, int where, int size, u32 *val);
 	int (*write)(struct pci_bus *bus, unsigned int devfn, int where, int size, u32 val);
+
+	void *suse_kabi_padding;
 };
 
 /*
@@ -853,6 +855,8 @@ struct pci_error_handlers {
 
 	/* Allow device driver to record more details of a correctable error */
 	void (*cor_error_detected)(struct pci_dev *dev);
+
+	void *suse_kabi_padding;
 };
 
 
--- a/include/linux/thunderbolt.h
+++ b/include/linux/thunderbolt.h
@@ -144,6 +144,7 @@ struct tb_property {
 		char *text;
 		u32 immediate;
 	} value;
+
 	void *suse_kabi_padding;
 };
 
@@ -260,6 +261,8 @@ struct tb_xdomain {
 	u8 target_link_width;
 	u8 link;
 	u8 depth;
+
+	void *suse_kabi_padding;
 };
 
 int tb_xdomain_lane_bonding_enable(struct tb_xdomain *xd);
@@ -363,6 +366,8 @@ struct tb_protocol_handler {
 	int (*callback)(const void *buf, size_t size, void *data);
 	void *data;
 	struct list_head list;
+
+	void *suse_kabi_padding;
 };
 
 int tb_register_protocol_handler(struct tb_protocol_handler *handler);
@@ -394,6 +399,8 @@ struct tb_service {
 	u32 prtcrevs;
 	u32 prtcstns;
 	struct dentry *debugfs_dir;
+
+	void *suse_kabi_padding;
 };
 
 static inline struct tb_service *tb_service_get(struct tb_service *svc)
@@ -435,6 +442,8 @@ struct tb_service_driver {
 	void (*remove)(struct tb_service *svc);
 	void (*shutdown)(struct tb_service *svc);
 	const struct tb_service_id *id_table;
+
+	void *suse_kabi_padding;
 };
 
 #define TB_SERVICE(key, id)				\
@@ -544,6 +553,7 @@ struct tb_ring {
 	u16 eof_mask;
 	void (*start_poll)(void *data);
 	void *poll_data;
+
 	void *suse_kabi_padding;
 };
 
