From b7ff61aa9de1cd398012a433932a5898e38de2eb Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 24 Jan 2023 10:37:16 +0100
Subject: [PATCH 2/3] paddings for mediatek 802.11
Patch-mainline: Never (kABI padding)
References: bsc#1207439

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 drivers/net/wireless/mediatek/mt76/mt76.h |   14 ++++++++++++++
 1 file changed, 14 insertions(+)

--- a/drivers/net/wireless/mediatek/mt76/mt76.h
+++ b/drivers/net/wireless/mediatek/mt76/mt76.h
@@ -72,6 +72,9 @@ struct mt76_bus_ops {
 		     const struct mt76_reg_pair *rp, int len);
 	int (*rd_rp)(struct mt76_dev *dev, u32 base,
 		     struct mt76_reg_pair *rp, int len);
+
+	void *suse_kabi_padding;
+
 	enum mt76_bus_type type;
 };
 
@@ -222,6 +225,8 @@ struct mt76_mcu_ops {
 	int (*mcu_rd_rp)(struct mt76_dev *dev, u32 base,
 			 struct mt76_reg_pair *rp, int len);
 	int (*mcu_restart)(struct mt76_dev *dev);
+
+	void *suse_kabi_padding;
 };
 
 struct mt76_queue_ops {
@@ -252,6 +257,8 @@ struct mt76_queue_ops {
 	void (*kick)(struct mt76_dev *dev, struct mt76_queue *q);
 
 	void (*reset_q)(struct mt76_dev *dev, struct mt76_queue *q);
+
+	void *suse_kabi_padding;
 };
 
 enum mt76_phy_type {
@@ -470,6 +477,8 @@ struct mt76_driver_ops {
 
 	void (*sta_remove)(struct mt76_dev *dev, struct ieee80211_vif *vif,
 			   struct ieee80211_sta *sta);
+
+	void *suse_kabi_padding;
 };
 
 struct mt76_channel_state {
@@ -559,6 +568,8 @@ struct mt76_usb {
 		int rp_len;
 		u32 base;
 	} mcu;
+
+	void *suse_kabi_padding;
 };
 
 #define MT76S_XMIT_BUF_SZ	0x3fe00
@@ -588,6 +599,7 @@ struct mt76_sdio {
 	} sched;
 
 	int (*parse_irq)(struct mt76_dev *dev, struct mt76s_intr *intr);
+	void *suse_kabi_padding;
 };
 
 struct mt76_mmio {
@@ -850,6 +862,8 @@ struct mt76_dev {
 #endif
 	struct workqueue_struct *wq;
 
+	void *suse_kabi_padding;
+
 	union {
 		struct mt76_mmio mmio;
 		struct mt76_usb usb;
