From: Yinghai Lu <yinghai@kernel.org>
Subject: PNP ACPI: Use real gsi
References: bnc#720750
Patch-Mainline: Submitted, should show up in 3.2 and stable

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 drivers/pnp/pnpacpi/rsparser.c |    8 ++++++++
 1 file changed, 8 insertions(+)

Index: linux-2.6/drivers/pnp/pnpacpi/rsparser.c
===================================================================
--- linux-2.6.orig/drivers/pnp/pnpacpi/rsparser.c
+++ linux-2.6/drivers/pnp/pnpacpi/rsparser.c
@@ -106,6 +106,14 @@ static void pnpacpi_parse_allocated_irqr
 		return;
 	}
 
+#ifdef CONFIG_X86_IO_APIC
+	/*
+	 * Convert it back to real gsi
+	 */
+	if (acpi_irq_model == ACPI_IRQ_MODEL_IOAPIC)
+		acpi_isa_irq_to_gsi(gsi, &gsi);
+#endif
+
 	/*
 	 * in IO-APIC mode, use overrided attribute. Two reasons:
 	 * 1. BIOS bug in DSDT
