From: Zhang Rui <rui.zhang@intel.com>
Subject: PNP ACPI:  Fix a buggy debug messagge.
References: bnc#720750
Patch-Mainline: submitted

Signed-off-by: Zhang Rui <rui.zhang@intel.com>
Acked-by: Yinghai Lu <yinghai@kernel.org>
Signed-off-by: Thomas Renninger <trenn@suse.de>

diff --git a/arch/x86/kernel/apic/io_apic.c b/arch/x86/kernel/apic/io_apic.c
index 3c31fa9..1a691f5 100644
--- a/arch/x86/kernel/apic/io_apic.c
+++ b/arch/x86/kernel/apic/io_apic.c
@@ -3641,8 +3641,8 @@ int io_apic_set_pci_routing(struct device *dev, int irq,
 	int node;
 
 	if (!IO_APIC_IRQ(irq)) {
-		apic_printk(APIC_QUIET,KERN_ERR "IOAPIC[%d]: Invalid reference to IRQ 0\n",
-			    irq_attr->ioapic);
+		apic_printk(APIC_QUIET,KERN_ERR "IOAPIC[%d]: Invalid reference to IRQ %d\n",
+			    irq_attr->ioapic, irq);
 		return -EINVAL;
 	}
 
