From: Michal Suchanek <msuchanek@suse.de>
Subject: fix build error due to missing header

References: bsc#1112178
Patch-mainline: no, not needed

--- a/arch/powerpc/kernel/security.c
+++ b/arch/powerpc/kernel/security.c
@@ -8,6 +8,7 @@
 #include <linux/device.h>
 #include <linux/seq_buf.h>
 #include <linux/debugfs.h>
+#include <linux/cpu.h>
 
 #include <asm/asm-prototypes.h>
 #include <asm/code-patching.h>
