Subject: [SCSI] ipr: increase dump size in ipr driver
From: wenxiong@linux.vnet.ibm.com <wenxiong@linux.vnet.ibm.com>
Patch-mainline: v3.13-rc4-69-g95d8a25
Git-commit: 95d8a25b52d551f631595ae4b5883d22a8d85a52
References: fate#317607

To help debugging several new SAS controllers, increase the dump size in ipr driver.

Signed-off-by: Wen Xiong <wenxiong@linux.vnet.ibm.com>
Acked-by: Dinar Valeev <dvaleev@suse.com>
---
 drivers/scsi/ipr.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: b/drivers/scsi/ipr.h
===================================================================
--- a/drivers/scsi/ipr.h	2014-12-10 02:28:50.471743011 -0600
+++ b/drivers/scsi/ipr.h	2014-12-10 02:29:01.551742999 -0600
@@ -301,7 +301,7 @@ IPR_PCII_NO_HOST_RRQ | IPR_PCII_IOARRIN_
  * Dump literals
  */
 #define IPR_FMT2_MAX_IOA_DUMP_SIZE			(4 * 1024 * 1024)
-#define IPR_FMT3_MAX_IOA_DUMP_SIZE			(32 * 1024 * 1024)
+#define IPR_FMT3_MAX_IOA_DUMP_SIZE			(80 * 1024 * 1024)
 #define IPR_FMT2_NUM_SDT_ENTRIES			511
 #define IPR_FMT3_NUM_SDT_ENTRIES			0xFFF
 #define IPR_FMT2_MAX_NUM_DUMP_PAGES	((IPR_FMT2_MAX_IOA_DUMP_SIZE / PAGE_SIZE) + 1)
