From 3970d80bf6fcab1caba1c0ebd7f524d89e2120b3 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Tue, 19 Nov 2019 13:38:51 +0100
Subject: [PATCH] powerpc: add link stack flush mitigation in debugfs.

References: CVE-2019-18660 bsc#1157038 bsc#1157923 ltc#182612
Patch-mainline: submitted https://patchwork.ozlabs.org/patch/1201822/

Signed-off-by: Michal Suchanek <msuchanek@suse.de>
---
 arch/powerpc/kernel/security.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/arch/powerpc/kernel/security.c b/arch/powerpc/kernel/security.c
index ec83a43b39c2..77ec2b1f6239 100644
--- a/arch/powerpc/kernel/security.c
+++ b/arch/powerpc/kernel/security.c
@@ -483,13 +483,24 @@ static int count_cache_flush_get(void *data, u64 *val)
 	return 0;
 }
 
+static int link_stack_flush_get(void *data, u64 *val)
+{
+	*val = link_stack_flush_enabled;
+
+	return 0;
+}
+
 DEFINE_SIMPLE_ATTRIBUTE(fops_count_cache_flush, count_cache_flush_get,
 			count_cache_flush_set, "%llu\n");
+DEFINE_SIMPLE_ATTRIBUTE(fops_link_stack_flush, link_stack_flush_get,
+			count_cache_flush_set, "%llu\n");
 
 static __init int count_cache_flush_debugfs_init(void)
 {
 	debugfs_create_file("count_cache_flush", 0600, powerpc_debugfs_root,
 			    NULL, &fops_count_cache_flush);
+	debugfs_create_file("link_stack_flush", 0600, powerpc_debugfs_root,
+			    NULL, &fops_link_stack_flush);
 	return 0;
 }
 device_initcall(count_cache_flush_debugfs_init);
-- 
2.12.3

