From: olh@suse.de
Subject: either dont call pm_prepare_console or resture the function delarations
in linux/pm.h

Signed-off-by: Olaf Hering <olh@suse.de>

Index: linux-2.6.15/drivers/macintosh/via-pmu.c
===================================================================
--- linux-2.6.15.orig/drivers/macintosh/via-pmu.c
+++ linux-2.6.15/drivers/macintosh/via-pmu.c
@@ -2075,7 +2075,6 @@ pmac_suspend_devices(void)
 {
 	int ret;
 
-	pm_prepare_console();
 	
 	/* Notify old-style device drivers & userland */
 	ret = broadcast_sleep(PBOOK_SLEEP_REQUEST, PBOOK_SLEEP_REJECT);
