Subject: crypto: nx - fix typo in nx driver config option
From: Seth Jennings <sjenning@linux.vnet.ibm.com>
Git-commit: 7c76bdd7c3baf6d2431bb801f5b11d2ac195fdd6
Patch-mainline: v3.5-rc2-33-g7c76bdd

Signed-off-by: Seth Jennings <sjenning@linux.vnet.ibm.com>
Acked-by: Kent Yoder <key@linux.vnet.ibm.com>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Acked-by: Dinar Valeev <dvaleev@suse.com>

Index: linux-3.0-SLE11-SP4/drivers/crypto/Kconfig
===================================================================
--- linux-3.0-SLE11-SP4.orig/drivers/crypto/Kconfig
+++ linux-3.0-SLE11-SP4/drivers/crypto/Kconfig
@@ -295,7 +295,7 @@ config CRYPTO_DEV_S5P
 	  algorithms execution.
 
 config CRYPTO_DEV_NX
-	tristate "Support for Power7+ in-Nest cryptographic accleration"
+	tristate "Support for Power7+ in-Nest cryptographic acceleration"
 	depends on PPC64 && IBMVIO
 	select CRYPTO_AES
 	select CRYPTO_CBC
