From: olh@suse.de
Subject: avoid include mess

 drivers/macintosh/via-pmu.c |    4 ++++
 1 files changed, 4 insertions(+)

Index: linux-2.6.16-rc2-olh/drivers/macintosh/via-pmu.c
===================================================================
--- linux-2.6.16-rc2-olh.orig/drivers/macintosh/via-pmu.c
+++ linux-2.6.16-rc2-olh/drivers/macintosh/via-pmu.c
@@ -2070,6 +2070,10 @@ restore_via_state(void)
 	out_8(&via[IER], IER_SET | SR_INT | CB1_INT);
 }
 
+/* to avoid include mess */
+extern int pm_prepare_console(void);
+extern void pm_restore_console(void);
+
 static int
 pmac_suspend_devices(void)
 {
