From: Tom Abraham <tabraham@suse.com>
Subject: proc_sys_revalidate: fix Oops on NULL nameidata
Git-commit: 0b728e1911cbe6e24020727c3870628b9653f32a
Patch-mainline: v3.6
References: bnc#907551

Acked-by: Miklos Szeredi <mszeredi@suse.cz>
---
 fs/proc/proc_sysctl.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/fs/proc/proc_sysctl.c
+++ b/fs/proc/proc_sysctl.c
@@ -389,7 +389,7 @@ static const struct inode_operations pro
 
 static int proc_sys_revalidate(struct dentry *dentry, struct nameidata *nd)
 {
-	if (nd->flags & LOOKUP_RCU)
+	if (nd && (nd->flags & LOOKUP_RCU))
 		return -ECHILD;
 	return !PROC_I(dentry->d_inode)->sysctl->unregistering;
 }
