From: "David S. Miller" <davem@davemloft.net>
Subject: ptp_pch: rename pch_pcidev to pci_driver to fix section mismatch
Git-commit:  d8d789497aa016f802b95606c6473e92361ec0b5
Patch-mainline: v3.4-rc1~177^2~100
References: fate#314848
Acked-by: Jiri Bohac <jbohac@suse.cz>

Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/drivers/ptp/ptp_pch.c b/drivers/ptp/ptp_pch.c
index 0b38ee1..375eb04 100644
--- a/drivers/ptp/ptp_pch.c
+++ b/drivers/ptp/ptp_pch.c
@@ -694,7 +694,7 @@ static DEFINE_PCI_DEVICE_TABLE(pch_ieee1588_pcidev_id) = {
 	{0}
 };
 
-static struct pci_driver pch_pcidev = {
+static struct pci_driver pch_driver = {
 	.name = KBUILD_MODNAME,
 	.id_table = pch_ieee1588_pcidev_id,
 	.probe = pch_probe,
@@ -705,7 +705,7 @@ static struct pci_driver pch_pcidev = {
 
 static void __exit ptp_pch_exit(void)
 {
-	pci_unregister_driver(&pch_pcidev);
+	pci_unregister_driver(&pch_driver);
 }
 
 static s32 __init ptp_pch_init(void)
@@ -713,7 +713,7 @@ static s32 __init ptp_pch_init(void)
 	s32 ret;
 
 	/* register the driver with the pci core */
-	ret = pci_register_driver(&pch_pcidev);
+	ret = pci_register_driver(&pch_driver);
 
 	return ret;
 }
-- 
1.7.11.4

