From 9eaea122699d0919cb7fc08078dbbdd864004067 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Thu, 4 Jun 2026 14:33:24 +0200
Subject: [PATCH] qcom: Kconfig: Allow disabling QCOM_SCM

References: Nvidia config
Patch-mainline: not yet, testing

---
 drivers/firmware/qcom/Kconfig | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/firmware/qcom/Kconfig b/drivers/firmware/qcom/Kconfig
index b477d54b495a6..5e421021e0e0a 100644
--- a/drivers/firmware/qcom/Kconfig
+++ b/drivers/firmware/qcom/Kconfig
@@ -9,9 +9,11 @@ menu "Qualcomm firmware drivers"
 config QCOM_SCM
 	select QCOM_TZMEM
 	tristate
+	default n
 
 config QCOM_TZMEM
 	tristate
+	default n
 	select GENERIC_ALLOCATOR
 
 choice
-- 
2.51.0

