From: Chad Dupuis <chad.dupuis@qlogic.com>
Date: Wed, 18 Sep 2013 07:02:40 -0400
Subject: qla2xxx: Add module parameter to override the default request queue size.
Patch-Mainline: Not yet
Reference: bnc#826756

Signed-off-by: Petr Uzel <petr.uzel@suse.cz>
Acked-by: Hannes Reinecke <hare@suse.de>

diff --git a/drivers/scsi/qla2xxx/qla_dbg.c b/drivers/scsi/qla2xxx/qla_dbg.c
index 82e5813..87fe609 100644
--- a/drivers/scsi/qla2xxx/qla_dbg.c
+++ b/drivers/scsi/qla2xxx/qla_dbg.c
@@ -11,7 +11,7 @@
  * ----------------------------------------------------------------------
  * |             Level            |   Last Value Used  |     Holes	|
  * ----------------------------------------------------------------------
- * | Module Init and Probe        |       0x0126       | 0x4b,0xba,0xfa,|
+ * | Module Init and Probe        |       0x0128       | 0x4b,0xba,0xfa,|
  * | Mailbox commands             |       0x117a       | 0x111a-0x111b  |
  * |                              |                    | 0x112c-0x112e  |
  * |                              |                    | 0x113a         |
diff --git a/drivers/scsi/qla2xxx/qla_def.h b/drivers/scsi/qla2xxx/qla_def.h
index 30165fc..f1b4966 100644
--- a/drivers/scsi/qla2xxx/qla_def.h
+++ b/drivers/scsi/qla2xxx/qla_def.h
@@ -260,6 +260,7 @@
 #define REQUEST_ENTRY_CNT_2100		128	/* Number of request entries. */
 #define REQUEST_ENTRY_CNT_2200		2048	/* Number of request entries. */
 #define REQUEST_ENTRY_CNT_24XX		2048	/* Number of request entries. */
+#define REQUEST_ENTRY_CNT_MAX		16384	/* Max # of request entries */
 #define RESPONSE_ENTRY_CNT_2100		64	/* Number of response entries.*/
 #define RESPONSE_ENTRY_CNT_2300		512	/* Number of response entries.*/
 #define RESPONSE_ENTRY_CNT_MQ		128	/* Number of response entries.*/
diff --git a/drivers/scsi/qla2xxx/qla_os.c b/drivers/scsi/qla2xxx/qla_os.c
index 825ffe8..bb77bf7 100644
--- a/drivers/scsi/qla2xxx/qla_os.c
+++ b/drivers/scsi/qla2xxx/qla_os.c
@@ -13,6 +13,8 @@
 #include <linux/mutex.h>
 #include <linux/kobject.h>
 #include <linux/slab.h>
+#include <linux/log2.h>
+
 #include <scsi/scsi_tcq.h>
 #include <scsi/scsicam.h>
 #include <scsi/scsi_transport.h>
@@ -226,6 +228,12 @@ MODULE_PARM_DESC(ql2xasynclogin,
 		"Default is 1 - Use asynchronous login if possible.");
 
 
+int ql2xreq_que_size = 0;
+module_param(ql2xreq_que_size, int, S_IRUGO);
+MODULE_PARM_DESC(ql2xreq_que_size,
+		"Override the default request queue size. Valid values must be "
+		"a power of 2 and between 2048 and 16384 inclusive.");
+
 /*
  * SCSI host template entry points
  */
@@ -2440,6 +2448,24 @@ qla2x00_probe_one(struct pci_dev *pdev, const struct pci_device_id *id)
 		ha->nvram_data_off = ~0;
 	}
 
+	/* Override the default request queue size if requested. */
+	if (ql2xreq_que_size >= req_length &&
+	    ql2xreq_que_size <= REQUEST_ENTRY_CNT_MAX &&
+	    is_power_of_2(ql2xreq_que_size) &&
+	    IS_FWI2_CAPABLE(ha) && !IS_QLA82XX(ha))
+		req_length = ql2xreq_que_size;
+	else if (ql2xreq_que_size > 0) {
+		if (!IS_FWI2_CAPABLE(ha) || IS_QLA82XX(ha))
+			ql_log_pci(ql_log_warn, pdev, 0x0127,
+			    "Cannot change the request queue size for this "
+			    "adapter.\n");
+		else
+			ql_log_pci(ql_log_warn, pdev, 0x128,
+			    "Changes in request queue size must be a power of "
+			    "2 between %d and %d, inclusive.\n", req_length,
+			    REQUEST_ENTRY_CNT_MAX);
+	}
+
 	ql_dbg_pci(ql_dbg_init, pdev, 0x001e,
 	    "mbx_count=%d, req_length=%d, "
 	    "rsp_length=%d, max_loop_id=%d, init_cb_size=%d, "
