From c0a8a61e7abbf66729687ee63659ee25983fbb1e Mon Sep 17 00:00:00 2001
From: Schspa Shi <schspa@gmail.com>
Date: Fri, 14 Jan 2022 16:12:16 +0800
Subject: [PATCH] random: fix typo in comments
Git-commit: c0a8a61e7abbf66729687ee63659ee25983fbb1e
References: git-fixes
Patch-mainline: v5.17-rc1

s/or/for

Signed-off-by: Schspa Shi <schspa@gmail.com>
Signed-off-by: Jason A. Donenfeld <Jason@zx2c4.com>
Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 drivers/char/random.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/char/random.c b/drivers/char/random.c
index 227fb7802738..ba8d63f52c05 100644
--- a/drivers/char/random.c
+++ b/drivers/char/random.c
@@ -101,7 +101,7 @@
  * ===============================
  *
  * There are four exported interfaces; two for use within the kernel,
- * and two or use from userspace.
+ * and two for use from userspace.
  *
  * Exported interfaces ---- userspace output
  * -----------------------------------------
-- 
2.35.3

