From a8196288de61cdcb19039141e700f35f63779739 Mon Sep 17 00:00:00 2001
From: David Disseldorp <ddiss@suse.de>
Date: Sat, 22 Aug 2015 15:30:33 +0200
Subject: [PATCH] rbd: add cmpsetattr device attribute for debugging
Patch-mainline: Not yet, SES2 clustered LIO/RBD
References: fate#318836

Signed-off-by: David Disseldorp <ddiss@suse.de>
---
 drivers/block/rbd.c |   32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

--- a/drivers/block/rbd.c
+++ b/drivers/block/rbd.c
@@ -4835,6 +4835,36 @@ static ssize_t rbd_setxattr_set(struct d
 		return size;
 }
 
+static ssize_t rbd_cmpsetxattr_set(struct device *dev,
+				   struct device_attribute *attr,
+				   const char *buf, size_t size)
+{
+	struct rbd_device *rbd_dev = dev_to_rbd_dev(dev);
+	char key[RBD_MAX_XATTR_STR_LEN];
+	char oldval[RBD_MAX_XATTR_STR_LEN];
+	char newval[RBD_MAX_XATTR_STR_LEN];
+	int ret;
+
+	ret = sscanf(buf, "%" __stringify(RBD_MAX_XATTR_STR_LEN) "s %"
+		     __stringify(RBD_MAX_XATTR_STR_LEN) "s %"
+		     __stringify(RBD_MAX_XATTR_STR_LEN) "s\n",
+		     key, oldval, newval);
+	if (ret != 3) {
+		rbd_warn(rbd_dev, "Invalid number of params. Got %d", ret);
+		return -EINVAL;
+	} else if (!strlen(key) || !strlen(oldval) || !strlen(newval)) {
+		rbd_warn(rbd_dev, "missing param");
+		return -EINVAL;
+	}
+
+	ret = rbd_dev_cmpsetxattr(rbd_dev, key, oldval, strlen(oldval) + 1,
+				  newval, strlen(newval) + 1);
+	if (ret)
+		return ret;
+	else
+		return size;
+}
+
 static DEVICE_ATTR(size, S_IRUGO, rbd_size_show, NULL);
 static DEVICE_ATTR(features, S_IRUGO, rbd_features_show, NULL);
 static DEVICE_ATTR(major, S_IRUGO, rbd_major_show, NULL);
@@ -4852,6 +4882,7 @@ static DEVICE_ATTR(current_snap, S_IRUGO
 static DEVICE_ATTR(snap_id, S_IRUGO, rbd_snap_id_show, NULL);
 static DEVICE_ATTR(parent, S_IRUGO, rbd_parent_show, NULL);
 static DEVICE_ATTR(setxattr, S_IWUSR, NULL, rbd_setxattr_set);
+static DEVICE_ATTR(cmpsetxattr, S_IWUSR, NULL, rbd_cmpsetxattr_set);
 
 static struct attribute *rbd_attrs[] = {
 	&dev_attr_size.attr,
@@ -4871,6 +4902,7 @@ static struct attribute *rbd_attrs[] = {
 	&dev_attr_parent.attr,
 	&dev_attr_refresh.attr,
 	&dev_attr_setxattr.attr,
+	&dev_attr_cmpsetxattr.attr,
 	NULL
 };
 
