Index: linux-2.6.3/fs/reiserfs/namei.c
===================================================================
--- linux-2.6.3.orig/fs/reiserfs/namei.c
+++ linux-2.6.3/fs/reiserfs/namei.c
@@ -667,7 +667,7 @@ static int reiserfs_mknod (struct inode 
         goto out_failed;
     }
 
-    init_special_inode(inode, mode, rdev) ;
+    init_special_inode(inode, inode->i_mode, rdev) ;
 
     //FIXME: needed for block and char devices only
     reiserfs_update_sd (&th, inode);
