From: Hans Reiser <reiser@namesys.com>
Subject: export find_get_pages

export find_get_pages

Signed-off-by: Andrew Morton <akpm@osdl.org>
Acked-by: Chris Mason <mason@suse.com>

---

 mm/filemap.c |    2 ++
 1 files changed, 2 insertions(+)

diff -puN mm/filemap.c~reiser4-export-find_get_pages mm/filemap.c
--- 25/mm/filemap.c~reiser4-export-find_get_pages	2005-06-25 01:17:26.000000000 -0700
+++ 25-akpm/mm/filemap.c	2005-06-25 01:17:26.000000000 -0700
@@ -665,6 +665,7 @@ unsigned find_get_pages(struct address_s
 	read_unlock_irq(&mapping->tree_lock);
 	return ret;
 }
+EXPORT_SYMBOL(find_get_pages);
 
 /*
  * Like find_get_pages, except we only return pages which are tagged with
@@ -686,6 +687,7 @@ unsigned find_get_pages_tag(struct addre
 	read_unlock_irq(&mapping->tree_lock);
 	return ret;
 }
+EXPORT_SYMBOL(find_get_pages_tag);
 
 /*
  * Same as grab_cache_page, but do not wait if the page is unavailable.
_
