

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/Kconfig.reiser4 |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN fs/Kconfig.reiser4~resier4-4kstacks-fix fs/Kconfig.reiser4
--- 25/fs/Kconfig.reiser4~resier4-4kstacks-fix	2004-08-21 14:21:30.716818728 -0700
+++ 25-akpm/fs/Kconfig.reiser4	2004-08-21 14:21:42.306056896 -0700
@@ -1,6 +1,6 @@
 config REISER4_FS
 	tristate "Reiser4 (EXPERIMENTAL very fast general purpose filesystem)"
-	depends on EXPERIMENTAL
+	depends on EXPERIMENTAL && !4KSTACKS
 	default y
 	---help---
 	  Reiser4 is more than twice as fast for both reads and writes as
_
