Subject: call panic if something unexpected happens
From: Olaf Hering <olh@suse.de>

 fs/reiser4/vfs_ops.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletion(-)

Index: linux-2.6.12/fs/reiser4/vfs_ops.c
===================================================================
--- linux-2.6.12.orig/fs/reiser4/vfs_ops.c
+++ linux-2.6.12/fs/reiser4/vfs_ops.c
@@ -1333,7 +1333,9 @@ reiser4_internal void reiser4_handle_err
 		sb->s_flags |= MS_RDONLY;
 		break;
 	case 2:
-		machine_restart(NULL);
+		panic("Hans tried to call machine_restart(NULL); here\n"
+		      "But we stopped him from doing so.\n"
+		      "A simple panic() call like this one does the job as well.\n");
 	}
 }
 
