From: Hans Reiser
Subject: export find_get_pages_tag

export find_get_pages_tag

Signed-off-by: Andrew Morton <akpm@osdl.org>
Acked-by: Chris Mason <mason@suse.com>

Index: linux.h/mm/filemap.c
===================================================================
--- linux.h.orig/mm/filemap.c	2005-01-05 11:13:26.878249909 -0500
+++ linux.h/mm/filemap.c	2005-01-05 11:13:27.944122341 -0500
@@ -618,7 +618,6 @@ unsigned find_get_pages(struct address_s
 	read_unlock_irq(&mapping->tree_lock);
 	return ret;
 }
-
 EXPORT_SYMBOL(find_get_pages);
 
 /*
@@ -641,6 +640,7 @@ unsigned find_get_pages_tag(struct addre
 	read_unlock_irq(&mapping->tree_lock);
 	return ret;
 }
+EXPORT_SYMBOL(find_get_pages_tag);
 
 /*
  * Same as grab_cache_page, but do not wait if the page is unavailable.
