
From: Hans Reiser <reiser@namesys.com>


Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/lib/radix-tree.c |    1 +
 1 files changed, 1 insertion(+)

diff -puN lib/radix-tree.c~reiser4-export-radix_tree_preload lib/radix-tree.c
--- 25/lib/radix-tree.c~reiser4-export-radix_tree_preload	Wed Aug 18 16:42:42 2004
+++ 25-akpm/lib/radix-tree.c	Wed Aug 18 16:42:42 2004
@@ -133,6 +133,7 @@ int radix_tree_preload(int gfp_mask)
 out:
 	return ret;
 }
+EXPORT_SYMBOL(radix_tree_preload);
 
 static inline void tag_set(struct radix_tree_node *node, int tag, int offset)
 {
_
