

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/inode.c |    1 -
 1 files changed, 1 deletion(-)

diff -puN fs/inode.c~reiser4-export-inode_lock-cleanup fs/inode.c
--- 25/fs/inode.c~reiser4-export-inode_lock-cleanup	Wed Aug 18 17:01:41 2004
+++ 25-akpm/fs/inode.c	Wed Aug 18 17:01:47 2004
@@ -234,7 +234,6 @@ void __iget(struct inode * inode)
 		list_move(&inode->i_list, &inode_in_use);
 	inodes_stat.nr_unused--;
 }
-
 EXPORT_SYMBOL(__iget);
 
 /**
_
