From: Jan Kara <jack@suse.cz>
References: fate#302681
Subject: [PATCH 08/28] reiserfs: Use sb_any_quota_loaded() instead of sb_any_quota_enabled().
Patch-mainline: 2.6.29?

Signed-off-by: Jan Kara <jack@suse.cz>
---
 fs/reiserfs/super.c |    8 +++-----
 1 files changed, 3 insertions(+), 5 deletions(-)

Index: linux-2.6.27/fs/reiserfs/super.c
===================================================================
--- linux-2.6.27.orig/fs/reiserfs/super.c	2008-10-24 11:40:32.000000000 +0200
+++ linux-2.6.27/fs/reiserfs/super.c	2008-10-24 12:06:32.000000000 +0200
@@ -1011,8 +1011,7 @@
 		if (c == 'u' || c == 'g') {
 			int qtype = c == 'u' ? USRQUOTA : GRPQUOTA;
 
-			if ((sb_any_quota_enabled(s) ||
-			     sb_any_quota_suspended(s)) &&
+			if (sb_any_quota_loaded(s) &&
 			    (!*arg != !REISERFS_SB(s)->s_qf_names[qtype])) {
 				reiserfs_warning(s, "super-6511",
 						 "cannot change journaled "
@@ -1065,8 +1064,7 @@
 						 "specified.");
 				return 0;
 			}
-			if ((sb_any_quota_enabled(s) ||
-			     sb_any_quota_suspended(s)) &&
+			if (sb_any_quota_loaded(s) &&
 			    *qfmt != REISERFS_SB(s)->s_jquota_fmt) {
 				reiserfs_warning(s, "super-6515",
 						 "cannot change journaled "
@@ -1093,7 +1091,7 @@
 	}
 	/* This checking is not precise wrt the quota type but for our purposes it is sufficient */
 	if (!(*mount_options & (1 << REISERFS_QUOTA))
-	    && sb_any_quota_enabled(s)) {
+	    && sb_any_quota_loaded(s)) {
 		reiserfs_warning(s, "super-6516", "quota options must "
 				 "be present when quota is turned on.");
 		return 0;
