From: jeffm@suse.com
Subject: reiserfs_file_write should not allow files > 2G on 3.5.x disks
References: 42191

reiserfs_file_write should not allow files > 2G on 3.5.x. disks.

Signed-off-by: mason@suse.com

Index: linux.head/fs/reiserfs/file.c
===================================================================
--- linux.head.orig/fs/reiserfs/file.c	2004-12-17 08:21:08.737487290 -0500
+++ linux.head/fs/reiserfs/file.c	2004-12-17 08:21:54.758007974 -0500
@@ -1166,6 +1166,23 @@ ssize_t reiserfs_file_write( struct file
     struct reiserfs_transaction_handle th;
     th.t_trans_id = 0;
 
+    /* If a filesystem is converted from 3.5 to 3.6, we'll have v3.5 items
+     * lying around (most of the disk, in fact). Despite the filesystem
+     * now being a v3.6 format, the old items still can't support large
+     * file sizes. Catch this case here, as the rest of the VFS layer is
+     * oblivious to the different limitations between old and new items.
+     * reiserfs_setattr catches this for truncates. This chunk is lifted
+     * from generic_write_checks. */
+    if (get_inode_item_key_version (inode) == KEY_FORMAT_3_5 && 
+            *ppos + count > MAX_NON_LFS) {
+            if (*ppos >= MAX_NON_LFS) {
+                send_sig(SIGXFSZ, current, 0);
+                return -EFBIG;
+            }
+            if (count > MAX_NON_LFS - (unsigned long)*ppos)
+                count = MAX_NON_LFS - (unsigned long)*ppos;
+    }
+
     if ( file->f_flags & O_DIRECT) { // Direct IO needs treatment
 	ssize_t result, after_file_end = 0;
 	if ( (*ppos + count >= inode->i_size) || (file->f_flags & O_APPEND) ) {
