

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/mm/swap.c |    2 --
 1 files changed, 2 deletions(-)

diff -puN mm/swap.c~reiser4-export-pagevec-funcs-cleanup mm/swap.c
--- 25/mm/swap.c~reiser4-export-pagevec-funcs-cleanup	Wed Aug 18 17:02:19 2004
+++ 25-akpm/mm/swap.c	Wed Aug 18 17:02:35 2004
@@ -258,7 +258,6 @@ void __pagevec_release(struct pagevec *p
 	release_pages(pvec->pages, pagevec_count(pvec), pvec->cold);
 	pagevec_reinit(pvec);
 }
-
 EXPORT_SYMBOL(__pagevec_release);
 
 /*
@@ -389,7 +388,6 @@ unsigned pagevec_lookup_tag(struct pagev
 					nr_pages, pvec->pages);
 	return pagevec_count(pvec);
 }
-
 EXPORT_SYMBOL(pagevec_lookup_tag);
 
 #ifdef CONFIG_SMP
_
