From: David Bond <dbond@suse.com>
Date: Tue, 31 May 2022 08:53:10 -0600
Subject: revert scsi: qla2xxx: Changes to support FCP2 Target
Patch-mainline: Never, revert of mainline commit causing customer issue for test package
References: bsc#1198438

Revert Adding changes to support FCP2 Target.
---

diff --git a/drivers/scsi/qla2xxx/qla_init.c b/drivers/scsi/qla2xxx/qla_init.c
index 88c5a8ec0e7c..4b27eebbc1b3 100644
--- a/drivers/scsi/qla2xxx/qla_init.c
+++ b/drivers/scsi/qla2xxx/qla_init.c
@@ -1821,13 +1821,6 @@ void qla2x00_handle_rscn(scsi_qla_host_t *vha, struct event_arg *ea)
 	case RSCN_PORT_ADDR:
 		fcport = qla2x00_find_fcport_by_nportid(vha, &ea->id, 1);
 		if (fcport) {
-			if (fcport->flags & FCF_FCP2_DEVICE) {
-				ql_dbg(ql_dbg_disc, vha, 0x2115,
-				       "Delaying session delete for FCP2 portid=%06x %8phC ",
-					fcport->d_id.b24, fcport->port_name);
-				return;
-			}
-
 			if (vha->hw->flags.edif_enabled && DBELL_ACTIVE(vha)) {
 				/*
 				 * On ipsec start by remote port, Target port
diff --git a/drivers/scsi/qla2xxx/qla_os.c b/drivers/scsi/qla2xxx/qla_os.c
index 850fc95f7033..9285db77b95d 100644
--- a/drivers/scsi/qla2xxx/qla_os.c
+++ b/drivers/scsi/qla2xxx/qla_os.c
@@ -4062,16 +4062,6 @@ qla2x00_mark_all_devices_lost(scsi_qla_host_t *vha)
 	    "Mark all dev lost\n");
 
 	list_for_each_entry(fcport, &vha->vp_fcports, list) {
-		if (fcport->loop_id != FC_NO_LOOP_ID &&
-		    (fcport->flags & FCF_FCP2_DEVICE) &&
-		    fcport->port_type == FCT_TARGET &&
-		    !qla2x00_reset_active(vha)) {
-			ql_dbg(ql_dbg_disc, vha, 0x211a,
-			       "Delaying session delete for FCP2 flags 0x%x port_type = 0x%x port_id=%06x %phC",
-			       fcport->flags, fcport->port_type,
-			       fcport->d_id.b24, fcport->port_name);
-			continue;
-		}
 		fcport->scan_state = 0;
 		qlt_schedule_sess_for_deletion(fcport);
 	}
