From 57730dea2ecb37b45fc80259f81d37aa7d8f1cb5 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 1 Feb 2022 15:20:30 +0100
Subject: [PATCH 1/2] rfkill: add kABI padding Patch-mainline: Never, kABI
 padding References: bsc#1179531
Patch-mainline: Never, kABI padding
References: bsc#1179531

To the main structure, cost is minimal.

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/rfkill.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/rfkill.h b/include/linux/rfkill.h
index 231e06b74..774c4226b 100644
--- a/include/linux/rfkill.h
+++ b/include/linux/rfkill.h
@@ -62,6 +62,8 @@ struct rfkill_ops {
 	void	(*poll)(struct rfkill *rfkill, void *data);
 	void	(*query)(struct rfkill *rfkill, void *data);
 	int	(*set_block)(void *data, bool blocked);
+
+	void * suse_kabi_padding;
 };
 
 #if defined(CONFIG_RFKILL) || defined(CONFIG_RFKILL_MODULE)
-- 
2.34.1

