From: Jan Glauber <jang@linux.vnet.ibm.com>
Subject: s390/pci: remove obsolete email addresses
Patch-mainline: v3.8-rc1
Git-commit: bedef75506ed40ca69ee7e1b7ebe47042bab72b2
References: bnc#848335,FATE#83037,LTC#94737

Signed-off-by: Jan Glauber <jang@linux.vnet.ibm.com>
Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/pci/pci.c |   14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

--- a/arch/s390/pci/pci.c
+++ b/arch/s390/pci/pci.c
@@ -6,13 +6,13 @@
  *
  * The System z PCI code is a rewrite from a prototype by
  * the following people (Kudoz!):
- *   Alexander Schmidt <alexschm@de.ibm.com>
- *   Christoph Raisch <raisch@de.ibm.com>
- *   Hannes Hering <hering2@de.ibm.com>
- *   Hoang-Nam Nguyen <hnguyen@de.ibm.com>
- *   Jan-Bernd Themann <themann@de.ibm.com>
- *   Stefan Roscher <stefan.roscher@de.ibm.com>
- *   Thomas Klein <tklein@de.ibm.com>
+ *   Alexander Schmidt
+ *   Christoph Raisch
+ *   Hannes Hering
+ *   Hoang-Nam Nguyen
+ *   Jan-Bernd Themann
+ *   Stefan Roscher
+ *   Thomas Klein
  */
 
 #define COMPONENT "zPCI"
