From: Michael Holzheu <holzheu@linux.vnet.ibm.com>
Subject: [S390] update default configuration
Date:   Wed Apr 11 14:28:08 2012 +0200
Patch-mainline: v3.4-rc4
Git-commit: ac2ac6e852716448d962033a041182cd1e2d3c01
References: bnc#786472,ltc#82694,fate#314077
    
Enable CRASH_DUMP    

Signed-off-by: Michael Holzheu <holzheu@linux.vnet.ibm.com>
Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John L. Jolly <jjolly@suse.de>

---
 arch/s390/defconfig |    1 +
 1 file changed, 1 insertion(+)

--- linux-3.0-SLE11-SP3.orig/arch/s390/defconfig	2012-11-05 22:35:21.000000000 +0100
+++ linux-3.0-SLE11-SP3/arch/s390/defconfig	2012-11-05 22:36:40.000000000 +0100
@@ -31,6 +31,7 @@ CONFIG_PREEMPT=y
 CONFIG_MEMORY_HOTPLUG=y
 CONFIG_MEMORY_HOTREMOVE=y
 CONFIG_KSM=y
+CONFIG_CRASH_DUMP=y
 CONFIG_BINFMT_MISC=m
 CONFIG_CMM=m
 CONFIG_HZ_100=y
