From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390/spinlock: add gmb memory barrier
References: bsc#1068032
Patch-mainline: Not yet, under development

Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>
---
 arch/s390/include/asm/system.h |    9 +++++++++
 1 file changed, 9 insertions(+)

--- a/arch/s390/include/asm/system.h
+++ b/arch/s390/include/asm/system.h
@@ -11,6 +11,7 @@
 #include <linux/kernel.h>
 #include <linux/errno.h>
 #include <linux/string.h>
+#include <asm/alternative.h>
 #include <asm/types.h>
 #include <asm/ptrace.h>
 #include <asm/setup.h>
@@ -176,6 +177,14 @@ extern int copy_from_user_real(void *des
  * all memory ops have completed wrt other CPU's ( see 7-15 POP  DJB ).
  */
 
+static inline void gmb(void)
+{
+	asm volatile(
+		ALTERNATIVE("", ".long 0xb2e8f000", 81)
+		: : : "memory");
+}
+#define gmb gmb
+
 #define eieio()	asm volatile("bcr 15,0" : : : "memory")
 #define SYNC_OTHER_CORES(x)   eieio()
 #define mb()    eieio()
