From: Michael Holzheu <holzheu@linux.vnet.ibm.com>
Subject: [S390] Remove useless newline in reserve_kdump_bootmem()
Date:   Tue Dec 27 11:27:06 2011 +0100
Patch-mainline: v3.3-rc1
Git-commit: 4999023aa95a00507d3f100ea75510c5c7270f74
References: bnc#786472,ltc#82694,fate#314077
    
Signed-off-by: Michael Holzheu <holzheu@linux.vnet.ibm.com>
Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John L. Jolly <jjolly@suse.de>

---
 arch/s390/kernel/setup.c |    1 -
 1 file changed, 1 deletion(-)

--- a/arch/s390/kernel/setup.c
+++ b/arch/s390/kernel/setup.c
@@ -662,7 +662,6 @@ static int __init verify_crash_base(unsi
 static void __init reserve_kdump_bootmem(unsigned long addr, unsigned long size,
 					 int type)
 {
-
 	create_mem_hole(memory_chunk, addr, size, type);
 }
 
