From 8e0bcc722289275d5586a08b0d33b737bac2836e Mon Sep 17 00:00:00 2001
From: Al Viro <viro@zeniv.linux.org.uk>
Date: Mon, 15 Apr 2013 13:29:15 -0400
Subject: [PATCH] fix a leak in /proc/schedstats
References: bsc#1094876
Patch-mainline: v3.10-rc1
Git-commit: 8e0bcc722289275d5586a08b0d33b737bac2836e

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: Oscar Salvador <osalvador@suse.de>
---
 kernel/sched_stats.h | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/kernel/sched_stats.h b/kernel/sched_stats.h
index 7d9eb0172cfa..44df48a56bde 100644
--- a/kernel/sched_stats.h
+++ b/kernel/sched_stats.h
@@ -124,16 +124,11 @@ static int schedstat_open(struct inode *inode, struct file *file)
 	return seq_open(file, &schedstat_sops);
 }
 
-static int schedstat_release(struct inode *inode, struct file *file)
-{
-	return 0;
-};
-
 static const struct file_operations proc_schedstat_operations = {
 	.open    = schedstat_open,
 	.read    = seq_read,
 	.llseek  = seq_lseek,
-	.release = schedstat_release,
+	.release = seq_release,
 };
 
 static int __init proc_schedstat_init(void)
-- 
2.13.6

