
From: Ingo Molnar <mingo@elte.hu>

misc fixes:

 - hotplug CPU fix
 - early init cleanup


Index: linux-2.6.5/init/main.c
===================================================================
--- linux-2.6.5.orig/init/main.c	2004-04-16 15:06:40.000000000 +0200
+++ linux-2.6.5/init/main.c	2004-04-16 15:06:43.000000000 +0200
@@ -416,6 +416,13 @@
 	 */
 	smp_prepare_boot_cpu();
 
+	/*
+	 * Set up the scheduler prior starting any interrupts (such as the
+	 * timer interrupt). Full topology setup happens at smp_init()
+	 * time - but meanwhile we still have a functioning scheduler.
+	 */
+	sched_init();
+
 	build_all_zonelists();
 	page_alloc_init();
 	printk("Kernel command line: %s\n", saved_command_line);
@@ -427,7 +434,7 @@
 	rcu_init();
 	init_IRQ();
 	pidhash_init();
-	sched_init();
+	init_timers();
 	softirq_init();
 	time_init();
 
Index: linux-2.6.5/kernel/sched.c
===================================================================
--- linux-2.6.5.orig/kernel/sched.c	2004-04-16 15:06:43.000000000 +0200
+++ linux-2.6.5/kernel/sched.c	2004-04-16 15:06:43.000000000 +0200
@@ -1133,6 +1133,7 @@
 		 * the migration.
 		 */
 		tlb_migrate_prepare(current->mm);
+		unlock_cpu_hotplug();
 
 		return;
 	}
@@ -3623,8 +3624,6 @@
 	set_task_cpu(current, smp_processor_id());
 	wake_up_forked_process(current);
 
-	init_timers();
-
 	/*
 	 * The boot idle thread does lazy MMU switching as well:
 	 */
