Index: linux-2.6.5/arch/ppc64/kernel/smp.c
===================================================================
--- linux-2.6.5.orig/arch/ppc64/kernel/smp.c	2004-04-16 15:06:45.000000000 +0200
+++ linux-2.6.5/arch/ppc64/kernel/smp.c	2004-04-16 15:06:45.000000000 +0200
@@ -967,8 +967,8 @@
 		first_cpu = last_cpu = NULL;
 
 		if (i != first_cpu(cpu_domain->span)) {
-			&per_cpu(cpu_domains, i)->flags |= SD_SHARE_CPUPOWER;
-			&per_cpu(cpu_domains, first_cpu(cpu_domain->span))->flags |=
+			per_cpu(cpu_domains, i).flags |= SD_SHARE_CPUPOWER;
+			per_cpu(cpu_domains, first_cpu(cpu_domain->span)).flags |=
 				SD_SHARE_CPUPOWER;
 			continue;
 		}
