From 87fa75e6a2ed60c532f3cfc74d9c0664971ec4b3 Mon Sep 17 00:00:00 2001
From: Saurav Kashyap <skashyap@marvell.com>
Date: Fri, 20 Oct 2023 14:01:22 +0530
Subject: [PATCH 1/2] qedf: Don't process stag work during unload.

Signed-off-by: Saurav Kashyap <skashyap@marvell.com>
Patch-mainline: not yet, waiting for upstream submission by Marvell
References: bsc#1214852
Acked-by: Martin Wilck <mwilck@suse.com>
---
 drivers/scsi/qedf/qedf_main.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/drivers/scsi/qedf/qedf_main.c
+++ b/drivers/scsi/qedf/qedf_main.c
@@ -3996,6 +3996,12 @@ void qedf_stag_change_work(struct work_s
 	struct qedf_ctx *qedf =
 	    container_of(work, struct qedf_ctx, stag_work.work);
 
+	if (test_bit(QEDF_UNLOADING, &qedf->flags)) {
+		QEDF_ERR(&qedf->dbg_ctx, "Driver unloading\n");
+		return;
+	}
+
+
 	printk_ratelimited("[%s]:[%s:%d]:%d: Performing software context reset.",
 			dev_name(&qedf->pdev->dev), __func__, __LINE__,
 			qedf->dbg_ctx.host_no);
