From: "Stewart, Sean" <Sean.Stewart@netapp.com>
Date: Wed, 11 Sep 2013 22:54:07 +0000
Subject: scsi_dh_rdac: Add new IBM 1813 product id to rdac devlist
References: bnc#846654
Patch-Mainline: queued for 3.14
Git-commit: scsi-misc 51fc01c4bcf8d89de14723813a59d716fdd5a56d

Signed-off-by: Sean Stewart <Sean.Stewart@netapp.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Acked-by: Hannes Reinecke <hare@suse.de>

diff --git a/drivers/scsi/device_handler/scsi_dh_rdac.c b/drivers/scsi/device_handler/scsi_dh_rdac.c
index 69c915a..4b9cf93 100644
--- a/drivers/scsi/device_handler/scsi_dh_rdac.c
+++ b/drivers/scsi/device_handler/scsi_dh_rdac.c
@@ -786,6 +786,7 @@ static const struct scsi_dh_devlist rdac_dev_list[] = {
 	{"IBM", "1742"},
 	{"IBM", "1745"},
 	{"IBM", "1746"},
+	{"IBM", "1813"},
 	{"IBM", "1814"},
 	{"IBM", "1815"},
 	{"IBM", "1818"},
