From: Fengguang Wu <fengguang.wu@intel.com>
Date: Wed, 3 Jul 2013 14:07:40 +0800
Subject: [PATCH v2 135/244] sfc: efx_ethtool_get_ts_info() can be static
Patch-mainline: v3.13-rc1
Git-commit: b51ca34a262f5a3556eb6f1d05fd88503927d3ae
References: bsc#909618 FATE#317521

Signed-off-by: Fengguang Wu <fengguang.wu@intel.com>
Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/sfc/ethtool.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/net/ethernet/sfc/ethtool.c
+++ b/drivers/net/ethernet/sfc/ethtool.c
@@ -1034,8 +1034,8 @@ static int efx_ethtool_set_rxfh_indir(st
 	return 0;
 }
 
-int efx_ethtool_get_ts_info(struct net_device *net_dev,
-			    struct ethtool_ts_info *ts_info)
+static int efx_ethtool_get_ts_info(struct net_device *net_dev,
+				   struct ethtool_ts_info *ts_info)
 {
 	struct efx_nic *efx = netdev_priv(net_dev);
 
