From: Ben Hutchings <bhutchings@solarflare.com>
Date: Tue, 8 Oct 2013 16:36:58 +0100
Subject: [PATCH v2 175/244] sfc: Map MCDI error MC_CMD_ERR_ENOTSUP to Linux
 EOPNOTSUPP
Patch-mainline: v3.14-rc1
Git-commit: ea136ae71f392341475116a371b92cc78183c2b2
References: bsc#909618 FATE#317521

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/sfc/mcdi.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/net/ethernet/sfc/mcdi.c
+++ b/drivers/net/ethernet/sfc/mcdi.c
@@ -192,6 +192,8 @@ static int efx_mcdi_errno(unsigned int m
 	TRANSLATE_ERROR(EALREADY);
 	TRANSLATE_ERROR(ENOSPC);
 #undef TRANSLATE_ERROR
+	case MC_CMD_ERR_ENOTSUP:
+		return -EOPNOTSUPP;
 	case MC_CMD_ERR_ALLOC_FAIL:
 		return -ENOBUFS;
 	case MC_CMD_ERR_MAC_EXIST:
