From: Andrew Rybchenko <Andrew.Rybchenko@oktetlabs.ru>
Date: Thu, 28 Nov 2013 08:28:13 +0400
Subject: [PATCH v2 203/244] sfc: remove unused 'enum efx_rx_alloc_method'
Patch-mainline: v3.14-rc1
Git-commit: 8c5270acf3f74a2bff04461ce7fbbcf0ae54b2ae
References: bsc#909618 FATE#317521

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/sfc/net_driver.h |    6 ------
 1 file changed, 6 deletions(-)

--- a/drivers/net/ethernet/sfc/net_driver.h
+++ b/drivers/net/ethernet/sfc/net_driver.h
@@ -358,12 +358,6 @@ struct efx_rx_queue {
 	unsigned int slow_fill_count;
 };
 
-enum efx_rx_alloc_method {
-	RX_ALLOC_METHOD_AUTO = 0,
-	RX_ALLOC_METHOD_SKB = 1,
-	RX_ALLOC_METHOD_PAGE = 2,
-};
-
 enum efx_sync_events_state {
 	SYNC_EVENTS_DISABLED = 0,
 	SYNC_EVENTS_QUIESCENT,
