From: Ben Hutchings <bhutchings@solarflare.com>
Date: Wed, 9 Oct 2013 14:14:41 +0100
Subject: [PATCH v2 177/244] sfc: Expose NVRAM_PARTITION_TYPE_LICENSE on EF10
Patch-mainline: v3.14-rc1
Git-commit: a84f3bf909c7ef5dbadd3401c95376d8ea44b2fe
References: bsc#909618 FATE#317521

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/sfc/ef10.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/ethernet/sfc/ef10.c
+++ b/drivers/net/ethernet/sfc/ef10.c
@@ -3296,6 +3296,7 @@ static const struct efx_ef10_nvram_type_
 	{ NVRAM_PARTITION_TYPE_EXPROM_CONFIG_PORT1, 0,   1, "sfc_exp_rom_cfg" },
 	{ NVRAM_PARTITION_TYPE_EXPROM_CONFIG_PORT2, 0,   2, "sfc_exp_rom_cfg" },
 	{ NVRAM_PARTITION_TYPE_EXPROM_CONFIG_PORT3, 0,   3, "sfc_exp_rom_cfg" },
+	{ NVRAM_PARTITION_TYPE_LICENSE,		   0,    0, "sfc_license" },
 	{ NVRAM_PARTITION_TYPE_PHY_MIN,		   0xff, 0, "sfc_phy_fw" },
 };
 
