From: Andrew Rybchenko <Andrew.Rybchenko@oktetlabs.ru>
Date: Sat, 23 Nov 2013 08:42:07 +0400
Subject: [PATCH v2 202/244] sfc: remove unused 'refcnt' from efx_rx_page_state
Patch-mainline: v3.14-rc1
Git-commit: ff2de51bfb26e8d6f2ee6d45d5a5eedf51c19d2f
References: bsc#909618 FATE#317521

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/sfc/net_driver.h |    3 ---
 1 file changed, 3 deletions(-)

--- a/drivers/net/ethernet/sfc/net_driver.h
+++ b/drivers/net/ethernet/sfc/net_driver.h
@@ -286,12 +286,9 @@ struct efx_rx_buffer {
  * Used to facilitate sharing dma mappings between recycled rx buffers
  * and those passed up to the kernel.
  *
- * @refcnt: Number of struct efx_rx_buffer's referencing this page.
- *	When refcnt falls to zero, the page is unmapped for dma
  * @dma_addr: The dma address of this page.
  */
 struct efx_rx_page_state {
-	unsigned refcnt;
 	dma_addr_t dma_addr;
 
 	unsigned int __pad[0] ____cacheline_aligned;
