From: Ben Hutchings <bhutchings@solarflare.com>
Date: Mon, 9 Jan 2012 19:54:16 +0000
Subject: sfc: Remove redundant 'rc' variable, always set to 0
Patch-mainline: v3.4-rc1
Git-commit: 6c8eef4ac8dfd888be5b46e363f8e42b9f0d5b6a
References: bnc#786035 FATE#314299

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/sfc/efx.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- a/drivers/net/ethernet/sfc/efx.c
+++ b/drivers/net/ethernet/sfc/efx.c
@@ -1808,7 +1808,6 @@ static void efx_watchdog(struct net_devi
 static int efx_change_mtu(struct net_device *net_dev, int new_mtu)
 {
 	struct efx_nic *efx = netdev_priv(net_dev);
-	int rc = 0;
 
 	EFX_ASSERT_RESET_SERIALISED(efx);
 
@@ -1833,7 +1832,7 @@ static int efx_change_mtu(struct net_dev
 
 	efx_start_all(efx);
 	netif_device_attach(efx->net_dev);
-	return rc;
+	return 0;
 }
 
 static int efx_set_mac_address(struct net_device *net_dev, void *data)
