From: Edward Cree <ecree@solarflare.com>
Date: Tue, 30 Jun 2020 13:14:13 +0100
Subject: sfc: initialise max_[tx_]channels in efx_init_channels()
Patch-mainline: v5.9-rc1
Git-commit: 937aa3ae4d14e2184b5ce5a801003723712e96bd
References: jsc#SLE-16683

Signed-off-by: Edward Cree <ecree@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 drivers/net/ethernet/sfc/efx_channels.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/net/ethernet/sfc/efx_channels.c
+++ b/drivers/net/ethernet/sfc/efx_channels.c
@@ -566,6 +566,9 @@ int efx_init_channels(struct efx_nic *ef
 	efx->interrupt_mode = min(efx->type->min_interrupt_mode,
 				  interrupt_mode);
 
+	efx->max_channels = EFX_MAX_CHANNELS;
+	efx->max_tx_channels = EFX_MAX_CHANNELS;
+
 	return 0;
 }
 
