From: Paul Bolle <pebolle@tiscali.nl>
Date: Fri, 15 Mar 2013 12:59:30 +0100
Subject: [PATCH v2 029/244] sfc: Fix Kconfig typo "----help---"
Patch-mainline: v3.11-rc1
Git-commit: 806b2139db236e0cbd0b5833ab0ce139f0196bcd
References: bsc#909618 FATE#317521

Signed-off-by: Paul Bolle <pebolle@tiscali.nl>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/sfc/Kconfig |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/ethernet/sfc/Kconfig
+++ b/drivers/net/ethernet/sfc/Kconfig
@@ -24,7 +24,7 @@ config SFC_MCDI_MON
 	bool "Solarflare SFC9000-family hwmon support"
 	depends on SFC && HWMON && !(SFC=y && HWMON=m)
 	default y
-	----help---
+	---help---
 	  This exposes the on-board firmware-managed sensors as a
 	  hardware monitor device.
 config SFC_SRIOV
