From: Ben Hutchings <bhutchings@solarflare.com>
Date: Mon, 9 Jan 2012 19:41:48 +0000
Subject: sfc: Remove obsolete function efx_dev_name()
Patch-mainline: v3.4-rc1
Git-commit: b249513e8ba3ca8bc2c87e78eb6e302d5d8abd6f
References: bnc#786035 FATE#314299

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/sfc/net_driver.h |    9 ---------
 1 file changed, 9 deletions(-)

--- a/drivers/net/ethernet/sfc/net_driver.h
+++ b/drivers/net/ethernet/sfc/net_driver.h
@@ -791,15 +791,6 @@ static inline int efx_dev_registered(str
 	return efx->net_dev->reg_state == NETREG_REGISTERED;
 }
 
-/* Net device name, for inclusion in log messages if it has been registered.
- * Use efx->name not efx->net_dev->name so that races with (un)registration
- * are harmless.
- */
-static inline const char *efx_dev_name(struct efx_nic *efx)
-{
-	return efx_dev_registered(efx) ? efx->name : "";
-}
-
 static inline unsigned int efx_port_num(struct efx_nic *efx)
 {
 	return efx->net_dev->dev_id;
