From: Edward Cree <ecree@solarflare.com>
Date: Thu, 2 Jul 2020 17:32:04 +0100
Subject: sfc_ef100: add EF100 to NIC-revision enumeration
Patch-mainline: v5.9-rc1
Git-commit: 805d22bf92f1541cbb5b5544f628864610e2c9a2
References: jsc#SLE-16683

Signed-off-by: Edward Cree <ecree@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 drivers/net/ethernet/sfc/nic_common.h |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/ethernet/sfc/nic_common.h
+++ b/drivers/net/ethernet/sfc/nic_common.h
@@ -21,6 +21,7 @@ enum {
 	 */
 	EFX_REV_SIENA_A0 = 3,
 	EFX_REV_HUNT_A0 = 4,
+	EFX_REV_EF100 = 5,
 };
 
 static inline int efx_nic_rev(struct efx_nic *efx)
