Subject: staging-hv: bind old hv_blkvsc to hv_storvsc
From: <ohering@suse.de>
Patch-mainline: never
---
 drivers/staging/hv/storvsc_drv.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/staging/hv/storvsc_drv.c
+++ b/drivers/staging/hv/storvsc_drv.c
@@ -1456,5 +1456,7 @@ static void __exit storvsc_drv_exit(void
 MODULE_LICENSE("GPL");
 MODULE_VERSION(HV_DRV_VERSION);
 MODULE_DESCRIPTION("Microsoft Hyper-V virtual storage driver");
+/* bind to old driver to simplify upgrade path */
+MODULE_ALIAS("hv_blkvsc");
 module_init(storvsc_drv_init);
 module_exit(storvsc_drv_exit);
