From 5e2308fcffc65b644a4bc4cbae8ec3a763245034 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Wed, 9 May 2018 15:34:32 +0200
Subject: [PATCH] STF: no POWER9 support on SLE12-SP2

References: CVE-2018-3639, bsc#1087082
Patch-mainline: no, not needed

Signed-off-by: Michal Suchanek <msuchanek@suse.de>
---
 arch/powerpc/kernel/security.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/powerpc/kernel/security.c b/arch/powerpc/kernel/security.c
--- a/arch/powerpc/kernel/security.c
+++ b/arch/powerpc/kernel/security.c
@@ -948,9 +948,12 @@ void setup_stf_barrier(void)
 	hv = cpu_has_feature(CPU_FTR_HVMODE);
 
 	/* Default to fallback in case fw-features are not available */
+#ifdef CPU_FTR_ARCH_300
 	if (cpu_has_feature(CPU_FTR_ARCH_300))
 		type = STF_BARRIER_EIEIO;
-	else if (cpu_has_feature(CPU_FTR_ARCH_207S))
+	else
+#endif
+	     if (cpu_has_feature(CPU_FTR_ARCH_207S))
 		type = STF_BARRIER_SYNC_ORI;
 	else if (cpu_has_feature(CPU_FTR_ARCH_206))
 		type = STF_BARRIER_FALLBACK;
-- 
2.12.3

