Subject: hv: mana: remove netdev_lockdep_set_classes usage
From: Olaf Hering <ohering@suse.de>
Patch-mainline: Never, old kernel
References: jsc#SLE-18779, bsc#1185726

--- a/drivers/net/ethernet/microsoft/mana/mana_en.c
+++ b/drivers/net/ethernet/microsoft/mana/mana_en.c
@@ -1787,7 +1787,7 @@ static int mana_probe_port(struct mana_c
 	if (err)
 		goto free_net;
 
-	netdev_lockdep_set_classes(ndev);
+
 
 	ndev->hw_features = NETIF_F_SG | NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM;
 	ndev->hw_features |= NETIF_F_RXCSUM;
