diff -purNX /home/olaf/kernel/kernel_exclude.txt linux-2.6.0-test11.SuSE.orig/arch/ppc/kernel/process.c linux-2.6.0-test11.SuSE/arch/ppc/kernel/process.c
--- linux-2.6.0-test11.SuSE.orig/arch/ppc/kernel/process.c	2003-11-26 20:44:06.000000000 +0000
+++ linux-2.6.0-test11.SuSE/arch/ppc/kernel/process.c	2003-12-14 18:15:40.000000000 +0000
@@ -534,7 +534,7 @@ void show_stack(struct task_struct *tsk,
 				next_exc = 0;
 			} else
 				ret = *(unsigned long *)(sp + 4);
-			printk(" [%08lx] ", ret);
+			printk(" [<%08lx>] ", ret);
 #ifdef CONFIG_KALLSYMS
 			print_symbol("%s", ret);
 			printk("\n");
