diff -p -purN linuxppc64-2.5.orig/arch/ppc64/kernel/process.c linuxppc64-2.5/arch/ppc64/kernel/process.c
--- linuxppc64-2.5.orig/arch/ppc64/kernel/process.c	2003-10-09 14:34:24.000000000 +0200
+++ linuxppc64-2.5/arch/ppc64/kernel/process.c	2003-12-14 22:26:59.000000000 +0100
@@ -412,7 +408,10 @@ void show_stack(struct task_struct *p, u
 
 	if (sp == 0)
 		sp = p->thread.ksp;
-	printk("Call Trace:\n");
+	printk("Call Trace:");
+#ifdef CONFIG_KALLSYMS
+	printk("\n");
+#endif
 	do {
 		if (__get_user(sp, (unsigned long *)sp))
 			break;
@@ -422,9 +421,13 @@ void show_stack(struct task_struct *p, u
 			break;
 		if (__get_user(ip, (unsigned long *)(sp + 16)))
 			break;
-		printk("[%016lx] ", ip);
+		printk("[<%016lx>] ", ip);
 		print_symbol("%s\n", ip);
 	} while (count++ < 32);
+#if !CONFIG_KALLSYMS
+	if (count > 0)
+		printk("\n");
+#endif
 }
 
 void dump_stack(void)
