--- ./arch/ppc/Makefile~	2004-06-05 10:12:20.536940417 +0200
+++ ./arch/ppc/Makefile	2004-06-05 10:14:02.342908377 +0200
@@ -90,7 +90,7 @@
 
 ifdef CONFIG_6xx
 # Ensure this is binutils 2.12.1 (or 2.12.90.0.7) or later
-NEW_AS	:= $(shell echo dssall | $(AS) -o /dev/null >/dev/null 2>&1 ; echo $$?)
+NEW_AS	:= $(shell echo dssall | $(AS) -many -o /dev/null >/dev/null 2>&1 ; echo $$?)
 GOODVER	:= 2.12.1
 else
 NEW_AS	:= 0
