diff -purNX /home/olaf/kernel/kernel_exclude.txt linux-2.6.0-test11.orig/drivers/pnp/isapnp/core.c linux-2.6.0-test11.SuSE/drivers/pnp/isapnp/core.c
--- linux-2.6.0-test11.orig/drivers/pnp/isapnp/core.c	2003-11-26 20:44:19.000000000 +0000
+++ linux-2.6.0-test11.SuSE/drivers/pnp/isapnp/core.c	2003-11-29 16:24:39.000000000 +0000
@@ -1081,6 +1081,11 @@ int __init isapnp_init(void)
 	struct pnp_card *card;
 	struct pnp_dev *dev;
 
+#if defined(CONFIG_PPC_MULTIPLATFORM)
+#include <asm/processor.h>
+	if(_machine == _MACH_Pmac)
+		isapnp_disable = 1;
+#endif
 	if (isapnp_disable) {
 		isapnp_detected = 0;
 		printk(KERN_INFO "isapnp: ISA Plug & Play support disabled\n");
