diff -p -purNX /suse/olh/kernel/kernel_exclude.txt linux-2.6.3/drivers/char/consolemap.c linux-2.6.3.setfont/drivers/char/consolemap.c
--- linux-2.6.3/drivers/char/consolemap.c	2004-02-18 04:58:48.000000000 +0100
+++ linux-2.6.3.setfont/drivers/char/consolemap.c	2004-03-06 16:51:38.000000000 +0100
@@ -450,6 +450,12 @@ int con_clear_unimap(int con, struct uni
 	struct vc_data *conp = vc_cons[con].d;
   
 	p = (struct uni_pagedir *)*conp->vc_uni_pagedir_loc;
+	/* FIXME */
+	if (!p) {
+		printk(KERN_ERR "p == NULL\n");
+		WARN_ON(1);
+		return -ENODEV;
+	}
 	if (p && p->readonly) return -EIO;
 	if (!p || --p->refcount) {
 		q = (struct uni_pagedir *)kmalloc(sizeof(*p), GFP_KERNEL);
@@ -477,6 +483,12 @@ con_set_unimap(int con, ushort ct, struc
 	struct vc_data *conp = vc_cons[con].d;
 
 	p = (struct uni_pagedir *)*conp->vc_uni_pagedir_loc;
+	/* FIXME */
+	if (!p) {
+		printk(KERN_ERR "p == NULL\n");
+		WARN_ON(1);
+		return -ENODEV;
+	}
 	if (p->readonly) return -EIO;
 	
 	if (!ct) return 0;
