From: olh@suse.de
Subject: define virt_to_phys and phys_to_virt


 drivers/macintosh/mol/include/alloc.h |    1 +
 1 files changed, 1 insertion(+)

Index: linux-2.6.12/drivers/macintosh/mol/include/alloc.h
===================================================================
--- linux-2.6.12.orig/drivers/macintosh/mol/include/alloc.h
+++ linux-2.6.12/drivers/macintosh/mol/include/alloc.h
@@ -21,6 +21,7 @@
 #include <linux/vmalloc.h>
 #include <linux/mm.h>
 #include <asm/uaccess.h>
+#include <asm/io.h>
 
 #ifdef LINUX_26
 #include <asm/cacheflush.h>
