Index: linux.t/drivers/char/consolemap.c
===================================================================
--- linux.t.orig/drivers/char/consolemap.c	2004-01-09 01:59:47.000000000 -0500
+++ linux.t/drivers/char/consolemap.c	2004-03-06 15:39:04.000000000 -0500
@@ -477,6 +477,12 @@
 	struct vc_data *conp = vc_cons[con].d;
 
 	p = (struct uni_pagedir *)*conp->vc_uni_pagedir_loc;
+	/* FIXME */
+	if (!p) {
+		printk(KERN_ERR "p == NULL\n");
+		WARN_ON(1);
+		return -ENODEV;
+	}
 	if (p->readonly) return -EIO;
 	
 	if (!ct) return 0;
