
From: Maneesh Soni <maneesh@in.ibm.com>

=> changes in version 0.5
  o Use a new struct sysfs_symlink to save information
    about the symlink name and the pointer to target kobject.
    This was required to accomodate the latest changes in symlink
    code in sysfs which implements readlink and follow_link
    operations.

=> changes in version 0.4
  o Nil, just re-diffed

=> changes in version 0.3
  o Nil, just re-diffed

=> changes in version 0.2
  o symlink name passed to sysfs_create_link() can be destroyed by the
    caller. So, the symlink name should be allocated and copied to the
    corresponding sysfs dirent instead of directly using the given name
    string. The allocated string is freed when the corresponding sysfs_dirent
    is freed through sysfs_put()

=================
  o sysfs_create_link() now does not create a dentry but allocates a
    sysfs_dirent and links it to the parent kobject.

  o sysfs_dirent corresponding to symlink has an array of two string. One of
    them is the name of the symlink and the second one is the target path of the
    symlink



---

 25-akpm/fs/sysfs/symlink.c |   42 +++++++++++++++++++++++++++++-------------
 1 files changed, 29 insertions(+), 13 deletions(-)

diff -puN fs/sysfs/symlink.c~sysfs-leaves-symlink fs/sysfs/symlink.c
--- 25/fs/sysfs/symlink.c~sysfs-leaves-symlink	2004-05-05 19:51:10.019216616 -0700
+++ 25-akpm/fs/sysfs/symlink.c	2004-05-05 19:51:10.023216008 -0700
@@ -13,7 +13,7 @@ static struct inode_operations sysfs_sym
 	.follow_link = sysfs_follow_link,
 };
 
-static int init_symlink(struct inode * inode)
+int init_symlink(struct inode * inode)
 {
 	inode->i_op = &sysfs_symlink_inode_operations;
 	return 0;
@@ -53,6 +53,30 @@ static void fill_object_path(struct kobj
 	}
 }
 
+static int sysfs_add_link(struct sysfs_dirent * parent_sd, char * name,
+			    struct kobject * target)
+{
+	struct sysfs_dirent * sd;
+	struct sysfs_symlink * sl;
+
+	sl = kmalloc(sizeof(*sl), GFP_KERNEL);
+	if (!sl)
+		return -ENOMEM;
+
+	sl->link_name = kmalloc(strlen(name) + 1, GFP_KERNEL);
+	strcpy(sl->link_name, name);
+	sl->target_kobj = kobject_get(target);
+
+	sd = sysfs_new_dirent(parent_sd, sl, SYSFS_KOBJ_LINK);
+	if (sd) {
+		sd->s_mode = S_IFLNK|S_IRWXUGO;
+		return 0;
+	}
+
+	kfree(sl);
+	return -ENOMEM;
+}
+
 /**
  *	sysfs_create_link - create symlink between two objects.
  *	@kobj:	object whose directory we're creating the link in.
@@ -62,21 +86,13 @@ static void fill_object_path(struct kobj
 int sysfs_create_link(struct kobject * kobj, struct kobject * target, char * name)
 {
 	struct dentry * dentry = kobj->dentry;
-	struct dentry * d;
 	int error = 0;
 
+	if (!name)
+		return -EINVAL;
+
 	down(&dentry->d_inode->i_sem);
-	d = sysfs_get_dentry(dentry,name);
-	if (!IS_ERR(d)) {
-		error = sysfs_create(d, S_IFLNK|S_IRWXUGO, init_symlink);
-		if (!error)
-			/*
-			 * associate the link dentry with the target kobject
-			 */
-			d->d_fsdata = kobject_get(target);
-		dput(d);
-	} else
-		error = PTR_ERR(d);
+	error = sysfs_add_link(dentry->d_fsdata, name, target);
 	up(&dentry->d_inode->i_sem);
 	return error;
 }

_
