From: Andy Grover <agrover@redhat.com>
Date: Thu, 19 Jan 2012 13:39:12 -0800
Subject: [PATCH] target: fix comment typos
Git-commit: 644df699d598c12989238b3f14fdabf78c6076ab
References: FATE#313550
Patch-Mainline: v3.4

Signed-off-by: Andy Grover <agrover@redhat.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 include/target/target_core_base.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/target/target_core_base.h b/include/target/target_core_base.h
index 5b7d5f3..f609af4 100644
--- a/include/target/target_core_base.h
+++ b/include/target/target_core_base.h
@@ -593,12 +593,12 @@ struct se_cmd {
 };
 
 struct se_tmr_req {
-	/* Task Management function to be preformed */
+	/* Task Management function to be performed */
 	u8			function;
 	/* Task Management response to send */
 	u8			response;
 	int			call_transport;
-	/* Reference to ITT that Task Mgmt should be preformed */
+	/* Reference to ITT that Task Mgmt should be performed */
 	u32			ref_task_tag;
 	/* 64-bit encoded SAM LUN from $FABRIC_MOD TMR header */
 	u64			ref_task_lun;
-- 
1.7.4.2

