From: Olaf Hering <ohering@suse.de>
Subject: TCM_RBD depends on BLK_DEV_RBD
Patch-mainline: Not yet, SES2 clustered LIO/RBD
References: fate#318836


--- a/drivers/target/Kconfig
+++ b/drivers/target/Kconfig
@@ -42,6 +42,7 @@ config TCM_USER2
 
 config TCM_RBD
 	tristate "TCM/RBD Subsystem Plugin for Linux/RBD"
+	depends on BLK_DEV_RBD
 	help
 	Say Y here to enable the TCM/RBD subsystem plugin for Ceph RBD
 	access.
