From: Christoph Hellwig <hch@lst.de>
Date: Tue, 13 Sep 2011 23:08:11 +0200
Subject: [PATCH] target: don't opencode transport_release_cmd in transport_release_fe_cmd
Git-commit: 31afc39c0c93edec5a117371f1bd2a264cceafac
References: FATE#313550
Patch-Mainline: v3.2

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/target/target_core_transport.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/drivers/target/target_core_transport.c b/drivers/target/target_core_transport.c
index 174c6ae..0bd437c 100644
--- a/drivers/target/target_core_transport.c
+++ b/drivers/target/target_core_transport.c
@@ -3804,8 +3804,7 @@ static void transport_release_fe_cmd(struct se_cmd *cmd)
 	transport_release_tasks(cmd);
 free_pages:
 	transport_free_pages(cmd);
-	transport_free_se_cmd(cmd);
-	cmd->se_tfo->release_cmd(cmd);
+	transport_release_cmd(cmd);
 }
 
 static int
-- 
1.7.4.2

