From: Roland Dreier <roland@purestorage.com>
Date: Mon, 16 Jul 2012 15:17:12 -0700
Subject: [PATCH] target: Make unnecessarily global se_dev_align_max_sectors() static
Git-commit: c8045372dd2d4784bc4cb32ddc27f4b88ff2140c
References: FATE#313550
Patch-Mainline: v3.6

Signed-off-by: Roland Dreier <roland@purestorage.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/target/target_core_device.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/target/target_core_device.c b/drivers/target/target_core_device.c
index 06be55c..1591e06 100644
--- a/drivers/target/target_core_device.c
+++ b/drivers/target/target_core_device.c
@@ -840,7 +840,7 @@ int se_dev_check_shutdown(struct se_device *dev)
 	return ret;
 }
 
-u32 se_dev_align_max_sectors(u32 max_sectors, u32 block_size)
+static u32 se_dev_align_max_sectors(u32 max_sectors, u32 block_size)
 {
 	u32 tmp, aligned_max_sectors;
 	/*
-- 
1.7.4.2

