From: Jesper Juhl <jj@chaosbits.net>
Date: Wed, 22 Jun 2011 01:55:50 -0700
Subject: loopback: Remove duplicate scsi/scsi_tcq.h include
Git-commit: 824cc5ff37a0dc3ec8c9e7002e3b1ca0b408917e
References: bnc#718950
Patch-Mainline: 3.1-rc1

Signed-off-by: Jesper Juhl <jj@chaosbits.net>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 drivers/target/loopback/tcm_loop.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/drivers/target/loopback/tcm_loop.c b/drivers/target/loopback/tcm_loop.c
index 083d6c5..e53cec6 100644
--- a/drivers/target/loopback/tcm_loop.c
+++ b/drivers/target/loopback/tcm_loop.c
@@ -31,7 +31,6 @@
 #include <scsi/scsi_host.h>
 #include <scsi/scsi_device.h>
 #include <scsi/scsi_cmnd.h>
-#include <scsi/scsi_tcq.h>
 
 #include <target/target_core_base.h>
 #include <target/target_core_transport.h>
-- 
1.6.0.2

