From: Christoph Hellwig <hch@infradead.org>
Date: Mon, 14 Nov 2011 11:36:31 -0500
Subject: [PATCH] target: remove the unused t_task_pt_sgl and t_task_pt_sgl_num se_cmd fields
Git-commit: 6fd126ffebef3897d8fca98644a9fd1cc5c7a5e3
References: FATE#313550
Patch-Mainline: v3.2

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 include/target/target_core_base.h |    8 --------
 1 files changed, 0 insertions(+), 8 deletions(-)

diff --git a/include/target/target_core_base.h b/include/target/target_core_base.h
index d1bfa99..df210bd 100644
--- a/include/target/target_core_base.h
+++ b/include/target/target_core_base.h
@@ -485,14 +485,6 @@ struct se_cmd {
 
 	struct work_struct	work;
 
-	/*
-	 * Used for pre-registered fabric SGL passthrough WRITE and READ
-	 * with the special SCF_PASSTHROUGH_CONTIG_TO_SG case for TCM_Loop
-	 * and other HW target mode fabric modules.
-	 */
-	struct scatterlist	*t_task_pt_sgl;
-	u32			t_task_pt_sgl_num;
-
 	struct scatterlist	*t_data_sg;
 	unsigned int		t_data_nents;
 	struct scatterlist	*t_bidi_data_sg;
-- 
1.7.4.2

