From: Andy Grover <agrover@redhat.com>
Date: Thu, 19 Jan 2012 13:39:13 -0800
Subject: [PATCH] target: Remove unused struct se_queue_req
Git-commit: ef28640497f5935bf614e1e29b16972d8b97cdb2
References: FATE#313550
Patch-Mainline: v3.4

Signed-off-by: Andy Grover <agrover@redhat.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 include/target/target_core_base.h |    6 ------
 1 files changed, 0 insertions(+), 6 deletions(-)

diff --git a/include/target/target_core_base.h b/include/target/target_core_base.h
index f609af4..41dd91b 100644
--- a/include/target/target_core_base.h
+++ b/include/target/target_core_base.h
@@ -474,12 +474,6 @@ struct t10_reservation {
 	struct t10_reservation_ops pr_ops;
 };
 
-struct se_queue_req {
-	int			state;
-	struct se_cmd		*cmd;
-	struct list_head	qr_list;
-};
-
 struct se_queue_obj {
 	atomic_t		queue_cnt;
 	spinlock_t		cmd_queue_lock;
-- 
1.7.4.2

