From: Christoph Hellwig <hch@infradead.org>
Date: Mon, 17 Oct 2011 13:56:52 -0400
Subject: [PATCH] target: remove unused TRANSPORT_ states
Git-commit: 59aaad1ec44d9a77c32b873b001f31c5af47fcc7
References: FATE#313550
Patch-Mainline: v3.2

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 include/target/target_core_base.h |    4 ----
 1 files changed, 0 insertions(+), 4 deletions(-)

diff --git a/include/target/target_core_base.h b/include/target/target_core_base.h
index ff005b4..2eb7ef1 100644
--- a/include/target/target_core_base.h
+++ b/include/target/target_core_base.h
@@ -91,11 +91,7 @@ enum transport_state_table {
 	TRANSPORT_COMPLETE_FAILURE = 7,
 	TRANSPORT_COMPLETE_TIMEOUT = 8,
 	TRANSPORT_PROCESS_TMR	= 9,
-	TRANSPORT_TMR_COMPLETE	= 10,
 	TRANSPORT_ISTATE_PROCESSING = 11,
-	TRANSPORT_ISTATE_PROCESSED = 12,
-	TRANSPORT_KILL		= 13,
-	TRANSPORT_FREE		= 15,
 	TRANSPORT_NEW_CMD_MAP	= 16,
 	TRANSPORT_FREE_CMD_INTR = 17,
 	TRANSPORT_COMPLETE_QF_WP = 18,
-- 
1.7.4.2

