From: Stephen Hemminger <shemminger@vyatta.com>
Date: Fri, 7 Sep 2012 09:33:15 -0700
Subject: [PATCH 087/105] netdev: make pci_error_handlers const
Patch-mainline: v3.7-rc1
Git-commit: 3646f0e5c97703cecfd96cdabb085e78d9b7f680 (partial)
References: bnc#790588 FATE#313912

Signed-off-by: Stephen Hemminger <shemminger@vyatta.com>
Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
Acked-by: Nithin Nayak Sujir <nsujir@broadcom.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/broadcom/tg3.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -16524,7 +16524,7 @@ done:
 	rtnl_unlock();
 }
 
-static struct pci_error_handlers tg3_err_handler = {
+static const struct pci_error_handlers tg3_err_handler = {
 	.error_detected	= tg3_io_error_detected,
 	.slot_reset	= tg3_io_slot_reset,
 	.resume		= tg3_io_resume
