From: Nithin Sujir <nsujir@broadcom.com>
Date: Fri, 20 Sep 2013 16:46:59 -0700
Subject: [PATCH 069/108] tg3: Remove redundant if check
Patch-mainline: v3.13-rc1
Git-commit: 4bc814ab71951cab2dd745c08772ed2eb849a5f8
References: bsc#908458 FATE#317507


Signed-off-by: Nithin Nayak Sujir <nsujir@broadcom.com>
Signed-off-by: Michael Chan <mchan@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Cho, Yu-Chen <acho@suse.com>
---
 drivers/net/ethernet/broadcom/tg3.c |    5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -9132,10 +9132,7 @@ static int tg3_halt(struct tg3 *tp, int
 		memset(tp->hw_stats, 0, sizeof(struct tg3_hw_stats));
 	}
 
-	if (err)
-		return err;
-
-	return 0;
+	return err;
 }
 
 static int tg3_set_mac_addr(struct net_device *dev, void *p)
