From: Brandon Philips <bphilips@suse.de>
Subject: [PATCH] tg3: entropy source
Patch-mainline: never
References: FATE#307517

Signed-off-by: Brandon Philips <bphilips@suse.de>

---
 drivers/net/tg3.c |   13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

--- a/drivers/net/tg3.c
+++ b/drivers/net/tg3.c
@@ -15,7 +15,6 @@
  *	notice is accompanying it.
  */
 
-
 #include <linux/module.h>
 #include <linux/moduleparam.h>
 #include <linux/stringify.h>
@@ -67,6 +66,10 @@
 #define tg3_flag_clear(tp, flag)			\
 	_tg3_flag_clear(TG3_FLAG_##flag, (tp)->tg3_flags)
 
+static int entropy = 0;
+module_param(entropy, int, 0);
+MODULE_PARM_DESC(entropy, "Allow tg3 to populate the /dev/random entropy pool");
+
 #define DRV_MODULE_NAME		"tg3"
 #define TG3_MAJ_NUM			3
 #define TG3_MIN_NUM			119
@@ -8590,10 +8593,13 @@ restart_timer:
 static int tg3_request_irq(struct tg3 *tp, int irq_num)
 {
 	irq_handler_t fn;
-	unsigned long flags;
+	unsigned long flags = 0;
 	char *name;
 	struct tg3_napi *tnapi = &tp->napi[irq_num];
 
+	if (entropy)
+		flags = IRQF_SAMPLE_RANDOM;
+
 	if (tp->irq_cnt == 1)
 		name = tp->dev->name;
 	else {
@@ -8606,12 +8612,11 @@ static int tg3_request_irq(struct tg3 *t
 		fn = tg3_msi;
 		if (tg3_flag(tp, 1SHOT_MSI))
 			fn = tg3_msi_1shot;
-		flags = 0;
 	} else {
 		fn = tg3_interrupt;
 		if (tg3_flag(tp, TAGGED_STATUS))
 			fn = tg3_interrupt_tagged;
-		flags = IRQF_SHARED;
+		flags |= IRQF_SHARED;
 	}
 
 	return request_irq(tnapi->irq_vec, fn, flags, name, tnapi);
