From: Nithin Sujir <nsujir@broadcom.com>
Date: Thu, 23 May 2013 11:11:26 +0000
Subject: [PATCH 045/108] tg3: Fix misplaced empty line
Patch-mainline: v3.11-rc1
Git-commit: 7c10ee32f32644c9efbb3cfebf5fb7cea4b48a2c
References: bsc#908458 FATE#317507


Signed-off-by: Nithin Nayak Sujir <nsujir@broadcom.com>
Signed-off-by: Michael Chan <mchan@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Cho, Yu-Chen <acho@suse.com>
---
 drivers/net/ethernet/broadcom/tg3.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -1314,8 +1314,8 @@ static int tg3_phy_toggle_auxctl_smdsp(s
 
 	if (err)
 		return err;
-	if (enable)
 
+	if (enable)
 		val |= MII_TG3_AUXCTL_ACTL_SMDSP_ENA;
 	else
 		val &= ~MII_TG3_AUXCTL_ACTL_SMDSP_ENA;
