Subject: get offsetof() for this userland file
From: olh@suse.de

---
 arch/um/sys-i386/user-offsets.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

Index: linux-2.6.19-rc6/arch/um/sys-i386/user-offsets.c
===================================================================
--- linux-2.6.19-rc6.orig/arch/um/sys-i386/user-offsets.c
+++ linux-2.6.19-rc6/arch/um/sys-i386/user-offsets.c
@@ -2,6 +2,6 @@
 #include <signal.h>
 #include <asm/ptrace.h>
 #include <asm/user.h>
-#include <linux/stddef.h>
+#include <stddef.h>
 #include <sys/poll.h>
 
 #define DEFINE(sym, val) \
