Subject: get offsetof() for this userland file
From: olh@suse.de

---
 arch/um/sys-i386/user-offsets.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

--- linux-2.6.18.orig/arch/um/sys-i386/user-offsets.c
+++ linux-2.6.18/arch/um/sys-i386/user-offsets.c
@@ -2,7 +2,7 @@
 #include <signal.h>
 #include <asm/ptrace.h>
 #include <asm/user.h>
-#include <linux/stddef.h>
+#include <stddef.h>
 #include <sys/poll.h>
 
 #define DEFINE(sym, val) \
