Subject: get offsetof() for this userland file
From: olh@suse.de

---
 arch/um/sys-i386/user-offsets.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

Index: linux-2.6.19-rc6/arch/um/sys-i386/user-offsets.c
===================================================================
--- linux-2.6.19-rc6.orig/arch/um/sys-i386/user-offsets.c
+++ linux-2.6.19-rc6/arch/um/sys-i386/user-offsets.c
@@ -2,8 +2,15 @@
 #include <signal.h>
 #include <asm/ptrace.h>
 #include <asm/user.h>
-#include <linux/stddef.h>
 #include <sys/poll.h>
+#ifndef __KERNEL__
+#define __KERNEL__
+#include <linux/stddef.h>
+#undef __KERNEL__
+#else
+#include <linux/stddef.h>
+#endif
+
 
 #define DEFINE(sym, val) \
 	asm volatile("\n->" #sym " %0 " #val : : "i" (val))
