From 55a46269ccedec140e2487a2344e8a247d48b385 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20Sroczy=C5=84ski?= <msroczyn@gmail.com>
Date: Thu, 25 Aug 2011 19:52:27 +0200
Subject: USB: pl2303: remove unreachable code
Patch-mainline: v3.2
Git-commit: 55a46269ccedec140e2487a2344e8a247d48b385

Signed-off-by: Michal Sroczynski <msroczyn@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

diff --git a/drivers/usb/serial/pl2303.c b/drivers/usb/serial/pl2303.c
index 1d33260..0ae840f 100644
--- a/drivers/usb/serial/pl2303.c
+++ b/drivers/usb/serial/pl2303.c
@@ -360,9 +360,6 @@ static void pl2303_set_termios(struct tty_struct *tty,
 				tmp >>= 2;
 				buf[1] <<= 1;
 			}
-			if (tmp > 256) {
-				tmp %= 256;
-			}
 			buf[0] = tmp;
 		}
 	}
