From: Thomas Renninger <trenn@suse.de>
Subject: Do not modify perf bias performance setting by default at boot
References: bnc#812259
Patch-Mainline: never

Signed-off-by: Thomas Renninger <trenn@suse.de>

Index: linux-3.0-SLE11-SP3/arch/x86/kernel/cpu/intel.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/arch/x86/kernel/cpu/intel.c
+++ linux-3.0-SLE11-SP3/arch/x86/kernel/cpu/intel.c
@@ -472,14 +472,11 @@ static void __cpuinit init_intel(struct
 		u64 epb;
 
 		rdmsrl(MSR_IA32_ENERGY_PERF_BIAS, epb);
-		if ((epb & 0xF) == ENERGY_PERF_BIAS_PERFORMANCE) {
+		if ((epb & 0xF) == ENERGY_PERF_BIAS_PERFORMANCE)
 			printk_once(KERN_WARNING "ENERGY_PERF_BIAS:"
-				" Set to 'normal', was 'performance'\n"
+				" Is set to 'performance'\n"
 				"ENERGY_PERF_BIAS: View and update with"
-				" x86_energy_perf_policy(8)\n");
-			epb = (epb & ~0xF) | ENERGY_PERF_BIAS_NORMAL;
-			wrmsrl(MSR_IA32_ENERGY_PERF_BIAS, epb);
-		}
+				" cpupower-set(8)\n");
 	}
 }
 
