Subject: introduce private definitions of register definitions to preserve kABI
From: Oliver Neukum <oneukum@suse.de>
References: bnc#772427
Patch-Mainline: Never

Do this in a manner compatible to the kABI

Signed-off-by: Oliver Neukum <oneukum@suse.de>

--- linux-3.0-SLE11-SP2/drivers/usb/host/xhci.h.alt	2012-08-10 13:50:49.000000000 +0200
+++ linux-3.0-SLE11-SP2/drivers/usb/host/xhci.h	2012-08-10 13:52:07.000000000 +0200
@@ -1777,4 +1777,7 @@ struct xhci_input_control_ctx *xhci_get_
 struct xhci_slot_ctx *xhci_get_slot_ctx(struct xhci_hcd *xhci, struct xhci_container_ctx *ctx);
 struct xhci_ep_ctx *xhci_get_ep_ctx(struct xhci_hcd *xhci, struct xhci_container_ctx *ctx, unsigned int ep_index);
 
+#define USB_INTEL_USB3_PSSEN   0xD8
+#define USB_INTEL_XUSB2PR      0xD0
+
 #endif /* __LINUX_XHCI_HCD_H */
