From: Daniel Wagner <dwagner@suse.de>
Date: Tue, 19 Jul 2022 16:11:13 +0200
Subject: blk-mq: fix kabi support concurrent queue quiesce unquiesce
Patch-mainline: never, kABI fix
References: bsc#1201651

---
 include/linux/blkdev.h |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -561,8 +561,6 @@ struct request_queue {
 	 */
 	struct mutex		mq_freeze_lock;
 
-	int			quiesce_depth;
-
 	struct blk_mq_tag_set	*tag_set;
 	struct list_head	tag_set_list;
 	struct bio_set		bio_split;
@@ -580,6 +578,10 @@ struct request_queue {
 
 #define BLK_MAX_WRITE_HINTS	5
 	u64			write_hints[BLK_MAX_WRITE_HINTS];
+
+#ifndef __GENKSYMS__
+	int			quiesce_depth;
+#endif /* !__GENKSYMS__ */
 };
 
 /* Keep blk_queue_flag_name[] in sync with the definitions below */
