From: Daniel Wagner <dwagner@suse.de>
Date: Fri, 13 Aug 2021 10:19:51 +0200
Subject: blk-cgroup: kABI fixes for new fc_app_id definition
Patch-Mainline: never, kABI fix for SLE15-SP2
References: bsc#1189385 jsc#SLE-18970

Signed-off-by: Daniel Wagner <dwagner@suse.de>
---
 include/linux/blk-cgroup.h |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

--- a/include/linux/blk-cgroup.h
+++ b/include/linux/blk-cgroup.h
@@ -57,13 +57,15 @@ struct blkcg {
 	struct blkcg_policy_data	*cpd[BLKCG_MAX_POLS];
 
 	struct list_head		all_blkcgs_node;
-#ifdef CONFIG_BLK_CGROUP_FC_APPID
-	char                            fc_app_id[FC_APPID_LEN];
-#endif
 #ifdef CONFIG_CGROUP_WRITEBACK
 	struct list_head		cgwb_list;
 	refcount_t			cgwb_refcnt;
 #endif
+#ifdef CONFIG_BLK_CGROUP_FC_APPID
+#ifndef __GENKSYMS__
+	char                            fc_app_id[FC_APPID_LEN];
+#endif	
+#endif
 };
 
 /*
