From b2e4908819fc5acae90637bc3bbd5e6f2a011889 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 3 Jun 2025 16:15:47 +0200
Subject: [PATCH] generic_pm_domain_data: hide new member in hole
References: git-fixes
Patch-mainline: Never (kABI fixup)

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/pm_domain.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/pm_domain.h b/include/linux/pm_domain.h
index c6716f474..a49c13c7f 100644
--- a/include/linux/pm_domain.h
+++ b/include/linux/pm_domain.h
@@ -251,7 +251,9 @@ struct generic_pm_domain_data {
 	unsigned int default_pstate;
 	unsigned int rpm_pstate;
 	bool hw_mode;
+#ifndef __GENKSYSMS__
 	bool rpm_always_on;
+#endif
 	void *data;
 };
 
-- 
2.49.0

