From: Davide Benini <davide.benini@suse.com>
Date: Thu Jun  6 17:40:00 CEST 2024
Patch-mainline: Never, kABI workaround
References: CVE-2024-35857 bsc#1224619
Subject: [PATCH] Kabi workaround for icmp: prevent possible NULL dereferences from
icmp_build_probe()

Signed-off-by: Davide Benini <davide.benini@suse.com>
Signed-off-by: Pedro Falcato <pfalcato@suse.de>
---
 net/ipv4/icmp.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/net/ipv4/icmp.c
+++ b/net/ipv4/icmp.c
@@ -92,8 +92,10 @@
 #include <net/inet_common.h>
 #include <net/ip_fib.h>
 #include <net/l3mdev.h>
+#ifndef __GENKSYMS__
 #include <net/addrconf.h>
 #include <net/inet_dscp.h>
+#endif
 /*
  *	Build xmit assembly blocks
  */
