From: Gabriel Krisman Bertazi <krisman@suse.de>
Subject: io_uring: Re-add dummy_ubuf for kABI purposes
Patch-mainline: Never, kabi
References: git-fixes

---
 include/linux/io_uring_types.h |    9 +++++++++
 1 file changed, 9 insertions(+)

--- a/include/linux/io_uring_types.h
+++ b/include/linux/io_uring_types.h
@@ -339,6 +339,15 @@ struct io_ring_ctx {
 	struct wait_queue_head		poll_wq;
 	struct io_restriction		restrictions;
 
+	/*
+	 * ->dummy_ubuf is no longer necessary.  Preserve it for kABI
+	 * purposes, but annotate the unused field for future use.
+	 */
+#ifdef __GENKSYMS__
+	struct io_mapped_ubuf		*dummy_ubuf;
+#else
+	void				*unused1;
+#endif
 	/* slow path rsrc auxilary data, used by update/register */
 	struct io_rsrc_data		*file_data;
 	struct io_rsrc_data		*buf_data;
