From: Gabriel Krisman Bertazi <krisman@suse.de>
Subject: io_uring: Handle kABI for rename of is_mapped
Patch-mainline: Never, kabi
References: git-fixes

---
 io_uring/kbuf.h |    4 ++++
 1 file changed, 4 insertions(+)

--- a/io_uring/kbuf.h
+++ b/io_uring/kbuf.h
@@ -28,7 +28,11 @@ struct io_buffer_list {
 	atomic_t refs;
 
 	/* ring mapped provided buffers */
+#ifdef __GENKSYMS__
+	__u8 is_mapped;
+#else
 	__u8 is_buf_ring;
+#endif
 
 	/* ring mapped provided buffers, but mmap'ed by application */
 	__u8 is_mmap;
