From: Pedro Falcato <pfalcato@suse.de>
Date: Mon May 5 12:53:58 2025 +0100
Patch-mainline: Never, kABI workaround
References: git-fixes
Subject: [PATCH] Kabi workaround for icmp: Unmask upper DSCP bits in icmp_route_lookup()
Signed-off-by: Pedro Falcato <pfalcato@suse.de>
---
 net/ipv4/icmp.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/net/ipv4/icmp.c
+++ b/net/ipv4/icmp.c
@@ -92,7 +92,9 @@
 #include <net/ip_fib.h>
 #include <net/l3mdev.h>
 #include <net/addrconf.h>
+#ifndef __GENKSYMS__
 #include <net/inet_dscp.h>
+#endif
 /*
  *	Build xmit assembly blocks
  */
