From: Thomas Abraham <tabraham@suse.com>
Subject: [PATCH] ipvs: Fix up kabi for expire_nodest_conn_work addition
References: bsc#1190467
Patch-mainline: Never, kabi fixup

Signed-off-by: Thomas Abraham <tabraham@suse.com>

---
 include/net/ip_vs.h | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/include/net/ip_vs.h b/include/net/ip_vs.h
index ae6963becca9..0362ea33983e 100644
--- a/include/net/ip_vs.h
+++ b/include/net/ip_vs.h
@@ -886,8 +886,6 @@ struct netns_ipvs {
 	atomic_t		conn_out_counter;
 
 #ifdef CONFIG_SYSCTL
-	/* delayed work for expiring no dest connections */
-	struct delayed_work	expire_nodest_conn_work;
 	/* 1/rate drop and drop-entry variables */
 	struct delayed_work	defense_work;   /* Work handler */
 	int			drop_rate;
@@ -963,6 +961,13 @@ struct netns_ipvs {
 	 * are not supported when synchronization is enabled.
 	 */
 	unsigned int		mixed_address_family_dests;
+
+#ifndef __GENKSYMS__
+#ifdef CONFIG_SYSCTL
+	/* delayed work for expiring no dest connections */
+	struct delayed_work	expire_nodest_conn_work;
+#endif
+#endif
 };
 
 #define DEFAULT_SYNC_THRESHOLD	3
-- 
2.26.2

