From: Jan Kara <jack@suse.cz>
Subject: jbd2: Move j_transaction_overhead_buffers into a hole
Patch-mainline: Never, kABI
References: bsc#1234042

To maintain journal_t kABI, move added j_transaction_overhead_buffers into
a hole.

Signed-off-by: Jan Kara <jack@suse.cz>

---
 include/linux/jbd2.h |   16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

--- a/include/linux/jbd2.h
+++ b/include/linux/jbd2.h
@@ -1098,13 +1098,6 @@ struct journal_s
 	int			j_revoke_records_per_block;
 
 	/**
-	 * @j_transaction_overhead_buffers:
-	 *
-	 * Number of blocks each transaction needs for its own bookkeeping
-	 */
-	int			j_transaction_overhead_buffers;
-
-	/**
 	 * @j_commit_interval:
 	 *
 	 * What is the maximum transaction lifetime before we begin a commit?
@@ -1168,6 +1161,15 @@ struct journal_s
 	 */
 	pid_t			j_last_sync_writer;
 
+#ifndef __GENKSYMS__
+	/**
+	 * @j_transaction_overhead_buffers:
+	 *
+	 * Number of blocks each transaction needs for its own bookkeeping
+	 */
+	int			j_transaction_overhead_buffers;
+#endif
+
 	/**
 	 * @j_average_commit_time:
 	 *
