From 766093e9de081ca53ecf565cb92ef3f914a4d56f Mon Sep 17 00:00:00 2001
From: "Lee, Chun-Yi" <jlee@suse.com>
Date: Wed, 19 Aug 2020 14:19:08 +0800
Subject: [PATCH] kABI: Fix kABI after EFI_RT_PROPERTIES table backport
Patch-mainline: Never, kABI fix only
References: bsc#1174029, bsc#1174110, bsc#1174111

Signed-off-by: Lee, Chun-Yi <jlee@suse.com>
---
 include/linux/efi.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/include/linux/efi.h b/include/linux/efi.h
index 8de6936..a8850a2 100644
--- a/include/linux/efi.h
+++ b/include/linux/efi.h
@@ -992,7 +992,6 @@ extern struct efi {
 	efi_system_table_t *systab;	/* EFI system table */
 	unsigned int runtime_version;	/* Runtime services version */
 	unsigned long mps;		/* MPS table */
-	unsigned int  runtime_supported_mask;
 	unsigned long acpi;		/* ACPI table  (IA64 ext 0.71) */
 	unsigned long acpi20;		/* ACPI table  (ACPI 2.0) */
 	unsigned long smbios;		/* SMBIOS table (32 bit entry point) */
@@ -1029,6 +1028,9 @@ extern struct efi {
 	efi_set_virtual_address_map_t *set_virtual_address_map;
 	struct efi_memory_map memmap;
 	unsigned long flags;
+#ifndef __GENKSYMS__
+	unsigned int  runtime_supported_mask;
+#endif
 } efi;
 
 #define EFI_RT_SUPPORTED_GET_TIME 				0x0001
-- 
2.16.4

