From 5ec8920edf266e95d4a60a8a226bb92a002f0713 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Tue, 6 Aug 2024 13:25:50 +0200
Subject: [PATCH] kABI: re-export tpm2_calc_ordinal_duration

References: bsc#1082555
---
 drivers/char/tpm/tpm2-cmd.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/char/tpm/tpm2-cmd.c b/drivers/char/tpm/tpm2-cmd.c
index ea2a646ff9b0..4564197f2feb 100644
--- a/drivers/char/tpm/tpm2-cmd.c
+++ b/drivers/char/tpm/tpm2-cmd.c
@@ -771,6 +771,7 @@ unsigned long tpm2_calc_ordinal_duration(struct tpm_chip *chip, u32 ordinal)
 
 	return duration;
 }
+EXPORT_SYMBOL_GPL(tpm2_calc_ordinal_duration);
 
 /**
  * tpm2_do_selftest() - ensure that all self tests have passed
-- 
2.46.0

