From: Fernando Fernandez Mancera <fmancera@suse.de>
Date: Thu Sep 4 12:58:11 2025 +0200
Subject: kABI fix for "netfilter: nf_tables: Audit log rule reset"
References: git-fixes
Patch-mainline: Never, kABI workaround

Signed-off-by: Fernando Fernandez Mancera <fmancera@suse.de>

---
 include/linux/audit.h |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/include/linux/audit.h
+++ b/include/linux/audit.h
@@ -117,8 +117,10 @@ enum audit_nfcfgop {
 	AUDIT_NFT_OP_OBJ_RESET,
 	AUDIT_NFT_OP_FLOWTABLE_REGISTER,
 	AUDIT_NFT_OP_FLOWTABLE_UNREGISTER,
-	AUDIT_NFT_OP_RULE_RESET,
 	AUDIT_NFT_OP_INVALID,
+#ifndef __GENKSYMS__
+	AUDIT_NFT_OP_RULE_RESET,
+#endif
 };
 
 extern int __init audit_register_class(int class, unsigned *list);
