From a274d096a8df2d450c90eb11461c6364d5c6419a Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Mon, 8 Jul 2024 13:13:54 +0200
Subject: [PATCH] kABI: rtas: Workaround false positive due to lost definition

References: bsc#1227487
---
 arch/powerpc/kernel/rtas.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/arch/powerpc/kernel/rtas.c b/arch/powerpc/kernel/rtas.c
index 9bc31566dba8..1af1f3e70133 100644
--- a/arch/powerpc/kernel/rtas.c
+++ b/arch/powerpc/kernel/rtas.c
@@ -25,6 +25,9 @@
 #include <linux/syscalls.h>
 #include <linux/types.h>
 #include <linux/uaccess.h>
+#ifdef __GENKSYMS__
+#include <linux/irqdomain.h>
+#endif
 
 #include <asm/delay.h>
 #include <asm/firmware.h>
-- 
2.45.2

