From: Joerg Roedel <jroedel@suse.de>
Date: Fri, 24 Jun 2022 11:38:55 +0200
Subject: [PATCH] kABI: Fix kABI after SNP-Guest backport
Patch-mainline: Never, kABI fix
References: jsc#SLE-19924, jsc#SLE-24814

Signed-off-by: Joerg Roedel <jroedel@suse.de>
---
 include/linux/cc_platform.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/cc_platform.h b/include/linux/cc_platform.h
index d08dd65b5c43..cf75ee9c8849 100644
--- a/include/linux/cc_platform.h
+++ b/include/linux/cc_platform.h
@@ -73,6 +73,7 @@ enum cc_attr {
 	 */
 	CC_ATTR_GUEST_UNROLL_STRING_IO,
 
+#ifndef __GENKSYMS__
 	/**
 	 * @CC_ATTR_SEV_SNP: Guest SNP is active.
 	 *
@@ -80,6 +81,7 @@ enum cc_attr {
 	 * using AMD SEV-SNP features.
 	 */
 	CC_ATTR_GUEST_SEV_SNP,
+#endif
 };
 
 #ifdef CONFIG_ARCH_HAS_CC_PLATFORM
-- 
2.36.1

