From: Nikolay Borisov <nik.borisov@suse.com>
Date: Mon, 26 Aug 2024 09:40:03 +0200
Subject: [PATCH] kabi: Preserve kABI when LAM is disabled
Patch-mainline: Never, kabi
References: bsc#1217845
---
 arch/x86/include/asm/mmu.h |    2 --
 1 file changed, 2 deletions(-)

--- a/arch/x86/include/asm/mmu.h
+++ b/arch/x86/include/asm/mmu.h
@@ -46,13 +46,11 @@ typedef struct {
 	unsigned long flags;
 #endif
 
-#ifdef CONFIG_ADDRESS_MASKING
 	/* Active LAM mode:  X86_CR3_LAM_U48 or X86_CR3_LAM_U57 or 0 (disabled) */
 	unsigned long lam_cr3_mask;
 
 	/* Significant bits of the virtual address. Excludes tag bits. */
 	u64 untag_mask;
-#endif
 
 	struct mutex lock;
 	void __user *vdso;			/* vdso base address */
