From: Joerg Roedel <jroedel@suse.de>
Date: Fri, 22 Jan 2021 16:56:17 +0100
Subject: [PATCH] kABI: Fix kABI for extended APIC-ID support
Patch-mainline: Never, kABI fix
References: bsc#1181259, jsc#ECO-3191

Signed-off-by: Joerg Roedel <jroedel@suse.de>
---
 arch/x86/include/asm/io_apic.h |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/arch/x86/include/asm/io_apic.h
+++ b/arch/x86/include/asm/io_apic.h
@@ -77,9 +77,12 @@ struct IO_APIC_route_entry {
 		trigger		:  1,	/* 0: edge, 1: level */
 		mask		:  1,	/* 0: enabled, 1: disabled */
 		__reserved_2	: 15;
-
+#ifndef __GENKSYMS__
 	__u32	__reserved_3	: 17,
 		ext_dest	:  7,
+#else
+	__u32   __reserved_3    : 24,
+#endif
 		dest		:  8;
 } __attribute__ ((packed));
 
