From: Juergen Gross <jgross@suse.com>
Date: Thu, 07 Jul 2022 08:01:13 +0200
Subject: [PATCH] kABI: Fix kABI after KVM: nVMX: handle nested posted interrupts when apicv is disabled for L1
Patch-mainline: Never, kABI fix
References: git-fixes

Signed-off-by: Juergen Gross <jgross@suse.com>
---
 arch/x86/include/asm/kvm_host.h | 4 +
 1 file changed, 4 insertion(+)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index 14fca1cbdd4d..9c636b607cf1 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1113,7 +1113,11 @@ struct kvm_x86_ops {
 	void (*load_eoi_exitmap)(struct kvm_vcpu *vcpu, u64 *eoi_exit_bitmap);
 	void (*set_virtual_apic_mode)(struct kvm_vcpu *vcpu);
 	void (*set_apic_access_page_addr)(struct kvm_vcpu *vcpu, hpa_t hpa);
+#ifdef __GENKSYMS__
+	void (*deliver_posted_interrupt)(struct kvm_vcpu *vcpu, int vector);
+#else
 	int (*deliver_posted_interrupt)(struct kvm_vcpu *vcpu, int vector);
+#endif
 	int (*sync_pir_to_irr)(struct kvm_vcpu *vcpu);
 	int (*set_tss_addr)(struct kvm *kvm, unsigned int addr);
 	int (*set_identity_map_addr)(struct kvm *kvm, u64 ident_addr);
-- 
2.36.1

