From: Juergen Gross <jgross@suse.com>
Date: Wed, 26 Oct 2022 16:29:29 +0200
Subject: [PATCH] kABI: Fix kABI after backport Always set kvm_run->if_flag
Patch-mainline: Never, kABI fix
References: git-fixes

Signed-off-by: Juergen Gross <jgross@suse.com>
---
 arch/x86/include/asm/kvm_host.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index e55e469cdd8c..eb282da0458a 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1320,7 +1320,6 @@ struct kvm_x86_ops {
 	void (*cache_reg)(struct kvm_vcpu *vcpu, enum kvm_reg reg);
 	unsigned long (*get_rflags)(struct kvm_vcpu *vcpu);
 	void (*set_rflags)(struct kvm_vcpu *vcpu, unsigned long rflags);
-	bool (*get_if_flag)(struct kvm_vcpu *vcpu);
 
 	void (*tlb_flush_all)(struct kvm_vcpu *vcpu);
 	void (*tlb_flush_current)(struct kvm_vcpu *vcpu);
@@ -1464,6 +1463,7 @@ struct kvm_x86_ops {
 
 #ifndef __GENKSYMS__
 	void (*guest_memory_reclaimed)(struct kvm *kvm);
+	bool (*get_if_flag)(struct kvm_vcpu *vcpu);
 #endif
 };
 
-- 
2.36.1

