From: Michal Suchanek <msuchanek@suse.de>
Subject: kABI: powerpc/pmem: Include pmem prototypes

References: bsc#1113295 git-fixes
Patch-mainline: never, kABI

Acked-by: Michal Suchanek <msuchanek@suse.de>
---

--- a/arch/powerpc/lib/pmem.c
+++ b/arch/powerpc/lib/pmem.c
@@ -6,7 +6,9 @@
 #include <linux/string.h>
 #include <linux/export.h>
 #include <linux/uaccess.h>
+#ifndef __GENKSYMS__
 #include <linux/libnvdimm.h>
+#endif
 
 #include <asm/cacheflush.h>
 
