From: Joerg Roedel <jroedel@suse.de>
Date: Fri, 16 Apr 2021 16:05:33 +0200
Subject: [PATCH] kABI: Fix kABI caused by fixes for bsc#1174426
Patch-mainline: Never, kABI fix
References: bsc#1174426

Signed-off-by: Joerg Roedel <jroedel@suse.de>
---
 include/linux/pci.h | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/include/linux/pci.h b/include/linux/pci.h
index ccf792ab5bc0..5882ee78ff86 100644
--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -313,10 +313,6 @@ struct pci_dev {
 #ifdef CONFIG_PCIEAER
 	u16		aer_cap;	/* AER capability offset */
 	struct aer_stats *aer_stats;	/* AER stats for this device */
-#endif
-#ifdef CONFIG_PCIEPORTBUS
-	struct rcec_ea	*rcec_ea;	/* RCEC cached endpoint association */
-	struct pci_dev  *rcec;          /* Associated RCEC device */
 #endif
 	u8		pcie_cap;	/* PCIe capability offset */
 	u8		msi_cap;	/* MSI capability offset */
@@ -489,6 +485,10 @@ struct pci_dev {
 	u16		dpc_cap;
 	unsigned int	dpc_rp_extensions:1;
 	u8		dpc_rp_log_size;
+#ifdef CONFIG_PCIEPORTBUS
+	struct rcec_ea	*rcec_ea;	/* RCEC cached endpoint association */
+	struct pci_dev  *rcec;          /* Associated RCEC device */
+#endif
 #endif /* __GENKSYMS__ */
 #endif
 };
-- 
2.25.0

