From: Juergen Gross <jgross@suse.com>
Date: Fri, 06 Dec 2024 13:39:15 +0200
Subject: kABI: kabi fix after vsock/virtio: fix `rx_bytes` accounting
Patch-mainline: Never, kABI fix
References: git-fixes

Signed-off-by: Juergen Gross <jgross@suse.com>
---
--- a/include/linux/virtio_vsock.h
+++ b/include/linux/virtio_vsock.h
@@ -140,9 +140,11 @@ struct virtio_vsock_sock {
 	u32 last_fwd_cnt;
 	u32 rx_bytes;
 	u32 buf_alloc;
-	u32 buf_used;
 	struct sk_buff_head rx_queue;
 	u32 msg_count;
+#ifndef __GENKSYMS__
+	u32 buf_used;
+#endif
 };
 
 struct virtio_vsock_pkt_info {
