From ea77907fa235907fbe2b63b166b1d5331df32c75 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Thu, 8 Aug 2024 18:58:54 +0200
Subject: [PATCH] kABI: tpm-interface: Hide new include from genksyms.

References: bsc#1082555
---
 drivers/char/tpm/tpm-interface.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/char/tpm/tpm-interface.c b/drivers/char/tpm/tpm-interface.c
index 1cb452a44f35..d1cd03b16490 100644
--- a/drivers/char/tpm/tpm-interface.c
+++ b/drivers/char/tpm/tpm-interface.c
@@ -28,7 +28,9 @@
 #include <linux/slab.h>
 #include <linux/mutex.h>
 #include <linux/spinlock.h>
+#ifndef __GENKSYMS__
 #include <linux/suspend.h>
+#endif
 #include <linux/freezer.h>
 
 #include "tpm.h"
-- 
2.46.0

