From: Michal Suchanek <msuchanek@suse.de>
Subject: libnvdimm: cover up nd_pfn_sb changes.

References: bsc#1171759
Patch-mainline: never, kABI
Signed-off-by: Michal Suchanek <msuchanek@suse.de>

--- a/drivers/nvdimm/pfn.h
+++ b/drivers/nvdimm/pfn.h
@@ -41,10 +41,14 @@ struct nd_pfn_sb {
 	/* minor-version-2 record the base alignment of the mapping */
 	__le32 align;
 	/* minor-version-3 guarantee the padding and flags are zero */
+#ifndef __GENKSYMS__
 	/* minor-version-4 record the page size and struct page size */
 	__le32 page_size;
 	__le16 page_struct_size;
 	u8 padding[3994];
+#else
+	u8 padding[4000];
+#endif
 	__le64 checksum;
 };
 
