From: Lee Duncan <lduncan@suse.com>
Date: Sun 11 May 2025 11:51:10 AM PDT
Subject: [PATCH] md: preserve KABI in struct md_personality v2
Patch-mainline: never, kabi
References: git-fixes

Signed-off-by: Lee Duncan <lduncan@suse.com>
---
 drivers/md/md.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/md/md.h
+++ b/drivers/md/md.h
@@ -682,11 +682,11 @@ struct md_personality
 	void *(*takeover) (struct mddev *mddev);
 	/* Changes the consistency policy of an active array. */
 	int (*change_consistency_policy)(struct mddev *mddev, const char *buf);
+#ifndef __GENKSYMS__
+	void (*prepare_suspend) (struct mddev *mddev);
 	/* convert io ranges from array to bitmap */
 	void (*bitmap_sector)(struct mddev *mddev, sector_t *offset,
 			      unsigned long *sectors);
-#ifndef __GENKSYMS__
-	void (*prepare_suspend) (struct mddev *mddev);
 #endif
 };
 
