From: Coly Li <colyli@suse.de>
Date: Mon 25 Mar 2024 06:59:08 +0800
Subject: [PATCH] md: preserve KABI in struct md_personality
Patch-mainline: never, kabi
References: jsc#PED-7542

Signed-off-by: Coly Li <colyli@suse.de>
---
 drivers/md/md.h |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/drivers/md/md.h
+++ b/drivers/md/md.h
@@ -664,7 +664,6 @@ struct md_personality
 	int (*start_reshape) (struct mddev *mddev);
 	void (*finish_reshape) (struct mddev *mddev);
 	void (*update_reshape_pos) (struct mddev *mddev);
-	void (*prepare_suspend) (struct mddev *mddev);
 	/* quiesce suspends or resumes internal processing.
 	 * 1 - stop new actions and wait for action io to complete
 	 * 0 - return to normal behaviour
@@ -682,6 +681,9 @@ struct md_personality
 	void *(*takeover) (struct mddev *mddev);
 	/* Changes the consistency policy of an active array. */
 	int (*change_consistency_policy)(struct mddev *mddev, const char *buf);
+#ifndef __GENKSYMS__
+	void (*prepare_suspend) (struct mddev *mddev);
+#endif
 };
 
 struct md_sysfs_entry {
