From 0494e8216182f9a4a98f6228a5a659ac7200fd0b Mon Sep 17 00:00:00 2001
From: Thomas Abraham <tabraham@suse.com>
Date: Thu, 12 Jan 2023 16:59:26 -0500
Subject: [PATCH] mm: compaction: kABI: avoid pglist_data kABI breakage
References: bsc#1207010
Patch-mainline: never, kABI workaround

---
 include/linux/mmzone.h | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/include/linux/mmzone.h b/include/linux/mmzone.h
index d8acf188b0de..f70817829f88 100644
--- a/include/linux/mmzone.h
+++ b/include/linux/mmzone.h
@@ -863,7 +863,6 @@ typedef struct pglist_data {
 	enum zone_type kcompactd_highest_zoneidx;
 	wait_queue_head_t kcompactd_wait;
 	struct task_struct *kcompactd;
-	bool proactive_compact_trigger;
 #endif
 	/*
 	 * This is a per-node reserve of pages that are not available
@@ -910,6 +909,12 @@ typedef struct pglist_data {
 	/* Per-node vmstats */
 	struct per_cpu_nodestat __percpu *per_cpu_nodestats;
 	atomic_long_t		vm_stat[NR_VM_NODE_STAT_ITEMS];
+
+#ifdef CONFIG_COMPACTION
+#ifndef __GENKSYMS__
+	bool proactive_compact_trigger;
+#endif
+#endif
 } pg_data_t;
 
 #define node_present_pages(nid)	(NODE_DATA(nid)->node_present_pages)
-- 
2.35.3

