From f6fa5aa1b5b916e3d92513070b046c2b2ff43684 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Wed, 28 Dec 2022 15:34:47 +0100
Subject: [PATCH] move new members of struct usbnet to end
Patch-mainline: Never (kABI fixup)
References: git-fixes

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/usb/usbnet.h |    3 +++
 1 file changed, 3 insertions(+)

--- a/include/linux/usb/usbnet.h
+++ b/include/linux/usb/usbnet.h
@@ -84,8 +84,11 @@ struct usbnet {
 #		define EVENT_NO_IP_ALIGN	13
 #		define EVENT_UNPLUG		31
 	void    *suse_kabi_padding;
+
+#ifndef __GENKSYMS__
 	u32			rx_speed;	/* in bps - NOT Mbps */
 	u32			tx_speed;	/* in bps - NOT Mbps */
+#endif
 };
 
 static inline bool usbnet_going_away(struct usbnet *ubn)
