From 65580d15462ab68214e22bc52101140e92b7aa7e Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Thu, 30 Jun 2022 12:33:46 +0200
Subject: [PATCH] move devm_allocate to end of structure for kABI
Patch-mainline: Never (kABI fix)
References: git-fixes

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/spi/spi.h |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/include/linux/spi/spi.h
+++ b/include/linux/spi/spi.h
@@ -501,7 +501,6 @@ struct spi_controller {
 #define SPI_MASTER_GPIO_SS		BIT(5)	/* GPIO CS must select slave */
 
 	/* flag indicating if the allocation of this struct is devres-managed */
-	bool			devm_allocated;
 
 	/* flag indicating this is an SPI slave controller */
 	bool			slave;
@@ -656,6 +655,10 @@ struct spi_controller {
 
 	/* Interrupt enable state during PTP system timestamping */
 	unsigned long		irq_flags;
+#ifndef __GENKSYMS__
+	/* flag indicating if the allocation of this struct is devres-managed */
+	bool			devm_allocated;
+#endif
 };
 
 static inline void *spi_controller_get_devdata(struct spi_controller *ctlr)
