From: Olaf Hering <ohering@suse.de>
Date: Fri, 18 Aug 2023 06:29:12 -0400
Patch-mainline: never, kABI
Subject: kabi: x86/hyperv: Set Virtual Trust Level in VMBus init message
References: bsc#1206453

--- a/include/asm-generic/mshyperv.h
+++ b/include/asm-generic/mshyperv.h
@@ -58,8 +58,10 @@ struct ms_hyperv_info {
 			u32 reserved2 : 20;
 		};
 	};
+#ifndef __GENKSYMS__
+	u8 vtl, _suse_res1, _suse_res2, _suse_res3;
+#endif
 	u64 shared_gpa_boundary;
-	u8 vtl;
 };
 extern struct ms_hyperv_info ms_hyperv;
 extern bool hv_nested;
--- a/include/linux/hyperv.h
+++ b/include/linux/hyperv.h
@@ -661,8 +661,13 @@ struct vmbus_channel_initiate_contact {
 		u64 interrupt_page;
 		struct {
 			u8	msg_sint;
+#ifdef __GENKSYMS__
+			u8      padding1[3];
+			u32     padding2;
+#else
 			u8	msg_vtl;
 			u8	reserved[6];
+#endif
 		};
 	};
 	u64 monitor_page1;
