From: Olaf Hering <ohering@suse.de>
Date: Fri, 18 Aug 2023 06:29:11 -0400
Patch-mainline: never, kABI
Subject: kabi: x86/hyperv: Add sev-snp enlightened guest static key
References: bsc#1206453

--- a/include/asm-generic/mshyperv.h
+++ b/include/asm-generic/mshyperv.h
@@ -36,21 +36,26 @@ struct ms_hyperv_info {
 	u32 nested_features;
 	u32 max_vp_index;
 	u32 max_lp_index;
+#ifdef __GENKSYMS__
+	u32 isolation_config_a;
+#else
 	union {
 		u32 isolation_config_a;
+
 		struct {
 			u32 paravisor_present : 1;
 			u32 reserved_a1 : 31;
 		};
 	};
+#endif
 	union {
 		u32 isolation_config_b;
 		struct {
 			u32 cvm_type : 4;
-			u32 reserved_b1 : 1;
+			u32 reserved1 : 1;
 			u32 shared_gpa_boundary_active : 1;
 			u32 shared_gpa_boundary_bits : 6;
-			u32 reserved_b2 : 20;
+			u32 reserved2 : 20;
 		};
 	};
 	u64 shared_gpa_boundary;
