From: Olaf Hering <ohering@suse.de>
Date: Sun, 26 Mar 2023 06:52:03 -0700
Patch-mainline: never, kABI
Subject: kabi: Drivers: hv: vmbus: Remove second mapping of VMBus monitor pages
References: bsc#1206453

--- a/drivers/hv/hyperv_vmbus.h
+++ b/drivers/hv/hyperv_vmbus.h
@@ -241,6 +241,8 @@ struct vmbus_connection {
 	 * is child->parent notification
 	 */
 	struct hv_monitor_page *monitor_pages[2];
+	void *monitor_pages_original[2];
+	phys_addr_t monitor_pages_pa[2];
 	struct list_head chn_msg_list;
 	spinlock_t channelmsg_lock;
 
