From: Olaf Hering <ohering@suse.de>
Date: Sat, 8 Apr 2023 14:34:41 -0700
Patch-mainline: never, kABI
Subject: kabi: Drivers: hv: vmbus: Remove the per-CPU post_msg_page
References: bsc#1206453

--- a/drivers/hv/hyperv_vmbus.h
+++ b/drivers/hv/hyperv_vmbus.h
@@ -122,6 +122,7 @@ enum {
 struct hv_per_cpu_context {
 	void *synic_message_page;
 	void *synic_event_page;
+	void *post_msg_page; /* kABI */
 
 	/*
 	 * Starting with win8, we can take channel interrupts on any CPU;
