From: Olaf Hering <ohering@suse.de>
Date: Sun, 26 Mar 2023 06:52:01 -0700
Patch-mainline: never, kABI
Subject: kabi: x86/hyperv: Change vTOM handling to use standard coco mechanisms
References: bsc#1206453

--- a/arch/x86/hyperv/ivm.c
+++ b/arch/x86/hyperv/ivm.c
@@ -321,7 +321,7 @@
 	cc_set_mask(ms_hyperv.shared_gpa_boundary);
 	physical_mask &= ms_hyperv.shared_gpa_boundary - 1;
 
-	x86_platform.hyper.is_private_mmio = hv_is_private_mmio;
+	suse_x86_hyper_runtime_is_private_mmio = hv_is_private_mmio;
 	x86_platform.guest.enc_cache_flush_required = hv_vtom_cache_flush_required;
 	x86_platform.guest.enc_tlb_flush_required = hv_vtom_tlb_flush_required;
 	x86_platform.guest.enc_status_change_finish = hv_vtom_set_host_visibility;
--- a/arch/x86/include/asm/coco.h
+++ b/arch/x86/include/asm/coco.h
@@ -7,6 +7,7 @@
 enum cc_vendor {
 	CC_VENDOR_NONE,
 	CC_VENDOR_AMD,
+	CC_VENDOR_HYPERV, /* kABI */
 	CC_VENDOR_INTEL,
 };
 
