From d7e6f5fc9f6341747d0dcb4f952a7e163c82f168 Mon Sep 17 00:00:00 2001
From: Denis Kirjanov <denis.kirjanov@suse.com>
Date: Tue, 27 Aug 2024 16:50:34 +0300
Subject: [PATCH] net: qdisc: preserve kabi for struct QDisc
Patch-mainline: Never, kabi
References: CVE-2024-27010 bsc#1223720
Signed-off-by: Denis Kirjanov <denis.kirjanov@suse.com>
---
 include/net/sch_generic.h | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/include/net/sch_generic.h b/include/net/sch_generic.h
index c83893db1a97..144e364801bd 100644
--- a/include/net/sch_generic.h
+++ b/include/net/sch_generic.h
@@ -106,7 +106,11 @@ struct Qdisc {
 	struct net_rate_estimator __rcu *rate_est;
 	struct gnet_stats_basic_sync __percpu *cpu_bstats;
 	struct gnet_stats_queue	__percpu *cpu_qstats;
+#ifndef __GENKSYMS__
+	int                     owner;
+#else
 	int			pad;
+#endif
 	refcount_t		refcnt;
 
 	/*
@@ -116,7 +120,6 @@ struct Qdisc {
 	struct qdisc_skb_head	q;
 	struct gnet_stats_basic_sync bstats;
 	struct gnet_stats_queue	qstats;
-	int                     owner;
 	unsigned long		state;
 	unsigned long		state2; /* must be written under qdisc spinlock */
 	struct Qdisc            *next_sched;
-- 
2.16.4

