From: Hannes Reinecke <hare@suse.de>
Date: Thu, 28 Jul 2022 09:20:16 +0200
Subject: [PATCH] nvme: kabi fixes for in-band authentication
Patch-Mainline: never, SLE15 kabi fix
References: bsc#1199086

Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 include/linux/nvme.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/include/linux/nvme.h b/include/linux/nvme.h
index 9527f92cd525..8c919e972d40 100644
--- a/include/linux/nvme.h
+++ b/include/linux/nvme.h
@@ -1679,11 +1679,13 @@ struct nvme_command {
 		struct nvmf_connect_command connect;
 		struct nvmf_property_set_command prop_set;
 		struct nvmf_property_get_command prop_get;
-		struct nvmf_auth_common_command auth_common;
 		struct nvmf_auth_send_command auth_send;
 		struct nvmf_auth_receive_command auth_receive;
 		struct nvme_dbbuf dbbuf;
 		struct nvme_directive_cmd directive;
+#ifndef __GENKSYMS__
+		struct nvmf_auth_common_command auth_common;
+#endif
 	};
 };
 
-- 
2.35.3

