From c92178eaae95bc5e45c26791927e9e442dae368b Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Tue, 2 Feb 2021 23:22:24 +0100
Subject: [PATCH] powerpc: kABI: add back suspend_disable_cpu in machdep_calls

References: bsc#1181674 ltc#189159
Patch-mainline: never, kABI

Signed-off-by: Michal Suchanek <msuchanek@suse.de>
---
 arch/powerpc/include/asm/machdep.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/powerpc/include/asm/machdep.h b/arch/powerpc/include/asm/machdep.h
index da8e1cafedf9..c43d6eca9edd 100644
--- a/arch/powerpc/include/asm/machdep.h
+++ b/arch/powerpc/include/asm/machdep.h
@@ -215,6 +215,7 @@ struct machdep_calls {
 	void (*suspend_disable_irqs)(void);
 	void (*suspend_enable_irqs)(void);
 #endif
+	int (*suspend_disable_cpu)(void);
 
 #ifdef CONFIG_ARCH_CPU_PROBE_RELEASE
 	ssize_t (*cpu_probe)(const char *, size_t);
-- 
2.26.2

