From f762c1c4a7452e8684db3a59fc856519d72f6f18 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Wed, 14 Dec 2022 22:33:20 +0100
Subject: [PATCH] restore m_can_lec_type
Patch-mainline: Never (kABI fixup)
References: git-fixes

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 drivers/net/can/m_can/m_can.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/can/m_can/m_can.h b/drivers/net/can/m_can/m_can.h
index f1f28d634..4f70fc8f1 100644
--- a/drivers/net/can/m_can/m_can.h
+++ b/drivers/net/can/m_can/m_can.h
@@ -38,7 +38,11 @@ enum m_can_lec_type {
 	LEC_BIT1_ERROR,
 	LEC_BIT0_ERROR,
 	LEC_CRC_ERROR,
+#ifdef __GENKSYMS__
+	LEC_UNUSED,
+#else
 	LEC_NO_CHANGE,
+#endif
 };
 
 enum m_can_mram_cfg {
-- 
2.35.3

