From: Liang Yan <lyan@suse.com>
Date: Mon, 23 Mar 2020 10:29:12 -0400
Subject: Revert sgi_iodev for kabi break
Patch-mainline: Never, kabi
References: bsc#1166547
 
Signed-off-by: Liang Yan <lyan@suse.com>
---
 include/kvm/arm_vgic.h |    1 +
 1 file changed, 1 insertion(+)

--- a/include/kvm/arm_vgic.h
+++ b/include/kvm/arm_vgic.h
@@ -311,6 +311,7 @@ struct vgic_cpu {
 	 * parts of the redistributor.
 	 */
 	struct vgic_io_device	rd_iodev;
+	struct vgic_io_device	sgi_iodev;
 	struct vgic_redist_region *rdreg;
 
 	/* Contains the attributes and gpa of the LPI pending tables. */
