From: Denis Kirjanov <dkirjanov@suse.com>
Date: Thu, 27 Aug 2020 20:52:44 +0200
Subject: scsi/fc: kABI fixes for new ELS_EDC, ELS_RDP definition
Patch-Mainline: never, kABI fix for SLE15-SP3
References: bsc#1171688 bsc#1174003 bsc#1190576

Signed-off-by: Denis Kirjanov <dkirjanov@suse.com>
---
 include/uapi/scsi/fc/fc_els.h |    4 ++++
 1 file changed, 4 insertions(+)
--- a/include/uapi/scsi/fc/fc_els.h
+++ b/include/uapi/scsi/fc/fc_els.h
@@ -41,7 +41,9 @@ enum fc_els_cmd {
 	ELS_REC =	0x13,	/* read exchange concise */
 	ELS_SRR =	0x14,	/* sequence retransmission request */
 	ELS_FPIN =	0x16,	/* Fabric Performance Impact Notification */
+#ifndef __GENKSYMS__
 	ELS_EDC =	0x17,	/* Exchange Diagnostic Capabilities */
+#endif
 	ELS_RDP =	0x18,	/* Read Diagnostic Parameters */
 	ELS_RDF =	0x19,	/* Register Diagnostic Functions */
 	ELS_PRLI =	0x20,	/* process login */
@@ -112,7 +114,9 @@ enum fc_els_cmd {
 	[ELS_REC] =	"REC",			\
 	[ELS_SRR] =	"SRR",			\
 	[ELS_FPIN] =	"FPIN",			\
+#ifndef __GENKSYMS__				\
 	[ELS_EDC] =	"EDC",			\
+#endif						\
 	[ELS_RDP] =	"RDP",			\
 	[ELS_RDF] =	"RDF",			\
 	[ELS_PRLI] =	"PRLI",			\

