From d8bcc1c7dc9486466f5b5707bc4fbd73f5572ffe Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Wed, 21 Dec 2022 20:04:49 +0100
Subject: [PATCH] struct usbnet: move new members to end
Patch-mainline: Never (kABI fixup)
References: git-fixes

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 include/linux/usb/usbnet.h | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/include/linux/usb/usbnet.h b/include/linux/usb/usbnet.h
index b801cd41e..8f21ca36a 100644
--- a/include/linux/usb/usbnet.h
+++ b/include/linux/usb/usbnet.h
@@ -83,10 +83,13 @@ struct usbnet {
 #		define EVENT_LINK_CHANGE	11
 #		define EVENT_SET_RX_MODE	12
 #		define EVENT_NO_IP_ALIGN	13
-	u32			rx_speed;	/* in bps - NOT Mbps */
-	u32			tx_speed;	/* in bps - NOT Mbps */
 
 	void *suse_kabi_padding;
+
+#ifndef __GENKSYMS__
+	u32			rx_speed;	/* in bps - NOT Mbps */
+	u32			tx_speed;	/* in bps - NOT Mbps */
+#endif
 };
 
 static inline struct usb_driver *driver_of(struct usb_interface *intf)
-- 
2.35.3

