From ca22149d7f95989a2f29a289bb9c339db9a29d46 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 7 Mar 2023 13:38:30 +0100
Subject: [PATCH] struct uvc_device move flush_status  new member to end
Patch-mainline: Never (kABI fixup)
References: git-fixes

Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 drivers/media/usb/uvc/uvcvideo.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/media/usb/uvc/uvcvideo.h b/drivers/media/usb/uvc/uvcvideo.h
index 74eade782..73dccb5c7 100644
--- a/drivers/media/usb/uvc/uvcvideo.h
+++ b/drivers/media/usb/uvc/uvcvideo.h
@@ -698,7 +698,6 @@ struct uvc_device {
 	struct usb_host_endpoint *int_ep;
 	struct urb *int_urb;
 	u8 *status;
-	bool flush_status;
 	struct input_dev *input;
 	char input_phys[64];
 
@@ -711,6 +710,9 @@ struct uvc_device {
 	} async_ctrl;
 
 	struct uvc_entity *gpio_unit;
+#ifndef __GENKSYMS__
+	bool flush_status;
+#endif
 };
 
 enum uvc_handle_state {
-- 
2.39.2

