From c8fbb2a6ccb47e0db824ea6e923f8eb431c20d3c Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.com>
Date: Tue, 27 May 2025 13:56:13 +0200
Subject: [PATCH] struct typec_port: move nre mutex to end
Patch-mainline: never (kABI fixup)
References: bsc#1242856 CVE-2025-37809

---
 drivers/usb/typec/class.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/typec/class.h b/drivers/usb/typec/class.h
index 0b96652f9..8503e5676 100644
--- a/drivers/usb/typec/class.h
+++ b/drivers/usb/typec/class.h
@@ -59,7 +59,6 @@ struct typec_port {
 	enum typec_pwr_opmode		pwr_opmode;
 	enum typec_port_type		port_type;
 	struct mutex			port_type_lock;
-	struct mutex			partner_link_lock;
 
 	enum typec_orientation		orientation;
 	struct typec_switch		*sw;
@@ -82,6 +81,9 @@ struct typec_port {
 	struct device			*usb3_dev;
 
 	void				*suse_kabi_padding;
+#ifndef __GENKSYMS__
+	struct mutex			partner_link_lock;
+#endif
 };
 
 #define to_typec_port(_dev_) container_of(_dev_, struct typec_port, dev)
-- 
2.49.0

