From: Olaf Hering <ohering@suse.de>
Date: Wed, 8 Mar 2023 18:40:02 -0800
Patch-mainline: never, kABI
Subject: kabi: x86/ioremap: Add hypervisor callback for private MMIO mapping in coco
References: bsc#1206453


--- a/arch/x86/include/asm/x86_init.h
+++ b/arch/x86/include/asm/x86_init.h
@@ -267,8 +267,8 @@ struct x86_hyper_runtime {
 	void (*pin_vcpu)(int cpu);
 	void (*sev_es_hcall_prepare)(struct ghcb *ghcb, struct pt_regs *regs);
 	bool (*sev_es_hcall_finish)(struct ghcb *ghcb, struct pt_regs *regs);
-	bool (*is_private_mmio)(u64 addr);
 };
+extern bool (*suse_x86_hyper_runtime_is_private_mmio)(u64 addr); /* kABI is_private_mmio */
 
 /**
  * struct x86_platform_ops - platform specific runtime functions
--- a/arch/x86/kernel/apic/io_apic.c
+++ b/arch/x86/kernel/apic/io_apic.c
@@ -2688,7 +2688,7 @@ static void io_apic_set_fixmap(enum fixe
 	 * bits, just like normal ioremap():
 	 */
 	if (cc_platform_has(CC_ATTR_GUEST_MEM_ENCRYPT)) {
-		if (x86_platform.hyper.is_private_mmio(phys))
+		if (suse_x86_hyper_runtime_is_private_mmio(phys))
 			flags = pgprot_encrypted(flags);
 		else
 			flags = pgprot_decrypted(flags);
--- a/arch/x86/kernel/x86_init.c
+++ b/arch/x86/kernel/x86_init.c
@@ -136,6 +136,8 @@ static bool enc_tlb_flush_required_noop(
 static bool enc_cache_flush_required_noop(void) { return false; }
 static bool is_private_mmio_noop(u64 addr) {return false; }
 
+bool (*suse_x86_hyper_runtime_is_private_mmio)(u64 addr) = is_private_mmio_noop;
+EXPORT_SYMBOL_GPL(suse_x86_hyper_runtime_is_private_mmio);
 struct x86_platform_ops x86_platform __ro_after_init = {
 	.calibrate_cpu			= native_calibrate_cpu_early,
 	.calibrate_tsc			= native_calibrate_tsc,
@@ -150,7 +152,6 @@ struct x86_platform_ops x86_platform __r
 	.realmode_reserve		= reserve_real_mode,
 	.realmode_init			= init_real_mode,
 	.hyper.pin_vcpu			= x86_op_int_noop,
-	.hyper.is_private_mmio		= is_private_mmio_noop,
 
 	.guest = {
 		.enc_status_change_prepare = enc_status_change_prepare_noop,
--- a/arch/x86/mm/ioremap.c
+++ b/arch/x86/mm/ioremap.c
@@ -115,7 +115,7 @@ static void __ioremap_check_other(resour
 	if (!cc_platform_has(CC_ATTR_GUEST_MEM_ENCRYPT))
 		return;
 
-	if (x86_platform.hyper.is_private_mmio(addr)) {
+	if (suse_x86_hyper_runtime_is_private_mmio(addr)) {
 		desc->flags |= IORES_MAP_ENCRYPTED;
 		return;
 	}
