Subject: xen3 i386 shared kernel PMD
From: http://xenbits.xensource.com/xen-unstable.hg (tip 9137)
Acked-by: jbeulich@novell.com

Index: head-2006-03-06/arch/i386/mm/pageattr.c
===================================================================
--- head-2006-03-06.orig/arch/i386/mm/pageattr.c	2006-03-06 11:09:32.000000000 +0100
+++ head-2006-03-06/arch/i386/mm/pageattr.c	2006-03-06 11:15:25.000000000 +0100
@@ -78,7 +78,11 @@ static void set_pmd_pte(pte_t *kpte, uns
 	unsigned long flags;
 
 	set_pte_atomic(kpte, pte); 	/* change init_mm */
+#ifndef HAVE_SHARED_KERNEL_PMD
 	if (PTRS_PER_PMD > 1)
+#else
+	if (HAVE_SHARED_KERNEL_PMD)
+#endif
 		return;
 
 	spin_lock_irqsave(&pgd_lock, flags);
