Subject: xen3 i386 shared kernel PMD
From: http://xenbits.xensource.com/xen-unstable.hg (tip 13816)
Acked-by: jbeulich@novell.com

Index: head-2007-01-16/arch/i386/mm/pageattr.c
===================================================================
--- head-2007-01-16.orig/arch/i386/mm/pageattr.c	2007-01-16 15:23:23.000000000 +0100
+++ head-2007-01-16/arch/i386/mm/pageattr.c	2007-01-17 09:58:53.000000000 +0100
@@ -90,7 +90,7 @@ static void set_pmd_pte(pte_t *kpte, uns
 	unsigned long flags;
 
 	set_pte_atomic(kpte, pte); 	/* change init_mm */
-	if (PTRS_PER_PMD > 1)
+	if (HAVE_SHARED_KERNEL_PMD)
 		return;
 
 	spin_lock_irqsave(&pgd_lock, flags);
