Subject: xen3 i386 shared kernel PMD
From: http://xenbits.xensource.com/xen-unstable.hg (tip 13816)
Acked-by: jbeulich@novell.com

---
 arch/i386/mm/pageattr.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

--- linux-2.6.20.orig/arch/i386/mm/pageattr.c
+++ linux-2.6.20/arch/i386/mm/pageattr.c
@@ -91,7 +91,7 @@ static void set_pmd_pte(pte_t *kpte, uns
 	unsigned long flags;
 
 	set_pte_atomic(kpte, pte); 	/* change init_mm */
-	if (PTRS_PER_PMD > 1)
+	if (HAVE_SHARED_KERNEL_PMD)
 		return;
 
 	spin_lock_irqsave(&pgd_lock, flags);
