Subject: 
From: http://xenbits.xensource.com/xen-unstable.hg (tip 10901)
Acked-by: jbeulich@novell.com

Index: head-2006-08-03/arch/i386/kernel/entry.S
===================================================================
--- head-2006-08-03.orig/arch/i386/kernel/entry.S	2006-08-03 09:29:15.000000000 +0200
+++ head-2006-08-03/arch/i386/kernel/entry.S	2006-08-03 11:27:13.000000000 +0200
@@ -265,7 +265,7 @@ ENTRY(sysenter_entry)
 	CFI_STARTPROC simple
 	CFI_DEF_CFA esp, 0
 	CFI_REGISTER esp, ebp
-	movl TSS_sysenter_esp0(%esp),%esp
+	movl SYSENTER_stack_esp0(%esp),%esp
 sysenter_past_esp:
 	/*
 	 * No need to follow this irqs on/off section: the syscall
@@ -703,7 +703,7 @@ device_not_available_emulate:
  * that sets up the real kernel stack. Check here, since we can't
  * allow the wrong stack to be used.
  *
- * "TSS_sysenter_esp0+12" is because the NMI/debug handler will have
+ * "SYSENTER_stack_esp0+12" is because the NMI/debug handler will have
  * already pushed 3 words if it hits on the sysenter instruction:
  * eflags, cs and eip.
  *
@@ -715,7 +715,7 @@ device_not_available_emulate:
 	cmpw $__KERNEL_CS,4(%esp);		\
 	jne ok;					\
 label:						\
-	movl TSS_sysenter_esp0+offset(%esp),%esp;	\
+	movl SYSENTER_stack_esp0+offset(%esp),%esp;	\
 	pushfl;					\
 	pushl $__KERNEL_CS;			\
 	pushl $sysenter_past_esp
