Subject: fix smp alternatives for uml.
From: kraxel@suse.de

Index: linux-2.6.12/include/asm-um/smp_alt.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.12/include/asm-um/smp_alt.h	2005-08-12 17:51:12.857994630 +0200
@@ -0,0 +1,6 @@
+#ifndef __UM_SMP_ALT_H
+#define __UM_SMP_ALT_H
+
+#include "asm/arch/smp_alt.h"
+
+#endif
