From: jbeulich@novell.com
Subject: prevent halted VCPUs from eating up CPU bandwidth
References: 17xxxx

Index: head-2006-06-07/include/asm-i386/bugs.h
===================================================================
--- head-2006-06-07.orig/include/asm-i386/bugs.h	2006-06-07 09:26:02.000000000 +0200
+++ head-2006-06-07/include/asm-i386/bugs.h	2006-06-07 09:35:29.000000000 +0200
@@ -92,6 +92,7 @@ static void __init check_fpu(void)
 
 static void __init check_hlt(void)
 {
+#ifndef CONFIG_XEN
 	printk(KERN_INFO "Checking 'hlt' instruction... ");
 	if (!boot_cpu_data.hlt_works_ok) {
 		printk("disabled\n");
@@ -102,6 +103,7 @@ static void __init check_hlt(void)
 	halt();
 	halt();
 	printk("OK.\n");
+#endif
 }
 
 /*
Index: head-2006-06-07/include/asm-i386/mach-xen/asm/hypervisor.h
===================================================================
--- head-2006-06-07.orig/include/asm-i386/mach-xen/asm/hypervisor.h	2006-06-07 09:26:02.000000000 +0200
+++ head-2006-06-07/include/asm-i386/mach-xen/asm/hypervisor.h	2006-06-07 09:35:29.000000000 +0200
@@ -41,6 +41,7 @@
 #include <xen/interface/xen.h>
 #include <xen/interface/dom0_ops.h>
 #include <xen/interface/sched.h>
+#include <xen/interface/vcpu.h>
 #include <xen/interface/nmi.h>
 #include <asm/ptrace.h>
 #include <asm/page.h>
Index: head-2006-06-07/include/asm-i386/mach-xen/asm/system.h
===================================================================
--- head-2006-06-07.orig/include/asm-i386/mach-xen/asm/system.h	2006-06-07 09:31:05.000000000 +0200
+++ head-2006-06-07/include/asm-i386/mach-xen/asm/system.h	2006-06-07 09:35:29.000000000 +0200
@@ -627,8 +627,11 @@ do {									\
 		preempt_enable_no_resched();				\
 } while (0)
 
-#define safe_halt()		((void)0)
-#define halt()			((void)0)
+#define safe_halt()		((void)HYPERVISOR_block())
+#define halt()			((void)					\
+	(HYPERVISOR_shared_info->vcpu_info[__vcpu_id].evtchn_upcall_mask \
+	 ? HYPERVISOR_vcpu_op(VCPUOP_down, __vcpu_id, NULL)		\
+	 : HYPERVISOR_block()))
 
 #define __save_and_cli(x)						\
 do {									\
Index: head-2006-06-07/include/asm-x86_64/mach-xen/asm/system.h
===================================================================
--- head-2006-06-07.orig/include/asm-x86_64/mach-xen/asm/system.h	2006-06-07 09:26:02.000000000 +0200
+++ head-2006-06-07/include/asm-x86_64/mach-xen/asm/system.h	2006-06-07 09:35:29.000000000 +0200
@@ -424,8 +424,11 @@ do {									\
 	preempt_enable_no_resched();					\
 	___x; })
 
-#define safe_halt()		((void)0)
-#define halt()			((void)0)
+#define safe_halt()		((void)HYPERVISOR_block())
+#define halt()			((void)					\
+	(HYPERVISOR_shared_info->vcpu_info[__vcpu_id].evtchn_upcall_mask \
+	 ? HYPERVISOR_vcpu_op(VCPUOP_down, __vcpu_id, NULL)		\
+	 : HYPERVISOR_block()))
 
 void cpu_idle_wait(void);
 
