From: garloff@suse.de
Subject: Fix from 2.6.11-rc3-bk9

Import from mainline.

Index: linux-2.6.10/arch/xen/i386/kernel/pci-dma.c
===================================================================
--- linux-2.6.10.orig/arch/xen/i386/kernel/pci-dma.c
+++ linux-2.6.10/arch/xen/i386/kernel/pci-dma.c
@@ -175,13 +175,13 @@ int dma_declare_coherent_memory(struct d
 	mem_base = ioremap(bus_addr, size);
 	if (!mem_base)
 		goto out;
 
-	dev->dma_mem = kmalloc(GFP_KERNEL, sizeof(struct dma_coherent_mem));
+	dev->dma_mem = kmalloc(sizeof(struct dma_coherent_mem), GFP_KERNEL);
 	if (!dev->dma_mem)
 		goto out;
 	memset(dev->dma_mem, 0, sizeof(struct dma_coherent_mem));
-	dev->dma_mem->bitmap = kmalloc(GFP_KERNEL, bitmap_size);
+	dev->dma_mem->bitmap = kmalloc(bitmap_size, GFP_KERNEL);
 	if (!dev->dma_mem->bitmap)
 		goto free1_out;
 	memset(dev->dma_mem->bitmap, 0, bitmap_size);
 
