From: rmaxfiel@novell.com
Subject: deal with mising media in CD drive exported to guest
References: 175114

Index: head-2006-05-11/drivers/xen/blkback/vbd.c
===================================================================
--- head-2006-05-11.orig/drivers/xen/blkback/vbd.c	2006-05-11 08:43:19.000000000 +0200
+++ head-2006-05-11/drivers/xen/blkback/vbd.c	2006-05-15 10:15:43.000000000 +0200
@@ -66,9 +66,11 @@ int vbd_create(blkif_t *blkif, blkif_vde
 	vbd->bdev = open_by_devnum(
 		vbd->pdevice,
 		vbd->readonly ? FMODE_READ : FMODE_WRITE);
+
 	if (IS_ERR(vbd->bdev)) {
-		DPRINTK("vbd_creat: device %08x doesn't exist.\n",
+		DPRINTK("vbd_creat: device %08x could not be opened.\n",
 			vbd->pdevice);
+		vbd->bdev = NULL;
 		return -ENOENT;
 	}
 
Index: head-2006-05-11/drivers/xen/blkback/xenbus.c
===================================================================
--- head-2006-05-11.orig/drivers/xen/blkback/xenbus.c	2006-05-12 09:24:36.000000000 +0200
+++ head-2006-05-11/drivers/xen/blkback/xenbus.c	2006-05-15 10:15:43.000000000 +0200
@@ -309,6 +309,9 @@ static void connect(struct backend_info 
 
 	DPRINTK("%s", dev->otherend);
 
+	if (be->blkif->vbd.bdev == NULL)
+		return;
+
 	/* Supply the information about the device the frontend needs */
 again:
 	err = xenbus_transaction_start(&xbt);
