From: olh@suse.de
Subject: every arch has to provide a ack_bad_irq

Index: linux-2.6.15/arch/x86_64/xen/kernel/apic.c
===================================================================
--- linux-2.6.15.orig/arch/x86_64/xen/kernel/apic.c
+++ linux-2.6.15/arch/x86_64/xen/kernel/apic.c
@@ -64,6 +64,10 @@ void smp_local_timer_interrupt(struct pt
 	 */
 }
 
+void ack_bad_irq(unsigned int irq)
+{
+	printk("unexpected IRQ trap at vector %02x\n", irq);
+}
 /*
  * Local APIC timer interrupt. This is the most natural way for doing
  * local interrupts, but local timer interrupts can be emulated by
