From: jbeulich@novell.com
Subject: adjust placement of reservation call for kexec crash kernel space

Addressing an apparent a mis-merge.

Index: head-2006-04-21/arch/i386/kernel/setup-xen.c
===================================================================
--- head-2006-04-21.orig/arch/i386/kernel/setup-xen.c	2006-04-21 13:29:39.000000000 +0200
+++ head-2006-04-21/arch/i386/kernel/setup-xen.c	2006-04-21 13:31:17.000000000 +0200
@@ -1353,6 +1353,11 @@ void __init setup_bootmem_allocator(void
 		}
 	}
 #endif
+#ifdef CONFIG_KEXEC
+	if (crashk_res.start != crashk_res.end)
+		reserve_bootmem(crashk_res.start,
+			crashk_res.end - crashk_res.start + 1);
+#endif
 
 	if (!xen_feature(XENFEAT_auto_translated_physmap))
 		phys_to_machine_mapping =
@@ -1475,11 +1480,6 @@ legacy_init_iomem_resources(struct resou
 		}
 	}
 #endif
-#ifdef CONFIG_KEXEC
-	if (crashk_res.start != crashk_res.end)
-		reserve_bootmem(crashk_res.start,
-			crashk_res.end - crashk_res.start + 1);
-#endif
 }
 
 /*
