From: jbeulich@novell.com
Subject: Use halt() instead of raw inline assembly.
References: 177511

Index: head-2006-05-19/arch/x86_64/kernel/smp-xen.c
===================================================================
--- head-2006-05-19.orig/arch/x86_64/kernel/smp-xen.c	2006-05-19 10:26:04.000000000 +0200
+++ head-2006-05-19/arch/x86_64/kernel/smp-xen.c	2006-05-23 14:44:13.000000000 +0200
@@ -488,7 +488,7 @@ static void smp_really_stop_cpu(void *du
 {
 	smp_stop_cpu(); 
 	for (;;) 
-		asm("hlt"); 
+		halt();
 } 
 
 void smp_send_stop(void)
