Subject: Update Xen-specific files for 2.6.16-rc2-git5
From: jbeulich@novell.com

Index: head-2006-02-13/arch/x86_64/kernel/setup-xen.c
===================================================================
--- head-2006-02-13.orig/arch/x86_64/kernel/setup-xen.c	2006-02-13 14:35:47.000000000 +0100
+++ head-2006-02-13/arch/x86_64/kernel/setup-xen.c	2006-02-13 14:37:42.000000000 +0100
@@ -130,7 +130,6 @@ unsigned long saved_video_mode;
 /*
  * Setup options
  */
-struct drive_info_struct { char dummy[32]; } drive_info;
 struct screen_info screen_info;
 struct sys_desc_table_struct {
 	unsigned short length;
@@ -628,7 +627,6 @@ void __init setup_arch(char **cmdline_p)
 
 #ifdef CONFIG_XEN
  	ROOT_DEV = MKDEV(RAMDISK_MAJOR,0); 
- 	drive_info = DRIVE_INFO;
 	kernel_end = 0;		/* dummy */
  	screen_info = SCREEN_INFO;
 
@@ -664,7 +662,6 @@ void __init setup_arch(char **cmdline_p)
 	ARCH_SETUP
 #else
  	ROOT_DEV = old_decode_dev(ORIG_ROOT_DEV);
- 	drive_info = DRIVE_INFO;
  	screen_info = SCREEN_INFO;
 	edid_info = EDID_INFO;
 	saved_video_mode = SAVED_VIDEO_MODE;
Index: head-2006-02-13/arch/x86_64/kernel/x8664_ksyms-xen.c
===================================================================
--- head-2006-02-13.orig/arch/x86_64/kernel/x8664_ksyms-xen.c	2006-02-13 12:17:53.000000000 +0100
+++ head-2006-02-13/arch/x86_64/kernel/x8664_ksyms-xen.c	2006-02-13 14:37:42.000000000 +0100
@@ -39,11 +39,6 @@ extern void __write_lock_failed(rwlock_t
 extern void __read_lock_failed(rwlock_t *rw);
 #endif
 
-#if defined(CONFIG_BLK_DEV_IDE) || defined(CONFIG_BLK_DEV_HD) || defined(CONFIG_BLK_DEV_IDE_MODULE) || defined(CONFIG_BLK_DEV_HD_MODULE)
-extern struct drive_info_struct drive_info;
-EXPORT_SYMBOL(drive_info);
-#endif
-
 /* platform dependent support */
 EXPORT_SYMBOL(boot_cpu_data);
 //EXPORT_SYMBOL(dump_fpu);
