From: jbeulich@novell.com
Subject: Forward information on buggy APICs to the hypervisor.
Reference: 116485

Equivalent to native i386 Linux behavior.

Index: head-2006-04-20/arch/i386/kernel/io_apic-xen.c
===================================================================
--- head-2006-04-20.orig/arch/i386/kernel/io_apic-xen.c	2006-04-20 15:15:19.000000000 +0200
+++ head-2006-04-20/arch/i386/kernel/io_apic-xen.c	2006-04-20 15:15:42.000000000 +0200
@@ -2481,8 +2481,17 @@ __setup("enable_8254_timer", setup_enabl
  
 static int __init io_apic_bug_finalize(void)
 {
-	if(sis_apic_bug == -1)
+	if(sis_apic_bug == -1) {
+		if (xen_start_info->flags & SIF_INITDOMAIN) {
+			dom0_op_t op = {
+				.cmd = DOM0_PLATFORM_QUIRK,
+				.u.platform_quirk.quirk_id = QUIRK_IOAPICMODIFY
+			};
+
+			HYPERVISOR_dom0_op(&op);
+		}
 		sis_apic_bug = 0;
+	}
 	return 0;
 }
 
Index: head-2006-04-20/include/xen/interface/dom0_ops.h
===================================================================
--- head-2006-04-20.orig/include/xen/interface/dom0_ops.h	2006-04-20 15:15:19.000000000 +0200
+++ head-2006-04-20/include/xen/interface/dom0_ops.h	2006-04-20 13:09:17.000000000 +0200
@@ -405,6 +405,7 @@ DEFINE_GUEST_HANDLE(dom0_getdomaininfoli
 
 #define DOM0_PLATFORM_QUIRK      39
 #define QUIRK_NOIRQBALANCING  1
+#define QUIRK_IOAPICMODIFY  2
 typedef struct dom0_platform_quirk {
     /* IN variables. */
     uint32_t quirk_id;
