Subject: Update Xen-specific files for 2.6.16-rc2-git2
From: jbeulich@novell.com

Index: head-2006-03-06/arch/x86_64/kernel/entry-xen.S
===================================================================
--- head-2006-03-06.orig/arch/x86_64/kernel/entry-xen.S	2006-03-06 11:15:34.000000000 +0100
+++ head-2006-03-06/arch/x86_64/kernel/entry-xen.S	2006-03-06 12:04:58.000000000 +0100
@@ -818,7 +818,7 @@ ENTRY(do_hypervisor_callback)   # do_hyp
 	jmp  error_exit
 
 #ifdef CONFIG_X86_LOCAL_APIC
-ENTRY(nmi)
+KPROBE_ENTRY(nmi)
 	zeroentry do_nmi_callback
 ENTRY(do_nmi_callback)
         addq $8, %rsp
@@ -828,6 +828,7 @@ ENTRY(do_nmi_callback)
         XEN_BLOCK_EVENTS(%rsi)
         GET_THREAD_INFO(%rcx)
         jmp  retint_restore_args
+	.previous .text
 #endif
 
         ALIGN
