Subject: Update Xen-specific files for 2.6.16-rc1-git3
From: jbeulich@novell.com

Index: head-2006-01-23/arch/i386/xen/kernel/quirks.c
===================================================================
--- head-2006-01-23.orig/arch/i386/xen/kernel/quirks.c	2006-01-23 15:19:19.742268216 +0100
+++ head-2006-01-23/arch/i386/xen/kernel/quirks.c	2006-01-23 15:20:29.073728232 +0100
@@ -25,8 +25,7 @@ static void __devinit quirk_intel_irqbal
 
 	/* enable access to config space*/
 	pci_read_config_byte(dev, 0xf4, &config);
-	config |= 0x2;
-	pci_write_config_byte(dev, 0xf4, config);
+	pci_write_config_byte(dev, 0xf4, config|0x2);
 
 	/* read xTPR register */
 	raw_pci_ops->read(0, 0, 0x40, 0x4c, 2, &word);
@@ -39,9 +38,9 @@ static void __devinit quirk_intel_irqbal
 		(void)HYPERVISOR_dom0_op(&op);
 	}
 
-	config &= ~0x2;
-	/* disable access to config space*/
-	pci_write_config_byte(dev, 0xf4, config);
+	/* put back the original value for config space*/
+	if (!(config & 0x2))
+		pci_write_config_byte(dev, 0xf4, config);
 }
 DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_INTEL,	PCI_DEVICE_ID_INTEL_E7320_MCH,	quirk_intel_irqbalance);
 DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_INTEL,	PCI_DEVICE_ID_INTEL_E7525_MCH,	quirk_intel_irqbalance);
Index: head-2006-01-23/arch/x86_64/xen/ia32/ia32entry.S
===================================================================
--- head-2006-01-23.orig/arch/x86_64/xen/ia32/ia32entry.S	2006-01-23 15:20:14.539937704 +0100
+++ head-2006-01-23/arch/x86_64/xen/ia32/ia32entry.S	2006-01-23 15:20:29.076727776 +0100
@@ -699,6 +699,19 @@ ia32_sys_call_table:
 	.quad sys_inotify_add_watch
 	.quad sys_inotify_rm_watch
 	.quad sys_migrate_pages
+	.quad compat_sys_openat		/* 295 */
+	.quad sys_mkdirat
+	.quad sys_mknodat
+	.quad sys_fchownat
+	.quad sys_futimesat
+	.quad compat_sys_newfstatat	/* 300 */
+	.quad sys_unlinkat
+	.quad sys_renameat
+	.quad sys_linkat
+	.quad sys_symlinkat
+	.quad sys_readlinkat		/* 305 */
+	.quad sys_fchmodat
+	.quad sys_faccessat
 ia32_syscall_end:		
 	.rept IA32_NR_syscalls-(ia32_syscall_end-ia32_sys_call_table)/8
 		.quad ni_syscall
Index: head-2006-01-23/arch/x86_64/xen/kernel/setup64.c
===================================================================
--- head-2006-01-23.orig/arch/x86_64/xen/kernel/setup64.c	2006-01-23 15:20:26.351142128 +0100
+++ head-2006-01-23/arch/x86_64/xen/kernel/setup64.c	2006-01-23 15:20:29.079727320 +0100
@@ -203,7 +203,7 @@ void pda_init(int cpu)
 	pda->irqstackptr += IRQSTACKSIZE-64;
 } 
 
-char boot_exception_stacks[(N_EXCEPTION_STACKS - 2) * EXCEPTION_STKSZ + DEBUG_STKSZ]
+char boot_exception_stacks[(N_EXCEPTION_STACKS - 1) * EXCEPTION_STKSZ + DEBUG_STKSZ]
 __attribute__((section(".bss.page_aligned")));
 
 /* May not be marked __init: used by software suspend */
