From: jbeulich@novell.com
Subject: eliminate static variables from assign_irq_vector()

Avoiding potential serialization issues upon their use.

Index: head-2006-04-24/arch/i386/kernel/io_apic-xen.c
===================================================================
--- head-2006-04-24.orig/arch/i386/kernel/io_apic-xen.c	2006-04-24 15:19:43.000000000 +0200
+++ head-2006-04-24/arch/i386/kernel/io_apic-xen.c	2006-04-25 14:40:50.000000000 +0200
@@ -1205,10 +1205,9 @@ u8 irq_vector[NR_IRQ_VECTORS] __read_mos
 
 int assign_irq_vector(int irq)
 {
-	static int current_vector = FIRST_DEVICE_VECTOR;
 	physdev_op_t op;
 
-	BUG_ON(irq >= NR_IRQ_VECTORS);
+	BUG_ON(irq != AUTO_ASSIGN && (unsigned)irq >= NR_IRQ_VECTORS);
 	if (irq != AUTO_ASSIGN && IO_APIC_VECTOR(irq) > 0)
 		return IO_APIC_VECTOR(irq);
 
@@ -1216,13 +1215,15 @@ int assign_irq_vector(int irq)
 	op.u.irq_op.irq = irq;
 	if (HYPERVISOR_physdev_op(&op))
 		return -ENOSPC;
-	current_vector = op.u.irq_op.vector;
 
-	vector_irq[current_vector] = irq;
-	if (irq != AUTO_ASSIGN)
-		IO_APIC_VECTOR(irq) = current_vector;
+	vector_irq[op.u.irq_op.vector] = irq;
+	if (irq != AUTO_ASSIGN) {
+		u8 vector = cmpxchg(&IO_APIC_VECTOR(irq), 0, op.u.irq_op.vector);
 
-	return current_vector;
+		BUG_ON(vector && vector != op.u.irq_op.vector);
+	}
+
+	return op.u.irq_op.vector;
 }
 
 #ifndef CONFIG_XEN
Index: head-2006-04-24/arch/x86_64/kernel/io_apic-xen.c
===================================================================
--- head-2006-04-24.orig/arch/x86_64/kernel/io_apic-xen.c	2006-04-24 15:19:43.000000000 +0200
+++ head-2006-04-24/arch/x86_64/kernel/io_apic-xen.c	2006-04-25 14:40:50.000000000 +0200
@@ -869,7 +869,6 @@ u8 irq_vector[NR_IRQ_VECTORS] __read_mos
 
 int assign_irq_vector(int irq)
 {
-	static int current_vector = FIRST_DEVICE_VECTOR;
 	physdev_op_t op;
   
   	BUG_ON(irq != AUTO_ASSIGN && (unsigned)irq >= NR_IRQ_VECTORS);
@@ -880,17 +879,19 @@ int assign_irq_vector(int irq)
 	op.u.irq_op.irq = irq;
 	if (HYPERVISOR_physdev_op(&op))
 		return -ENOSPC;
-	current_vector = op.u.irq_op.vector;
 
-	vector_irq[current_vector] = irq;
-	if (irq != AUTO_ASSIGN)
-		IO_APIC_VECTOR(irq) = current_vector;
+	vector_irq[op.u.irq_op.vector] = irq;
+	if (irq != AUTO_ASSIGN) {
+		u8 vector = cmpxchg(&IO_APIC_VECTOR(irq), 0, op.u.irq_op.vector);
 
-	return current_vector;
+		BUG_ON(vector && vector != op.u.irq_op.vector);
+	}
+
+	return op.u.irq_op.vector;
 }
 
-extern void (*interrupt[NR_IRQS])(void);
 #ifndef CONFIG_XEN
+extern void (*interrupt[NR_IRQS])(void);
 static struct hw_interrupt_type ioapic_level_type;
 static struct hw_interrupt_type ioapic_edge_type;
 
