Subject: Update Xen-specific files for 2.6.16-rc4
From: jbeulich@novell.com

Index: head-2006-02-20/arch/x86_64/kernel/head-xen.S
===================================================================
--- head-2006-02-20.orig/arch/x86_64/kernel/head-xen.S	2006-02-20 09:08:44.000000000 +0100
+++ head-2006-02-20/arch/x86_64/kernel/head-xen.S	2006-02-20 09:50:39.000000000 +0100
@@ -75,6 +75,8 @@ ENTRY(early_idt_handler)
 
 early_idt_msg:
 	.asciz "PANIC: early exception rip %lx error %lx cr2 %lx\n"
+early_idt_ripmsg:
+	.asciz "RIP %s\n"
 
 #if 0
 ENTRY(lgdt_finish)
Index: head-2006-02-20/arch/x86_64/kernel/io_apic-xen.c
===================================================================
--- head-2006-02-20.orig/arch/x86_64/kernel/io_apic-xen.c	2006-02-20 09:50:32.000000000 +0100
+++ head-2006-02-20/arch/x86_64/kernel/io_apic-xen.c	2006-02-20 09:52:54.000000000 +0100
@@ -30,6 +30,9 @@
 #include <linux/mc146818rtc.h>
 #include <linux/acpi.h>
 #include <linux/sysdev.h>
+#ifdef CONFIG_ACPI
+#include <acpi/acpi_bus.h>
+#endif
 
 #include <asm/io.h>
 #include <asm/smp.h>
@@ -310,6 +313,8 @@ __setup("apic", enable_ioapic_setup);
 
    And another hack to disable the IOMMU on VIA chipsets.
 
+   ... and others. Really should move this somewhere else.
+
    Kludge-O-Rama. */
 void __init check_ioapic(void) 
 { 
@@ -358,6 +363,17 @@ void __init check_ioapic(void) 
 #ifndef CONFIG_XEN
 					if (apic_runs_main_timer != 0)
 						break;
+#ifdef CONFIG_ACPI
+					/* Don't do this for laptops right
+					   right now because their timer
+					   doesn't necessarily tick in C2/3 */
+					if (acpi_fadt.revision >= 3 &&
+			(acpi_fadt.plvl2_lat + acpi_fadt.plvl3_lat) < 1100) {
+						printk(KERN_INFO
+"ATI board detected, but seems to be a laptop. Timer might be shakey, sorry\n");
+						break;
+					}
+#endif
 					printk(KERN_INFO
 	     "ATI board detected. Using APIC/PM timer.\n");
 					apic_runs_main_timer = 1;
Index: head-2006-02-20/arch/x86_64/kernel/mpparse-xen.c
===================================================================
--- head-2006-02-20.orig/arch/x86_64/kernel/mpparse-xen.c	2006-02-20 09:08:44.000000000 +0100
+++ head-2006-02-20/arch/x86_64/kernel/mpparse-xen.c	2006-02-20 09:50:39.000000000 +0100
@@ -295,9 +295,9 @@ static int __init smp_read_mpc(struct mp
 
 	memcpy(str,mpc->mpc_productid,12);
 	str[12]=0;
-	printk(KERN_INFO "Product ID: %s ",str);
+	printk("Product ID: %s ",str);
 
-	printk(KERN_INFO "APIC at: 0x%X\n",mpc->mpc_lapic);
+	printk("APIC at: 0x%X\n",mpc->mpc_lapic);
 
 	/* save the local APIC address, it might be non-default */
 	if (!acpi_lapic)
