Subject: Update Xen-specific files for 2.6.16-rc5-git9
From: jbeulich@novell.com

Index: head-2006-03-10/arch/i386/kernel/acpi/boot-xen.c
===================================================================
--- head-2006-03-10.orig/arch/i386/kernel/acpi/boot-xen.c	2006-03-10 11:08:16.000000000 +0100
+++ head-2006-03-10/arch/i386/kernel/acpi/boot-xen.c	2006-03-10 11:08:45.000000000 +0100
@@ -44,6 +44,9 @@ extern void __init clustered_apic_check(
 extern int gsi_irq_sharing(int gsi);
 #include <asm/proto.h>
 
+static inline int acpi_madt_oem_check(char *oem_id, char *oem_table_id) { return 0; }
+
+
 #else				/* X86 */
 
 #ifdef	CONFIG_X86_LOCAL_APIC
Index: head-2006-03-10/arch/i386/kernel/time-xen.c
===================================================================
--- head-2006-03-10.orig/arch/i386/kernel/time-xen.c	2006-03-10 11:08:16.000000000 +0100
+++ head-2006-03-10/arch/i386/kernel/time-xen.c	2006-03-10 11:08:45.000000000 +0100
@@ -829,9 +829,9 @@ static int timer_resume(struct sys_devic
 	write_seqlock_irqsave(&xtime_lock, flags);
 	xtime.tv_sec = sec;
 	xtime.tv_nsec = 0;
-	write_sequnlock_irqrestore(&xtime_lock, flags);
-	jiffies += sleep_length;
+	jiffies_64 += sleep_length;
 	wall_jiffies += sleep_length;
+	write_sequnlock_irqrestore(&xtime_lock, flags);
 	if (last_timer->resume)
 		last_timer->resume();
 	cur_timer = last_timer;
