From: jbeulich@novell.com
Subject: add appropriate module license for xenbus_be
References: 173143
Patch-mainline: unstable c/s 9957, 3.0-testing c/s 9668

Index: head-2006-05-03/drivers/xen/xenbus/xenbus_backend_client.c
===================================================================
--- head-2006-05-03.orig/drivers/xen/xenbus/xenbus_backend_client.c	2006-04-27 17:01:16.000000000 +0200
+++ head-2006-05-03/drivers/xen/xenbus/xenbus_backend_client.c	2006-05-08 09:14:03.000000000 +0200
@@ -139,6 +139,7 @@ int xenbus_unmap_ring(struct xenbus_devi
 }
 EXPORT_SYMBOL_GPL(xenbus_unmap_ring);
 
+MODULE_LICENSE("Dual BSD/GPL");
 
 /*
  * Local variables:
