From: Ross Maxfield
Subject: This patch removes the erratic mouse behavior of SLES10, SLES9 hvm guests.
Patch-mainline: never.  A patch in the mm tree will eventually remove polling of the 8042. (ftp://ftp.kernel.org/pub/linux/kernel/people/akpm/patches/2.6/2.6.18-rc5/2.6.18-rc5-mm1/broken-out/input-i8042-get-rid-of-polling-timer.patch)
References: 167187

Acked-by: 
Signed-off-by: Ross Maxfield <rmaxfiel@novell.com>

Index: linux-2.6.18/drivers/input/serio/i8042.c
===================================================================
--- linux-2.6.18.orig/drivers/input/serio/i8042.c
+++ linux-2.6.18/drivers/input/serio/i8042.c
@@ -428,13 +428,11 @@ static irqreturn_t i8042_interrupt(int i
 	spin_lock_irqsave(&i8042_lock, flags);
 	str = i8042_read_status();
 	if (unlikely(~str & I8042_STR_OBF)) {
-		spin_unlock_irqrestore(&i8042_lock, flags);
 		if (irq) dbg("Interrupt %d, without any data", irq);
 		ret = 0;
 		goto out;
 	}
 	data = i8042_read_data();
-	spin_unlock_irqrestore(&i8042_lock, flags);
 
 	if (i8042_mux_present && (str & I8042_STR_AUXDATA)) {
 		static unsigned long last_transmit;
@@ -490,6 +488,7 @@ static irqreturn_t i8042_interrupt(int i
 
 	ret = 1;
  out:
+	spin_unlock_irqrestore(&i8042_lock, flags);
 	return IRQ_RETVAL(ret);
 }
 
