From: Andrea Arcangeli <andrea@suse.de>
Subject: avoid silent stack overflow over the heap
Patch-mainline: no
References: 44807

x

Signed-off-by: Andrea Arcangeli <andrea@suse.de>

diff -uNrp linux-2.6.10.vanilla/arch/xen/i386/mm/fault.c linux-2.6.10.xen203/arch/xen/i386/mm/fault.c
--- linux-2.6.10.vanilla/arch/xen/i386/mm/fault.c	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.10.xen203/arch/xen/i386/mm/fault.c	2005-01-20 05:08:34.000000000 +0100
@@ -42,7 +42,7 @@
 		if (address + 32 < regs->esp)
 			goto bad_area;
 	}
-	if (expand_stack(vma, address))
+	if (expand_stack(vma, address, NULL))
 		goto bad_area;
 /*
  * Ok, we have a good vm_area for this memory access, so
