From: Andrea Arcangeli <andrea@suse.de>
Subject: avoid silent stack overflow over the heap
Patch-mainline: no
References: SUSE44807

x

Signed-off-by: Andrea Arcangeli <andrea@suse.de>

Index: linux-2.6.10/arch/xen/i386/mm/fault.c
===================================================================
--- linux-2.6.10.orig/arch/xen/i386/mm/fault.c	2005-02-22 12:44:19.000000000 +0100
+++ linux-2.6.10/arch/xen/i386/mm/fault.c	2005-02-22 12:48:09.000000000 +0100
@@ -315,7 +315,7 @@
 		if (address + 32 < regs->esp)
 			goto bad_area;
 	}
-	if (expand_stack(vma, address))
+	if (expand_stack(vma, address, NULL))
 		goto bad_area;
 /*
  * Ok, we have a good vm_area for this memory access, so

