From: garloff@suse.de
Subject: bootloader-type misses on arch xen

.

Index: linux-2.6.10/kernel/sysctl.c
===================================================================
--- linux-2.6.10.orig/kernel/sysctl.c	2005-02-22 12:41:22.000000000 +0100
+++ linux-2.6.10/kernel/sysctl.c	2005-02-22 12:55:57.000000000 +0100
@@ -653,7 +653,7 @@
 		.proc_handler   = &proc_unknown_nmi_panic,
 	},
 #endif
-#if defined(CONFIG_X86)
+#if defined(CONFIG_X86) && !defined(CONFIG_XEN)
 	{
 		.ctl_name	= KERN_BOOTLOADER_TYPE,
 		.procname	= "bootloader_type",
