Subject: xen3 arch-x86_64
From: http://xenbits.xensource.com/xen-3.0.3-testing.hg (tip 11772)
Patch-mainline: obsolete
Acked-by: jbeulich@novell.com

Index: head-2006-10-16/arch/x86_64/ia32/Makefile
===================================================================
--- head-2006-10-16.orig/arch/x86_64/ia32/Makefile	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/ia32/Makefile	2006-10-16 10:09:24.000000000 +0200
@@ -27,9 +27,25 @@ quiet_cmd_syscall = SYSCALL $@
 			   -Wl,-soname=linux-gate.so.1 -o $@ \
 			   -Wl,-T,$(filter-out FORCE,$^)
 
+$(obj)/vsyscall-int80.so \
 $(obj)/vsyscall-sysenter.so $(obj)/vsyscall-syscall.so: \
 $(obj)/vsyscall-%.so: $(src)/vsyscall.lds $(obj)/vsyscall-%.o FORCE
 	$(call if_changed,syscall)
 
-AFLAGS_vsyscall-sysenter.o = -m32 -Wa,-32
-AFLAGS_vsyscall-syscall.o = -m32 -Wa,-32
+AFLAGS_vsyscall-sysenter.o = -m32 -Wa,-32 -Iarch/i386/kernel
+AFLAGS_vsyscall-syscall.o = -m32 -Wa,-32 -Iarch/i386/kernel
+
+ifdef CONFIG_XEN
+AFLAGS_vsyscall-int80.o = -m32 -Wa,-32 -Iarch/i386/kernel
+CFLAGS_syscall32-xen.o += -DUSE_INT80
+AFLAGS_syscall32_syscall-xen.o += -DUSE_INT80
+
+$(obj)/syscall32_syscall-xen.o: \
+	$(foreach F,int80 sysenter syscall,$(obj)/vsyscall-$F.so)
+
+targets := $(foreach F,int80 sysenter syscall,vsyscall-$F.o vsyscall-$F.so)
+
+include $(srctree)/scripts/Makefile.xen
+
+obj-y := $(call cherrypickxen, $(obj-y))
+endif
Index: head-2006-10-16/arch/x86_64/ia32/vsyscall-sigreturn.S
===================================================================
--- head-2006-10-16.orig/arch/x86_64/ia32/vsyscall-sigreturn.S	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/ia32/vsyscall-sigreturn.S	2006-10-16 10:09:24.000000000 +0200
@@ -139,5 +139,5 @@ __kernel_rt_sigreturn:
 	.align 4
 .LENDFDE3:
 
-#include "../../i386/kernel/vsyscall-note.S"
+#include <vsyscall-note.S>
 
Index: head-2006-10-16/arch/x86_64/kernel/acpi/Makefile
===================================================================
--- head-2006-10-16.orig/arch/x86_64/kernel/acpi/Makefile	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/kernel/acpi/Makefile	2006-10-16 10:09:24.000000000 +0200
@@ -7,3 +7,4 @@ obj-y			+= processor.o
 processor-y		:= ../../../i386/kernel/acpi/processor.o ../../../i386/kernel/acpi/cstate.o
 endif
 
+boot-$(CONFIG_XEN)		:= ../../../i386/kernel/acpi/boot-xen.o
Index: head-2006-10-16/arch/x86_64/kernel/asm-offsets.c
===================================================================
--- head-2006-10-16.orig/arch/x86_64/kernel/asm-offsets.c	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/kernel/asm-offsets.c	2006-10-16 10:09:24.000000000 +0200
@@ -67,8 +67,10 @@ int main(void)
 	DEFINE(pbe_address, offsetof(struct pbe, address));
 	DEFINE(pbe_orig_address, offsetof(struct pbe, orig_address));
 	DEFINE(pbe_next, offsetof(struct pbe, next));
+#ifndef CONFIG_X86_NO_TSS
 	BLANK();
 	DEFINE(TSS_ist, offsetof(struct tss_struct, ist));
+#endif
 	BLANK();
 	DEFINE(crypto_tfm_ctx_offset, offsetof(struct crypto_tfm, __crt_ctx));
 	return 0;
Index: head-2006-10-16/arch/x86_64/kernel/init_task.c
===================================================================
--- head-2006-10-16.orig/arch/x86_64/kernel/init_task.c	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/kernel/init_task.c	2006-10-16 10:09:24.000000000 +0200
@@ -37,6 +37,8 @@ union thread_union init_thread_union 
 struct task_struct init_task = INIT_TASK(init_task);
 
 EXPORT_SYMBOL(init_task);
+
+#ifndef CONFIG_X86_NO_TSS
 /*
  * per-CPU TSS segments. Threads are completely 'soft' on Linux,
  * no more per-task TSS's. The TSS size is kept cacheline-aligned
@@ -45,6 +47,7 @@ EXPORT_SYMBOL(init_task);
  * on exact cacheline boundaries, to eliminate cacheline ping-pong.
  */ 
 DEFINE_PER_CPU(struct tss_struct, init_tss) ____cacheline_internodealigned_in_smp = INIT_TSS;
+#endif
 
 /* Copies of the original ist values from the tss are only accessed during
  * debugging, no special alignment required.
Index: head-2006-10-16/arch/x86_64/kernel/Makefile
===================================================================
--- head-2006-10-16.orig/arch/x86_64/kernel/Makefile	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/kernel/Makefile	2006-10-16 10:09:24.000000000 +0200
@@ -21,11 +21,13 @@ obj-$(CONFIG_MICROCODE)		+= microcode.o
 obj-$(CONFIG_X86_CPUID)		+= cpuid.o
 obj-$(CONFIG_SMP)		+= smp.o smpboot.o trampoline.o
 obj-$(CONFIG_X86_LOCAL_APIC)	+= apic.o  nmi.o
+obj-$(CONFIG_X86_XEN_GENAPIC)	+= genapic.o genapic_xen.o
 obj-$(CONFIG_X86_IO_APIC)	+= io_apic.o mpparse.o \
 		genapic.o genapic_cluster.o genapic_flat.o
 obj-$(CONFIG_KEXEC)		+= machine_kexec.o relocate_kernel.o crash.o
 obj-$(CONFIG_CRASH_DUMP)	+= crash_dump.o
-obj-$(CONFIG_PM)		+= suspend.o
+obj-$(CONFIG_SOFTWARE_SUSPEND)	+= suspend.o
+obj-$(CONFIG_ACPI_SLEEP)	+= suspend.o
 obj-$(CONFIG_SOFTWARE_SUSPEND)	+= suspend_asm.o
 obj-$(CONFIG_CPU_FREQ)		+= cpufreq/
 obj-$(CONFIG_EARLY_PRINTK)	+= early_printk.o
@@ -55,3 +57,17 @@ i8237-y				+= ../../i386/kernel/i8237.o
 msr-$(subst m,y,$(CONFIG_X86_MSR))  += ../../i386/kernel/msr.o
 alternative-y			+= ../../i386/kernel/alternative.o
 
+ifdef CONFIG_XEN
+time-y				+= ../../i386/kernel/time-xen.o
+pci-dma-y			+= ../../i386/kernel/pci-dma-xen.o
+microcode-$(subst m,y,$(CONFIG_MICROCODE))  := ../../i386/kernel/microcode-xen.o
+quirks-y			:= ../../i386/kernel/quirks-xen.o
+
+n-obj-xen := i8259.o reboot.o i8237.o smpboot.o trampoline.o
+
+include $(srctree)/scripts/Makefile.xen
+
+obj-y := $(call filterxen, $(obj-y), $(n-obj-xen))
+obj-y := $(call cherrypickxen, $(obj-y))
+extra-y := $(call cherrypickxen, $(extra-y))
+endif
Index: head-2006-10-16/arch/x86_64/Makefile
===================================================================
--- head-2006-10-16.orig/arch/x86_64/Makefile	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/Makefile	2006-10-16 10:09:24.000000000 +0200
@@ -56,8 +56,12 @@ cflags-y += $(call cc-option,-funit-at-a
 # prevent gcc from generating any FP code by mistake
 cflags-y += $(call cc-option,-mno-sse -mno-mmx -mno-sse2 -mno-3dnow,)
 
+cppflags-$(CONFIG_XEN) += \
+	-D__XEN_INTERFACE_VERSION__=$(CONFIG_XEN_INTERFACE_VERSION)
+
 CFLAGS += $(cflags-y)
 CFLAGS_KERNEL += $(cflags-kernel-y)
+CPPFLAGS += $(cppflags-y)
 AFLAGS += -m64
 
 head-y := arch/x86_64/kernel/head.o arch/x86_64/kernel/head64.o arch/x86_64/kernel/init_task.o
@@ -76,6 +80,21 @@ boot := arch/x86_64/boot
 PHONY += bzImage bzlilo install archmrproper \
 	 fdimage fdimage144 fdimage288 isoimage archclean
 
+ifdef CONFIG_XEN
+CPPFLAGS := -Iinclude$(if $(KBUILD_SRC),2)/asm/mach-xen $(CPPFLAGS)
+head-y := arch/x86_64/kernel/head-xen.o arch/x86_64/kernel/head64-xen.o arch/x86_64/kernel/init_task.o
+LDFLAGS_vmlinux := -e _start
+boot := arch/i386/boot-xen
+.PHONY: vmlinuz
+#Default target when executing "make"
+all: vmlinuz
+
+vmlinuz: vmlinux
+	$(Q)$(MAKE) $(build)=$(boot) $@
+
+install:
+	$(Q)$(MAKE) $(build)=$(boot) XENGUEST=$(XENGUEST) $@
+else
 #Default target when executing "make"
 all: bzImage
 
@@ -96,6 +115,7 @@ fdimage fdimage144 fdimage288 isoimage: 
 
 install:
 	$(Q)$(MAKE) $(build)=$(boot) BOOTIMAGE=$(BOOTIMAGE) $@ 
+endif
 
 archclean:
 	$(Q)$(MAKE) $(clean)=$(boot)
Index: head-2006-10-16/arch/x86_64/mm/Makefile
===================================================================
--- head-2006-10-16.orig/arch/x86_64/mm/Makefile	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/mm/Makefile	2006-10-16 10:09:24.000000000 +0200
@@ -9,3 +9,13 @@ obj-$(CONFIG_K8_NUMA) += k8topology.o
 obj-$(CONFIG_ACPI_NUMA) += srat.o
 
 hugetlbpage-y = ../../i386/mm/hugetlbpage.o
+
+ifdef CONFIG_XEN
+include $(srctree)/scripts/Makefile.xen
+
+ioremap-y	+= ../../i386/mm/ioremap-xen.o
+hypervisor-y	+= ../../i386/mm/hypervisor.o
+obj-y		+= hypervisor.o
+
+obj-y := $(call cherrypickxen, $(obj-y))
+endif
Index: head-2006-10-16/arch/x86_64/oprofile/Makefile
===================================================================
--- head-2006-10-16.orig/arch/x86_64/oprofile/Makefile	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/oprofile/Makefile	2006-10-16 10:09:24.000000000 +0200
@@ -11,9 +11,12 @@ DRIVER_OBJS = $(addprefix ../../../drive
 	oprofilefs.o oprofile_stats.o \
 	timer_int.o )
 
+ifdef CONFIG_XEN
+OPROFILE-y := xenoprof.o
+else
 OPROFILE-y := init.o backtrace.o
 OPROFILE-$(CONFIG_X86_LOCAL_APIC) += nmi_int.o op_model_athlon.o op_model_p4.o \
 				     op_model_ppro.o
 OPROFILE-$(CONFIG_X86_IO_APIC)    += nmi_timer_int.o 
-
+endif
 oprofile-y = $(DRIVER_OBJS) $(addprefix ../../i386/oprofile/, $(OPROFILE-y))
Index: head-2006-10-16/arch/x86_64/pci/Makefile
===================================================================
--- head-2006-10-16.orig/arch/x86_64/pci/Makefile	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/arch/x86_64/pci/Makefile	2006-10-16 10:09:24.000000000 +0200
@@ -15,11 +15,23 @@ obj-$(CONFIG_PCI_MMCONFIG) += mmconfig.o
 
 obj-$(CONFIG_NUMA)	+= k8-bus.o
 
+# pcifront should be after mmconfig.o and direct.o as it should only
+# take over if direct access to the PCI bus is unavailable
+obj-$(CONFIG_XEN_PCIDEV_FRONTEND)	+= pcifront.o
+
 direct-y += ../../i386/pci/direct.o
 acpi-y   += ../../i386/pci/acpi.o
+pcifront-y += ../../i386/pci/pcifront.o
 legacy-y += ../../i386/pci/legacy.o
 irq-y    += ../../i386/pci/irq.o
 common-y += ../../i386/pci/common.o
 fixup-y  += ../../i386/pci/fixup.o
 i386-y  += ../../i386/pci/i386.o
 init-y += ../../i386/pci/init.o
+
+ifdef CONFIG_XEN
+irq-y := ../../i386/pci/irq-xen.o
+include $(srctree)/scripts/Makefile.xen
+
+obj-y := $(call cherrypickxen, $(obj-y))
+endif
Index: head-2006-10-16/include/asm-x86_64/apic.h
===================================================================
--- head-2006-10-16.orig/include/asm-x86_64/apic.h	2006-10-16 10:08:00.000000000 +0200
+++ head-2006-10-16/include/asm-x86_64/apic.h	2006-10-16 10:09:24.000000000 +0200
@@ -98,11 +98,13 @@ extern void setup_APIC_extened_lvt(unsig
 extern int disable_timer_pin_1;
 
 
+#ifndef CONFIG_XEN
 void smp_send_timer_broadcast_ipi(void);
 void switch_APIC_timer_to_ipi(void *cpumask);
 void switch_ipi_to_APIC_timer(void *cpumask);
 
 #define ARCH_APICTIMER_STOPS_ON_C3	1
+#endif
 
 #endif /* CONFIG_X86_LOCAL_APIC */
 
